{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Darwin Core Conversion of eDNA Sequence Data From the FAIRe (NOAA Version) metadata template \n",
    "\n",
    "**Version:** 3.0\n",
    "\n",
    "**Authors:** Katherine Silliman, Bayden Willms\n",
    "\n",
    "**Last Updated:** 2-June-2025\n",
    "\n",
    "This notebook is for converting a [FAIR-eDNA](https://fair-edna.github.io/index.html)-based data sheet to DarwinCore for submission to OBIS. It has been testing on a Windows 11 laptop, with Python 3.11. \n",
    "\n",
    "To generate the input files for edna2obis, please run [FAIRe2NODE](https://github.com/aomlomics/FAIReSheets/tree/FAIRe2NODE) to generate your own FAIR-eDNA (NOAA) template. Once you've filled in your data, you are ready to begin.\n",
    "\n",
    "This newest version of edna2obis takes the same input files as the ODE, the [Ocean DNA Explorer](https://www.oceandnaexplorer.org/). Explore your data (publically or privately) with visualizations, API capabilities, and more through ODE. \n",
    "\n",
    "[FAIR-eDNA NOAA Google Sheet](https://docs.google.com/spreadsheets/d/1mkjfUQW3gTn3ezhMQmFDQn4EBoQ2Xv4SZeSd9sqagoU/edit?gid=0#gid=0)\n",
    "\n",
    "**Requirements:**\n",
    "- Python 3\n",
    "- Python 3 packages:\n",
    "    - os\n",
    "- External packages:\n",
    "    - Bio.Entrez from biopython\n",
    "    - numpy\n",
    "    - pandas\n",
    "    - openpyxl\n",
    "    - pyworms\n",
    "    - multiprocess\n",
    "- Custom modules:\n",
    "    - WoRMS_matching\n",
    "    - analysis_helpers\n",
    "\n",
    "**Resources:**\n",
    "- Abarenkov K, Andersson AF, Bissett A, Finstad AG, Fossøy F, Grosjean M, Hope M, Jeppesen TS, Kõljalg U, Lundin D, Nilsson RN, Prager M, Provoost P, Schigel D, Suominen S, Svenningsen C & Frøslev TG (2023) Publishing DNA-derived data through biodiversity data platforms, v1.3. Copenhagen: GBIF Secretariat. https://doi.org/10.35035/doc-vf1a-nr22.https://doi.org/10.35035/doc-vf1a-nr22.\n",
    "- [OBIS manual](https://manual.obis.org/dna_data.html)\n",
    "- [TDWG Darwin Core Occurrence Core](https://dwc.tdwg.org/terms/#occurrence)\n",
    "- [GBIF DNA Derived Data Extension](https://tools.gbif.org/dwca-validator/extension.do?id=http://rs.gbif.org/terms/1.0/DNADerivedData)\n",
    "- https://github.com/iobis/dataset-edna\n",
    "\n",
    "**Citation**  \n",
    "Silliman K, Anderson S, Storo R, Thompson L (2023) A Case Study in Sharing Marine eDNA Metabarcoding Data to OBIS. Biodiversity Information Science and Standards 7: e111048. https://doi.org/10.3897/biss.7.111048\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation  \n",
    "\n",
    "```bash\n",
    "conda create -n edna2obis\n",
    "conda activate edna2obis\n",
    "conda install -c conda-forge notebook\n",
    "conda install -c conda-forge nb_conda_kernels\n",
    "\n",
    "conda install -c conda-forge numpy pandas\n",
    "conda install -c conda-forge openpyxl\n",
    "\n",
    "#worms conversion\n",
    "conda install -c conda-forge pyworms\n",
    "conda install -c conda-forge multiprocess\n",
    "conda install -c conda-forge biopython\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import WoRMS_matching # custom functions for querying WoRMS API\n",
    "import WoRMS_v3_matching # new custom functions for querying WoRMS API\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# jupyter notebook parameters\n",
    "pd.set_option('display.max_colwidth', 150)\n",
    "pd.set_option('display.max_columns', 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that in a Jupyter Notebook, the current directory is always where the .ipynb file is being run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Prepare Input Data \n",
    "\n",
    "**Project data and metadata**  \n",
    "This workflow assumes that you have your project metadata in an Excel sheet formatted like the FAIR-eDNA template located **TODO NEED TO ADD CORRECT LINK**[here](https://docs.google.com/spreadsheets/d/1YBXFU9PuMqm7IT1tp0LTxQ1v2j0tlCWFnhSpy-EBwPw/edit?usp=drive_link). Instructions for filling out the metadata template are located in the 'README' sheet and at the [documentation website](https://noaa-omics-templates.readthedocs.io/en/latest/).\n",
    "\n",
    "**eDNA and taxonomy data**  \n",
    "The eDNA data and assigned taxonomy should be in a specific tab-delimited format. ![asv_table format](../images/asv_table.png)\n",
    "\n",
    "This file is generated automatically by [Tourmaline 2](https://github.com/aomlomics/tourmaline/tree/develop), in X location. If your data was generated with Qiime2 or a previous version of Tourmaline, you can convert the `table.qza`, `taxonomy.qza`, and `repseqs.qza` outputs to the correct format using the `create_asv_seq_taxa_obis.sh` shell script.\n",
    "\n",
    "Example:  \n",
    "\n",
    "``` bash\n",
    "#Run this with a qiime2 environment. \n",
    "bash create_asv_seq_taxa_obis.sh -f \\\n",
    "../gomecc_v2_raw/table-16S-merge.qza -t ../gomecc_v2_raw/taxonomy-16S-merge.qza -r ../gomecc_v2_raw/repseqs-16S-merge.qza \\\n",
    "-o ../gomecc_v2_raw/gomecc-16S-asv.tsv\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Set configs  \n",
    "\n",
    "Below you can set definitions for parameters used in the code. \n",
    "\n",
    "| Parameter           | Description                                                                                                       | Example                                                                                              |\n",
    "|---------------------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|\n",
    "| `sampleMetadata`    | Name of sheet in FAIRe template data Excel file with sample metadata.                                                    | \"sampleMetadata\"                                                                                  |\n",
    "| `experimentRunMetadata`         | Name of sheet in FAIRe template data Excel file with data about molecular preparation methods.                           | \"experimentRunMetadata\"                                                                                 |\n",
    "| `projectMetadata`        | Name of sheet in FAIRe template data Excel file with metadata about the study.                                           | \"projectMetadata\"                                                                                         |\n",
    "| `excel_file`        | Path of the FAIRe data Excel file.                                                                                  | \"../raw-v3/FAIRe_noaa-aoml-gomecc4.xlsx\"                                                  |\n",
    "| `FAIRe_NOAA_checklist`        | Path of the FAIRe NOAA Checklist, which contains information on mapping to DarwinCore and the expected files for OBIS submission.                                                                                  | \"../raw-v3/FAIRe_NOAA_checklist_v1.0.xlsx\"                                                  |\n",
    "| `datafiles`         | Python dictionary, where keys are the amplicon names and the values are the paths to the cooresponding ASV table. |   See example below to format raw data per analysis  |\n",
    "| `skip_sample_types` | Python list of sample_category values to skip from OBIS submission, such as controls or blanks.                       | [`negative control`, `positive control`] |\n",
    "| `skip_columns`      | Python list of columns to ignore when submitting to OBIS.                                                         | [`samp_collect_notes`]                                                                                   |\n",
    "| `taxonomic_api_source`      | Specify whether you want taxonomic assignment from either WoRMS or GBIF APIs.     |  `WoRMS` | \n",
    "<!-- | `analysisMetadata`     | Name of sheet in FAIRe template data Excel file with data about analysis methods.                                        | \"analysisMetadata_gomecc4_16s_p1\"                                                                                      | -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EDIT THIS CELL\n",
    "# Here, you specify where all of your data is, set a few other parameters, and set parameters related to the taxonomic assignment.\n",
    "\n",
    "# STEP 1\n",
    "# Assign sheets from your FAIRe Excel metadata file\n",
    "# NOTE: Left side is what edna2obis calls that data. Right side is the actual sheetname from your FAIRe Excel metadata file:\n",
    "params = {}\n",
    "params['sampleMetadata'] = \"sampleMetadata\"\n",
    "params['experimentRunMetadata']= \"experimentRunMetadata\"\n",
    "params['projectMetadata'] = \"projectMetadata\"\n",
    "\n",
    "params['excel_file'] = \"../raw-v3/FAIRe_NOAA_noaa-aoml-gomecc4_SHARING.xlsx\"\n",
    "params['FAIRe_NOAA_checklist'] = \"../raw-v3/FAIRe_NOAA_checklist_v1.0.xlsx\"\n",
    "# Be sure to also include the paths for the FAIRe metadata Excel sheet, and the FAIRe NOAA Checklist\n",
    "\n",
    "\n",
    "# STEP 2\n",
    "# Assign pathnames for your raw data. Each analysis should have 2 raw data files associated with it.\n",
    "params['datafiles'] = {\n",
    "    'gomecc4_18s_p1-6_v2024.10_241122': {\n",
    "        'taxonomy_file': '../raw-v3/asvTaxaFeatures_gomecc4_18s_p1-6_v2024.10_241122.tsv',\n",
    "        'occurrence_file': '../raw-v3/table_gomecc4_18s_p1-6_v2024.10_241122.tsv'\n",
    "    }, \n",
    "    'gomecc4_16s_p3-6_v2024.10_241122': {\n",
    "        'taxonomy_file': '../raw-v3/asvTaxaFeatures_gomecc4_16s_p3-6_v2024.10_241122.tsv',\n",
    "        'occurrence_file': '../raw-v3/table_gomecc4_16s_p3-6_v2024.10_241122.tsv'\n",
    "    }, \n",
    "    'gomecc4_16s_p1-2_v2024.10_241122': {\n",
    "        'taxonomy_file': '../raw-v3/asvTaxaFeatures_gomecc4_16s_p1-2_v2024.10_241122.tsv',\n",
    "        'occurrence_file': '../raw-v3/table_gomecc4_16s_p1-2_v2024.10_241122.tsv'\n",
    "    },\n",
    "    # Add other analysis runs here, following the pattern:\n",
    "    # 'your_analysis_run_name': {\n",
    "    #     'taxonomy_file': 'path/to/your/asvTaxaFeatures_your_analysis_run_name.tsv',\n",
    "    #     'occurrence_file': 'path/to/your/table_your_analysis_run_name.tsv'\n",
    "    # },\n",
    "}\n",
    "\n",
    "\n",
    "# STEP 3:\n",
    "# However you denote control / blank samples, specify here. \n",
    "params['skip_sample_types'] = ['negative control','positive control']\n",
    "params['skip_columns']= ['samp_collect_notes','date_modified','modified_by']\n",
    "\n",
    "\n",
    "# STEP 4:\n",
    "# Specify which API you would like to use to assign taxonomy. Options are either WoRMS or GBIF\n",
    "# Taxonomic Assignment Parameters:\n",
    "params['taxonomic_api_source'] = 'WoRMS'\n",
    "# params['taxonomic_api_source'] = 'GBIF'\n",
    "\n",
    "\n",
    "# STEP 5:\n",
    "# Define which assays should not consider 'species' rank in their taxonomic assignment.\n",
    "# This is because certain assays (for example, 16S) species' level assignments are not useful / correct.\n",
    "# \tAdditionally, for example, 18S species' level assignments are good, and we want them!\n",
    "# This should be the exact 'assay_name' value as found in your analysisMetadata sheets (cell D3)\n",
    "# and subsequently in the 'assay_name' column of the intermediate occurrence.csv\n",
    "params['user_defined_assays_to_skip_species'] = [\n",
    "    'ssu16sv4v5-emp',  # Example, replace with your actual 16S assay names\n",
    "    # 'another_16s_assay_name_if_any',\n",
    "]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load data\n",
    "\n",
    "Note that in a Jupyter Notebook, the current directory is always where the .ipynb file is being run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load project, sample, experimentRun, and analysis data from the FAIRe Excel file\n",
    "\n",
    "projectMetadata, sampleMetadata, and experimentRunMetadata can be loaded normally, but we dynamically load the analysisMetadata sheet(s). The user may have any number of analysisMetadata sheets in their submission, and the cell below will detect each one automatically and load their data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Discovered 3 analysis metadata sheets:\n",
      "  - analysisMetadata_gomecc4_16s_p1\n",
      "  - analysisMetadata_gomecc4_16s_p3\n",
      "  - analysisMetadata_gomecc4_18s_p1\n",
      "  - Processing sheet 'analysisMetadata_gomecc4_16s_p1': Found assay 'ssu16sv4v5-emp' with analysis run 'gomecc4_16s_p1-2_v2024.10_241122'\n",
      "  - Processing sheet 'analysisMetadata_gomecc4_16s_p3': Found assay 'ssu16sv4v5-emp' with analysis run 'gomecc4_16s_p3-6_v2024.10_241122'\n",
      "  - Processing sheet 'analysisMetadata_gomecc4_18s_p1': Found assay 'ssu18sv9-emp' with analysis run 'gomecc4_18s_p1-6_v2024.10_241122'\n",
      "\n",
      "Summary of analyses by assay (from analysisMetadata sheets):\n",
      "  - Assay 'ssu16sv4v5-emp': 2 analysis run(s)\n",
      "    - gomecc4_16s_p1-2_v2024.10_241122\n",
      "    - gomecc4_16s_p3-6_v2024.10_241122\n",
      "  - Assay 'ssu18sv9-emp': 1 analysis run(s)\n",
      "    - gomecc4_18s_p1-6_v2024.10_241122\n",
      "\n",
      "User-defined params['datafiles'] content (from Cell 8):\n",
      "  Analysis Run Name: 'gomecc4_18s_p1-6_v2024.10_241122'\n",
      "    Taxonomy File: ../raw-v3/asvTaxaFeatures_gomecc4_18s_p1-6_v2024.10_241122.tsv\n",
      "    Occurrence File: ../raw-v3/table_gomecc4_18s_p1-6_v2024.10_241122.tsv\n",
      "  Analysis Run Name: 'gomecc4_16s_p3-6_v2024.10_241122'\n",
      "    Taxonomy File: ../raw-v3/asvTaxaFeatures_gomecc4_16s_p3-6_v2024.10_241122.tsv\n",
      "    Occurrence File: ../raw-v3/table_gomecc4_16s_p3-6_v2024.10_241122.tsv\n",
      "  Analysis Run Name: 'gomecc4_16s_p1-2_v2024.10_241122'\n",
      "    Taxonomy File: ../raw-v3/asvTaxaFeatures_gomecc4_16s_p1-2_v2024.10_241122.tsv\n",
      "    Occurrence File: ../raw-v3/table_gomecc4_16s_p1-2_v2024.10_241122.tsv\n"
     ]
    }
   ],
   "source": [
    "# Discover all sheets in the Excel file\n",
    "excel = pd.ExcelFile(params['excel_file'])\n",
    "all_sheets = excel.sheet_names\n",
    "\n",
    "# Find analysis metadata sheets\n",
    "analysis_sheets = [sheet for sheet in all_sheets if sheet.startswith('analysisMetadata')]\n",
    "print(f\"Discovered {len(analysis_sheets)} analysis metadata sheets:\")\n",
    "for sheet_name_iter in analysis_sheets: # Renamed 'sheet' to 'sheet_name_iter'\n",
    "    print(f\"  - {sheet_name_iter}\")\n",
    "\n",
    "# Load the main data sheets (projectMetadata, sampleMetadata, experimentRunMetadata)\n",
    "data = pd.read_excel(\n",
    "    params['excel_file'],\n",
    "    [params['projectMetadata'], params['sampleMetadata'], params['experimentRunMetadata']],\n",
    "    index_col=None, na_values=[\"\"], comment=\"#\"\n",
    ")\n",
    "\n",
    "# Load all analysis metadata sheets\n",
    "# This dictionary stores the actual DataFrames from each analysisMetadata sheet.\n",
    "analysis_data_by_assay = {}\n",
    "\n",
    "for sheet_name_iter in analysis_sheets:\n",
    "    # Load the sheet\n",
    "    analysis_df = pd.read_excel(params['excel_file'], sheet_name_iter)\n",
    "    \n",
    "    # Get assay_name and analysis_run_name from specific cells (D3 and D4 in FAIRe template)\n",
    "    # Ensure these are strings for reliable dictionary keys.\n",
    "    assay_name = str(analysis_df.iloc[1, 3])        # Corresponds to Excel cell D3\n",
    "    analysis_run_name = str(analysis_df.iloc[2, 3]) # Corresponds to Excel cell D4\n",
    "    \n",
    "    print(f\"  - Processing sheet '{sheet_name_iter}': Found assay '{assay_name}' with analysis run '{analysis_run_name}'\")\n",
    "    \n",
    "    # Store the analysis DataFrame in analysis_data_by_assay, organized by assay_name then analysis_run_name\n",
    "    if assay_name not in analysis_data_by_assay:\n",
    "        analysis_data_by_assay[assay_name] = {}\n",
    "    analysis_data_by_assay[assay_name][analysis_run_name] = analysis_df\n",
    "\n",
    "# Add the structured analysis data to the main 'data' dictionary\n",
    "data['analysis_data_by_assay'] = analysis_data_by_assay\n",
    "\n",
    "# For backward compatibility or general reference, store the DataFrame of the first analysis sheet found\n",
    "if analysis_sheets:\n",
    "    data['analysisMetadata'] = pd.read_excel(params['excel_file'], analysis_sheets[0])\n",
    "else:\n",
    "    print(\"Warning: No analysis metadata sheets found! 'data['analysisMetadata']' will not be populated.\")\n",
    "\n",
    "# Print summary of analyses by assay from the loaded metadata sheets\n",
    "print(\"\\nSummary of analyses by assay (from analysisMetadata sheets):\")\n",
    "for assay, analyses_dict in analysis_data_by_assay.items():\n",
    "    print(f\"  - Assay '{assay}': {len(analyses_dict)} analysis run(s)\")\n",
    "    for run_name_key in analyses_dict.keys():\n",
    "        print(f\"    - {run_name_key}\")\n",
    "\n",
    "# Verify the user-defined params['analysis_files'] from Cell 8\n",
    "print(\"\\nUser-defined params['datafiles'] content (from Cell 8):\")\n",
    "if 'datafiles' in params and params['datafiles']:\n",
    "    for run_name, files_dict in params['datafiles'].items():\n",
    "        print(f\"  Analysis Run Name: '{run_name}'\")\n",
    "        print(f\"    Taxonomy File: {files_dict.get('taxonomy_file', 'Not specified')}\")\n",
    "        print(f\"    Occurrence File: {files_dict.get('occurrence_file', 'Not specified')}\")\n",
    "        # Check if this run_name from params matches one found in the Excel sheets\n",
    "        found_in_excel = any(run_name in an_dict for an_dict in analysis_data_by_assay.values())\n",
    "        if not found_in_excel:\n",
    "            print(f\"    Warning: Analysis run name '{run_name}' from params['datafiles'] was not found as an analysis_run_name in any analysisMetadata sheet.\")\n",
    "else:\n",
    "    print(\"  params['datafiles'] is empty or not defined in Cell 8.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename keys in data dictionary to a general term\n",
    "data['sampleMetadata'] = data.pop(params['sampleMetadata'])\n",
    "data['experimentRunMetadata'] = data.pop(params['experimentRunMetadata'])\n",
    "# The line below is already done in Cell 4. It is treated differently because the exact sheet name is not 'analysisMetadata'\n",
    "# data['analysisMetadata'] = data.pop(params['analysisMetadata'])\n",
    "data['projectMetadata'] = data.pop(params['projectMetadata'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sampleMetadata \n",
    "Contextual data about the samples collected, such as when it was collected, where it was collected from, what kind of sample it is, and what were the properties of the environment or experimental condition from which the sample was taken. Each row is a distinct sample, or Event. Most of this information is recorded during sample collection. This sheet contains terms from the FAIRe NOAA data template. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_name</th>\n",
       "      <th>samp_category</th>\n",
       "      <th>neg_cont_type</th>\n",
       "      <th>pos_cont_type</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sample_derived_from</th>\n",
       "      <th>sample_composed_of</th>\n",
       "      <th>rel_cont_id</th>\n",
       "      <th>biological_rep_relation</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>verbatimLongitude</th>\n",
       "      <th>verbatimLatitude</th>\n",
       "      <th>verbatimCoordinateSystem</th>\n",
       "      <th>verbatimSRS</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>eventDurationValue</th>\n",
       "      <th>verbatimEventDate</th>\n",
       "      <th>verbatimEventTime</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>habitat_natural_artificial_0_1</th>\n",
       "      <th>samp_collect_method</th>\n",
       "      <th>...</th>\n",
       "      <th>phosphate</th>\n",
       "      <th>phosphate_unit</th>\n",
       "      <th>pressure</th>\n",
       "      <th>pressure_unit</th>\n",
       "      <th>silicate</th>\n",
       "      <th>silicate_unit</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_alkalinity_unit</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>transmittance_unit</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station_id</th>\n",
       "      <th>ctd_cast_number</th>\n",
       "      <th>ctd_bottle_number</th>\n",
       "      <th>replicate_number</th>\n",
       "      <th>extract_id</th>\n",
       "      <th>extract_plate</th>\n",
       "      <th>extract_well_number</th>\n",
       "      <th>extract_well_position</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>organism</th>\n",
       "      <th>samp_collect_notes</th>\n",
       "      <th>dna_yield</th>\n",
       "      <th>dna_yield_unit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>26.997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>0</td>\n",
       "      <td>https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md</td>\n",
       "      <td>...</td>\n",
       "      <td>1.94489</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>623</td>\n",
       "      <td>dbar</td>\n",
       "      <td>20.3569</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>2318.9</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>not provided</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>52</td>\n",
       "      <td>D7</td>\n",
       "      <td>SAMN37516091</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>DCM = deep chlorophyl max.</td>\n",
       "      <td>12.057</td>\n",
       "      <td>ng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>26.997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>0</td>\n",
       "      <td>https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md</td>\n",
       "      <td>...</td>\n",
       "      <td>1.94489</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>623</td>\n",
       "      <td>dbar</td>\n",
       "      <td>20.3569</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>2318.9</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>not provided</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>60</td>\n",
       "      <td>D8</td>\n",
       "      <td>SAMN37516092</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>DCM was around 80 m and not well defined.</td>\n",
       "      <td>17.115</td>\n",
       "      <td>ng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>26.997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>0</td>\n",
       "      <td>https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md</td>\n",
       "      <td>...</td>\n",
       "      <td>1.94489</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>623</td>\n",
       "      <td>dbar</td>\n",
       "      <td>20.3569</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>2318.9</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>not provided</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>62</td>\n",
       "      <td>F8</td>\n",
       "      <td>SAMN37516093</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>Surface CTD bottles did not fire correctly; hand niskin bottle used for the surface cast. PM cast.</td>\n",
       "      <td>10.8345</td>\n",
       "      <td>ng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>26.997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>0</td>\n",
       "      <td>https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0517</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>49</td>\n",
       "      <td>dbar</td>\n",
       "      <td>1.05635</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>2371</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>4.665</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>not provided</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>53</td>\n",
       "      <td>E7</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>Only enough water for 2 surface replicates.</td>\n",
       "      <td>223.5</td>\n",
       "      <td>ng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>26.997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>0</td>\n",
       "      <td>https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0517</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>49</td>\n",
       "      <td>dbar</td>\n",
       "      <td>1.05635</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>2371</td>\n",
       "      <td>µmol/kg</td>\n",
       "      <td>4.665</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>not provided</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>46</td>\n",
       "      <td>F6</td>\n",
       "      <td>SAMN37516095</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>103.26</td>\n",
       "      <td>ng</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 166 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 samp_name samp_category neg_cont_type pos_cont_type  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A        sample           NaN           NaN   \n",
       "1  GOMECC4_27N_Sta1_Deep_B        sample           NaN           NaN   \n",
       "2  GOMECC4_27N_Sta1_Deep_C        sample           NaN           NaN   \n",
       "3   GOMECC4_27N_Sta1_DCM_A        sample           NaN           NaN   \n",
       "4   GOMECC4_27N_Sta1_DCM_B        sample           NaN           NaN   \n",
       "\n",
       "        materialSampleID  sample_derived_from  sample_composed_of  \\\n",
       "0  GOMECC4_27N_Sta1_Deep                  NaN                 NaN   \n",
       "1  GOMECC4_27N_Sta1_Deep                  NaN                 NaN   \n",
       "2  GOMECC4_27N_Sta1_Deep                  NaN                 NaN   \n",
       "3   GOMECC4_27N_Sta1_DCM                  NaN                 NaN   \n",
       "4   GOMECC4_27N_Sta1_DCM                  NaN                 NaN   \n",
       "\n",
       "   rel_cont_id  biological_rep_relation decimalLongitude decimalLatitude  \\\n",
       "0          NaN                      NaN          -79.618          26.997   \n",
       "1          NaN                      NaN          -79.618          26.997   \n",
       "2          NaN                      NaN          -79.618          26.997   \n",
       "3          NaN                      NaN          -79.618          26.997   \n",
       "4          NaN                      NaN          -79.618          26.997   \n",
       "\n",
       "   verbatimLongitude  verbatimLatitude  verbatimCoordinateSystem verbatimSRS  \\\n",
       "0                NaN               NaN                       NaN       WGS84   \n",
       "1                NaN               NaN                       NaN       WGS84   \n",
       "2                NaN               NaN                       NaN       WGS84   \n",
       "3                NaN               NaN                       NaN       WGS84   \n",
       "4                NaN               NaN                       NaN       WGS84   \n",
       "\n",
       "                                  geo_loc_name               eventDate  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)  2021-09-14T11:00-04:00   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)  2021-09-14T11:00-04:00   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)  2021-09-14T11:00-04:00   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)  2021-09-14T11:00-04:00   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)  2021-09-14T11:00-04:00   \n",
       "\n",
       "   eventDurationValue  verbatimEventDate  verbatimEventTime  \\\n",
       "0                 NaN                NaN                NaN   \n",
       "1                 NaN                NaN                NaN   \n",
       "2                 NaN                NaN                NaN   \n",
       "3                 NaN                NaN                NaN   \n",
       "4                 NaN                NaN                NaN   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "4  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "\n",
       "                  env_medium  habitat_natural_artificial_0_1  \\\n",
       "0  sea water [ENVO:00002149]                               0   \n",
       "1  sea water [ENVO:00002149]                               0   \n",
       "2  sea water [ENVO:00002149]                               0   \n",
       "3  sea water [ENVO:00002149]                               0   \n",
       "4  sea water [ENVO:00002149]                               0   \n",
       "\n",
       "                                                              samp_collect_method  \\\n",
       "0  https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md   \n",
       "1  https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md   \n",
       "2  https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md   \n",
       "3  https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md   \n",
       "4  https://zenodo.org/records/14224755 (v1.1.0) protocol_sampling_sterivex_dry.md   \n",
       "\n",
       "   ... phosphate phosphate_unit pressure pressure_unit silicate silicate_unit  \\\n",
       "0  ...   1.94489        µmol/kg      623          dbar  20.3569       µmol/kg   \n",
       "1  ...   1.94489        µmol/kg      623          dbar  20.3569       µmol/kg   \n",
       "2  ...   1.94489        µmol/kg      623          dbar  20.3569       µmol/kg   \n",
       "3  ...    0.0517        µmol/kg       49          dbar  1.05635       µmol/kg   \n",
       "4  ...    0.0517        µmol/kg       49          dbar  1.05635       µmol/kg   \n",
       "\n",
       "  tot_alkalinity tot_alkalinity_unit  transmittance transmittance_unit  \\\n",
       "0         2318.9             µmol/kg         4.7221                NaN   \n",
       "1         2318.9             µmol/kg         4.7221                NaN   \n",
       "2         2318.9             µmol/kg         4.7221                NaN   \n",
       "3           2371             µmol/kg          4.665                NaN   \n",
       "4           2371             µmol/kg          4.665                NaN   \n",
       "\n",
       "  serial_number  line_id  station_id  ctd_cast_number  ctd_bottle_number  \\\n",
       "0   GOMECC4_001      27N        Sta1     not provided                  3   \n",
       "1   GOMECC4_002      27N        Sta1     not provided                  3   \n",
       "2   GOMECC4_003      27N        Sta1     not provided                  3   \n",
       "3   GOMECC4_004      27N        Sta1     not provided                 14   \n",
       "4   GOMECC4_005      27N        Sta1     not provided                 14   \n",
       "\n",
       "  replicate_number extract_id      extract_plate  extract_well_number  \\\n",
       "0                A  Plate4_52  GOMECC2021_Plate4                   52   \n",
       "1                B  Plate4_60  GOMECC2021_Plate4                   60   \n",
       "2                C  Plate4_62  GOMECC2021_Plate4                   62   \n",
       "3                A  Plate4_53  GOMECC2021_Plate4                   53   \n",
       "4                B  Plate4_46  GOMECC2021_Plate4                   46   \n",
       "\n",
       "  extract_well_position biosample_accession             organism  \\\n",
       "0                    D7        SAMN37516091  seawater metagenome   \n",
       "1                    D8        SAMN37516092  seawater metagenome   \n",
       "2                    F8        SAMN37516093  seawater metagenome   \n",
       "3                    E7        SAMN37516094  seawater metagenome   \n",
       "4                    F6        SAMN37516095  seawater metagenome   \n",
       "\n",
       "                                                                                   samp_collect_notes  \\\n",
       "0                                                                          DCM = deep chlorophyl max.   \n",
       "1                                                           DCM was around 80 m and not well defined.   \n",
       "2  Surface CTD bottles did not fire correctly; hand niskin bottle used for the surface cast. PM cast.   \n",
       "3                                                         Only enough water for 2 surface replicates.   \n",
       "4                                                                                                 NaN   \n",
       "\n",
       "  dna_yield dna_yield_unit  \n",
       "0    12.057             ng  \n",
       "1    17.115             ng  \n",
       "2   10.8345             ng  \n",
       "3     223.5             ng  \n",
       "4    103.26             ng  \n",
       "\n",
       "[5 rows x 166 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['sampleMetadata'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### experimentRunMetadata  \n",
    "Contextual data about how the samples were prepared for sequencing. Includes how they were extracted, what amplicon was targeted, how they were sequenced. Each row is a separate sequencing library preparation, distinguished by a unique lib_id. **TODO: MIGHT NEED HELP WITH THIS DESCRIPTION**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samp_name</th>\n",
       "      <th>assay_name</th>\n",
       "      <th>pcr_plate_id</th>\n",
       "      <th>lib_id</th>\n",
       "      <th>seq_run_id</th>\n",
       "      <th>lib_conc</th>\n",
       "      <th>lib_conc_unit</th>\n",
       "      <th>lib_conc_meth</th>\n",
       "      <th>phix_perc</th>\n",
       "      <th>mid_forward</th>\n",
       "      <th>mid_reverse</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>checksum_filename</th>\n",
       "      <th>checksum_filename2</th>\n",
       "      <th>associatedSequences</th>\n",
       "      <th>input_read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_NegativeControl_1</td>\n",
       "      <td>ssu16sv4v5-emp</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>GOMECC16S_Neg1</td>\n",
       "      <td>20220613_Amplicon_PE250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>CGTCGCTA</td>\n",
       "      <td>GOMECC16S_Neg1_S499_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Neg1_S499_L001_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26148505 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516589 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>29319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_NegativeControl_2</td>\n",
       "      <td>ssu16sv4v5-emp</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>GOMECC16S_Neg2</td>\n",
       "      <td>20220613_Amplicon_PE250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAGCAGCT</td>\n",
       "      <td>CTAGAGCT</td>\n",
       "      <td>GOMECC16S_Neg2_S500_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Neg2_S500_L001_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26148503 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516590 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>30829</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   samp_name      assay_name    pcr_plate_id          lib_id  \\\n",
       "0  GOMECC4_NegativeControl_1  ssu16sv4v5-emp  not applicable  GOMECC16S_Neg1   \n",
       "1  GOMECC4_NegativeControl_2  ssu16sv4v5-emp  not applicable  GOMECC16S_Neg2   \n",
       "\n",
       "                seq_run_id  lib_conc  lib_conc_unit  lib_conc_meth  phix_perc  \\\n",
       "0  20220613_Amplicon_PE250       NaN            NaN            NaN        NaN   \n",
       "1  20220613_Amplicon_PE250       NaN            NaN            NaN        NaN   \n",
       "\n",
       "  mid_forward mid_reverse                                  filename  \\\n",
       "0    TAGCAGCT    CGTCGCTA  GOMECC16S_Neg1_S499_L001_R1_001.fastq.gz   \n",
       "1    TAGCAGCT    CTAGAGCT  GOMECC16S_Neg2_S500_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                  filename2  checksum_filename  \\\n",
       "0  GOMECC16S_Neg1_S499_L001_R2_001.fastq.gz                NaN   \n",
       "1  GOMECC16S_Neg2_S500_L001_R2_001.fastq.gz                NaN   \n",
       "\n",
       "   checksum_filename2  \\\n",
       "0                 NaN   \n",
       "1                 NaN   \n",
       "\n",
       "                                                                                                                                     associatedSequences  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/sra/SRR26148505 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516589 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "1  https://www.ncbi.nlm.nih.gov/sra/SRR26148503 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516590 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "\n",
       "   input_read_count  \n",
       "0             29319  \n",
       "1             30829  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['experimentRunMetadata'].head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load ASV data  \n",
    "There is one ASV file for each marker that was sequenced. The ASV data files have one row for each unique amplicon sequence variants (ASVs). They contain the ASV DNA sequence, a unique hash identifier the taxonomic assignment for each ASV, the confidence given that assignment by the naive-bayes classifier, and then the number of reads observed in each sample. \n",
    "\n",
    "This file is created automatically with [Tourmaline v.2023.5+](https://github.com/aomlomics/tourmaline), and is found in `01-taxonomy/asv_taxa_sample_table.tsv`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| column name    | definition                                                                                                                                                                                                                                                                                                                                                                                              |\n",
    "|----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| featureid      | A hash of the ASV sequence, used as a unique identifier for the ASV.                                                                                                                                                                                                                                                                                                                                    |\n",
    "| taxonomy       | The full taxonomy assigned to an ASV sequence. This string could be formatted in very different ways depending on the reference database used during classification, however it should always be in reverse rank order separated by ;. We provide examples for how to process results from a Silva classifier and the PR2 18S classifier. For other taxonomy formats, the code will need to be adapted. |\n",
    "| Confidence     | This is the confidence score assigned the taxonomic classification with a naive-bayes classifier.                                                                                                                                                                                                                                                                                                       |\n",
    "| sample columns | The next columns each represent a sample (or eventID), and the number of reads for that ASV observed in the sample.                                                                                                                                                                                                                                                                                     |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading raw data files based on params['datafiles']:\n",
      "  Processing analysis run: gomecc4_18s_p1-6_v2024.10_241122\n",
      "    Successfully loaded taxonomy file: ../raw-v3/asvTaxaFeatures_gomecc4_18s_p1-6_v2024.10_241122.tsv\n",
      "    Successfully loaded occurrence file: ../raw-v3/table_gomecc4_18s_p1-6_v2024.10_241122.tsv\n",
      "      Shape after loading: (24473, 501)\n",
      "      Column names (first 5): ['OTU ID', 'GOMECC4_27N_Sta1_DCM_A', 'GOMECC4_27N_Sta1_DCM_B', 'GOMECC4_27N_Sta1_DCM_C', 'GOMECC4_27N_Sta1_Deep_A']\n",
      "  Processing analysis run: gomecc4_16s_p3-6_v2024.10_241122\n",
      "    Successfully loaded taxonomy file: ../raw-v3/asvTaxaFeatures_gomecc4_16s_p3-6_v2024.10_241122.tsv\n",
      "    Successfully loaded occurrence file: ../raw-v3/table_gomecc4_16s_p3-6_v2024.10_241122.tsv\n",
      "      Shape after loading: (49523, 312)\n",
      "      Column names (first 5): ['OTU ID', 'GOMECC4_27N_Sta1_DCM_A', 'GOMECC4_27N_Sta1_DCM_B', 'GOMECC4_27N_Sta1_DCM_C', 'GOMECC4_27N_Sta1_Deep_A']\n",
      "  Processing analysis run: gomecc4_16s_p1-2_v2024.10_241122\n",
      "    Successfully loaded taxonomy file: ../raw-v3/asvTaxaFeatures_gomecc4_16s_p1-2_v2024.10_241122.tsv\n",
      "    Successfully loaded occurrence file: ../raw-v3/table_gomecc4_16s_p1-2_v2024.10_241122.tsv\n",
      "      Shape after loading: (19540, 198)\n",
      "      Column names (first 5): ['OTU ID', 'GOMECC4_Blank_DIW_20210930_A', 'GOMECC4_Blank_DIW_20210930_B', 'GOMECC4_Blank_DIW_20210930_C', 'GOMECC4_Blank_DIW_20211011_A']\n"
     ]
    }
   ],
   "source": [
    "# Read in ASV tables for each analysis run.\n",
    "# We now have separate taxonomy and occurrence files per analysis_run_name.\n",
    "\n",
    "raw_data_tables = {} # This will store DataFrames for taxonomy and occurrences for each run\n",
    "\n",
    "print(\"Loading raw data files based on params['datafiles']:\")\n",
    "if 'datafiles' in params and params['datafiles']:\n",
    "    for analysis_run_name, file_paths in params['datafiles'].items():\n",
    "        print(f\"  Processing analysis run: {analysis_run_name}\")\n",
    "        raw_data_tables[analysis_run_name] = {}\n",
    "        \n",
    "        # Load taxonomy file\n",
    "        if 'taxonomy_file' in file_paths:\n",
    "            tax_path = file_paths['taxonomy_file']\n",
    "            try:\n",
    "                raw_data_tables[analysis_run_name]['taxonomy'] = pd.read_table(tax_path, sep='\\t', low_memory=False)\n",
    "                print(f\"    Successfully loaded taxonomy file: {tax_path}\")\n",
    "                # Optional: print shape or head if useful for verification\n",
    "                # print(f\"      Shape: {raw_data_tables[analysis_run_name]['taxonomy'].shape}\")\n",
    "            except FileNotFoundError:\n",
    "                print(f\"    ERROR: Taxonomy file not found at {tax_path}\")\n",
    "            except Exception as e:\n",
    "                print(f\"    ERROR: Could not load taxonomy file {tax_path}. Error: {e}\")\n",
    "        else:\n",
    "            print(f\"    Warning: Taxonomy file path not specified for {analysis_run_name}\")\n",
    "            \n",
    "        # Load occurrence file\n",
    "        if 'occurrence_file' in file_paths:\n",
    "            occ_path = file_paths['occurrence_file']\n",
    "            try:\n",
    "                # Row 1 is a comment, Row 2 is header, Data starts Row 3.\n",
    "                # skiprows=1 means skip the first row (the comment).\n",
    "                # header=0 (after skipping 1 row) means use the NEW first row (original Row 2) as headers.\n",
    "                df_occ = pd.read_table(occ_path, \n",
    "                                       sep='\\t', \n",
    "                                       skiprows=1,  # Skip the comment line (original Row 1)\n",
    "                                       header=0,    # Use the next line (original Row 2) as column headers\n",
    "                                       low_memory=False)\n",
    "\n",
    "                # It's common for TSV tools to export the first column header with a '#' prefix, e.g., \"#OTU ID\".\n",
    "                # If so, remove the leading '#'.\n",
    "                if df_occ.columns[0].startswith('#'):\n",
    "                    df_occ.rename(columns={df_occ.columns[0]: df_occ.columns[0][1:]}, inplace=True)\n",
    "                \n",
    "                raw_data_tables[analysis_run_name]['occurrence'] = df_occ\n",
    "                print(f\"    Successfully loaded occurrence file: {occ_path}\")\n",
    "                print(f\"      Shape after loading: {raw_data_tables[analysis_run_name]['occurrence'].shape}\")\n",
    "                print(f\"      Column names (first 5): {raw_data_tables[analysis_run_name]['occurrence'].columns.tolist()[:5]}\")\n",
    "            except FileNotFoundError:\n",
    "                print(f\"    ERROR: Occurrence file not found at {occ_path}\")\n",
    "            except Exception as e:\n",
    "                print(f\"    ERROR: Could not load occurrence file {occ_path}. Error: {e}\")\n",
    "        else:\n",
    "            print(f\"    Warning: Occurrence file path not specified for {analysis_run_name}\")\n",
    "\n",
    "# Downstream code will need to be adapted to use 'raw_data_tables'.\n",
    "# The old 'asv_tables' variable, which assumed one merged table per \"gene\",\n",
    "# is now replaced by 'raw_data_tables' which holds separate 'taxonomy' and 'occurrence'\n",
    "# DataFrames for each 'analysis_run_name'.\n",
    "\n",
    "# For example, to access the taxonomy DataFrame for 'gomecc4_18s_p1-6_v2024.10_241122':\n",
    "# raw_data_tables['gomecc4_18s_p1-6_v2024.10_241122']['taxonomy']\n",
    "# And its occurrence DataFrame:\n",
    "# raw_data_tables['gomecc4_18s_p1-6_v2024.10_241122']['occurrence']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis runs for which data was loaded:\n",
      "  - gomecc4_18s_p1-6_v2024.10_241122\n",
      "    - Taxonomy table shape: (24473, 14)\n",
      "    - Occurrence table shape: (24473, 501)\n",
      "  - gomecc4_16s_p3-6_v2024.10_241122\n",
      "    - Taxonomy table shape: (49523, 12)\n",
      "    - Occurrence table shape: (49523, 312)\n",
      "  - gomecc4_16s_p1-2_v2024.10_241122\n",
      "    - Taxonomy table shape: (19540, 12)\n",
      "    - Occurrence table shape: (19540, 198)\n"
     ]
    }
   ],
   "source": [
    "# Inspect the keys of the newly loaded raw_data_tables\n",
    "# This will show the analysis_run_names for which data was loaded.\n",
    "if 'raw_data_tables' in locals() and raw_data_tables: # Check if it exists and is not empty\n",
    "    print(\"Analysis runs for which data was loaded:\")\n",
    "    for run_name in raw_data_tables.keys():\n",
    "        print(f\"  - {run_name}\")\n",
    "        if 'taxonomy' in raw_data_tables[run_name]:\n",
    "            print(f\"    - Taxonomy table shape: {raw_data_tables[run_name]['taxonomy'].shape}\")\n",
    "        else:\n",
    "            print(f\"    - Taxonomy table: Not loaded or error during load.\")\n",
    "        if 'occurrence' in raw_data_tables[run_name]:\n",
    "            print(f\"    - Occurrence table shape: {raw_data_tables[run_name]['occurrence'].shape}\")\n",
    "        else:\n",
    "            print(f\"    - Occurrence table: Not loaded or error during load.\")\n",
    "else:\n",
    "    print(\"raw_data_tables dictionary is not defined or is empty.\")\n",
    "\n",
    "# Example of how to access a specific table (optional, for your direct inspection):\n",
    "# if 'gomecc4_18s_p1-6_v2024.10_241122' in raw_data_tables:\n",
    "#     print(\"\\nHead of taxonomy table for 'gomecc4_18s_p1-6_v2024.10_241122':\")\n",
    "#     if 'taxonomy' in raw_data_tables['gomecc4_18s_p1-6_v2024.10_241122']:\n",
    "#         print(raw_data_tables['gomecc4_18s_p1-6_v2024.10_241122']['taxonomy'].head(2))\n",
    "#     print(\"\\nHead of occurrence table for 'gomecc4_18s_p1-6_v2024.10_241122':\")\n",
    "#     if 'occurrence' in raw_data_tables['gomecc4_18s_p1-6_v2024.10_241122']:\n",
    "#         print(raw_data_tables['gomecc4_18s_p1-6_v2024.10_241122']['occurrence'].head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Head of OCCURRENCE table for 'gomecc4_18s_p1-6_v2024.10_241122' (first 20 columns):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OTU ID</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_C</th>\n",
       "      <th>GOMECC4_27N_Sta6_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta6_DCM_B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>1518.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4268.0</td>\n",
       "      <td>2002.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9532.0</td>\n",
       "      <td>1930.0</td>\n",
       "      <td>2037.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>963.0</td>\n",
       "      <td>316.0</td>\n",
       "      <td>543.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>613.0</td>\n",
       "      <td>561.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>395.0</td>\n",
       "      <td>297.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>915.0</td>\n",
       "      <td>1447.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5d4df37251121c08397c6fbc27b06175</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>409.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f863f671a575c6ab587e8de0190d3335</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>2672.0</td>\n",
       "      <td>2424.0</td>\n",
       "      <td>2605.0</td>\n",
       "      <td>1918.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>1165.0</td>\n",
       "      <td>2267.0</td>\n",
       "      <td>2206.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1103.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>941.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             OTU ID  GOMECC4_27N_Sta1_DCM_A  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb                  1518.0   \n",
       "1  4e38e8ced9070952b314e1880bede1ca                   963.0   \n",
       "2  5d4df37251121c08397c6fbc27b06175                     0.0   \n",
       "3  f863f671a575c6ab587e8de0190d3335                     0.0   \n",
       "4  2a31e5c01634165da99e7381279baa75                  1165.0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_B  GOMECC4_27N_Sta1_DCM_C  GOMECC4_27N_Sta1_Deep_A  \\\n",
       "0                     0.0                     0.0                      6.0   \n",
       "1                   316.0                   543.0                     19.0   \n",
       "2                     4.0                     0.0                     12.0   \n",
       "3                     0.0                     0.0                      0.0   \n",
       "4                  2267.0                  2206.0                      2.0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_B  GOMECC4_27N_Sta1_Deep_C  \\\n",
       "0                      0.0                      0.0   \n",
       "1                     10.0                      0.0   \n",
       "2                      5.0                      0.0   \n",
       "3                      0.0                      0.0   \n",
       "4                      0.0                      0.0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Surface_A  GOMECC4_27N_Sta1_Surface_B  \\\n",
       "0                         0.0                      4268.0   \n",
       "1                         0.0                         0.0   \n",
       "2                         0.0                         0.0   \n",
       "3                         0.0                         0.0   \n",
       "4                         0.0                         0.0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_DCM_A  GOMECC4_27N_Sta4_DCM_B  GOMECC4_27N_Sta4_DCM_C  \\\n",
       "0                  2002.0                     0.0                    14.0   \n",
       "1                   613.0                   561.0                   434.0   \n",
       "2                     9.0                     0.0                     0.0   \n",
       "3                     0.0                     0.0                     5.0   \n",
       "4                     0.0                     0.0                     0.0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_Deep_A  GOMECC4_27N_Sta4_Deep_B  GOMECC4_27N_Sta4_Deep_C  \\\n",
       "0                      0.0                      0.0                      0.0   \n",
       "1                      0.0                    395.0                    297.0   \n",
       "2                      0.0                      0.0                      0.0   \n",
       "3                      0.0                      0.0                      0.0   \n",
       "4                      0.0                      2.0                      0.0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_Surface_A  GOMECC4_27N_Sta4_Surface_B  \\\n",
       "0                      9532.0                      1930.0   \n",
       "1                        76.0                       915.0   \n",
       "2                         0.0                       300.0   \n",
       "3                        92.0                      2672.0   \n",
       "4                      1103.0                       157.0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_Surface_C  GOMECC4_27N_Sta6_DCM_A  GOMECC4_27N_Sta6_DCM_B  \n",
       "0                      2037.0                     0.0                     0.0  \n",
       "1                      1447.0                   140.0                     0.0  \n",
       "2                         0.0                   864.0                   409.0  \n",
       "3                      2424.0                  2605.0                  1918.0  \n",
       "4                         0.0                   104.0                   941.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Choose an analysis run name that you have defined in params['datafiles']\n",
    "# For example, the first one:\n",
    "analysis_to_inspect = list(params['datafiles'].keys())[0] if params['datafiles'] else None\n",
    "\n",
    "if analysis_to_inspect and analysis_to_inspect in raw_data_tables:\n",
    "    if 'occurrence' in raw_data_tables[analysis_to_inspect]:\n",
    "        print(f\"Head of OCCURRENCE table for '{analysis_to_inspect}' (first 20 columns):\")\n",
    "        display(raw_data_tables[analysis_to_inspect]['occurrence'].iloc[:, 0:20].head())\n",
    "    else:\n",
    "        print(f\"Occurrence table not found for '{analysis_to_inspect}'.\")\n",
    "else:\n",
    "    if not analysis_to_inspect:\n",
    "        print(\"params['datafiles'] is empty. No analysis run to inspect.\")\n",
    "    else:\n",
    "        print(f\"Analysis run '{analysis_to_inspect}' not found in loaded raw_data_tables.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Drop samples with unwanted sample types  \n",
    "\n",
    "Often with eDNA projects, we have control samples that are sequenced along with our survey samples. These can include filtering distilled water, using pure water instead of DNA in a PCR or DNA extraction protocol, or a mock community of known microbial taxa. Controls can help identify and mitigate contaminant DNA in our samples, but are not useful for biodiversity platforms like OBIS. You can select which sample_type values to drop with the `skip_sample_types` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG in Cell 23: samples_to_drop (first 5): ['GOMECC4_Blank_DIW_20210915_A', 'GOMECC4_Blank_DIW_20210915_B', 'GOMECC4_Blank_DIW_20210915_C', 'GOMECC4_Blank_DIW_20210930_A', 'GOMECC4_Blank_DIW_20210930_B']\n"
     ]
    }
   ],
   "source": [
    "# This should be your Cell 23 (or equivalent)\n",
    "samps_to_remove = data['sampleMetadata']['samp_category'].isin(params['skip_sample_types'])\n",
    "samples_to_drop = data['sampleMetadata']['samp_name'][samps_to_remove].astype(str).str.strip().tolist() \n",
    "print(f\"DEBUG in Cell 23: samples_to_drop (first 5): {samples_to_drop[:5]}\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can view the list of samples to be dropped below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GOMECC4_Blank_DIW_20210915_A',\n",
       " 'GOMECC4_Blank_DIW_20210915_B',\n",
       " 'GOMECC4_Blank_DIW_20210915_C',\n",
       " 'GOMECC4_Blank_DIW_20210930_A',\n",
       " 'GOMECC4_Blank_DIW_20210930_B',\n",
       " 'GOMECC4_Blank_DIW_20210930_C',\n",
       " 'GOMECC4_Blank_DIW_20211011_A',\n",
       " 'GOMECC4_Blank_DIW_20211011_B',\n",
       " 'GOMECC4_Blank_DIW_20211011_C',\n",
       " 'GOMECC4_Blank_DIW_20211016_A',\n",
       " 'GOMECC4_Blank_DIW_20211016_B',\n",
       " 'GOMECC4_Blank_DIW_20211016_C',\n",
       " 'GOMECC4_ExtractionBlank_1',\n",
       " 'GOMECC4_ExtractionBlank_11',\n",
       " 'GOMECC4_ExtractionBlank_12',\n",
       " 'GOMECC4_ExtractionBlank_3',\n",
       " 'GOMECC4_ExtractionBlank_5',\n",
       " 'GOMECC4_ExtractionBlank_7',\n",
       " 'GOMECC4_ExtractionBlank_9',\n",
       " 'GOMECC4_MSUControl_1',\n",
       " 'GOMECC4_MSUControl_2',\n",
       " 'GOMECC4_MSUControl_3',\n",
       " 'GOMECC4_MSUControl_4',\n",
       " 'GOMECC4_MSUControl_5',\n",
       " 'GOMECC4_MSUControl_6',\n",
       " 'GOMECC4_MSUControl_7',\n",
       " 'GOMECC4_NegativeControl_1',\n",
       " 'GOMECC4_NegativeControl_2',\n",
       " 'GOMECC4_PositiveControl_1',\n",
       " 'GOMECC4_PositiveControl_2']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_to_drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove samples from sampleMetadata sheet\n",
    "data['sampleMetadata'] = data['sampleMetadata'][~samps_to_remove]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['sample'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the samp_category values left in your sampleMetadata. We only want 'sample' (indicating it is not a control or blank).\n",
    "data['sampleMetadata']['samp_category'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove samples from experimentRunMetadata\n",
    "prep_samps_to_remove = data['experimentRunMetadata']['samp_name'].isin(samples_to_drop)\n",
    "data['experimentRunMetadata'] = data['experimentRunMetadata'][~prep_samps_to_remove]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Drop unwanted samples from ASV files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to remove blank/control samples from loaded occurrence tables...\n",
      "  For analysis run 'gomecc4_18s_p1-6_v2024.10_241122':\n",
      "    Original columns: 501, Columns after removal: 473\n",
      "    Removed (28 total for this run): ['GOMECC4_Blank_DIW_20210915_A', 'GOMECC4_Blank_DIW_20210915_B', 'GOMECC4_Blank_DIW_20210915_C', 'GOMECC4_Blank_DIW_20210930_A', 'GOMECC4_Blank_DIW_20210930_B', 'GOMECC4_Blank_DIW_20210930_C', 'GOMECC4_Blank_DIW_20211011_A', 'GOMECC4_Blank_DIW_20211011_B', 'GOMECC4_Blank_DIW_20211011_C', 'GOMECC4_Blank_DIW_20211016_A', 'GOMECC4_Blank_DIW_20211016_B', 'GOMECC4_Blank_DIW_20211016_C', 'GOMECC4_ExtractionBlank_1', 'GOMECC4_ExtractionBlank_11', 'GOMECC4_ExtractionBlank_12', 'GOMECC4_ExtractionBlank_3', 'GOMECC4_ExtractionBlank_5', 'GOMECC4_ExtractionBlank_7', 'GOMECC4_ExtractionBlank_9', 'GOMECC4_MSUControl_1', 'GOMECC4_MSUControl_2', 'GOMECC4_MSUControl_3', 'GOMECC4_MSUControl_4', 'GOMECC4_MSUControl_5', 'GOMECC4_MSUControl_6', 'GOMECC4_MSUControl_7', 'GOMECC4_NegativeControl_1', 'GOMECC4_NegativeControl_2']\n",
      "  For analysis run 'gomecc4_16s_p3-6_v2024.10_241122':\n",
      "    Original columns: 312, Columns after removal: 291\n",
      "    Removed (21 total for this run): ['GOMECC4_Blank_DIW_20210915_A', 'GOMECC4_Blank_DIW_20210915_B', 'GOMECC4_Blank_DIW_20210915_C', 'GOMECC4_Blank_DIW_20211016_A', 'GOMECC4_Blank_DIW_20211016_B', 'GOMECC4_Blank_DIW_20211016_C', 'GOMECC4_ExtractionBlank_11', 'GOMECC4_ExtractionBlank_12', 'GOMECC4_ExtractionBlank_5', 'GOMECC4_ExtractionBlank_7', 'GOMECC4_ExtractionBlank_9', 'GOMECC4_MSUControl_1', 'GOMECC4_MSUControl_2', 'GOMECC4_MSUControl_3', 'GOMECC4_MSUControl_4', 'GOMECC4_MSUControl_5', 'GOMECC4_MSUControl_6', 'GOMECC4_NegativeControl_1', 'GOMECC4_NegativeControl_2', 'GOMECC4_PositiveControl_1', 'GOMECC4_PositiveControl_2']\n",
      "  For analysis run 'gomecc4_16s_p1-2_v2024.10_241122':\n",
      "    Original columns: 198, Columns after removal: 183\n",
      "    Removed (15 total for this run): ['GOMECC4_Blank_DIW_20210930_A', 'GOMECC4_Blank_DIW_20210930_B', 'GOMECC4_Blank_DIW_20210930_C', 'GOMECC4_Blank_DIW_20211011_A', 'GOMECC4_Blank_DIW_20211011_B', 'GOMECC4_Blank_DIW_20211011_C', 'GOMECC4_ExtractionBlank_1', 'GOMECC4_ExtractionBlank_3', 'GOMECC4_MSUControl_1', 'GOMECC4_MSUControl_2', 'GOMECC4_MSUControl_3', 'GOMECC4_MSUControl_4', 'GOMECC4_MSUControl_5', 'GOMECC4_MSUControl_6', 'GOMECC4_PositiveControl_1']\n",
      "Finished blank/control sample removal process.\n"
     ]
    }
   ],
   "source": [
    "# This cell REMOVES blank/control samples from the ALREADY LOADED occurrence tables\n",
    "\n",
    "print(\"Attempting to remove blank/control samples from loaded occurrence tables...\")\n",
    "if 'raw_data_tables' in locals() and raw_data_tables:\n",
    "    if 'samples_to_drop' in locals() and samples_to_drop:\n",
    "        for analysis_run_name, tables_dict in raw_data_tables.items():\n",
    "            if 'occurrence' in tables_dict and not tables_dict['occurrence'].empty:\n",
    "                occ_df = tables_dict['occurrence'] # Get a reference to the occurrence DataFrame\n",
    "                original_cols_count = len(occ_df.columns)\n",
    "                \n",
    "                # Identify which of the samples_to_drop are actual columns in THIS occurrence table\n",
    "                cols_to_remove_in_this_df = [col for col in samples_to_drop if col in occ_df.columns]\n",
    "                \n",
    "                if cols_to_remove_in_this_df:\n",
    "                    # Perform the drop operation. This MODIFIES the DataFrame in raw_data_tables.\n",
    "                    raw_data_tables[analysis_run_name]['occurrence'] = occ_df.drop(columns=cols_to_remove_in_this_df)\n",
    "                    \n",
    "                    print(f\"  For analysis run '{analysis_run_name}':\")\n",
    "                    print(f\"    Original columns: {original_cols_count}, Columns after removal: {len(raw_data_tables[analysis_run_name]['occurrence'].columns)}\")\n",
    "                    print(f\"    Removed ({len(cols_to_remove_in_this_df)} total for this run): {cols_to_remove_in_this_df}\") # Print all removed\n",
    "                else:\n",
    "                    print(f\"  For analysis run '{analysis_run_name}': No specified blank/control samples found to remove in this table's columns.\")\n",
    "            else:\n",
    "                print(f\"  Skipping analysis run '{analysis_run_name}': No 'occurrence' table found or it is empty.\")\n",
    "        print(\"Finished blank/control sample removal process.\")\n",
    "    else:\n",
    "        print(\"WARNING: 'samples_to_drop' list not found or empty. No columns removed from occurrence tables.\")\n",
    "else:\n",
    "    print(\"WARNING: 'raw_data_tables' not found or empty. Cannot remove columns.\")\n",
    "\n",
    "# You can add a verification step here if you like, for example:\n",
    "# target_run_to_check = 'gomecc4_18s_p1-6_v2024.10_241122' # Use one of your actual analysis run names\n",
    "# if target_run_to_check in raw_data_tables and 'occurrence' in raw_data_tables[target_run_to_check]:\n",
    "# print(f\"\\nColumns in occurrence table for '{target_run_to_check}' AFTER blank removal:\")\n",
    "# print(raw_data_tables[target_run_to_check]['occurrence'].columns.tolist()[:30])\n",
    "# else:\n",
    "# print(f\"\\nCould not verify columns for '{target_run_to_check}'.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Drop columns with all NAs  \n",
    "\n",
    "If your project data file has columns with only NAs, this code will check for those, provide their column headers for verification, then remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identifies columns that have ALL NA values in 'sampleMetadata' and 'experimentRunMetadata'.\n",
    "# This is based on your original code, modified to exclude 'analysisMetadata'.\n",
    "# TODO: remove empty rows from analysisMetadata. Inform user of rows / fields deleted \n",
    "\n",
    "dropped = pd.DataFrame()\n",
    "\n",
    "# Only check 'sampleMetadata' and 'experimentRunMetadata'\n",
    "for sheet_name in ['sampleMetadata', 'experimentRunMetadata']:\n",
    "    # Safety check: ensure the sheet exists in data and is not empty\n",
    "    if sheet_name in data and not data[sheet_name].empty:\n",
    "        all_na_cols = data[sheet_name].columns[data[sheet_name].isnull().all(axis=0)]\n",
    "        res = pd.Series(all_na_cols, name=sheet_name)\n",
    "        dropped = pd.concat([dropped, res], axis=1)\n",
    "    elif sheet_name not in data:\n",
    "        print(f\"FYI: Sheet '{sheet_name}' not found in 'data' dictionary. Cannot check for all-NA columns.\")\n",
    "    else: # In data but empty\n",
    "        print(f\"FYI: Sheet '{sheet_name}' is empty. Cannot check for all-NA columns.\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which columns in each sheet have only NA values?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleMetadata</th>\n",
       "      <th>experimentRunMetadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>neg_cont_type</td>\n",
       "      <td>lib_conc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pos_cont_type</td>\n",
       "      <td>lib_conc_unit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_derived_from</td>\n",
       "      <td>lib_conc_meth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_composed_of</td>\n",
       "      <td>phix_perc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rel_cont_id</td>\n",
       "      <td>checksum_filename</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>org_matter</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>org_matter_unit</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>org_nitro</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>org_nitro_unit</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>transmittance_unit</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sampleMetadata experimentRunMetadata\n",
       "0         neg_cont_type              lib_conc\n",
       "1         pos_cont_type         lib_conc_unit\n",
       "2   sample_derived_from         lib_conc_meth\n",
       "3    sample_composed_of             phix_perc\n",
       "4           rel_cont_id     checksum_filename\n",
       "..                  ...                   ...\n",
       "71           org_matter                   NaN\n",
       "72      org_matter_unit                   NaN\n",
       "73            org_nitro                   NaN\n",
       "74       org_nitro_unit                   NaN\n",
       "75   transmittance_unit                   NaN\n",
       "\n",
       "[76 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dropped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you are fine with leaving these columns out, proceed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping from data['sampleMetadata']: ['neg_cont_type', 'pos_cont_type', 'sample_derived_from', 'sample_composed_of', 'rel_cont_id', 'biological_rep_relation', 'verbatimLongitude', 'verbatimLatitude', 'verbatimCoordinateSystem', 'eventDurationValue', 'verbatimEventDate', 'verbatimEventTime', 'samp_store_method_additional', 'stationed_sample_dur', 'pump_flow_rate', 'pump_flow_rate_unit', 'prefilter_material', 'filter_diameter', 'filter_surface_area', 'prepped_samp_store_temp', 'prepped_samp_store_sol', 'prepped_samp_store_dur', 'prep_method_additional', 'date_ext', 'nucl_acid_ext_modify', 'dna_cleanup_0_1', 'dna_cleanup_method', 'concentration_method', 'ratioOfAbsorbance260_280', 'pool_dna_num', 'nucl_acid_ext_method_additional', 'samp_weather', 'elev', 'light_intensity', 'suspend_part_matter', 'tidal_stage', 'turbidity', 'water_current', 'solar_irradiance', 'wind_direction', 'wind_speed', 'diss_inorg_nitro', 'diss_inorg_nitro_unit', 'diss_org_carb', 'diss_org_carb_unit', 'diss_org_nitro', 'diss_org_nitro_unit', 'tot_diss_nitro', 'tot_diss_nitro_unit', 'tot_inorg_nitro', 'tot_inorg_nitro_unit', 'tot_nitro', 'tot_nitro_unit', 'tot_part_carb', 'tot_part_carb_unit', 'tot_org_carb', 'tot_org_carb_unit', 'tot_org_c_meth', 'tot_nitro_content', 'tot_nitro_content_unit', 'tot_nitro_cont_meth', 'tot_carb', 'tot_carb_unit', 'part_org_carb', 'part_org_carb_unit', 'part_org_nitro', 'part_org_nitro_unit', 'nitro', 'nitro_unit', 'org_carb', 'org_carb_unit', 'org_matter', 'org_matter_unit', 'org_nitro', 'org_nitro_unit', 'transmittance_unit']\n",
      "Dropping from data['experimentRunMetadata']: ['lib_conc', 'lib_conc_unit', 'lib_conc_meth', 'phix_perc', 'checksum_filename', 'checksum_filename2']\n"
     ]
    }
   ],
   "source": [
    "# Drops all-NA columns from 'sampleMetadata' and 'experimentRunMetadata'\n",
    "# based on the 'dropped' DataFrame.\n",
    "\n",
    "# These are the sheets Cell 19 might have put into the 'dropped' DataFrame\n",
    "sheets_to_clean = ['sampleMetadata', 'experimentRunMetadata']\n",
    "\n",
    "for sheet_name in sheets_to_clean:\n",
    "    # Check if 'dropped' has a column for this sheet AND if that column lists any actual columns to drop\n",
    "    if sheet_name in dropped.columns and not dropped[sheet_name].dropna().empty:\n",
    "        cols_to_drop = list(dropped[sheet_name].dropna())\n",
    "        \n",
    "        # Ensure the target DataFrame exists in 'data'\n",
    "        if sheet_name in data:\n",
    "            print(f\"Dropping from data['{sheet_name}']: {cols_to_drop}\")\n",
    "            data[sheet_name].drop(columns=cols_to_drop, inplace=True, errors='ignore')\n",
    "        # else: # Optional: warning if data[sheet_name] is missing, but Cell 19 should prevent this\n",
    "            # print(f\"Warning: data['{sheet_name}'] not found, cannot drop columns.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now lets drop NA rows of each analysisMetadata sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Identifying rows with empty 'values' column in analysisMetadata sheets:\n",
      "  In analysis sheet - Assay: 'ssu16sv4v5-emp', Run: 'gomecc4_16s_p1-2_v2024.10_241122':\n",
      "    Found 36 row(s) where 'values' is empty.\n",
      "      - Index: 30, Term Name: screen_contam_method\n",
      "      - Index: 31, Term Name: screen_geograph_method\n",
      "      - Index: 32, Term Name: screen_nontarget_method\n",
      "      - Index: 33, Term Name: screen_other\n",
      "      - Index: 36, Term Name: bioinfo_method_additional\n",
      "      - Index: 37, Term Name: output_read_count\n",
      "      - Index: 38, Term Name: output_otu_num\n",
      "      - Index: 39, Term Name: otu_num_tax_assigned\n",
      "      - Index: 40, Term Name: discard_untrimmed\n",
      "      - Index: 41, Term Name: qiime2_version\n",
      "      - Index: 42, Term Name: tourmaline_asv_method\n",
      "      - Index: 43, Term Name: dada2_trunc_len_f\n",
      "      - Index: 44, Term Name: dada2pe_trunc_len_r\n",
      "      - Index: 45, Term Name: dada2_trim_left_f\n",
      "      - Index: 46, Term Name: dada2pe_trim_left_r\n",
      "      - Index: 47, Term Name: dada2_max_ee_f\n",
      "      - Index: 48, Term Name: dada2pe_max_ee_r\n",
      "      - Index: 49, Term Name: dada2_trunc_q\n",
      "      - Index: 50, Term Name: dada2_pooling_method\n",
      "      - Index: 51, Term Name: dada2_chimera_method\n",
      "      - Index: 52, Term Name: dada2_min_fold_parent_over_abundance\n",
      "      - Index: 53, Term Name: dada2_n_reads_learn\n",
      "      - Index: 54, Term Name: deblur_trim_length\n",
      "      - Index: 55, Term Name: deblur_mean_error\n",
      "      - Index: 56, Term Name: deblur_indel_prob\n",
      "      - Index: 57, Term Name: deblur_indel_max\n",
      "      - Index: 58, Term Name: deblur_min_reads\n",
      "      - Index: 59, Term Name: deblur_min_size\n",
      "      - Index: 60, Term Name: repseqs_min_abundance\n",
      "      - Index: 61, Term Name: repseqs_min_length\n",
      "      - Index: 62, Term Name: repseqs_max_length\n",
      "      - Index: 63, Term Name: repseqs_min_prevalence\n",
      "      - Index: 64, Term Name: skl_confidence\n",
      "      - Index: 65, Term Name: min_consensus\n",
      "      - Index: 66, Term Name: tourmaline_classify_method\n",
      "      - Index: 67, Term Name: blca_confidence\n",
      "  In analysis sheet - Assay: 'ssu16sv4v5-emp', Run: 'gomecc4_16s_p3-6_v2024.10_241122':\n",
      "    Found 36 row(s) where 'values' is empty.\n",
      "      - Index: 30, Term Name: screen_contam_method\n",
      "      - Index: 31, Term Name: screen_geograph_method\n",
      "      - Index: 32, Term Name: screen_nontarget_method\n",
      "      - Index: 33, Term Name: screen_other\n",
      "      - Index: 36, Term Name: bioinfo_method_additional\n",
      "      - Index: 37, Term Name: output_read_count\n",
      "      - Index: 38, Term Name: output_otu_num\n",
      "      - Index: 39, Term Name: otu_num_tax_assigned\n",
      "      - Index: 40, Term Name: discard_untrimmed\n",
      "      - Index: 41, Term Name: qiime2_version\n",
      "      - Index: 42, Term Name: tourmaline_asv_method\n",
      "      - Index: 43, Term Name: dada2_trunc_len_f\n",
      "      - Index: 44, Term Name: dada2pe_trunc_len_r\n",
      "      - Index: 45, Term Name: dada2_trim_left_f\n",
      "      - Index: 46, Term Name: dada2pe_trim_left_r\n",
      "      - Index: 47, Term Name: dada2_max_ee_f\n",
      "      - Index: 48, Term Name: dada2pe_max_ee_r\n",
      "      - Index: 49, Term Name: dada2_trunc_q\n",
      "      - Index: 50, Term Name: dada2_pooling_method\n",
      "      - Index: 51, Term Name: dada2_chimera_method\n",
      "      - Index: 52, Term Name: dada2_min_fold_parent_over_abundance\n",
      "      - Index: 53, Term Name: dada2_n_reads_learn\n",
      "      - Index: 54, Term Name: deblur_trim_length\n",
      "      - Index: 55, Term Name: deblur_mean_error\n",
      "      - Index: 56, Term Name: deblur_indel_prob\n",
      "      - Index: 57, Term Name: deblur_indel_max\n",
      "      - Index: 58, Term Name: deblur_min_reads\n",
      "      - Index: 59, Term Name: deblur_min_size\n",
      "      - Index: 60, Term Name: repseqs_min_abundance\n",
      "      - Index: 61, Term Name: repseqs_min_length\n",
      "      - Index: 62, Term Name: repseqs_max_length\n",
      "      - Index: 63, Term Name: repseqs_min_prevalence\n",
      "      - Index: 64, Term Name: skl_confidence\n",
      "      - Index: 65, Term Name: min_consensus\n",
      "      - Index: 66, Term Name: tourmaline_classify_method\n",
      "      - Index: 67, Term Name: blca_confidence\n",
      "  In analysis sheet - Assay: 'ssu18sv9-emp', Run: 'gomecc4_18s_p1-6_v2024.10_241122':\n",
      "    Found 37 row(s) where 'values' is empty.\n",
      "      - Index: 5, Term Name: trim_param\n",
      "      - Index: 30, Term Name: screen_contam_method\n",
      "      - Index: 31, Term Name: screen_geograph_method\n",
      "      - Index: 32, Term Name: screen_nontarget_method\n",
      "      - Index: 33, Term Name: screen_other\n",
      "      - Index: 36, Term Name: bioinfo_method_additional\n",
      "      - Index: 37, Term Name: output_read_count\n",
      "      - Index: 38, Term Name: output_otu_num\n",
      "      - Index: 39, Term Name: otu_num_tax_assigned\n",
      "      - Index: 40, Term Name: discard_untrimmed\n",
      "      - Index: 41, Term Name: qiime2_version\n",
      "      - Index: 42, Term Name: tourmaline_asv_method\n",
      "      - Index: 43, Term Name: dada2_trunc_len_f\n",
      "      - Index: 44, Term Name: dada2pe_trunc_len_r\n",
      "      - Index: 45, Term Name: dada2_trim_left_f\n",
      "      - Index: 46, Term Name: dada2pe_trim_left_r\n",
      "      - Index: 47, Term Name: dada2_max_ee_f\n",
      "      - Index: 48, Term Name: dada2pe_max_ee_r\n",
      "      - Index: 49, Term Name: dada2_trunc_q\n",
      "      - Index: 50, Term Name: dada2_pooling_method\n",
      "      - Index: 51, Term Name: dada2_chimera_method\n",
      "      - Index: 52, Term Name: dada2_min_fold_parent_over_abundance\n",
      "      - Index: 53, Term Name: dada2_n_reads_learn\n",
      "      - Index: 54, Term Name: deblur_trim_length\n",
      "      - Index: 55, Term Name: deblur_mean_error\n",
      "      - Index: 56, Term Name: deblur_indel_prob\n",
      "      - Index: 57, Term Name: deblur_indel_max\n",
      "      - Index: 58, Term Name: deblur_min_reads\n",
      "      - Index: 59, Term Name: deblur_min_size\n",
      "      - Index: 60, Term Name: repseqs_min_abundance\n",
      "      - Index: 61, Term Name: repseqs_min_length\n",
      "      - Index: 62, Term Name: repseqs_max_length\n",
      "      - Index: 63, Term Name: repseqs_min_prevalence\n",
      "      - Index: 64, Term Name: skl_confidence\n",
      "      - Index: 65, Term Name: min_consensus\n",
      "      - Index: 66, Term Name: tourmaline_classify_method\n",
      "      - Index: 67, Term Name: blca_confidence\n",
      "\n",
      "Summary of rows with an empty 'values' column that can be dropped:\n",
      "  - Assay: 'ssu16sv4v5-emp', Run: 'gomecc4_16s_p1-2_v2024.10_241122' - 36 row(s) at Indices: [30, 31, 32, 33, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67]\n",
      "  - Assay: 'ssu16sv4v5-emp', Run: 'gomecc4_16s_p3-6_v2024.10_241122' - 36 row(s) at Indices: [30, 31, 32, 33, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67]\n",
      "  - Assay: 'ssu18sv9-emp', Run: 'gomecc4_18s_p1-6_v2024.10_241122' - 37 row(s) at Indices: [5, 30, 31, 32, 33, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67]\n",
      "Proceed to the next cell to remove these rows if desired.\n"
     ]
    }
   ],
   "source": [
    "# Cell to identify and report rows with empty 'values' in analysisMetadata sheets\n",
    "\n",
    "analysis_rows_to_drop_info = {} \n",
    "expected_value_col_name = 'values' # As per your Excel sheet (column D header)\n",
    "\n",
    "print(f\"Identifying rows with empty '{expected_value_col_name}' column in analysisMetadata sheets:\")\n",
    "if 'analysis_data_by_assay' in data and data['analysis_data_by_assay']:\n",
    "    for assay_name, analyses_dict in data['analysis_data_by_assay'].items():\n",
    "        if not isinstance(analyses_dict, dict): continue\n",
    "        for run_name, analysis_df in analyses_dict.items():\n",
    "            if not isinstance(analysis_df, pd.DataFrame) or analysis_df.empty: continue\n",
    "\n",
    "            # Ensure the 'values' column exists\n",
    "            if expected_value_col_name not in analysis_df.columns:\n",
    "                print(f\"  Warning: Column '{expected_value_col_name}' not found in Assay: '{assay_name}', Run: '{run_name}'. Skipping this sheet for row deletion.\")\n",
    "                continue\n",
    "            \n",
    "            # Identify rows where the 'values' column is NA\n",
    "            # This will correctly ignore the 'assay_name' and 'analysis_run_name' rows\n",
    "            # as their 'values' column is populated.\n",
    "            empty_values_mask = analysis_df[expected_value_col_name].isnull()\n",
    "            rows_to_drop_indices = analysis_df[empty_values_mask].index.tolist()\n",
    "\n",
    "            if rows_to_drop_indices:\n",
    "                print(f\"  In analysis sheet - Assay: '{assay_name}', Run: '{run_name}':\")\n",
    "                # To provide more context, let's also show the 'term_name' for rows to be dropped\n",
    "                terms_to_drop = analysis_df.loc[rows_to_drop_indices, 'term_name'].tolist() if 'term_name' in analysis_df.columns else [\"N/A\"]*len(rows_to_drop_indices)\n",
    "                print(f\"    Found {len(rows_to_drop_indices)} row(s) where '{expected_value_col_name}' is empty.\")\n",
    "                for i, idx in enumerate(rows_to_drop_indices):\n",
    "                    print(f\"      - Index: {idx}, Term Name: {terms_to_drop[i]}\")\n",
    "                analysis_rows_to_drop_info[(assay_name, run_name)] = rows_to_drop_indices\n",
    "\n",
    "if not analysis_rows_to_drop_info:\n",
    "    print(f\"\\nNo rows with an empty '{expected_value_col_name}' column found in any analysisMetadata sheet.\")\n",
    "else:\n",
    "    print(f\"\\nSummary of rows with an empty '{expected_value_col_name}' column that can be dropped:\")\n",
    "    for (assay, run), indices in analysis_rows_to_drop_info.items():\n",
    "        print(f\"  - Assay: '{assay}', Run: '{run}' - {len(indices)} row(s) at Indices: {indices}\")\n",
    "    print(\"Proceed to the next cell to remove these rows if desired.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If you are okay with those rows being deleted, continue:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Removing identified rows with empty 'values' from analysisMetadata sheets...\n",
      "  For Assay: 'ssu16sv4v5-emp', Run: 'gomecc4_16s_p1-2_v2024.10_241122': Removed 36 row(s).\n",
      "  For Assay: 'ssu16sv4v5-emp', Run: 'gomecc4_16s_p3-6_v2024.10_241122': Removed 36 row(s).\n",
      "  For Assay: 'ssu18sv9-emp', Run: 'gomecc4_18s_p1-6_v2024.10_241122': Removed 37 row(s).\n",
      "\n",
      "Finished removing rows with empty 'values' from analysisMetadata sheets.\n"
     ]
    }
   ],
   "source": [
    "# Cell to perform the deletion of rows with empty 'values' from analysisMetadata sheets\n",
    "\n",
    "if 'analysis_rows_to_drop_info' in locals() and analysis_rows_to_drop_info:\n",
    "    print(f\"\\nRemoving identified rows with empty 'values' from analysisMetadata sheets...\")\n",
    "    for (assay_name, run_name), indices_to_drop in analysis_rows_to_drop_info.items():\n",
    "        if indices_to_drop: \n",
    "            try:\n",
    "                original_count = len(data['analysis_data_by_assay'][assay_name][run_name])\n",
    "                data['analysis_data_by_assay'][assay_name][run_name].drop(index=indices_to_drop, inplace=True)\n",
    "                new_count = len(data['analysis_data_by_assay'][assay_name][run_name])\n",
    "                print(f\"  For Assay: '{assay_name}', Run: '{run_name}': Removed {original_count - new_count} row(s).\")\n",
    "            except Exception as e:\n",
    "                print(f\"  An error occurred while dropping rows for Assay: '{assay_name}', Run: '{run_name}': {e}\")\n",
    "    print(\"\\nFinished removing rows with empty 'values' from analysisMetadata sheets.\")\n",
    "    # del analysis_rows_to_drop_info # Optional: clear the info variable\n",
    "else:\n",
    "    print(\"\\nNo rows with empty 'values' were previously identified for deletion from analysisMetadata sheets, or 'analysis_rows_to_drop_info' not found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now lets check for columns or rows with SOME missing values. OBIS wants those deleted as well"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's check which columns have missing values in some of the rows. These should be filled in on the Excel sheet with the appropriate term ('not applicable', 'missing', or 'not collected'). Alternatively, you can drop the column if it is not needed for submission to OBIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identifies columns in 'sampleMetadata' and 'experimentRunMetadata'\n",
    "# that have *some* (but not all) missing NA values.\n",
    "\n",
    "some = pd.DataFrame()\n",
    "\n",
    "# Sheets to check for columns with *some* NAs.\n",
    "# We focus on the wide-format sheets where this is most informative.\n",
    "sheets_to_examine_for_some_na = ['sampleMetadata', 'experimentRunMetadata']\n",
    "\n",
    "for sheet_name in sheets_to_examine_for_some_na:\n",
    "    if sheet_name in data and not data[sheet_name].empty:\n",
    "        # Find columns that have at least one NA, but not all NAs\n",
    "        # .any() finds columns with at least one NA\n",
    "        # .all() finds columns with all NAs (we want to exclude these if they were already handled,\n",
    "        # but for \"some\", .any() is sufficient to identify columns that are not completely full)\n",
    "        cols_with_some_na = data[sheet_name].columns[data[sheet_name].isnull().any(axis=0)]\n",
    "        \n",
    "        res = pd.Series(cols_with_some_na.tolist(), name=sheet_name) # .tolist() for cleaner Series\n",
    "        some = pd.concat([some, res], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleMetadata</th>\n",
       "      <th>experimentRunMetadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ph_meth</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>carbonate_unit</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pco2_unit</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>samp_collect_notes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleMetadata experimentRunMetadata\n",
       "0             ph_meth                   NaN\n",
       "1      carbonate_unit                   NaN\n",
       "2           pco2_unit                   NaN\n",
       "3  samp_collect_notes                   NaN"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "some"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I'm going to drop all the columns with some missing data, as I don't need them for submission to OBIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "From data['sampleMetadata']: Dropped columns (had any NA): ['ph_meth', 'carbonate_unit', 'pco2_unit', 'samp_collect_notes']\n",
      "No columns dropped from data['experimentRunMetadata'] by 'any NA' rule.\n"
     ]
    }
   ],
   "source": [
    "# Drops any column from 'sampleMetadata' and 'experimentRunMetadata'\n",
    "# if that column contains ANY missing NA values. Lists the dropped columns.\n",
    "\n",
    "sheets_to_aggressively_clean = ['sampleMetadata', 'experimentRunMetadata']\n",
    "\n",
    "for sheet_name in sheets_to_aggressively_clean:\n",
    "    if sheet_name in data and not data[sheet_name].empty: # Check if DataFrame exists and is not empty\n",
    "        original_columns = data[sheet_name].columns.tolist() # Get column names before dropping\n",
    "        \n",
    "        data[sheet_name].dropna(axis=1, how='any', inplace=True)\n",
    "        \n",
    "        current_columns = data[sheet_name].columns.tolist() # Get column names after dropping\n",
    "        dropped_column_names = [col for col in original_columns if col not in current_columns]\n",
    "        \n",
    "        if dropped_column_names:\n",
    "            print(f\"From data['{sheet_name}']: Dropped columns (had any NA): {dropped_column_names}\")\n",
    "        else: # Optional: if no columns were dropped\n",
    "            print(f\"No columns dropped from data['{sheet_name}'] by 'any NA' rule.\")\n",
    "    elif sheet_name not in data:\n",
    "        print(f\"FYI: Sheet '{sheet_name}' not found in 'data'. No columns dropped.\")\n",
    "    else: # Sheet is in data but empty\n",
    "        print(f\"FYI: Sheet '{sheet_name}' is empty. No columns dropped.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load data dictionary Excel file \n",
    "This FAIRe NOAA Checklist Excel file also contains columns for mapping FAIRe fields to the appropriate Darwin Core terms which OBIS is expecting. Currently, we are only preparing an Occurrence core file and a DNA-derived extension file, with Event information in the Occurrence file. Future versions of this workflow will prepare an extendedMeasurementOrFact file as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dwc_data created. Occurrence mappings: 26, dnaDerived mappings: 26\n"
     ]
    }
   ],
   "source": [
    "dwc_data = {}\n",
    "checklist_df = pd.DataFrame()\n",
    "\n",
    "try:\n",
    "    # Load the 'checklist' sheet from the FAIRe NOAA Checklist Excel file\n",
    "    checklist_df = pd.read_excel(\n",
    "        params['FAIRe_NOAA_checklist'], # Ensure this param is set in your params cell\n",
    "        sheet_name='checklist',\n",
    "        na_values=[\"\"]\n",
    "    )\n",
    "except Exception as e:\n",
    "    print(f\"Error loading 'checklist' sheet: {e}\")\n",
    "\n",
    "# Define relevant column names from your checklist\n",
    "col_faire_term = 'term_name'\n",
    "col_output_spec = 'edna2obis_output_file'\n",
    "col_dwc_mapping = 'dwc_term'\n",
    "\n",
    "occurrence_maps = []\n",
    "dna_derived_maps = []\n",
    "\n",
    "# Process the checklist if it loaded successfully and has the required columns\n",
    "if not checklist_df.empty and all(col in checklist_df.columns for col in [col_faire_term, col_output_spec, col_dwc_mapping]):\n",
    "    for _, row in checklist_df.iterrows():\n",
    "        faire_term = row[col_faire_term]\n",
    "        output_file = str(row[col_output_spec]).lower() # Convert to string and lowercase\n",
    "        dwc_term = row[col_dwc_mapping]\n",
    "\n",
    "        # Add to lists if terms are valid\n",
    "        if pd.notna(faire_term) and pd.notna(dwc_term) and str(faire_term).strip() and str(dwc_term).strip():\n",
    "            if 'occurrence' in output_file:\n",
    "                occurrence_maps.append({'DwC_term': dwc_term, 'FAIRe_term': faire_term})\n",
    "            if 'dnaderived' in output_file:\n",
    "                dna_derived_maps.append({'DwC_term': dwc_term, 'FAIRe_term': faire_term})\n",
    "    \n",
    "    # Create DataFrames, using DwC_term as index\n",
    "    dwc_data['occurrence'] = pd.DataFrame(occurrence_maps).drop_duplicates().set_index('DwC_term') if occurrence_maps else \\\n",
    "                             pd.DataFrame(columns=['FAIRe_term']).set_index(pd.Index([], name='DwC_term'))\n",
    "    \n",
    "    dwc_data['dnaDerived'] = pd.DataFrame(dna_derived_maps).drop_duplicates().set_index('DwC_term') if dna_derived_maps else \\\n",
    "                             pd.DataFrame(columns=['FAIRe_term']).set_index(pd.Index([], name='DwC_term'))\n",
    "else:\n",
    "    # If checklist is empty or missing columns, create empty structures for dwc_data\n",
    "    if checklist_df.empty and 'FAIRe_NOAA_checklist_excel' in params: # Avoid double error message if file load failed\n",
    "        print(f\"Checklist DataFrame is empty or required columns are missing. Creating empty DwC mappings.\")\n",
    "    dwc_data['occurrence'] = pd.DataFrame(columns=['FAIRe_term']).set_index(pd.Index([], name='DwC_term'))\n",
    "    dwc_data['dnaDerived'] = pd.DataFrame(columns=['FAIRe_term']).set_index(pd.Index([], name='DwC_term'))\n",
    "\n",
    "# Print summary\n",
    "print(f\"dwc_data created. Occurrence mappings: {len(dwc_data['occurrence'])}, dnaDerived mappings: {len(dwc_data['dnaDerived'])}\")\n",
    "\n",
    "# The original notebook's Cell 26 was dwc_data['event'].head(). This will now be skipped or adapted.\n",
    "# Original Cell 27 was dwc_data['occurrence'].head().\n",
    "# Original Cell 28 was dwc_data['dna'].head() (now 'dnaDerived')."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FAIRe_term</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DwC_term</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>recordedBy</th>\n",
       "      <td>recordedBy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>datasetID</th>\n",
       "      <td>project_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>parentEventID</th>\n",
       "      <td>samp_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>materialSampleID</th>\n",
       "      <td>materialSampleID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLongitude</th>\n",
       "      <td>decimalLongitude</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLatitude</th>\n",
       "      <td>decimalLatitude</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geodeticDatum</th>\n",
       "      <td>verbatimSRS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locality</th>\n",
       "      <td>geo_loc_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventDate</th>\n",
       "      <td>eventDate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <td>samp_size</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <td>samp_size_unit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventID</th>\n",
       "      <td>lib_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>associatedSequences</th>\n",
       "      <td>associatedSequences</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kingdom</th>\n",
       "      <td>kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phylum</th>\n",
       "      <td>phylum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>class</th>\n",
       "      <td>class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>order</th>\n",
       "      <td>order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td>family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genus</th>\n",
       "      <td>genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scientificName</th>\n",
       "      <td>scientificName</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxonRank</th>\n",
       "      <td>taxonRank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxonID</th>\n",
       "      <td>taxonID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <td>verbatimIdentification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identificationRemarks</th>\n",
       "      <td>identificationRemarks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <td>minimumDepthInMeters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <td>maximumDepthInMeters</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    FAIRe_term\n",
       "DwC_term                                      \n",
       "recordedBy                          recordedBy\n",
       "datasetID                           project_id\n",
       "parentEventID                        samp_name\n",
       "materialSampleID              materialSampleID\n",
       "decimalLongitude              decimalLongitude\n",
       "decimalLatitude                decimalLatitude\n",
       "geodeticDatum                      verbatimSRS\n",
       "locality                          geo_loc_name\n",
       "eventDate                            eventDate\n",
       "sampleSizeValue                      samp_size\n",
       "sampleSizeUnit                  samp_size_unit\n",
       "eventID                                 lib_id\n",
       "associatedSequences        associatedSequences\n",
       "kingdom                                kingdom\n",
       "phylum                                  phylum\n",
       "class                                    class\n",
       "order                                    order\n",
       "family                                  family\n",
       "genus                                    genus\n",
       "scientificName                  scientificName\n",
       "taxonRank                            taxonRank\n",
       "taxonID                                taxonID\n",
       "verbatimIdentification  verbatimIdentification\n",
       "identificationRemarks    identificationRemarks\n",
       "minimumDepthInMeters      minimumDepthInMeters\n",
       "maximumDepthInMeters      maximumDepthInMeters"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print the mapping for the Occurrence Core\n",
    "# Keep in mind, some terms are hard coded later in this workflow, or are derived by more than one FAIRe term\n",
    "# This means there may be less fields listed below than what the Occurrence Core will have upon completion of edna2obis\n",
    "dwc_data['occurrence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FAIRe_term</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DwC_term</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>env_broad_scale</th>\n",
       "      <td>env_broad_scale</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env_local_scale</th>\n",
       "      <td>env_local_scale</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env_medium</th>\n",
       "      <td>env_medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samp_collect_device</th>\n",
       "      <td>samp_collect_device</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samp_mat_process</th>\n",
       "      <td>samp_mat_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>size_frac</th>\n",
       "      <td>size_frac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <td>samp_vol_we_dna_ext</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <td>nucl_acid_ext</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>concentration</th>\n",
       "      <td>concentration</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>concentrationUnit</th>\n",
       "      <td>concentration_unit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target_gene</th>\n",
       "      <td>target_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target_subfragment</th>\n",
       "      <td>target_subfragment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ampliconSize</th>\n",
       "      <td>ampliconSize</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <td>pcr_primer_forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <td>pcr_primer_reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <td>pcr_primer_name_forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <td>pcr_primer_name_reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <td>pcr_primer_reference_forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <td>nucl_acid_amp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pcr_cond</th>\n",
       "      <td>pcr_cond</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lib_layout</th>\n",
       "      <td>lib_layout</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventID</th>\n",
       "      <td>lib_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_sequence</th>\n",
       "      <td>dna_sequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>otu_class_appr</th>\n",
       "      <td>otu_clust_tool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>otu_db</th>\n",
       "      <td>otu_db</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <td>otu_seq_comp_appr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           FAIRe_term\n",
       "DwC_term                                             \n",
       "env_broad_scale                       env_broad_scale\n",
       "env_local_scale                       env_local_scale\n",
       "env_medium                                 env_medium\n",
       "samp_collect_device               samp_collect_device\n",
       "samp_mat_process                     samp_mat_process\n",
       "size_frac                                   size_frac\n",
       "samp_vol_we_dna_ext               samp_vol_we_dna_ext\n",
       "nucl_acid_ext                           nucl_acid_ext\n",
       "concentration                           concentration\n",
       "concentrationUnit                  concentration_unit\n",
       "target_gene                               target_gene\n",
       "target_subfragment                 target_subfragment\n",
       "ampliconSize                             ampliconSize\n",
       "pcr_primer_forward                 pcr_primer_forward\n",
       "pcr_primer_reverse                 pcr_primer_reverse\n",
       "pcr_primer_name_forward       pcr_primer_name_forward\n",
       "pcr_primer_name_reverse       pcr_primer_name_reverse\n",
       "pcr_primer_reference     pcr_primer_reference_forward\n",
       "nucl_acid_amp                           nucl_acid_amp\n",
       "pcr_cond                                     pcr_cond\n",
       "lib_layout                                 lib_layout\n",
       "eventID                                        lib_id\n",
       "DNA_sequence                             dna_sequence\n",
       "otu_class_appr                         otu_clust_tool\n",
       "otu_db                                         otu_db\n",
       "otu_seq_comp_appr                   otu_seq_comp_appr"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print the mapping for the DNA Derived Extension\n",
    "# Keep in mind, some terms are hard coded later in this workflow, or are derived by more than one FAIRe term\n",
    "# This means there may be less fields listed below than what the DNA Derived Extension will have upon completion of edna2obis\n",
    "dwc_data['dnaDerived']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert to Occurrence file\n",
    "In order to link the DNA-derived extension metadata to our OBIS occurrence records, we have to use the Occurrence core. For this data set, a `parentEvent` is a filtered water sample that was DNA extracted, a sequencing library from that DNA extraction is an `event`, and an `occurrence` is an ASV observed within a library. We will have an an occurence file and a DNA derived data file. Future versions will generate a measurements file.   \n",
    "**Define files**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build a combined Occurrence Core for all Analyses\n",
    "This code creates an Occurrence Core for each analysis in the submission, then combines them into one Occurence Core. This operation is complex, includes merging dataframes together, adding the missing fields which OBIS and GBIF expect, and parsing taxonomy raw data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚀 Starting data processing for 3 analysis run(s) to generate occurrence records.\n",
      "  ℹ️ Project Metadata Fetched: recordedBy='Luke Thompson', datasetID (from project_id)='noaa-aoml-gomecc4'\n",
      "\n",
      "Processing Analysis Run: gomecc4_18s_p1-6_v2024.10_241122\n",
      "  Successfully processed gomecc4_18s_p1-6_v2024.10_241122: Generated 147083 records.\n",
      "\n",
      "Processing Analysis Run: gomecc4_16s_p3-6_v2024.10_241122\n",
      "  Successfully processed gomecc4_16s_p3-6_v2024.10_241122: Generated 118671 records.\n",
      "\n",
      "Processing Analysis Run: gomecc4_16s_p1-2_v2024.10_241122\n",
      "  Successfully processed gomecc4_16s_p1-2_v2024.10_241122: Generated 46460 records.\n",
      "\n",
      "🏁 LOOP COMPLETED: Successful runs: 3, Failed runs: 0, Total DataFrames to combine: 3\n",
      "🔄 No duplicate occurrenceID records found to drop.\n",
      "\n",
      "💾 Combined occurrence file 'occurrence.csv' saved to '../processed-v3/occurrence.csv' with 312214 records.\n",
      "\n",
      "👀 Preview of final combined occurrence data (first 5 rows, selected columns):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>assay_name</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>geodeticDatum</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>locationID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>Luke Thompson</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.99999999990944, against reference database: PR2 v5.0.1; V9 1391f-1510r r...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>Luke Thompson</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.999067062720315, against reference database: PR2 v5.0.1; V9 1391f-1510r ...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>Luke Thompson</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.8911679667827849, against reference database: PR2 v5.0.1; V9 1391f-1510r...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>Luke Thompson</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.9996383713806553, against reference database: PR2 v5.0.1; V9 1391f-1510r...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>Luke Thompson</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.97066987594629, against reference database: PR2 v5.0.1; V9 1391f-1510r r...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               eventID  \\\n",
       "0  GOMECC18S_Plate4_53   \n",
       "1  GOMECC18S_Plate4_53   \n",
       "2  GOMECC18S_Plate4_53   \n",
       "3  GOMECC18S_Plate4_53   \n",
       "4  GOMECC18S_Plate4_53   \n",
       "\n",
       "                                               occurrenceID    assay_name  \\\n",
       "0  GOMECC18S_Plate4_53_occ_36aa75f9b28f5f831c2d631ba65c2bcb  ssu18sv9-emp   \n",
       "1  GOMECC18S_Plate4_53_occ_4e38e8ced9070952b314e1880bede1ca  ssu18sv9-emp   \n",
       "2  GOMECC18S_Plate4_53_occ_2a31e5c01634165da99e7381279baa75  ssu18sv9-emp   \n",
       "3  GOMECC18S_Plate4_53_occ_ecee60339b2fb88ea6d1c8d18054bed4  ssu18sv9-emp   \n",
       "4  GOMECC18S_Plate4_53_occ_fa1f1a97dd4ae7c826009186bad26384  ssu18sv9-emp   \n",
       "\n",
       "            parentEventID          datasetID     recordedBy  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  Luke Thompson   \n",
       "1  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  Luke Thompson   \n",
       "2  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  Luke Thompson   \n",
       "3  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  Luke Thompson   \n",
       "4  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  Luke Thompson   \n",
       "\n",
       "                                      locality decimalLatitude  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "\n",
       "  decimalLongitude geodeticDatum  \\\n",
       "0          -79.618         WGS84   \n",
       "1          -79.618         WGS84   \n",
       "2          -79.618         WGS84   \n",
       "3          -79.618         WGS84   \n",
       "4          -79.618         WGS84   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.99999999990944, against reference database: PR2 v5.0.1; V9 1391f-1510r r...   \n",
       "1  qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.999067062720315, against reference database: PR2 v5.0.1; V9 1391f-1510r ...   \n",
       "2  qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.8911679667827849, against reference database: PR2 v5.0.1; V9 1391f-1510r...   \n",
       "3  qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.9996383713806553, against reference database: PR2 v5.0.1; V9 1391f-1510r...   \n",
       "4  qiime2-2021.2; naive bayes classifier; scikit-learn 0.23.1, confidence: 0.97066987594629, against reference database: PR2 v5.0.1; V9 1391f-1510r r...   \n",
       "\n",
       "  locationID  \n",
       "0   27N_Sta1  \n",
       "1   27N_Sta1  \n",
       "2   27N_Sta1  \n",
       "3   27N_Sta1  \n",
       "4   27N_Sta1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# --- MAIN DATA PROCESSING LOOP --- Loop through each analysis run defined in params['datafiles']\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import warnings\n",
    "\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "print(f\"🚀 Starting data processing for {len(params['datafiles'])} analysis run(s) to generate occurrence records.\")\n",
    "\n",
    "# Define the desired final columns for occurrence.csv IN THE SPECIFIC ORDER REQUIRED\n",
    "DESIRED_OCCURRENCE_COLUMNS_IN_ORDER = [\n",
    "    'eventID', 'organismQuantity', 'assay_name', 'occurrenceID', 'verbatimIdentification',\n",
    "    'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species',  \n",
    "    'scientificName', 'scientificNameID', \n",
    "    'taxonRank', 'identificationRemarks',\n",
    "    'taxonID', 'basisOfRecord', 'nameAccordingTo', 'organismQuantityType',\n",
    "    'recordedBy', 'materialSampleID', 'sampleSizeValue', 'sampleSizeUnit',\n",
    "    'associatedSequences', 'locationID', 'eventDate', 'minimumDepthInMeters', 'maximumDepthInMeters',\n",
    "    'locality', 'decimalLatitude', 'decimalLongitude',\n",
    "    'geodeticDatum', 'parentEventID', 'datasetID', 'occurrenceStatus'\n",
    "]\n",
    "\n",
    "output_dir = \"../processed-v3/\" \n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "output_filename = \"occurrence.csv\"\n",
    "output_path = os.path.join(output_dir, output_filename)\n",
    "\n",
    "all_processed_occurrence_dfs = []\n",
    "successful_runs = 0\n",
    "failed_runs = 0\n",
    "\n",
    "# --- Fetch Project-Level Metadata ONCE ---\n",
    "project_recorded_by = \"recordedBy_NotProvided\"\n",
    "project_dataset_id = \"DatasetID_NotProvided\"\n",
    "\n",
    "def get_project_meta_value(project_meta_df, term_to_find, default_val=pd.NA):\n",
    "    if not all(col in project_meta_df.columns for col in ['term_name', 'project_level']):\n",
    "        return default_val\n",
    "    term_to_find_stripped = str(term_to_find).strip()\n",
    "    match = project_meta_df[project_meta_df['term_name'].astype(str).str.strip().str.lower() == term_to_find_stripped.lower()]\n",
    "    if not match.empty:\n",
    "        value = match['project_level'].iloc[0]\n",
    "        if pd.notna(value):\n",
    "            return str(value).strip()\n",
    "    return default_val\n",
    "\n",
    "if 'projectMetadata' in data and not data['projectMetadata'].empty:\n",
    "    project_meta_df = data['projectMetadata']\n",
    "    project_recorded_by = get_project_meta_value(project_meta_df, 'recordedBy', project_recorded_by)\n",
    "    project_dataset_id = get_project_meta_value(project_meta_df, 'project_id', project_dataset_id)\n",
    "    print(f\"  ℹ️ Project Metadata Fetched: recordedBy='{project_recorded_by}', datasetID (from project_id)='{project_dataset_id}'\")\n",
    "else:\n",
    "    print(\"  ⚠️ Warning: projectMetadata is empty or not found. 'recordedBy' and 'datasetID' will use default placeholder values.\")\n",
    "\n",
    "# --- MAIN DATA PROCESSING LOOP ---\n",
    "for analysis_run_name, file_paths_dict in params['datafiles'].items():\n",
    "    print(f\"\\nProcessing Analysis Run: {analysis_run_name}\")\n",
    "    try:\n",
    "        # --- STEP 1: Load and Prepare Raw Taxonomy and Abundance Data ---\n",
    "        if not (analysis_run_name in raw_data_tables and\n",
    "                'taxonomy' in raw_data_tables[analysis_run_name] and not raw_data_tables[analysis_run_name]['taxonomy'].empty and\n",
    "                'occurrence' in raw_data_tables[analysis_run_name] and not raw_data_tables[analysis_run_name]['occurrence'].empty):\n",
    "            print(f\"  Skipping {analysis_run_name}: Raw taxonomy or occurrence data is missing or empty.\")\n",
    "            failed_runs += 1\n",
    "            continue\n",
    "\n",
    "        current_tax_df_raw = raw_data_tables[analysis_run_name]['taxonomy'].copy()\n",
    "        current_abundance_df_raw = raw_data_tables[analysis_run_name]['occurrence'].copy()\n",
    "\n",
    "        featureid_col_tax = current_tax_df_raw.columns[0]\n",
    "        current_tax_df_raw.rename(columns={featureid_col_tax: 'featureid'}, inplace=True)\n",
    "        featureid_col_abun = current_abundance_df_raw.columns[0] \n",
    "        current_abundance_df_raw.rename(columns={featureid_col_abun: 'featureid'}, inplace=True)\n",
    "\n",
    "        sequence_col_dwc = 'DNA_sequence'\n",
    "        sequence_col_input = 'sequence' \n",
    "        confidence_col_original_case = 'Confidence' \n",
    "        \n",
    "        if 'dna_sequence' in current_tax_df_raw.columns and sequence_col_input not in current_tax_df_raw.columns:\n",
    "             current_tax_df_raw.rename(columns={'dna_sequence': sequence_col_input}, inplace=True)\n",
    "        elif sequence_col_input not in current_tax_df_raw.columns:\n",
    "            current_tax_df_raw[sequence_col_input] = pd.NA\n",
    "            \n",
    "        # Determine the source for DwC 'verbatimIdentification'.\n",
    "        # User wants to use the column named 'taxonomy' from the input asvTaxaFeatures file.\n",
    "        # NOTE: we are basically using the taxonomy string instead of verbatimIdentification, for the Occurrence Core's verbatimIdentifiction.\n",
    "        # Don't be confused- this is to help with taxonomic matching. The 'taxonomy' field is the same as verbatimIdentification, except it just is more machine readable. \n",
    "        \n",
    "        # This is the name of the FAIRe column in current_tax_df_raw that holds the string to be used.\n",
    "        # It will be renamed to 'verbatimIdentification' in current_tax_df_processed.\n",
    "        source_column_for_verbatim_id = 'taxonomy' \n",
    "\n",
    "        if source_column_for_verbatim_id in current_tax_df_raw.columns:\n",
    "            verbatim_id_source_col = source_column_for_verbatim_id \n",
    "            # print(f\"  Using column '{verbatim_id_source_col}' from input as source for DwC 'verbatimIdentification'.\") # Kept for debug, can be commented\n",
    "        else:\n",
    "            # If the explicitly desired 'taxonomy' column is MISSING from the input file.\n",
    "            print(f\"  CRITICAL WARNING: Column '{source_column_for_verbatim_id}' not found in input taxonomy table for '{analysis_run_name}'. DwC 'verbatimIdentification' will use a placeholder.\")\n",
    "            current_tax_df_raw['verbatimIdentification_placeholder'] = f\"Data from '{source_column_for_verbatim_id}' column not available in source\"\n",
    "            verbatim_id_source_col = 'verbatimIdentification_placeholder'\n",
    "            # This placeholder column will be picked up by tax_cols_to_keep and then renamed.                               \n",
    "        \n",
    "        tax_cols_to_keep = ['featureid', sequence_col_input, verbatim_id_source_col]\n",
    "        if confidence_col_original_case in current_tax_df_raw.columns:\n",
    "            tax_cols_to_keep.append(confidence_col_original_case)\n",
    "        else:\n",
    "            current_tax_df_raw[confidence_col_original_case] = pd.NA \n",
    "            tax_cols_to_keep.append(confidence_col_original_case)\n",
    "\n",
    "        current_tax_df_processed = current_tax_df_raw[[col for col in tax_cols_to_keep if col in current_tax_df_raw.columns]].copy()\n",
    "        current_tax_df_processed.rename(columns={verbatim_id_source_col: 'verbatimIdentification', \n",
    "                                                 sequence_col_input: sequence_col_dwc}, inplace=True)\n",
    "\n",
    "        # --- STEP 2: Melt Abundance and Merge with Taxonomy ---\n",
    "        current_assay_occ_melted = pd.melt(\n",
    "            current_abundance_df_raw, id_vars=['featureid'],\n",
    "            var_name='samp_name', value_name='organismQuantity'\n",
    "        )\n",
    "        current_assay_occ_melted = current_assay_occ_melted[current_assay_occ_melted['organismQuantity'] > 0.0]\n",
    "        \n",
    "        current_assay_occurrence_intermediate_df = pd.merge(\n",
    "            current_assay_occ_melted, current_tax_df_processed,\n",
    "            on='featureid', how='left'\n",
    "        )\n",
    "\n",
    "        # --- STEP 3: Initialize ALL DwC Fields from DESIRED_OCCURRENCE_COLUMNS_IN_ORDER ---\n",
    "        for col in DESIRED_OCCURRENCE_COLUMNS_IN_ORDER:\n",
    "            if col not in current_assay_occurrence_intermediate_df.columns:\n",
    "                 current_assay_occurrence_intermediate_df[col] = pd.NA\n",
    "        \n",
    "        # Set values for fields that are constructed or have fixed values for this process\n",
    "        current_assay_occurrence_intermediate_df['taxonID'] = 'ASV:' + current_assay_occurrence_intermediate_df['featureid'].astype(str)\n",
    "        current_assay_occurrence_intermediate_df['organismQuantityType'] = 'DNA sequence reads'\n",
    "        current_assay_occurrence_intermediate_df['occurrenceStatus'] = 'present'\n",
    "        current_assay_occurrence_intermediate_df['basisOfRecord'] = 'MaterialSample'\n",
    "        current_assay_occurrence_intermediate_df['nameAccordingTo'] = 'Original Classification; WoRMS/GBIF (pending further matching)'\n",
    "        \n",
    "        # --- STEP 4: Assign Project-Level Metadata ---\n",
    "        current_assay_occurrence_intermediate_df['recordedBy'] = project_recorded_by\n",
    "        current_assay_occurrence_intermediate_df['datasetID'] = project_dataset_id\n",
    "\n",
    "        # --- STEP 5: Merge `sampleMetadata` and map FAIRe terms to DwC terms ---\n",
    "        if 'sampleMetadata' in data and not data['sampleMetadata'].empty:\n",
    "            sm_df_to_merge = data['sampleMetadata'].copy()\n",
    "            sm_df_to_merge['samp_name'] = sm_df_to_merge['samp_name'].astype(str).str.strip()\n",
    "            current_assay_occurrence_intermediate_df['samp_name'] = current_assay_occurrence_intermediate_df['samp_name'].astype(str).str.strip()\n",
    "\n",
    "            current_assay_occurrence_intermediate_df = pd.merge(\n",
    "                current_assay_occurrence_intermediate_df, sm_df_to_merge,\n",
    "                on='samp_name', how='left', suffixes=('', '_sm') \n",
    "            )\n",
    "            \n",
    "            # These columns have specific assignment logic/calculation later or are core IDs\n",
    "            # They are filled by mapping only if currently NA. They all have different, specific logic to construct.\n",
    "            cols_with_specific_logic_or_origin = [\n",
    "                'datasetID', 'recordedBy', 'eventID', 'occurrenceID', 'taxonID', \n",
    "                'organismQuantityType', 'occurrenceStatus', 'basisOfRecord', 'nameAccordingTo',\n",
    "                'parentEventID', 'associatedSequences', \n",
    "                'sampleSizeValue', 'sampleSizeUnit', 'identificationRemarks'\n",
    "            ]\n",
    "\n",
    "            # Populate DwC columns using the dwc_data['occurrence'] mapping\n",
    "            for dwc_col_target, faire_row in dwc_data['occurrence'].iterrows():\n",
    "                if dwc_col_target not in DESIRED_OCCURRENCE_COLUMNS_IN_ORDER: # Ensure we only care about desired output columns\n",
    "                    continue \n",
    "\n",
    "                faire_col_source_original = str(faire_row['FAIRe_term']).strip()\n",
    "                source_col_in_df = None\n",
    "                \n",
    "                # Check for the column with _sm suffix first (if a clash occurred during merge with sampleMetadata)\n",
    "                if faire_col_source_original + '_sm' in current_assay_occurrence_intermediate_df.columns:\n",
    "                    source_col_in_df = faire_col_source_original + '_sm'\n",
    "                # Else, check for the original FAIRe term name (if no clash)\n",
    "                elif faire_col_source_original in current_assay_occurrence_intermediate_df.columns:\n",
    "                    source_col_in_df = faire_col_source_original\n",
    "                \n",
    "                if source_col_in_df:\n",
    "                    # If the DwC target column has specific logic for its creation or is a core ID,\n",
    "                    # only fill it from sampleMetadata if it's currently NA.\n",
    "                    if dwc_col_target in cols_with_specific_logic_or_origin:\n",
    "                        current_assay_occurrence_intermediate_df[dwc_col_target] = current_assay_occurrence_intermediate_df[dwc_col_target].fillna(current_assay_occurrence_intermediate_df[source_col_in_df])\n",
    "                    else: \n",
    "                        # For other \"standard\" DwC columns (like locality, lat, lon, geodeticDatum, etc.),\n",
    "                        # directly assign from the source FAIRe column.\n",
    "                        current_assay_occurrence_intermediate_df[dwc_col_target] = current_assay_occurrence_intermediate_df[source_col_in_df]\n",
    "                elif dwc_col_target in ['locality', 'decimalLatitude', 'decimalLongitude', 'geodeticDatum', 'eventDate']: # Only print diagnostic for key terms if mapping is missing in checklist\n",
    "                     print(f\"  DIAGNOSTIC: For DwC term '{dwc_col_target}', its mapped FAIRe term '{faire_col_source_original}' (from checklist) was NOT found as a column in the merged sample data (checked as '{faire_col_source_original}' and '{faire_col_source_original}_sm'). The DwC column '{dwc_col_target}' will likely be empty if not populated by other means.\")\n",
    "        else:\n",
    "            print(f\"  Warning: 'sampleMetadata' is empty or not found. Cannot merge for DwC term population for run {analysis_run_name}.\")\n",
    "\n",
    "        # Construct 'locationID' \n",
    "        line_id_col_sm = 'line_id_sm' if 'line_id_sm' in current_assay_occurrence_intermediate_df.columns else 'line_id'\n",
    "        station_id_col_sm = 'station_id_sm' if 'station_id_sm' in current_assay_occurrence_intermediate_df.columns else 'station_id'\n",
    "\n",
    "        if line_id_col_sm in current_assay_occurrence_intermediate_df.columns and \\\n",
    "           station_id_col_sm in current_assay_occurrence_intermediate_df.columns:\n",
    "            line_ids = current_assay_occurrence_intermediate_df[line_id_col_sm].astype(str).fillna('NoLineID')\n",
    "            station_ids = current_assay_occurrence_intermediate_df[station_id_col_sm].astype(str).fillna('NoStationID')\n",
    "            current_assay_occurrence_intermediate_df['locationID'] = line_ids + \"_\" + station_ids\n",
    "            # print(f\"    Constructed 'locationID'.\") # Reduced verbosity\n",
    "        else:\n",
    "            print(f\"    Warning: Could not construct 'locationID' using '{line_id_col_sm}' or '{station_id_col_sm}'.\")\n",
    "            if 'locationID' not in current_assay_occurrence_intermediate_df.columns or current_assay_occurrence_intermediate_df['locationID'].isna().all():\n",
    "                 current_assay_occurrence_intermediate_df['locationID'] = \"LocationID_NotAvailable\"\n",
    "\n",
    "\n",
    "        # --- STEP 6: Merge `experimentRunMetadata` & Define `eventID`, `associatedSequences` ---\n",
    "        assay_name_for_current_run = next((an_key for an_key, runs_dict_val in data.get('analysis_data_by_assay', {}).items() if isinstance(runs_dict_val, dict) and analysis_run_name in runs_dict_val), None)\n",
    "        \n",
    "        current_assay_occurrence_intermediate_df['assay_name'] = assay_name_for_current_run # <-- ADD THIS LINE\n",
    "\n",
    "        if not assay_name_for_current_run:\n",
    "            print(f\"    ERROR: Could not determine assay_name for '{analysis_run_name}'.\")\n",
    "            current_assay_occurrence_intermediate_df['eventID'] = current_assay_occurrence_intermediate_df['eventID'].fillna(f\"ERROR_eventID_for_{analysis_run_name}\")\n",
    "            # Also ensure a placeholder for assay_name if it couldn't be found, though it should be an error condition\n",
    "            current_assay_occurrence_intermediate_df['assay_name'] = current_assay_occurrence_intermediate_df['assay_name'].fillna(f\"UNKNOWN_ASSAY_FOR_{analysis_run_name}\")\n",
    "        elif 'experimentRunMetadata' in data and not data['experimentRunMetadata'].empty:\n",
    "            erm_df = data['experimentRunMetadata'].copy()\n",
    "            erm_df['samp_name'] = erm_df['samp_name'].astype(str).str.strip()\n",
    "            erm_df_assay_specific = erm_df[erm_df['assay_name'].astype(str).str.strip() == str(assay_name_for_current_run).strip()]\n",
    "\n",
    "            if not erm_df_assay_specific.empty:\n",
    "                faire_lib_id_col = str(dwc_data['occurrence'].loc['eventID', 'FAIRe_term']).strip() if 'eventID' in dwc_data['occurrence'].index else 'lib_id'\n",
    "                faire_assoc_seq_col = str(dwc_data['occurrence'].loc['associatedSequences', 'FAIRe_term']).strip() if 'associatedSequences' in dwc_data['occurrence'].index else 'associatedSequences'\n",
    "\n",
    "                cols_to_select_from_erm = {'samp_name'}\n",
    "                if faire_lib_id_col in erm_df_assay_specific.columns: cols_to_select_from_erm.add(faire_lib_id_col)\n",
    "                if faire_assoc_seq_col in erm_df_assay_specific.columns: cols_to_select_from_erm.add(faire_assoc_seq_col)\n",
    "                \n",
    "                erm_to_merge = erm_df_assay_specific[list(cols_to_select_from_erm)].drop_duplicates(subset=['samp_name']).copy()\n",
    "                \n",
    "                current_assay_occurrence_intermediate_df = pd.merge(\n",
    "                    current_assay_occurrence_intermediate_df, erm_to_merge,\n",
    "                    on='samp_name', how='left', suffixes=('', '_erm')\n",
    "                )\n",
    "                \n",
    "                source_lib_id_col_actual = faire_lib_id_col + '_erm' if faire_lib_id_col + '_erm' in current_assay_occurrence_intermediate_df.columns else faire_lib_id_col\n",
    "                if source_lib_id_col_actual in current_assay_occurrence_intermediate_df.columns:\n",
    "                    current_assay_occurrence_intermediate_df['eventID'] = current_assay_occurrence_intermediate_df['eventID'].fillna(current_assay_occurrence_intermediate_df[source_lib_id_col_actual])\n",
    "                \n",
    "                source_assoc_seq_col_actual = faire_assoc_seq_col + '_erm' if faire_assoc_seq_col + '_erm' in current_assay_occurrence_intermediate_df.columns else faire_assoc_seq_col\n",
    "                if source_assoc_seq_col_actual in current_assay_occurrence_intermediate_df.columns:\n",
    "                     current_assay_occurrence_intermediate_df['associatedSequences'] = current_assay_occurrence_intermediate_df['associatedSequences'].fillna(current_assay_occurrence_intermediate_df[source_assoc_seq_col_actual])\n",
    "            else:\n",
    "                current_assay_occurrence_intermediate_df['eventID'] = current_assay_occurrence_intermediate_df['eventID'].fillna(f\"NoExpMeta_eventID_for_{analysis_run_name}\")\n",
    "        \n",
    "        current_assay_occurrence_intermediate_df['eventID'] = current_assay_occurrence_intermediate_df['eventID'].astype(str)\n",
    "        \n",
    "        # --- STEP 7: Construct `occurrenceID` ---\n",
    "        current_assay_occurrence_intermediate_df['occurrenceID'] = \\\n",
    "            current_assay_occurrence_intermediate_df['eventID'] + \\\n",
    "            '_occ_' + \\\n",
    "            current_assay_occurrence_intermediate_df['featureid'].astype(str)\n",
    "\n",
    "        # --- STEP 8: `identificationRemarks`, `sampleSizeValue`/`Unit`, `parentEventID` ---\n",
    "        otu_seq_comp_appr_str = \"Unknown sequence comparison approach\"\n",
    "        taxa_class_method_str = \"\"\n",
    "        taxa_ref_db_str = \"Unknown reference DB\"\n",
    "\n",
    "        if assay_name_for_current_run and analysis_run_name in data.get('analysis_data_by_assay', {}).get(assay_name_for_current_run, {}):\n",
    "            analysis_meta_df_for_run = data['analysis_data_by_assay'][assay_name_for_current_run][analysis_run_name]\n",
    "            if 'term_name' in analysis_meta_df_for_run.columns and 'values' in analysis_meta_df_for_run.columns:\n",
    "                def get_analysis_meta(term, df, default):\n",
    "                    val_series = df[df['term_name'].astype(str).str.strip() == term]['values']\n",
    "                    return str(val_series.iloc[0]).strip() if not val_series.empty and pd.notna(val_series.iloc[0]) else default\n",
    "                otu_seq_comp_appr_str = get_analysis_meta('otu_seq_comp_appr', analysis_meta_df_for_run, otu_seq_comp_appr_str)\n",
    "                taxa_class_method_str = get_analysis_meta('taxa_class_method', analysis_meta_df_for_run, taxa_class_method_str)\n",
    "                taxa_ref_db_str = get_analysis_meta('otu_db', analysis_meta_df_for_run, taxa_ref_db_str)\n",
    "        \n",
    "        confidence_value_series = pd.Series([\"unknown confidence\"] * len(current_assay_occurrence_intermediate_df), index=current_assay_occurrence_intermediate_df.index, dtype=object)\n",
    "        if confidence_col_original_case in current_assay_occurrence_intermediate_df.columns:\n",
    "            confidence_value_series = current_assay_occurrence_intermediate_df[confidence_col_original_case].astype(str).fillna(\"unknown confidence\")\n",
    "        \n",
    "        current_assay_occurrence_intermediate_df['identificationRemarks'] = \\\n",
    "            f\"{otu_seq_comp_appr_str}, confidence: \" + \\\n",
    "            confidence_value_series + \\\n",
    "            f\", against reference database: {taxa_ref_db_str}\"\n",
    "        \n",
    "        if 'eventID' in current_assay_occurrence_intermediate_df.columns and not current_assay_occurrence_intermediate_df['eventID'].isna().all():\n",
    "            sample_size_map = current_assay_occurrence_intermediate_df.groupby('eventID')['organismQuantity'].sum().to_dict()\n",
    "            current_assay_occurrence_intermediate_df['sampleSizeValue'] = current_assay_occurrence_intermediate_df['eventID'].map(sample_size_map)\n",
    "        current_assay_occurrence_intermediate_df['sampleSizeUnit'] = 'DNA sequence reads'\n",
    "        \n",
    "        if 'parentEventID' in dwc_data['occurrence'].index:\n",
    "            faire_parent_event_id_col_name = str(dwc_data['occurrence'].loc['parentEventID','FAIRe_term']).strip() \n",
    "            actual_parent_event_col_name = faire_parent_event_id_col_name + \"_sm\" if faire_parent_event_id_col_name + \"_sm\" in current_assay_occurrence_intermediate_df.columns else faire_parent_event_id_col_name\n",
    "            if actual_parent_event_col_name in current_assay_occurrence_intermediate_df.columns :\n",
    "                 current_assay_occurrence_intermediate_df['parentEventID'] = current_assay_occurrence_intermediate_df['parentEventID'].fillna(current_assay_occurrence_intermediate_df[actual_parent_event_col_name])\n",
    "\n",
    "        # --- STEP 9: Final Column Selection and Order for this assay's DataFrame ---\n",
    "        for col_final_desired in DESIRED_OCCURRENCE_COLUMNS_IN_ORDER:\n",
    "            if col_final_desired not in current_assay_occurrence_intermediate_df.columns:\n",
    "                current_assay_occurrence_intermediate_df[col_final_desired] = pd.NA\n",
    "        \n",
    "        current_assay_occurrence_final_df = current_assay_occurrence_intermediate_df[DESIRED_OCCURRENCE_COLUMNS_IN_ORDER].copy()\n",
    "        \n",
    "        all_processed_occurrence_dfs.append(current_assay_occurrence_final_df)\n",
    "        successful_runs += 1\n",
    "        print(f\"  Successfully processed {analysis_run_name}: Generated {len(current_assay_occurrence_final_df)} records.\")\n",
    "\n",
    "    except Exception as e:\n",
    "        import traceback\n",
    "        print(f\"  ❌ Error processing {analysis_run_name}: {str(e)}\")\n",
    "        print(f\"  Traceback for {analysis_run_name}: {traceback.format_exc()}\")\n",
    "        failed_runs += 1\n",
    "\n",
    "# --- POST-LOOP CONCATENATION & FINALIZATION ---\n",
    "print(f\"\\n🏁 LOOP COMPLETED: Successful runs: {successful_runs}, Failed runs: {failed_runs}, Total DataFrames to combine: {len(all_processed_occurrence_dfs)}\")\n",
    "\n",
    "if all_processed_occurrence_dfs:\n",
    "    occ_all_final_combined = pd.concat(all_processed_occurrence_dfs, ignore_index=True, sort=False)\n",
    "    \n",
    "    for col_final_desired in DESIRED_OCCURRENCE_COLUMNS_IN_ORDER:\n",
    "        if col_final_desired not in occ_all_final_combined.columns:\n",
    "            occ_all_final_combined[col_final_desired] = pd.NA\n",
    "    \n",
    "    occ_all_final_output = occ_all_final_combined.reindex(columns=DESIRED_OCCURRENCE_COLUMNS_IN_ORDER)\n",
    "    \n",
    "    original_rows_before_dedup = len(occ_all_final_output)\n",
    "    if 'occurrenceID' in occ_all_final_output.columns and not occ_all_final_output['occurrenceID'].isna().all():\n",
    "        num_duplicates = occ_all_final_output.duplicated(subset=['occurrenceID']).sum()\n",
    "        if num_duplicates > 0:\n",
    "            occ_all_final_output.drop_duplicates(subset=['occurrenceID'], keep='first', inplace=True)\n",
    "            print(f\"🔄 Dropped {num_duplicates} duplicate occurrenceID records. Final rows: {len(occ_all_final_output)}.\")\n",
    "        else:\n",
    "            print(\"🔄 No duplicate occurrenceID records found to drop.\")\n",
    "    else:\n",
    "        print(\"  ⚠️ WARNING: 'occurrenceID' column not found or is all NA. Cannot effectively drop duplicates based on it.\")\n",
    "\n",
    "    try:\n",
    "        occ_all_final_output.to_csv(output_path, index=False, na_rep='') \n",
    "        print(f\"\\n💾 Combined occurrence file '{output_filename}' saved to '{output_path}' with {len(occ_all_final_output)} records.\")\n",
    "        # final_cols_list_written = occ_all_final_output.columns.tolist() # Removed for brevity\n",
    "        # print(f\"📋 Final columns written ({len(final_cols_list_written)} total), in order: {final_cols_list_written}\")\n",
    "\n",
    "        print(f\"\\n👀 Preview of final combined occurrence data (first 5 rows, selected columns):\")\n",
    "        preview_cols_subset = ['eventID', 'occurrenceID', 'assay_name', 'parentEventID', 'datasetID', 'recordedBy', \n",
    "                               'locality', 'decimalLatitude', 'decimalLongitude', 'geodeticDatum', \n",
    "                               'identificationRemarks', 'locationID']\n",
    "        preview_cols_to_show = [col for col in preview_cols_subset if col in occ_all_final_output.columns]\n",
    "        display(occ_all_final_output[preview_cols_to_show].head())\n",
    "    except Exception as e:\n",
    "        print(f\"  ❌ Error saving combined occurrence file: {str(e)}\")\n",
    "else:\n",
    "    print(f\"❌ No data to combine - all analysis runs may have failed or yielded no occurrence records.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NOTE:\n",
    "The Occurrence Core at this step (before taxonomic assignment through WoRMS or GBIF), contains an assay_name column. This will be removed from the final Occurrence Core (after taxonomic assignment) but it is used by the taxonomic assignment code to know which assay's data you want to remove the 'species' rank from consideration. This is because some assays, like 16S for example, return non-usable assignments at species level, while, for example, 18S species assignments ARE useful."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DEBUGGING ONLY: Create Smaller Subset for Faster Taxonomic Matching\n",
    "\n",
    "**Run this cell ONLY if you want to test taxonomic matching on a small subset.**\n",
    "This cell will take the `occ_df_for_processing_step` (which currently holds the full dataset from Cell 33), create a small subset from it, and then **overwrite** `occ_df_for_processing_step` with this smaller subset.\n",
    "If you want to run taxonomic matching on the **full dataset, simply DO NOT RUN THIS CELL.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to load full occurrence data from: ../processed-v3/occurrence.csv\n",
      "Successfully loaded ../processed-v3/occurrence.csv with 312214 records.\n",
      "  Available assays in occurrence.csv: ['ssu18sv9-emp' 'ssu16sv4v5-emp']\n",
      "  Selected 10 rows for assay 'ssu18sv9-emp' in the subset.\n",
      "  Selected 10 rows for assay 'ssu16sv4v5-emp' in the subset.\n",
      "Successfully created subset with 20 total rows for taxonomic matching.\n",
      "  Unique assay_names in subset: ['ssu18sv9-emp' 'ssu16sv4v5-emp']\n",
      "\n",
      "Proceeding with the SUBSET (20 rows) for taxonomic matching.\n"
     ]
    }
   ],
   "source": [
    "# Create a small subset of occurrence.csv for faster testing of taxonomic matching\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Determine the correct output directory path\n",
    "current_output_dir = None\n",
    "if 'output_dir' in globals() and isinstance(output_dir, str):\n",
    "    current_output_dir = output_dir # From a previous run of cell 28\n",
    "elif 'params' in globals() and 'output_dir' in params and isinstance(params['output_dir'], str):\n",
    "    current_output_dir = params['output_dir'] # From cell 8\n",
    "else:\n",
    "    current_output_dir = \"../processed-v3/\" # Fallback\n",
    "    print(f\"Warning: 'output_dir' not found in globals or params, using default: {current_output_dir}\")\n",
    "\n",
    "full_occurrence_csv_path = os.path.join(current_output_dir, \"occurrence.csv\")\n",
    "occ_df_subset_for_matching = pd.DataFrame() # Initialize\n",
    "full_occ_df = pd.DataFrame() # Initialize\n",
    "\n",
    "N_ROWS_PER_ASSAY_SUBSET = 10 # Number of rows to take per assay for the subset\n",
    "print(f\"Attempting to load full occurrence data from: {full_occurrence_csv_path}\")\n",
    "if os.path.exists(full_occurrence_csv_path):\n",
    "    try:\n",
    "        full_occ_df = pd.read_csv(full_occurrence_csv_path, low_memory=False, dtype=str)\n",
    "        if 'organismQuantity' in full_occ_df.columns:\n",
    "            full_occ_df['organismQuantity'] = pd.to_numeric(full_occ_df['organismQuantity'], errors='coerce').fillna(0).astype(float)\n",
    "        print(f\"Successfully loaded {full_occurrence_csv_path} with {len(full_occ_df)} records.\")\n",
    "\n",
    "        if not full_occ_df.empty and 'assay_name' in full_occ_df.columns and 'verbatimIdentification' in full_occ_df.columns:\n",
    "            available_assays = full_occ_df['assay_name'].dropna().unique()\n",
    "            print(f\"  Available assays in occurrence.csv: {available_assays}\")\n",
    "            \n",
    "            subset_dfs = []\n",
    "            for assay in available_assays:\n",
    "                assay_subset = full_occ_df[full_occ_df['assay_name'] == assay].head(N_ROWS_PER_ASSAY_SUBSET)\n",
    "                if not assay_subset.empty:\n",
    "                    subset_dfs.append(assay_subset)\n",
    "                    print(f\"  Selected {len(assay_subset)} rows for assay '{assay}' in the subset.\")\n",
    "                else:\n",
    "                    print(f\"  No rows found for assay '{assay}' to include in subset.\")\n",
    "\n",
    "            if subset_dfs:\n",
    "                occ_df_subset_for_matching = pd.concat(subset_dfs, ignore_index=True)\n",
    "                print(f\"Successfully created subset with {len(occ_df_subset_for_matching)} total rows for taxonomic matching.\")\n",
    "                print(f\"  Unique assay_names in subset: {occ_df_subset_for_matching['assay_name'].unique()}\")\n",
    "                # For inspection:\n",
    "                # display(occ_df_subset_for_matching[['eventID', 'occurrenceID', 'verbatimIdentification', 'assay_name']].head())\n",
    "            else:\n",
    "                print(\"Could not create a valid subset. No assay data found or all assays resulted in empty subsets.\")\n",
    "        else:\n",
    "            print(\"Full occurrence DataFrame is empty or missing 'assay_name'/'verbatimIdentification' columns. Cannot create subset.\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error loading or processing {full_occurrence_csv_path} for subsetting: {e}\")\n",
    "        import traceback\n",
    "        print(traceback.format_exc())\n",
    "else:\n",
    "    print(f\"ERROR: Full occurrence file not found at {full_occurrence_csv_path}. Cannot create subset. Ensure Cell 28 (occurrence.csv generation) ran successfully.\")\n",
    "\n",
    "# Determine which DataFrame to use for the actual matching process\n",
    "current_run_type = \"NO DATA\"\n",
    "if not occ_df_subset_for_matching.empty:\n",
    "    occ_df_for_actual_matching = occ_df_subset_for_matching\n",
    "    current_run_type = \"SUBSET\"\n",
    "    print(f\"\\nProceeding with the {current_run_type} ({len(occ_df_for_actual_matching)} rows) for taxonomic matching.\")\n",
    "elif not full_occ_df.empty: # Fallback to full if subset is empty but full loaded\n",
    "    print(\"\\nWARNING: Subset creation failed or resulted in an empty DataFrame. Proceeding with FULL dataset.\")\n",
    "    occ_df_for_actual_matching = full_occ_df\n",
    "    current_run_type = \"FULL DATASET\"\n",
    "else:\n",
    "    print(\"\\nERROR: Neither subset nor full occurrence data is available for matching. Please check previous cells.\")\n",
    "    occ_df_for_actual_matching = pd.DataFrame() # Ensure it's an empty DF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Taxonomic Assignment\n",
    "\n",
    "The following cells will perform taxonomic matching using the API specified in the notebook parameters (`params['taxonomic_api_source']`). This process will:\n",
    "1. Dynamically import the appropriate matching script (e.g., `WoRMS_v3_matching.py` or a future `GBIF_v3_matching.py`).\n",
    "2. Read the `occurrence.csv` file generated previously (which includes raw `verbatimIdentification` strings and `assay_name`).\n",
    "3. Use the imported script to query the selected API.\n",
    "4. Implement caching to avoid redundant API calls.\n",
    "5. Handle assay-specific rules (e.g., skipping species-level matching for specified assays).\n",
    "6. Update the occurrence data with API-derived `scientificName`, `scientificNameID`, `taxonRank`, and hierarchical rank columns.\n",
    "7. Set `nameAccordingTo` to reflect the API source.\n",
    "8. Handle specific post-processing rules, such as for \"Eukaryota\"-only verbatim strings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully imported and reloaded WoRMS_v3_matching.py\n",
      "\n",
      "Taxonomic Matching Setup:\n",
      "  API Source: WoRMS\n",
      "  Assays configured by user to skip species-level matching: ['ssu16sv4v5-emp']\n",
      "  Number of processes for WoRMS: All available CPUs\n",
      "  Output directory for reading/writing files: ../processed-v3/\n"
     ]
    }
   ],
   "source": [
    "import importlib\n",
    "import pandas as pd\n",
    "import os\n",
    "import WoRMS_v3_matching # Import the new script\n",
    "\n",
    "# Reload the module to pick up any changes if you edit the .py file\n",
    "importlib.reload(WoRMS_v3_matching)\n",
    "\n",
    "print(\"Successfully imported and reloaded WoRMS_v3_matching.py\")\n",
    "\n",
    "# --- Define Parameters for Taxonomic Matching ---\n",
    "\n",
    "# This should be defined in your main parameters cell (e.g., Cell 8 of your notebook)\n",
    "# Ensure 'params' dictionary exists from that cell.\n",
    "if 'params' not in globals():\n",
    "    print(\"CRITICAL ERROR: 'params' dictionary not found. It should be defined in an early cell (e.g., Cell 8).\")\n",
    "    params = {} # Initialize to prevent immediate crash, but this is not ideal\n",
    "\n",
    "# 1. API Source (already in your params from Cell 8)\n",
    "# params['taxonomic_api_source'] = 'WoRMS' # or 'GBIF' when ready\n",
    "\n",
    "# 2. Assays to skip SPECIES-level matching for (already in your params from Cell 8)\n",
    "# params['user_defined_assays_to_skip_species'] = ['ssu16sv4v5-emp'] # Example\n",
    "\n",
    "# 3. Number of processes for matching (0 means use all available CPUs)\n",
    "# This can also be in your main params cell or set here.\n",
    "params['worms_n_proc'] = params.get('worms_n_proc', 0) \n",
    "params['gbif_n_proc'] = params.get('gbif_n_proc', 0) # For future GBIF use\n",
    "\n",
    "# 4. Output directory (should be consistent with where occurrence.csv was saved)\n",
    "# This should also ideally come from your main params cell (Cell 8) or the occurrence.csv generation cell (Cell 28)\n",
    "if 'output_dir' not in params:\n",
    "    if 'output_dir' in globals() and isinstance(output_dir, str): # If set by cell 28\n",
    "         params['output_dir'] = output_dir\n",
    "    else:\n",
    "        params['output_dir'] = \"../processed-v3/\" # Fallback\n",
    "        print(f\"Warning: 'output_dir' not found in params, using fallback: {params['output_dir']}\")\n",
    "else: # If it is in params, ensure it's a string\n",
    "    if not isinstance(params['output_dir'], str):\n",
    "        params['output_dir'] = \"../processed-v3/\"\n",
    "        print(f\"Warning: params['output_dir'] was not a string, reset to default: {params['output_dir']}\")\n",
    "\n",
    "\n",
    "# --- Consolidate parameters for the matching function ---\n",
    "# The WoRMS_OBIS_matcher.py script will primarily use:\n",
    "# - params_dict['taxonomic_api_source']\n",
    "# - params_dict['assays_to_skip_species_match'] (derived from user_defined_assays_to_skip_species)\n",
    "# - params_dict['worms_n_proc'] (or 'gbif_n_proc' for GBIF)\n",
    "\n",
    "current_api_source = params.get('taxonomic_api_source', 'WoRMS') # Default to WoRMS if not set\n",
    "print(f\"\\nTaxonomic Matching Setup:\")\n",
    "print(f\"  API Source: {current_api_source}\")\n",
    "print(f\"  Assays configured by user to skip species-level matching: {params.get('user_defined_assays_to_skip_species', 'NONE DEFINED')}\")\n",
    "if current_api_source == 'WoRMS':\n",
    "    print(f\"  Number of processes for WoRMS: {params['worms_n_proc'] if params['worms_n_proc'] > 0 else 'All available CPUs'}\")\n",
    "elif current_api_source == 'GBIF':\n",
    "    print(f\"  Number of processes for GBIF: {params['gbif_n_proc'] if params['gbif_n_proc'] > 0 else 'All available CPUs'} (GBIF matching not yet fully implemented)\")\n",
    "print(f\"  Output directory for reading/writing files: {params['output_dir']}\")\n",
    "\n",
    "# Standard DwC ranks that the matching script will try to populate\n",
    "DWC_RANKS_STD = ['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform Taxonomic Matching\n",
    "\n",
    "This cell loads the intermediate `occurrence.csv` file (which should include the `assay_name` column) and applies the taxonomic matching using the selected API and its parameters. DEBUG statements will show progress.\n",
    "\n",
    "NOTE: This operation may take a few minutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Starting taxonomic matching using 'WoRMS' API on SUBSET data (20 rows)...\n",
      "Found 13 unique, non-empty (verbatimIdentification, assay_name) combinations for WoRMS matching.\n",
      "Starting WoRMS queries with 8 processes using multiprocess.Pool...\n",
      "  Processed 13/13 unique combinations from API/pool.\n",
      "\n",
      "Finished applying WoRMS taxonomic matches to DataFrame.\n",
      "\n",
      "Taxonomic matching process via 'WoRMS' finished.\n",
      "\n",
      "Preview of DataFrame after matching (selected columns):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>assay_name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>match_type_debug</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:1066</td>\n",
       "      <td>Subphylum</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Success_Query_Crustacea_EffectiveRank_genus_WoRMSRank_Subphylum</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Species</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Success_Query_Clausocalanus_furcatus_EffectiveRank_species_WoRMSRank_Species</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Oithona;Oithona_sp.;</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>Oithona</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:106485</td>\n",
       "      <td>Genus</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Success_Query_Oithona_EffectiveRank_genus_WoRMSRank_Genus</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Cyclopoida</td>\n",
       "      <td>Oithonidae</td>\n",
       "      <td>Oithona</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Class</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Success_Query_Dinophyceae_EffectiveRank_species_WoRMSRank_Class</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>ssu18sv9-emp</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Family</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Success_Query_Gymnodiniaceae_EffectiveRank_species_WoRMSRank_Family</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               eventID  \\\n",
       "0  GOMECC18S_Plate4_53   \n",
       "1  GOMECC18S_Plate4_53   \n",
       "2  GOMECC18S_Plate4_53   \n",
       "3  GOMECC18S_Plate4_53   \n",
       "4  GOMECC18S_Plate4_53   \n",
       "\n",
       "                                               occurrenceID  \\\n",
       "0  GOMECC18S_Plate4_53_occ_36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC18S_Plate4_53_occ_4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC18S_Plate4_53_occ_2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC18S_Plate4_53_occ_ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC18S_Plate4_53_occ_fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                         verbatimIdentification  \\\n",
       "0                                        Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2                   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Oithona;Oithona_sp.;   \n",
       "3                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "     assay_name          scientificName  \\\n",
       "0  ssu18sv9-emp               Crustacea   \n",
       "1  ssu18sv9-emp  Clausocalanus furcatus   \n",
       "2  ssu18sv9-emp                 Oithona   \n",
       "3  ssu18sv9-emp             Dinophyceae   \n",
       "4  ssu18sv9-emp          Gymnodiniaceae   \n",
       "\n",
       "                            scientificNameID  taxonRank nameAccordingTo  \\\n",
       "0    urn:lsid:marinespecies.org:taxname:1066  Subphylum           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104503    Species           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:106485      Genus           WoRMS   \n",
       "3   urn:lsid:marinespecies.org:taxname:19542      Class           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:109410     Family           WoRMS   \n",
       "\n",
       "                                                               match_type_debug  \\\n",
       "0               Success_Query_Crustacea_EffectiveRank_genus_WoRMSRank_Subphylum   \n",
       "1  Success_Query_Clausocalanus_furcatus_EffectiveRank_species_WoRMSRank_Species   \n",
       "2                     Success_Query_Oithona_EffectiveRank_genus_WoRMSRank_Genus   \n",
       "3               Success_Query_Dinophyceae_EffectiveRank_species_WoRMSRank_Class   \n",
       "4           Success_Query_Gymnodiniaceae_EffectiveRank_species_WoRMSRank_Family   \n",
       "\n",
       "     kingdom      phylum        class          order           family  \\\n",
       "0   Animalia  Arthropoda         None           None             None   \n",
       "1   Animalia  Arthropoda     Copepoda      Calanoida  Clausocalanidae   \n",
       "2   Animalia  Arthropoda     Copepoda     Cyclopoida       Oithonidae   \n",
       "3  Chromista     Myzozoa  Dinophyceae           None             None   \n",
       "4  Chromista     Myzozoa  Dinophyceae  Gymnodiniales   Gymnodiniaceae   \n",
       "\n",
       "           genus species  \n",
       "0           None    None  \n",
       "1  Clausocalanus    None  \n",
       "2        Oithona    None  \n",
       "3           None    None  \n",
       "4           None    None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counts of 'WoRMS' match_type_debug:\n",
      "match_type_debug\n",
      "Success_Query_Alphaproteobacteria_EffectiveRank_order_WoRMSRank_Class           6\n",
      "Success_Query_Dinophyceae_EffectiveRank_species_WoRMSRank_Class                 3\n",
      "Success_Query_Synechococcales_EffectiveRank_family_WoRMSRank_Order              3\n",
      "Success_Query_Hydrozoa_EffectiveRank_species_WoRMSRank_Class                    2\n",
      "Success_Query_Clausocalanus_furcatus_EffectiveRank_species_WoRMSRank_Species    1\n",
      "Success_Query_Crustacea_EffectiveRank_genus_WoRMSRank_Subphylum                 1\n",
      "Success_Query_Gymnodiniaceae_EffectiveRank_species_WoRMSRank_Family             1\n",
      "Success_Query_Oithona_EffectiveRank_genus_WoRMSRank_Genus                       1\n",
      "Success_Query_Pelagomonas_calceolata_EffectiveRank_species_WoRMSRank_Species    1\n",
      "Success_Query_Alphaproteobacteria_EffectiveRank_family_WoRMSRank_Class          1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Counts of resulting 'WoRMS' scientificName (top 10 unique non-NA values):\n",
      "scientificName\n",
      "Alphaproteobacteria       7\n",
      "Synechococcales           3\n",
      "Dinophyceae               3\n",
      "Hydrozoa                  2\n",
      "Crustacea                 1\n",
      "Gymnodiniaceae            1\n",
      "Oithona                   1\n",
      "Clausocalanus furcatus    1\n",
      "Pelagomonas calceolata    1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Perform Taxonomic Matching\n",
    "\n",
    "# 'occ_df_for_actual_matching' should be available from the subsetting cell (or fallback to full_occ_df)\n",
    "# 'params' dictionary should be available from the previous cell and Cell 8.\n",
    "\n",
    "if 'occ_df_for_actual_matching' not in globals() or occ_df_for_actual_matching.empty:\n",
    "    print(\"ERROR: 'occ_df_for_actual_matching' is not defined or is empty. Cannot proceed with taxonomic matching.\")\n",
    "    print(\"Please ensure the subsetting cell (or the cell loading occurrence.csv) ran correctly.\")\n",
    "    occ_df_matched_with_api = pd.DataFrame() # Initialize as empty to prevent error in next cell\n",
    "elif 'params' not in globals():\n",
    "    print(\"ERROR: 'params' dictionary not found. Cannot proceed with taxonomic matching.\")\n",
    "    occ_df_matched_with_api = pd.DataFrame()\n",
    "else:\n",
    "    api_to_use = params.get('taxonomic_api_source')\n",
    "    print(f\"\\nStarting taxonomic matching using '{api_to_use}' API on {current_run_type} data ({len(occ_df_for_actual_matching)} rows)...\")\n",
    "\n",
    "    if api_to_use == 'WoRMS':\n",
    "        if hasattr(WoRMS_v3_matching, 'get_worms_match_for_dataframe'):\n",
    "            # The function now directly uses the 'user_defined_assays_to_skip_species' from params for clarity\n",
    "            # It also needs 'worms_n_proc' from params\n",
    "            matching_params_for_script = {\n",
    "                'taxonomic_api_source': 'WoRMS', # Explicitly pass for clarity inside the script\n",
    "                'assays_to_skip_species_match': params.get('user_defined_assays_to_skip_species', []),\n",
    "                'worms_n_proc': params.get('worms_n_proc', 0) # Pass n_proc for WoRMS\n",
    "            }\n",
    "            occ_df_matched_with_api = WoRMS_v3_matching.get_worms_match_for_dataframe(\n",
    "                occurrence_df=occ_df_for_actual_matching,\n",
    "                params_dict=matching_params_for_script, # Pass the consolidated dict\n",
    "                n_proc=params.get('worms_n_proc', 0) # Also passed directly for Pool\n",
    "            )\n",
    "        else:\n",
    "            print(\"ERROR: WoRMS_OBIS_matcher.py does not have 'get_worms_match_for_dataframe' function.\")\n",
    "            occ_df_matched_with_api = occ_df_for_actual_matching.copy() # Avoid error\n",
    "    \n",
    "    elif api_to_use == 'GBIF':\n",
    "        print(\"GBIF matching is selected but not yet implemented in this workflow.\")\n",
    "        print(\"The 'occ_df_matched_with_api' will be a copy of the input or empty.\")\n",
    "        # Placeholder for GBIF call when ready:\n",
    "        # matching_params_for_script = {\n",
    "        #     'taxonomic_api_source': 'GBIF',\n",
    "        #     'assays_to_skip_species_match': params.get('user_defined_assays_to_skip_species', []), # or a GBIF specific list\n",
    "        #     'gbif_n_proc': params.get('gbif_n_proc', 0)\n",
    "        # }\n",
    "        # occ_df_matched_with_api = GBIF_OBIS_matcher.get_gbif_match_for_dataframe(\n",
    "        #     occurrence_df=occ_df_for_actual_matching,\n",
    "        #     params_dict=matching_params_for_script\n",
    "        # )\n",
    "        occ_df_matched_with_api = occ_df_for_actual_matching.copy() # For now\n",
    "    else:\n",
    "        print(f\"ERROR: Unknown or unsupported taxonomic_api_source: {api_to_use}\")\n",
    "        occ_df_matched_with_api = occ_df_for_actual_matching.copy() # Avoid error\n",
    "\n",
    "    print(f\"\\nTaxonomic matching process via '{api_to_use}' finished.\")\n",
    "    \n",
    "    if not occ_df_matched_with_api.empty:\n",
    "        print(\"\\nPreview of DataFrame after matching (selected columns):\")\n",
    "        preview_cols = ['eventID', 'occurrenceID', 'verbatimIdentification', 'assay_name', \n",
    "                        'scientificName', 'scientificNameID', 'taxonRank', 'nameAccordingTo', 'match_type_debug'] + DWC_RANKS_STD\n",
    "        display_cols = [col for col in preview_cols if col in occ_df_matched_with_api.columns]\n",
    "        display(occ_df_matched_with_api[display_cols].head())\n",
    "        \n",
    "        if 'match_type_debug' in occ_df_matched_with_api.columns:\n",
    "            print(f\"\\nCounts of '{api_to_use}' match_type_debug:\")\n",
    "            print(occ_df_matched_with_api['match_type_debug'].value_counts(dropna=False))\n",
    "        if 'scientificName' in occ_df_matched_with_api.columns:\n",
    "            print(f\"\\nCounts of resulting '{api_to_use}' scientificName (top 10 unique non-NA values):\")\n",
    "            print(occ_df_matched_with_api['scientificName'].dropna().value_counts().head(10))\n",
    "    else:\n",
    "        print(f\"DataFrame 'occ_df_matched_with_api' is empty after '{api_to_use}' matching attempt.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Post-Matching Processing and Final Save\n",
    "\n",
    "This cell handles API-specific post-processing. For example, if WoRMS was used, it checks for verbatim strings that were effectively \"Eukaryota\"-only and ensures they are set to \"incertae sedis\" if not already resolved to a lower rank. The temporary `assay_name` and `match_type_debug` columns are then removed, and the file is saved."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting post-matching processing.\n",
      "  Removed temporary columns: ['assay_name', 'match_type_debug']\n",
      "\n",
      "💾 Taxonomically updated SUBSET occurrence file 'occurrence_SUBSET_worms_matched.csv' saved to '../processed-v3/occurrence_SUBSET_worms_matched.csv' with 20 records.\n",
      "  Final columns written (35 total): ['eventID', 'organismQuantity', 'occurrenceID', 'verbatimIdentification', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species', 'scientificName', 'scientificNameID', 'taxonRank', 'identificationRemarks', 'taxonID', 'basisOfRecord', 'nameAccordingTo', 'organismQuantityType', 'recordedBy', 'materialSampleID', 'sampleSizeValue', 'sampleSizeUnit', 'associatedSequences', 'locationID', 'eventDate', 'minimumDepthInMeters', 'maximumDepthInMeters', 'locality', 'decimalLatitude', 'decimalLongitude', 'geodeticDatum', 'parentEventID', 'datasetID', 'occurrenceStatus']\n",
      "\n",
      "👀 Preview of final taxonomically updated SUBSET occurrence data (first 5 rows, selected columns):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:1066</td>\n",
       "      <td>Subphylum</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Species</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Oithona;Oithona_sp.;</td>\n",
       "      <td>Oithona</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:106485</td>\n",
       "      <td>Genus</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Cyclopoida</td>\n",
       "      <td>Oithonidae</td>\n",
       "      <td>Oithona</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Class</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>GOMECC18S_Plate4_53_occ_fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Family</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               eventID  \\\n",
       "0  GOMECC18S_Plate4_53   \n",
       "1  GOMECC18S_Plate4_53   \n",
       "2  GOMECC18S_Plate4_53   \n",
       "3  GOMECC18S_Plate4_53   \n",
       "4  GOMECC18S_Plate4_53   \n",
       "\n",
       "                                               occurrenceID  \\\n",
       "0  GOMECC18S_Plate4_53_occ_36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC18S_Plate4_53_occ_4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC18S_Plate4_53_occ_2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC18S_Plate4_53_occ_ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC18S_Plate4_53_occ_fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                         verbatimIdentification  \\\n",
       "0                                        Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2                   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Oithona;Oithona_sp.;   \n",
       "3                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "           scientificName                           scientificNameID  \\\n",
       "0               Crustacea    urn:lsid:marinespecies.org:taxname:1066   \n",
       "1  Clausocalanus furcatus  urn:lsid:marinespecies.org:taxname:104503   \n",
       "2                 Oithona  urn:lsid:marinespecies.org:taxname:106485   \n",
       "3             Dinophyceae   urn:lsid:marinespecies.org:taxname:19542   \n",
       "4          Gymnodiniaceae  urn:lsid:marinespecies.org:taxname:109410   \n",
       "\n",
       "   taxonRank nameAccordingTo    kingdom      phylum        class  \\\n",
       "0  Subphylum           WoRMS   Animalia  Arthropoda         None   \n",
       "1    Species           WoRMS   Animalia  Arthropoda     Copepoda   \n",
       "2      Genus           WoRMS   Animalia  Arthropoda     Copepoda   \n",
       "3      Class           WoRMS  Chromista     Myzozoa  Dinophyceae   \n",
       "4     Family           WoRMS  Chromista     Myzozoa  Dinophyceae   \n",
       "\n",
       "           order           family          genus species  \n",
       "0           None             None           None    None  \n",
       "1      Calanoida  Clausocalanidae  Clausocalanus    None  \n",
       "2     Cyclopoida       Oithonidae        Oithona    None  \n",
       "3           None             None           None    None  \n",
       "4  Gymnodiniales   Gymnodiniaceae           None    None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Post-Matching Processing and Final Save\n",
    "\n",
    "if 'occ_df_matched_with_api' not in globals() or occ_df_matched_with_api.empty:\n",
    "    print(\"Skipping post-matching and save: 'occ_df_matched_with_api' is not defined or is empty.\")\n",
    "else:\n",
    "    print(\"Starting post-matching processing.\")\n",
    "    api_source_used = params.get('taxonomic_api_source', 'UnknownAPI') # Get from params\n",
    "    \n",
    "    # --- Handle \"Eukaryota\"-only verbatim strings specifically after WoRMS matching ---\n",
    "    # This logic is simplified here. The matching script itself tries to be robust.\n",
    "    # However, a final check can be useful.\n",
    "    if api_source_used == 'WoRMS':\n",
    "        # Example condition: if scientificName is 'Eukaryota' and taxonRank is 'kingdom' (or similar high rank)\n",
    "        # after matching, and the original verbatim was simple like \"Eukaryota\", convert to \"incertae sedis\".\n",
    "        # This specific rule might need more nuanced conditions based on your actual WoRMS script's behavior.\n",
    "        \n",
    "        # A more robust check for \"effectively Eukaryota only\" verbatim strings that resulted in high-level Eukaryota match:\n",
    "        eukaryota_verbatim_mask = pd.Series([False] * len(occ_df_matched_with_api), index=occ_df_matched_with_api.index)\n",
    "        if 'verbatimIdentification' in occ_df_matched_with_api.columns and \\\n",
    "           'scientificName' in occ_df_matched_with_api.columns and \\\n",
    "           'taxonRank' in occ_df_matched_with_api.columns:\n",
    "\n",
    "            for index, row in occ_df_matched_with_api.iterrows():\n",
    "                vi = str(row.get('verbatimIdentification', '')).strip().lower()\n",
    "                parsed_vi = [term.strip() for term in vi.split(';') if term.strip()]\n",
    "                \n",
    "                is_effectively_eukaryota_only_verbatim = False\n",
    "                if len(parsed_vi) == 1 and parsed_vi[0] == 'eukaryota':\n",
    "                    is_effectively_eukaryota_only_verbatim = True\n",
    "                elif len(parsed_vi) > 0 and parsed_vi[-1] == 'eukaryota': \n",
    "                    preceding_meaningful = [term for term in parsed_vi[:-1] if term and term != 'unassigned']\n",
    "                    if not preceding_meaningful: # If only \"eukaryota\" is meaningful at the end\n",
    "                        is_effectively_eukaryota_only_verbatim = True\n",
    "                \n",
    "                if is_effectively_eukaryota_only_verbatim:\n",
    "                    current_sn = str(row.get('scientificName', '')).lower()\n",
    "                    current_tr = str(row.get('taxonRank', '')).lower()\n",
    "                    high_ranks = ['kingdom', 'superkingdom', 'domain', 'subkingdom', 'no rank', 'unranked', ''] # Include empty/None as high rank\n",
    "                    \n",
    "                    if current_sn == 'eukaryota' and current_tr in high_ranks:\n",
    "                        eukaryota_verbatim_mask[index] = True\n",
    "                    # Also, if WoRMS already returned incertae sedis for such a string, ensure it's standardized.\n",
    "                    elif current_sn == 'incertae sedis' and is_effectively_eukaryota_only_verbatim:\n",
    "                         eukaryota_verbatim_mask[index] = True\n",
    "\n",
    "\n",
    "            num_eukaryota_verbatim_to_convert = eukaryota_verbatim_mask.sum()\n",
    "            if num_eukaryota_verbatim_to_convert > 0:\n",
    "                print(f\"  Post-processing ({api_source_used}): Found {num_eukaryota_verbatim_to_convert} records with 'Eukaryota'-only verbatim strings that matched to high-level Eukaryota or were already 'incertae sedis'. Standardizing to 'incertae sedis'.\")\n",
    "                occ_df_matched_with_api.loc[eukaryota_verbatim_mask, 'scientificName'] = 'incertae sedis'\n",
    "                occ_df_matched_with_api.loc[eukaryota_verbatim_mask, 'scientificNameID'] = 'urn:lsid:marinespecies.org:taxname:12' # AphiaID for Incertae Sedis in WoRMS\n",
    "                occ_df_matched_with_api.loc[eukaryota_verbatim_mask, 'taxonRank'] = 'no rank'\n",
    "                for rank_col in DWC_RANKS_STD: # Clear out other ranks\n",
    "                    if rank_col in occ_df_matched_with_api.columns:\n",
    "                        occ_df_matched_with_api.loc[eukaryota_verbatim_mask, rank_col] = None \n",
    "                if 'match_type_debug' in occ_df_matched_with_api.columns:\n",
    "                     occ_df_matched_with_api.loc[eukaryota_verbatim_mask, 'match_type_debug'] = 'Standardized_IncertaeSedis_From_EukaryotaOnlyVerbatim'\n",
    "                if 'nameAccordingTo' in occ_df_matched_with_api.columns:\n",
    "                    occ_df_matched_with_api.loc[eukaryota_verbatim_mask, 'nameAccordingTo'] = api_source_used + \"; Local rule for Eukaryota-only verbatim\"\n",
    "\n",
    "\n",
    "    # --- Remove temporary/intermediate columns ---\n",
    "    columns_to_drop_final = []\n",
    "    if 'assay_name' in occ_df_matched_with_api.columns: columns_to_drop_final.append('assay_name')\n",
    "    if 'match_type_debug' in occ_df_matched_with_api.columns: columns_to_drop_final.append('match_type_debug')\n",
    "    \n",
    "    if columns_to_drop_final:\n",
    "        occ_df_matched_with_api.drop(columns=columns_to_drop_final, inplace=True, errors='ignore')\n",
    "        print(f\"  Removed temporary columns: {columns_to_drop_final}\")\n",
    "\n",
    "    # --- Define final column order (should match DESIRED_OCCURRENCE_COLUMNS_IN_ORDER from Cell 28, *without* 'assay_name') ---\n",
    "    # This list is defined in your cell 28 (occurrence core generation)\n",
    "    # We re-use it here, excluding 'assay_name' if it was present.\n",
    "    if 'DESIRED_OCCURRENCE_COLUMNS_IN_ORDER' in globals():\n",
    "        final_dwc_columns_ordered_post_match = [col for col in DESIRED_OCCURRENCE_COLUMNS_IN_ORDER if col != 'assay_name']\n",
    "    else: # Fallback if the list isn't found (shouldn't happen if cell 28 ran)\n",
    "        print(\"Warning: DESIRED_OCCURRENCE_COLUMNS_IN_ORDER not found. Using columns present in DataFrame.\")\n",
    "        final_dwc_columns_ordered_post_match = occ_df_matched_with_api.columns.tolist()\n",
    "\n",
    "    # Ensure all desired columns exist, add as NA if missing\n",
    "    for col in final_dwc_columns_ordered_post_match:\n",
    "        if col not in occ_df_matched_with_api.columns:\n",
    "            occ_df_matched_with_api[col] = pd.NA \n",
    "            print(f\"  Warning: Final column '{col}' was missing from matched data and added as NA.\")\n",
    "\n",
    "    occ_df_final_output = occ_df_matched_with_api.reindex(columns=final_dwc_columns_ordered_post_match)\n",
    "    \n",
    "    # --- Save the final, taxonomically enriched occurrence file ---\n",
    "    # Filename depends on whether it was a subset or full run for clarity\n",
    "    file_prefix = \"occurrence_SUBSET\" if current_run_type == \"SUBSET\" else \"occurrence\"\n",
    "    final_occurrence_filename = f\"{file_prefix}_{api_source_used.lower()}_matched.csv\" \n",
    "    \n",
    "    # Ensure params['output_dir'] is a valid path\n",
    "    output_directory_for_save = params.get('output_dir', \"../processed-v3/\")\n",
    "    if not isinstance(output_directory_for_save, str): # Safeguard\n",
    "        output_directory_for_save = \"../processed-v3/\"\n",
    "    os.makedirs(output_directory_for_save, exist_ok=True) # Ensure directory exists\n",
    "    \n",
    "    final_output_path = os.path.join(output_directory_for_save, final_occurrence_filename)\n",
    "\n",
    "    try:\n",
    "        occ_df_final_output.to_csv(final_output_path, index=False, na_rep='') \n",
    "        print(f\"\\n💾 Taxonomically updated {current_run_type} occurrence file '{final_occurrence_filename}' saved to '{final_output_path}' with {len(occ_df_final_output)} records.\")\n",
    "        print(f\"  Final columns written ({len(occ_df_final_output.columns)} total): {occ_df_final_output.columns.tolist()}\")\n",
    "        \n",
    "        print(f\"\\n👀 Preview of final taxonomically updated {current_run_type} occurrence data (first 5 rows, selected columns):\")\n",
    "        preview_cols_final = ['eventID', 'occurrenceID', 'verbatimIdentification', \n",
    "                              'scientificName', 'scientificNameID', 'taxonRank', 'nameAccordingTo'] + DWC_RANKS_STD\n",
    "        display_cols_final = [col for col in preview_cols_final if col in occ_df_final_output.columns]\n",
    "        display(occ_df_final_output[display_cols_final].head())\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"  ❌ ERROR saving final taxonomically updated {current_run_type} occurrence file: {str(e)}\")\n",
    "        import traceback\n",
    "        print(traceback.format_exc())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# WoRMS conversion. \n",
    "Note, can't use `multiprocessing` library in a Jupyter notebook, need `multiprocess`. See [here](https://stackoverflow.com/questions/41385708/multiprocessing-example-giving-attributeerror)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OBIS currently requires taxonomy assignments that match WoRMS, however none of the commonly used metabarcoding reference databases use WoRMS as the basis of their taxonomy. This means the taxonomic ranks for any given scientific name on WoRMS may not directly compare to what is assigned. There are ongoing discussions about this problem (see [this](https://github.com/iobis/Project-team-Genetic-Data/issues/5) GitHub issue).     \n",
    "\n",
    "Many of them, especially for microbes, include taxa that aren't on WoRMS at all. This is because the name may not have been fully and officially adopted by the scientific community (or at least not adopted by WoRMS yet). We therefore need a system for searching through the higher taxonomic ranks given, finding the lowest one that will match on WoRMS, and putting that name in the `scientificName` column. The assigned taxonomy is then recorded in `verbatimIdentification`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Had some [issues with the parallelization](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr) on Mac M1. Adding 'OBJC_DISABLE_INITIALIZE_FORK_SAFETY = YES' to .bash_profile and then [This](https://github.com/python/cpython/issues/74570) fixed it.   \n",
    "Try to run without the bash_profile fix LATER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"no_proxy\"]=\"*\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 16S worms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Species level IDs might be trash, [see here](https://forum.qiime2.org/t/processing-filtering-and-evaluating-the-silva-database-and-other-reference-sequence-data-with-rescript/15494), so look at genus and up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "import WoRMS_matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'WoRMS_matching' from 'c:\\\\Users\\\\bayde\\\\OneDrive\\\\Documents\\\\NOAA_AOML_Code\\\\edna2obis\\\\edna2obis\\\\src\\\\WoRMS_matching.py'>"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import importlib\n",
    "importlib.reload(WoRMS_matching)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_16S = asv_tables['16S V4-V5'][['taxonomy','domain','phylum','class','order','family','genus','species']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ignore_index is important!\n",
    "tax_16S = tax_16S.drop_duplicates(ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2729, 8)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_16S.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    worms_16s = WoRMS_matching.get_worms_from_scientific_name_parallel(\n",
    "    tax_df = tax_16S,ordered_rank_columns=['genus','family','order','class','phylum','domain'],\n",
    "    full_tax_column=\"taxonomy\",full_tax_vI=True,n_proc=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea</td>\n",
       "      <td>class</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales</td>\n",
       "      <td>d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales</td>\n",
       "      <td>class</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:559846</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium</td>\n",
       "      <td>d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium</td>\n",
       "      <td>domain</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...</td>\n",
       "      <td>order</td>\n",
       "      <td>Rhodospirillales</td>\n",
       "      <td>Rhodospirillales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392751</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodospirillales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium</td>\n",
       "      <td>genus</td>\n",
       "      <td>Sphingobium</td>\n",
       "      <td>Sphingobium</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:571470</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Sphingomonadales</td>\n",
       "      <td>Sphingomonadaceae</td>\n",
       "      <td>Sphingobium</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                full_tax  \\\n",
       "0                   d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea   \n",
       "1                                                                                       d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales   \n",
       "2                       d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...   \n",
       "4                                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                   d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea   \n",
       "1                                                                                       d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales   \n",
       "2                       d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...   \n",
       "4                                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium   \n",
       "\n",
       "  old_taxonRank             old name       scientificName  \\\n",
       "0         class  Gammaproteobacteria  Gammaproteobacteria   \n",
       "1         class          Bacteroidia          Bacteroidia   \n",
       "2        domain             Bacteria             Bacteria   \n",
       "3         order     Rhodospirillales     Rhodospirillales   \n",
       "4         genus          Sphingobium          Sphingobium   \n",
       "\n",
       "                            scientificNameID   kingdom          phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "1  urn:lsid:marinespecies.org:taxname:559846  Bacteria   Bacteroidetes   \n",
       "2       urn:lsid:marinespecies.org:taxname:6  Bacteria            None   \n",
       "3  urn:lsid:marinespecies.org:taxname:392751  Bacteria  Proteobacteria   \n",
       "4  urn:lsid:marinespecies.org:taxname:571470  Bacteria  Proteobacteria   \n",
       "\n",
       "                 class             order             family        genus  \\\n",
       "0  Gammaproteobacteria              None               None         None   \n",
       "1          Bacteroidia              None               None         None   \n",
       "2                 None              None               None         None   \n",
       "3  Alphaproteobacteria  Rhodospirillales               None         None   \n",
       "4  Alphaproteobacteria  Sphingomonadales  Sphingomonadaceae  Sphingobium   \n",
       "\n",
       "  taxonRank  \n",
       "0     Class  \n",
       "1     Class  \n",
       "2   Kingdom  \n",
       "3     Order  \n",
       "4     Genus  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>d__Eukaryota</td>\n",
       "      <td>d__Eukaryota</td>\n",
       "      <td>domain</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         full_tax verbatimIdentification old_taxonRank   old name  \\\n",
       "242  d__Eukaryota           d__Eukaryota        domain  Eukaryota   \n",
       "\n",
       "    scientificName scientificNameID kingdom phylum class order family genus  \\\n",
       "242       No match             None     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "\n",
       "    taxonRank  \n",
       "242       NaN  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s[worms_16s[\"scientificName\"]==\"No match\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_16s.loc[worms_16s[\"scientificName\"]==\"No match\",'scientificName'] = \"Biota\"\n",
    "worms_16s.loc[worms_16s[\"scientificName\"]==\"Biota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      full_tax verbatimIdentification old_taxonRank old name scientificName  \\\n",
       "97  Unassigned             Unassigned           NaN      NaN            NaN   \n",
       "\n",
       "   scientificNameID kingdom phylum class order family genus taxonRank  \n",
       "97              NaN     NaN    NaN   NaN   NaN    NaN   NaN       NaN  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s[worms_16s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# incertae sedis    is controlled vocab for saying unnassigned \n",
    "# also insert that if it just matched to Eukaryota"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 13)\n",
      "(0, 13)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(worms_16s[worms_16s['scientificName'].isna() == True].shape)\n",
    "worms_16s.loc[worms_16s['scientificName'].isna() == True,'scientificName'] = 'incertae sedis'\n",
    "worms_16s.loc[worms_16s['scientificName'] == 'incertae sedis','scientificNameID'] =  'urn:lsid:marinespecies.org:taxname:12'\n",
    "print(worms_16s[worms_16s['scientificName'].isna() == True].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_16s.to_csv(\"../processed/worms_16S_matching.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales</td>\n",
       "      <td>d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:559846</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium</td>\n",
       "      <td>d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...</td>\n",
       "      <td>Rhodospirillales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392751</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodospirillales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium</td>\n",
       "      <td>Sphingobium</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:571470</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Sphingomonadales</td>\n",
       "      <td>Sphingomonadaceae</td>\n",
       "      <td>Sphingobium</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                full_tax  \\\n",
       "0                   d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea   \n",
       "1                                                                                       d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales   \n",
       "2                       d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...   \n",
       "4                                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                   d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Cellvibrionales; f__Halieaceae; g__OM60(NOR5)_clade; s__uncultured_Haliea   \n",
       "1                                                                                       d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Chitinophagales   \n",
       "2                       d__Bacteria; p__Verrucomicrobiota; c__Omnitrophia; o__Omnitrophales; f__Omnitrophales; g__Omnitrophales; s__uncultured_bacterium   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodospirillales; f__AEGEAN-169_marine_group; g__AEGEAN-169_marine_group; s__alpha_prot...   \n",
       "4                                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae; g__Sphingobium   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Gammaproteobacteria  urn:lsid:marinespecies.org:taxname:393018  Bacteria   \n",
       "1          Bacteroidia  urn:lsid:marinespecies.org:taxname:559846  Bacteria   \n",
       "2             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "3     Rhodospirillales  urn:lsid:marinespecies.org:taxname:392751  Bacteria   \n",
       "4          Sphingobium  urn:lsid:marinespecies.org:taxname:571470  Bacteria   \n",
       "\n",
       "           phylum                class             order             family  \\\n",
       "0  Proteobacteria  Gammaproteobacteria              None               None   \n",
       "1   Bacteroidetes          Bacteroidia              None               None   \n",
       "2            None                 None              None               None   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rhodospirillales               None   \n",
       "4  Proteobacteria  Alphaproteobacteria  Sphingomonadales  Sphingomonadaceae   \n",
       "\n",
       "         genus taxonRank  \n",
       "0         None     Class  \n",
       "1         None     Class  \n",
       "2         None   Kingdom  \n",
       "3         None     Order  \n",
       "4  Sphingobium     Genus  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s.drop(columns=['old name','old_taxonRank'],inplace=True)\n",
    "worms_16s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Parvibaculales</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>uncultured marine</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Puniceispirillales</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Thermoplasmatota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            featureid  \\\n",
       "182  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "225  00e6c13fe86364a5084987093afa1916   \n",
       "347  015dad1fafca90944d905beb2a980bc3   \n",
       "412  019c88c6ade406f731954f38e3461564   \n",
       "719  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                  sequence  \\\n",
       "182  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "225  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "347  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "412  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "719  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                  taxonomy  \\\n",
       "182                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "225                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "347                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "412                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "719  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "     Confidence    domain                         phylum  \\\n",
       "182    0.832190  Bacteria                 Proteobacteria   \n",
       "225    0.867040  Bacteria                 Proteobacteria   \n",
       "347    1.000000   Archaea               Thermoplasmatota   \n",
       "412    0.952911  Bacteria                 Proteobacteria   \n",
       "719    0.818195  Bacteria  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                             class                          order  \\\n",
       "182            Alphaproteobacteria                 Parvibaculales   \n",
       "225            Alphaproteobacteria             Puniceispirillales   \n",
       "347                 Thermoplasmata                Marine Group II   \n",
       "412            Alphaproteobacteria                  Rickettsiales   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                            family                          genus  \\\n",
       "182                   OCS116 clade                   OCS116 clade   \n",
       "225                   SAR116 clade                   SAR116 clade   \n",
       "347                Marine Group II                Marine Group II   \n",
       "412                   Mitochondria                   Mitochondria   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                  species                 eventID  organismQuantity  \\\n",
       "182     uncultured marine  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "225                   NaN  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "347                   NaN  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "412  uncultured bacterium  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "719  uncultured bacterium  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                                       occurrenceID  \n",
       "182  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d  \n",
       "225  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916  \n",
       "347  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3  \n",
       "412  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564  \n",
       "719  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S V4-V5'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Occurrence and worms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(169470, 14)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S V4-V5'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                taxonomy  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "   Confidence                 eventID  organismQuantity  \\\n",
       "0    0.832190  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "1    0.867040  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "2    1.000000  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "3    0.952911  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "4    0.818195  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class  \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class  \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class  \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order  \n",
       "4            None                 None           None   None  None   Kingdom  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "occ16_test = occ['16S V4-V5'].copy()\n",
    "occ16_test.drop(columns=['domain','phylum','class','order','family','genus','species'],inplace=True)\n",
    "\n",
    "occ16_test = occ16_test.merge(worms_16s, how='left', left_on ='taxonomy', right_on='full_tax')\n",
    "occ16_test.drop(columns='full_tax', inplace=True)\n",
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks  \n",
    "\n",
    "```\n",
    "occ16_test['identificationRemarks'] = occ16_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+occ16_test['Confidence'].astype(str) +\", against reference database: \"+occ16_test['taxa_ref_db']\n",
    "```\n",
    "\n",
    "'Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>controls_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>411</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S V9</td>\n",
       "      <td>260</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  amplicon_sequenced  ampliconSize trim_method  \\\n",
       "0          16S V4-V5           411    cutadapt   \n",
       "1             18S V9           260    cutadapt   \n",
       "\n",
       "                     cluster_method pid_clustering  \\\n",
       "0  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "1  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1             PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706   \n",
       "\n",
       "                             code_repo  \\\n",
       "0  https://github.com/aomlomics/gomecc   \n",
       "1  https://github.com/aomlomics/gomecc   \n",
       "\n",
       "                                                                       identificationReferences  \\\n",
       "0                              10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline   \n",
       "1  10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline   \n",
       "\n",
       "                                                                                                                            controls_used  \n",
       "0  12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community  \n",
       "1                          12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls   "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '16S V4-V5','taxa_class_method'].item()\n",
    "occ16_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '16S V4-V5','taxa_ref_db'].item()\n",
    "\n",
    "occ16_test['identificationRemarks'] = occ16_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+occ16_test['Confidence'].astype(str) +\", against reference database: \"+occ16_test['taxa_ref_db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695'"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test['identificationRemarks'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### taxonID, basisOfRecord, eventID, nameAccordingTo, organismQuantityType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['taxonID'] = 'ASV:'+occ16_test['featureid']\n",
    "occ16_test['basisOfRecord'] = 'MaterialSample'\n",
    "occ16_test['nameAccordingTo'] = \"WoRMS\"\n",
    "occ16_test['organismQuantityType'] = \"DNA sequence reads\"\n",
    "occ16_test['recordedBy'] = data['study_data']['recordedBy'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### associatedSequences, materialSampleID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_name', 'library_id', 'title', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'lib_layout', 'platform',\n",
       "       'instrument_model', 'design_description', 'filetype', 'filename',\n",
       "       'filename2', 'biosample_accession', 'sra_accession', 'seq_meth',\n",
       "       'nucl_acid_ext', 'amplicon_sequenced', 'target_gene',\n",
       "       'target_subfragment', 'pcr_primer_forward', 'pcr_primer_reverse',\n",
       "       'pcr_primer_name_forward', 'pcr_primer_name_reverse',\n",
       "       'pcr_primer_reference', 'pcr_cond', 'nucl_acid_amp', 'adapters',\n",
       "       'mid_barcode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test = occ16_test.merge(data['prep_data'].loc[data['prep_data']['amplicon_sequenced'] == '16S V4-V5',['sample_name','sra_accession','biosample_accession']], how='left', left_on ='eventID', right_on='sample_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                taxonomy  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "   Confidence                 eventID  organismQuantity  \\\n",
       "0    0.832190  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "1    0.867040  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "2    1.000000  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "3    0.952911  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "4    0.818195  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \\\n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order   \n",
       "4            None                 None           None   None  None   Kingdom   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "\n",
       "              sample_name sra_accession biosample_accession  \n",
       "0  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "1  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "2  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "3  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "4  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### eventID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['eventID'] = occ16_test['eventID']+\"_16S\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sampleSize "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sampleSize by total number of reads per sample\n",
    "x = asv_tables['16S V4-V5'].sum(numeric_only=True).astype('int')\n",
    "x.index = x.index+\"_16S\"\n",
    "occ16_test['sampleSizeValue'] = occ16_test['eventID'].map(x).astype('str')\n",
    "occ16_test['sampleSizeUnit'] = 'DNA sequence reads'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                taxonomy  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "   Confidence                     eventID  organismQuantity  \\\n",
       "0    0.832190  GOMECC4_27N_Sta1_DCM_A_16S                18   \n",
       "1    0.867040  GOMECC4_27N_Sta1_DCM_A_16S                36   \n",
       "2    1.000000  GOMECC4_27N_Sta1_DCM_A_16S                49   \n",
       "3    0.952911  GOMECC4_27N_Sta1_DCM_A_16S                 2   \n",
       "4    0.818195  GOMECC4_27N_Sta1_DCM_A_16S                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \\\n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order   \n",
       "4            None                 None           None   None  None   Kingdom   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "\n",
       "              sample_name sra_accession biosample_accession sampleSizeValue  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "1  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "2  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "3  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "4  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "\n",
       "       sampleSizeUnit  \n",
       "0  DNA sequence reads  \n",
       "1  DNA sequence reads  \n",
       "2  DNA sequence reads  \n",
       "3  DNA sequence reads  \n",
       "4  DNA sequence reads  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ16_test.drop(columns=['sample_name','featureid','taxonomy','Confidence','taxa_class_method','taxa_ref_db'],inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge event and occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>countryCode</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>geodeticDatum</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    eventID        locationID  \\\n",
       "939      GOMECC4_CAPECORAL_Sta141_DCM_B_18S  CAPECORAL_Sta141   \n",
       "940      GOMECC4_CAPECORAL_Sta141_DCM_C_18S  CAPECORAL_Sta141   \n",
       "941  GOMECC4_CAPECORAL_Sta141_Surface_A_18S  CAPECORAL_Sta141   \n",
       "942  GOMECC4_CAPECORAL_Sta141_Surface_B_18S  CAPECORAL_Sta141   \n",
       "943  GOMECC4_CAPECORAL_Sta141_Surface_C_18S  CAPECORAL_Sta141   \n",
       "\n",
       "                  eventDate minimumDepthInMeters maximumDepthInMeters  \\\n",
       "939  2021-10-20T12:47-04:00                   59                   59   \n",
       "940  2021-10-20T12:47-04:00                   59                   59   \n",
       "941  2021-10-20T12:47-04:00                    4                    4   \n",
       "942  2021-10-20T12:47-04:00                    4                    4   \n",
       "943  2021-10-20T12:47-04:00                    4                    4   \n",
       "\n",
       "                locality        waterBody countryCode decimalLatitude  \\\n",
       "939  USA: Gulf of Mexico  Mexico, Gulf of          US          25.574   \n",
       "940  USA: Gulf of Mexico  Mexico, Gulf of          US          25.574   \n",
       "941  USA: Gulf of Mexico  Mexico, Gulf of          US          25.574   \n",
       "942  USA: Gulf of Mexico  Mexico, Gulf of          US          25.574   \n",
       "943  USA: Gulf of Mexico  Mexico, Gulf of          US          25.574   \n",
       "\n",
       "    decimalLongitude geodeticDatum samplingProtocol  \\\n",
       "939          -84.843         WGS84      CTD rosette   \n",
       "940          -84.843         WGS84      CTD rosette   \n",
       "941          -84.843         WGS84      CTD rosette   \n",
       "942          -84.843         WGS84      CTD rosette   \n",
       "943          -84.843         WGS84      CTD rosette   \n",
       "\n",
       "                          parentEventID          datasetID  \n",
       "939      GOMECC4_CAPECORAL_Sta141_DCM_B  noaa-aoml-gomecc4  \n",
       "940      GOMECC4_CAPECORAL_Sta141_DCM_C  noaa-aoml-gomecc4  \n",
       "941  GOMECC4_CAPECORAL_Sta141_Surface_A  noaa-aoml-gomecc4  \n",
       "942  GOMECC4_CAPECORAL_Sta141_Surface_B  noaa-aoml-gomecc4  \n",
       "943  GOMECC4_CAPECORAL_Sta141_Surface_C  noaa-aoml-gomecc4  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_event_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_merged = occ16_test.merge(all_event_data,how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>countryCode</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>geodeticDatum</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                            DNA_sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S                18   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S                36   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S                49   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S                 2   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \\\n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order   \n",
       "4            None                 None           None   None  None   Kingdom   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "  sampleSizeValue      sampleSizeUnit  \\\n",
       "0           16187  DNA sequence reads   \n",
       "1           16187  DNA sequence reads   \n",
       "2           16187  DNA sequence reads   \n",
       "3           16187  DNA sequence reads   \n",
       "4           16187  DNA sequence reads   \n",
       "\n",
       "                                                                                                                                     associatedSequences  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "1  https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "2  https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "3  https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "4  https://www.ncbi.nlm.nih.gov/sra/SRR26148187 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "\n",
       "  locationID               eventDate minimumDepthInMeters  \\\n",
       "0   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "1   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "2   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "3   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "4   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "\n",
       "  maximumDepthInMeters                                     locality  \\\n",
       "0                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "         waterBody countryCode decimalLatitude decimalLongitude geodeticDatum  \\\n",
       "0  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "1  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "2  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "3  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "4  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "\n",
       "  samplingProtocol           parentEventID          datasetID  \n",
       "0      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "1      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "2      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "3      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "4      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_merged.drop(columns=['DNA_sequence']).to_csv(\"../processed/occurrence_16S.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 18S worms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "18S PR2 database provides WORMS IDs for species that are in worms. We will read in that file, assign known worms ids, the do a search for unannotated taxa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update the path to the PR2 database\n",
    "pr2_18S = pd.read_excel(\"C:/Users/bayde/OneDrive/Documents/NOAA_AOML_Code/reference_DBs/pr2_version_5.0.0_taxonomy.xlsx\",\n",
    "    index_col=None, na_values=[\"\"])\n",
    "pr2_18S = pr2_18S.dropna(subset=['worms_id'])\n",
    "pr2_18S['worms_id'] = pr2_18S['worms_id'].astype('int').astype('str')\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('_',' ',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace(' sp\\.','',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace(' spp\\.','',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('-',' ',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('\\/',' ',regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "pr2_18S_dict = dict(zip(pr2_18S.species,pr2_18S.worms_id))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'614894'"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(pr2_18S_dict['Aphanocapsa feldmannii'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### code to get record from aphia id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Had some [issues with the parallelization](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr) on Mac M1. Adding 'OBJC_DISABLE_INITIALIZE_FORK_SAFETY = YES' to .bash_profile and then [This](https://github.com/python/cpython/issues/74570) fixed it.   \n",
    "Try to run without the bash_profile fix LATER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"no_proxy\"]=\"*\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_18S = asv_tables['18S V9'][['taxonomy','domain','supergroup','division','subdivision','class','order','family','genus','species']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1374, 10)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_18S = tax_18S.drop_duplicates(ignore_index=True)\n",
    "tax_18S.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aspergillus penicillioides: No match, speciesProtoscenium cf intricatum: No match, species\n",
      "\n",
      "Euglypha acanthophora: No match, species\n",
      "RAD B X Group IVe X: No match, species\n",
      "RAD B X Group IVe X: No match, genus\n",
      "Eimeriida: No match, order\n",
      "RAD B X Group IVe: No match, family\n",
      "Coccidiomorphea: No match, class\n",
      "MAST 12A: No match, species\n",
      "MAST 12A: No match, genus\n",
      "Nibbleromonas: No match, genus\n",
      "RAD B X: No match, order\n",
      "MAST 12: No match, family\n",
      "Nibbleridae: No match, family\n",
      "RAD B: No match, class\n",
      "Opalozoa X: No match, order\n",
      "Nibbleridida: No match, order\n",
      "Malus x: No match, species\n",
      "Nibbleridea: No match, class\n",
      "Euduboscquella cachoni: No match, species\n",
      "Malus: No match, genus\n",
      "Obazoa: No match, supergroup\n",
      "Nibbleridia X: No match, subdivision\n",
      "Pectinoida: No match, family\n",
      "Embryophyceae XX: No match, family\n",
      "Nibbleridia: No match, division\n",
      "Embryophyceae X: No match, order\n",
      "Provora: No match, supergroup\n",
      "Skeletonema menzellii: No match, species\n",
      "Dictyochales X: No match, species\n",
      "Dictyochales X: No match, genus\n",
      "Embryophyceae: No match, class\n",
      "Delibus: No match, species\n",
      "Delibus: No match, genus\n",
      "Streptophyta X: No match, subdivision\n",
      "Archaeplastida: No match, supergroup\n",
      "Oligotrichida XX: No match, species\n",
      "Oligotrichida XX: No match, genus\n",
      "Haplosporidium littoralis: No match, species\n",
      "Oligotrichida X: No match, family\n",
      "MAST 4D: No match, species\n",
      "MAST 4D: No match, genus\n",
      "Chlamydomonadales X: No match, family\n",
      "Uroleptus gallina: No match, species\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "Dinophyceae XX: No match, family\n",
      "Choreotrichida Tintinnina: No match, order\n",
      "Siphonosphaera cyathina: No match, species\n",
      "Dinophyceae X: No match, order\n",
      "MAST 3I: No match, species\n",
      "MAST 3I: No match, genus\n",
      "Dino Group II Clade 30 X: No match, species\n",
      "Dino Group II Clade 30 X: No match, genus\n",
      "MAST 3: No match, family\n",
      "Maxillopoda X: No match, species\n",
      "Maxillopoda X: No match, genus\n",
      "Dino Group II Clade 30: No match, family\n",
      "Dino Group II: No match, order\n",
      "LOS7N I lineage: No match, family\n",
      "Caecitellaceae X: No match, species\n",
      "Caecitellaceae X: No match, genus\n",
      "Rhodococcus rhodochrous: No match, species\n",
      "Dino Group II Clade 4 X: No match, species\n",
      "Dino Group II Clade 4 X: No match, genus\n",
      "Oikopleura dioica: No match, species\n",
      "Haliphthorales X: No match, species\n",
      "Haliphthorales X: No match, genus\n",
      "Dino Group II Clade 4: No match, family\n",
      "Gigartinales X: No match, family\n",
      "Haliphthorales: No match, familyDino Group II: No match, order\n",
      "\n",
      "Chlorellales X: No match, family\n",
      "Peronosporomycetes X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "Lychnaspis E2 giltschii: No match, species\n",
      "Eustigmatophyceae XX: No match, family\n",
      "Dino Group II Clade 16 X: No match, species\n",
      "Dino Group II Clade 16 X: No match, genus\n",
      "Echinodermata XXX: No match, species\n",
      "Echinodermata XXX: No match, genus\n",
      "Gyrista: No match, subdivision\n",
      "Lychnaspis E2: No match, genus\n",
      "Eustigmatophyceae X: No match, order\n",
      "Dino Group II Clade 16: No match, family\n",
      "Echinodermata XX: No match, family\n",
      "Stramenopiles: No match, division\n",
      "Acantharea E2: No match, family\n",
      "Dino Group II: No match, order\n",
      "Echinodermata X: No match, order\n",
      "TSAR: No match, supergroup\n",
      "Acantharea E: No match, order\n",
      "Gregarinomorphea: No match, class\n",
      "Peronosporomycetes XXX: No match, species\n",
      "Peronosporomycetes XXX: No match, genus\n",
      "Parmales env 2 X: No match, species\n",
      "Parmales env 2 X: No match, genus\n",
      "Peronosporomycetes XX: No match, family\n",
      "Peronosporomycetes X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Microrhizoidea pickettheapsiorum: No match, species\n",
      "Parmales env 2: No match, family\n",
      "Ochromonadales clade VII X: No match, species\n",
      "Ochromonadales clade VII X: No match, genus\n",
      "Dino Group II Clade 1 X: No match, species\n",
      "Dino Group II Clade 1 X: No match, genus\n",
      "Microrhizoidea: No match, genus\n",
      "Ochromonadales clade VII: No match, family\n",
      "Dino Group II Clade 1: No match, family\n",
      "Dino Group II: No match, order\n",
      "Picozoa X: No match, subdivision\n",
      "Dino Group II Clade 33 X: No match, species\n",
      "Dino Group II Clade 33 X: No match, genus\n",
      "Dino Group II Clade 7 X: No match, species\n",
      "Dino Group II Clade 7 X: No match, genus\n",
      "Dino Group II Clade 7: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group II Clade 33: No match, family\n",
      "Dino Group II: No match, order\n",
      "Eupelagonemidae X: No match, species\n",
      "Eupelagonemidae X: No match, genus\n",
      "MAST 8: No match, family\n",
      "Rotifera XXX: No match, species\n",
      "Rotifera XXX: No match, genus\n",
      "Plagiopylea X: No match, order\n",
      "Sagenista X: No match, order\n",
      "Rotifera XX: No match, family\n",
      "RAD B X Group IVd X: No match, species\n",
      "RAD B X Group IVd X: No match, genus\n",
      "RAD B X Group IVd: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Rotifera X: No match, order\n",
      "Teratosphaeria: No match, genus\n",
      "Eupetalomonads XX: No match, species\n",
      "Eupetalomonads XX: No match, genus\n",
      "MAST 1B: No match, species\n",
      "MAST 1B: No match, genus\n",
      "Eupetalomonads X: No match, family\n",
      "MAST 1: No match, family\n",
      "Chloroparvula B2: No match, species\n",
      "Obazoa: No match, supergroup\n",
      "Gyrista XX: No match, order\n",
      "MAST 3: No match, family\n",
      "Gyrista X: No match, class\n",
      "Peronosporomycetes X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group I Clade 1 X: No match, species\n",
      "Dino Group I Clade 1 X: No match, genus\n",
      "Gyrista: No match, subdivision\n",
      "Dino Group I Clade 1: No match, family\n",
      "Stramenopiles: No match, division\n",
      "Bicoecea XXX: No match, species\n",
      "Bicoecea XXX: No match, genus\n",
      "Nemertea XX: No match, family\n",
      "Dino Group I: No match, order\n",
      "Blastocladiomycetes: No match, family\n",
      "TSAR: No match, supergroup\n",
      "Nemertea X: No match, orderBicoecea XX: No match, family\n",
      "\n",
      "RAD B X Group Ic X: No match, species\n",
      "RAD B X Group Ic X: No match, genus\n",
      "Blastocladiomycotina: No match, order\n",
      "Bicoecea X: No match, order\n",
      "Dino Group I Clade 4 X: No match, species\n",
      "Dino Group I Clade 4 X: No match, genus\n",
      "RAD B X Group Ic: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, classBlastocladiomycota: No match, class\n",
      "\n",
      "Ochromonadales clade VIII X: No match, speciesDino Group I Clade 4: No match, family\n",
      "\n",
      "Dino Group I: No match, orderOchromonadales clade VIII X: No match, genus\n",
      "\n",
      "Dino Group II Clade 26 X: No match, species\n",
      "Dino Group II Clade 26 X: No match, genus\n",
      "Ochromonadales clade VIII: No match, family\n",
      "Picozoa XXXXX: No match, species\n",
      "Picozoa XXXXX: No match, genus\n",
      "Dino Group II Clade 26: No match, family\n",
      "Dino Group II: No match, order\n",
      "Gamasiphis pulchellus: No match, species\n",
      "Peromelissa tricuspidata: No match, species\n",
      "Picozoa XXXX: No match, family\n",
      "Gamasiphis: No match, genus\n",
      "Haptophyta XXXXX: No match, species\n",
      "Haptophyta XXXXX: No match, genus\n",
      "Acantharea 4: No match, order\n",
      "Picozoa XXX: No match, order\n",
      "Dino Group II Clade 5 X: No match, species\n",
      "Dino Group II Clade 5 X: No match, genus\n",
      "Haptophyta XXXX: No match, family\n",
      "Picozoa XX: No match, class\n",
      "Cryptosporidium1: No match, genus\n",
      "Dino Group II Clade 5: No match, family\n",
      "Dino Group II: No match, order\n",
      "Haptophyta XXX: No match, order\n",
      "Picozoa X: No match, subdivision\n",
      "RAD B X Group IVb X: No match, species\n",
      "RAD B X Group IVb X: No match, genus\n",
      "Haptophyta XX: No match, class\n",
      "RAD B X Group IVb: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Haptophyta X: No match, subdivision\n",
      "Spumellaria 3 X: No match, species\n",
      "Spumellaria 3 X: No match, genus\n",
      "Spumellaria 3: No match, family\n",
      "Pyloniidae 3: No match, species\n",
      "Pyloniidae 3: No match, genus\n",
      "Acantharea 1 XX: No match, species\n",
      "Acantharea 1 XX: No match, genus\n",
      "Acantharea 1 X: No match, family\n",
      "Acantharea 1: No match, order\n",
      "Acantharea C: No match, order\n",
      "LC104 lineage X: No match, species\n",
      "LC104 lineage X: No match, genus\n",
      "LC104 lineage: No match, family\n",
      "Dolichomastigaceae A: No match, species\n",
      "Dolichomastigaceae A: No match, genus\n",
      "Blastodinium mangini: No match, species\n",
      "Chlorarachniophyceae X: No match, order\n",
      "Williopsis pratensis: No match, species\n",
      "MAST 10 X: No match, species\n",
      "MAST 10 X: No match, genus\n",
      "Williopsis: No match, genus\n",
      "MAST 10: No match, family\n",
      "Sagenista X: No match, order\n",
      "Heterocapsa nei rotundata: No match, species\n",
      "Dino Group II Clade 32 X: No match, species\n",
      "Dino Group II Clade 32 X: No match, genus\n",
      "Dino Group II Clade 32: No match, family\n",
      "Dino Group II: No match, order\n",
      "Annelida XX: No match, family\n",
      "Antetintinnidium mucicola: No match, species\n",
      "Annelida X: No match, order\n",
      "MAST 8A: No match, species\n",
      "MAST 8A: No match, genus\n",
      "Antetintinnidium: No match, genus\n",
      "MAST 8: No match, family\n",
      "Sagenista X: No match, order\n",
      "Telonemia Group 2 X: No match, species\n",
      "Telonemia Group 2 X: No match, genus\n",
      "Cnidaria X: No match, order\n",
      "Telonemia Group 2: No match, family\n",
      "Colpodea X: No match, orderAnnelida XX: No match, family\n",
      "\n",
      "Annelida X: No match, order\n",
      "Telonemia XXX: No match, order\n",
      "Telonemia XX: No match, class\n",
      "Centroplasthelida XXXXX: No match, species\n",
      "Centroplasthelida XXXXX: No match, genus\n",
      "Telonemia X: No match, subdivision\n",
      "Centroplasthelida XXXX: No match, family\n",
      "Leptophryidae: No match, family\n",
      "Phaeodarea: No match, class\n",
      "Telonemia: No match, division\n",
      "Centroplasthelida XXX: No match, order\n",
      "TSAR: No match, supergroup\n",
      "Craniata X: No match, order\n",
      "Centroplasthelida XX: No match, class\n",
      "Mariager Fjord lineage X: No match, species\n",
      "Mariager Fjord lineage X: No match, genus\n",
      "Centroplasthelida X: No match, subdivision\n",
      "Mariager Fjord lineage: No match, family\n",
      "Mataza lineage X: No match, species\n",
      "Mataza lineage X: No match, genus\n",
      "Centroplasthelida: No match, division\n",
      "Suessiaceae X: No match, species\n",
      "Suessiaceae X: No match, genus\n",
      "Variosea X: No match, order\n",
      "Mataza lineage: No match, family\n",
      "RAD B X Group Ie X: No match, species\n",
      "RAD B X Group Ie X: No match, genus\n",
      "Filosa Thecofilosea X: No match, order\n",
      "Dino Group II Clade 3 X: No match, species\n",
      "Dino Group II Clade 3 X: No match, genus\n",
      "RAD B X Group Ie: No match, family\n",
      "Filosa Thecofilosea: No match, class\n",
      "Dipodomys ordii: No match, species\n",
      "Dino Group II Clade 3: No match, family\n",
      "Dino Group II: No match, order\n",
      "Micromonas bravo B2: No match, species\n",
      "RAD B X: No match, order\n",
      "Dipodomys: No match, genus\n",
      "Dino Group II Clade 27 X: No match, species\n",
      "Dino Group II Clade 27 X: No match, genus\n",
      "Thraustochytriaceae HK10: No match, species\n",
      "Thraustochytriaceae HK10: No match, genus\n",
      "RAD B: No match, class\n",
      "Hemiselmis aquamarina: No match, species\n",
      "Dino Group II Clade 27: No match, family\n",
      "Dino Group II: No match, order\n",
      "Cryptomonadales:nucl: No match, order\n",
      "Gigartacon C1 muelleri: No match, species\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Gigartacon C1: No match, genus\n",
      "Goussia6 janae: No match, species\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "Acantharea C1: No match, family\n",
      "Goussia6: No match, genus\n",
      "Haptophyta X: No match, subdivision\n",
      "Cryptophyta:nucl: No match, division\n",
      "Acantharea C: No match, order\n",
      "Cryptista:nucl: No match, supergroup\n",
      "Liosphaeroidea XX: No match, species\n",
      "Liosphaeroidea XX: No match, genus\n",
      "Stephanoecidae Group D X: No match, species\n",
      "Stephanoecidae Group D X: No match, genus\n",
      "Liosphaeroidea X: No match, family\n",
      "Stephanoecidae Group D: No match, family\n",
      "Discocephalus pararotatorius: No match, species\n",
      "Nusuttodinium desymbiontum: No match, species\n",
      "MAST 7C: No match, species\n",
      "MAST 7C: No match, genus\n",
      "Gyrista: No match, subdivision\n",
      "MAST 7: No match, familyPelagomonadaceae clade D1: No match, species\n",
      "\n",
      "Sagenista X: No match, orderPelagomonadaceae clade D1: No match, genus\n",
      "\n",
      "Rogersella griseliniae: No match, species\n",
      "Coccidiomorphea: No match, class\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Prymnesiophyceae Clade B3 X: No match, species\n",
      "Prymnesiophyceae Clade B3 X: No match, genus\n",
      "Rogersella: No match, genus\n",
      "Nebulomonas marisrubri: No match, species\n",
      "Dino Group I Clade 5 X: No match, species\n",
      "Dino Group I Clade 5 X: No match, genus\n",
      "Prymnesiophyceae Clade B3: No match, family\n",
      "Nebulomonas: No match, genus\n",
      "Dino Group I Clade 5: No match, family\n",
      "Dino Group I: No match, order\n",
      "Lophophaenidae: No match, family\n",
      "Nebulidae: No match, family\n",
      "Prasino Clade 9 XXX: No match, species\n",
      "Prasino Clade 9 XXX: No match, genus\n",
      "Nebulidida: No match, order\n",
      "Prasino Clade 9 XX: No match, family\n",
      "Haptophyta X: No match, subdivision\n",
      "Nebulidea: No match, class\n",
      "Prasino Clade 9 X: No match, order\n",
      "Pseudoplatyophrya nana: No match, species\n",
      "Thraustochytriaceae LAB17: No match, species\n",
      "Thraustochytriaceae LAB17: No match, genus\n",
      "Stephanoecidae Group E X: No match, species\n",
      "Stephanoecidae Group E X: No match, genus\n",
      "Olpidiopsis porphyrae: No match, species\n",
      "Prasino Clade 9: No match, class\n",
      "Nebulidia X: No match, subdivision\n",
      "Pseudoplatyophrya: No match, genus\n",
      "Stephanoecidae Group E: No match, family\n",
      "Chlorophyta X: No match, subdivision\n",
      "Nebulidia: No match, division\n",
      "Provora: No match, supergroup\n",
      "Chlorellales X: No match, family\n",
      "Drosophila ananassae: No match, species\n",
      "Philaenus spumarius: No match, species\n",
      "Dino Group V XX: No match, species\n",
      "Dino Group V XX: No match, genus\n",
      "Nemastomatales X: No match, family\n",
      "Philaenus: No match, genus\n",
      "Apionsoma misakianum: No match, species\n",
      "Dino Group V X: No match, family\n",
      "Dino Group V: No match, order\n",
      "Corallinales X: No match, family\n",
      "Dino Group II Clade 57 X: No match, species\n",
      "Dino Group II Clade 57 X: No match, genus\n",
      "Diplonemida DSPDII X: No match, species\n",
      "Diplonemida DSPDII X: No match, genus\n",
      "Dino Group III XX: No match, species\n",
      "Dino Group III XX: No match, genus\n",
      "Dino Group II Clade 57: No match, family\n",
      "Diplonemida DSPDII: No match, family\n",
      "Dino Group III X: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group III: No match, order\n",
      "Bryozoa XX: No match, family\n",
      "RAD A X: No match, order\n",
      "MAST 1D: No match, species\n",
      "MAST 1D: No match, genus\n",
      "Bryozoa X: No match, order\n",
      "Chytridiales XX: No match, species\n",
      "Chytridiales XX: No match, genus\n",
      "RAD A: No match, class\n",
      "MAST 1: No match, family\n",
      "Chytridiales X: No match, family\n",
      "Gyrista XX: No match, orderMicromonas clade B4: No match, species\n",
      "\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Chaetoceros debilis 2: No match, species\n",
      "Bicosoeca vacillans: No match, species\n",
      "Crustomastigaceae AB: No match, species\n",
      "Crustomastigaceae AB: No match, genus\n",
      "Vermamoebidae: No match, family\n",
      "Conchoecia annelida: No match, species\n",
      "Conchoecia annelida: No match, genus\n",
      "Crustomastigaceae: No match, family\n",
      "Paraphysomonadales Clade EC1Ga X: No match, species\n",
      "Paraphysomonadales Clade EC1Ga X: No match, genus\n",
      "Annelida XX: No match, family\n",
      "Paraphysomonadales Clade EC1Ga: No match, family\n",
      "Annelida X: No match, order\n",
      "TSAR: No match, supergroup\n",
      "Breviata lineage X: No match, species\n",
      "Breviata lineage X: No match, genus\n",
      "Monosigidae Group B X: No match, species\n",
      "Monosigidae Group B X: No match, genus\n",
      "Breviata lineage: No match, family\n",
      "Monosigidae Group B: No match, family\n",
      "Spumellaria 6 X: No match, species\n",
      "Spumellaria 6 X: No match, genus\n",
      "Breviatea XXX: No match, order\n",
      "MAST 4C: No match, species\n",
      "MAST 4C: No match, genus\n",
      "Spumellaria 6: No match, family\n",
      "Breviatea XX: No match, class\n",
      "MAST 4: No match, family\n",
      "Alveolata XXXXX: No match, species\n",
      "Alveolata XXXXX: No match, genus\n",
      "Breviatea X: No match, subdivision\n",
      "Sagenista X: No match, order\n",
      "Alveolata XXXX: No match, family\n",
      "Alveolata XXX: No match, order\n",
      "Ctenophora XX: No match, family\n",
      "Obazoa: No match, supergroup\n",
      "MOCH 5 XXX: No match, species\n",
      "MOCH 5 XXX: No match, genus\n",
      "Alveolata XX: No match, class\n",
      "Ctenophora X: No match, order\n",
      "Prymnesiophyceae Clade E XX: No match, species\n",
      "Prymnesiophyceae Clade E XX: No match, genus\n",
      "MOCH 5 XX: No match, family\n",
      "Alveolata X: No match, subdivision\n",
      "MOCH 5 X: No match, order\n",
      "Prymnesiophyceae Clade E X: No match, family\n",
      "MOCH 5: No match, classPrymnesiophyceae Clade E: No match, order\n",
      "\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Sagenista X: No match, order\n",
      "Strobilidiidae X: No match, species\n",
      "Strobilidiidae X: No match, genus\n",
      "Haptophyta X: No match, subdivision\n",
      "Obazoa: No match, supergroup\n",
      "Hyphodontia rimosissima: No match, species\n",
      "Demodex: No match, genus\n",
      "Labyrinthulaceae ANT10 3: No match, species\n",
      "Labyrinthulaceae ANT10 3: No match, genus\n",
      "Acantharea 4a X: No match, species\n",
      "Acantharea 4a X: No match, genus\n",
      "Spongopylidae: No match, family\n",
      "Acantharea 4a: No match, family\n",
      "Hexaconus serratus: No match, species\n",
      "Cephalopoda X: No match, family\n",
      "Acantharea 4: No match, order\n",
      "Dino Group I Clade 7 X: No match, species\n",
      "Dino Group I Clade 7 X: No match, genus\n",
      "Hexaconus: No match, genus\n",
      "Dino Group I Clade 7: No match, family\n",
      "Dino Group I: No match, order\n",
      "Acantharea E4: No match, family\n",
      "Acantharea E: No match, order\n",
      "Microhelida XX: No match, species\n",
      "Microhelida XX: No match, genus\n",
      "MAST 8C: No match, species\n",
      "MAST 8C: No match, genus\n",
      "MAST 8: No match, family\n",
      "Sagenista X: No match, order\n",
      "Florenciellales X: No match, species\n",
      "Florenciellales X: No match, genus\n",
      "MAST 3B: No match, species\n",
      "MAST 3B: No match, genus\n",
      "MAST 7E: No match, species\n",
      "MAST 7E: No match, genus\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "Microhelida X: No match, family\n",
      "MAST 3: No match, family\n",
      "Melkoniana: No match, species\n",
      "Melkoniana: No match, genus\n",
      "Microhelida: No match, order\n",
      "MAST 7: No match, familyAdelina grylli: No match, species\n",
      "\n",
      "Sagenista X: No match, order\n",
      "RAD B X Group IV XX: No match, species\n",
      "RAD B X Group IV XX: No match, genus\n",
      "RAD B X Group IV X: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Gorilla gorilla: No match, species\n",
      "Acantharea F3: No match, family\n",
      "Ventricleftida XX: No match, species\n",
      "Ventricleftida XX: No match, genus\n",
      "Gorilla: No match, genus\n",
      "Acantharea F: No match, order\n",
      "Ventricleftida X: No match, family\n",
      "Chlorellales X: No match, family\n",
      "Craniata XX: No match, family\n",
      "Craniata X: No match, order\n",
      "Papiliotrema laurentii: No match, species\n",
      "RAD A Xb X: No match, species\n",
      "RAD A Xb X: No match, genus\n",
      "Dino Group II Clade 35 X: No match, species\n",
      "Dino Group II Clade 35 X: No match, genus\n",
      "RAD A Xb: No match, family\n",
      "Dino Group II Clade 39 X: No match, species\n",
      "Dino Group II Clade 39 X: No match, genus\n",
      "Cryptomonas:nucl: No match, genus\n",
      "Dino Group II Clade 35: No match, family\n",
      "Dino Group II: No match, order\n",
      "RAD A X: No match, order\n",
      "Dino Group II Clade 39: No match, family\n",
      "Dino Group II: No match, order\n",
      "Cryptomonadales X:nucl: No match, family\n",
      "RAD A: No match, class\n",
      "Micromonas clade B5: No match, species\n",
      "Ventrifissuridae X: No match, species\n",
      "Ventrifissuridae X: No match, genus\n",
      "Cryptomonadales:nucl: No match, order\n",
      "MAST 4B: No match, species\n",
      "MAST 4B: No match, genus\n",
      "MAST 4: No match, familyMAST 7D: No match, species\n",
      "\n",
      "Sagenista X: No match, orderMAST 7D: No match, genus\n",
      "\n",
      "Stramenopiles: No match, divisionMAST 7: No match, family\n",
      "\n",
      "TSAR: No match, supergroupSagenista X: No match, order\n",
      "\n",
      "Cryptophyceae:nucl: No match, class\n",
      "MAST 22D: No match, species\n",
      "MAST 22D: No match, genus\n",
      "Cannobotryidae X: No match, species\n",
      "Cannobotryidae X: No match, genus\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "MAST 22: No match, family\n",
      "Cryptophyta:nucl: No match, division\n",
      "Prymnesiales X: No match, family\n",
      "Monosigidae Group A X: No match, speciesBicosoecida: No match, order\n",
      "\n",
      "Monosigidae Group A X: No match, genus\n",
      "Phyllostaurus B2 echinoides: No match, species\n",
      "Cryptista:nucl: No match, supergroup\n",
      "Dino Group II Clade 6 X: No match, species\n",
      "Dino Group II Clade 6 X: No match, genus\n",
      "Monosigidae Group A: No match, family\n",
      "Phyllostaurus B2: No match, genus\n",
      "Dino Group II Clade 6: No match, family\n",
      "Dino Group II: No match, order\n",
      "Acantharea B2: No match, family\n",
      "Gastrotricha XX: No match, family\n",
      "Knufia cryptophialidica: No match, species\n",
      "Acantharea B: No match, order\n",
      "Gastrotricha X: No match, order\n",
      "MAST 7B: No match, species\n",
      "MAST 7B: No match, genus\n",
      "MAST 1A: No match, species\n",
      "MAST 1A: No match, genus\n",
      "MAST 1: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "Dactylodinium arachnoides: No match, species\n",
      "Minorisa minuta: No match, species\n",
      "MAST 12: No match, family\n",
      "Kathablepharidida XX: No match, species\n",
      "Kathablepharidida XX: No match, genus\n",
      "Opalozoa X: No match, orderMinorisa: No match, genus\n",
      "\n",
      "Kathablepharidida X: No match, family\n",
      "Minorisa lineage: No match, family\n",
      "Stramenopiles X Group 7 XXX: No match, species\n",
      "Stramenopiles X Group 7 XXX: No match, genus\n",
      "Kathablepharidida: No match, order\n",
      "Stramenopiles X Group 7 XX: No match, family\n",
      "Chlorarachnida: No match, order\n",
      "Kathablepharidea: No match, class\n",
      "Stramenopiles X Group 7 X: No match, order\n",
      "Hemichordata XX: No match, family\n",
      "Pseudo nitzschia: No match, species\n",
      "Pseudo nitzschia: No match, genus\n",
      "Kathablepharida: No match, subdivision\n",
      "Stramenopiles X Group 7: No match, class\n",
      "Hemichordata X: No match, order\n",
      "Chaetoceros brevis 2: No match, species\n",
      "Kathablepharidacea: No match, division\n",
      "Stramenopiles X: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Exobasidiomycetes X: No match, species\n",
      "Exobasidiomycetes X: No match, genus\n",
      "Pseudo nitzschia: No match, genus\n",
      "Exobasidiomycetes: No match, family\n",
      "Rhinomonas nottbecki: No match, species\n",
      "Labyrinthulomycetes LAB15 X: No match, species\n",
      "Labyrinthulomycetes LAB15 X: No match, genus\n",
      "Sphaeropleales X: No match, family\n",
      "Labyrinthulomycetes LAB15: No match, family\n",
      "Clausidium vancouverense: No match, species\n",
      "Rhodomonas:nucl: No match, species\n",
      "Rhodomonas:nucl: No match, genus\n",
      "Acantharea F: No match, order\n",
      "Pyrenomonadaceae:nucl: No match, family\n",
      "Dino Group I Clade 2 X: No match, species\n",
      "Dino Group I Clade 2 X: No match, genus\n",
      "Pyrenomonadales:nucl: No match, order\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "Cryptophyta:nucl: No match, division\n",
      "Cryptista:nucl: No match, supergroup\n",
      "RAD A Xe X: No match, species\n",
      "RAD A Xe X: No match, genus\n",
      "Dino Group I Clade 2: No match, family\n",
      "Dino Group I: No match, order\n",
      "RAD A Xe: No match, family\n",
      "RAD A X: No match, order\n",
      "RAD A: No match, class\n",
      "Paratetrahymena parawassi: No match, species\n",
      "Ostreococcus clade B: No match, species\n",
      "Dino Group II Clade 22 X: No match, species\n",
      "Dino Group II Clade 22 X: No match, genus\n",
      "Malassezia sympodialis: No match, species\n",
      "Dino Group II Clade 22: No match, family\n",
      "Dino Group II: No match, order\n",
      "Acantharea 4d X: No match, species\n",
      "Acantharea 4d X: No match, genus\n",
      "Malassezia: No match, genus\n",
      "MAST 12B: No match, species\n",
      "MAST 12B: No match, genus\n",
      "MAST 12: No match, family\n",
      "Opalozoa X: No match, order\n",
      "Acantharea 4d: No match, family\n",
      "Exobasidiomycetes: No match, family\n",
      "Micromonas commoda A1: No match, species\n",
      "Acantharea 4: No match, order\n",
      "Prymnesiophyceae Clade D XX: No match, species\n",
      "Prymnesiophyceae Clade D XX: No match, genus\n",
      "Prymnesiophyceae Clade D X: No match, family\n",
      "Chaetoceros pumilum: No match, species\n",
      "Prymnesiophyceae Clade D: No match, order\n",
      "Haptophyta X: No match, subdivision\n",
      "Brassica rapa: No match, species\n",
      "Mediocremonas mediterraneus: No match, species\n",
      "Mediocremonas: No match, genus\n",
      "Ctenophora XXX: No match, species\n",
      "Ctenophora XXX: No match, genus\n",
      "Developea XX: No match, family\n",
      "Ctenophora XX: No match, family\n",
      "Pedinellales X: No match, species\n",
      "Pedinellales X: No match, genus\n",
      "Developea X: No match, order\n",
      "Ctenophora X: No match, order\n",
      "Developea: No match, class\n",
      "Obazoa: No match, supergroup\n",
      "Gyrista: No match, subdivision\n",
      "Perkinsida XXX: No match, species\n",
      "Perkinsida XXX: No match, genus\n",
      "Stramenopiles: No match, division\n",
      "Perkinsida XX: No match, family\n",
      "TSAR: No match, supergroup\n",
      "Dictyochophyceae X: No match, order\n",
      "Perkinsida X: No match, order\n",
      "Symbiontida XXX: No match, species\n",
      "Symbiontida XXX: No match, genus\n",
      "Voromonas pontica: No match, species\n",
      "Bacillariophyceae XXX: No match, species\n",
      "Bacillariophyceae XXX: No match, genus\n",
      "Symbiontida XX: No match, family\n",
      "Voromonas: No match, genus\n",
      "Bacillariophyceae XX: No match, family\n",
      "Symbiontida X: No match, orderConchellium capsula: No match, species\n",
      "\n",
      "Bacillariophyceae X: No match, order\n",
      "Colpodellaceae: No match, family\n",
      "Symbiontida: No match, class\n",
      "Conchellium: No match, genus\n",
      "Conchariidae: No match, family\n",
      "Malassezia globosa: No match, species\n",
      "Fibrophrys: No match, species\n",
      "Fibrophrys: No match, genus\n",
      "Malassezia: No match, genus\n",
      "Amphifilaceae: No match, family\n",
      "Thoracosphaeraceae X: No match, species\n",
      "Thoracosphaeraceae X: No match, genus\n",
      "Dino Group II Clade 2 X: No match, species\n",
      "Dino Group II Clade 2 X: No match, genus\n",
      "Exobasidiomycetes: No match, family\n",
      "Dino Group II Clade 2: No match, family\n",
      "Dino Group II: No match, order\n",
      "Haptophyta Clade HAP4 XXX: No match, species\n",
      "Haptophyta Clade HAP4 XXX: No match, genus\n",
      "Appendicularia X: No match, species\n",
      "Appendicularia X: No match, genus\n",
      "Haptophyta Clade HAP4 XX: No match, family\n",
      "Mediophyceae XXX: No match, species\n",
      "Mediophyceae XXX: No match, genus\n",
      "Haptophyta Clade HAP4 X: No match, orderChaetoceros muellerii: No match, species\n",
      "\n",
      "Mediophyceae XX: No match, family\n",
      "Urochordata X: No match, order\n",
      "Rhizopus oryzae: No match, species\n",
      "Haptophyta Clade HAP4: No match, class\n",
      "Mediophyceae X: No match, order\n",
      "Haptophyta X: No match, subdivision\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Ebriacea: No match, order\n",
      "Obazoa: No match, supergroup\n",
      "Filosa Imbricatea XXX: No match, species\n",
      "Filosa Imbricatea XXX: No match, genus\n",
      "Filosa Thecofilosea: No match, class\n",
      "Filosa Imbricatea XX: No match, family\n",
      "Metasychis disparidentata: No match, species\n",
      "RAD A Xc X: No match, species\n",
      "RAD A Xc X: No match, genus\n",
      "Filosa Imbricatea X: No match, order\n",
      "RAD A Xc: No match, family\n",
      "Filosa Imbricatea: No match, classRAD A X: No match, order\n",
      "\n",
      "RAD A: No match, class\n",
      "Dino Group II Clade 17 X: No match, species\n",
      "Dino Group II Clade 17 X: No match, genus\n",
      "Dino Group II Clade 17: No match, family\n",
      "Dino Group II: No match, order\n",
      "RAD C Xa X: No match, species\n",
      "RAD C Xa X: No match, genus\n",
      "RAD C Xa: No match, family\n",
      "MAST 11 X: No match, species\n",
      "MAST 11 X: No match, genus\n",
      "RAD C X: No match, order\n",
      "MAST 11: No match, family\n",
      "Sagenista X: No match, order\n",
      "RAD C: No match, class\n",
      "MAST 2: No match, family\n",
      "Dino Group II Clade 10 and 11 X: No match, species\n",
      "Dino Group II Clade 10 and 11 X: No match, genus\n",
      "Dino Group II Clade 13 X: No match, species\n",
      "Dino Group II Clade 13 X: No match, genus\n",
      "Gyrista XX: No match, order\n",
      "Dino Group II Clade 10 and 11: No match, family\n",
      "Dino Group II Clade 13: No match, familyDino Group II: No match, order\n",
      "\n",
      "Dino Group II: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Choreotrichida Tintinnina X: No match, family\n",
      "Choreotrichida Tintinnina: No match, order\n",
      "Trentepohliales X: No match, family\n",
      "Echinodermata XX: No match, family\n",
      "Echinodermata X: No match, order\n",
      "Chytridiomycota XXX: No match, species\n",
      "Chytridiomycota XXX: No match, genus\n",
      "Saccamminidae X: No match, species\n",
      "Saccamminidae X: No match, genus\n",
      "Dino Group II: No match, order\n",
      "Chytridiomycota XX: No match, family\n",
      "CCW10 lineage X: No match, species\n",
      "CCW10 lineage X: No match, genus\n",
      "RAD A Xa X: No match, species\n",
      "Chytridiomycota X: No match, orderRAD A Xa X: No match, genus\n",
      "\n",
      "CCW10 lineage: No match, family\n",
      "RAD A Xa: No match, family\n",
      "RAD A X: No match, order\n",
      "RAD A: No match, class\n",
      "Rhodotorula lamellibrachiae: No match, species\n",
      "Rhizosphaera: No match, species\n",
      "Rhizosphaera: No match, genus\n",
      "Rhizosphaeridae: No match, family\n",
      "Dino Group II Clade 20 X: No match, species\n",
      "Dino Group II Clade 20 X: No match, genus\n",
      "Dino Group II Clade 20: No match, family\n",
      "Dino Group II: No match, order\n",
      "Annelida XX: No match, family\n",
      "Annelida X: No match, order\n",
      "Vannella contorta: No match, species\n",
      "MAST 2C: No match, species\n",
      "MAST 2C: No match, genus\n",
      "MAST 2: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group II XX: No match, species\n",
      "Dino Group II XX: No match, genus\n",
      "Dino Group II X: No match, family\n",
      "Dino Group II: No match, order\n",
      "Syndiniales XXX: No match, species\n",
      "Syndiniales XXX: No match, genus\n",
      "Syndiniales XX: No match, family\n",
      "Syndiniales X: No match, order\n",
      "Spizellomycetales XX: No match, species\n",
      "Spizellomycetales XX: No match, genus\n",
      "Cryomonadida XX: No match, species\n",
      "Cryomonadida XX: No match, genus\n",
      "Spizellomycetales X: No match, family\n",
      "Cryomonadida X: No match, family\n",
      "Dino Group II Clade 4: No match, family\n",
      "Dino Group II: No match, order\n",
      "MAST 4E: No match, species\n",
      "MAST 4E: No match, genus\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "Chlorarachnida: No match, order\n",
      "Paralecudina polymorpha: No match, species\n",
      "MAST 3E: No match, species\n",
      "MAST 3E: No match, genus\n",
      "MAST 3: No match, family\n",
      "Paralecudina: No match, genus\n",
      "Paralecudinids: No match, family\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "RAD B X Group II X: No match, species\n",
      "RAD B X Group II X: No match, genus\n",
      "RAD B X Group II: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Dino Group II Clade 9 X: No match, species\n",
      "Dino Group II Clade 9 X: No match, genus\n",
      "Dino Group II Clade 9: No match, family\n",
      "Dino Group II: No match, order\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Monosigidae Group A: No match, family\n",
      "Vexillifera tasmaniana: No match, species\n",
      "Cryptomonadales XX: No match, species\n",
      "Cryptomonadales XX: No match, genus\n",
      "Phalachroma: No match, genus\n",
      "Cryptomonadales X: No match, family\n",
      "Oxyphysiaceae: No match, family\n",
      "Rhodophyta X: No match, subdivision\n",
      "Stagonosporopsis cucurbitacearum: No match, species\n",
      "Excentroconchidae X: No match, species\n",
      "Excentroconchidae X: No match, genus\n",
      "Parvilucifera prorocentri: No match, species\n",
      "Excentroconchidae: No match, family\n",
      "Dino Group I Clade 3 X: No match, species\n",
      "Dino Group I Clade 3 X: No match, genus\n",
      "Cercozoa XXXX: No match, species\n",
      "Cercozoa XXXX: No match, genus\n",
      "MAST 2D: No match, species\n",
      "MAST 2D: No match, genus\n",
      "Dino Group I Clade 3: No match, family\n",
      "Dino Group I: No match, order\n",
      "Cercozoa XXX: No match, family\n",
      "MAST 2: No match, family\n",
      "MAST 1C: No match, species\n",
      "MAST 1C: No match, genus\n",
      "MAST 1: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Cercozoa XX: No match, order\n",
      "Gyrista XX: No match, order\n",
      "Dino Group II Clade 29 X: No match, species\n",
      "Dino Group II Clade 29 X: No match, genus\n",
      "Cercozoa X: No match, class\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group II Clade 29: No match, family\n",
      "Dino Group II: No match, order\n",
      "Chrysophyceae XXX: No match, species\n",
      "Chrysophyceae XXX: No match, genus\n",
      "Chrysophyceae XX: No match, family\n",
      "Chrysophyceae X: No match, order\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Dino Group II Clade 12 X: No match, species\n",
      "Dino Group II Clade 12 X: No match, genus\n",
      "Chaetoceros sp Clade Na26B1: No match, species\n",
      "Dino Group II Clade 12: No match, family\n",
      "Dino Group II: No match, order\n",
      "Chrysophyceae Clade EC2H XX: No match, species\n",
      "Chrysophyceae Clade EC2H XX: No match, genus\n",
      "Chrysophyceae Clade EC2H X: No match, family\n",
      "Chrysophyceae Clade EC2H: No match, order\n",
      "Perichlamydium venustum: No match, species\n",
      "Dinophyceae XXX: No match, species\n",
      "Dinophyceae XXX: No match, genus\n",
      "Perichlamydium: No match, genus\n",
      "Dinophyceae XX: No match, family\n",
      "Stylodictyidae: No match, family\n",
      "Acantharea E XX: No match, species\n",
      "Acantharea E XX: No match, genus\n",
      "Dinophyceae X: No match, order\n",
      "Frankliniella occidentalis: No match, species\n",
      "Acantharea E X: No match, family\n",
      "Dictyocoryne truncata: No match, species\n",
      "Acantharea E: No match, order\n",
      "Tricolocamptra aff urnula: No match, species\n",
      "Tricolocamptra: No match, genus\n",
      "MAST 7A: No match, species\n",
      "MAST 7A: No match, genus\n",
      "Paradinidae X: No match, species\n",
      "Paradinidae X: No match, genus\n",
      "Theopiliidae: No match, family\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "RAD B X Group If X: No match, species\n",
      "RAD B X Group If X: No match, genus\n",
      "Paradinidae: No match, family\n",
      "RAD B X Group If: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "RAD B X Group III X: No match, species\n",
      "RAD B X Group III X: No match, genus\n",
      "Gregarinomorphea X GRE1 XX: No match, species\n",
      "Gregarinomorphea X GRE1 XX: No match, genus\n",
      "RAD B X Group III: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Gregarinomorphea X GRE1 X: No match, family\n",
      "Apusomonadidae Group 1 XX: No match, species\n",
      "Apusomonadidae Group 1 XX: No match, genus\n",
      "Gregarinomorphea X GRE1: No match, order\n",
      "Apusomonadidae Group 1 X: No match, family\n",
      "Gregarinomorphea: No match, class\n",
      "Pelagodinium beii: No match, species\n",
      "Apusomonadidae Group 1: No match, order\n",
      "Chlorellales X: No match, family\n",
      "Bjerkandera adusta: No match, species\n",
      "Bjerkandera: No match, genus\n",
      "Parmales env 3 X: No match, species\n",
      "Parmales env 3 X: No match, genus\n",
      "Haptophyta Clade HAP2 XXX: No match, species\n",
      "Haptophyta Clade HAP2 XXX: No match, genus\n",
      "Pterocorys clausa: No match, species\n",
      "Haptophyta Clade HAP2 XX: No match, familyParmales env 3: No match, family\n",
      "\n",
      "Tunicaraptor unikontum: No match, species\n",
      "Haptophyta Clade HAP2 X: No match, order\n",
      "Tunicaraptor: No match, genus\n",
      "Haptophyta Clade HAP2: No match, class\n",
      "Haptophyta X: No match, subdivision\n",
      "Thaumatomonadida XX: No match, species\n",
      "Thaumatomonadida XX: No match, genus\n",
      "Pinus taeda: No match, species\n",
      "Opisthokonta XXXX: No match, family\n",
      "Haliommatidium: No match, species\n",
      "Haliommatidium: No match, genus\n",
      "Thaumatomonadida X: No match, family\n",
      "Opisthokonta XXX: No match, order\n",
      "Acantharea F3: No match, family\n",
      "Opisthokonta XX: No match, class\n",
      "Acantharea F: No match, order\n",
      "Protaspa lineage X: No match, species\n",
      "Protaspa lineage X: No match, genus\n",
      "Opisthokonta X: No match, subdivision\n",
      "Litholophus spB: No match, species\n",
      "Protaspa lineage: No match, family\n",
      "Chromadorea X: No match, family\n",
      "NPK2 lineage X: No match, species\n",
      "NPK2 lineage X: No match, genus\n",
      "Litholophus: No match, genus\n",
      "Obazoa: No match, supergroup\n",
      "NPK2 lineage: No match, family\n",
      "Chlorarachniophyceae X: No match, order\n",
      "Acantharea C3: No match, family\n",
      "Acantharea C: No match, order\n",
      "Rhodotorula toruloides: No match, species\n",
      "MAST 3F: No match, species\n",
      "MAST 3F: No match, genus\n",
      "MAST 3: No match, family\n",
      "Ochromonadales clade XII X: No match, species\n",
      "Ochromonadales clade XII X: No match, genus\n",
      "Ochromonadales clade XII: No match, family\n",
      "MOCH 4 XXX: No match, species\n",
      "MOCH 4 XXX: No match, genus\n",
      "Novel clade 2 X: No match, species\n",
      "Novel clade 2 X: No match, genus\n",
      "MOCH 4 XX: No match, family\n",
      "Novel clade 2: No match, family\n",
      "MOCH 4 X: No match, order\n",
      "Filosa Imbricatea X: No match, order\n",
      "Filosa Imbricatea: No match, class\n",
      "MOCH 4: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Radiolaria XXXX: No match, species\n",
      "Radiolaria XXXX: No match, genus\n",
      "Penicillium capsulatum: No match, species\n",
      "Radiolaria XXX: No match, family\n",
      "Derxomyces: No match, genus\n",
      "Radiolaria XX: No match, order\n",
      "Radiolaria X: No match, class\n",
      "Prasinoderma singulare: No match, species\n",
      "Dino Group II Clade 52 XX: No match, species\n",
      "Dino Group II Clade 52 XX: No match, genus\n",
      "Exobasidiomycetes: No match, family\n",
      "Euchitoniidae: No match, family\n",
      "Dino Group II X: No match, family\n",
      "Dino Group II: No match, order\n",
      "Hydrozoa X: No match, species\n",
      "Hydrozoa X: No match, genus\n",
      "Chattonaria mesnili: No match, species\n",
      "Chattonaria: No match, genus\n",
      "Centroplasthelida X: No match, subdivision\n",
      "Hematodinium perezi: No match, species\n",
      "Gregarinomorphea XX: No match, family\n",
      "Centroplasthelida: No match, division\n",
      "Gregarinomorphea X: No match, order\n",
      "Paralecudina anankea: No match, species\n",
      "Protaspa lineage: No match, family\n",
      "Gregarinomorphea: No match, class\n",
      "Paralecudina: No match, genus\n",
      "MAST 2 X: No match, species\n",
      "MAST 2 X: No match, genus\n",
      "MAST 2: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Paralecudinids: No match, family\n",
      "MAST 2B: No match, species\n",
      "MAST 2B: No match, genus\n",
      "MAST 2: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group II Clade 55 X: No match, species\n",
      "Dino Group II Clade 55 X: No match, genus\n",
      "Nassellaria 3 X: No match, species\n",
      "Nassellaria 3 X: No match, genus\n",
      "Dino Group II Clade 55: No match, family\n",
      "Dino Group II: No match, order\n",
      "Prymnesiophyceae Clade B5 X: No match, species\n",
      "Prymnesiophyceae Clade B5 X: No match, genus\n",
      "Nassellaria 3: No match, family\n",
      "Prymnesiophyceae Clade B5: No match, family\n",
      "Corallicolidae: No match, family\n",
      "Telonemia Group 1 X: No match, species\n",
      "Telonemia Group 1 X: No match, genus\n",
      "Ichthyobodonidae X: No match, species\n",
      "Ichthyobodonidae X: No match, genus\n",
      "Eimeriida: No match, order\n",
      "Telonemia Group 1: No match, family\n",
      "Telonemia XXX: No match, order\n",
      "Telonemia XX: No match, class\n",
      "Telonemia X: No match, subdivision\n",
      "Telonemia: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Strombidiidae X: No match, species\n",
      "Strombidiidae X: No match, genus\n",
      "Ichthyobodonidae: No match, family\n",
      "Coccidiomorphea: No match, class\n",
      "Opisthokonta XXXXX: No match, species\n",
      "Opisthokonta XXXXX: No match, genus\n",
      "Labyrinthulomycetes LAB6: No match, species\n",
      "Labyrinthulomycetes LAB6: No match, genus\n",
      "Opisthokonta XXXX: No match, family\n",
      "Labyrinthulomycetes LAB1 6 8: No match, family\n",
      "Opisthokonta XXX: No match, order\n",
      "Dino Group I: No match, order\n",
      "Opisthokonta XX: No match, class\n",
      "RAD B X Group Ib X: No match, species\n",
      "RAD B X Group Ib X: No match, genus\n",
      "Opisthokonta X: No match, subdivision\n",
      "Obazoa: No match, supergroup\n",
      "RAD B X Group Ib: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Dipodascus ingens: No match, species\n",
      "Dino Group II Clade 21 X: No match, species\n",
      "Dino Group II Clade 21 X: No match, genus\n",
      "MAST 6 X: No match, species\n",
      "MAST 6 X: No match, genus\n",
      "Dino Group II Clade 21: No match, family\n",
      "Dino Group II: No match, order\n",
      "Phaeophyceae XX: No match, family\n",
      "MAST 6: No match, family\n",
      "OLIGO5 XX: No match, species\n",
      "OLIGO5 XX: No match, genus\n",
      "Urochordata X: No match, order\n",
      "Obazoa: No match, supergroupPhaeophyceae X: No match, order\n",
      "\n",
      "Pseudophyllomitidae: No match, order\n",
      "OLIGO5 X: No match, family\n",
      "Hexalonchidae: No match, family\n",
      "OLIGO5: No match, order\n",
      "Polycladida X: No match, species\n",
      "Polycladida X: No match, genus\n",
      "Dino Group I XX: No match, species\n",
      "Dino Group I XX: No match, genus\n",
      "Dino Group II Clade 44 X: No match, species\n",
      "Dino Group II Clade 44 X: No match, genus\n",
      "Favellidae: No match, family\n",
      "Choreotrichida Tintinnina: No match, order\n",
      "Dino Group I X: No match, family\n",
      "Dino Group I: No match, order\n",
      "Dino Group II Clade 44: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group II Clade 38 X: No match, species\n",
      "Dino Group II Clade 38 X: No match, genus\n",
      "Chromadorea X: No match, family\n",
      "Dino Group I Clade 8 X: No match, species\n",
      "Dino Group I Clade 8 X: No match, genus\n",
      "Dino Group II Clade 38: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group I Clade 8: No match, family\n",
      "Dino Group I: No match, order\n",
      "Anurofeca: No match, species\n",
      "Anurofeca: No match, genus\n",
      "Selaginopsis cornigera: No match, species\n",
      "Dino Group I Clade 6 X: No match, species\n",
      "Dino Group I Clade 6 X: No match, genus\n",
      "Streptophyta X: No match, subdivision\n",
      "Pseudoperkinsidae X: No match, family\n",
      "Pectinoida: No match, family\n",
      "Dino Group I Clade 6: No match, family\n",
      "Dino Group I: No match, order\n",
      "Pseudoperkinsidae: No match, order\n",
      "Ichthyophonida: No match, class\n",
      "Dinophyta XXXX: No match, species\n",
      "Dinophyta XXXX: No match, genus\n",
      "Monothalamids: No match, class\n",
      "Dinophyta XXX: No match, family\n",
      "Sey055 lineage X: No match, species\n",
      "Sey055 lineage X: No match, genus\n",
      "Dinophyta XX: No match, order\n",
      "Sey055 lineage: No match, family\n",
      "TAGIRI1 lineage X: No match, species\n",
      "TAGIRI1 lineage X: No match, genus\n",
      "Dino Group II Clade 15 X: No match, species\n",
      "Dino Group II Clade 15 X: No match, genus\n",
      "Dino Group IV: No match, order\n",
      "Dinophyta X: No match, class\n",
      "TAGIRI1 lineage: No match, family\n",
      "Ebriacea: No match, order\n",
      "Filosa Thecofilosea: No match, class\n",
      "Dino Group II Clade 15: No match, family\n",
      "Dino Group II: No match, order\n",
      "Ustilaginomycetes X: No match, species\n",
      "Ustilaginomycetes X: No match, genus\n",
      "Pichia membranifaciens: No match, species\n",
      "RAD B X Group Ia X: No match, species\n",
      "RAD B X Group Ia X: No match, genus\n",
      "MAST 3K: No match, species\n",
      "MAST 3K: No match, genus\n",
      "MAST 3: No match, family\n",
      "Ichthyobodonidae: No match, family\n",
      "RAD B X Group Ia: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Tintinnidae X: No match, species\n",
      "Tintinnidae X: No match, genus\n",
      "Lophophaenidae X: No match, species\n",
      "Lophophaenidae X: No match, genus\n",
      "Lophophaenidae: No match, family\n",
      "Stramenopiles XXXXX: No match, species\n",
      "Stramenopiles XXXXX: No match, genus\n",
      "Stramenopiles XXXX: No match, family\n",
      "Dino Group I Clade 4: No match, family\n",
      "Cryptomonadales X:nucl: No match, family\n",
      "Stramenopiles XXX: No match, order\n",
      "Dino Group I: No match, order\n",
      "Cryptomonadales:nucl: No match, order\n",
      "Watanabea Clade X: No match, family\n",
      "Stramenopiles XX: No match, class\n",
      "RAD B X Group IVc X: No match, species\n",
      "RAD B X Group IVc X: No match, genus\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Watanabea Clade: No match, order\n",
      "Stramenopiles X: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "RAD B X Group IVc: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Cryptophyta:nucl: No match, division\n",
      "Cryptista:nucl: No match, supergroup\n",
      "Chlorodendrales X: No match, species\n",
      "MAST 9A: No match, species\n",
      "MAST 9A: No match, genus\n",
      "Chlorodendrales XX: No match, genus\n",
      "MAST 9: No match, family\n",
      "Sagenista X: No match, order\n",
      "Porosporidae X: No match, species\n",
      "Porosporidae X: No match, genus\n",
      "Dino Group II Clade 8 X: No match, species\n",
      "Dino Group II Clade 8 X: No match, genus\n",
      "Stephanoecidae Group I: No match, family\n",
      "233780: Unaccepted, using Chattonella marina var. ovata\n",
      "Dino Group II Clade 8: No match, family\n",
      "Dino Group II: No match, order\n",
      "MAST 4F: No match, species\n",
      "MAST 4F: No match, genus\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "Sarcinochrysidaceae X: No match, species\n",
      "Sarcinochrysidaceae X: No match, genus\n",
      "Pterocystida XXX: No match, species\n",
      "Pterocystida XXX: No match, genus\n",
      "MAST 3J: No match, species\n",
      "MAST 3J: No match, genus\n",
      "MAST 3: No match, family\n",
      "Pterocystida XX: No match, family\n",
      "Echinamoeba exundans: No match, species\n",
      "Pterocystida X: No match, order\n",
      "MAST 4A: No match, speciesDino Group I Clade 3: No match, family\n",
      "\n",
      "MAST 4A: No match, genusDino Group I: No match, order\n",
      "\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "Dino Group II Clade 18 X: No match, species\n",
      "Dino Group II Clade 18 X: No match, genus\n",
      "Liburna glaciale: No match, species\n",
      "Paracoccidioides: No match, genus\n",
      "Dino Group II Clade 18: No match, family\n",
      "Dino Group II: No match, order\n",
      "115104: Unaccepted, using Gephyrocapsa huxleyi\n",
      "MAST 8D: No match, species\n",
      "MAST 8D: No match, genus\n",
      "MAST 8: No match, family\n",
      "Sagenista X: No match, order\n",
      "Galactomyces: No match, species\n",
      "Galactomyces: No match, genus\n",
      "Dino Group II Clade 28 X: No match, species\n",
      "Dino Group II Clade 28 X: No match, genus\n",
      "Dino Group II Clade 28: No match, family\n",
      "Dino Group II: No match, order\n",
      "Chrysomerophyceae XX: No match, family\n",
      "Dino Group II Clade 14 X: No match, species\n",
      "Dino Group II Clade 14 X: No match, genus\n",
      "Chrysomerophyceae X: No match, order\n",
      "Dino Group II Clade 14: No match, family\n",
      "Dino Group II: No match, order\n",
      "Malassezia slooffiae: No match, species\n",
      "MAST 3C: No match, species\n",
      "MAST 3C: No match, genus\n",
      "MAST 3: No match, family\n",
      "Malassezia: No match, genus\n",
      "Basal Cryptophyceae 2 X: No match, species\n",
      "Basal Cryptophyceae 2 X: No match, genus\n",
      "Ceraceomyces americanus: No match, species\n",
      "Exobasidiomycetes: No match, family\n",
      "Basal Cryptophyceae 2: No match, family\n",
      "RAD C Xb X: No match, species\n",
      "RAD C Xb X: No match, genus\n",
      "Ceraceomyces: No match, genus\n",
      "Cryptophyceae X: No match, order\n",
      "RAD C Xb: No match, family\n",
      "RAD C X: No match, order\n",
      "RAD C: No match, class\n",
      "Ichthyophonus irregularis: No match, species\n",
      "MOCH 2 XXX: No match, species\n",
      "MOCH 2 XXX: No match, genus\n",
      "Haliphthorales: No match, family\n",
      "MOCH 2 XX: No match, family\n",
      "Peronosporomycetes X: No match, order\n",
      "MOCH 2 X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "MOCH 2: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Gyrista: No match, subdivision\n",
      "Eugregarinorida EUG1 X: No match, species\n",
      "Eugregarinorida EUG1 X: No match, genus\n",
      "Stramenopiles: No match, division\n",
      "Chrysophyceae Clade EC2I XX: No match, species\n",
      "Chrysophyceae Clade EC2I XX: No match, genus\n",
      "Eugregarinorida EUG1: No match, family\n",
      "TSAR: No match, supergroup\n",
      "Chrysophyceae Clade EC2I X: No match, family\n",
      "Acanthoecida XX: No match, species\n",
      "Acanthoecida XX: No match, genus\n",
      "Chrysophyceae Clade EC2I: No match, order\n",
      "Acanthoecida X: No match, family\n",
      "Amphifilaceae: No match, family\n",
      "Embryophyceae XX: No match, family\n",
      "Embryophyceae X: No match, order\n",
      "Embryophyceae: No match, class\n",
      "Streptophyta X: No match, subdivision\n",
      "Paradinidae: No match, family\n",
      "Chaunacanthid: No match, species\n",
      "Chaunacanthid: No match, genus\n",
      "Acantharea C4: No match, family\n",
      "Acantharea C: No match, order\n",
      "Stigmostylus ferrusi: No match, species\n",
      "RAD A Xg X: No match, species\n",
      "RAD A Xg X: No match, genus\n",
      "Stigmostylus: No match, genus\n",
      "RAD A Xg: No match, family\n",
      "RAD A X: No match, order\n",
      "RAD A: No match, class\n",
      "Acantharea C3: No match, family\n",
      "Acantharea C: No match, order\n",
      "Hexalonchidae: No match, family\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Telonemia Group 2: No match, family\n",
      "Telonemia XXX: No match, order\n",
      "Telonemia XX: No match, class\n",
      "Spumellaria 1 X: No match, species\n",
      "Spumellaria 1 X: No match, genus\n",
      "Telonemia X: No match, subdivision\n",
      "Spumellaria 1: No match, family\n",
      "Chlorophyta X: No match, subdivision\n",
      "Telonemia: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Spumellaria 4b X: No match, species\n",
      "Spumellaria 4b X: No match, genus\n",
      "Spumellaria 4b: No match, family\n",
      "CONThreeP XXX: No match, species\n",
      "CONThreeP XXX: No match, genus\n",
      "CONThreeP XX: No match, family\n",
      "CONThreeP X: No match, order\n",
      "Spongopylidae: No match, family\n",
      "CONThreeP: No match, class\n",
      "Cryptomonadales X: No match, family\n",
      "Echinamoeba silvestris: No match, species\n",
      "Kodamaea ohmeri: No match, species\n",
      "Kodamaea: No match, genus\n",
      "Eutintinnus fraknoi: No match, species\n",
      "MAST 8B: No match, species\n",
      "MAST 8B: No match, genus\n",
      "MAST 8: No match, family\n",
      "Sagenista X: No match, order\n",
      "Dino Group II Clade 40 X: No match, species\n",
      "Dino Group II Clade 40 X: No match, genus\n",
      "Dino Group II Clade 40: No match, family\n",
      "Dino Group II: No match, order\n",
      "Stephanoecidae Group H X: No match, species\n",
      "Stephanoecidae Group H X: No match, genus\n",
      "Stephanoecidae Group H: No match, family\n",
      "MAST 3D: No match, species\n",
      "MAST 3D: No match, genus\n",
      "MAST 3: No match, family\n",
      "RAD B X Group IVa X: No match, species\n",
      "RAD B X Group IVa X: No match, genus\n",
      "RAD B X Group IVa: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Echinodermata XX: No match, family\n",
      "Margolisiellidae X: No match, species\n",
      "Margolisiellidae X: No match, genus\n",
      "Echinodermata X: No match, order\n",
      "Margolisiellidae: No match, family\n",
      "Marosporida: No match, order\n",
      "Coccidiomorphea: No match, class\n",
      "Colpodida X: No match, species\n",
      "Colpodida X: No match, genus\n",
      "Thraustochytriaceae QPX: No match, species\n",
      "Thraustochytriaceae QPX: No match, genus\n",
      "Telonemia Group 1: No match, family\n",
      "Telonemia XXX: No match, order\n",
      "Chaetoceros curvisetus 3: No match, species\n",
      "Telonemia XX: No match, class\n",
      "Telonemia X: No match, subdivision\n",
      "Telonemia: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Malassezia: No match, genus\n",
      "Exobasidiomycetes: No match, family\n",
      "Astrosphaeridae X: No match, species\n",
      "Astrosphaeridae X: No match, genus\n",
      "Dino Group II Clade 23 X: No match, species\n",
      "Dino Group II Clade 23 X: No match, genus\n",
      "Dino Group II Clade 23: No match, family\n",
      "Dino Group II: No match, order\n",
      "Heliocephala: No match, genus\n",
      "Pezizomycotina X: No match, family\n",
      "Neoparamoeba branchiphila: No match, species\n",
      "Euglyphida XX: No match, species\n",
      "Euglyphida XX: No match, genus\n",
      "Euglyphida X: No match, family\n",
      "Dino Group II Clade 25 X: No match, species\n",
      "Dino Group II Clade 25 X: No match, genus\n",
      "Dino Group II Clade 25: No match, family\n",
      "Dino Group II: No match, order\n",
      "Chlorarachnida XX: No match, species\n",
      "Chlorarachnida XX: No match, genus\n",
      "Chlorarachnida X: No match, family\n",
      "Rhizosphaeridae: No match, family\n",
      "Chlorarachnida: No match, order\n",
      "RAD B X Group Id X: No match, species\n",
      "RAD B X Group Id X: No match, genus\n",
      "RAD B X Group Id: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Termitomyces heimii: No match, species\n",
      "Termitomyces: No match, genus\n",
      "Stephanoecidae Group I X: No match, species\n",
      "Stephanoecidae Group I X: No match, genus\n",
      "Stephanoecidae Group I: No match, family\n",
      "Leidyana2: No match, species\n",
      "Leidyana2: No match, genus\n",
      "Stylodictya subtilis: No match, species\n",
      "Scaphopoda X: No match, family\n",
      "Vannella ebro: No match, species\n",
      "Colpodella angusta: No match, species\n",
      "Halocrusticida baliensis: No match, species\n",
      "Halocrusticida: No match, genus\n",
      "Haliphthorales: No match, family\n",
      "Peronosporomycetes X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Pterista: No match, order\n",
      "Chroomonas:nucl: No match, species\n",
      "Chroomonas:nucl: No match, genus\n",
      "Hemiselmidaceae:nucl: No match, family\n",
      "Cryptomonadales:nucl: No match, order\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "Cryptophyta:nucl: No match, division\n",
      "Cryptista:nucl: No match, supergroup\n",
      "MAST 3 X: No match, species\n",
      "MAST 3 X: No match, genus\n",
      "MAST 3: No match, family\n",
      "Difficilina paramertis: No match, species\n",
      "Wickerhamomyces anomalus: No match, species\n",
      "Wickerhamomyces: No match, genus\n",
      "Leidyana2: No match, genus\n",
      "Ochroconis musae: No match, species\n",
      "Ochroconis: No match, genus\n",
      "Corallinales XX: No match, species\n",
      "Corallinales XX: No match, genus\n",
      "Corallinales X: No match, family\n",
      "Foraminifera XXXX: No match, species\n",
      "Foraminifera XXXX: No match, genus\n",
      "Foraminifera XXX: No match, family\n",
      "Foraminifera XX: No match, order\n",
      "Foraminifera X: No match, class\n",
      "Ochromonadales clade V X: No match, species\n",
      "Ochromonadales clade V X: No match, genus\n",
      "Ochromonadales clade V: No match, family\n",
      "Obazoa: No match, supergroup\n",
      "Filosa Thecofilosea: No match, class\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    worms_18s = WoRMS_matching.get_worms_from_aphiaid_or_name_parallel(\n",
    "    tax_df = tax_18S,worms_dict=pr2_18S_dict,ordered_rank_columns=['species','genus','family','order','class','subdivision','division','supergroup'],\n",
    "    full_tax_column=\"taxonomy\",full_tax_vI=True,n_proc=6)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;</td>\n",
       "      <td>Paraphysomonas</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:291417</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Chrysophyceae</td>\n",
       "      <td>Chromulinales</td>\n",
       "      <td>Paraphysomonadaceae</td>\n",
       "      <td>Paraphysomonas</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;</td>\n",
       "      <td>Gonyaulax polygramma</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:110035</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gonyaulacales</td>\n",
       "      <td>Gonyaulacaceae</td>\n",
       "      <td>Gonyaulax</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae</td>\n",
       "      <td>Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae</td>\n",
       "      <td>Mamiellaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:17663</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>Chlorophyta</td>\n",
       "      <td>Mamiellophyceae</td>\n",
       "      <td>Mamiellales</td>\n",
       "      <td>Mamiellaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;</td>\n",
       "      <td>Blastodinium galatheanum</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:623673</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Blastodiniales</td>\n",
       "      <td>Blastodinidae</td>\n",
       "      <td>Blastodinium</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:22565</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Subphylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                       full_tax  \\\n",
       "0  Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;   \n",
       "1              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;   \n",
       "2                                   Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae   \n",
       "3       Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;   \n",
       "4                                                                Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida   \n",
       "\n",
       "                                                                                                         verbatimIdentification  \\\n",
       "0  Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;   \n",
       "1              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;   \n",
       "2                                   Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae   \n",
       "3       Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;   \n",
       "4                                                                Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida   \n",
       "\n",
       "             scientificName                           scientificNameID  \\\n",
       "0            Paraphysomonas  urn:lsid:marinespecies.org:taxname:291417   \n",
       "1      Gonyaulax polygramma  urn:lsid:marinespecies.org:taxname:110035   \n",
       "2              Mamiellaceae   urn:lsid:marinespecies.org:taxname:17663   \n",
       "3  Blastodinium galatheanum  urn:lsid:marinespecies.org:taxname:623673   \n",
       "4               Apicomplexa   urn:lsid:marinespecies.org:taxname:22565   \n",
       "\n",
       "     kingdom       phylum            class           order  \\\n",
       "0  Chromista   Ochrophyta    Chrysophyceae   Chromulinales   \n",
       "1  Chromista      Myzozoa      Dinophyceae   Gonyaulacales   \n",
       "2    Plantae  Chlorophyta  Mamiellophyceae     Mamiellales   \n",
       "3  Chromista      Myzozoa      Dinophyceae  Blastodiniales   \n",
       "4  Chromista      Myzozoa             None            None   \n",
       "\n",
       "                family           genus  taxonRank  \n",
       "0  Paraphysomonadaceae  Paraphysomonas      Genus  \n",
       "1       Gonyaulacaceae       Gonyaulax    Species  \n",
       "2         Mamiellaceae            None     Family  \n",
       "3        Blastodinidae    Blastodinium    Species  \n",
       "4                 None            None  Subphylum  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which taxa had absolutely no matches\n",
    "worms_18s[worms_18s[\"scientificName\"]==\"No match\"]['old name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Haliphthorales_X;Haliphthorales_X_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Haliphthorales_X;Haliphthorales_X_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Peronosporomycetes_XX;Peronosporomycetes_XXX;Peronosporomycetes_XXX_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Peronosporomycetes_XX;Peronosporomycetes_XXX;Peronosporomycetes_XXX_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Stramenopiles_X;Stramenopiles_X-Group-7;Stramenopiles_X-Group-7_X;Stramenopiles_X-Group-7_XX;Stramenopiles_X-Group-7_...</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Stramenopiles_X;Stramenopiles_X-Group-7;Stramenopiles_X-Group-7_X;Stramenopiles_X-Group-7_XX;Stramenopiles_X-Group-7_...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Mediophyceae_X;Mediophyceae_XX;Mediophyceae_XXX;Mediophyceae_XXX_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Mediophyceae_X;Mediophyceae_XX;Mediophyceae_XXX;Mediophyceae_XXX_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2C;MAST-2C_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2C;MAST-2C_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X</td>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2_X;MAST-2_X_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2_X;MAST-2_X_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2B;MAST-2B_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2B;MAST-2B_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl</td>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Cryptista:nucl</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl;Cryptomonas:nucl</td>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl;Cryptomonas:nucl</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Cryptista:nucl</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Pyrenomonadales:nucl;Pyrenomonadaceae:nucl;Rhodomonas:nucl;Rh...</td>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Pyrenomonadales:nucl;Pyrenomonadaceae:nucl;Rhodomonas:nucl;Rh...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Cryptista:nucl</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Developea;Developea_X;Developea_XX;Mediocremonas;Mediocremonas_mediterraneus;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Developea;Developea_X;Developea_XX;Mediocremonas;Mediocremonas_mediterraneus;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Chaetocerotales</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Chaetocerotales</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Bacillariophyta</td>\n",
       "      <td>Bacillariophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Superorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2D;MAST-2D_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2D;MAST-2D_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Tunicaraptor;Tunicaraptor_unikontum;</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Tunicaraptor;Tunicaraptor_unikontum;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2;Telonemia-Group-2_X;Telonemia-Group-2_X_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2;Telonemia-Group-2_X;Telonemia-Group-2_X_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                  full_tax  \\\n",
       "18                      Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Haliphthorales_X;Haliphthorales_X_sp.;   \n",
       "19   Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Peronosporomycetes_XX;Peronosporomycetes_XXX;Peronosporomycetes_XXX_sp.;   \n",
       "28                                                                            Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X   \n",
       "91   Eukaryota;TSAR;Stramenopiles;Stramenopiles_X;Stramenopiles_X-Group-7;Stramenopiles_X-Group-7_X;Stramenopiles_X-Group-7_XX;Stramenopiles_X-Group-7_...   \n",
       "116                                Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Mediophyceae_X;Mediophyceae_XX;Mediophyceae_XXX;Mediophyceae_XXX_sp.;   \n",
       "125                                                                                       Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2   \n",
       "131                                                                  Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2C;MAST-2C_sp.;   \n",
       "185                                                                                               Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X   \n",
       "187                                                                Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2_X;MAST-2_X_sp.;   \n",
       "189                                                                  Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2B;MAST-2B_sp.;   \n",
       "214                       Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl   \n",
       "90       Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl;Cryptomonas:nucl   \n",
       "101  Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Pyrenomonadales:nucl;Pyrenomonadaceae:nucl;Rhodomonas:nucl;Rh...   \n",
       "114                                     Eukaryota;TSAR;Stramenopiles;Gyrista;Developea;Developea_X;Developea_XX;Mediocremonas;Mediocremonas_mediterraneus;   \n",
       "144                                                                                      Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Chaetocerotales   \n",
       "150                                                                  Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2D;MAST-2D_sp.;   \n",
       "170                   Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Tunicaraptor;Tunicaraptor_unikontum;   \n",
       "6                                                                                                                            Eukaryota;Obazoa;Opisthokonta   \n",
       "34                                                                                                                   Eukaryota;Obazoa;Opisthokonta;Metazoa   \n",
       "46                          Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2;Telonemia-Group-2_X;Telonemia-Group-2_X_sp.;   \n",
       "\n",
       "                                                                                                                                    verbatimIdentification  \\\n",
       "18                      Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Haliphthorales_X;Haliphthorales_X_sp.;   \n",
       "19   Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Peronosporomycetes_XX;Peronosporomycetes_XXX;Peronosporomycetes_XXX_sp.;   \n",
       "28                                                                            Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X   \n",
       "91   Eukaryota;TSAR;Stramenopiles;Stramenopiles_X;Stramenopiles_X-Group-7;Stramenopiles_X-Group-7_X;Stramenopiles_X-Group-7_XX;Stramenopiles_X-Group-7_...   \n",
       "116                                Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Mediophyceae_X;Mediophyceae_XX;Mediophyceae_XXX;Mediophyceae_XXX_sp.;   \n",
       "125                                                                                       Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2   \n",
       "131                                                                  Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2C;MAST-2C_sp.;   \n",
       "185                                                                                               Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X   \n",
       "187                                                                Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2_X;MAST-2_X_sp.;   \n",
       "189                                                                  Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2B;MAST-2B_sp.;   \n",
       "214                       Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl   \n",
       "90       Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Cryptomonadales_X:nucl;Cryptomonas:nucl   \n",
       "101  Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Pyrenomonadales:nucl;Pyrenomonadaceae:nucl;Rhodomonas:nucl;Rh...   \n",
       "114                                     Eukaryota;TSAR;Stramenopiles;Gyrista;Developea;Developea_X;Developea_XX;Mediocremonas;Mediocremonas_mediterraneus;   \n",
       "144                                                                                      Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae;Chaetocerotales   \n",
       "150                                                                  Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-2;MAST-2D;MAST-2D_sp.;   \n",
       "170                   Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Tunicaraptor;Tunicaraptor_unikontum;   \n",
       "6                                                                                                                            Eukaryota;Obazoa;Opisthokonta   \n",
       "34                                                                                                                   Eukaryota;Obazoa;Opisthokonta;Metazoa   \n",
       "46                          Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2;Telonemia-Group-2_X;Telonemia-Group-2_X_sp.;   \n",
       "\n",
       "    old_taxonRank        old name scientificName scientificNameID    kingdom  \\\n",
       "18     supergroup            TSAR       No match             None        NaN   \n",
       "19     supergroup            TSAR       No match             None        NaN   \n",
       "28     supergroup            TSAR       No match             None        NaN   \n",
       "91     supergroup            TSAR       No match             None        NaN   \n",
       "116    supergroup            TSAR       No match             None        NaN   \n",
       "125    supergroup            TSAR       No match             None        NaN   \n",
       "131    supergroup            TSAR       No match             None        NaN   \n",
       "185    supergroup        Haptista       No match             None        NaN   \n",
       "187    supergroup            TSAR       No match             None        NaN   \n",
       "189    supergroup            TSAR       No match             None        NaN   \n",
       "214    supergroup  Cryptista:nucl       No match             None        NaN   \n",
       "90     supergroup  Cryptista:nucl       No match             None        NaN   \n",
       "101    supergroup  Cryptista:nucl       No match             None        NaN   \n",
       "114    supergroup            TSAR       No match             None        NaN   \n",
       "144    supergroup            TSAR       No match             None  Chromista   \n",
       "150    supergroup            TSAR       No match             None        NaN   \n",
       "170    supergroup          Obazoa       No match             None        NaN   \n",
       "6      supergroup          Obazoa       No match             None        NaN   \n",
       "34     supergroup          Obazoa       No match             None   Animalia   \n",
       "46     supergroup            TSAR       No match             None        NaN   \n",
       "\n",
       "              phylum              class order family genus   taxonRank  \n",
       "18               NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "19               NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "28               NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "91               NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "116              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "125              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "131              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "185              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "187              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "189              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "214              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "90               NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "101              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "114              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "144  Bacillariophyta  Bacillariophyceae  None   None  None  Superorder  \n",
       "150              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "170              NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "6                NaN                NaN   NaN    NaN   NaN         NaN  \n",
       "34              None               None  None   None  None     Kingdom  \n",
       "46               NaN                NaN   NaN    NaN   NaN         NaN  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s[\"scientificName\"]==\"No match\"].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.loc[worms_18s[\"scientificName\"]==\"No match\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"scientificName\"]==\"Biota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Eukaryota;Haptista</td>\n",
       "      <td>Eukaryota;Haptista</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aphiaID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                  full_tax  \\\n",
       "120                                                                                                     Eukaryota;Haptista   \n",
       "77                                                                                                               Eukaryota   \n",
       "109                                                                                                             Unassigned   \n",
       "189  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;   \n",
       "\n",
       "                                                                                                    verbatimIdentification  \\\n",
       "120                                                                                                     Eukaryota;Haptista   \n",
       "77                                                                                                               Eukaryota   \n",
       "109                                                                                                             Unassigned   \n",
       "189  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;   \n",
       "\n",
       "    old_taxonRank  old name scientificName scientificNameID kingdom phylum  \\\n",
       "120    supergroup  Haptista            NaN              NaN     NaN    NaN   \n",
       "77            NaN       NaN            NaN              NaN     NaN    NaN   \n",
       "109           NaN       NaN            NaN              NaN     NaN    NaN   \n",
       "189           NaN   aphiaID            NaN              NaN     NaN    NaN   \n",
       "\n",
       "    class order family genus taxonRank  \n",
       "120   NaN   NaN    NaN   NaN       NaN  \n",
       "77    NaN   NaN    NaN   NaN       NaN  \n",
       "109   NaN   NaN    NaN   NaN       NaN  \n",
       "189   NaN   NaN    NaN   NaN   species  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota;Haptista\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota;Haptista\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aphiaID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                  full_tax  \\\n",
       "109                                                                                                             Unassigned   \n",
       "189  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;   \n",
       "\n",
       "                                                                                                    verbatimIdentification  \\\n",
       "109                                                                                                             Unassigned   \n",
       "189  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Spiniferites;Spiniferites_mirabilis;   \n",
       "\n",
       "    old_taxonRank old name scientificName scientificNameID kingdom phylum  \\\n",
       "109           NaN      NaN            NaN              NaN     NaN    NaN   \n",
       "189           NaN  aphiaID            NaN              NaN     NaN    NaN   \n",
       "\n",
       "    class order family genus taxonRank  \n",
       "109   NaN   NaN    NaN   NaN       NaN  \n",
       "189   NaN   NaN    NaN   NaN   species  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2, 13)\n",
      "(0, 13)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(worms_18s[worms_18s['scientificName'].isna() == True].shape)\n",
    "worms_18s.loc[worms_18s['scientificName'].isna() == True,'scientificName'] = 'incertae sedis'\n",
    "worms_18s.loc[worms_18s['scientificName'] == 'incertae sedis','scientificNameID'] =  'urn:lsid:marinespecies.org:taxname:12'\n",
    "print(worms_18s[worms_18s['scientificName'].isna() == True].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(332, 13)"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s[\"old name\"]==\"aphiaID\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.to_csv(\"../processed/worms_18S_matching.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;</td>\n",
       "      <td>Paraphysomonas</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:291417</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Ochrophyta</td>\n",
       "      <td>Chrysophyceae</td>\n",
       "      <td>Chromulinales</td>\n",
       "      <td>Paraphysomonadaceae</td>\n",
       "      <td>Paraphysomonas</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;</td>\n",
       "      <td>Gonyaulax polygramma</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:110035</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gonyaulacales</td>\n",
       "      <td>Gonyaulacaceae</td>\n",
       "      <td>Gonyaulax</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae</td>\n",
       "      <td>Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae</td>\n",
       "      <td>Mamiellaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:17663</td>\n",
       "      <td>Plantae</td>\n",
       "      <td>Chlorophyta</td>\n",
       "      <td>Mamiellophyceae</td>\n",
       "      <td>Mamiellales</td>\n",
       "      <td>Mamiellaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;</td>\n",
       "      <td>Blastodinium galatheanum</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:623673</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Blastodiniales</td>\n",
       "      <td>Blastodinidae</td>\n",
       "      <td>Blastodinium</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:22565</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Subphylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                       full_tax  \\\n",
       "0  Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;   \n",
       "1              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;   \n",
       "2                                   Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae   \n",
       "3       Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;   \n",
       "4                                                                Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida   \n",
       "\n",
       "                                                                                                         verbatimIdentification  \\\n",
       "0  Eukaryota;TSAR;Stramenopiles;Gyrista;Chrysophyceae;Paraphysomonadales;Paraphysomonadaceae;Paraphysomonas;Paraphysomonas_sp.;   \n",
       "1              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gonyaulacales;Gonyaulacaceae;Gonyaulax;Gonyaulax_polygramma;   \n",
       "2                                   Eukaryota;Archaeplastida;Chlorophyta;Chlorophyta_X;Mamiellophyceae;Mamiellales;Mamiellaceae   \n",
       "3       Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Peridiniales;Blastodiniaceae;Blastodinium;Blastodinium_galatheanum;   \n",
       "4                                                                Eukaryota;TSAR;Alveolata;Apicomplexa;Coccidiomorphea;Eimeriida   \n",
       "\n",
       "             scientificName                           scientificNameID  \\\n",
       "0            Paraphysomonas  urn:lsid:marinespecies.org:taxname:291417   \n",
       "1      Gonyaulax polygramma  urn:lsid:marinespecies.org:taxname:110035   \n",
       "2              Mamiellaceae   urn:lsid:marinespecies.org:taxname:17663   \n",
       "3  Blastodinium galatheanum  urn:lsid:marinespecies.org:taxname:623673   \n",
       "4               Apicomplexa   urn:lsid:marinespecies.org:taxname:22565   \n",
       "\n",
       "     kingdom       phylum            class           order  \\\n",
       "0  Chromista   Ochrophyta    Chrysophyceae   Chromulinales   \n",
       "1  Chromista      Myzozoa      Dinophyceae   Gonyaulacales   \n",
       "2    Plantae  Chlorophyta  Mamiellophyceae     Mamiellales   \n",
       "3  Chromista      Myzozoa      Dinophyceae  Blastodiniales   \n",
       "4  Chromista      Myzozoa             None            None   \n",
       "\n",
       "                family           genus  taxonRank  \n",
       "0  Paraphysomonadaceae  Paraphysomonas      Genus  \n",
       "1       Gonyaulacaceae       Gonyaulax    Species  \n",
       "2         Mamiellaceae            None     Family  \n",
       "3        Blastodinidae    Blastodinium    Species  \n",
       "4                 None            None  Subphylum  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s.drop(columns=['old name','old_taxonRank'],inplace=True)\n",
    "worms_18s.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Occurrence and worms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(149182, 16)"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S V9'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "2  2a31e5c01634165da99e7381279baa75   \n",
       "3  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                                                           sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "4    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                                                                                                       taxonomy  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "   Confidence                 eventID  organismQuantity  \\\n",
       "0    0.922099  GOMECC4_27N_Sta1_DCM_A              1516   \n",
       "1    0.999947  GOMECC4_27N_Sta1_DCM_A               962   \n",
       "2    0.779948  GOMECC4_27N_Sta1_DCM_A              1164   \n",
       "3    0.999931  GOMECC4_27N_Sta1_DCM_A               287   \n",
       "4    0.986908  GOMECC4_27N_Sta1_DCM_A               250   \n",
       "\n",
       "                                                 occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                         verbatimIdentification  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "           scientificName                           scientificNameID  \\\n",
       "0    Neocalanus cristatus  urn:lsid:marinespecies.org:taxname:104470   \n",
       "1  Clausocalanus furcatus  urn:lsid:marinespecies.org:taxname:104503   \n",
       "2             Acrocalanus  urn:lsid:marinespecies.org:taxname:104192   \n",
       "3             Dinophyceae   urn:lsid:marinespecies.org:taxname:19542   \n",
       "4          Gymnodiniaceae  urn:lsid:marinespecies.org:taxname:109410   \n",
       "\n",
       "     kingdom      phylum        class          order           family  \\\n",
       "0   Animalia  Arthropoda     Copepoda      Calanoida        Calanidae   \n",
       "1   Animalia  Arthropoda     Copepoda      Calanoida  Clausocalanidae   \n",
       "2   Animalia  Arthropoda     Copepoda      Calanoida    Paracalanidae   \n",
       "3  Chromista     Myzozoa  Dinophyceae           None             None   \n",
       "4  Chromista     Myzozoa  Dinophyceae  Gymnodiniales   Gymnodiniaceae   \n",
       "\n",
       "           genus taxonRank  \n",
       "0     Neocalanus   Species  \n",
       "1  Clausocalanus   Species  \n",
       "2    Acrocalanus     Genus  \n",
       "3           None     Class  \n",
       "4           None    Family  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get identificationRemarks\n",
    "occ18_test = occ['18S V9'].copy()\n",
    "occ18_test.drop(columns=['domain','supergroup','division','subdivision','class','order','family','genus','species'],inplace=True)\n",
    "#occ18_test.drop(columns=['old name'],inplace=True)\n",
    "\n",
    "occ18_test = occ18_test.merge(worms_18s, how='left', left_on ='taxonomy', right_on='full_tax')\n",
    "occ18_test.drop(columns='full_tax', inplace=True)\n",
    "occ18_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>controls_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>411</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S V9</td>\n",
       "      <td>260</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  amplicon_sequenced  ampliconSize trim_method  \\\n",
       "0          16S V4-V5           411    cutadapt   \n",
       "1             18S V9           260    cutadapt   \n",
       "\n",
       "                     cluster_method pid_clustering  \\\n",
       "0  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "1  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1             PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706   \n",
       "\n",
       "                             code_repo  \\\n",
       "0  https://github.com/aomlomics/gomecc   \n",
       "1  https://github.com/aomlomics/gomecc   \n",
       "\n",
       "                                                                       identificationReferences  \\\n",
       "0                              10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline   \n",
       "1  10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline   \n",
       "\n",
       "                                                                                                                            controls_used  \n",
       "0  12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community  \n",
       "1                          12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls   "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '18S V9','taxa_class_method'].item()\n",
    "occ18_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '18S V9','taxa_ref_db'].item()\n",
    "\n",
    "occ18_test['identificationRemarks'] = occ18_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+occ18_test['Confidence'].astype(str) +\", against reference database: \"+occ18_test['taxa_ref_db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706'"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test['identificationRemarks'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### taxonID, basisOfRecord, eventID, nameAccordingTo, organismQuantityType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['taxonID'] = 'ASV:'+occ18_test['featureid']\n",
    "occ18_test['basisOfRecord'] = 'MaterialSample'\n",
    "occ18_test['nameAccordingTo'] = \"WoRMS\"\n",
    "occ18_test['organismQuantityType'] = \"DNA sequence reads\"\n",
    "occ18_test['recordedBy'] = data['study_data']['recordedBy'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### associatedSequences, materialSampleID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_name', 'library_id', 'title', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'lib_layout', 'platform',\n",
       "       'instrument_model', 'design_description', 'filetype', 'filename',\n",
       "       'filename2', 'biosample_accession', 'sra_accession', 'seq_meth',\n",
       "       'nucl_acid_ext', 'amplicon_sequenced', 'target_gene',\n",
       "       'target_subfragment', 'pcr_primer_forward', 'pcr_primer_reverse',\n",
       "       'pcr_primer_name_forward', 'pcr_primer_name_reverse',\n",
       "       'pcr_primer_reference', 'pcr_cond', 'nucl_acid_amp', 'adapters',\n",
       "       'mid_barcode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test = occ18_test.merge(data['prep_data'].loc[data['prep_data']['amplicon_sequenced'] == '18S V9',['sample_name','sra_accession','biosample_accession']], how='left', left_on ='eventID', right_on='sample_name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### eventID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['eventID'] = occ18_test['eventID']+\"_18S\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sampleSize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sampleSize by total number of reads per sample\n",
    "x = asv_tables['18S V9'].sum(numeric_only=True).astype('int')\n",
    "x.index = x.index+\"_18S\"\n",
    "occ18_test['sampleSizeValue'] = occ18_test['eventID'].map(x).astype('str')\n",
    "occ18_test['sampleSizeUnit'] = 'DNA sequence reads'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ18_test.drop(columns=['sample_name','featureid','taxonomy','Confidence','taxa_class_method','taxa_ref_db'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['associatedSequences'] = \"https://www.ncbi.nlm.nih.gov/sra/\"+occ18_test['sra_accession']+' | '+ \"https://www.ncbi.nlm.nih.gov/biosample/\"+occ18_test['biosample_accession']+' | '+\"https://www.ncbi.nlm.nih.gov/bioproject/\"+data['study_data']['bioproject_accession'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test.rename(columns={'biosample_accession': 'materialSampleID',\n",
    "                  'sequence': 'DNA_sequence'},inplace=True)\n",
    "                   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ18_test.drop(columns=['sra_accession'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DNA_sequence', 'eventID', 'organismQuantity', 'occurrenceID',\n",
       "       'verbatimIdentification', 'scientificName', 'scientificNameID',\n",
       "       'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'taxonRank',\n",
       "       'identificationRemarks', 'taxonID', 'basisOfRecord', 'nameAccordingTo',\n",
       "       'organismQuantityType', 'recordedBy', 'materialSampleID',\n",
       "       'sampleSizeValue', 'sampleSizeUnit', 'associatedSequences'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Genus</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                       DNA_sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "4    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_18S              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_18S               962   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_18S              1164   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_18S               287   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_18S               250   \n",
       "\n",
       "                                                 occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                         verbatimIdentification  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "           scientificName                           scientificNameID  \\\n",
       "0    Neocalanus cristatus  urn:lsid:marinespecies.org:taxname:104470   \n",
       "1  Clausocalanus furcatus  urn:lsid:marinespecies.org:taxname:104503   \n",
       "2             Acrocalanus  urn:lsid:marinespecies.org:taxname:104192   \n",
       "3             Dinophyceae   urn:lsid:marinespecies.org:taxname:19542   \n",
       "4          Gymnodiniaceae  urn:lsid:marinespecies.org:taxname:109410   \n",
       "\n",
       "     kingdom      phylum        class          order           family  \\\n",
       "0   Animalia  Arthropoda     Copepoda      Calanoida        Calanidae   \n",
       "1   Animalia  Arthropoda     Copepoda      Calanoida  Clausocalanidae   \n",
       "2   Animalia  Arthropoda     Copepoda      Calanoida    Paracalanidae   \n",
       "3  Chromista     Myzozoa  Dinophyceae           None             None   \n",
       "4  Chromista     Myzozoa  Dinophyceae  Gymnodiniales   Gymnodiniaceae   \n",
       "\n",
       "           genus taxonRank  \\\n",
       "0     Neocalanus   Species   \n",
       "1  Clausocalanus   Species   \n",
       "2    Acrocalanus     Genus   \n",
       "3           None     Class   \n",
       "4           None    Family   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:36aa75f9b28f5f831c2d631ba65c2bcb  MaterialSample           WoRMS   \n",
       "1  ASV:4e38e8ced9070952b314e1880bede1ca  MaterialSample           WoRMS   \n",
       "2  ASV:2a31e5c01634165da99e7381279baa75  MaterialSample           WoRMS   \n",
       "3  ASV:ecee60339b2fb88ea6d1c8d18054bed4  MaterialSample           WoRMS   \n",
       "4  ASV:fa1f1a97dd4ae7c826009186bad26384  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "  sampleSizeValue      sampleSizeUnit  \\\n",
       "0            9838  DNA sequence reads   \n",
       "1            9838  DNA sequence reads   \n",
       "2            9838  DNA sequence reads   \n",
       "3            9838  DNA sequence reads   \n",
       "4            9838  DNA sequence reads   \n",
       "\n",
       "                                                                                                                                     associatedSequences  \n",
       "0  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...  \n",
       "1  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...  \n",
       "2  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...  \n",
       "3  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...  \n",
       "4  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge event and occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_merged = occ18_test.merge(all_event_data,how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>countryCode</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>geodeticDatum</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Genus</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>US</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>WGS84</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                       DNA_sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "4    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_18S              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_18S               962   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_18S              1164   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_18S               287   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_18S               250   \n",
       "\n",
       "                                                 occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                         verbatimIdentification  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "           scientificName                           scientificNameID  \\\n",
       "0    Neocalanus cristatus  urn:lsid:marinespecies.org:taxname:104470   \n",
       "1  Clausocalanus furcatus  urn:lsid:marinespecies.org:taxname:104503   \n",
       "2             Acrocalanus  urn:lsid:marinespecies.org:taxname:104192   \n",
       "3             Dinophyceae   urn:lsid:marinespecies.org:taxname:19542   \n",
       "4          Gymnodiniaceae  urn:lsid:marinespecies.org:taxname:109410   \n",
       "\n",
       "     kingdom      phylum        class          order           family  \\\n",
       "0   Animalia  Arthropoda     Copepoda      Calanoida        Calanidae   \n",
       "1   Animalia  Arthropoda     Copepoda      Calanoida  Clausocalanidae   \n",
       "2   Animalia  Arthropoda     Copepoda      Calanoida    Paracalanidae   \n",
       "3  Chromista     Myzozoa  Dinophyceae           None             None   \n",
       "4  Chromista     Myzozoa  Dinophyceae  Gymnodiniales   Gymnodiniaceae   \n",
       "\n",
       "           genus taxonRank  \\\n",
       "0     Neocalanus   Species   \n",
       "1  Clausocalanus   Species   \n",
       "2    Acrocalanus     Genus   \n",
       "3           None     Class   \n",
       "4           None    Family   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:36aa75f9b28f5f831c2d631ba65c2bcb  MaterialSample           WoRMS   \n",
       "1  ASV:4e38e8ced9070952b314e1880bede1ca  MaterialSample           WoRMS   \n",
       "2  ASV:2a31e5c01634165da99e7381279baa75  MaterialSample           WoRMS   \n",
       "3  ASV:ecee60339b2fb88ea6d1c8d18054bed4  MaterialSample           WoRMS   \n",
       "4  ASV:fa1f1a97dd4ae7c826009186bad26384  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "  sampleSizeValue      sampleSizeUnit  \\\n",
       "0            9838  DNA sequence reads   \n",
       "1            9838  DNA sequence reads   \n",
       "2            9838  DNA sequence reads   \n",
       "3            9838  DNA sequence reads   \n",
       "4            9838  DNA sequence reads   \n",
       "\n",
       "                                                                                                                                     associatedSequences  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "1  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "2  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "3  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "4  https://www.ncbi.nlm.nih.gov/sra/SRR26161153 | https://www.ncbi.nlm.nih.gov/biosample/SAMN37516094 | https://www.ncbi.nlm.nih.gov/bioproject/PRJNA...   \n",
       "\n",
       "  locationID               eventDate minimumDepthInMeters  \\\n",
       "0   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "1   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "2   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "3   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "4   27N_Sta1  2021-09-14T11:00-04:00                   49   \n",
       "\n",
       "  maximumDepthInMeters                                     locality  \\\n",
       "0                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "         waterBody countryCode decimalLatitude decimalLongitude geodeticDatum  \\\n",
       "0  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "1  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "2  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "3  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "4  Mexico, Gulf of          US          26.997          -79.618         WGS84   \n",
       "\n",
       "  samplingProtocol           parentEventID          datasetID  \n",
       "0      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "1      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "2      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "3      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  \n",
       "4      CTD rosette  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_merged.drop(columns=['DNA_sequence']).to_csv(\"../processed/occurrence_18S.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### combine 16s and 18s occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(149182, 37)"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all = pd.concat([occ16_merged,occ18_merged],axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all['occurrenceStatus'] = 'present' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(318652, 38)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all.drop(columns=['DNA_sequence']).to_csv(\"../processed/occurrence.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA DERIVED AFTER THIS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create DNA Derived Extension\n",
    "\n",
    "First we clearly define the desired columns for the DNA Derived Extension output file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define desired columns for DNA derived extension in output order\n",
    "DESIRED_DNA_DERIVED_COLUMNS = [\n",
    "    'eventID', 'source_mat_id', 'env_broad_scale', 'env_local_scale', 'env_medium', \n",
    "    'samp_vol_we_dna_ext', 'samp_collect_device', 'samp_mat_process', 'size_frac', \n",
    "    'concentration', 'lib_layout', 'seq_meth', 'nucl_acid_ext', 'target_gene', \n",
    "    'target_subfragment', 'pcr_primer_forward', 'pcr_primer_reverse', \n",
    "    'pcr_primer_name_forward', 'pcr_primer_name_reverse', 'pcr_primer_reference', \n",
    "    'pcr_cond', 'nucl_acid_amp', 'ampliconSize', 'otu_seq_comp_appr', 'otu_db', \n",
    "    'occurrenceID', 'DNA_sequence', 'concentrationUnit', 'otu_class_appr'\n",
    "]\n",
    "\n",
    "print(f\"DNA derived extension will have {len(DESIRED_DNA_DERIVED_COLUMNS)} columns\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Start building DNA derived extension by merging occurrence data with experiment run metadata to get library and sample information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['eventID', 'samp_name', 'occurrenceID', 'DNA_sequence', 'sop', 'nucl_acid_ext', 'samp_vol_we_dna_ext', 'samp_mat_process', 'nucl_acid_amp', 'target_gene', 'target_subfragment', 'ampliconSize', 'lib_layout', 'pcr_primer_forward', 'pcr_primer_reverse', 'pcr_primer_name_forward', 'pcr_primer_name_reverse', 'pcr_primer_reference', 'pcr_cond', 'seq_meth', 'otu_class_appr', 'otu_seq_comp_appr', 'otu_db', 'env_broad_scale', 'env_local_scale', 'env_medium', 'size_frac', 'concentration', 'concentrationUnit', 'samp_collec_device', 'source_mat_id'])"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Start with occurrence data and merge with experimentRunMetadata on eventID (lib_id)\n",
    "dna_base = occ_all_final_output[['eventID', 'occurrenceID', 'DNA_sequence']].copy()\n",
    "\n",
    "# Merge with experimentRunMetadata to get samp_name and library metadata\n",
    "dna_with_exp = dna_base.merge(data['experimentRunMetadata'], left_on='eventID', right_on='lib_id', how='left')\n",
    "\n",
    "print(f\"DNA base rows: {len(dna_base)}\")\n",
    "print(f\"After exp merge: {len(dna_with_exp)}\")\n",
    "print(f\"Columns added: {set(dna_with_exp.columns) - set(dna_base.columns)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge with sample metadata to get sample collection information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge with sampleMetadata to get sample collection metadata\n",
    "dna_with_sample = dna_with_exp.merge(data['sampleMetadata'], on='samp_name', how='left')\n",
    "\n",
    "print(f\"After sample merge: {len(dna_with_sample)}\")\n",
    "print(f\"Total columns now: {len(dna_with_sample.columns)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Apply Darwin Core mapping to transform raw field names to standardized terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply Darwin Core mapping using the dwc_data['dnaDerived'] dictionary\n",
    "dna_mapped = dna_with_sample.copy()\n",
    "\n",
    "# Apply mappings for each column that has a Darwin Core equivalent\n",
    "for col in dna_mapped.columns:\n",
    "    if col in dwc_data['dnaDerived']:\n",
    "        dwc_term = dwc_data['dnaDerived'][col]\n",
    "        if dwc_term != col:  # Only rename if different\n",
    "            dna_mapped = dna_mapped.rename(columns={col: dwc_term})\n",
    "            print(f\"Mapped: {col} -> {dwc_term}\")\n",
    "\n",
    "print(f\"Mapping complete. Final columns: {len(dna_mapped.columns)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add hardcoded fields and create final DNA derived extension output."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CHECK THIS TOMORROW. DO WE WANT TO INCLUDE THIS MAPPING IN THE CHECKLIST? WHY OR WHY NOT?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add source_mat_id using materialSampleID \n",
    "if 'materialSampleID' in dna_mapped.columns:\n",
    "    dna_mapped['source_mat_id'] = dna_mapped['materialSampleID']\n",
    "    print(\"Added source_mat_id from materialSampleID\")\n",
    "\n",
    "# Select only the desired columns that exist in our data\n",
    "available_columns = [col for col in DESIRED_DNA_DERIVED_COLUMNS if col in dna_mapped.columns]\n",
    "missing_columns = [col for col in DESIRED_DNA_DERIVED_COLUMNS if col not in dna_mapped.columns]\n",
    "\n",
    "print(f\"Available columns: {len(available_columns)}\")\n",
    "print(f\"Missing columns: {missing_columns}\")\n",
    "\n",
    "# Create final DNA derived extension with only available columns\n",
    "dna_derived_final = dna_mapped[available_columns].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Final DNA derived extension output and save to file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Final DNA derived extension shape: {dna_derived_final.shape}\")\n",
    "print(f\"Columns: {list(dna_derived_final.columns)}\")\n",
    "\n",
    "# Display first few rows\n",
    "print(\"\\nFirst 3 rows:\")\n",
    "display(dna_derived_final.head(3))\n",
    "\n",
    "# Save to CSV\n",
    "output_filename = 'dna_derived_extension.csv'\n",
    "dna_derived_final.to_csv(output_filename, index=False)\n",
    "print(f\"\\nDNA derived extension saved as: {output_filename}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BELOW IS FROM THE OLD VERSION OF EDNA2OBIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collec_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>concentration</th>\n",
       "      <th>concentrationUnit</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>eventID</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>otu_class_appr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9dd2f0b0dd33100b7e</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTTCGTAGGCGTCTTGCCAAGTTGATCGTTAAAGCCACCGGCTTAACCGGTGATCTGCGATCAAAACTGGCGAGATAGAATATGTGAGGGGAATGTGGAATTC...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b7f992a6764fa69e97</td>\n",
       "      <td>TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGATGTAGGCGGTCTGTCAAGTCGGATGTGAAATCCCTGGGCTCAACCCAGGAACTGCATTCGAAACTGTCAGACTAGAGTCTCGGAGGGGGTGGCGGAATTC...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041c787e5a481888c27e</td>\n",
       "      <td>TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTCCGCAGGCGGACTATTAAGTCAGTGGTGAAAGTCTGCAGCTTAACTGTAGAATTGCCATTGAAACTGATAGTCTTGAGTGTGGTTGAAGTGGGCGGAATAT...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705bdf3f93218ac93a030</td>\n",
       "      <td>TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTGCGTAGGCGTCTACGTAAGTTGTTTGTTAAATCCATCGGCTTAACCGATGATCTGCAAACAAAACTGCATAGATAGAGTTTGGAAGAGGAAAGTGGAATTC...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db7812103fe73fdefb7d2b</td>\n",
       "      <td>TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAGCGCAGGTGGCTGAGTTCGTCCGTGGTGCAAGCTCCAGGCCTAACCTGGAGAGGTCTACGGATACTGCTCGGCTTGAGGGCGGTAGAGGAGCACGGAATTC...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source_mat_id               env_broad_scale  \\\n",
       "0  GOMECC4_27N_Sta1_Deep  marine biome [ENVO:00000447]   \n",
       "1  GOMECC4_27N_Sta1_Deep  marine biome [ENVO:00000447]   \n",
       "2  GOMECC4_27N_Sta1_Deep  marine biome [ENVO:00000447]   \n",
       "3  GOMECC4_27N_Sta1_Deep  marine biome [ENVO:00000447]   \n",
       "4  GOMECC4_27N_Sta1_Deep  marine biome [ENVO:00000447]   \n",
       "\n",
       "                           env_local_scale                 env_medium  \\\n",
       "0  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "1  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "2  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "3  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "4  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "\n",
       "  samp_vol_we_dna_ext samp_collec_device  \\\n",
       "0             1920 ml      Niskin bottle   \n",
       "1             1920 ml      Niskin bottle   \n",
       "2             1920 ml      Niskin bottle   \n",
       "3             1920 ml      Niskin bottle   \n",
       "4             1920 ml      Niskin bottle   \n",
       "\n",
       "                                                  samp_mat_process size_frac  \\\n",
       "0  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "1  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "2  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "3  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "4  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "\n",
       "           concentration      concentrationUnit lib_layout  \\\n",
       "0  missing: not provided  missing: not provided     paired   \n",
       "1  missing: not provided  missing: not provided     paired   \n",
       "2  missing: not provided  missing: not provided     paired   \n",
       "3  missing: not provided  missing: not provided     paired   \n",
       "4  missing: not provided  missing: not provided     paired   \n",
       "\n",
       "               seq_meth  \\\n",
       "0  Illumina MiSeq 2x250   \n",
       "1  Illumina MiSeq 2x250   \n",
       "2  Illumina MiSeq 2x250   \n",
       "3  Illumina MiSeq 2x250   \n",
       "4  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                          nucl_acid_ext  \\\n",
       "0  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "1  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "2  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "3  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "4  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "  target_gene target_subfragment   pcr_primer_forward    pcr_primer_reverse  \\\n",
       "0    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "1    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "2    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "3    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "4    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_name_reverse     pcr_primer_reference  \\\n",
       "0                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "1                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "2                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "3                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "4                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "\n",
       "                                                                                                      pcr_cond  \\\n",
       "0  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "1  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "2  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "3  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "4  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "\n",
       "             nucl_acid_amp  ampliconSize  \\\n",
       "0  10.1111/1462-2920.13023           411   \n",
       "1  10.1111/1462-2920.13023           411   \n",
       "2  10.1111/1462-2920.13023           411   \n",
       "3  10.1111/1462-2920.13023           411   \n",
       "4  10.1111/1462-2920.13023           411   \n",
       "\n",
       "                                   otu_seq_comp_appr  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                                 otu_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "\n",
       "                       eventID  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S   \n",
       "\n",
       "                                                      occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9dd2f0b0dd33100b7e   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b7f992a6764fa69e97   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041c787e5a481888c27e   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705bdf3f93218ac93a030   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db7812103fe73fdefb7d2b   \n",
       "\n",
       "                                                                                                                                            DNA_sequence  \\\n",
       "0  TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTTCGTAGGCGTCTTGCCAAGTTGATCGTTAAAGCCACCGGCTTAACCGGTGATCTGCGATCAAAACTGGCGAGATAGAATATGTGAGGGGAATGTGGAATTC...   \n",
       "1  TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGATGTAGGCGGTCTGTCAAGTCGGATGTGAAATCCCTGGGCTCAACCCAGGAACTGCATTCGAAACTGTCAGACTAGAGTCTCGGAGGGGGTGGCGGAATTC...   \n",
       "2  TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTCCGCAGGCGGACTATTAAGTCAGTGGTGAAAGTCTGCAGCTTAACTGTAGAATTGCCATTGAAACTGATAGTCTTGAGTGTGGTTGAAGTGGGCGGAATAT...   \n",
       "3  TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTGCGTAGGCGTCTACGTAAGTTGTTTGTTAAATCCATCGGCTTAACCGATGATCTGCAAACAAAACTGCATAGATAGAGTTTGGAAGAGGAAAGTGGAATTC...   \n",
       "4  TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAGCGCAGGTGGCTGAGTTCGTCCGTGGTGCAAGCTCCAGGCCTAACCTGGAGAGGTCTACGGATACTGCTCGGCTTGAGGGCGGTAGAGGAGCACGGAATTC...   \n",
       "\n",
       "                          otu_class_appr  \n",
       "0  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "1  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "2  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "3  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "4  Tourmaline; qiime2-2021.2; dada2; ASV  "
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_occ.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "edna2obis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
