{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Darwin Core Conversion of eDNA Sequence Data From the AOML_MIMARKS metadata template \n",
    "\n",
    "**Version:** 1.0.0\n",
    "\n",
    "**Author:** Katherine Silliman\n",
    "\n",
    "**Last Updated:** 2-Oct-2023\n",
    "\n",
    "This notebook is for converting a [MIMARKS](https://fairsharing.org/FAIRsharing.zvrep1)-based data sheet to DarwinCore for submission to OBIS. It has been testing on a Mac M1 laptop running in Rosetta mode, with Python 3.11. \n",
    "\n",
    "[Metadata template Google Sheet](https://docs.google.com/spreadsheets/d/1jof9MBEll7Xluu8-_znLRBIP9JpyAd_5YvdioZ-REoY/edit?usp=sharing)\n",
    "\n",
    "**Requirements:**\n",
    "- Python 3\n",
    "- Python 3 packages:\n",
    "    - os\n",
    "- External packages:\n",
    "    - Bio.Entrez from biopython\n",
    "    - numpy\n",
    "    - pandas\n",
    "    - openpyxl\n",
    "    - pyworms\n",
    "    - multiprocess\n",
    "- Custom modules:\n",
    "    - WoRMS_matching\n",
    "\n",
    "**Resources:**\n",
    "- Abarenkov K, Andersson AF, Bissett A, Finstad AG, Fossøy F, Grosjean M, Hope M, Jeppesen TS, Kõljalg U, Lundin D, Nilsson RN, Prager M, Provoost P, Schigel D, Suominen S, Svenningsen C & Frøslev TG (2023) Publishing DNA-derived data through biodiversity data platforms, v1.3. Copenhagen: GBIF Secretariat. https://doi.org/10.35035/doc-vf1a-nr22.https://doi.org/10.35035/doc-vf1a-nr22.\n",
    "- [OBIS manual](https://manual.obis.org/dna_data.html)\n",
    "- [TDWG Darwin Core Occurrence Core](https://dwc.tdwg.org/terms/#occurrence)\n",
    "- [GBIF DNA Derived Data Extension](https://tools.gbif.org/dwca-validator/extension.do?id=http://rs.gbif.org/terms/1.0/DNADerivedData)\n",
    "- https://github.com/iobis/dataset-edna\n",
    "\n",
    "**Citation**  \n",
    "Silliman K, Anderson S, Storo R, Thompson L (2023) A Case Study in Sharing Marine eDNA Metabarcoding Data to OBIS. Biodiversity Information Science and Standards 7: e111048. https://doi.org/10.3897/biss.7.111048\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation  \n",
    "\n",
    "```bash\n",
    "conda create -n edna2obis\n",
    "conda activate edna2obis\n",
    "conda install -c conda-forge notebook\n",
    "conda install -c conda-forge nb_conda_kernels\n",
    "\n",
    "conda install -c conda-forge numpy pandas\n",
    "conda install -c conda-forge openpyxl\n",
    "\n",
    "#worms conversion\n",
    "conda install -c conda-forge pyworms\n",
    "conda install -c conda-forge multiprocess\n",
    "conda install -c conda-forge biopython\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import WoRMS_matching # custom functions for querying WoRMS API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# jupyter notebook parameters\n",
    "pd.set_option('display.max_colwidth', 150)\n",
    "pd.set_option('display.max_columns', 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that in a Jupyter Notebook, the current directory is always where the .ipynb file is being run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Prepare input data \n",
    "\n",
    "**Project data and metadata**  \n",
    "This workflow assumes that you have your project metadata in an Excel sheet formatted like the template located [here](https://docs.google.com/spreadsheets/d/1jof9MBEll7Xluu8-_znLRBIP9JpyAd_5YvdioZ-REoY/edit?usp=sharing). Instructions for filling out the metadata template are located in the 'Readme' sheet.\n",
    "\n",
    "**eDNA and taxonomy data**  \n",
    "The eDNA data and assigned taxonomy should be in a specific tab-delimited format. ![asv_table format](../images/asv_table.png)\n",
    "\n",
    "This file is generated automatically by [Tourmaline v2023.5+](https://github.com/aomlomics/tourmaline), in X location. If your data was generated with Qiime2 or a previous version of Tourmaline, you can convert the `table.qza`, `taxonomy.qza`, and `repseqs.qza` outputs to the correct format using the `create_asv_seq_taxa_obis.sh` shell script.\n",
    "\n",
    "Example:  \n",
    "\n",
    "``` bash\n",
    "#Run this with a qiime2 environment. \n",
    "bash create_asv_seq_taxa_obis.sh -f \\\n",
    "../gomecc_v2_raw/table-16S-merge.qza -t ../gomecc_v2_raw/taxonomy-16S-merge.qza -r ../gomecc_v2_raw/repseqs-16S-merge.qza \\\n",
    "-o ../gomecc_v2_raw/gomecc-16S-asv.tsv\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Set configs  \n",
    "\n",
    "Below you can set definitions for parameters used in the code. \n",
    "\n",
    "| Parameter           | Description                                                                                                       | Example                                                                                              |\n",
    "|---------------------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|\n",
    "| `sample_data`       | Name of sheet in project data Excel file with sample data.                                                        | \"water_sample_data\"                                                                                  |\n",
    "| `prep_data`         | Name of sheet in project data Excel file with data about molecular preparation methods.                           | \"amplicon_prep_data\"                                                                                 |\n",
    "| `analysis_data`     | Name of sheet in project data Excel file with data about analysis methods.                                        | \"analysis_data\"                                                                                      |\n",
    "| `study_data`        | Name of sheet in project data Excel file with metadata about the study.                                           | \"study_data\"                                                                                         |\n",
    "| `msmt_metadata`     | Name of sheet in project data Excel file with metadata about additional measurements. Not used in current code.   | \"measurement_metadata\"                                                                               |\n",
    "| `excel_file`        | Path of project data Excel file.                                                                                  | \"../raw/gomecc4_AOML_MIMARKS.survey.water.6.0.xlsx\"                                                  |\n",
    "| `md_excel`          | Path of data dictionary Excel file.                                                                               | \"../raw/gomecc_AOML2DwC standards.xlsx\"                                                              |\n",
    "| `datafiles`         | Python dictionary, where keys are the amplicon names and the values are the paths to the cooresponding ASV table. | {'16S V4-V5': '../raw/gomecc-16S-asv.tsv', '18S V9': '../raw/gomecc-18S-asv.tsv'}                    |\n",
    "| `skip_sample_types` | Python list of sample_type values to skip from OBIS submission, such as controls or blanks.                       | ['mock community','distilled water blank','extraction blank','PCR no-template control','RTSF blank'] |\n",
    "| `skip_columns`      | Python list of columns to ignore when submitting to OBIS.                                                         | ['notes_sampling']                                                                                   |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['sample_data'] = \"water_sample_data\"\n",
    "params['prep_data']= \"amplicon_prep_data\"\n",
    "params['analysis_data'] = \"analysis_data\"\n",
    "params['study_data'] = \"study_data\"\n",
    "params['msmt_metadata'] = \"measurement_metadata\"\n",
    "params['excel_file'] = \"../raw/gomecc4_AOML_MIMARKS.survey.water.6.0.xlsx\"\n",
    "\n",
    "params['datafiles'] = {'16S V4-V5': '../raw/gomecc-16S-asv.tsv',\n",
    "                       '18S V9': '../raw/gomecc-18S-asv.tsv'}\n",
    "\n",
    "params['skip_sample_types'] = ['mock community','distilled water blank','extraction blank','PCR no-template control','RTSF blank']\n",
    "params['skip_columns']= ['notes_sampling']\n",
    "params['md_excel'] = \"../raw/gomecc_AOML2DwC standards.xlsx\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load data\n",
    "\n",
    "Note that in a Jupyter Notebook, the current directory is always where the .ipynb file is being run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load project data Excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data = pd.read_excel(\n",
    "    params['excel_file'], \n",
    "    [params['study_data'],params['sample_data'],params['prep_data'],params['analysis_data'],params['msmt_metadata']],\n",
    "    index_col=None, na_values=[\"\"], comment=\"#\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename keys in data dictionary to a general term\n",
    "data['sample_data'] = data.pop(params['sample_data'])\n",
    "data['prep_data'] = data.pop(params['prep_data'])\n",
    "data['analysis_data'] = data.pop(params['analysis_data'])\n",
    "data['study_data'] = data.pop(params['study_data'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove * from headers (was required for NCBI submission, but no longer needed)\n",
    "data['sample_data'].columns = data['sample_data'].columns.str.replace(\"*\",\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sample_data  \n",
    "Contextual data about the samples collected, such as when it was collected, where it was collected from, what kind of sample it is, and what were the properties of the environment or experimental condition from which the sample was taken. Each row is a distinct sample, or Event. Most of this information is recorded during sample collection. This sheet contains terms from the MIMARKS survey water 6.0 package. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>metagenome_sequenced</th>\n",
       "      <th>organism</th>\n",
       "      <th>collection_date_local</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>depth</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>...</th>\n",
       "      <th>carbonate</th>\n",
       "      <th>diss_inorg_carb</th>\n",
       "      <th>diss_oxygen</th>\n",
       "      <th>fluor</th>\n",
       "      <th>hydrogen_ion</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>nitrite</th>\n",
       "      <th>nitrate_plus_nitrite</th>\n",
       "      <th>omega_arag</th>\n",
       "      <th>pco2</th>\n",
       "      <th>ph</th>\n",
       "      <th>phosphate</th>\n",
       "      <th>pressure</th>\n",
       "      <th>salinity</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>date_sheet_modified</th>\n",
       "      <th>modified_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516091</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>...</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>2023-10-03 13:28:31.916</td>\n",
       "      <td>luke.thompson@noaa.gov</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516092</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>...</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516093</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>...</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>...</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516095</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>...</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_name serial_number        cruise_id line_id station  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A   GOMECC4_001  GOMECC-4 (2021)     27N    Sta1   \n",
       "1  GOMECC4_27N_Sta1_Deep_B   GOMECC4_002  GOMECC-4 (2021)     27N    Sta1   \n",
       "2  GOMECC4_27N_Sta1_Deep_C   GOMECC4_003  GOMECC-4 (2021)     27N    Sta1   \n",
       "3   GOMECC4_27N_Sta1_DCM_A   GOMECC4_004  GOMECC-4 (2021)     27N    Sta1   \n",
       "4   GOMECC4_27N_Sta1_DCM_B   GOMECC4_005  GOMECC-4 (2021)     27N    Sta1   \n",
       "\n",
       "  ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0             3                A  GOMECC4_27N_Sta1_Deep   \n",
       "1             3                B  GOMECC4_27N_Sta1_Deep   \n",
       "2             3                C  GOMECC4_27N_Sta1_Deep   \n",
       "3            14                A   GOMECC4_27N_Sta1_DCM   \n",
       "4            14                B   GOMECC4_27N_Sta1_DCM   \n",
       "\n",
       "                              biological_replicates extract_number  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C      Plate4_52   \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C      Plate4_60   \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B      Plate4_62   \n",
       "3    GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C      Plate4_53   \n",
       "4    GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C      Plate4_46   \n",
       "\n",
       "                                             sample_title  \\\n",
       "0  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A   \n",
       "1  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B   \n",
       "2  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C   \n",
       "3   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A   \n",
       "4   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B   \n",
       "\n",
       "  bioproject_accession biosample_accession  amplicon_sequenced  \\\n",
       "0          PRJNA887898        SAMN37516091  16S V4-V5 | 18S V9   \n",
       "1          PRJNA887898        SAMN37516092  16S V4-V5 | 18S V9   \n",
       "2          PRJNA887898        SAMN37516093  16S V4-V5 | 18S V9   \n",
       "3          PRJNA887898        SAMN37516094  16S V4-V5 | 18S V9   \n",
       "4          PRJNA887898        SAMN37516095  16S V4-V5 | 18S V9   \n",
       "\n",
       "  metagenome_sequenced             organism   collection_date_local  \\\n",
       "0     planned for FY24  seawater metagenome  2021-09-14T11:00-04:00   \n",
       "1     planned for FY24  seawater metagenome  2021-09-14T11:00-04:00   \n",
       "2     planned for FY24  seawater metagenome  2021-09-14T11:00-04:00   \n",
       "3     planned for FY24  seawater metagenome  2021-09-14T11:00-04:00   \n",
       "4     planned for FY24  seawater metagenome  2021-09-14T11:00-04:00   \n",
       "\n",
       "    collection_date  depth               env_broad_scale  \\\n",
       "0  2021-09-14T07:00  618 m  marine biome [ENVO:00000447]   \n",
       "1  2021-09-14T07:00  618 m  marine biome [ENVO:00000447]   \n",
       "2  2021-09-14T07:00  618 m  marine biome [ENVO:00000447]   \n",
       "3  2021-09-14T07:00   49 m  marine biome [ENVO:00000447]   \n",
       "4  2021-09-14T07:00   49 m  marine biome [ENVO:00000447]   \n",
       "\n",
       "                           env_local_scale                 env_medium  \\\n",
       "0  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "1  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "2  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "3       marine photic zone [ENVO:00000209]  sea water [ENVO:00002149]   \n",
       "4       marine photic zone [ENVO:00000209]  sea water [ENVO:00002149]   \n",
       "\n",
       "                                  geo_loc_name            lat_lon  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "\n",
       "  decimalLatitude  ...       carbonate  diss_inorg_carb      diss_oxygen  \\\n",
       "0          26.997  ...  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg   \n",
       "1          26.997  ...  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg   \n",
       "2          26.997  ...  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg   \n",
       "3          26.997  ...  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   \n",
       "4          26.997  ...  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   \n",
       "\n",
       "    fluor    hydrogen_ion          nitrate          nitrite  \\\n",
       "0  0.0308  0.0000000142 M  29.3256 µmol/kg  0.00391 µmol/kg   \n",
       "1  0.0308  0.0000000142 M  29.3256 µmol/kg  0.00391 µmol/kg   \n",
       "2  0.0308  0.0000000142 M  29.3256 µmol/kg  0.00391 µmol/kg   \n",
       "3   0.036  0.0000000094 M        0 µmol/kg        0 µmol/kg   \n",
       "4   0.036  0.0000000094 M        0 µmol/kg        0 µmol/kg   \n",
       "\n",
       "  nitrate_plus_nitrite omega_arag      pco2     ph        phosphate  pressure  \\\n",
       "0      29.3295 µmol/kg      1.168  624 µatm  7.849  1.94489 µmol/kg  623 dbar   \n",
       "1      29.3295 µmol/kg      1.168  624 µatm  7.849  1.94489 µmol/kg  623 dbar   \n",
       "2      29.3295 µmol/kg      1.168  624 µatm  7.849  1.94489 µmol/kg  623 dbar   \n",
       "3            0 µmol/kg      3.805  423 µatm  8.027   0.0517 µmol/kg   49 dbar   \n",
       "4            0 µmol/kg      3.805  423 µatm  8.027   0.0517 µmol/kg   49 dbar   \n",
       "\n",
       "     salinity      samp_store_loc samp_store_temp         silicate  \\\n",
       "0  34.946 psu  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg   \n",
       "1  34.946 psu  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg   \n",
       "2  34.946 psu  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg   \n",
       "3  36.325 psu  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg   \n",
       "4  36.325 psu  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg   \n",
       "\n",
       "   size_frac_low size_frac_up       temp  tot_alkalinity tot_depth_water_col  \\\n",
       "0  no pre-filter      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m   \n",
       "1  no pre-filter      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m   \n",
       "2  no pre-filter      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m   \n",
       "3  no pre-filter      0.22 µm  28.592 °C    2371 µmol/kg               623 m   \n",
       "4  no pre-filter      0.22 µm  28.592 °C    2371 µmol/kg               623 m   \n",
       "\n",
       "  transmittance     date_sheet_modified             modified_by  \n",
       "0        4.7221 2023-10-03 13:28:31.916  luke.thompson@noaa.gov  \n",
       "1        4.7221                     NaT                     NaN  \n",
       "2        4.7221                     NaT                     NaN  \n",
       "3         4.665                     NaT                     NaN  \n",
       "4         4.665                     NaT                     NaN  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['sample_data'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### prep_data  \n",
    "Contextual data about how the samples were prepared for sequencing. Includes how they were extracted, what amplicon was targeted, how they were sequenced. Each row is a separate sequencing library preparation, distinguished by a unique library_id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>drive_location</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>seq_method</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "      <th>date_sheet_modified</th>\n",
       "      <th>modified_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_NegativeControl_1</td>\n",
       "      <td>GOMECC16S_Neg1</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Neg1_S499_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Neg1_S499_L001_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN37516589</td>\n",
       "      <td>SRR26148505</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC18S_Plate4_53_S340_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC18S_Plate4_53_S340_L001_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>SRR26161153</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S V9</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>2023-10-03 12:49:23.878</td>\n",
       "      <td>katherine.silliman@noaa.gov</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_name           library_id  \\\n",
       "0  GOMECC4_NegativeControl_1       GOMECC16S_Neg1   \n",
       "1     GOMECC4_27N_Sta1_DCM_A  GOMECC18S_Plate4_53   \n",
       "\n",
       "                                                                   title  \\\n",
       "0  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "1  18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "\n",
       "  library_strategy library_source library_selection lib_layout  platform  \\\n",
       "0         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "1         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "\n",
       "  instrument_model  \\\n",
       "0   Illumina MiSeq   \n",
       "1   Illumina MiSeq   \n",
       "\n",
       "                                                                                                                                      design_description  \\\n",
       "0  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "1  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "\n",
       "  filetype                                       filename  \\\n",
       "0    fastq       GOMECC16S_Neg1_S499_L001_R1_001.fastq.gz   \n",
       "1    fastq  GOMECC18S_Plate4_53_S340_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                       filename2  drive_location  \\\n",
       "0       GOMECC16S_Neg1_S499_L001_R2_001.fastq.gz             NaN   \n",
       "1  GOMECC18S_Plate4_53_S340_L001_R2_001.fastq.gz             NaN   \n",
       "\n",
       "  biosample_accession sra_accession            seq_method  \\\n",
       "0        SAMN37516589   SRR26148505  Illumina MiSeq 2x250   \n",
       "1        SAMN37516094   SRR26161153  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                          nucl_acid_ext  \\\n",
       "0  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "1  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "  amplicon_sequenced target_gene target_subfragment   pcr_primer_forward  \\\n",
       "0          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "1             18S V9    18S rRNA                 V9     GTACACACCGCCCGTC   \n",
       "\n",
       "         pcr_primer_reverse pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "0      CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "\n",
       "           pcr_primer_reference  \\\n",
       "0       10.1111/1462-2920.13023   \n",
       "1  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                                                                                              pcr_cond  \\\n",
       "0          initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "1  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "\n",
       "                  nucl_acid_amp  \\\n",
       "0       10.1111/1462-2920.13023   \n",
       "1  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                        adapters            mid_barcode  \\\n",
       "0  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided   \n",
       "1  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided   \n",
       "\n",
       "      date_sheet_modified                  modified_by  \n",
       "0                     NaT                          NaN  \n",
       "1 2023-10-03 12:49:23.878  katherine.silliman@noaa.gov  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load ASV data  \n",
    "There is one ASV file for each marker that was sequenced. The ASV data files have one row for each unique amplicon sequence variants (ASVs). They contain the ASV DNA sequence, a unique hash identifier the taxonomic assignment for each ASV, the confidence given that assignment by the naive-bayes classifier, and then the number of reads observed in each sample. \n",
    "\n",
    "This file is created automatically with [Tourmaline v.2023.5+](https://github.com/aomlomics/tourmaline), and is found in `01-taxonomy/asv_taxa_sample_table.tsv`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| column name    | definition                                                                                                                                                                                                                                                                                                                                                                                              |\n",
    "|----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| featureid      | A hash of the ASV sequence, used as a unique identifier for the ASV.                                                                                                                                                                                                                                                                                                                                    |\n",
    "| sequence       | The DNA sequence of the ASV                                                                                                                                                                                                                                                                                                                                                                             |\n",
    "| taxonomy       | The full taxonomy assigned to an ASV sequence. This string could be formatted in very different ways depending on the reference database used during classification, however it should always be in reverse rank order separated by ;. We provide examples for how to process results from a Silva classifier and the PR2 18S classifier. For other taxonomy formats, the code will need to be adapted. |\n",
    "| Confidence     | This is the confidence score assigned the taxonomic classification with a naive-bayes classifier.                                                                                                                                                                                                                                                                                                       |\n",
    "| sample columns | The next columns each represent a sample (or eventID), and the number of reads for that ASV observed in the sample.                                                                                                                                                                                                                                                                                     |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in ASV tables, looping through amplicons\n",
    "asv_tables = {}\n",
    "\n",
    "for gene in params['datafiles'].keys():\n",
    "    asv_tables[gene] = pd.read_table(params['datafiles'][gene])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['16S V4-V5', '18S V9'])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...   \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...   \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...   \n",
       "\n",
       "                                                                                                             taxonomy  \\\n",
       "0                  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1      d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4              d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0    0.978926                       0                       0   \n",
       "1    0.881698                       0                       0   \n",
       "2    0.762793                       0                       0   \n",
       "3    0.997619                       0                       0   \n",
       "4    0.999961                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C  GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  \\\n",
       "0                       0                        0                        0   \n",
       "1                       0                        0                        0   \n",
       "2                       0                        0                        0   \n",
       "3                       0                        0                        0   \n",
       "4                       0                        0                        0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_C  GOMECC4_27N_Sta1_Surface_A  \\\n",
       "0                        0                           0   \n",
       "1                        0                           0   \n",
       "2                        0                           0   \n",
       "3                        0                           0   \n",
       "4                        0                           0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Surface_B  GOMECC4_27N_Sta4_DCM_A  GOMECC4_27N_Sta4_DCM_B  \\\n",
       "0                           0                       0                       0   \n",
       "1                           0                       0                       0   \n",
       "2                           0                       0                       0   \n",
       "3                           0                       0                       0   \n",
       "4                           0                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_DCM_C  GOMECC4_27N_Sta4_Deep_A  GOMECC4_27N_Sta4_Deep_B  \\\n",
       "0                       0                        0                        0   \n",
       "1                       0                        0                        0   \n",
       "2                       0                        0                        0   \n",
       "3                       0                        0                        0   \n",
       "4                       0                        0                        0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_Deep_C  GOMECC4_27N_Sta4_Surface_A  \\\n",
       "0                        0                           0   \n",
       "1                        0                           0   \n",
       "2                        0                           0   \n",
       "3                        0                           0   \n",
       "4                        0                           0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_Surface_B  \n",
       "0                          25  \n",
       "1                           0  \n",
       "2                           0  \n",
       "3                           0  \n",
       "4                           0  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S V4-V5'].iloc[:,0:20].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Drop samples with unwanted sample types  \n",
    "\n",
    "Often with eDNA projects, we have control samples that are sequenced along with our survey samples. These can include filtering distilled water, using pure water instead of DNA in a PCR or DNA extraction protocol, or a mock community of known microbial taxa. Controls can help identify and mitigate contaminant DNA in our samples, but are not useful for biodiversity platforms like OBIS. You can select which sample_type values to drop with the `skip_sample_types` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "samps_to_remove = data['sample_data']['sample_type'].isin(params['skip_sample_types'])\n",
    "#data['sample_data'][samps_to_remove]\n",
    "# list of samples to drop\n",
    "samples_to_drop = data['sample_data']['sample_name'][samps_to_remove]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can view the list of samples to be dropped below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26     GOMECC4_Blank_DIW_20210915_A\n",
       "27     GOMECC4_Blank_DIW_20210915_B\n",
       "28     GOMECC4_Blank_DIW_20210915_C\n",
       "200    GOMECC4_Blank_DIW_20210930_A\n",
       "201    GOMECC4_Blank_DIW_20210930_B\n",
       "202    GOMECC4_Blank_DIW_20210930_C\n",
       "334    GOMECC4_Blank_DIW_20211011_A\n",
       "335    GOMECC4_Blank_DIW_20211011_B\n",
       "336    GOMECC4_Blank_DIW_20211011_C\n",
       "409    GOMECC4_Blank_DIW_20211016_A\n",
       "410    GOMECC4_Blank_DIW_20211016_B\n",
       "411    GOMECC4_Blank_DIW_20211016_C\n",
       "484       GOMECC4_ExtractionBlank_1\n",
       "485      GOMECC4_ExtractionBlank_11\n",
       "486      GOMECC4_ExtractionBlank_12\n",
       "487       GOMECC4_ExtractionBlank_3\n",
       "488       GOMECC4_ExtractionBlank_5\n",
       "489       GOMECC4_ExtractionBlank_7\n",
       "490       GOMECC4_ExtractionBlank_9\n",
       "491            GOMECC4_MSUControl_1\n",
       "492            GOMECC4_MSUControl_2\n",
       "493            GOMECC4_MSUControl_3\n",
       "494            GOMECC4_MSUControl_4\n",
       "495            GOMECC4_MSUControl_5\n",
       "496            GOMECC4_MSUControl_6\n",
       "497            GOMECC4_MSUControl_7\n",
       "498       GOMECC4_NegativeControl_1\n",
       "499       GOMECC4_NegativeControl_2\n",
       "500       GOMECC4_PositiveControl_1\n",
       "501       GOMECC4_PositiveControl_2\n",
       "Name: sample_name, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_to_drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove samples from sample_data sheet\n",
    "data['sample_data'] = data['sample_data'][~samps_to_remove]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['seawater'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the sample_type values left in your sample_data. We only want seawater.\n",
    "data['sample_data']['sample_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove samples from prep_data\n",
    "prep_samps_to_remove = data['prep_data']['sample_name'].isin(samples_to_drop)\n",
    "data['prep_data'] = data['prep_data'][~prep_samps_to_remove]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### drop unwanted samples from ASV files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in params['datafiles'].keys():\n",
    "    asv_tables[gene] = asv_tables[gene].drop(columns=samples_to_drop,errors='ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Drop columns with all NAs  \n",
    "\n",
    "If your project data file has columns with only NAs, this code will check for those, provide their column headers for verification, then remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which have all NAs?\n",
    "dropped = pd.DataFrame()\n",
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    res = pd.Series(data[sheet].columns[data[sheet].isnull().all(0)],\n",
    "                name=sheet)\n",
    "    dropped=pd.concat([dropped,res],axis=1)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which columns in each sheet have only NA values?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_data</th>\n",
       "      <th>prep_data</th>\n",
       "      <th>analysis_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samp_size</td>\n",
       "      <td>drive_location</td>\n",
       "      <td>sop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_data       prep_data analysis_data\n",
       "0   samp_size  drive_location           sop"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dropped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you are fine with leaving these columns out, proceed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    data[sheet].dropna(axis=1, how='all',inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's check which columns have missing values in some of the rows. These should be filled in on the Excel sheet with the appropriate term ('not applicable', 'missing', or 'not collected'). Alternatively, you can drop the column if it is not needed for submission to OBIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which columns have missing data (NAs) in some rows\n",
    "some = pd.DataFrame()\n",
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    res = pd.Series(data[sheet].columns[data[sheet].isnull().any()].tolist(),\n",
    "                name=sheet)\n",
    "    some=pd.concat([some,res],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_data</th>\n",
       "      <th>prep_data</th>\n",
       "      <th>analysis_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>notes_bottle_metadata</td>\n",
       "      <td>date_sheet_modified</td>\n",
       "      <td>date_sheet_modified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>date_sheet_modified</td>\n",
       "      <td>modified_by</td>\n",
       "      <td>modified_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>modified_by</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample_data            prep_data        analysis_data\n",
       "0  notes_bottle_metadata  date_sheet_modified  date_sheet_modified\n",
       "1    date_sheet_modified          modified_by          modified_by\n",
       "2            modified_by                  NaN                  NaN"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "some"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I'm going to drop all the columns with some missing data, as I don't need them for submission to OBIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop columns with any missing data\n",
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    data[sheet].dropna(axis=1, how='any',inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load data dictionary Excel file \n",
    "This Excel file is used as a data dictionary for converting between terms used in the project data Excel file and Darwin Core terms for submission to OBIS. Currently, we are only preparing an Occurrence core file and a DNA-derived extension file, with Event information in the Occurrence file. Future versions of this workflow will prepare an extendedMeasurementOrFact file as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in data dictionary excel file\n",
    "dwc_data = pd.read_excel(\n",
    "    params['md_excel'], \n",
    "    ['event','occurrence','dna'],\n",
    "    index_col=0, na_values=[\"\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AOML_term</th>\n",
       "      <th>AOML_file</th>\n",
       "      <th>DwC_definition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DwC_term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>eventID</th>\n",
       "      <td>sample_name</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventDate</th>\n",
       "      <td>collection_date_local</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.td...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplingProtocol</th>\n",
       "      <td>collection_method</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locationID</th>\n",
       "      <td>station</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLatitude</th>\n",
       "      <td>decimalLatitude</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              AOML_term    AOML_file  \\\n",
       "DwC_term                                               \n",
       "eventID                     sample_name  sample_data   \n",
       "eventDate         collection_date_local  sample_data   \n",
       "samplingProtocol      collection_method  sample_data   \n",
       "locationID                      station  sample_data   \n",
       "decimalLatitude         decimalLatitude  sample_data   \n",
       "\n",
       "                                                                                                                                                         DwC_definition  \n",
       "DwC_term                                                                                                                                                                 \n",
       "eventID           An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:...  \n",
       "eventDate         this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.td...  \n",
       "samplingProtocol                                                      The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.  \n",
       "locationID                      An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.  \n",
       "decimalLatitude   The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example of a sheet in the data dictionary\n",
    "dwc_data['event'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert to Occurrence file\n",
    "In order to link the DNA-derived extension metadata to our OBIS occurrence records, we have to use the Occurrence core. For this data set, a `parentEvent` is a filtered water sample that was DNA extracted, a sequencing library from that DNA extraction is an `event`, and an `occurrence` is an ASV observed within a library. We will have an an occurence file and a DNA derived data file. Future versions will generate a measurements file.   \n",
    "**Define files**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Sampling event info \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AOML_term</th>\n",
       "      <th>AOML_file</th>\n",
       "      <th>DwC_definition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DwC_term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>eventID</th>\n",
       "      <td>sample_name</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventDate</th>\n",
       "      <td>collection_date_local</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.td...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplingProtocol</th>\n",
       "      <td>collection_method</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locationID</th>\n",
       "      <td>station</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLatitude</th>\n",
       "      <td>decimalLatitude</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLongitude</th>\n",
       "      <td>decimalLongitude</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The geographic longitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geodeticDatum</th>\n",
       "      <td>none</td>\n",
       "      <td>pipeline</td>\n",
       "      <td>The ellipsoid, geodetic datum, or spatial reference system (SRS) upon which the geographic coordinates given in dwc:decimalLatitude and dwc:decima...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>countryCode</th>\n",
       "      <td>none</td>\n",
       "      <td>pipeline</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <td>depth</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <td>derived: depth</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>datasetID</th>\n",
       "      <td>project_id_external</td>\n",
       "      <td>study_data</td>\n",
       "      <td>An identifier for the set of data. May be a global unique identifier or an identifier specific to a collection or institution.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>waterBody</th>\n",
       "      <td>derived</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locality</th>\n",
       "      <td>geo_loc_name</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventRemarks</th>\n",
       "      <td>derived: controls_used</td>\n",
       "      <td>analysis_data</td>\n",
       "      <td>Comments or notes about the dwc:Event.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   AOML_term      AOML_file  \\\n",
       "DwC_term                                                      \n",
       "eventID                          sample_name    sample_data   \n",
       "eventDate              collection_date_local    sample_data   \n",
       "samplingProtocol           collection_method    sample_data   \n",
       "locationID                           station    sample_data   \n",
       "decimalLatitude              decimalLatitude    sample_data   \n",
       "decimalLongitude            decimalLongitude    sample_data   \n",
       "geodeticDatum                           none       pipeline   \n",
       "countryCode                             none       pipeline   \n",
       "minimumDepthInMeters                   depth    sample_data   \n",
       "maximumDepthInMeters          derived: depth    sample_data   \n",
       "datasetID                project_id_external     study_data   \n",
       "waterBody                            derived    sample_data   \n",
       "locality                        geo_loc_name    sample_data   \n",
       "eventRemarks          derived: controls_used  analysis_data   \n",
       "\n",
       "                                                                                                                                                             DwC_definition  \n",
       "DwC_term                                                                                                                                                                     \n",
       "eventID               An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:...  \n",
       "eventDate             this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.td...  \n",
       "samplingProtocol                                                          The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.  \n",
       "locationID                          An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.  \n",
       "decimalLatitude       The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:...  \n",
       "decimalLongitude      The geographic longitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms...  \n",
       "geodeticDatum         The ellipsoid, geodetic datum, or spatial reference system (SRS) upon which the geographic coordinates given in dwc:decimalLatitude and dwc:decima...  \n",
       "countryCode                                                                                                                                                             NaN  \n",
       "minimumDepthInMeters                                                                                                                                                    NaN  \n",
       "maximumDepthInMeters                                                                                                                                                    NaN  \n",
       "datasetID                                    An identifier for the set of data. May be a global unique identifier or an identifier specific to a collection or institution.  \n",
       "waterBody             The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the G...  \n",
       "locality                                                                                                                                                                NaN  \n",
       "eventRemarks                                                                                                                         Comments or notes about the dwc:Event.  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dwc_data['event']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "event_dict = dwc_data['event'].to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AOML_term': 'sample_name',\n",
       " 'AOML_file': 'sample_data',\n",
       " 'DwC_definition': 'An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:eventID'}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_dict['eventID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "maximumDepthInMeters {'AOML_term': 'derived: depth', 'AOML_file': 'sample_data', 'DwC_definition': nan}\n",
      "waterBody {'AOML_term': 'derived', 'AOML_file': 'sample_data', 'DwC_definition': 'The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the Getty Thesaurus of Geographic Names.'}\n"
     ]
    }
   ],
   "source": [
    "# check which event terms are not in sample_data sheet\n",
    "for key in event_dict.keys():\n",
    "    if event_dict[key]['AOML_file'] == 'sample_data':\n",
    "        if event_dict[key]['AOML_term'] not in data['sample_data'].columns:\n",
    "            print(key,event_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_f/3hfwkwps2rq9q60vkv4fnd_n9rf1vk/T/ipykernel_5945/2121370784.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Atlantic Ocean' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  data['sample_data'].loc[data['sample_data']['geo_loc_name'].str.contains(\"Atlantic Ocean\"), 'waterBody']= \"Atlantic Ocean\"\n"
     ]
    }
   ],
   "source": [
    "# custom add waterBody\n",
    "\n",
    "data['sample_data'].loc[data['sample_data']['geo_loc_name'].str.contains(\"Atlantic Ocean\"), 'waterBody']= \"Atlantic Ocean\"\n",
    "data['sample_data'].loc[data['sample_data']['geo_loc_name'].str.contains(\"Gulf\"), 'waterBody']= \"Mexico, Gulf of\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change locationID to line_id+station\n",
    "data['sample_data']['station'] = data['sample_data']['line_id']+ \"_\"+data['sample_data']['station'] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename sample_data columns to fit DwC standard\n",
    "gen = (x for x in event_dict.keys() if event_dict[x]['AOML_file'] == 'sample_data')\n",
    "rename_dict = {}\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[event_dict[x]['AOML_term']] = x\n",
    "\n",
    "event_sample = data['sample_data'].rename(columns=rename_dict)\n",
    "event_sample = event_sample.drop(columns=[col for col in event_sample if col not in rename_dict.values()])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add maximumDepthInMeters\n",
    "#remove m in depth\n",
    "event_sample['minimumDepthInMeters'] = event_sample['minimumDepthInMeters'].str.strip(\" m\")\n",
    "event_sample['maximumDepthInMeters'] = event_sample['minimumDepthInMeters']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "3   GOMECC4_27N_Sta1_DCM_A   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "4   GOMECC4_27N_Sta1_DCM_B   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "1          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "2          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "3          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "4          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters  \n",
       "0                  618  \n",
       "1                  618  \n",
       "2                  618  \n",
       "3                   49  \n",
       "4                   49  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "event_sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add amplicon_sequenced back \n",
    "event_sample['amplicon_sequenced'] = data['sample_data']['amplicon_sequenced']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now add an event for each sequencing library, with replicate water sample as the parentEvent.  \n",
    "\n",
    "**Future Update**: make this a for loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>parentEventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "1          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "2          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "3          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "4          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters  amplicon_sequenced            parentEventID  \n",
       "0                  618  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_A  \n",
       "1                  618  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_B  \n",
       "2                  618  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_C  \n",
       "3                   49  16S V4-V5 | 18S V9   GOMECC4_27N_Sta1_DCM_A  \n",
       "4                   49  16S V4-V5 | 18S V9   GOMECC4_27N_Sta1_DCM_B  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "child_data_16S = event_sample[event_sample['amplicon_sequenced'].str.contains('16S V4-V5')].copy()\n",
    "child_data_16S['parentEventID'] = child_data_16S['eventID']\n",
    "child_data_16S['eventID'] = child_data_16S['eventID']+\"_16S\"\n",
    "child_data_16S.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>parentEventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_18S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_18S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_18S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_18S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_18S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "1          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "2          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "3          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "4          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters  amplicon_sequenced            parentEventID  \n",
       "0                  618  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_A  \n",
       "1                  618  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_B  \n",
       "2                  618  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_C  \n",
       "3                   49  16S V4-V5 | 18S V9   GOMECC4_27N_Sta1_DCM_A  \n",
       "4                   49  16S V4-V5 | 18S V9   GOMECC4_27N_Sta1_DCM_B  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "child_data_18S = event_sample[event_sample['amplicon_sequenced'].str.contains('18S V9')].copy()\n",
    "child_data_18S['parentEventID'] = child_data_18S['eventID']\n",
    "child_data_18S['eventID'] = child_data_18S['eventID']+\"_18S\"\n",
    "child_data_18S.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is your full event file\n",
    "all_event_data = pd.concat([child_data_16S,child_data_18S],axis=0,ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_event_data = all_event_data.drop(columns=['amplicon_sequenced'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>59</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>59</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    eventID        locationID  \\\n",
       "939      GOMECC4_CAPECORAL_Sta141_DCM_B_18S  CAPECORAL_Sta141   \n",
       "940      GOMECC4_CAPECORAL_Sta141_DCM_C_18S  CAPECORAL_Sta141   \n",
       "941  GOMECC4_CAPECORAL_Sta141_Surface_A_18S  CAPECORAL_Sta141   \n",
       "942  GOMECC4_CAPECORAL_Sta141_Surface_B_18S  CAPECORAL_Sta141   \n",
       "943  GOMECC4_CAPECORAL_Sta141_Surface_C_18S  CAPECORAL_Sta141   \n",
       "\n",
       "                  eventDate minimumDepthInMeters             locality  \\\n",
       "939  2021-10-20T12:47-04:00                   59  USA: Gulf of Mexico   \n",
       "940  2021-10-20T12:47-04:00                   59  USA: Gulf of Mexico   \n",
       "941  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "942  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "943  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "\n",
       "    decimalLatitude decimalLongitude samplingProtocol        waterBody  \\\n",
       "939          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "940          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "941          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "942          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "943          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "\n",
       "    maximumDepthInMeters                       parentEventID  \n",
       "939                   59      GOMECC4_CAPECORAL_Sta141_DCM_B  \n",
       "940                   59      GOMECC4_CAPECORAL_Sta141_DCM_C  \n",
       "941                    4  GOMECC4_CAPECORAL_Sta141_Surface_A  \n",
       "942                    4  GOMECC4_CAPECORAL_Sta141_Surface_B  \n",
       "943                    4  GOMECC4_CAPECORAL_Sta141_Surface_C  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_event_data.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which terms are still missing from the event info?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "geodeticDatum {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': 'The ellipsoid, geodetic datum, or spatial reference system (SRS) upon which the geographic coordinates given in dwc:decimalLatitude and dwc:decimalLongitude are based.'}\n",
      "countryCode {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': nan}\n",
      "datasetID {'AOML_term': 'project_id_external', 'AOML_file': 'study_data', 'DwC_definition': 'An identifier for the set of data. May be a global unique identifier or an identifier specific to a collection or institution.'}\n",
      "eventRemarks {'AOML_term': 'derived: controls_used', 'AOML_file': 'analysis_data', 'DwC_definition': 'Comments or notes about the dwc:Event.'}\n"
     ]
    }
   ],
   "source": [
    "for key in event_dict.keys():\n",
    "    if event_dict[key]['AOML_file'] != 'sample_data':\n",
    "        print(key,event_dict[key])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "countryCode, leave blank because it spans multiple countries. eventRemarks will be added later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#datasetID\n",
    "all_event_data['datasetID'] = data['study_data']['project_id_external'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#geodeticDatum\n",
    "all_event_data['geodeticDatum'] = \"WGS84\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_16S   27N_Sta1  2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2                  618  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4                   49  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "1          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "2          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "3          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "4          26.997          -79.618      CTD rosette  Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters            parentEventID          datasetID  \\\n",
       "0                  618  GOMECC4_27N_Sta1_Deep_A  noaa-aoml-gomecc4   \n",
       "1                  618  GOMECC4_27N_Sta1_Deep_B  noaa-aoml-gomecc4   \n",
       "2                  618  GOMECC4_27N_Sta1_Deep_C  noaa-aoml-gomecc4   \n",
       "3                   49   GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4   \n",
       "4                   49   GOMECC4_27N_Sta1_DCM_B  noaa-aoml-gomecc4   \n",
       "\n",
       "  geodeticDatum  \n",
       "0         WGS84  \n",
       "1         WGS84  \n",
       "2         WGS84  \n",
       "3         WGS84  \n",
       "4         WGS84  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_event_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Occurrence file  \n",
    "Now get the occurrence info from the ASV tables, format it, then merge it with the event info."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# create a dictionary to hold both markers\n",
    "occ = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 18S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>1516</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4257</td>\n",
       "      <td>2005</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>962</td>\n",
       "      <td>316</td>\n",
       "      <td>548</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>613</td>\n",
       "      <td>561</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5d4df37251121c08397c6fbc27b06175</td>\n",
       "      <td>GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCCTGGCGGTTAACGCTGCCTAGTTGGCGAAAAGACGACCAAACTGTAGCACTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Sinocalanus;Sinocalanus_sinensis;</td>\n",
       "      <td>0.992300</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f863f671a575c6ab587e8de0190d3335</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCCAGGCGGGTCGCCCTGCCTGGTCTACGGGAAGACGACCAAACTGTAGTGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Paracalanus;Paracalanus_parvus;</td>\n",
       "      <td>0.998393</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>1164</td>\n",
       "      <td>2272</td>\n",
       "      <td>2208</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>287</td>\n",
       "      <td>414</td>\n",
       "      <td>335</td>\n",
       "      <td>195</td>\n",
       "      <td>228</td>\n",
       "      <td>298</td>\n",
       "      <td>252</td>\n",
       "      <td>349</td>\n",
       "      <td>175</td>\n",
       "      <td>102</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>d70494a723d85d66aa88d2d8a975aeec</td>\n",
       "      <td>GCTACTACCGATTGAATGGTTCCGTGAATTCTTGAGATCGGCGCGGGAACAACTGGCAACGGTTGATCCCGATTGCTGAGAACTTGTGTAAACGCGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta</td>\n",
       "      <td>0.992451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>250</td>\n",
       "      <td>323</td>\n",
       "      <td>194</td>\n",
       "      <td>51</td>\n",
       "      <td>59</td>\n",
       "      <td>55</td>\n",
       "      <td>222</td>\n",
       "      <td>250</td>\n",
       "      <td>230</td>\n",
       "      <td>163</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bbaaf7bb4e71c80de970677779e3bf3a</td>\n",
       "      <td>GCTACTACCGATTGAATGGTTTAGTGAGATCTTCGGATTGGCACAATCGCGGCCTAACGGAAGTGATGGTGCCGAAAAGTTGCTCAAACTTGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Cnidaria;Cnidaria_X;Hydrozoa;Sulculeolaria;Sulculeolaria_quadrivalvis;</td>\n",
       "      <td>0.864777</td>\n",
       "      <td>212</td>\n",
       "      <td>50</td>\n",
       "      <td>237</td>\n",
       "      <td>552</td>\n",
       "      <td>1278</td>\n",
       "      <td>480</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>24</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7a8324bb4448b65f7adc73d70e5901da</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTATTTGGACTGGGCCTTTGGAGGATTCGTTCTCCAATGTTGCTCGGGAAGACTCCCAAACTTGAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Delibus;Delibus_sp.;</td>\n",
       "      <td>0.992088</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "2  5d4df37251121c08397c6fbc27b06175   \n",
       "3  f863f671a575c6ab587e8de0190d3335   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "5  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "6  d70494a723d85d66aa88d2d8a975aeec   \n",
       "7  fa1f1a97dd4ae7c826009186bad26384   \n",
       "8  bbaaf7bb4e71c80de970677779e3bf3a   \n",
       "9  7a8324bb4448b65f7adc73d70e5901da   \n",
       "\n",
       "                                                                                                                                sequence  \\\n",
       "0        GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1        GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2        GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCCTGGCGGTTAACGCTGCCTAGTTGGCGAAAAGACGACCAAACTGTAGCACTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3        GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCCAGGCGGGTCGCCCTGCCTGGTCTACGGGAAGACGACCAAACTGTAGTGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "4       GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "5         GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "6  GCTACTACCGATTGAATGGTTCCGTGAATTCTTGAGATCGGCGCGGGAACAACTGGCAACGGTTGATCCCGATTGCTGAGAACTTGTGTAAACGCGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "7         GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "8      GCTACTACCGATTGAATGGTTTAGTGAGATCTTCGGATTGGCACAATCGCGGCCTAACGGAAGTGATGGTGCCGAAAAGTTGCTCAAACTTGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "9      GCTACTACCGATTGAACGTTTTAGTGAGGTATTTGGACTGGGCCTTTGGAGGATTCGTTCTCCAATGTTGCTCGGGAAGACTCCCAAACTTGAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                                                                                                       taxonomy  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2      Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Sinocalanus;Sinocalanus_sinensis;   \n",
       "3        Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Paracalanus;Paracalanus_parvus;   \n",
       "4           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "5                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "6                                                                                 Eukaryota;Obazoa;Opisthokonta   \n",
       "7                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "8  Eukaryota;Obazoa;Opisthokonta;Metazoa;Cnidaria;Cnidaria_X;Hydrozoa;Sulculeolaria;Sulculeolaria_quadrivalvis;   \n",
       "9                   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Delibus;Delibus_sp.;   \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0    0.922099                    1516                       0   \n",
       "1    0.999947                     962                     316   \n",
       "2    0.992300                       0                       4   \n",
       "3    0.998393                       0                       0   \n",
       "4    0.779948                    1164                    2272   \n",
       "5    0.999931                     287                     414   \n",
       "6    0.992451                       0                       0   \n",
       "7    0.986908                     250                     323   \n",
       "8    0.864777                     212                      50   \n",
       "9    0.992088                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C  GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  \\\n",
       "0                       0                        6                        0   \n",
       "1                     548                       19                       10   \n",
       "2                       0                       12                        5   \n",
       "3                       0                        0                        0   \n",
       "4                    2208                        2                        0   \n",
       "5                     335                      195                      228   \n",
       "6                       0                        4                        0   \n",
       "7                     194                       51                       59   \n",
       "8                     237                      552                     1278   \n",
       "9                       0                       15                        0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_C  GOMECC4_27N_Sta1_Surface_A  \\\n",
       "0                        0                           0   \n",
       "1                        0                           0   \n",
       "2                        0                           0   \n",
       "3                        0                           0   \n",
       "4                        0                           0   \n",
       "5                      298                         252   \n",
       "6                        0                           0   \n",
       "7                       55                         222   \n",
       "8                      480                           0   \n",
       "9                        0                           0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Surface_B  GOMECC4_27N_Sta4_DCM_A  GOMECC4_27N_Sta4_DCM_B  \\\n",
       "0                        4257                    2005                       0   \n",
       "1                           0                     613                     561   \n",
       "2                           0                       9                       0   \n",
       "3                           0                       0                       0   \n",
       "4                           0                       0                       0   \n",
       "5                         349                     175                     102   \n",
       "6                           0                       5                       0   \n",
       "7                         250                     230                     163   \n",
       "8                           0                      26                      24   \n",
       "9                           0                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_DCM_C  \n",
       "0                      14  \n",
       "1                     434  \n",
       "2                       0  \n",
       "3                       5  \n",
       "4                       0  \n",
       "5                     216  \n",
       "6                      22  \n",
       "7                     214  \n",
       "8                      21  \n",
       "9                       0  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['18S V9'].iloc[0:10,0:15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### format taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to automate this? Everyone's taxonomy might be different?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#18S \n",
    "taxa_ranks_18S = ['domain','supergroup','division','subdivision','class','order','family','genus','species']\n",
    "\n",
    "asv_tables['18S V9'][['domain','supergroup','division','subdivision','class','order','family','genus','species']] = [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\"]\n",
    "for index, row in asv_tables['18S V9'].iterrows():\n",
    "    taxa = row['taxonomy'].split(\";\")\n",
    "    for i in range(0,len(taxa)):\n",
    "        if i < len(taxa_ranks_18S):\n",
    "            asv_tables['18S V9'].loc[index,taxa_ranks_18S[i]] = taxa[i]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>domain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>subdivision</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>1516</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>962</td>\n",
       "      <td>316</td>\n",
       "      <td>548</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5d4df37251121c08397c6fbc27b06175</td>\n",
       "      <td>GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCCTGGCGGTTAACGCTGCCTAGTTGGCGAAAAGACGACCAAACTGTAGCACTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Sinocalanus;Sinocalanus_sinensis;</td>\n",
       "      <td>0.992300</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Sinocalanus</td>\n",
       "      <td>Sinocalanus sinensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f863f671a575c6ab587e8de0190d3335</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCCAGGCGGGTCGCCCTGCCTGGTCTACGGGAAGACGACCAAACTGTAGTGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Paracalanus;Paracalanus_parvus;</td>\n",
       "      <td>0.998393</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Paracalanus parvus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>1164</td>\n",
       "      <td>2272</td>\n",
       "      <td>2208</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Acrocalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>287</td>\n",
       "      <td>414</td>\n",
       "      <td>335</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>d70494a723d85d66aa88d2d8a975aeec</td>\n",
       "      <td>GCTACTACCGATTGAATGGTTCCGTGAATTCTTGAGATCGGCGCGGGAACAACTGGCAACGGTTGATCCCGATTGCTGAGAACTTGTGTAAACGCGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta</td>\n",
       "      <td>0.992451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>250</td>\n",
       "      <td>323</td>\n",
       "      <td>194</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bbaaf7bb4e71c80de970677779e3bf3a</td>\n",
       "      <td>GCTACTACCGATTGAATGGTTTAGTGAGATCTTCGGATTGGCACAATCGCGGCCTAACGGAAGTGATGGTGCCGAAAAGTTGCTCAAACTTGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Cnidaria;Cnidaria_X;Hydrozoa;Sulculeolaria;Sulculeolaria_quadrivalvis;</td>\n",
       "      <td>0.864777</td>\n",
       "      <td>212</td>\n",
       "      <td>50</td>\n",
       "      <td>237</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>Cnidaria X</td>\n",
       "      <td>Hydrozoa</td>\n",
       "      <td>Sulculeolaria</td>\n",
       "      <td>Sulculeolaria quadrivalvis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7a8324bb4448b65f7adc73d70e5901da</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTATTTGGACTGGGCCTTTGGAGGATTCGTTCTCCAATGTTGCTCGGGAAGACTCCCAAACTTGAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Delibus;Delibus_sp.;</td>\n",
       "      <td>0.992088</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Delibus</td>\n",
       "      <td>Delibus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "2  5d4df37251121c08397c6fbc27b06175   \n",
       "3  f863f671a575c6ab587e8de0190d3335   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "5  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "6  d70494a723d85d66aa88d2d8a975aeec   \n",
       "7  fa1f1a97dd4ae7c826009186bad26384   \n",
       "8  bbaaf7bb4e71c80de970677779e3bf3a   \n",
       "9  7a8324bb4448b65f7adc73d70e5901da   \n",
       "\n",
       "                                                                                                                                sequence  \\\n",
       "0        GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1        GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2        GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCCTGGCGGTTAACGCTGCCTAGTTGGCGAAAAGACGACCAAACTGTAGCACTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3        GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCCAGGCGGGTCGCCCTGCCTGGTCTACGGGAAGACGACCAAACTGTAGTGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "4       GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "5         GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "6  GCTACTACCGATTGAATGGTTCCGTGAATTCTTGAGATCGGCGCGGGAACAACTGGCAACGGTTGATCCCGATTGCTGAGAACTTGTGTAAACGCGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "7         GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "8      GCTACTACCGATTGAATGGTTTAGTGAGATCTTCGGATTGGCACAATCGCGGCCTAACGGAAGTGATGGTGCCGAAAAGTTGCTCAAACTTGATCATTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "9      GCTACTACCGATTGAACGTTTTAGTGAGGTATTTGGACTGGGCCTTTGGAGGATTCGTTCTCCAATGTTGCTCGGGAAGACTCCCAAACTTGAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                                                                                                       taxonomy  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2      Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Sinocalanus;Sinocalanus_sinensis;   \n",
       "3        Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Paracalanus;Paracalanus_parvus;   \n",
       "4           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "5                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "6                                                                                 Eukaryota;Obazoa;Opisthokonta   \n",
       "7                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "8  Eukaryota;Obazoa;Opisthokonta;Metazoa;Cnidaria;Cnidaria_X;Hydrozoa;Sulculeolaria;Sulculeolaria_quadrivalvis;   \n",
       "9                   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Delibus;Delibus_sp.;   \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0    0.922099                    1516                       0   \n",
       "1    0.999947                     962                     316   \n",
       "2    0.992300                       0                       4   \n",
       "3    0.998393                       0                       0   \n",
       "4    0.779948                    1164                    2272   \n",
       "5    0.999931                     287                     414   \n",
       "6    0.992451                       0                       0   \n",
       "7    0.986908                     250                     323   \n",
       "8    0.864777                     212                      50   \n",
       "9    0.992088                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C     domain supergroup      division     subdivision  \\\n",
       "0                       0  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "1                     548  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "2                       0  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "3                       0  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "4                    2208  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "5                     335  Eukaryota       TSAR     Alveolata  Dinoflagellata   \n",
       "6                       0  Eukaryota     Obazoa  Opisthokonta             NaN   \n",
       "7                     194  Eukaryota       TSAR     Alveolata  Dinoflagellata   \n",
       "8                     237  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "9                       0  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "\n",
       "         class          order          family          genus  \\\n",
       "0   Arthropoda      Crustacea     Maxillopoda     Neocalanus   \n",
       "1   Arthropoda      Crustacea     Maxillopoda  Clausocalanus   \n",
       "2   Arthropoda      Crustacea     Maxillopoda    Sinocalanus   \n",
       "3   Arthropoda      Crustacea     Maxillopoda    Paracalanus   \n",
       "4   Arthropoda      Crustacea     Maxillopoda    Acrocalanus   \n",
       "5  Dinophyceae            NaN             NaN            NaN   \n",
       "6          NaN            NaN             NaN            NaN   \n",
       "7  Dinophyceae  Gymnodiniales  Gymnodiniaceae            NaN   \n",
       "8     Cnidaria     Cnidaria X        Hydrozoa  Sulculeolaria   \n",
       "9   Arthropoda      Crustacea     Maxillopoda        Delibus   \n",
       "\n",
       "                      species  \n",
       "0        Neocalanus cristatus  \n",
       "1      Clausocalanus furcatus  \n",
       "2        Sinocalanus sinensis  \n",
       "3          Paracalanus parvus  \n",
       "4                 Acrocalanus  \n",
       "5                         NaN  \n",
       "6                         NaN  \n",
       "7                         NaN  \n",
       "8  Sulculeolaria quadrivalvis  \n",
       "9                     Delibus  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# replace None with NA\n",
    "asv_tables['18S V9'] = asv_tables['18S V9'].fillna(value=np.nan)\n",
    "## Replace 'unknown', 'unassigned', etc. in species and taxonomy columns with NaN\n",
    "\n",
    "asv_tables['18S V9'][taxa_ranks_18S] = asv_tables['18S V9'][taxa_ranks_18S].replace({'unassigned':np.nan,\n",
    "                            'Unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan,\n",
    "                               '':np.nan})\n",
    "asv_tables['18S V9'].iloc[0:10,[0,1,2,3,4,5,6,-9,-8,-7,-6,-5,-4,-3,-2,-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace _,- with space, remove sp. \n",
    "\n",
    "asv_tables['18S V9'][taxa_ranks_18S] = asv_tables['18S V9'][taxa_ranks_18S].replace('_',' ',regex=True)\n",
    "asv_tables['18S V9'][taxa_ranks_18S] = asv_tables['18S V9'][taxa_ranks_18S].replace(' sp\\.','',regex=True)\n",
    "asv_tables['18S V9'][taxa_ranks_18S] = asv_tables['18S V9'][taxa_ranks_18S].replace(' spp\\.','',regex=True)\n",
    "asv_tables['18S V9'][taxa_ranks_18S] = asv_tables['18S V9'][taxa_ranks_18S].replace('-',' ',regex=True)\n",
    "asv_tables['18S V9'][taxa_ranks_18S] = asv_tables['18S V9'][taxa_ranks_18S].replace('\\/',' ',regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24067, 485)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['18S V9'].shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ['18S V9'] = pd.melt(asv_tables['18S V9'],id_vars=['featureid','sequence','taxonomy','Confidence','domain','supergroup','division','subdivision','class','order','family','genus','species'],\n",
    "               var_name='eventID',value_name='organismQuantity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11359624, 15)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S V9'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(146232, 15)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)\n",
    "\n",
    "occ['18S V9'] = occ['18S V9'][occ['18S V9']['organismQuantity'] > 0]\n",
    "print(occ['18S V9'].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>subdivision</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "5  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "7  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                                                           sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "4  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "5    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "7    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                                                                                                       taxonomy  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "4           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "5                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "7                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "   Confidence     domain supergroup      division     subdivision  \\\n",
       "0    0.922099  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "1    0.999947  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "4    0.779948  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "5    0.999931  Eukaryota       TSAR     Alveolata  Dinoflagellata   \n",
       "7    0.986908  Eukaryota       TSAR     Alveolata  Dinoflagellata   \n",
       "\n",
       "         class          order          family          genus  \\\n",
       "0   Arthropoda      Crustacea     Maxillopoda     Neocalanus   \n",
       "1   Arthropoda      Crustacea     Maxillopoda  Clausocalanus   \n",
       "4   Arthropoda      Crustacea     Maxillopoda    Acrocalanus   \n",
       "5  Dinophyceae            NaN             NaN            NaN   \n",
       "7  Dinophyceae  Gymnodiniales  Gymnodiniaceae            NaN   \n",
       "\n",
       "                  species                 eventID  organismQuantity  \n",
       "0    Neocalanus cristatus  GOMECC4_27N_Sta1_DCM_A              1516  \n",
       "1  Clausocalanus furcatus  GOMECC4_27N_Sta1_DCM_A               962  \n",
       "4             Acrocalanus  GOMECC4_27N_Sta1_DCM_A              1164  \n",
       "5                     NaN  GOMECC4_27N_Sta1_DCM_A               287  \n",
       "7                     NaN  GOMECC4_27N_Sta1_DCM_A               250  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S V9'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add occurenceID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create an occurrenceID that will uniquely identify each ASV observed within a water sample\n",
    "\n",
    "occ['18S V9']['occurrenceID'] = occ['18S V9']['featureid']\n",
    "occ['18S V9']['occurrenceID'] = occ['18S V9']['eventID'] + '_occ' + occ['18S V9']['occurrenceID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>subdivision</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "5  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "7  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                                                           sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "4  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "5    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "7    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                                                                                                       taxonomy  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "4           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "5                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "7                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "   Confidence     domain supergroup      division     subdivision  \\\n",
       "0    0.922099  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "1    0.999947  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "4    0.779948  Eukaryota     Obazoa  Opisthokonta         Metazoa   \n",
       "5    0.999931  Eukaryota       TSAR     Alveolata  Dinoflagellata   \n",
       "7    0.986908  Eukaryota       TSAR     Alveolata  Dinoflagellata   \n",
       "\n",
       "         class          order          family          genus  \\\n",
       "0   Arthropoda      Crustacea     Maxillopoda     Neocalanus   \n",
       "1   Arthropoda      Crustacea     Maxillopoda  Clausocalanus   \n",
       "4   Arthropoda      Crustacea     Maxillopoda    Acrocalanus   \n",
       "5  Dinophyceae            NaN             NaN            NaN   \n",
       "7  Dinophyceae  Gymnodiniales  Gymnodiniaceae            NaN   \n",
       "\n",
       "                  species                 eventID  organismQuantity  \\\n",
       "0    Neocalanus cristatus  GOMECC4_27N_Sta1_DCM_A              1516   \n",
       "1  Clausocalanus furcatus  GOMECC4_27N_Sta1_DCM_A               962   \n",
       "4             Acrocalanus  GOMECC4_27N_Sta1_DCM_A              1164   \n",
       "5                     NaN  GOMECC4_27N_Sta1_DCM_A               287   \n",
       "7                     NaN  GOMECC4_27N_Sta1_DCM_A               250   \n",
       "\n",
       "                                                 occurrenceID  \n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb  \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca  \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75  \n",
       "5  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4  \n",
       "7  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S V9'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 16S"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### 1st, format ASV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0003aeafc4bc0522877d4804829e65b7</td>\n",
       "      <td>CACCGGCATCTCGAGTGGTATCCACTTTTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGAACTAGGGAGTGGGAGAGGTAGACGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Crenarchaeota; c__Nitrososphaeria; o__Nitrosopumilales; f__Nitrosopumilaceae; g__Candidatus_Nitrosopelagicus; s__marine_metagenome</td>\n",
       "      <td>0.779476</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0003b46ae196127658c07aeb11b36b1a</td>\n",
       "      <td>TACGGAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGTTTAGCAAGTTGAATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATTCAAAACTACTAAGCTAGAGTACGAGAGAGGAGAGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Thiomicrospirales; f__Thioglobaceae; g__SUP05_cluster</td>\n",
       "      <td>0.810269</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>00065c7f5701f8db77fc2c50a1204c71</td>\n",
       "      <td>TACGGGAGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAAAAATAAGTCTATTGTTAAATCCAGAAGCTTAACTTCTGTCAAGCGATATGAAACTATTCTTCTTGAGAATGGTAGGGGTAGAAGGAATT...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0006da4e1ff162826badd8bdcfaf9dfe</td>\n",
       "      <td>GACGGAGGATGCAAGTGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAGAAATAAGCCTATTGTTAAATCCAGGAGCTTAACTTCTGTCCAGCGATATGAAACTATTTTTCTTGAGGGTGGTAGGGGTAGAAGGAATT...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0006fe6033cca4da30ea5ce9cba446f0</td>\n",
       "      <td>TACGAATGCTGCAAGCGTAGTTCGGAATCACTGGGCATAAAGAGCACGTAGGCGGCCTATTAAGTCAGCTGTGAAATCCCTCGGCTTAACCGAGGAACTGCAGCTGATACTGATAGGCTTGAGTACGGGAGGGGAGAGCGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Planctomycetota; c__Pla3_lineage; o__Pla3_lineage; f__Pla3_lineage; g__Pla3_lineage; s__uncultured_Planctomycetaceae</td>\n",
       "      <td>0.949296</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "5  0003aeafc4bc0522877d4804829e65b7   \n",
       "6  0003b46ae196127658c07aeb11b36b1a   \n",
       "7  00065c7f5701f8db77fc2c50a1204c71   \n",
       "8  0006da4e1ff162826badd8bdcfaf9dfe   \n",
       "9  0006fe6033cca4da30ea5ce9cba446f0   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...   \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...   \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...   \n",
       "5  CACCGGCATCTCGAGTGGTATCCACTTTTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGAACTAGGGAGTGGGAGAGGTAGACGGTACTC...   \n",
       "6  TACGGAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGTTTAGCAAGTTGAATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATTCAAAACTACTAAGCTAGAGTACGAGAGAGGAGAGTAGAATTT...   \n",
       "7  TACGGGAGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAAAAATAAGTCTATTGTTAAATCCAGAAGCTTAACTTCTGTCAAGCGATATGAAACTATTCTTCTTGAGAATGGTAGGGGTAGAAGGAATT...   \n",
       "8  GACGGAGGATGCAAGTGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAGAAATAAGCCTATTGTTAAATCCAGGAGCTTAACTTCTGTCCAGCGATATGAAACTATTTTTCTTGAGGGTGGTAGGGGTAGAAGGAATT...   \n",
       "9  TACGAATGCTGCAAGCGTAGTTCGGAATCACTGGGCATAAAGAGCACGTAGGCGGCCTATTAAGTCAGCTGTGAAATCCCTCGGCTTAACCGAGGAACTGCAGCTGATACTGATAGGCTTGAGTACGGGAGGGGAGAGCGGAATTC...   \n",
       "\n",
       "                                                                                                                                            taxonomy  \\\n",
       "0                                                 d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1                                     d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2                                       d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3                                 d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4                                             d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "5  d__Archaea; p__Crenarchaeota; c__Nitrososphaeria; o__Nitrosopumilales; f__Nitrosopumilaceae; g__Candidatus_Nitrosopelagicus; s__marine_metagenome   \n",
       "6                                   d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Thiomicrospirales; f__Thioglobaceae; g__SUP05_cluster   \n",
       "7                        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale   \n",
       "8                        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale   \n",
       "9               d__Bacteria; p__Planctomycetota; c__Pla3_lineage; o__Pla3_lineage; f__Pla3_lineage; g__Pla3_lineage; s__uncultured_Planctomycetaceae   \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0    0.978926                       0                       0   \n",
       "1    0.881698                       0                       0   \n",
       "2    0.762793                       0                       0   \n",
       "3    0.997619                       0                       0   \n",
       "4    0.999961                       0                       0   \n",
       "5    0.779476                       0                       0   \n",
       "6    0.810269                       0                       0   \n",
       "7    1.000000                       0                       0   \n",
       "8    1.000000                       0                       0   \n",
       "9    0.949296                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C  GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  \\\n",
       "0                       0                        0                        0   \n",
       "1                       0                        0                        0   \n",
       "2                       0                        0                        0   \n",
       "3                       0                        0                        0   \n",
       "4                       0                        0                        0   \n",
       "5                       0                        0                        0   \n",
       "6                       0                        0                        0   \n",
       "7                       0                        0                        0   \n",
       "8                       0                        0                        0   \n",
       "9                       0                        0                        0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_C  GOMECC4_27N_Sta1_Surface_A  \\\n",
       "0                        0                           0   \n",
       "1                        0                           0   \n",
       "2                        0                           0   \n",
       "3                        0                           0   \n",
       "4                        0                           0   \n",
       "5                        0                           0   \n",
       "6                        0                           0   \n",
       "7                        0                           0   \n",
       "8                        0                           0   \n",
       "9                        0                           0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Surface_B  GOMECC4_27N_Sta4_DCM_A  GOMECC4_27N_Sta4_DCM_B  \\\n",
       "0                           0                       0                       0   \n",
       "1                           0                       0                       0   \n",
       "2                           0                       0                       0   \n",
       "3                           0                       0                       0   \n",
       "4                           0                       0                       0   \n",
       "5                           0                       0                       0   \n",
       "6                           0                       0                       0   \n",
       "7                           0                       0                       0   \n",
       "8                           0                       0                       0   \n",
       "9                           0                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_DCM_C  GOMECC4_27N_Sta4_Deep_A  GOMECC4_27N_Sta4_Deep_B  \\\n",
       "0                       0                        0                        0   \n",
       "1                       0                        0                        0   \n",
       "2                       0                        0                        0   \n",
       "3                       0                        0                        0   \n",
       "4                       0                        0                        0   \n",
       "5                       0                        0                        0   \n",
       "6                       0                        0                        0   \n",
       "7                       0                        0                        0   \n",
       "8                       0                        0                        0   \n",
       "9                       0                        0                        0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_Deep_C  GOMECC4_27N_Sta4_Surface_A  \\\n",
       "0                        0                           0   \n",
       "1                        0                           0   \n",
       "2                        0                           0   \n",
       "3                        0                           0   \n",
       "4                        0                           0   \n",
       "5                        0                           0   \n",
       "6                        0                           0   \n",
       "7                        0                           0   \n",
       "8                        0                           0   \n",
       "9                        0                           0   \n",
       "\n",
       "   GOMECC4_27N_Sta4_Surface_B  \n",
       "0                          25  \n",
       "1                           0  \n",
       "2                           0  \n",
       "3                           0  \n",
       "4                           0  \n",
       "5                           0  \n",
       "6                           0  \n",
       "7                           0  \n",
       "8                           0  \n",
       "9                           0  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S V4-V5'].iloc[0:10,0:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio'"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S V4-V5']['taxonomy'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxa_ranks_16S = ['domain','phylum','class','order','family','genus','species']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_C</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Vibrionales</td>\n",
       "      <td>f__Vibrionaceae</td>\n",
       "      <td>g__Vibrio</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__HOC36</td>\n",
       "      <td>f__HOC36</td>\n",
       "      <td>g__HOC36</td>\n",
       "      <td>s__Candidatus_Thioglobus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Cyanobacteria</td>\n",
       "      <td>c__Cyanobacteriia</td>\n",
       "      <td>o__Synechococcales</td>\n",
       "      <td>f__Cyanobiaceae</td>\n",
       "      <td>g__Cyanobium_PCC-6307</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__Myxococcia</td>\n",
       "      <td>o__Myxococcales</td>\n",
       "      <td>f__Myxococcaceae</td>\n",
       "      <td>g__P3OB-42</td>\n",
       "      <td>s__uncultured_bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__SAR86_clade</td>\n",
       "      <td>f__SAR86_clade</td>\n",
       "      <td>g__SAR86_clade</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0003aeafc4bc0522877d4804829e65b7</td>\n",
       "      <td>CACCGGCATCTCGAGTGGTATCCACTTTTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGAACTAGGGAGTGGGAGAGGTAGACGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Crenarchaeota; c__Nitrososphaeria; o__Nitrosopumilales; f__Nitrosopumilaceae; g__Candidatus_Nitrosopelagicus; s__marine_metagenome</td>\n",
       "      <td>0.779476</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__Crenarchaeota</td>\n",
       "      <td>c__Nitrososphaeria</td>\n",
       "      <td>o__Nitrosopumilales</td>\n",
       "      <td>f__Nitrosopumilaceae</td>\n",
       "      <td>g__Candidatus_Nitrosopelagicus</td>\n",
       "      <td>s__marine_metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0003b46ae196127658c07aeb11b36b1a</td>\n",
       "      <td>TACGGAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGTTTAGCAAGTTGAATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATTCAAAACTACTAAGCTAGAGTACGAGAGAGGAGAGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Thiomicrospirales; f__Thioglobaceae; g__SUP05_cluster</td>\n",
       "      <td>0.810269</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Thiomicrospirales</td>\n",
       "      <td>f__Thioglobaceae</td>\n",
       "      <td>g__SUP05_cluster</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>00065c7f5701f8db77fc2c50a1204c71</td>\n",
       "      <td>TACGGGAGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAAAAATAAGTCTATTGTTAAATCCAGAAGCTTAACTTCTGTCAAGCGATATGAAACTATTCTTCTTGAGAATGGTAGGGGTAGAAGGAATT...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Cyanobacteria</td>\n",
       "      <td>c__Cyanobacteriia</td>\n",
       "      <td>o__Chloroplast</td>\n",
       "      <td>f__Chloroplast</td>\n",
       "      <td>g__Chloroplast</td>\n",
       "      <td>s__Prasinoderma_coloniale</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0006da4e1ff162826badd8bdcfaf9dfe</td>\n",
       "      <td>GACGGAGGATGCAAGTGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAGAAATAAGCCTATTGTTAAATCCAGGAGCTTAACTTCTGTCCAGCGATATGAAACTATTTTTCTTGAGGGTGGTAGGGGTAGAAGGAATT...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Cyanobacteria</td>\n",
       "      <td>c__Cyanobacteriia</td>\n",
       "      <td>o__Chloroplast</td>\n",
       "      <td>f__Chloroplast</td>\n",
       "      <td>g__Chloroplast</td>\n",
       "      <td>s__Prasinoderma_coloniale</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0006fe6033cca4da30ea5ce9cba446f0</td>\n",
       "      <td>TACGAATGCTGCAAGCGTAGTTCGGAATCACTGGGCATAAAGAGCACGTAGGCGGCCTATTAAGTCAGCTGTGAAATCCCTCGGCTTAACCGAGGAACTGCAGCTGATACTGATAGGCTTGAGTACGGGAGGGGAGAGCGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Planctomycetota; c__Pla3_lineage; o__Pla3_lineage; f__Pla3_lineage; g__Pla3_lineage; s__uncultured_Planctomycetaceae</td>\n",
       "      <td>0.949296</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Planctomycetota</td>\n",
       "      <td>c__Pla3_lineage</td>\n",
       "      <td>o__Pla3_lineage</td>\n",
       "      <td>f__Pla3_lineage</td>\n",
       "      <td>g__Pla3_lineage</td>\n",
       "      <td>s__uncultured_Planctomycetaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "5  0003aeafc4bc0522877d4804829e65b7   \n",
       "6  0003b46ae196127658c07aeb11b36b1a   \n",
       "7  00065c7f5701f8db77fc2c50a1204c71   \n",
       "8  0006da4e1ff162826badd8bdcfaf9dfe   \n",
       "9  0006fe6033cca4da30ea5ce9cba446f0   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...   \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...   \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...   \n",
       "5  CACCGGCATCTCGAGTGGTATCCACTTTTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGAACTAGGGAGTGGGAGAGGTAGACGGTACTC...   \n",
       "6  TACGGAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGTTTAGCAAGTTGAATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATTCAAAACTACTAAGCTAGAGTACGAGAGAGGAGAGTAGAATTT...   \n",
       "7  TACGGGAGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAAAAATAAGTCTATTGTTAAATCCAGAAGCTTAACTTCTGTCAAGCGATATGAAACTATTCTTCTTGAGAATGGTAGGGGTAGAAGGAATT...   \n",
       "8  GACGGAGGATGCAAGTGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAGAAATAAGCCTATTGTTAAATCCAGGAGCTTAACTTCTGTCCAGCGATATGAAACTATTTTTCTTGAGGGTGGTAGGGGTAGAAGGAATT...   \n",
       "9  TACGAATGCTGCAAGCGTAGTTCGGAATCACTGGGCATAAAGAGCACGTAGGCGGCCTATTAAGTCAGCTGTGAAATCCCTCGGCTTAACCGAGGAACTGCAGCTGATACTGATAGGCTTGAGTACGGGAGGGGAGAGCGGAATTC...   \n",
       "\n",
       "                                                                                                                                            taxonomy  \\\n",
       "0                                                 d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1                                     d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2                                       d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3                                 d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4                                             d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "5  d__Archaea; p__Crenarchaeota; c__Nitrososphaeria; o__Nitrosopumilales; f__Nitrosopumilaceae; g__Candidatus_Nitrosopelagicus; s__marine_metagenome   \n",
       "6                                   d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Thiomicrospirales; f__Thioglobaceae; g__SUP05_cluster   \n",
       "7                        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale   \n",
       "8                        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale   \n",
       "9               d__Bacteria; p__Planctomycetota; c__Pla3_lineage; o__Pla3_lineage; f__Pla3_lineage; g__Pla3_lineage; s__uncultured_Planctomycetaceae   \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0    0.978926                       0                       0   \n",
       "1    0.881698                       0                       0   \n",
       "2    0.762793                       0                       0   \n",
       "3    0.997619                       0                       0   \n",
       "4    0.999961                       0                       0   \n",
       "5    0.779476                       0                       0   \n",
       "6    0.810269                       0                       0   \n",
       "7    1.000000                       0                       0   \n",
       "8    1.000000                       0                       0   \n",
       "9    0.949296                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C  GOMECC4_YUCATAN_Sta102_Surface_C       domain  \\\n",
       "0                       0                                 7  d__Bacteria   \n",
       "1                       0                                 0  d__Bacteria   \n",
       "2                       0                                 0  d__Bacteria   \n",
       "3                       0                                 0  d__Bacteria   \n",
       "4                       0                                 0  d__Bacteria   \n",
       "5                       0                                 0   d__Archaea   \n",
       "6                       0                                 0  d__Bacteria   \n",
       "7                       0                                 0  d__Bacteria   \n",
       "8                       0                                 0  d__Bacteria   \n",
       "9                       0                                 0  d__Bacteria   \n",
       "\n",
       "               phylum                   class                 order  \\\n",
       "0   p__Proteobacteria  c__Gammaproteobacteria        o__Vibrionales   \n",
       "1   p__Proteobacteria  c__Gammaproteobacteria              o__HOC36   \n",
       "2    p__Cyanobacteria       c__Cyanobacteriia    o__Synechococcales   \n",
       "3      p__Myxococcota           c__Myxococcia       o__Myxococcales   \n",
       "4   p__Proteobacteria  c__Gammaproteobacteria        o__SAR86_clade   \n",
       "5    p__Crenarchaeota      c__Nitrososphaeria   o__Nitrosopumilales   \n",
       "6   p__Proteobacteria  c__Gammaproteobacteria  o__Thiomicrospirales   \n",
       "7    p__Cyanobacteria       c__Cyanobacteriia        o__Chloroplast   \n",
       "8    p__Cyanobacteria       c__Cyanobacteriia        o__Chloroplast   \n",
       "9  p__Planctomycetota         c__Pla3_lineage       o__Pla3_lineage   \n",
       "\n",
       "                 family                           genus  \\\n",
       "0       f__Vibrionaceae                       g__Vibrio   \n",
       "1              f__HOC36                        g__HOC36   \n",
       "2       f__Cyanobiaceae           g__Cyanobium_PCC-6307   \n",
       "3      f__Myxococcaceae                      g__P3OB-42   \n",
       "4        f__SAR86_clade                  g__SAR86_clade   \n",
       "5  f__Nitrosopumilaceae  g__Candidatus_Nitrosopelagicus   \n",
       "6      f__Thioglobaceae                g__SUP05_cluster   \n",
       "7        f__Chloroplast                  g__Chloroplast   \n",
       "8        f__Chloroplast                  g__Chloroplast   \n",
       "9       f__Pla3_lineage                 g__Pla3_lineage   \n",
       "\n",
       "                           species  \n",
       "0                             None  \n",
       "1         s__Candidatus_Thioglobus  \n",
       "2                             None  \n",
       "3          s__uncultured_bacterium  \n",
       "4                             None  \n",
       "5             s__marine_metagenome  \n",
       "6                             None  \n",
       "7        s__Prasinoderma_coloniale  \n",
       "8        s__Prasinoderma_coloniale  \n",
       "9  s__uncultured_Planctomycetaceae  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S V4-V5'][['domain','phylum','class','order','family','genus','species']] = asv_tables['16S V4-V5']['taxonomy'].str.split(\"; \",expand=True)\n",
    "asv_tables['16S V4-V5'].iloc[0:10,[0,1,2,3,4,5,6,-8,-7,-6,-5,-4,-3,-2,-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_tables['16S V4-V5']['domain'] = asv_tables['16S V4-V5']['domain'].str.replace(\"d__\", \"\")\n",
    "asv_tables['16S V4-V5']['phylum'] = asv_tables['16S V4-V5']['phylum'].str.replace(\"p__\", \"\")\n",
    "asv_tables['16S V4-V5']['class'] = asv_tables['16S V4-V5']['class'].str.replace(\"c__\", \"\")\n",
    "asv_tables['16S V4-V5']['order'] = asv_tables['16S V4-V5']['order'].str.replace(\"o__\", \"\")\n",
    "asv_tables['16S V4-V5']['family'] = asv_tables['16S V4-V5']['family'].str.replace(\"f__\", \"\")\n",
    "asv_tables['16S V4-V5']['genus'] = asv_tables['16S V4-V5']['genus'].str.replace(\"g__\", \"\")\n",
    "asv_tables['16S V4-V5']['species'] = asv_tables['16S V4-V5']['species'].str.replace(\"s__\", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_C</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>HOC36</td>\n",
       "      <td>HOC36</td>\n",
       "      <td>HOC36</td>\n",
       "      <td>Candidatus_Thioglobus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Cyanobium_PCC-6307</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>Myxococcia</td>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>P3OB-42</td>\n",
       "      <td>uncultured_bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>SAR86_clade</td>\n",
       "      <td>SAR86_clade</td>\n",
       "      <td>SAR86_clade</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0003aeafc4bc0522877d4804829e65b7</td>\n",
       "      <td>CACCGGCATCTCGAGTGGTATCCACTTTTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGAACTAGGGAGTGGGAGAGGTAGACGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Crenarchaeota; c__Nitrososphaeria; o__Nitrosopumilales; f__Nitrosopumilaceae; g__Candidatus_Nitrosopelagicus; s__marine_metagenome</td>\n",
       "      <td>0.779476</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Nitrososphaeria</td>\n",
       "      <td>Nitrosopumilales</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>Candidatus_Nitrosopelagicus</td>\n",
       "      <td>marine_metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0003b46ae196127658c07aeb11b36b1a</td>\n",
       "      <td>TACGGAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGTTTAGCAAGTTGAATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATTCAAAACTACTAAGCTAGAGTACGAGAGAGGAGAGTAGAATTT...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Thiomicrospirales; f__Thioglobaceae; g__SUP05_cluster</td>\n",
       "      <td>0.810269</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiomicrospirales</td>\n",
       "      <td>Thioglobaceae</td>\n",
       "      <td>SUP05_cluster</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>00065c7f5701f8db77fc2c50a1204c71</td>\n",
       "      <td>TACGGGAGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAAAAATAAGTCTATTGTTAAATCCAGAAGCTTAACTTCTGTCAAGCGATATGAAACTATTCTTCTTGAGAATGGTAGGGGTAGAAGGAATT...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Prasinoderma_coloniale</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0006da4e1ff162826badd8bdcfaf9dfe</td>\n",
       "      <td>GACGGAGGATGCAAGTGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAGAAATAAGCCTATTGTTAAATCCAGGAGCTTAACTTCTGTCCAGCGATATGAAACTATTTTTCTTGAGGGTGGTAGGGGTAGAAGGAATT...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Prasinoderma_coloniale</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0006fe6033cca4da30ea5ce9cba446f0</td>\n",
       "      <td>TACGAATGCTGCAAGCGTAGTTCGGAATCACTGGGCATAAAGAGCACGTAGGCGGCCTATTAAGTCAGCTGTGAAATCCCTCGGCTTAACCGAGGAACTGCAGCTGATACTGATAGGCTTGAGTACGGGAGGGGAGAGCGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Planctomycetota; c__Pla3_lineage; o__Pla3_lineage; f__Pla3_lineage; g__Pla3_lineage; s__uncultured_Planctomycetaceae</td>\n",
       "      <td>0.949296</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Planctomycetota</td>\n",
       "      <td>Pla3_lineage</td>\n",
       "      <td>Pla3_lineage</td>\n",
       "      <td>Pla3_lineage</td>\n",
       "      <td>Pla3_lineage</td>\n",
       "      <td>uncultured_Planctomycetaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "5  0003aeafc4bc0522877d4804829e65b7   \n",
       "6  0003b46ae196127658c07aeb11b36b1a   \n",
       "7  00065c7f5701f8db77fc2c50a1204c71   \n",
       "8  0006da4e1ff162826badd8bdcfaf9dfe   \n",
       "9  0006fe6033cca4da30ea5ce9cba446f0   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTT...   \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTT...   \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTC...   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTC...   \n",
       "5  CACCGGCATCTCGAGTGGTATCCACTTTTATTGGGCCTAAAGCATCCGTAGCCTGTTCTGTAAGTTTTCGGTTAAATCCATAAGCTCAACTTATGGGCTGCCGAAAATACTGCAGAACTAGGGAGTGGGAGAGGTAGACGGTACTC...   \n",
       "6  TACGGAGGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGTTTAGCAAGTTGAATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATTCAAAACTACTAAGCTAGAGTACGAGAGAGGAGAGTAGAATTT...   \n",
       "7  TACGGGAGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAAAAATAAGTCTATTGTTAAATCCAGAAGCTTAACTTCTGTCAAGCGATATGAAACTATTCTTCTTGAGAATGGTAGGGGTAGAAGGAATT...   \n",
       "8  GACGGAGGATGCAAGTGTTATCCGGAATTATTGGGCGTAAAGCGTTTGTAGGTGGAGAAATAAGCCTATTGTTAAATCCAGGAGCTTAACTTCTGTCCAGCGATATGAAACTATTTTTCTTGAGGGTGGTAGGGGTAGAAGGAATT...   \n",
       "9  TACGAATGCTGCAAGCGTAGTTCGGAATCACTGGGCATAAAGAGCACGTAGGCGGCCTATTAAGTCAGCTGTGAAATCCCTCGGCTTAACCGAGGAACTGCAGCTGATACTGATAGGCTTGAGTACGGGAGGGGAGAGCGGAATTC...   \n",
       "\n",
       "                                                                                                                                            taxonomy  \\\n",
       "0                                                 d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1                                     d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2                                       d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3                                 d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4                                             d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "5  d__Archaea; p__Crenarchaeota; c__Nitrososphaeria; o__Nitrosopumilales; f__Nitrosopumilaceae; g__Candidatus_Nitrosopelagicus; s__marine_metagenome   \n",
       "6                                   d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Thiomicrospirales; f__Thioglobaceae; g__SUP05_cluster   \n",
       "7                        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale   \n",
       "8                        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast; s__Prasinoderma_coloniale   \n",
       "9               d__Bacteria; p__Planctomycetota; c__Pla3_lineage; o__Pla3_lineage; f__Pla3_lineage; g__Pla3_lineage; s__uncultured_Planctomycetaceae   \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0    0.978926                       0                       0   \n",
       "1    0.881698                       0                       0   \n",
       "2    0.762793                       0                       0   \n",
       "3    0.997619                       0                       0   \n",
       "4    0.999961                       0                       0   \n",
       "5    0.779476                       0                       0   \n",
       "6    0.810269                       0                       0   \n",
       "7    1.000000                       0                       0   \n",
       "8    1.000000                       0                       0   \n",
       "9    0.949296                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C  GOMECC4_YUCATAN_Sta102_Surface_C    domain  \\\n",
       "0                       0                                 7  Bacteria   \n",
       "1                       0                                 0  Bacteria   \n",
       "2                       0                                 0  Bacteria   \n",
       "3                       0                                 0  Bacteria   \n",
       "4                       0                                 0  Bacteria   \n",
       "5                       0                                 0   Archaea   \n",
       "6                       0                                 0  Bacteria   \n",
       "7                       0                                 0  Bacteria   \n",
       "8                       0                                 0  Bacteria   \n",
       "9                       0                                 0  Bacteria   \n",
       "\n",
       "            phylum                class              order             family  \\\n",
       "0   Proteobacteria  Gammaproteobacteria        Vibrionales       Vibrionaceae   \n",
       "1   Proteobacteria  Gammaproteobacteria              HOC36              HOC36   \n",
       "2    Cyanobacteria       Cyanobacteriia    Synechococcales       Cyanobiaceae   \n",
       "3      Myxococcota           Myxococcia       Myxococcales      Myxococcaceae   \n",
       "4   Proteobacteria  Gammaproteobacteria        SAR86_clade        SAR86_clade   \n",
       "5    Crenarchaeota      Nitrososphaeria   Nitrosopumilales  Nitrosopumilaceae   \n",
       "6   Proteobacteria  Gammaproteobacteria  Thiomicrospirales      Thioglobaceae   \n",
       "7    Cyanobacteria       Cyanobacteriia        Chloroplast        Chloroplast   \n",
       "8    Cyanobacteria       Cyanobacteriia        Chloroplast        Chloroplast   \n",
       "9  Planctomycetota         Pla3_lineage       Pla3_lineage       Pla3_lineage   \n",
       "\n",
       "                         genus                       species  \n",
       "0                       Vibrio                           NaN  \n",
       "1                        HOC36         Candidatus_Thioglobus  \n",
       "2           Cyanobium_PCC-6307                           NaN  \n",
       "3                      P3OB-42          uncultured_bacterium  \n",
       "4                  SAR86_clade                           NaN  \n",
       "5  Candidatus_Nitrosopelagicus             marine_metagenome  \n",
       "6                SUP05_cluster                           NaN  \n",
       "7                  Chloroplast        Prasinoderma_coloniale  \n",
       "8                  Chloroplast        Prasinoderma_coloniale  \n",
       "9                 Pla3_lineage  uncultured_Planctomycetaceae  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# replace None with NA\n",
    "asv_tables['16S V4-V5'] = asv_tables['16S V4-V5'].fillna(value=np.nan)\n",
    "## Replace 'unknown', 'unassigned', etc. in species and taxonomy columns with NaN\n",
    "\n",
    "asv_tables['16S V4-V5'][taxa_ranks_16S] = asv_tables['16S V4-V5'][taxa_ranks_16S].replace({'unassigned':np.nan,'Unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan,\n",
    "                               '':np.nan})\n",
    "asv_tables['16S V4-V5'].iloc[0:10,[0,1,2,3,4,5,6,-8,-7,-6,-5,-4,-3,-2,-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace _,- with space, remove sp. \n",
    "\n",
    "asv_tables['16S V4-V5'][taxa_ranks_16S] = asv_tables['16S V4-V5'][taxa_ranks_16S].replace('_',' ',regex=True)\n",
    "asv_tables['16S V4-V5'][taxa_ranks_16S] = asv_tables['16S V4-V5'][taxa_ranks_16S].replace(' sp\\.','',regex=True)\n",
    "asv_tables['16S V4-V5'][taxa_ranks_16S] = asv_tables['16S V4-V5'][taxa_ranks_16S].replace('-',' ',regex=True)\n",
    "asv_tables['16S V4-V5'][taxa_ranks_16S] = asv_tables['16S V4-V5'][taxa_ranks_16S].replace(' spp\\.','',regex=True)\n",
    "asv_tables['16S V4-V5'][taxa_ranks_16S] = asv_tables['16S V4-V5'][taxa_ranks_16S].replace('\\/',' ',regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Melt asv_tables to long format\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65048, 483)"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S V4-V5'].shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ['16S V4-V5'] = pd.melt(asv_tables['16S V4-V5'],id_vars=['featureid','sequence','taxonomy','Confidence','domain','phylum','class','order','family','genus','species'],\n",
    "               var_name='eventID',value_name='organismQuantity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30702656, 13)"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S V4-V5'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(165158, 13)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)\n",
    "\n",
    "occ['16S V4-V5'] = occ['16S V4-V5'][occ['16S V4-V5']['organismQuantity'] > 0]\n",
    "print(occ['16S V4-V5'].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create an occurrenceID that will uniquely identify each ASV observed within a water sample\n",
    "\n",
    "occ['16S V4-V5']['occurrenceID'] = occ['16S V4-V5']['featureid']\n",
    "occ['16S V4-V5']['occurrenceID'] = occ['16S V4-V5']['eventID'] + '_16S_occ' + occ['16S V4-V5']['occurrenceID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Parvibaculales</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>uncultured marine</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Puniceispirillales</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Thermoplasmatota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            featureid  \\\n",
       "182  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "225  00e6c13fe86364a5084987093afa1916   \n",
       "347  015dad1fafca90944d905beb2a980bc3   \n",
       "412  019c88c6ade406f731954f38e3461564   \n",
       "719  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                  sequence  \\\n",
       "182  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "225  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "347  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "412  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "719  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                  taxonomy  \\\n",
       "182                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "225                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "347                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "412                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "719  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "     Confidence    domain                         phylum  \\\n",
       "182    0.832190  Bacteria                 Proteobacteria   \n",
       "225    0.867040  Bacteria                 Proteobacteria   \n",
       "347    1.000000   Archaea               Thermoplasmatota   \n",
       "412    0.952911  Bacteria                 Proteobacteria   \n",
       "719    0.818195  Bacteria  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                             class                          order  \\\n",
       "182            Alphaproteobacteria                 Parvibaculales   \n",
       "225            Alphaproteobacteria             Puniceispirillales   \n",
       "347                 Thermoplasmata                Marine Group II   \n",
       "412            Alphaproteobacteria                  Rickettsiales   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                            family                          genus  \\\n",
       "182                   OCS116 clade                   OCS116 clade   \n",
       "225                   SAR116 clade                   SAR116 clade   \n",
       "347                Marine Group II                Marine Group II   \n",
       "412                   Mitochondria                   Mitochondria   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                  species                 eventID  organismQuantity  \\\n",
       "182     uncultured marine  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "225                   NaN  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "347                   NaN  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "412  uncultured bacterium  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "719  uncultured bacterium  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                                       occurrenceID  \n",
       "182  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d  \n",
       "225  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916  \n",
       "347  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3  \n",
       "412  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564  \n",
       "719  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S V4-V5'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### WORMS conversion. \n",
    "Note, can't use `multiprocessing` library in a Jupyter notebook, need `multiprocess`. See [here](https://stackoverflow.com/questions/41385708/multiprocessing-example-giving-attributeerror)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OBIS currently requires taxonomy assignments that match WoRMS, however none of the commonly used metabarcoding reference databases use WoRMS as the basis of their taxonomy. This means the taxonomic ranks for any given scientific name on WoRMS may not directly compare to what is assigned. There are ongoing discussions about this problem (see [this](https://github.com/iobis/Project-team-Genetic-Data/issues/5) GitHub issue).     \n",
    "\n",
    "Many of them, especially for microbes, include taxa that aren't on WoRMS at all. This is because the name may not have been fully and officially adopted by the scientific community (or at least not adopted by WoRMS yet). We therefore need a system for searching through the higher taxonomic ranks given, finding the lowest one that will match on WoRMS, and putting that name in the `scientificName` column. The assigned taxonomy is then recorded elsewhere."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Had some [issues with the parallelization](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr) on Mac M1. Adding 'OBJC_DISABLE_INITIALIZE_FORK_SAFETY = YES' to .bash_profile and then [This](https://github.com/python/cpython/issues/74570) fixed it.   \n",
    "Try to run without the bash_profile fix LATER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"no_proxy\"]=\"*\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 16S worms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Species level IDs might be trash, [see here](https://forum.qiime2.org/t/processing-filtering-and-evaluating-the-silva-database-and-other-reference-sequence-data-with-rescript/15494), so look at genus and up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "import WoRMS_matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'WoRMS_matching' from '/Users/katherine.silliman/Projects/NOAA/DwC/edna2obis/src/WoRMS_matching.py'>"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import importlib\n",
    "importlib.reload(WoRMS_matching)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_16S = asv_tables['16S V4-V5'][['taxonomy','domain','phylum','class','order','family','genus','species']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ignore_index is important!\n",
    "tax_16S = tax_16S.drop_duplicates(ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2729, 8)"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_16S.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    worms_16s = WoRMS_matching.get_worms_from_scientific_name_parallel(\n",
    "    tax_df = tax_16S,ordered_rank_columns=['genus','family','order','class','phylum','domain'],\n",
    "    full_tax_column=\"taxonomy\",full_tax_vI=True,n_proc=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>genus</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:480248</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>class</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>order</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:345514</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanophyceae</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>family</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570956</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>class</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                             full_tax  \\\n",
       "0                  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1      d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4              d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "\n",
       "                                                                                               verbatimIdentification  \\\n",
       "0                  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1      d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4              d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "\n",
       "  old_taxonRank             old name       scientificName  \\\n",
       "0         genus               Vibrio               Vibrio   \n",
       "1         class  Gammaproteobacteria  Gammaproteobacteria   \n",
       "2         order      Synechococcales      Synechococcales   \n",
       "3        family        Myxococcaceae        Myxococcaceae   \n",
       "4         class  Gammaproteobacteria  Gammaproteobacteria   \n",
       "\n",
       "                            scientificNameID   kingdom          phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:480248  Bacteria  Proteobacteria   \n",
       "1  urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "2  urn:lsid:marinespecies.org:taxname:345514  Bacteria   Cyanobacteria   \n",
       "3  urn:lsid:marinespecies.org:taxname:570956  Bacteria  Proteobacteria   \n",
       "4  urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "\n",
       "                 class            order         family   genus taxonRank  \n",
       "0  Gammaproteobacteria      Vibrionales   Vibrionaceae  Vibrio     Genus  \n",
       "1  Gammaproteobacteria             None           None    None     Class  \n",
       "2         Cyanophyceae  Synechococcales           None    None     Order  \n",
       "3  Deltaproteobacteria     Myxococcales  Myxococcaceae    None    Family  \n",
       "4  Gammaproteobacteria             None           None    None     Class  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>d__Eukaryota</td>\n",
       "      <td>d__Eukaryota</td>\n",
       "      <td>domain</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         full_tax verbatimIdentification old_taxonRank   old name  \\\n",
       "242  d__Eukaryota           d__Eukaryota        domain  Eukaryota   \n",
       "\n",
       "    scientificName scientificNameID kingdom phylum class order family genus  \\\n",
       "242       No match             None     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "\n",
       "    taxonRank  \n",
       "242       NaN  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s[worms_16s[\"scientificName\"]==\"No match\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_16s.loc[worms_16s[\"scientificName\"]==\"No match\",'scientificName'] = \"Biota\"\n",
    "worms_16s.loc[worms_16s[\"scientificName\"]==\"Biota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      full_tax verbatimIdentification old_taxonRank old name scientificName  \\\n",
       "97  Unassigned             Unassigned           NaN      NaN            NaN   \n",
       "\n",
       "   scientificNameID kingdom phylum class order family genus taxonRank  \n",
       "97              NaN     NaN    NaN   NaN   NaN    NaN   NaN       NaN  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s[worms_16s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 13)\n",
      "(0, 13)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(worms_16s[worms_16s['scientificName'].isna() == True].shape)\n",
    "worms_16s.loc[worms_16s['scientificName'].isna() == True,'scientificName'] = 'incertae sedis'\n",
    "worms_16s.loc[worms_16s['scientificName'] == 'incertae sedis','scientificNameID'] =  'urn:lsid:marinespecies.org:taxname:12'\n",
    "print(worms_16s[worms_16s['scientificName'].isna() == True].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_16s.to_csv(\"../processed/worms_16S_matching.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:480248</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:345514</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanophyceae</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570956</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                             full_tax  \\\n",
       "0                  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1      d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4              d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "\n",
       "                                                                                               verbatimIdentification  \\\n",
       "0                  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio   \n",
       "1      d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus   \n",
       "2        d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4              d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0               Vibrio  urn:lsid:marinespecies.org:taxname:480248  Bacteria   \n",
       "1  Gammaproteobacteria  urn:lsid:marinespecies.org:taxname:393018  Bacteria   \n",
       "2      Synechococcales  urn:lsid:marinespecies.org:taxname:345514  Bacteria   \n",
       "3        Myxococcaceae  urn:lsid:marinespecies.org:taxname:570956  Bacteria   \n",
       "4  Gammaproteobacteria  urn:lsid:marinespecies.org:taxname:393018  Bacteria   \n",
       "\n",
       "           phylum                class            order         family  \\\n",
       "0  Proteobacteria  Gammaproteobacteria      Vibrionales   Vibrionaceae   \n",
       "1  Proteobacteria  Gammaproteobacteria             None           None   \n",
       "2   Cyanobacteria         Cyanophyceae  Synechococcales           None   \n",
       "3  Proteobacteria  Deltaproteobacteria     Myxococcales  Myxococcaceae   \n",
       "4  Proteobacteria  Gammaproteobacteria             None           None   \n",
       "\n",
       "    genus taxonRank  \n",
       "0  Vibrio     Genus  \n",
       "1    None     Class  \n",
       "2    None     Order  \n",
       "3    None    Family  \n",
       "4    None     Class  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s.drop(columns=['old name','old_taxonRank'],inplace=True)\n",
    "worms_16s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Parvibaculales</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>uncultured marine</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Puniceispirillales</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Thermoplasmatota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            featureid  \\\n",
       "182  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "225  00e6c13fe86364a5084987093afa1916   \n",
       "347  015dad1fafca90944d905beb2a980bc3   \n",
       "412  019c88c6ade406f731954f38e3461564   \n",
       "719  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                  sequence  \\\n",
       "182  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "225  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "347  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "412  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "719  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                  taxonomy  \\\n",
       "182                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "225                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "347                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "412                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "719  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "     Confidence    domain                         phylum  \\\n",
       "182    0.832190  Bacteria                 Proteobacteria   \n",
       "225    0.867040  Bacteria                 Proteobacteria   \n",
       "347    1.000000   Archaea               Thermoplasmatota   \n",
       "412    0.952911  Bacteria                 Proteobacteria   \n",
       "719    0.818195  Bacteria  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                             class                          order  \\\n",
       "182            Alphaproteobacteria                 Parvibaculales   \n",
       "225            Alphaproteobacteria             Puniceispirillales   \n",
       "347                 Thermoplasmata                Marine Group II   \n",
       "412            Alphaproteobacteria                  Rickettsiales   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                            family                          genus  \\\n",
       "182                   OCS116 clade                   OCS116 clade   \n",
       "225                   SAR116 clade                   SAR116 clade   \n",
       "347                Marine Group II                Marine Group II   \n",
       "412                   Mitochondria                   Mitochondria   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                  species                 eventID  organismQuantity  \\\n",
       "182     uncultured marine  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "225                   NaN  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "347                   NaN  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "412  uncultured bacterium  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "719  uncultured bacterium  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                                       occurrenceID  \n",
       "182  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d  \n",
       "225  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916  \n",
       "347  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3  \n",
       "412  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564  \n",
       "719  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S V4-V5'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Occurrence and worms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(165158, 14)"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S V4-V5'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                taxonomy  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "   Confidence                 eventID  organismQuantity  \\\n",
       "0    0.832190  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "1    0.867040  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "2    1.000000  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "3    0.952911  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "4    0.818195  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class  \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class  \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class  \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order  \n",
       "4            None                 None           None   None  None   Kingdom  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "occ16_test = occ['16S V4-V5'].copy()\n",
    "occ16_test.drop(columns=['domain','phylum','class','order','family','genus','species'],inplace=True)\n",
    "\n",
    "occ16_test = occ16_test.merge(worms_16s, how='left', left_on ='taxonomy', right_on='full_tax')\n",
    "occ16_test.drop(columns='full_tax', inplace=True)\n",
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>controls_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>411</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S V9</td>\n",
       "      <td>260</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  amplicon_sequenced  ampliconSize trim_method  \\\n",
       "0          16S V4-V5           411    cutadapt   \n",
       "1             18S V9           260    cutadapt   \n",
       "\n",
       "                     cluster_method pid_clustering  \\\n",
       "0  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "1  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1             PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706   \n",
       "\n",
       "                             code_repo  \\\n",
       "0  https://github.com/aomlomics/gomecc   \n",
       "1  https://github.com/aomlomics/gomecc   \n",
       "\n",
       "                                                                       identificationReferences  \\\n",
       "0                              10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline   \n",
       "1  10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline   \n",
       "\n",
       "                                                                                                                            controls_used  \n",
       "0  12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community  \n",
       "1                          12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls   "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>controls_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>411</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S V9</td>\n",
       "      <td>260</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  amplicon_sequenced  ampliconSize trim_method  \\\n",
       "0          16S V4-V5           411    cutadapt   \n",
       "1             18S V9           260    cutadapt   \n",
       "\n",
       "                     cluster_method pid_clustering  \\\n",
       "0  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "1  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1             PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706   \n",
       "\n",
       "                             code_repo  \\\n",
       "0  https://github.com/aomlomics/gomecc   \n",
       "1  https://github.com/aomlomics/gomecc   \n",
       "\n",
       "                                                                       identificationReferences  \\\n",
       "0                              10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline   \n",
       "1  10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline   \n",
       "\n",
       "                                                                                                                            controls_used  \n",
       "0  12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community  \n",
       "1                          12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls   "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '16S V4-V5','taxa_class_method'].item()\n",
    "occ16_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '16S V4-V5','taxa_ref_db'].item()\n",
    "\n",
    "occ16_test['identificationRemarks'] = occ16_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+occ16_test['Confidence'].astype(str) +\", against reference database: \"+occ16_test['taxa_ref_db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695'"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test['identificationRemarks'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### taxonID, basisOfRecord, eventID, nameAccordingTo, organismQuantityType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['taxonID'] = 'ASV:'+occ16_test['featureid']\n",
    "occ16_test['basisOfRecord'] = 'MaterialSample'\n",
    "occ16_test['nameAccordingTo'] = \"WoRMS\"\n",
    "occ16_test['organismQuantityType'] = \"DNA sequence reads\"\n",
    "occ16_test['recordedBy'] = data['study_data']['recordedBy'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### associatedSequences, materialSampleID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_name', 'library_id', 'title', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'lib_layout', 'platform',\n",
       "       'instrument_model', 'design_description', 'filetype', 'filename',\n",
       "       'filename2', 'biosample_accession', 'sra_accession', 'seq_method',\n",
       "       'nucl_acid_ext', 'amplicon_sequenced', 'target_gene',\n",
       "       'target_subfragment', 'pcr_primer_forward', 'pcr_primer_reverse',\n",
       "       'pcr_primer_name_forward', 'pcr_primer_name_reverse',\n",
       "       'pcr_primer_reference', 'pcr_cond', 'nucl_acid_amp', 'adapters',\n",
       "       'mid_barcode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test = occ16_test.merge(data['prep_data'].loc[data['prep_data']['amplicon_sequenced'] == '16S V4-V5',['sample_name','sra_accession','biosample_accession']], how='left', left_on ='eventID', right_on='sample_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                taxonomy  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "   Confidence                 eventID  organismQuantity  \\\n",
       "0    0.832190  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "1    0.867040  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "2    1.000000  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "3    0.952911  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "4    0.818195  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \\\n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order   \n",
       "4            None                 None           None   None  None   Kingdom   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "\n",
       "              sample_name sra_accession biosample_accession  \n",
       "0  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "1  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "2  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "3  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "4  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### eventID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['eventID'] = occ16_test['eventID']+\"_16S\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sampleSize "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sampleSize by total number of reads per sample\n",
    "x = asv_tables['16S V4-V5'].sum(numeric_only=True).astype('int')\n",
    "x.index = x.index+\"_16S\"\n",
    "occ16_test['sampleSizeValue'] = occ16_test['eventID'].map(x).astype('str')\n",
    "occ16_test['sampleSizeUnit'] = 'DNA sequence reads'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                                sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                                                                                                                                                taxonomy  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "   Confidence                     eventID  organismQuantity  \\\n",
       "0    0.832190  GOMECC4_27N_Sta1_DCM_A_16S                18   \n",
       "1    0.867040  GOMECC4_27N_Sta1_DCM_A_16S                36   \n",
       "2    1.000000  GOMECC4_27N_Sta1_DCM_A_16S                49   \n",
       "3    0.952911  GOMECC4_27N_Sta1_DCM_A_16S                 2   \n",
       "4    0.818195  GOMECC4_27N_Sta1_DCM_A_16S                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \\\n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order   \n",
       "4            None                 None           None   None  None   Kingdom   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "\n",
       "              sample_name sra_accession biosample_accession sampleSizeValue  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "1  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "2  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "3  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "4  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094           16187   \n",
       "\n",
       "       sampleSizeUnit  \n",
       "0  DNA sequence reads  \n",
       "1  DNA sequence reads  \n",
       "2  DNA sequence reads  \n",
       "3  DNA sequence reads  \n",
       "4  DNA sequence reads  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ16_test.drop(columns=['sample_name','featureid','taxonomy','Confidence','taxa_class_method','taxa_ref_db'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['associatedSequences'] = occ16_test['sra_accession']+' | '+ occ16_test['biosample_accession']+' | '+data['study_data']['bioproject_accession'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test.rename(columns={'biosample_accession': 'materialSampleID',\n",
    "                  'sequence': 'DNA_sequence'},inplace=True)\n",
    "                   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ16_test.drop(columns=['sra_accession'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DNA_sequence', 'eventID', 'organismQuantity', 'occurrenceID',\n",
       "       'verbatimIdentification', 'scientificName', 'scientificNameID',\n",
       "       'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'taxonRank',\n",
       "       'identificationRemarks', 'taxonID', 'basisOfRecord', 'nameAccordingTo',\n",
       "       'organismQuantityType', 'recordedBy', 'materialSampleID',\n",
       "       'sampleSizeValue', 'sampleSizeUnit', 'associatedSequences'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                            DNA_sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S                18   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S                36   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S                49   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S                 2   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \\\n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order   \n",
       "4            None                 None           None   None  None   Kingdom   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "  sampleSizeValue      sampleSizeUnit  \\\n",
       "0           16187  DNA sequence reads   \n",
       "1           16187  DNA sequence reads   \n",
       "2           16187  DNA sequence reads   \n",
       "3           16187  DNA sequence reads   \n",
       "4           16187  DNA sequence reads   \n",
       "\n",
       "                        associatedSequences  \n",
       "0  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "1  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "2  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "3  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "4  SRR26148187 | SAMN37516094 | PRJNA887898  "
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge event and occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>59</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>59</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    eventID        locationID  \\\n",
       "939      GOMECC4_CAPECORAL_Sta141_DCM_B_18S  CAPECORAL_Sta141   \n",
       "940      GOMECC4_CAPECORAL_Sta141_DCM_C_18S  CAPECORAL_Sta141   \n",
       "941  GOMECC4_CAPECORAL_Sta141_Surface_A_18S  CAPECORAL_Sta141   \n",
       "942  GOMECC4_CAPECORAL_Sta141_Surface_B_18S  CAPECORAL_Sta141   \n",
       "943  GOMECC4_CAPECORAL_Sta141_Surface_C_18S  CAPECORAL_Sta141   \n",
       "\n",
       "                  eventDate minimumDepthInMeters             locality  \\\n",
       "939  2021-10-20T12:47-04:00                   59  USA: Gulf of Mexico   \n",
       "940  2021-10-20T12:47-04:00                   59  USA: Gulf of Mexico   \n",
       "941  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "942  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "943  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "\n",
       "    decimalLatitude decimalLongitude samplingProtocol        waterBody  \\\n",
       "939          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "940          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "941          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "942          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "943          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "\n",
       "    maximumDepthInMeters                       parentEventID  \\\n",
       "939                   59      GOMECC4_CAPECORAL_Sta141_DCM_B   \n",
       "940                   59      GOMECC4_CAPECORAL_Sta141_DCM_C   \n",
       "941                    4  GOMECC4_CAPECORAL_Sta141_Surface_A   \n",
       "942                    4  GOMECC4_CAPECORAL_Sta141_Surface_B   \n",
       "943                    4  GOMECC4_CAPECORAL_Sta141_Surface_C   \n",
       "\n",
       "             datasetID geodeticDatum  \n",
       "939  noaa-aoml-gomecc4         WGS84  \n",
       "940  noaa-aoml-gomecc4         WGS84  \n",
       "941  noaa-aoml-gomecc4         WGS84  \n",
       "942  noaa-aoml-gomecc4         WGS84  \n",
       "943  noaa-aoml-gomecc4         WGS84  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_event_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_merged = occ16_test.merge(all_event_data,how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16187</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                            DNA_sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGACAGTTGAGGGTGAAATCCCGGAGCTTAACTTCGGAACTGCCCCCAATACTACTAATCTAGAGTTCGGAAGAGGTGAGTGGAATTC...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGCTCTTTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAAGACTGGAGAGCTAGAAAACGGAAGAGGGTAGTGGAATTC...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTCCGTAGCCGGTCTGGTACATTCGTGGGTAAATCAACTCGCTTAACGAGTTGAATTCTGCGAGGACGGCCAGACTTGGGACCGGGAGAGGTGTGGGGTACTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTCATTTAAGCGGTCCGATAAGTTAAAAGCCAACAGTTAGAGCCTAACTCTTTCAAGCTTTTAATACTGTCAGACTAGAGTATATCAGAGAATAGTAGAATTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTCCGTAGGCGTTTTGCTAAGTTGATCGTTAAATCCATCGGCTTAACCGATGACATGCGATCAAAACTGGCAGAATAGAATATGTGAGGGGAATGTAGAATTC...   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S                18   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S                36   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S                49   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S                 2   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S                 3   \n",
       "\n",
       "                                                     occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5084987093afa1916   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca90944d905beb2a980bc3   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f731954f38e3461564   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                                                                                                                  verbatimIdentification  \\\n",
       "0                      d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Parvibaculales; f__OCS116_clade; g__OCS116_clade; s__uncultured_marine   \n",
       "1                                        d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Puniceispirillales; f__SAR116_clade; g__SAR116_clade   \n",
       "2                                         d__Archaea; p__Thermoplasmatota; c__Thermoplasmata; o__Marine_Group_II; f__Marine_Group_II; g__Marine_Group_II   \n",
       "3                    d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade); c__Marinimicrobia_(SAR406_clade); o__Marinimicrobia_(SAR406_clade); f__Marinimicrobia_(SAR406_clade...   \n",
       "\n",
       "        scientificName                           scientificNameID   kingdom  \\\n",
       "0  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "1  Alphaproteobacteria  urn:lsid:marinespecies.org:taxname:392750  Bacteria   \n",
       "2       Thermoplasmata  urn:lsid:marinespecies.org:taxname:416268   Archaea   \n",
       "3        Rickettsiales  urn:lsid:marinespecies.org:taxname:570969  Bacteria   \n",
       "4             Bacteria       urn:lsid:marinespecies.org:taxname:6  Bacteria   \n",
       "\n",
       "           phylum                class          order family genus taxonRank  \\\n",
       "0  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "1  Proteobacteria  Alphaproteobacteria           None   None  None     Class   \n",
       "2   Euryarchaeota       Thermoplasmata           None   None  None     Class   \n",
       "3  Proteobacteria  Alphaproteobacteria  Rickettsiales   None  None     Order   \n",
       "4            None                 None           None   None  None   Kingdom   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.867040054, against reference database: Silva SSU Ref ...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 1.0, against reference database: Silva SSU Ref NR 99 v1...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.952910602, against reference database: Silva SSU Ref ...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.818195053, against reference database: Silva SSU Ref ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "  sampleSizeValue      sampleSizeUnit  \\\n",
       "0           16187  DNA sequence reads   \n",
       "1           16187  DNA sequence reads   \n",
       "2           16187  DNA sequence reads   \n",
       "3           16187  DNA sequence reads   \n",
       "4           16187  DNA sequence reads   \n",
       "\n",
       "                        associatedSequences locationID  \\\n",
       "0  SRR26148187 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "1  SRR26148187 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "2  SRR26148187 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "3  SRR26148187 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "4  SRR26148187 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "\n",
       "                eventDate minimumDepthInMeters  \\\n",
       "0  2021-09-14T11:00-04:00                   49   \n",
       "1  2021-09-14T11:00-04:00                   49   \n",
       "2  2021-09-14T11:00-04:00                   49   \n",
       "3  2021-09-14T11:00-04:00                   49   \n",
       "4  2021-09-14T11:00-04:00                   49   \n",
       "\n",
       "                                      locality decimalLatitude  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "\n",
       "  decimalLongitude samplingProtocol       waterBody maximumDepthInMeters  \\\n",
       "0          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "1          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "2          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "3          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "4          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "\n",
       "            parentEventID          datasetID geodeticDatum  \n",
       "0  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "1  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "2  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "3  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "4  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_merged.drop(columns=['DNA_sequence']).to_csv(\"../processed/occurrence_16S.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 18S worms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "18S PR2 database provides WORMS IDs for species that are in worms. We will read in that file, assign known worms ids, the do a search for unannotated taxa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "pr2_18S = pd.read_excel(\"../../../databases/18S_PR2/pr2_v5.0.0_SSU/pr2_version_5.0.0_taxonomy.xlsx\",\n",
    "    index_col=None, na_values=[\"\"])\n",
    "pr2_18S = pr2_18S.dropna(subset=['worms_id'])\n",
    "pr2_18S['worms_id'] = pr2_18S['worms_id'].astype('int').astype('str')\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('_',' ',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace(' sp\\.','',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace(' spp\\.','',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('-',' ',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('\\/',' ',regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "pr2_18S_dict = dict(zip(pr2_18S.species,pr2_18S.worms_id))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'614894'"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(pr2_18S_dict['Aphanocapsa feldmannii'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### code to get record from aphia id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Had some [issues with the parallelization](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr) on Mac M1. Adding 'OBJC_DISABLE_INITIALIZE_FORK_SAFETY = YES' to .bash_profile and then [This](https://github.com/python/cpython/issues/74570) fixed it.   \n",
    "Try to run without the bash_profile fix LATER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"no_proxy\"]=\"*\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_18S = asv_tables['18S V9'][['taxonomy','domain','supergroup','division','subdivision','class','order','family','genus','species']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1374, 10)"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_18S = tax_18S.drop_duplicates(ignore_index=True)\n",
    "tax_18S.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    worms_18s = WoRMS_matching.get_worms_from_aphiaid_or_name_parallel(\n",
    "    tax_df = tax_18S,worms_dict=pr2_18S_dict,ordered_rank_columns=['species','genus','family','order','class','subdivision','division','supergroup'],\n",
    "    full_tax_column=\"taxonomy\",full_tax_vI=True,n_proc=6)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;</td>\n",
       "      <td>genus</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:100211</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Trichocomaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;</td>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;</td>\n",
       "      <td>species</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:106286</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Cryptophyta</td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td>Cryptomonadales</td>\n",
       "      <td>Cryptomonadaceae</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;</td>\n",
       "      <td>species</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:101195</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligotrichea</td>\n",
       "      <td>Oligotrichida</td>\n",
       "      <td>Strombidiidae</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida;Annelida_X;Annelida_XX;Prionospio;Prionospio_dubia;</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aphiaID</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:131155</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>Polychaeta</td>\n",
       "      <td>Spionida</td>\n",
       "      <td>Spionidae</td>\n",
       "      <td>Prionospio</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;</td>\n",
       "      <td>class</td>\n",
       "      <td>Opalozoa</td>\n",
       "      <td>Opalozoa</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:582466</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Bigyra</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Subphylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                 full_tax  \\\n",
       "0    Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;   \n",
       "2               Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;   \n",
       "3                      Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida;Annelida_X;Annelida_XX;Prionospio;Prionospio_dubia;   \n",
       "4                                  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;   \n",
       "\n",
       "                                                                                                   verbatimIdentification  \\\n",
       "0    Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;   \n",
       "2               Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;   \n",
       "3                                                                                                        Prionospio dubia   \n",
       "4                                  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;   \n",
       "\n",
       "  old_taxonRank     old name    scientificName  \\\n",
       "0         genus  Aspergillus       Aspergillus   \n",
       "1       species   Goniomonas        Goniomonas   \n",
       "2       species  Strombidium       Strombidium   \n",
       "3           NaN      aphiaID  Prionospio dubia   \n",
       "4         class     Opalozoa          Opalozoa   \n",
       "\n",
       "                            scientificNameID    kingdom       phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:100211      Fungi   Ascomycota   \n",
       "1  urn:lsid:marinespecies.org:taxname:106286  Chromista  Cryptophyta   \n",
       "2  urn:lsid:marinespecies.org:taxname:101195  Chromista   Ciliophora   \n",
       "3  urn:lsid:marinespecies.org:taxname:131155   Animalia     Annelida   \n",
       "4  urn:lsid:marinespecies.org:taxname:582466  Chromista       Bigyra   \n",
       "\n",
       "            class            order            family        genus  taxonRank  \n",
       "0  Eurotiomycetes       Eurotiales    Trichocomaceae  Aspergillus      Genus  \n",
       "1   Cryptophyceae  Cryptomonadales  Cryptomonadaceae   Goniomonas      Genus  \n",
       "2    Oligotrichea    Oligotrichida     Strombidiidae  Strombidium      Genus  \n",
       "3      Polychaeta         Spionida         Spionidae   Prionospio    Species  \n",
       "4            None             None              None         None  Subphylum  "
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Haptista', 'Provora', 'Obazoa', 'TSAR', 'Cryptista:nucl',\n",
       "       'Archaeplastida'], dtype=object)"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# which taxa had absolutely no matches\n",
    "worms_18s[worms_18s[\"scientificName\"]==\"No match\"]['old name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida</td>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Nibbleridea;Nibbleridida;Nibbleridae;Nibbleromonas</td>\n",
       "      <td>Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Nibbleridea;Nibbleridida;Nibbleridae;Nibbleromonas</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Provora</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Centroplasthelida_XX;Centroplasthelida_XXX;Centroplasthelida_XXXX;Centroplasthelida_XXXXX...</td>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Centroplasthelida_XX;Centroplasthelida_XXX;Centroplasthelida_XXXX;Centroplasthelida_XXXXX...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebulidea;Nebulidida;Nebulidae;Nebulomonas;Nebulomonas_marisrubri;</td>\n",
       "      <td>Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebulidea;Nebulidida;Nebulidae;Nebulomonas;Nebulomonas_marisrubri;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Provora</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviatea_XX;Breviatea_XXX;Breviata-lineage;Breviata-lineage_X;Breviata-lineage_X_sp.;</td>\n",
       "      <td>Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviatea_XX;Breviatea_XXX;Breviata-lineage;Breviata-lineage_X;Breviata-lineage_X_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2</td>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Halocrusticida;Halocrusticida_baliensis;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Halocrusticida;Halocrusticida_baliensis;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida;Pterista</td>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida;Pterista</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Hemiselmidaceae:nucl;Chroomonas:nucl;Chr...</td>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Hemiselmidaceae:nucl;Chroomonas:nucl;Chr...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Cryptista:nucl</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>Eukaryota;Obazoa</td>\n",
       "      <td>Eukaryota;Obazoa</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;Archaeplastida</td>\n",
       "      <td>Eukaryota;Archaeplastida</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Archaeplastida</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1B;MAST-1B_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1B;MAST-1B_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl</td>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Cryptista:nucl</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Eukaryota;TSAR</td>\n",
       "      <td>Eukaryota;TSAR</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenophora;Ctenophora_X;Ctenophora_XX</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenophora;Ctenophora_X;Ctenophora_XX</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1A;MAST-1A_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1A;MAST-1A_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Opisthokonta_XXXXX;Opisthokonta_XXXXX_sp.;</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Opisthokonta_XXXXX;Opisthokonta_XXXXX_sp.;</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-1</td>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-1</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                  full_tax  \\\n",
       "121                                                                                  Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida   \n",
       "4                                                           Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Nibbleridea;Nibbleridida;Nibbleridae;Nibbleromonas   \n",
       "48   Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Centroplasthelida_XX;Centroplasthelida_XXX;Centroplasthelida_XXXX;Centroplasthelida_XXXXX...   \n",
       "62                                              Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebulidea;Nebulidida;Nebulidae;Nebulomonas;Nebulomonas_marisrubri;   \n",
       "72                           Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviatea_XX;Breviatea_XXX;Breviata-lineage;Breviata-lineage_X;Breviata-lineage_X_sp.;   \n",
       "108                                                            Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales   \n",
       "120                                                                      Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2   \n",
       "188                   Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Halocrusticida;Halocrusticida_baliensis;   \n",
       "191                                                                         Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida;Pterista   \n",
       "195  Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Hemiselmidaceae:nucl;Chroomonas:nucl;Chr...   \n",
       "227                                                                                                                                       Eukaryota;Obazoa   \n",
       "4                                                                                                                                 Eukaryota;Archaeplastida   \n",
       "19                                                                   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1B;MAST-1B_sp.;   \n",
       "46                                               Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl   \n",
       "60                                                                                                                                          Eukaryota;TSAR   \n",
       "65                                                                             Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenophora;Ctenophora_X;Ctenophora_XX   \n",
       "81                                                                   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1A;MAST-1A_sp.;   \n",
       "119             Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Opisthokonta_XXXXX;Opisthokonta_XXXXX_sp.;   \n",
       "177                                                                                                      Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae   \n",
       "203                                                                      Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-1   \n",
       "\n",
       "                                                                                                                                    verbatimIdentification  \\\n",
       "121                                                                                  Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida   \n",
       "4                                                           Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Nibbleridea;Nibbleridida;Nibbleridae;Nibbleromonas   \n",
       "48   Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Centroplasthelida_XX;Centroplasthelida_XXX;Centroplasthelida_XXXX;Centroplasthelida_XXXXX...   \n",
       "62                                              Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebulidea;Nebulidida;Nebulidae;Nebulomonas;Nebulomonas_marisrubri;   \n",
       "72                           Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviatea_XX;Breviatea_XXX;Breviata-lineage;Breviata-lineage_X;Breviata-lineage_X_sp.;   \n",
       "108                                                            Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales   \n",
       "120                                                                      Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-2   \n",
       "188                   Eukaryota;TSAR;Stramenopiles;Gyrista;Peronosporomycetes;Peronosporomycetes_X;Haliphthorales;Halocrusticida;Halocrusticida_baliensis;   \n",
       "191                                                                         Eukaryota;Haptista;Centroplasthelida;Centroplasthelida_X;Pterocystida;Pterista   \n",
       "195  Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl;Hemiselmidaceae:nucl;Chroomonas:nucl;Chr...   \n",
       "227                                                                                                                                       Eukaryota;Obazoa   \n",
       "4                                                                                                                                 Eukaryota;Archaeplastida   \n",
       "19                                                                   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1B;MAST-1B_sp.;   \n",
       "46                                               Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl;Cryptophyta_X:nucl;Cryptophyceae:nucl;Cryptomonadales:nucl   \n",
       "60                                                                                                                                          Eukaryota;TSAR   \n",
       "65                                                                             Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenophora;Ctenophora_X;Ctenophora_XX   \n",
       "81                                                                   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X;Gyrista_XX;MAST-1;MAST-1A;MAST-1A_sp.;   \n",
       "119             Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;Opisthokonta_XX;Opisthokonta_XXX;Opisthokonta_XXXX;Opisthokonta_XXXXX;Opisthokonta_XXXXX_sp.;   \n",
       "177                                                                                                      Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae   \n",
       "203                                                                      Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia_XX;Telonemia_XXX;Telonemia-Group-1   \n",
       "\n",
       "    old_taxonRank        old name scientificName scientificNameID   kingdom  \\\n",
       "121    supergroup        Haptista       No match             None       NaN   \n",
       "4      supergroup         Provora       No match             None       NaN   \n",
       "48     supergroup        Haptista       No match             None       NaN   \n",
       "62     supergroup         Provora       No match             None       NaN   \n",
       "72     supergroup          Obazoa       No match             None       NaN   \n",
       "108    supergroup            TSAR       No match             None       NaN   \n",
       "120    supergroup            TSAR       No match             None       NaN   \n",
       "188    supergroup            TSAR       No match             None       NaN   \n",
       "191    supergroup        Haptista       No match             None       NaN   \n",
       "195    supergroup  Cryptista:nucl       No match             None       NaN   \n",
       "227    supergroup          Obazoa       No match             None       NaN   \n",
       "4      supergroup  Archaeplastida       No match             None       NaN   \n",
       "19     supergroup            TSAR       No match             None       NaN   \n",
       "46     supergroup  Cryptista:nucl       No match             None       NaN   \n",
       "60     supergroup            TSAR       No match             None       NaN   \n",
       "65     supergroup          Obazoa       No match             None  Animalia   \n",
       "81     supergroup            TSAR       No match             None       NaN   \n",
       "119    supergroup          Obazoa       No match             None       NaN   \n",
       "177    supergroup            TSAR       No match             None       NaN   \n",
       "203    supergroup            TSAR       No match             None       NaN   \n",
       "\n",
       "    phylum class order family genus taxonRank  \n",
       "121    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "4      NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "48     NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "62     NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "72     NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "108    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "120    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "188    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "191    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "195    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "227    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "4      NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "19     NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "46     NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "60     NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "65    None  None  None   None  None   Kingdom  \n",
       "81     NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "119    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "177    NaN   NaN   NaN    NaN   NaN       NaN  \n",
       "203    NaN   NaN   NaN    NaN   NaN       NaN  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s[\"scientificName\"]==\"No match\"].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.loc[worms_18s[\"scientificName\"]==\"No match\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"scientificName\"]==\"Biota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Eukaryota;Haptista</td>\n",
       "      <td>Eukaryota;Haptista</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               full_tax verbatimIdentification old_taxonRank  old name  \\\n",
       "109          Unassigned             Unassigned           NaN       NaN   \n",
       "120  Eukaryota;Haptista     Eukaryota;Haptista    supergroup  Haptista   \n",
       "77            Eukaryota              Eukaryota           NaN       NaN   \n",
       "\n",
       "    scientificName scientificNameID kingdom phylum class order family genus  \\\n",
       "109            NaN              NaN     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "120            NaN              NaN     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "77             NaN              NaN     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "\n",
       "    taxonRank  \n",
       "109       NaN  \n",
       "120       NaN  \n",
       "77        NaN  "
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota;Haptista\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota;Haptista\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       full_tax verbatimIdentification old_taxonRank old name scientificName  \\\n",
       "109  Unassigned             Unassigned           NaN      NaN            NaN   \n",
       "\n",
       "    scientificNameID kingdom phylum class order family genus taxonRank  \n",
       "109              NaN     NaN    NaN   NaN   NaN    NaN   NaN       NaN  "
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 13)\n",
      "(0, 13)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(worms_18s[worms_18s['scientificName'].isna() == True].shape)\n",
    "worms_18s.loc[worms_18s['scientificName'].isna() == True,'scientificName'] = 'incertae sedis'\n",
    "worms_18s.loc[worms_18s['scientificName'] == 'incertae sedis','scientificNameID'] =  'urn:lsid:marinespecies.org:taxname:12'\n",
    "print(worms_18s[worms_18s['scientificName'].isna() == True].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(332, 13)"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s[\"old name\"]==\"aphiaID\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.to_csv(\"../processed/worms_18S_matching.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:100211</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Trichocomaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;</td>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:106286</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Cryptophyta</td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td>Cryptomonadales</td>\n",
       "      <td>Cryptomonadaceae</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:101195</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligotrichea</td>\n",
       "      <td>Oligotrichida</td>\n",
       "      <td>Strombidiidae</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida;Annelida_X;Annelida_XX;Prionospio;Prionospio_dubia;</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:131155</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>Polychaeta</td>\n",
       "      <td>Spionida</td>\n",
       "      <td>Spionidae</td>\n",
       "      <td>Prionospio</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;</td>\n",
       "      <td>Opalozoa</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:582466</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Bigyra</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Subphylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                 full_tax  \\\n",
       "0    Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;   \n",
       "2               Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;   \n",
       "3                      Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida;Annelida_X;Annelida_XX;Prionospio;Prionospio_dubia;   \n",
       "4                                  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;   \n",
       "\n",
       "                                                                                                   verbatimIdentification  \\\n",
       "0    Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota;Pezizomycotina;Eurotiomycetes;Aspergillus;Aspergillus_penicillioides;   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;Cryptophyceae;Goniomonadales;Goniomonadales_X;Goniomonas;Goniomonas_sp.;   \n",
       "2               Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrichea;Oligotrichida;Strombidiidae;Strombidium;Strombidium_sp.;   \n",
       "3                                                                                                        Prionospio dubia   \n",
       "4                                  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;Opalozoa_X;MAST-12;MAST-12A;MAST-12A_sp.;   \n",
       "\n",
       "     scientificName                           scientificNameID    kingdom  \\\n",
       "0       Aspergillus  urn:lsid:marinespecies.org:taxname:100211      Fungi   \n",
       "1        Goniomonas  urn:lsid:marinespecies.org:taxname:106286  Chromista   \n",
       "2       Strombidium  urn:lsid:marinespecies.org:taxname:101195  Chromista   \n",
       "3  Prionospio dubia  urn:lsid:marinespecies.org:taxname:131155   Animalia   \n",
       "4          Opalozoa  urn:lsid:marinespecies.org:taxname:582466  Chromista   \n",
       "\n",
       "        phylum           class            order            family  \\\n",
       "0   Ascomycota  Eurotiomycetes       Eurotiales    Trichocomaceae   \n",
       "1  Cryptophyta   Cryptophyceae  Cryptomonadales  Cryptomonadaceae   \n",
       "2   Ciliophora    Oligotrichea    Oligotrichida     Strombidiidae   \n",
       "3     Annelida      Polychaeta         Spionida         Spionidae   \n",
       "4       Bigyra            None             None              None   \n",
       "\n",
       "         genus  taxonRank  \n",
       "0  Aspergillus      Genus  \n",
       "1   Goniomonas      Genus  \n",
       "2  Strombidium      Genus  \n",
       "3   Prionospio    Species  \n",
       "4         None  Subphylum  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s.drop(columns=['old name','old_taxonRank'],inplace=True)\n",
    "worms_18s.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Occurrence and worms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146232, 16)"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S V9'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "2  2a31e5c01634165da99e7381279baa75   \n",
       "3  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                                                           sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "4    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                                                                                                       taxonomy  \\\n",
       "0       Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2           Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                           Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                              Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "   Confidence                 eventID  organismQuantity  \\\n",
       "0    0.922099  GOMECC4_27N_Sta1_DCM_A              1516   \n",
       "1    0.999947  GOMECC4_27N_Sta1_DCM_A               962   \n",
       "2    0.779948  GOMECC4_27N_Sta1_DCM_A              1164   \n",
       "3    0.999931  GOMECC4_27N_Sta1_DCM_A               287   \n",
       "4    0.986908  GOMECC4_27N_Sta1_DCM_A               250   \n",
       "\n",
       "                                                 occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                verbatimIdentification  \\\n",
       "0                                                                                 Neocalanus cristatus   \n",
       "1                                                                               Clausocalanus furcatus   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                     Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "           scientificName                           scientificNameID  \\\n",
       "0    Neocalanus cristatus  urn:lsid:marinespecies.org:taxname:104470   \n",
       "1  Clausocalanus furcatus  urn:lsid:marinespecies.org:taxname:104503   \n",
       "2             Acrocalanus  urn:lsid:marinespecies.org:taxname:104192   \n",
       "3             Dinophyceae   urn:lsid:marinespecies.org:taxname:19542   \n",
       "4          Gymnodiniaceae  urn:lsid:marinespecies.org:taxname:109410   \n",
       "\n",
       "     kingdom      phylum        class          order           family  \\\n",
       "0   Animalia  Arthropoda     Copepoda      Calanoida        Calanidae   \n",
       "1   Animalia  Arthropoda     Copepoda      Calanoida  Clausocalanidae   \n",
       "2   Animalia  Arthropoda     Copepoda      Calanoida    Paracalanidae   \n",
       "3  Chromista     Myzozoa  Dinophyceae           None             None   \n",
       "4  Chromista     Myzozoa  Dinophyceae  Gymnodiniales   Gymnodiniaceae   \n",
       "\n",
       "           genus taxonRank  \n",
       "0     Neocalanus   Species  \n",
       "1  Clausocalanus   Species  \n",
       "2    Acrocalanus     Genus  \n",
       "3           None     Class  \n",
       "4           None    Family  "
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get identificationRemarks\n",
    "occ18_test = occ['18S V9'].copy()\n",
    "occ18_test.drop(columns=['domain','supergroup','division','subdivision','class','order','family','genus','species'],inplace=True)\n",
    "#occ18_test.drop(columns=['old name'],inplace=True)\n",
    "\n",
    "occ18_test = occ18_test.merge(worms_18s, how='left', left_on ='taxonomy', right_on='full_tax')\n",
    "occ18_test.drop(columns='full_tax', inplace=True)\n",
    "occ18_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>controls_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>411</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S V9</td>\n",
       "      <td>260</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  amplicon_sequenced  ampliconSize trim_method  \\\n",
       "0          16S V4-V5           411    cutadapt   \n",
       "1             18S V9           260    cutadapt   \n",
       "\n",
       "                     cluster_method pid_clustering  \\\n",
       "0  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "1  Tourmaline; qiime2-2021.2; dada2            ASV   \n",
       "\n",
       "                                   taxa_class_method  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                            taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1             PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706   \n",
       "\n",
       "                             code_repo  \\\n",
       "0  https://github.com/aomlomics/gomecc   \n",
       "1  https://github.com/aomlomics/gomecc   \n",
       "\n",
       "                                                                       identificationReferences  \\\n",
       "0                              10.5281/zenodo.8392695 | https://github.com/aomlomics/tourmaline   \n",
       "1  10.5281/zenodo.8392706 | https://pr2-database.org/ | https://github.com/aomlomics/tourmaline   \n",
       "\n",
       "                                                                                                                            controls_used  \n",
       "0  12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 12 2nd PCR no-template controls | 3 Zymo mock community  \n",
       "1                          12 distilled water blanks | 2 PCR no-template controls | 7 extraction blanks | 7 2nd PCR no-template controls   "
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '18S V9','taxa_class_method'].item()\n",
    "occ18_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['amplicon_sequenced'] == '18S V9','taxa_ref_db'].item()\n",
    "\n",
    "occ18_test['identificationRemarks'] = occ18_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+occ18_test['Confidence'].astype(str) +\", against reference database: \"+occ18_test['taxa_ref_db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706'"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test['identificationRemarks'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### taxonID, basisOfRecord, eventID, nameAccordingTo, organismQuantityType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['taxonID'] = 'ASV:'+occ18_test['featureid']\n",
    "occ18_test['basisOfRecord'] = 'MaterialSample'\n",
    "occ18_test['nameAccordingTo'] = \"WoRMS\"\n",
    "occ18_test['organismQuantityType'] = \"DNA sequence reads\"\n",
    "occ18_test['recordedBy'] = data['study_data']['recordedBy'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### associatedSequences, materialSampleID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_name', 'library_id', 'title', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'lib_layout', 'platform',\n",
       "       'instrument_model', 'design_description', 'filetype', 'filename',\n",
       "       'filename2', 'biosample_accession', 'sra_accession', 'seq_method',\n",
       "       'nucl_acid_ext', 'amplicon_sequenced', 'target_gene',\n",
       "       'target_subfragment', 'pcr_primer_forward', 'pcr_primer_reverse',\n",
       "       'pcr_primer_name_forward', 'pcr_primer_name_reverse',\n",
       "       'pcr_primer_reference', 'pcr_cond', 'nucl_acid_amp', 'adapters',\n",
       "       'mid_barcode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test = occ18_test.merge(data['prep_data'].loc[data['prep_data']['amplicon_sequenced'] == '18S V9',['sample_name','sra_accession','biosample_accession']], how='left', left_on ='eventID', right_on='sample_name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### eventID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['eventID'] = occ18_test['eventID']+\"_18S\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sampleSize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sampleSize by total number of reads per sample\n",
    "x = asv_tables['18S V9'].sum(numeric_only=True).astype('int')\n",
    "x.index = x.index+\"_18S\"\n",
    "occ18_test['sampleSizeValue'] = occ18_test['eventID'].map(x).astype('str')\n",
    "occ18_test['sampleSizeUnit'] = 'DNA sequence reads'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ18_test.drop(columns=['sample_name','featureid','taxonomy','Confidence','taxa_class_method','taxa_ref_db'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['associatedSequences'] = occ18_test['sra_accession']+' | '+ occ18_test['biosample_accession']+' | '+data['study_data']['bioproject_accession'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test.rename(columns={'biosample_accession': 'materialSampleID',\n",
    "                  'sequence': 'DNA_sequence'},inplace=True)\n",
    "                   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ18_test.drop(columns=['sra_accession'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DNA_sequence', 'eventID', 'organismQuantity', 'occurrenceID',\n",
       "       'verbatimIdentification', 'scientificName', 'scientificNameID',\n",
       "       'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'taxonRank',\n",
       "       'identificationRemarks', 'taxonID', 'basisOfRecord', 'nameAccordingTo',\n",
       "       'organismQuantityType', 'recordedBy', 'materialSampleID',\n",
       "       'sampleSizeValue', 'sampleSizeUnit', 'associatedSequences'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Genus</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                       DNA_sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "4    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_18S              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_18S               962   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_18S              1164   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_18S               287   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_18S               250   \n",
       "\n",
       "                                                 occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                verbatimIdentification  \\\n",
       "0                                                                                 Neocalanus cristatus   \n",
       "1                                                                               Clausocalanus furcatus   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                     Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "           scientificName                           scientificNameID  \\\n",
       "0    Neocalanus cristatus  urn:lsid:marinespecies.org:taxname:104470   \n",
       "1  Clausocalanus furcatus  urn:lsid:marinespecies.org:taxname:104503   \n",
       "2             Acrocalanus  urn:lsid:marinespecies.org:taxname:104192   \n",
       "3             Dinophyceae   urn:lsid:marinespecies.org:taxname:19542   \n",
       "4          Gymnodiniaceae  urn:lsid:marinespecies.org:taxname:109410   \n",
       "\n",
       "     kingdom      phylum        class          order           family  \\\n",
       "0   Animalia  Arthropoda     Copepoda      Calanoida        Calanidae   \n",
       "1   Animalia  Arthropoda     Copepoda      Calanoida  Clausocalanidae   \n",
       "2   Animalia  Arthropoda     Copepoda      Calanoida    Paracalanidae   \n",
       "3  Chromista     Myzozoa  Dinophyceae           None             None   \n",
       "4  Chromista     Myzozoa  Dinophyceae  Gymnodiniales   Gymnodiniaceae   \n",
       "\n",
       "           genus taxonRank  \\\n",
       "0     Neocalanus   Species   \n",
       "1  Clausocalanus   Species   \n",
       "2    Acrocalanus     Genus   \n",
       "3           None     Class   \n",
       "4           None    Family   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:36aa75f9b28f5f831c2d631ba65c2bcb  MaterialSample           WoRMS   \n",
       "1  ASV:4e38e8ced9070952b314e1880bede1ca  MaterialSample           WoRMS   \n",
       "2  ASV:2a31e5c01634165da99e7381279baa75  MaterialSample           WoRMS   \n",
       "3  ASV:ecee60339b2fb88ea6d1c8d18054bed4  MaterialSample           WoRMS   \n",
       "4  ASV:fa1f1a97dd4ae7c826009186bad26384  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "  sampleSizeValue      sampleSizeUnit  \\\n",
       "0            9838  DNA sequence reads   \n",
       "1            9838  DNA sequence reads   \n",
       "2            9838  DNA sequence reads   \n",
       "3            9838  DNA sequence reads   \n",
       "4            9838  DNA sequence reads   \n",
       "\n",
       "                        associatedSequences  \n",
       "0  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "1  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "2  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "3  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "4  SRR26161153 | SAMN37516094 | PRJNA887898  "
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge event and occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_merged = occ18_test.merge(all_event_data,how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Genus</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...</td>\n",
       "      <td>ASV:ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...</td>\n",
       "      <td>ASV:fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>9838</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                       DNA_sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "3    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAGTGTTCAGTTCCTGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "4    GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAATGTTTGGATCCCGAACGTTGCAGCGGAAAGTTTAGTGAACCTTATCACTTAGAGGAAGGAGAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_18S              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_18S               962   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_18S              1164   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_18S               287   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_18S               250   \n",
       "\n",
       "                                                 occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314e1880bede1ca   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e7381279baa75   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                                                                verbatimIdentification  \\\n",
       "0                                                                                 Neocalanus cristatus   \n",
       "1                                                                               Clausocalanus furcatus   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;   \n",
       "3                                                  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae   \n",
       "4                     Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinophyceae;Gymnodiniales;Gymnodiniaceae   \n",
       "\n",
       "           scientificName                           scientificNameID  \\\n",
       "0    Neocalanus cristatus  urn:lsid:marinespecies.org:taxname:104470   \n",
       "1  Clausocalanus furcatus  urn:lsid:marinespecies.org:taxname:104503   \n",
       "2             Acrocalanus  urn:lsid:marinespecies.org:taxname:104192   \n",
       "3             Dinophyceae   urn:lsid:marinespecies.org:taxname:19542   \n",
       "4          Gymnodiniaceae  urn:lsid:marinespecies.org:taxname:109410   \n",
       "\n",
       "     kingdom      phylum        class          order           family  \\\n",
       "0   Animalia  Arthropoda     Copepoda      Calanoida        Calanidae   \n",
       "1   Animalia  Arthropoda     Copepoda      Calanoida  Clausocalanidae   \n",
       "2   Animalia  Arthropoda     Copepoda      Calanoida    Paracalanidae   \n",
       "3  Chromista     Myzozoa  Dinophyceae           None             None   \n",
       "4  Chromista     Myzozoa  Dinophyceae  Gymnodiniales   Gymnodiniaceae   \n",
       "\n",
       "           genus taxonRank  \\\n",
       "0     Neocalanus   Species   \n",
       "1  Clausocalanus   Species   \n",
       "2    Acrocalanus     Genus   \n",
       "3           None     Class   \n",
       "4           None    Family   \n",
       "\n",
       "                                                                                                                                   identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 ...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999946735, against reference database: PR2 v5.0.1; V9...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.779948049, against reference database: PR2 v5.0.1; V9...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.999930607, against reference database: PR2 v5.0.1; V9...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.98690791, against reference database: PR2 v5.0.1; V9 ...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:36aa75f9b28f5f831c2d631ba65c2bcb  MaterialSample           WoRMS   \n",
       "1  ASV:4e38e8ced9070952b314e1880bede1ca  MaterialSample           WoRMS   \n",
       "2  ASV:2a31e5c01634165da99e7381279baa75  MaterialSample           WoRMS   \n",
       "3  ASV:ecee60339b2fb88ea6d1c8d18054bed4  MaterialSample           WoRMS   \n",
       "4  ASV:fa1f1a97dd4ae7c826009186bad26384  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "  sampleSizeValue      sampleSizeUnit  \\\n",
       "0            9838  DNA sequence reads   \n",
       "1            9838  DNA sequence reads   \n",
       "2            9838  DNA sequence reads   \n",
       "3            9838  DNA sequence reads   \n",
       "4            9838  DNA sequence reads   \n",
       "\n",
       "                        associatedSequences locationID  \\\n",
       "0  SRR26161153 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "1  SRR26161153 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "2  SRR26161153 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "3  SRR26161153 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "4  SRR26161153 | SAMN37516094 | PRJNA887898   27N_Sta1   \n",
       "\n",
       "                eventDate minimumDepthInMeters  \\\n",
       "0  2021-09-14T11:00-04:00                   49   \n",
       "1  2021-09-14T11:00-04:00                   49   \n",
       "2  2021-09-14T11:00-04:00                   49   \n",
       "3  2021-09-14T11:00-04:00                   49   \n",
       "4  2021-09-14T11:00-04:00                   49   \n",
       "\n",
       "                                      locality decimalLatitude  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "\n",
       "  decimalLongitude samplingProtocol       waterBody maximumDepthInMeters  \\\n",
       "0          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "1          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "2          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "3          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "4          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "\n",
       "            parentEventID          datasetID geodeticDatum  \n",
       "0  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "1  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "2  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "3  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "4  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  "
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_merged.drop(columns=['DNA_sequence']).to_csv(\"../processed/occurrence_18S.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### combine 16s and 18s occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146232, 36)"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all = pd.concat([occ16_merged,occ18_merged],axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all['occurrenceStatus'] = 'present' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(311390, 37)"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all.drop(columns=['DNA_sequence']).to_csv(\"../processed/occurrence.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNA-derived"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_dict = dwc_data['dna'].to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['eventID', 'occurrenceID', 'DNA_sequence', 'sop', 'nucl_acid_ext', 'samp_vol_we_dna_ext', 'samp_mat_process', 'nucl_acid_amp', 'target_gene', 'target_subfragment', 'ampliconSize', 'lib_layout', 'pcr_primer_forward', 'pcr_primer_reverse', 'pcr_primer_name_forward', 'pcr_primer_name_reverse', 'pcr_primer_reference', 'pcr_cond', 'seq_meth', 'otu_class_appr', 'otu_seq_comp_appr', 'otu_db', 'env_broad_scale', 'env_local_scale', 'env_medium', 'size_frac', 'concentration', 'concentrationUnit', 'samp_collect_device', 'source_mat_id'])"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_dict.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### sample_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samp_vol_we_dna_ext {'AOML_term': 'samp_vol_we_dna_ext', 'AOML_file': 'sample_data', 'DwC_definition': 'Volume (ml) or mass (g) of total collected sample processed for DNA extraction.MIXS:0000111', 'Example': nan}\n",
      "samp_mat_process {'AOML_term': 'samp_mat_process', 'AOML_file': 'sample_data', 'DwC_definition': 'Any processing applied to the sample during or after retrieving the sample from environment. This field accepts OBI, for a browser of OBI (v 2018-02-12) terms please see http://purl.bioontology.org/ontology/OBI', 'Example': nan}\n",
      "env_broad_scale {'AOML_term': 'env_broad_scale', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "env_local_scale {'AOML_term': 'env_local_scale', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "env_medium {'AOML_term': 'env_medium', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "size_frac {'AOML_term': 'size_frac', 'AOML_file': 'sample_data', 'DwC_definition': 'Filtering pore size used in sample preparation. Examples: 0-0.22 micrometer', 'Example': nan}\n",
      "concentration {'AOML_term': 'dna_conc', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "concentrationUnit {'AOML_term': 'derived: dna_conc', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "samp_collect_device {'AOML_term': 'samp_collect_device', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "source_mat_id {'AOML_term': 'source_mat_id', 'AOML_file': 'sample_data', 'DwC_definition': 'used here for the nisken bottle sample', 'Example': nan}\n"
     ]
    }
   ],
   "source": [
    "# check which dna file terms are in sample_data\n",
    "for key in dna_dict.keys():\n",
    "    if dna_dict[key]['AOML_file'] == 'sample_data':\n",
    "        print(key,dna_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename sample_data columns to fit DwC standard\n",
    "rename_dict = {}\n",
    "gen = (x for x in dna_dict.keys() if dna_dict[x]['AOML_file'] == 'sample_data')\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[dna_dict[x]['AOML_term']] = x\n",
    "\n",
    "gen = (x for x in dna_dict.keys() if dna_dict[x]['AOML_file'] == 'prep_data')\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[dna_dict[x]['AOML_term']] = x\n",
    "\n",
    "gen = (x for x in dna_dict.keys() if dna_dict[x]['AOML_file'] == 'analysis_data')\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[dna_dict[x]['AOML_term']] = x\n",
    "\n",
    "dna_sample = data['sample_data'].rename(columns=rename_dict).copy()\n",
    "dna_prep = data['prep_data'].rename(columns=rename_dict).copy()\n",
    "dna_analysis = data['analysis_data'].rename(columns=rename_dict).copy()\n",
    "\n",
    "#dna_sample = dna_sample.drop(columns=[col for col in dna_sample if col not in rename_dict.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>metagenome_sequenced</th>\n",
       "      <th>organism</th>\n",
       "      <th>collection_date_local</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>depth</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>...</th>\n",
       "      <th>ammonium</th>\n",
       "      <th>carbonate</th>\n",
       "      <th>diss_inorg_carb</th>\n",
       "      <th>diss_oxygen</th>\n",
       "      <th>fluor</th>\n",
       "      <th>hydrogen_ion</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>nitrite</th>\n",
       "      <th>nitrate_plus_nitrite</th>\n",
       "      <th>omega_arag</th>\n",
       "      <th>pco2</th>\n",
       "      <th>ph</th>\n",
       "      <th>phosphate</th>\n",
       "      <th>pressure</th>\n",
       "      <th>salinity</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>waterBody</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516091</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516092</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516093</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516095</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID serial_number        cruise_id line_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S   GOMECC4_001  GOMECC-4 (2021)     27N   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S   GOMECC4_002  GOMECC-4 (2021)     27N   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S   GOMECC4_003  GOMECC-4 (2021)     27N   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_16S   GOMECC4_004  GOMECC-4 (2021)     27N   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_16S   GOMECC4_005  GOMECC-4 (2021)     27N   \n",
       "\n",
       "    station ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0  27N_Sta1             3                A  GOMECC4_27N_Sta1_Deep   \n",
       "1  27N_Sta1             3                B  GOMECC4_27N_Sta1_Deep   \n",
       "2  27N_Sta1             3                C  GOMECC4_27N_Sta1_Deep   \n",
       "3  27N_Sta1            14                A   GOMECC4_27N_Sta1_DCM   \n",
       "4  27N_Sta1            14                B   GOMECC4_27N_Sta1_DCM   \n",
       "\n",
       "                              biological_replicates extract_number  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C      Plate4_52   \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C      Plate4_60   \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B      Plate4_62   \n",
       "3    GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C      Plate4_53   \n",
       "4    GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C      Plate4_46   \n",
       "\n",
       "                                             sample_title  \\\n",
       "0  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A   \n",
       "1  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B   \n",
       "2  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C   \n",
       "3   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A   \n",
       "4   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B   \n",
       "\n",
       "  bioproject_accession biosample_accession metagenome_sequenced  \\\n",
       "0          PRJNA887898        SAMN37516091     planned for FY24   \n",
       "1          PRJNA887898        SAMN37516092     planned for FY24   \n",
       "2          PRJNA887898        SAMN37516093     planned for FY24   \n",
       "3          PRJNA887898        SAMN37516094     planned for FY24   \n",
       "4          PRJNA887898        SAMN37516095     planned for FY24   \n",
       "\n",
       "              organism   collection_date_local   collection_date  depth  \\\n",
       "0  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "1  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "2  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "3  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00   49 m   \n",
       "4  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00   49 m   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "4  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "\n",
       "                  env_medium                                 geo_loc_name  \\\n",
       "0  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "             lat_lon decimalLatitude decimalLongitude  ...         ammonium  \\\n",
       "0  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "1  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "2  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "3  26.997 N 79.618 W          26.997          -79.618  ...  0.32968 µmol/kg   \n",
       "4  26.997 N 79.618 W          26.997          -79.618  ...  0.32968 µmol/kg   \n",
       "\n",
       "        carbonate  diss_inorg_carb      diss_oxygen   fluor    hydrogen_ion  \\\n",
       "0  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "1  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "2  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "3  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   0.036  0.0000000094 M   \n",
       "4  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   0.036  0.0000000094 M   \n",
       "\n",
       "           nitrate          nitrite nitrate_plus_nitrite omega_arag      pco2  \\\n",
       "0  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "1  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "2  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "3        0 µmol/kg        0 µmol/kg            0 µmol/kg      3.805  423 µatm   \n",
       "4        0 µmol/kg        0 µmol/kg            0 µmol/kg      3.805  423 µatm   \n",
       "\n",
       "      ph        phosphate  pressure    salinity      samp_store_loc  \\\n",
       "0  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "1  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "2  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "3  8.027   0.0517 µmol/kg   49 dbar  36.325 psu  NOAA/AOML Room 248   \n",
       "4  8.027   0.0517 µmol/kg   49 dbar  36.325 psu  NOAA/AOML Room 248   \n",
       "\n",
       "  samp_store_temp         silicate  size_frac_low size_frac_up       temp  \\\n",
       "0          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "1          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "2          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "3          -20 °C  1.05635 µmol/kg  no pre-filter      0.22 µm  28.592 °C   \n",
       "4          -20 °C  1.05635 µmol/kg  no pre-filter      0.22 µm  28.592 °C   \n",
       "\n",
       "   tot_alkalinity tot_depth_water_col transmittance       waterBody  \n",
       "0  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "1  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "2  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "3    2371 µmol/kg               623 m         4.665  Atlantic Ocean  \n",
       "4    2371 µmol/kg               623 m         4.665  Atlantic Ocean  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_16 = dna_sample[dna_sample['amplicon_sequenced'].str.contains('16S V4-V5')].copy()\n",
    "dna_16['eventID'] = dna_16['eventID']+\"_16S\"\n",
    "dna_16.drop(columns=['amplicon_sequenced'],inplace=True)\n",
    "dna_16.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>metagenome_sequenced</th>\n",
       "      <th>organism</th>\n",
       "      <th>collection_date_local</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>depth</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>...</th>\n",
       "      <th>ammonium</th>\n",
       "      <th>carbonate</th>\n",
       "      <th>diss_inorg_carb</th>\n",
       "      <th>diss_oxygen</th>\n",
       "      <th>fluor</th>\n",
       "      <th>hydrogen_ion</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>nitrite</th>\n",
       "      <th>nitrate_plus_nitrite</th>\n",
       "      <th>omega_arag</th>\n",
       "      <th>pco2</th>\n",
       "      <th>ph</th>\n",
       "      <th>phosphate</th>\n",
       "      <th>pressure</th>\n",
       "      <th>salinity</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>waterBody</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_18S</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516091</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_18S</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516092</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_18S</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516093</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_18S</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516095</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID serial_number        cruise_id line_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_18S   GOMECC4_001  GOMECC-4 (2021)     27N   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_18S   GOMECC4_002  GOMECC-4 (2021)     27N   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_18S   GOMECC4_003  GOMECC-4 (2021)     27N   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_18S   GOMECC4_004  GOMECC-4 (2021)     27N   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_18S   GOMECC4_005  GOMECC-4 (2021)     27N   \n",
       "\n",
       "    station ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0  27N_Sta1             3                A  GOMECC4_27N_Sta1_Deep   \n",
       "1  27N_Sta1             3                B  GOMECC4_27N_Sta1_Deep   \n",
       "2  27N_Sta1             3                C  GOMECC4_27N_Sta1_Deep   \n",
       "3  27N_Sta1            14                A   GOMECC4_27N_Sta1_DCM   \n",
       "4  27N_Sta1            14                B   GOMECC4_27N_Sta1_DCM   \n",
       "\n",
       "                              biological_replicates extract_number  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C      Plate4_52   \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C      Plate4_60   \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B      Plate4_62   \n",
       "3    GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C      Plate4_53   \n",
       "4    GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C      Plate4_46   \n",
       "\n",
       "                                             sample_title  \\\n",
       "0  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A   \n",
       "1  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B   \n",
       "2  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C   \n",
       "3   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A   \n",
       "4   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B   \n",
       "\n",
       "  bioproject_accession biosample_accession metagenome_sequenced  \\\n",
       "0          PRJNA887898        SAMN37516091     planned for FY24   \n",
       "1          PRJNA887898        SAMN37516092     planned for FY24   \n",
       "2          PRJNA887898        SAMN37516093     planned for FY24   \n",
       "3          PRJNA887898        SAMN37516094     planned for FY24   \n",
       "4          PRJNA887898        SAMN37516095     planned for FY24   \n",
       "\n",
       "              organism   collection_date_local   collection_date  depth  \\\n",
       "0  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "1  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "2  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "3  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00   49 m   \n",
       "4  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00   49 m   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "4  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "\n",
       "                  env_medium                                 geo_loc_name  \\\n",
       "0  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "             lat_lon decimalLatitude decimalLongitude  ...         ammonium  \\\n",
       "0  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "1  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "2  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "3  26.997 N 79.618 W          26.997          -79.618  ...  0.32968 µmol/kg   \n",
       "4  26.997 N 79.618 W          26.997          -79.618  ...  0.32968 µmol/kg   \n",
       "\n",
       "        carbonate  diss_inorg_carb      diss_oxygen   fluor    hydrogen_ion  \\\n",
       "0  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "1  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "2  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "3  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   0.036  0.0000000094 M   \n",
       "4  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   0.036  0.0000000094 M   \n",
       "\n",
       "           nitrate          nitrite nitrate_plus_nitrite omega_arag      pco2  \\\n",
       "0  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "1  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "2  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "3        0 µmol/kg        0 µmol/kg            0 µmol/kg      3.805  423 µatm   \n",
       "4        0 µmol/kg        0 µmol/kg            0 µmol/kg      3.805  423 µatm   \n",
       "\n",
       "      ph        phosphate  pressure    salinity      samp_store_loc  \\\n",
       "0  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "1  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "2  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "3  8.027   0.0517 µmol/kg   49 dbar  36.325 psu  NOAA/AOML Room 248   \n",
       "4  8.027   0.0517 µmol/kg   49 dbar  36.325 psu  NOAA/AOML Room 248   \n",
       "\n",
       "  samp_store_temp         silicate  size_frac_low size_frac_up       temp  \\\n",
       "0          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "1          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "2          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "3          -20 °C  1.05635 µmol/kg  no pre-filter      0.22 µm  28.592 °C   \n",
       "4          -20 °C  1.05635 µmol/kg  no pre-filter      0.22 µm  28.592 °C   \n",
       "\n",
       "   tot_alkalinity tot_depth_water_col transmittance       waterBody  \n",
       "0  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "1  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "2  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "3    2371 µmol/kg               623 m         4.665  Atlantic Ocean  \n",
       "4    2371 µmol/kg               623 m         4.665  Atlantic Ocean  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_18 = dna_sample[dna_sample['amplicon_sequenced'].str.contains('18S V9')].copy()\n",
    "dna_18['eventID'] = dna_18['eventID']+\"_18S\"\n",
    "dna_18.drop(columns=['amplicon_sequenced'],inplace=True)\n",
    "dna_18.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>metagenome_sequenced</th>\n",
       "      <th>organism</th>\n",
       "      <th>collection_date_local</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>depth</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>...</th>\n",
       "      <th>ammonium</th>\n",
       "      <th>carbonate</th>\n",
       "      <th>diss_inorg_carb</th>\n",
       "      <th>diss_oxygen</th>\n",
       "      <th>fluor</th>\n",
       "      <th>hydrogen_ion</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>nitrite</th>\n",
       "      <th>nitrate_plus_nitrite</th>\n",
       "      <th>omega_arag</th>\n",
       "      <th>pco2</th>\n",
       "      <th>ph</th>\n",
       "      <th>phosphate</th>\n",
       "      <th>pressure</th>\n",
       "      <th>salinity</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>waterBody</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516091</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516092</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516093</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516095</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID serial_number        cruise_id line_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S   GOMECC4_001  GOMECC-4 (2021)     27N   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S   GOMECC4_002  GOMECC-4 (2021)     27N   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S   GOMECC4_003  GOMECC-4 (2021)     27N   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_16S   GOMECC4_004  GOMECC-4 (2021)     27N   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_16S   GOMECC4_005  GOMECC-4 (2021)     27N   \n",
       "\n",
       "    station ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0  27N_Sta1             3                A  GOMECC4_27N_Sta1_Deep   \n",
       "1  27N_Sta1             3                B  GOMECC4_27N_Sta1_Deep   \n",
       "2  27N_Sta1             3                C  GOMECC4_27N_Sta1_Deep   \n",
       "3  27N_Sta1            14                A   GOMECC4_27N_Sta1_DCM   \n",
       "4  27N_Sta1            14                B   GOMECC4_27N_Sta1_DCM   \n",
       "\n",
       "                              biological_replicates extract_number  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C      Plate4_52   \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C      Plate4_60   \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B      Plate4_62   \n",
       "3    GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C      Plate4_53   \n",
       "4    GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C      Plate4_46   \n",
       "\n",
       "                                             sample_title  \\\n",
       "0  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A   \n",
       "1  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B   \n",
       "2  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C   \n",
       "3   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A   \n",
       "4   Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B   \n",
       "\n",
       "  bioproject_accession biosample_accession metagenome_sequenced  \\\n",
       "0          PRJNA887898        SAMN37516091     planned for FY24   \n",
       "1          PRJNA887898        SAMN37516092     planned for FY24   \n",
       "2          PRJNA887898        SAMN37516093     planned for FY24   \n",
       "3          PRJNA887898        SAMN37516094     planned for FY24   \n",
       "4          PRJNA887898        SAMN37516095     planned for FY24   \n",
       "\n",
       "              organism   collection_date_local   collection_date  depth  \\\n",
       "0  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "1  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "2  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00  618 m   \n",
       "3  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00   49 m   \n",
       "4  seawater metagenome  2021-09-14T11:00-04:00  2021-09-14T07:00   49 m   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "4  marine biome [ENVO:00000447]       marine photic zone [ENVO:00000209]   \n",
       "\n",
       "                  env_medium                                 geo_loc_name  \\\n",
       "0  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4  sea water [ENVO:00002149]  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "             lat_lon decimalLatitude decimalLongitude  ...         ammonium  \\\n",
       "0  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "1  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "2  26.997 N 79.618 W          26.997          -79.618  ...  0.25971 µmol/kg   \n",
       "3  26.997 N 79.618 W          26.997          -79.618  ...  0.32968 µmol/kg   \n",
       "4  26.997 N 79.618 W          26.997          -79.618  ...  0.32968 µmol/kg   \n",
       "\n",
       "        carbonate  diss_inorg_carb      diss_oxygen   fluor    hydrogen_ion  \\\n",
       "0  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "1  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "2  88.434 µmol/kg  2215.45 µmol/kg   129.44 µmol/kg  0.0308  0.0000000142 M   \n",
       "3  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   0.036  0.0000000094 M   \n",
       "4  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg   0.036  0.0000000094 M   \n",
       "\n",
       "           nitrate          nitrite nitrate_plus_nitrite omega_arag      pco2  \\\n",
       "0  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "1  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "2  29.3256 µmol/kg  0.00391 µmol/kg      29.3295 µmol/kg      1.168  624 µatm   \n",
       "3        0 µmol/kg        0 µmol/kg            0 µmol/kg      3.805  423 µatm   \n",
       "4        0 µmol/kg        0 µmol/kg            0 µmol/kg      3.805  423 µatm   \n",
       "\n",
       "      ph        phosphate  pressure    salinity      samp_store_loc  \\\n",
       "0  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "1  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "2  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "3  8.027   0.0517 µmol/kg   49 dbar  36.325 psu  NOAA/AOML Room 248   \n",
       "4  8.027   0.0517 µmol/kg   49 dbar  36.325 psu  NOAA/AOML Room 248   \n",
       "\n",
       "  samp_store_temp         silicate  size_frac_low size_frac_up       temp  \\\n",
       "0          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "1          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "2          -20 °C  20.3569 µmol/kg  no pre-filter      0.22 µm   7.479 °C   \n",
       "3          -20 °C  1.05635 µmol/kg  no pre-filter      0.22 µm  28.592 °C   \n",
       "4          -20 °C  1.05635 µmol/kg  no pre-filter      0.22 µm  28.592 °C   \n",
       "\n",
       "   tot_alkalinity tot_depth_water_col transmittance       waterBody  \n",
       "0  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "1  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "2  2318.9 µmol/kg               623 m        4.7221  Atlantic Ocean  \n",
       "3    2371 µmol/kg               623 m         4.665  Atlantic Ocean  \n",
       "4    2371 µmol/kg               623 m         4.665  Atlantic Ocean  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_sample = pd.concat([dna_16,dna_18],axis=0,ignore_index=True)\n",
    "dna_sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_1</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516307</td>\n",
       "      <td>SRR26148474</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_10</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_10_S10_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_10_S10_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516268</td>\n",
       "      <td>SRR26148413</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_11</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_11_S11_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_11_S11_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516264</td>\n",
       "      <td>SRR26148140</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta49_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_12</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_12_S12_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_12_S12_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516246</td>\n",
       "      <td>SRR26148197</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S</td>\n",
       "      <td>GOMECC16S_Plate1_13</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_13_S13_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_13_S13_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516308</td>\n",
       "      <td>SRR26148464</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                eventID           library_id  \\\n",
       "4   GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S   GOMECC16S_Plate1_1   \n",
       "6     GOMECC4_GALVESTON_Sta54_DCM_B_16S  GOMECC16S_Plate1_10   \n",
       "8    GOMECC4_GALVESTON_Sta54_Deep_A_16S  GOMECC16S_Plate1_11   \n",
       "10   GOMECC4_GALVESTON_Sta49_Deep_A_16S  GOMECC16S_Plate1_12   \n",
       "12  GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S  GOMECC16S_Plate1_13   \n",
       "\n",
       "                                                                    title  \\\n",
       "4   16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "6   16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "8   16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "10  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "12  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "\n",
       "   library_strategy library_source library_selection lib_layout  platform  \\\n",
       "4          AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "6          AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "8          AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "10         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "12         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "\n",
       "   instrument_model  \\\n",
       "4    Illumina MiSeq   \n",
       "6    Illumina MiSeq   \n",
       "8    Illumina MiSeq   \n",
       "10   Illumina MiSeq   \n",
       "12   Illumina MiSeq   \n",
       "\n",
       "                                                                                                                                       design_description  \\\n",
       "4   Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "6   Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "8   Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "10  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "12  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "\n",
       "   filetype                                      filename  \\\n",
       "4     fastq    GOMECC16S_Plate1_1_S1_L001_R1_001.fastq.gz   \n",
       "6     fastq  GOMECC16S_Plate1_10_S10_L001_R1_001.fastq.gz   \n",
       "8     fastq  GOMECC16S_Plate1_11_S11_L001_R1_001.fastq.gz   \n",
       "10    fastq  GOMECC16S_Plate1_12_S12_L001_R1_001.fastq.gz   \n",
       "12    fastq  GOMECC16S_Plate1_13_S13_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                       filename2 biosample_accession  \\\n",
       "4     GOMECC16S_Plate1_1_S1_L001_R2_001.fastq.gz        SAMN37516307   \n",
       "6   GOMECC16S_Plate1_10_S10_L001_R2_001.fastq.gz        SAMN37516268   \n",
       "8   GOMECC16S_Plate1_11_S11_L001_R2_001.fastq.gz        SAMN37516264   \n",
       "10  GOMECC16S_Plate1_12_S12_L001_R2_001.fastq.gz        SAMN37516246   \n",
       "12  GOMECC16S_Plate1_13_S13_L001_R2_001.fastq.gz        SAMN37516308   \n",
       "\n",
       "   sra_accession              seq_meth  \\\n",
       "4    SRR26148474  Illumina MiSeq 2x250   \n",
       "6    SRR26148413  Illumina MiSeq 2x250   \n",
       "8    SRR26148140  Illumina MiSeq 2x250   \n",
       "10   SRR26148197  Illumina MiSeq 2x250   \n",
       "12   SRR26148464  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                           nucl_acid_ext  \\\n",
       "4   https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "6   https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "8   https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "10  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "12  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "   amplicon_sequenced target_gene target_subfragment   pcr_primer_forward  \\\n",
       "4           16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "6           16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "8           16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "10          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "12          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "\n",
       "      pcr_primer_reverse pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "4   CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "6   CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "8   CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "10  CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "12  CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "\n",
       "       pcr_primer_reference  \\\n",
       "4   10.1111/1462-2920.13023   \n",
       "6   10.1111/1462-2920.13023   \n",
       "8   10.1111/1462-2920.13023   \n",
       "10  10.1111/1462-2920.13023   \n",
       "12  10.1111/1462-2920.13023   \n",
       "\n",
       "                                                                                                       pcr_cond  \\\n",
       "4   initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "6   initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "8   initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "10  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "12  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "\n",
       "              nucl_acid_amp                                       adapters  \\\n",
       "4   10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "6   10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "8   10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "10  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "12  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "\n",
       "              mid_barcode  \n",
       "4   missing: not provided  \n",
       "6   missing: not provided  \n",
       "8   missing: not provided  \n",
       "10  missing: not provided  \n",
       "12  missing: not provided  "
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prep_16 = dna_prep[dna_prep['amplicon_sequenced'].str.contains('16S V4-V5')].copy()\n",
    "prep_16['eventID'] = prep_16['eventID']+\"_16S\"\n",
    "prep_16.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC18S_Plate4_53_S340_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC18S_Plate4_53_S340_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>SRR26161153</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S V9</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_18S</td>\n",
       "      <td>GOMECC18S_Plate4_46</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC18S_Plate4_46_S333_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC18S_Plate4_46_S333_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516095</td>\n",
       "      <td>SRR26161138</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S V9</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_C_18S</td>\n",
       "      <td>GOMECC18S_Plate4_54</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC18S_Plate4_54_S341_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC18S_Plate4_54_S341_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516096</td>\n",
       "      <td>SRR26160919</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S V9</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_18S</td>\n",
       "      <td>GOMECC18S_Plate4_52</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC18S_Plate4_52_S339_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC18S_Plate4_52_S339_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516091</td>\n",
       "      <td>SRR26160709</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S V9</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_18S</td>\n",
       "      <td>GOMECC18S_Plate4_60</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC18S_Plate4_60_S347_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC18S_Plate4_60_S347_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516092</td>\n",
       "      <td>SRR26160970</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S V9</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID           library_id  \\\n",
       "1   GOMECC4_27N_Sta1_DCM_A_18S  GOMECC18S_Plate4_53   \n",
       "3   GOMECC4_27N_Sta1_DCM_B_18S  GOMECC18S_Plate4_46   \n",
       "5   GOMECC4_27N_Sta1_DCM_C_18S  GOMECC18S_Plate4_54   \n",
       "7  GOMECC4_27N_Sta1_Deep_A_18S  GOMECC18S_Plate4_52   \n",
       "9  GOMECC4_27N_Sta1_Deep_B_18S  GOMECC18S_Plate4_60   \n",
       "\n",
       "                                                                   title  \\\n",
       "1  18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "3  18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "5  18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "7  18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "9  18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "\n",
       "  library_strategy library_source library_selection lib_layout  platform  \\\n",
       "1         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "3         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "5         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "7         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "9         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "\n",
       "  instrument_model  \\\n",
       "1   Illumina MiSeq   \n",
       "3   Illumina MiSeq   \n",
       "5   Illumina MiSeq   \n",
       "7   Illumina MiSeq   \n",
       "9   Illumina MiSeq   \n",
       "\n",
       "                                                                                                                                      design_description  \\\n",
       "1  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "3  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "5  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "7  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "9  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "\n",
       "  filetype                                       filename  \\\n",
       "1    fastq  GOMECC18S_Plate4_53_S340_L001_R1_001.fastq.gz   \n",
       "3    fastq  GOMECC18S_Plate4_46_S333_L001_R1_001.fastq.gz   \n",
       "5    fastq  GOMECC18S_Plate4_54_S341_L001_R1_001.fastq.gz   \n",
       "7    fastq  GOMECC18S_Plate4_52_S339_L001_R1_001.fastq.gz   \n",
       "9    fastq  GOMECC18S_Plate4_60_S347_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                       filename2 biosample_accession  \\\n",
       "1  GOMECC18S_Plate4_53_S340_L001_R2_001.fastq.gz        SAMN37516094   \n",
       "3  GOMECC18S_Plate4_46_S333_L001_R2_001.fastq.gz        SAMN37516095   \n",
       "5  GOMECC18S_Plate4_54_S341_L001_R2_001.fastq.gz        SAMN37516096   \n",
       "7  GOMECC18S_Plate4_52_S339_L001_R2_001.fastq.gz        SAMN37516091   \n",
       "9  GOMECC18S_Plate4_60_S347_L001_R2_001.fastq.gz        SAMN37516092   \n",
       "\n",
       "  sra_accession              seq_meth  \\\n",
       "1   SRR26161153  Illumina MiSeq 2x250   \n",
       "3   SRR26161138  Illumina MiSeq 2x250   \n",
       "5   SRR26160919  Illumina MiSeq 2x250   \n",
       "7   SRR26160709  Illumina MiSeq 2x250   \n",
       "9   SRR26160970  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                          nucl_acid_ext  \\\n",
       "1  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "3  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "5  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "7  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "9  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "  amplicon_sequenced target_gene target_subfragment pcr_primer_forward  \\\n",
       "1             18S V9    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "3             18S V9    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "5             18S V9    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "7             18S V9    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "9             18S V9    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "\n",
       "         pcr_primer_reverse pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "1  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "5  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "7  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "9  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "\n",
       "           pcr_primer_reference  \\\n",
       "1  10.1371/journal.pone.0006372   \n",
       "3  10.1371/journal.pone.0006372   \n",
       "5  10.1371/journal.pone.0006372   \n",
       "7  10.1371/journal.pone.0006372   \n",
       "9  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                                                                                              pcr_cond  \\\n",
       "1  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "3  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "5  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "7  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "9  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "\n",
       "                  nucl_acid_amp  \\\n",
       "1  10.1371/journal.pone.0006372   \n",
       "3  10.1371/journal.pone.0006372   \n",
       "5  10.1371/journal.pone.0006372   \n",
       "7  10.1371/journal.pone.0006372   \n",
       "9  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                        adapters            mid_barcode  \n",
       "1  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "3  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "5  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "7  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "9  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  "
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prep_18 = dna_prep[dna_prep['amplicon_sequenced'].str.contains('18S V9')].copy()\n",
    "prep_18['eventID'] = prep_18['eventID']+\"_18S\"\n",
    "prep_18.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_1</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516307</td>\n",
       "      <td>SRR26148474</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_10</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_10_S10_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_10_S10_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516268</td>\n",
       "      <td>SRR26148413</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_11</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_11_S11_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_11_S11_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516264</td>\n",
       "      <td>SRR26148140</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta49_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_12</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_12_S12_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_12_S12_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516246</td>\n",
       "      <td>SRR26148197</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S</td>\n",
       "      <td>GOMECC16S_Plate1_13</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Plate1_13_S13_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Plate1_13_S13_L001_R2_001.fastq.gz</td>\n",
       "      <td>SAMN37516308</td>\n",
       "      <td>SRR26148464</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               eventID           library_id  \\\n",
       "0  GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S   GOMECC16S_Plate1_1   \n",
       "1    GOMECC4_GALVESTON_Sta54_DCM_B_16S  GOMECC16S_Plate1_10   \n",
       "2   GOMECC4_GALVESTON_Sta54_Deep_A_16S  GOMECC16S_Plate1_11   \n",
       "3   GOMECC4_GALVESTON_Sta49_Deep_A_16S  GOMECC16S_Plate1_12   \n",
       "4  GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S  GOMECC16S_Plate1_13   \n",
       "\n",
       "                                                                   title  \\\n",
       "0  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "1  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "2  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "3  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "4  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "\n",
       "  library_strategy library_source library_selection lib_layout  platform  \\\n",
       "0         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "1         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "2         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "3         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "4         AMPLICON    METAGENOMIC               PCR     paired  ILLUMINA   \n",
       "\n",
       "  instrument_model  \\\n",
       "0   Illumina MiSeq   \n",
       "1   Illumina MiSeq   \n",
       "2   Illumina MiSeq   \n",
       "3   Illumina MiSeq   \n",
       "4   Illumina MiSeq   \n",
       "\n",
       "                                                                                                                                      design_description  \\\n",
       "0  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "1  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "2  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "3  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "4  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic b...   \n",
       "\n",
       "  filetype                                      filename  \\\n",
       "0    fastq    GOMECC16S_Plate1_1_S1_L001_R1_001.fastq.gz   \n",
       "1    fastq  GOMECC16S_Plate1_10_S10_L001_R1_001.fastq.gz   \n",
       "2    fastq  GOMECC16S_Plate1_11_S11_L001_R1_001.fastq.gz   \n",
       "3    fastq  GOMECC16S_Plate1_12_S12_L001_R1_001.fastq.gz   \n",
       "4    fastq  GOMECC16S_Plate1_13_S13_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                      filename2 biosample_accession  \\\n",
       "0    GOMECC16S_Plate1_1_S1_L001_R2_001.fastq.gz        SAMN37516307   \n",
       "1  GOMECC16S_Plate1_10_S10_L001_R2_001.fastq.gz        SAMN37516268   \n",
       "2  GOMECC16S_Plate1_11_S11_L001_R2_001.fastq.gz        SAMN37516264   \n",
       "3  GOMECC16S_Plate1_12_S12_L001_R2_001.fastq.gz        SAMN37516246   \n",
       "4  GOMECC16S_Plate1_13_S13_L001_R2_001.fastq.gz        SAMN37516308   \n",
       "\n",
       "  sra_accession              seq_meth  \\\n",
       "0   SRR26148474  Illumina MiSeq 2x250   \n",
       "1   SRR26148413  Illumina MiSeq 2x250   \n",
       "2   SRR26148140  Illumina MiSeq 2x250   \n",
       "3   SRR26148197  Illumina MiSeq 2x250   \n",
       "4   SRR26148464  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                          nucl_acid_ext  \\\n",
       "0  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "1  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "2  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "3  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "4  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "  amplicon_sequenced target_gene target_subfragment   pcr_primer_forward  \\\n",
       "0          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "1          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "2          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "3          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "4          16S V4-V5    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA   \n",
       "\n",
       "     pcr_primer_reverse pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "0  CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "1  CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "2  CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "3  CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "4  CCGYCAATTYMTTTRAGTTT                  515F-Y                    926R   \n",
       "\n",
       "      pcr_primer_reference  \\\n",
       "0  10.1111/1462-2920.13023   \n",
       "1  10.1111/1462-2920.13023   \n",
       "2  10.1111/1462-2920.13023   \n",
       "3  10.1111/1462-2920.13023   \n",
       "4  10.1111/1462-2920.13023   \n",
       "\n",
       "                                                                                                      pcr_cond  \\\n",
       "0  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "1  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "2  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "3  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "4  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "\n",
       "             nucl_acid_amp                                       adapters  \\\n",
       "0  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "1  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "2  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "3  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "4  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "\n",
       "             mid_barcode  \n",
       "0  missing: not provided  \n",
       "1  missing: not provided  \n",
       "2  missing: not provided  \n",
       "3  missing: not provided  \n",
       "4  missing: not provided  "
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_prep = pd.concat([prep_16,prep_18],axis=0,ignore_index=True)\n",
    "dna_prep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge prep and sample\n",
    "dna = dna_sample.merge(dna_prep, how='outer', on='eventID')\n",
    "dna = dna.merge(dna_analysis,how='outer',on='amplicon_sequenced')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_values(['samp_vol_we_dna_ext', 'samp_mat_process', 'env_broad_scale', 'env_local_scale', 'env_medium', 'size_frac', 'concentration', 'concentrationUnit', 'samp_collect_device', 'source_mat_id', 'eventID', 'nucl_acid_ext', 'nucl_acid_amp', 'target_gene', 'target_subfragment', 'lib_layout', 'pcr_primer_forward', 'pcr_primer_reverse', 'pcr_primer_name_forward', 'pcr_primer_name_reverse', 'pcr_primer_reference', 'pcr_cond', 'seq_meth', 'sop', 'ampliconSize', 'otu_class_appr', 'otu_seq_comp_appr', 'otu_db'])"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rename_dict.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['serial_number',\n",
       " 'cruise_id',\n",
       " 'line_id',\n",
       " 'station',\n",
       " 'ctd_bottle_no',\n",
       " 'sample_replicate',\n",
       " 'biological_replicates',\n",
       " 'extract_number',\n",
       " 'sample_title',\n",
       " 'bioproject_accession',\n",
       " 'biosample_accession_x',\n",
       " 'metagenome_sequenced',\n",
       " 'organism',\n",
       " 'collection_date_local',\n",
       " 'collection_date',\n",
       " 'depth',\n",
       " 'geo_loc_name',\n",
       " 'lat_lon',\n",
       " 'decimalLatitude',\n",
       " 'decimalLongitude',\n",
       " 'sample_type',\n",
       " 'collection_method',\n",
       " 'basisOfRecord',\n",
       " 'cluster_16s',\n",
       " 'cluster_18s',\n",
       " 'notes_sampling',\n",
       " 'line_position',\n",
       " 'offshore_inshore_200m_isobath',\n",
       " 'depth_category',\n",
       " 'ocean_acidification_status',\n",
       " 'seascape_class',\n",
       " 'seascape_probability',\n",
       " 'seascape_window',\n",
       " 'dna_sample_number',\n",
       " 'dna_yield',\n",
       " 'extraction_plate_name',\n",
       " 'extraction_well_number',\n",
       " 'extraction_well_position',\n",
       " 'ship_crs_expocode',\n",
       " 'woce_sect',\n",
       " 'ammonium',\n",
       " 'carbonate',\n",
       " 'diss_inorg_carb',\n",
       " 'diss_oxygen',\n",
       " 'fluor',\n",
       " 'hydrogen_ion',\n",
       " 'nitrate',\n",
       " 'nitrite',\n",
       " 'nitrate_plus_nitrite',\n",
       " 'omega_arag',\n",
       " 'pco2',\n",
       " 'ph',\n",
       " 'phosphate',\n",
       " 'pressure',\n",
       " 'salinity',\n",
       " 'samp_store_loc',\n",
       " 'samp_store_temp',\n",
       " 'silicate',\n",
       " 'size_frac_low',\n",
       " 'size_frac_up',\n",
       " 'temp',\n",
       " 'tot_alkalinity',\n",
       " 'tot_depth_water_col',\n",
       " 'transmittance',\n",
       " 'waterBody',\n",
       " 'library_id',\n",
       " 'title',\n",
       " 'library_strategy',\n",
       " 'library_source',\n",
       " 'library_selection',\n",
       " 'platform',\n",
       " 'instrument_model',\n",
       " 'design_description',\n",
       " 'filetype',\n",
       " 'filename',\n",
       " 'filename2',\n",
       " 'biosample_accession_y',\n",
       " 'sra_accession',\n",
       " 'amplicon_sequenced',\n",
       " 'adapters',\n",
       " 'mid_barcode',\n",
       " 'trim_method',\n",
       " 'cluster_method',\n",
       " 'pid_clustering',\n",
       " 'code_repo',\n",
       " 'identificationReferences',\n",
       " 'controls_used']"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#which columns are not in the list of values for dna-derived extension?\n",
    "[col for col in dna if col not in rename_dict.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = dna.drop(columns=[col for col in dna if col not in rename_dict.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>concentration</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2040 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.634 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2080 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>2.307 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2100 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.286 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2000 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.831 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2000 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.849 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    eventID                     source_mat_id  \\\n",
       "939      GOMECC4_CAPECORAL_Sta141_DCM_B_18S      GOMECC4_CAPECORAL_Sta141_DCM   \n",
       "940      GOMECC4_CAPECORAL_Sta141_DCM_C_18S      GOMECC4_CAPECORAL_Sta141_DCM   \n",
       "941  GOMECC4_CAPECORAL_Sta141_Surface_A_18S  GOMECC4_CAPECORAL_Sta141_Surface   \n",
       "942  GOMECC4_CAPECORAL_Sta141_Surface_B_18S  GOMECC4_CAPECORAL_Sta141_Surface   \n",
       "943  GOMECC4_CAPECORAL_Sta141_Surface_C_18S  GOMECC4_CAPECORAL_Sta141_Surface   \n",
       "\n",
       "                  env_broad_scale                     env_local_scale  \\\n",
       "939  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "940  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "941  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "942  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "943  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "\n",
       "                    env_medium samp_vol_we_dna_ext samp_collect_device  \\\n",
       "939  sea water [ENVO:00002149]             2040 ml       Niskin bottle   \n",
       "940  sea water [ENVO:00002149]             2080 ml       Niskin bottle   \n",
       "941  sea water [ENVO:00002149]             2100 ml       Niskin bottle   \n",
       "942  sea water [ENVO:00002149]             2000 ml       Niskin bottle   \n",
       "943  sea water [ENVO:00002149]             2000 ml       Niskin bottle   \n",
       "\n",
       "                                                    samp_mat_process  \\\n",
       "939  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "940  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "941  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "942  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "943  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "\n",
       "    size_frac concentration lib_layout              seq_meth  \\\n",
       "939   0.22 µm   1.634 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "940   0.22 µm   2.307 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "941   0.22 µm   1.286 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "942   0.22 µm   1.831 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "943   0.22 µm   1.849 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                            nucl_acid_ext  \\\n",
       "939  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "940  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "941  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "942  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "943  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "    target_gene target_subfragment pcr_primer_forward  \\\n",
       "939    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "940    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "941    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "942    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "943    18S rRNA                 V9   GTACACACCGCCCGTC   \n",
       "\n",
       "           pcr_primer_reverse pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "939  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "940  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "941  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "942  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "943  TGATCCTTCTGCAGGTTCACCTAC                   1391f                   EukBr   \n",
       "\n",
       "             pcr_primer_reference  \\\n",
       "939  10.1371/journal.pone.0006372   \n",
       "940  10.1371/journal.pone.0006372   \n",
       "941  10.1371/journal.pone.0006372   \n",
       "942  10.1371/journal.pone.0006372   \n",
       "943  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                                                                                                pcr_cond  \\\n",
       "939  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "940  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "941  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "942  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "943  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "\n",
       "                    nucl_acid_amp  ampliconSize  \\\n",
       "939  10.1371/journal.pone.0006372           260   \n",
       "940  10.1371/journal.pone.0006372           260   \n",
       "941  10.1371/journal.pone.0006372           260   \n",
       "942  10.1371/journal.pone.0006372           260   \n",
       "943  10.1371/journal.pone.0006372           260   \n",
       "\n",
       "                                     otu_seq_comp_appr  \\\n",
       "939  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "940  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "941  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "942  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "943  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                        otu_db  \n",
       "939  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706  \n",
       "940  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706  \n",
       "941  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706  \n",
       "942  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706  \n",
       "943  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706  "
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### merge with occurrenceID, DNA_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(944, 25)"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ = dna.merge(occ_all[['eventID','occurrenceID','DNA_sequence']],how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(311390, 27)"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_occ.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>concentration</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9dd2f0b0dd33100b7e</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTTCGTAGGCGTCTTGCCAAGTTGATCGTTAAAGCCACCGGCTTAACCGGTGATCTGCGATCAAAACTGGCGAGATAGAATATGTGAGGGGAATGTGGAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b7f992a6764fa69e97</td>\n",
       "      <td>TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGATGTAGGCGGTCTGTCAAGTCGGATGTGAAATCCCTGGGCTCAACCCAGGAACTGCATTCGAAACTGTCAGACTAGAGTCTCGGAGGGGGTGGCGGAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041c787e5a481888c27e</td>\n",
       "      <td>TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTCCGCAGGCGGACTATTAAGTCAGTGGTGAAAGTCTGCAGCTTAACTGTAGAATTGCCATTGAAACTGATAGTCTTGAGTGTGGTTGAAGTGGGCGGAATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705bdf3f93218ac93a030</td>\n",
       "      <td>TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTGCGTAGGCGTCTACGTAAGTTGTTTGTTAAATCCATCGGCTTAACCGATGATCTGCAAACAAAACTGCATAGATAGAGTTTGGAAGAGGAAAGTGGAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db7812103fe73fdefb7d2b</td>\n",
       "      <td>TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAGCGCAGGTGGCTGAGTTCGTCCGTGGTGCAAGCTCCAGGCCTAACCTGGAGAGGTCTACGGATACTGCTCGGCTTGAGGGCGGTAGAGGAGCACGGAATTC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID          source_mat_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "4  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "\n",
       "                  env_medium samp_vol_we_dna_ext samp_collect_device  \\\n",
       "0  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "1  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "2  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "3  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "4  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "\n",
       "                                                  samp_mat_process size_frac  \\\n",
       "0  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "1  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "2  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "3  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "4  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "\n",
       "   concentration lib_layout              seq_meth  \\\n",
       "0  0.08038 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "1  0.08038 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "2  0.08038 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "3  0.08038 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "4  0.08038 ng/µl     paired  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                          nucl_acid_ext  \\\n",
       "0  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "1  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "2  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "3  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "4  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "  target_gene target_subfragment   pcr_primer_forward    pcr_primer_reverse  \\\n",
       "0    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "1    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "2    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "3    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "4    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_name_reverse     pcr_primer_reference  \\\n",
       "0                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "1                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "2                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "3                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "4                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "\n",
       "                                                                                                      pcr_cond  \\\n",
       "0  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "1  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "2  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "3  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "4  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "\n",
       "             nucl_acid_amp  ampliconSize  \\\n",
       "0  10.1111/1462-2920.13023           411   \n",
       "1  10.1111/1462-2920.13023           411   \n",
       "2  10.1111/1462-2920.13023           411   \n",
       "3  10.1111/1462-2920.13023           411   \n",
       "4  10.1111/1462-2920.13023           411   \n",
       "\n",
       "                                   otu_seq_comp_appr  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                                 otu_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "\n",
       "                                                      occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9dd2f0b0dd33100b7e   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b7f992a6764fa69e97   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041c787e5a481888c27e   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705bdf3f93218ac93a030   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db7812103fe73fdefb7d2b   \n",
       "\n",
       "                                                                                                                                            DNA_sequence  \n",
       "0  TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTTCGTAGGCGTCTTGCCAAGTTGATCGTTAAAGCCACCGGCTTAACCGGTGATCTGCGATCAAAACTGGCGAGATAGAATATGTGAGGGGAATGTGGAATTC...  \n",
       "1  TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGATGTAGGCGGTCTGTCAAGTCGGATGTGAAATCCCTGGGCTCAACCCAGGAACTGCATTCGAAACTGTCAGACTAGAGTCTCGGAGGGGGTGGCGGAATTC...  \n",
       "2  TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTCCGCAGGCGGACTATTAAGTCAGTGGTGAAAGTCTGCAGCTTAACTGTAGAATTGCCATTGAAACTGATAGTCTTGAGTGTGGTTGAAGTGGGCGGAATAT...  \n",
       "3  TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTGCGTAGGCGTCTACGTAAGTTGTTTGTTAAATCCATCGGCTTAACCGATGATCTGCAAACAAAACTGCATAGATAGAGTTTGGAAGAGGAAAGTGGAATTC...  \n",
       "4  TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAGCGCAGGTGGCTGAGTTCGTCCGTGGTGCAAGCTCCAGGCCTAACCTGGAGAGGTCTACGGATACTGCTCGGCTTGAGGGCGGTAGAGGAGCACGGAATTC...  "
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ['concentration'] = dna_occ['concentration'].str.strip(\" ng/µl\")\n",
    "dna_occ['concentrationUnit'] = \"ng/µl\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sop {'AOML_term': 'sop', 'AOML_file': 'analysis_data', 'DwC_definition': 'Standard operating procedures used in assembly and/or annotation of genomes, metagenomes or environmental sequences. Or A reference to a well documented protocol, e.g. using protocols.io', 'Example': nan}\n",
      "otu_class_appr {'AOML_term': 'derived: cluster_method, pid_clustering', 'AOML_file': 'analysis_data', 'DwC_definition': 'Approach/algorithm when defining OTUs or ASVs, include version and parameters separated by semicolons', 'Example': '\"dada2; 1.14.0; ASV\"'}\n"
     ]
    }
   ],
   "source": [
    "# check if all DwC terms are in dna file\n",
    "for key in dna_dict.keys():\n",
    "    if key not in dna_occ.columns:\n",
    "        print(key,dna_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; dada2'"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data']['cluster_method'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ['seq_meth'] = 'Illumina MiSeq 2x250'\n",
    "dna_occ['otu_class_appr']= data['analysis_data']['cluster_method'][0]+\"; \"+data['analysis_data']['pid_clustering'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sop {'AOML_term': 'sop', 'AOML_file': 'analysis_data', 'DwC_definition': 'Standard operating procedures used in assembly and/or annotation of genomes, metagenomes or environmental sequences. Or A reference to a well documented protocol, e.g. using protocols.io', 'Example': nan}\n"
     ]
    }
   ],
   "source": [
    "# check if all DwC terms are in dna file\n",
    "for key in dna_dict.keys():\n",
    "    if key not in dna_occ.columns:\n",
    "        print(key,dna_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>concentration</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>concentrationUnit</th>\n",
       "      <th>otu_class_appr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9dd2f0b0dd33100b7e</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTTCGTAGGCGTCTTGCCAAGTTGATCGTTAAAGCCACCGGCTTAACCGGTGATCTGCGATCAAAACTGGCGAGATAGAATATGTGAGGGGAATGTGGAATTC...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b7f992a6764fa69e97</td>\n",
       "      <td>TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGATGTAGGCGGTCTGTCAAGTCGGATGTGAAATCCCTGGGCTCAACCCAGGAACTGCATTCGAAACTGTCAGACTAGAGTCTCGGAGGGGGTGGCGGAATTC...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041c787e5a481888c27e</td>\n",
       "      <td>TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTCCGCAGGCGGACTATTAAGTCAGTGGTGAAAGTCTGCAGCTTAACTGTAGAATTGCCATTGAAACTGATAGTCTTGAGTGTGGTTGAAGTGGGCGGAATAT...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705bdf3f93218ac93a030</td>\n",
       "      <td>TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTGCGTAGGCGTCTACGTAAGTTGTTTGTTAAATCCATCGGCTTAACCGATGATCTGCAAACAAAACTGCATAGATAGAGTTTGGAAGAGGAAAGTGGAATTC...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>paired</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db7812103fe73fdefb7d2b</td>\n",
       "      <td>TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAGCGCAGGTGGCTGAGTTCGTCCGTGGTGCAAGCTCCAGGCCTAACCTGGAGAGGTCTACGGATACTGCTCGGCTTGAGGGCGGTAGAGGAGCACGGAATTC...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID          source_mat_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "4  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "\n",
       "                  env_medium samp_vol_we_dna_ext samp_collect_device  \\\n",
       "0  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "1  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "2  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "3  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "4  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "\n",
       "                                                  samp_mat_process size_frac  \\\n",
       "0  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "1  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "2  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "3  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "4  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   0.22 µm   \n",
       "\n",
       "  concentration lib_layout              seq_meth  \\\n",
       "0       0.08038     paired  Illumina MiSeq 2x250   \n",
       "1       0.08038     paired  Illumina MiSeq 2x250   \n",
       "2       0.08038     paired  Illumina MiSeq 2x250   \n",
       "3       0.08038     paired  Illumina MiSeq 2x250   \n",
       "4       0.08038     paired  Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                          nucl_acid_ext  \\\n",
       "0  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "1  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "2  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "3  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "4  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "  target_gene target_subfragment   pcr_primer_forward    pcr_primer_reverse  \\\n",
       "0    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "1    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "2    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "3    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "4    16S rRNA              V4-V5  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT   \n",
       "\n",
       "  pcr_primer_name_forward pcr_primer_name_reverse     pcr_primer_reference  \\\n",
       "0                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "1                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "2                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "3                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "4                  515F-Y                    926R  10.1111/1462-2920.13023   \n",
       "\n",
       "                                                                                                      pcr_cond  \\\n",
       "0  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "1  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "2  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "3  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "4  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25   \n",
       "\n",
       "             nucl_acid_amp  ampliconSize  \\\n",
       "0  10.1111/1462-2920.13023           411   \n",
       "1  10.1111/1462-2920.13023           411   \n",
       "2  10.1111/1462-2920.13023           411   \n",
       "3  10.1111/1462-2920.13023           411   \n",
       "4  10.1111/1462-2920.13023           411   \n",
       "\n",
       "                                   otu_seq_comp_appr  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                                 otu_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695   \n",
       "\n",
       "                                                      occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9dd2f0b0dd33100b7e   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b7f992a6764fa69e97   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041c787e5a481888c27e   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705bdf3f93218ac93a030   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db7812103fe73fdefb7d2b   \n",
       "\n",
       "                                                                                                                                            DNA_sequence  \\\n",
       "0  TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTTCGTAGGCGTCTTGCCAAGTTGATCGTTAAAGCCACCGGCTTAACCGGTGATCTGCGATCAAAACTGGCGAGATAGAATATGTGAGGGGAATGTGGAATTC...   \n",
       "1  TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGATGTAGGCGGTCTGTCAAGTCGGATGTGAAATCCCTGGGCTCAACCCAGGAACTGCATTCGAAACTGTCAGACTAGAGTCTCGGAGGGGGTGGCGGAATTC...   \n",
       "2  TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTCCGCAGGCGGACTATTAAGTCAGTGGTGAAAGTCTGCAGCTTAACTGTAGAATTGCCATTGAAACTGATAGTCTTGAGTGTGGTTGAAGTGGGCGGAATAT...   \n",
       "3  TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTGCGTAGGCGTCTACGTAAGTTGTTTGTTAAATCCATCGGCTTAACCGATGATCTGCAAACAAAACTGCATAGATAGAGTTTGGAAGAGGAAAGTGGAATTC...   \n",
       "4  TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAGCGCAGGTGGCTGAGTTCGTCCGTGGTGCAAGCTCCAGGCCTAACCTGGAGAGGTCTACGGATACTGCTCGGCTTGAGGGCGGTAGAGGAGCACGGAATTC...   \n",
       "\n",
       "  concentrationUnit                         otu_class_appr  \n",
       "0             ng/µl  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "1             ng/µl  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "2             ng/µl  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "3             ng/µl  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "4             ng/µl  Tourmaline; qiime2-2021.2; dada2; ASV  "
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ.to_csv(\"../processed/dna-derived.csv\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:edna2obis]",
   "language": "python",
   "name": "conda-env-edna2obis-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
