{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Darwin Core Conversion of eDNA Sequence Data From the AOML_MIMARKS metadata template \n",
    "\n",
    "**Version:** 1.0.0\n",
    "\n",
    "**Author:** Katherine Silliman\n",
    "\n",
    "**Last Updated:** 2-Oct-2023\n",
    "\n",
    "This notebook is for converting a [MIMARKS](https://fairsharing.org/FAIRsharing.zvrep1)-based data sheet to DarwinCore for submission to OBIS. It has been testing on a Mac M1 laptop running in Rosetta mode, with Python 3.11. \n",
    "\n",
    "[Metadata template Google Sheet](https://docs.google.com/spreadsheets/d/1jof9MBEll7Xluu8-_znLRBIP9JpyAd_5YvdioZ-REoY/edit?usp=sharing)\n",
    "\n",
    "**Requirements:**\n",
    "- Python 3\n",
    "- Python 3 packages:\n",
    "    - os\n",
    "- External packages:\n",
    "    - Bio.Entrez from biopython\n",
    "    - numpy\n",
    "    - pandas\n",
    "    - openpyxl\n",
    "    - pyworms\n",
    "    - multiprocess\n",
    "- Custom modules:\n",
    "    - WoRMS_matching\n",
    "\n",
    "**Resources:**\n",
    "- Abarenkov K, Andersson AF, Bissett A, Finstad AG, Fossøy F, Grosjean M, Hope M, Jeppesen TS, Kõljalg U, Lundin D, Nilsson RN, Prager M, Provoost P, Schigel D, Suominen S, Svenningsen C & Frøslev TG (2023) Publishing DNA-derived data through biodiversity data platforms, v1.3. Copenhagen: GBIF Secretariat. https://doi.org/10.35035/doc-vf1a-nr22.https://doi.org/10.35035/doc-vf1a-nr22.\n",
    "- [OBIS manual](https://manual.obis.org/dna_data.html)\n",
    "- [TDWG Darwin Core Occurrence Core](https://dwc.tdwg.org/terms/#occurrence)\n",
    "- [GBIF DNA Derived Data Extension](https://tools.gbif.org/dwca-validator/extension.do?id=http://rs.gbif.org/terms/1.0/DNADerivedData)\n",
    "- https://github.com/iobis/dataset-edna\n",
    "\n",
    "**Citation**  \n",
    "Silliman K, Anderson S, Storo R, Thompson L (2023) A Case Study in Sharing Marine eDNA Metabarcoding Data to OBIS. Biodiversity Information Science and Standards 7: e111048. https://doi.org/10.3897/biss.7.111048\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation\n",
    "```\n",
    "conda create -n edna2obis\n",
    "conda activate edna2obis\n",
    "conda install -c conda-forge notebook\n",
    "conda install -c conda-forge nb_conda_kernels\n",
    "\n",
    "conda install -c conda-forge numpy pandas\n",
    "conda install -c conda-forge openpyxl\n",
    "\n",
    "#worms conversion\n",
    "conda install -c conda-forge pyworms\n",
    "conda install -c conda-forge multiprocess\n",
    "conda install -c conda-forge biopython\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import WoRMS_matching # custom functions for querying WoRMS API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# jupyter notebook parameters\n",
    "pd.set_option('display.max_colwidth', 0)\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that in a Jupyter Notebook, the current directory is always where the .ipynb file is being run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Prepare input data \n",
    "\n",
    "**Project data and metadata**  \n",
    "This workflow assumes that you have your project metadata in an Excel sheet formatted like the template located [here](https://docs.google.com/spreadsheets/d/1jof9MBEll7Xluu8-_znLRBIP9JpyAd_5YvdioZ-REoY/edit?usp=sharing). Instructions for filling out the metadata template are located in the 'Readme' sheet.\n",
    "\n",
    "**eDNA and taxonomy data**  \n",
    "The eDNA data and assigned taxonomy should be in a specific tab-delimited format. ![asv_table format](../images/asv_table.png)\n",
    "\n",
    "This file is generated automatically by [Tourmaline v2023.5+](https://github.com/aomlomics/tourmaline), in X location. If your data was generated with Qiime2 or a previous version of Tourmaline, you can convert the `table.qza`, `taxonomy.qza`, and `repseqs.qza` outputs to the correct format using the `create_asv_seq_taxa_obis.sh` shell script.\n",
    "\n",
    "Example:  \n",
    "\n",
    "``` \n",
    "#Run this with a qiime2 environment. \n",
    "bash create_asv_seq_taxa_obis.sh -f \\\n",
    "../gomecc_v2_raw/table-16S-merge.qza -t ../gomecc_v2_raw/taxonomy-16S-merge.qza -r ../gomecc_v2_raw/repseqs-16S-merge.qza \\\n",
    "-o ../gomecc_v2_raw/gomecc-16S-asv.tsv\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Set configs  \n",
    "\n",
    "Below you can set definitions for parameters used in the code. \n",
    "\n",
    "| Parameter           | Description                                                                                                       | Example                                                                                              |\n",
    "|---------------------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|\n",
    "| `sample_data`       | Name of sheet in project data Excel file with sample data.                                                        | \"water_sample_data\"                                                                                  |\n",
    "| `prep_data`         | Name of sheet in project data Excel file with data about molecular preparation methods.                           | \"amplicon_prep_data\"                                                                                 |\n",
    "| `analysis_data`     | Name of sheet in project data Excel file with data about analysis methods.                                        | \"analysis_data\"                                                                                      |\n",
    "| `study_data`        | Name of sheet in project data Excel file with metadata about the study.                                           | \"study_data\"                                                                                         |\n",
    "| `msmt_metadata`     | Name of sheet in project data Excel file with metadata about additional measurements. Not used in current code.   | \"measurement_metadata\"                                                                               |\n",
    "| `excel_file`        | Path of project data Excel file.                                                                                  | \"../raw/gomecc4_AOML_MIMARKS.survey.water.6.0.xlsx\"                                                  |\n",
    "| `md_excel`          | Path of data dictionary Excel file.                                                                               | \"../raw/gomecc_AOML2DwC standards.xlsx\"                                                              |\n",
    "| `datafiles`         | Python dictionary, where keys are the amplicon names and the values are the paths to the cooresponding ASV table. | {'16S V4-V5': '../raw/gomecc-16S-asv.tsv', '18S V9': '../raw/gomecc-18S-asv.tsv'}                    |\n",
    "| `skip_sample_types` | Python list of sample_type values to skip from OBIS submission, such as controls or blanks.                       | ['mock community','distilled water blank','extraction blank','PCR no-template control','RTSF blank'] |\n",
    "| `skip_columns`      | Python list of columns to ignore when submitting to OBIS.                                                         | ['notes_sampling']                                                                                   |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {}\n",
    "params['sample_data'] = \"water_sample_data\"\n",
    "params['prep_data']= \"amplicon_prep_data\"\n",
    "params['analysis_data'] = \"analysis_data\"\n",
    "params['study_data'] = \"study_data\"\n",
    "params['msmt_metadata'] = \"measurement_metadata\"\n",
    "params['excel_file'] = \"../raw/gomecc4_AOML_MIMARKS.survey.water.6.0.xlsx\"\n",
    "\n",
    "params['datafiles'] = {'16S V4-V5': '../raw/gomecc-16S-asv.tsv',\n",
    "                       '18S V9': '../raw/gomecc-18S-asv.tsv'}\n",
    "\n",
    "params['skip_sample_types'] = ['mock community','distilled water blank','extraction blank','PCR no-template control','RTSF blank']\n",
    "params['skip_columns']= ['notes_sampling']\n",
    "params['md_excel'] = \"../raw/gomecc_AOML2DwC standards.xlsx\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load data\n",
    "\n",
    "Note that in a Jupyter Notebook, the current directory is always where the .ipynb file is being run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load project data Excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data = pd.read_excel(\n",
    "    params['excel_file'], \n",
    "    [params['study_data'],params['sample_data'],params['prep_data'],params['analysis_data'],params['msmt_metadata']],\n",
    "    index_col=None, na_values=[\"\"], comment=\"#\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename keys in data dictionary to a general term\n",
    "data['sample_data'] = data.pop(params['sample_data'])\n",
    "data['prep_data'] = data.pop(params['prep_data'])\n",
    "data['analysis_data'] = data.pop(params['analysis_data'])\n",
    "data['study_data'] = data.pop(params['study_data'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove * from headers (was required for NCBI submission, but no longer needed)\n",
    "data['sample_data'].columns = data['sample_data'].columns.str.replace(\"*\",\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sample_data  \n",
    "Contextual data about the samples collected, such as when it was collected, where it was collected from, what kind of sample it is, and what were the properties of the environment or experimental condition from which the sample was taken. Each row is a distinct sample, or Event. Most of this information is recorded during sample collection. This sheet contains terms from the MIMARKS survey water 6.0 package. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>metagenome_sequenced</th>\n",
       "      <th>organism</th>\n",
       "      <th>collection_date_local</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>depth</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>samp_size</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>collection_method</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>cluster_16s</th>\n",
       "      <th>cluster_18s</th>\n",
       "      <th>notes_sampling</th>\n",
       "      <th>notes_bottle_metadata</th>\n",
       "      <th>line_position</th>\n",
       "      <th>offshore_inshore_200m_isobath</th>\n",
       "      <th>depth_category</th>\n",
       "      <th>ocean_acidification_status</th>\n",
       "      <th>seascape_class</th>\n",
       "      <th>seascape_probability</th>\n",
       "      <th>seascape_window</th>\n",
       "      <th>dna_sample_number</th>\n",
       "      <th>dna_conc</th>\n",
       "      <th>dna_yield</th>\n",
       "      <th>extraction_plate_name</th>\n",
       "      <th>extraction_well_number</th>\n",
       "      <th>extraction_well_position</th>\n",
       "      <th>ship_crs_expocode</th>\n",
       "      <th>woce_sect</th>\n",
       "      <th>ammonium</th>\n",
       "      <th>carbonate</th>\n",
       "      <th>diss_inorg_carb</th>\n",
       "      <th>diss_oxygen</th>\n",
       "      <th>fluor</th>\n",
       "      <th>hydrogen_ion</th>\n",
       "      <th>nitrate</th>\n",
       "      <th>nitrite</th>\n",
       "      <th>nitrate_plus_nitrite</th>\n",
       "      <th>omega_arag</th>\n",
       "      <th>pco2</th>\n",
       "      <th>ph</th>\n",
       "      <th>phosphate</th>\n",
       "      <th>pressure</th>\n",
       "      <th>salinity</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>date_sheet_modified</th>\n",
       "      <th>modified_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516091</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>seawater</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>Cluster 3</td>\n",
       "      <td>Cluster 3</td>\n",
       "      <td>DCM = deep chlorophyl max.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Offshore</td>\n",
       "      <td>offshore</td>\n",
       "      <td>Deep</td>\n",
       "      <td>Low</td>\n",
       "      <td>13</td>\n",
       "      <td>0.507214</td>\n",
       "      <td>8-day</td>\n",
       "      <td>1</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>12.057 ng</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>52</td>\n",
       "      <td>D7</td>\n",
       "      <td>WBTSRHB</td>\n",
       "      <td>RB2103</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>2023-10-03 13:28:31.916</td>\n",
       "      <td>luke.thompson@noaa.gov</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516092</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>1940 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>seawater</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>Cluster 3</td>\n",
       "      <td>Cluster 3</td>\n",
       "      <td>DCM was around 80 m and not well defined.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Offshore</td>\n",
       "      <td>offshore</td>\n",
       "      <td>Deep</td>\n",
       "      <td>Low</td>\n",
       "      <td>13</td>\n",
       "      <td>0.507214</td>\n",
       "      <td>8-day</td>\n",
       "      <td>2</td>\n",
       "      <td>0.1141 ng/µl</td>\n",
       "      <td>17.115 ng</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>60</td>\n",
       "      <td>D8</td>\n",
       "      <td>WBTSRHB</td>\n",
       "      <td>RB2103</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516093</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>618 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>2000 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>seawater</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>Cluster 3</td>\n",
       "      <td>Cluster 3</td>\n",
       "      <td>Surface CTD bottles did not fire correctly; hand niskin bottle used for the surface cast. PM cast.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Offshore</td>\n",
       "      <td>offshore</td>\n",
       "      <td>Deep</td>\n",
       "      <td>Low</td>\n",
       "      <td>13</td>\n",
       "      <td>0.507214</td>\n",
       "      <td>8-day</td>\n",
       "      <td>3</td>\n",
       "      <td>0.07223 ng/µl</td>\n",
       "      <td>10.8345 ng</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>62</td>\n",
       "      <td>F8</td>\n",
       "      <td>WBTSRHB</td>\n",
       "      <td>RB2103</td>\n",
       "      <td>0.25971 µmol/kg</td>\n",
       "      <td>88.434 µmol/kg</td>\n",
       "      <td>2215.45 µmol/kg</td>\n",
       "      <td>129.44 µmol/kg</td>\n",
       "      <td>0.0308</td>\n",
       "      <td>0.0000000142 M</td>\n",
       "      <td>29.3256 µmol/kg</td>\n",
       "      <td>0.00391 µmol/kg</td>\n",
       "      <td>29.3295 µmol/kg</td>\n",
       "      <td>1.168</td>\n",
       "      <td>624 µatm</td>\n",
       "      <td>7.849</td>\n",
       "      <td>1.94489 µmol/kg</td>\n",
       "      <td>623 dbar</td>\n",
       "      <td>34.946 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>1540 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>seawater</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>Cluster 1</td>\n",
       "      <td>Cluster 2</td>\n",
       "      <td>Only enough water for 2 surface replicates.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Offshore</td>\n",
       "      <td>offshore</td>\n",
       "      <td>DCM</td>\n",
       "      <td>High</td>\n",
       "      <td>13</td>\n",
       "      <td>0.507214</td>\n",
       "      <td>8-day</td>\n",
       "      <td>4</td>\n",
       "      <td>1.49 ng/µl</td>\n",
       "      <td>223.5 ng</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>53</td>\n",
       "      <td>E7</td>\n",
       "      <td>WBTSRHB</td>\n",
       "      <td>RB2103</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>PRJNA887898</td>\n",
       "      <td>SAMN37516095</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>planned for FY24</td>\n",
       "      <td>seawater metagenome</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>2021-09-14T07:00</td>\n",
       "      <td>49 m</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997 N 79.618 W</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>1720 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp</td>\n",
       "      <td>seawater</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>Cluster 1</td>\n",
       "      <td>Cluster 2</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>Offshore</td>\n",
       "      <td>offshore</td>\n",
       "      <td>DCM</td>\n",
       "      <td>High</td>\n",
       "      <td>13</td>\n",
       "      <td>0.507214</td>\n",
       "      <td>8-day</td>\n",
       "      <td>5</td>\n",
       "      <td>0.6884 ng/µl</td>\n",
       "      <td>103.26 ng</td>\n",
       "      <td>GOMECC2021_Plate4</td>\n",
       "      <td>46</td>\n",
       "      <td>F6</td>\n",
       "      <td>WBTSRHB</td>\n",
       "      <td>RB2103</td>\n",
       "      <td>0.32968 µmol/kg</td>\n",
       "      <td>229.99 µmol/kg</td>\n",
       "      <td>2033.19 µmol/kg</td>\n",
       "      <td>193.443 µmol/kg</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.0000000094 M</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>0 µmol/kg</td>\n",
       "      <td>3.805</td>\n",
       "      <td>423 µatm</td>\n",
       "      <td>8.027</td>\n",
       "      <td>0.0517 µmol/kg</td>\n",
       "      <td>49 dbar</td>\n",
       "      <td>36.325 psu</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_name serial_number        cruise_id line_id station  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A  GOMECC4_001   GOMECC-4 (2021)  27N     Sta1     \n",
       "1  GOMECC4_27N_Sta1_Deep_B  GOMECC4_002   GOMECC-4 (2021)  27N     Sta1     \n",
       "2  GOMECC4_27N_Sta1_Deep_C  GOMECC4_003   GOMECC-4 (2021)  27N     Sta1     \n",
       "3  GOMECC4_27N_Sta1_DCM_A   GOMECC4_004   GOMECC-4 (2021)  27N     Sta1     \n",
       "4  GOMECC4_27N_Sta1_DCM_B   GOMECC4_005   GOMECC-4 (2021)  27N     Sta1     \n",
       "\n",
       "  ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0  3             A                GOMECC4_27N_Sta1_Deep   \n",
       "1  3             B                GOMECC4_27N_Sta1_Deep   \n",
       "2  3             C                GOMECC4_27N_Sta1_Deep   \n",
       "3  14            A                GOMECC4_27N_Sta1_DCM    \n",
       "4  14            B                GOMECC4_27N_Sta1_DCM    \n",
       "\n",
       "                              biological_replicates extract_number  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C  Plate4_52       \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C  Plate4_60       \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B  Plate4_62       \n",
       "3  GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C    Plate4_53       \n",
       "4  GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C    Plate4_46       \n",
       "\n",
       "                                             sample_title  \\\n",
       "0  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_A   \n",
       "1  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_B   \n",
       "2  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_Deep_C   \n",
       "3  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_A    \n",
       "4  Atlantic Ocean seawater sample GOMECC4_27N_Sta1_DCM_B    \n",
       "\n",
       "  bioproject_accession biosample_accession  amplicon_sequenced  \\\n",
       "0  PRJNA887898          SAMN37516091        16S V4-V5 | 18S V9   \n",
       "1  PRJNA887898          SAMN37516092        16S V4-V5 | 18S V9   \n",
       "2  PRJNA887898          SAMN37516093        16S V4-V5 | 18S V9   \n",
       "3  PRJNA887898          SAMN37516094        16S V4-V5 | 18S V9   \n",
       "4  PRJNA887898          SAMN37516095        16S V4-V5 | 18S V9   \n",
       "\n",
       "  metagenome_sequenced             organism   collection_date_local  \\\n",
       "0  planned for FY24     seawater metagenome  2021-09-14T11:00-04:00   \n",
       "1  planned for FY24     seawater metagenome  2021-09-14T11:00-04:00   \n",
       "2  planned for FY24     seawater metagenome  2021-09-14T11:00-04:00   \n",
       "3  planned for FY24     seawater metagenome  2021-09-14T11:00-04:00   \n",
       "4  planned for FY24     seawater metagenome  2021-09-14T11:00-04:00   \n",
       "\n",
       "    collection_date  depth               env_broad_scale  \\\n",
       "0  2021-09-14T07:00  618 m  marine biome [ENVO:00000447]   \n",
       "1  2021-09-14T07:00  618 m  marine biome [ENVO:00000447]   \n",
       "2  2021-09-14T07:00  618 m  marine biome [ENVO:00000447]   \n",
       "3  2021-09-14T07:00  49 m   marine biome [ENVO:00000447]   \n",
       "4  2021-09-14T07:00  49 m   marine biome [ENVO:00000447]   \n",
       "\n",
       "                           env_local_scale                 env_medium  \\\n",
       "0  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "1  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "2  marine mesopelagic zone [ENVO:00000213]  sea water [ENVO:00002149]   \n",
       "3  marine photic zone [ENVO:00000209]       sea water [ENVO:00002149]   \n",
       "4  marine photic zone [ENVO:00000209]       sea water [ENVO:00002149]   \n",
       "\n",
       "                                  geo_loc_name            lat_lon  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)  26.997 N 79.618 W   \n",
       "\n",
       "  decimalLatitude decimalLongitude samp_vol_we_dna_ext samp_collect_device  \\\n",
       "0  26.997         -79.618           1920 ml             Niskin bottle        \n",
       "1  26.997         -79.618           1940 ml             Niskin bottle        \n",
       "2  26.997         -79.618           2000 ml             Niskin bottle        \n",
       "3  26.997         -79.618           1540 ml             Niskin bottle        \n",
       "4  26.997         -79.618           1720 ml             Niskin bottle        \n",
       "\n",
       "                                                  samp_mat_process  \\\n",
       "0  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "1  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "2  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "3  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "4  Pumped through Sterivex filter (0.22-µm) using peristaltic pµmp   \n",
       "\n",
       "  sample_type  samp_size size_frac collection_method   basisOfRecord  \\\n",
       "0  seawater   NaN         0.22 µm   CTD rosette       MaterialSample   \n",
       "1  seawater   NaN         0.22 µm   CTD rosette       MaterialSample   \n",
       "2  seawater   NaN         0.22 µm   CTD rosette       MaterialSample   \n",
       "3  seawater   NaN         0.22 µm   CTD rosette       MaterialSample   \n",
       "4  seawater   NaN         0.22 µm   CTD rosette       MaterialSample   \n",
       "\n",
       "  cluster_16s cluster_18s  \\\n",
       "0  Cluster 3   Cluster 3    \n",
       "1  Cluster 3   Cluster 3    \n",
       "2  Cluster 3   Cluster 3    \n",
       "3  Cluster 1   Cluster 2    \n",
       "4  Cluster 1   Cluster 2    \n",
       "\n",
       "                                                                                       notes_sampling  \\\n",
       "0  DCM = deep chlorophyl max.                                                                           \n",
       "1  DCM was around 80 m and not well defined.                                                            \n",
       "2  Surface CTD bottles did not fire correctly; hand niskin bottle used for the surface cast. PM cast.   \n",
       "3  Only enough water for 2 surface replicates.                                                          \n",
       "4                                                                                                       \n",
       "\n",
       "  notes_bottle_metadata line_position offshore_inshore_200m_isobath  \\\n",
       "0  NaN                   Offshore      offshore                       \n",
       "1  NaN                   Offshore      offshore                       \n",
       "2  NaN                   Offshore      offshore                       \n",
       "3  NaN                   Offshore      offshore                       \n",
       "4  NaN                   Offshore      offshore                       \n",
       "\n",
       "  depth_category ocean_acidification_status seascape_class  \\\n",
       "0  Deep           Low                        13              \n",
       "1  Deep           Low                        13              \n",
       "2  Deep           Low                        13              \n",
       "3  DCM            High                       13              \n",
       "4  DCM            High                       13              \n",
       "\n",
       "  seascape_probability seascape_window dna_sample_number       dna_conc  \\\n",
       "0  0.507214             8-day           1                 0.08038 ng/µl   \n",
       "1  0.507214             8-day           2                 0.1141 ng/µl    \n",
       "2  0.507214             8-day           3                 0.07223 ng/µl   \n",
       "3  0.507214             8-day           4                 1.49 ng/µl      \n",
       "4  0.507214             8-day           5                 0.6884 ng/µl    \n",
       "\n",
       "    dna_yield extraction_plate_name extraction_well_number  \\\n",
       "0  12.057 ng   GOMECC2021_Plate4     52                      \n",
       "1  17.115 ng   GOMECC2021_Plate4     60                      \n",
       "2  10.8345 ng  GOMECC2021_Plate4     62                      \n",
       "3  223.5 ng    GOMECC2021_Plate4     53                      \n",
       "4  103.26 ng   GOMECC2021_Plate4     46                      \n",
       "\n",
       "  extraction_well_position ship_crs_expocode woce_sect         ammonium  \\\n",
       "0  D7                       WBTSRHB           RB2103    0.25971 µmol/kg   \n",
       "1  D8                       WBTSRHB           RB2103    0.25971 µmol/kg   \n",
       "2  F8                       WBTSRHB           RB2103    0.25971 µmol/kg   \n",
       "3  E7                       WBTSRHB           RB2103    0.32968 µmol/kg   \n",
       "4  F6                       WBTSRHB           RB2103    0.32968 µmol/kg   \n",
       "\n",
       "        carbonate  diss_inorg_carb      diss_oxygen   fluor    hydrogen_ion  \\\n",
       "0  88.434 µmol/kg  2215.45 µmol/kg  129.44 µmol/kg   0.0308  0.0000000142 M   \n",
       "1  88.434 µmol/kg  2215.45 µmol/kg  129.44 µmol/kg   0.0308  0.0000000142 M   \n",
       "2  88.434 µmol/kg  2215.45 µmol/kg  129.44 µmol/kg   0.0308  0.0000000142 M   \n",
       "3  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg  0.036   0.0000000094 M   \n",
       "4  229.99 µmol/kg  2033.19 µmol/kg  193.443 µmol/kg  0.036   0.0000000094 M   \n",
       "\n",
       "           nitrate          nitrite nitrate_plus_nitrite omega_arag      pco2  \\\n",
       "0  29.3256 µmol/kg  0.00391 µmol/kg  29.3295 µmol/kg      1.168      624 µatm   \n",
       "1  29.3256 µmol/kg  0.00391 µmol/kg  29.3295 µmol/kg      1.168      624 µatm   \n",
       "2  29.3256 µmol/kg  0.00391 µmol/kg  29.3295 µmol/kg      1.168      624 µatm   \n",
       "3  0 µmol/kg        0 µmol/kg        0 µmol/kg            3.805      423 µatm   \n",
       "4  0 µmol/kg        0 µmol/kg        0 µmol/kg            3.805      423 µatm   \n",
       "\n",
       "      ph        phosphate  pressure    salinity      samp_store_loc  \\\n",
       "0  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "1  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "2  7.849  1.94489 µmol/kg  623 dbar  34.946 psu  NOAA/AOML Room 248   \n",
       "3  8.027  0.0517 µmol/kg   49 dbar   36.325 psu  NOAA/AOML Room 248   \n",
       "4  8.027  0.0517 µmol/kg   49 dbar   36.325 psu  NOAA/AOML Room 248   \n",
       "\n",
       "  samp_store_temp         silicate  size_frac_low size_frac_up       temp  \\\n",
       "0  -20 °C          20.3569 µmol/kg  no pre-filter  0.22 µm      7.479 °C    \n",
       "1  -20 °C          20.3569 µmol/kg  no pre-filter  0.22 µm      7.479 °C    \n",
       "2  -20 °C          20.3569 µmol/kg  no pre-filter  0.22 µm      7.479 °C    \n",
       "3  -20 °C          1.05635 µmol/kg  no pre-filter  0.22 µm      28.592 °C   \n",
       "4  -20 °C          1.05635 µmol/kg  no pre-filter  0.22 µm      28.592 °C   \n",
       "\n",
       "   tot_alkalinity tot_depth_water_col transmittance     date_sheet_modified  \\\n",
       "0  2318.9 µmol/kg  623 m               4.7221       2023-10-03 13:28:31.916   \n",
       "1  2318.9 µmol/kg  623 m               4.7221       NaT                       \n",
       "2  2318.9 µmol/kg  623 m               4.7221       NaT                       \n",
       "3  2371 µmol/kg    623 m               4.665        NaT                       \n",
       "4  2371 µmol/kg    623 m               4.665        NaT                       \n",
       "\n",
       "              modified_by  \n",
       "0  luke.thompson@noaa.gov  \n",
       "1  NaN                     \n",
       "2  NaN                     \n",
       "3  NaN                     \n",
       "4  NaN                     "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['sample_data'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### prep_data  \n",
    "Contextual data about how the samples were prepared for sequencing. Includes how they were extracted, what amplicon was targeted, how they were sequenced. Each row is a separate sequencing library preparation, distinguished by a unique library_id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "      <th>filename2</th>\n",
       "      <th>drive_location</th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>seq_method</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "      <th>date_sheet_modified</th>\n",
       "      <th>modified_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_NegativeControl_1</td>\n",
       "      <td>GOMECC16S_Neg1</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic bead-based extraction kits (ZymoBIOMICS 96 DNA/RNA MagBead kit). Following bead beating in the cartridges, extractions were finished on an automated KingFisher Flex instrument (Thermo Fisher) in 96-well plates. A two-step PCR approach was used, targeting 16S V4-V5 rRNA with primers 515F (5-GTGYCAGCMGCCGCGGTAA-3) and 926R (5-CCGYCAATTYMTTTRAGTTT-3). Primers were constructed with Fluidigm common oligos CS1 forward (CS1-TS-F: 5-ACACTGACGACATGGTTCTACA-3) and CS2 reverse (CS2-TS-R: 5-TACGGTAGCAGAGACTTGGTCT-3) fused to their 5' ends. PCR products were sent to the Michigan State University Research Technology Support Facility Genomics Core for secondary PCR and sequencing. Secondary PCR used dual-indexed, Illumina-compatible primers, targeting the Fluidigm CS1/CS2 oligomers at the ends of the PCR products. Sequencing runs were performed on an Illumina MiSeq to produce 250+250 nt paired reads.</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC16S_Neg1_S499_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC16S_Neg1_S499_L001_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN37516589</td>\n",
       "      <td>SRR26148505</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>16S V4-V5</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic bead-based extraction kits (ZymoBIOMICS 96 DNA/RNA MagBead kit). Following bead beating in the cartridges, extractions were finished on an automated KingFisher Flex instrument (Thermo Fisher) in 96-well plates. A two-step PCR approach was used, targeting 18S V0 rRNA with primers 1391f; 5’-GTACACACCGCCCGTC-3’ and EukBr; 5’-TGATCCTTCTGCAGGTTCACCTAC-3’. Primers were constructed with Fluidigm common oligos CS1 forward (CS1-TS-F: 5-ACACTGACGACATGGTTCTACA-3) and CS2 reverse (CS2-TS-R: 5-TACGGTAGCAGAGACTTGGTCT-3) fused to their 5' ends. PCR products were sent to the Michigan State University Research Technology Support Facility Genomics Core for secondary PCR and sequencing. Secondary PCR used dual-indexed, Illumina-compatible primers, targeting the Fluidigm CS1/CS2 oligomers at the ends of the PCR products. Sequencing runs were performed on an Illumina MiSeq to produce 250+250 nt paired reads</td>\n",
       "      <td>fastq</td>\n",
       "      <td>GOMECC18S_Plate4_53_S340_L001_R1_001.fastq.gz</td>\n",
       "      <td>GOMECC18S_Plate4_53_S340_L001_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>SRR26161153</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md</td>\n",
       "      <td>18S V9</td>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "      <td>2023-10-03 12:49:23.878</td>\n",
       "      <td>katherine.silliman@noaa.gov</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_name           library_id  \\\n",
       "0  GOMECC4_NegativeControl_1  GOMECC16S_Neg1        \n",
       "1  GOMECC4_27N_Sta1_DCM_A     GOMECC18S_Plate4_53   \n",
       "\n",
       "                                                                   title  \\\n",
       "0  16S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "1  18S amplicon metabarcoding of marine metagenome: Gulf of Mexico (USA)   \n",
       "\n",
       "  library_strategy library_source library_selection lib_layout  platform  \\\n",
       "0  AMPLICON         METAGENOMIC    PCR               paired     ILLUMINA   \n",
       "1  AMPLICON         METAGENOMIC    PCR               paired     ILLUMINA   \n",
       "\n",
       "  instrument_model  \\\n",
       "0  Illumina MiSeq    \n",
       "1  Illumina MiSeq    \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     design_description  \\\n",
       "0  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic bead-based extraction kits (ZymoBIOMICS 96 DNA/RNA MagBead kit). Following bead beating in the cartridges, extractions were finished on an automated KingFisher Flex instrument (Thermo Fisher) in 96-well plates. A two-step PCR approach was used, targeting 16S V4-V5 rRNA with primers 515F (5-GTGYCAGCMGCCGCGGTAA-3) and 926R (5-CCGYCAATTYMTTTRAGTTT-3). Primers were constructed with Fluidigm common oligos CS1 forward (CS1-TS-F: 5-ACACTGACGACATGGTTCTACA-3) and CS2 reverse (CS2-TS-R: 5-TACGGTAGCAGAGACTTGGTCT-3) fused to their 5' ends. PCR products were sent to the Michigan State University Research Technology Support Facility Genomics Core for secondary PCR and sequencing. Secondary PCR used dual-indexed, Illumina-compatible primers, targeting the Fluidigm CS1/CS2 oligomers at the ends of the PCR products. Sequencing runs were performed on an Illumina MiSeq to produce 250+250 nt paired reads.    \n",
       "1  Samples were collected and filtered onto Sterivex 0.22 um cartridge filters. DNA was extracted from Sterivex by adding lysis buffer and magnetic bead-based extraction kits (ZymoBIOMICS 96 DNA/RNA MagBead kit). Following bead beating in the cartridges, extractions were finished on an automated KingFisher Flex instrument (Thermo Fisher) in 96-well plates. A two-step PCR approach was used, targeting 18S V0 rRNA with primers 1391f; 5’-GTACACACCGCCCGTC-3’ and EukBr; 5’-TGATCCTTCTGCAGGTTCACCTAC-3’. Primers were constructed with Fluidigm common oligos CS1 forward (CS1-TS-F: 5-ACACTGACGACATGGTTCTACA-3) and CS2 reverse (CS2-TS-R: 5-TACGGTAGCAGAGACTTGGTCT-3) fused to their 5' ends. PCR products were sent to the Michigan State University Research Technology Support Facility Genomics Core for secondary PCR and sequencing. Secondary PCR used dual-indexed, Illumina-compatible primers, targeting the Fluidigm CS1/CS2 oligomers at the ends of the PCR products. Sequencing runs were performed on an Illumina MiSeq to produce 250+250 nt paired reads   \n",
       "\n",
       "  filetype                                       filename  \\\n",
       "0  fastq    GOMECC16S_Neg1_S499_L001_R1_001.fastq.gz        \n",
       "1  fastq    GOMECC18S_Plate4_53_S340_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                       filename2  drive_location  \\\n",
       "0  GOMECC16S_Neg1_S499_L001_R2_001.fastq.gz      NaN               \n",
       "1  GOMECC18S_Plate4_53_S340_L001_R2_001.fastq.gz NaN               \n",
       "\n",
       "  biosample_accession sra_accession            seq_method  \\\n",
       "0  SAMN37516589        SRR26148505   Illumina MiSeq 2x250   \n",
       "1  SAMN37516094        SRR26161153   Illumina MiSeq 2x250   \n",
       "\n",
       "                                                                          nucl_acid_ext  \\\n",
       "0  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "1  https://github.com/aomlomics/protocols/blob/main/protocol_DNA_extraction_Sterivex.md   \n",
       "\n",
       "  amplicon_sequenced target_gene target_subfragment   pcr_primer_forward  \\\n",
       "0  16S V4-V5          16S rRNA    V4-V5              GTGYCAGCMGCCGCGGTAA   \n",
       "1  18S V9             18S rRNA    V9                 GTACACACCGCCCGTC      \n",
       "\n",
       "         pcr_primer_reverse pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "0  CCGYCAATTYMTTTRAGTTT      515F-Y                  926R                     \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  1391f                   EukBr                    \n",
       "\n",
       "           pcr_primer_reference  \\\n",
       "0  10.1111/1462-2920.13023        \n",
       "1  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                                                                                              pcr_cond  \\\n",
       "0  initial denaturation:95_2;denaturation:95_0.75;annealing:50_0.75;elongation:68_1.5;final elongation:68_5;25           \n",
       "1  initial denaturation:94_3;denaturation:94_0.75;annealing:65_0.25;57_0.5;elongation:72_1.5;final elongation:72_10;35   \n",
       "\n",
       "                  nucl_acid_amp  \\\n",
       "0  10.1111/1462-2920.13023        \n",
       "1  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                        adapters            mid_barcode  \\\n",
       "0  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided   \n",
       "1  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided   \n",
       "\n",
       "      date_sheet_modified                  modified_by  \n",
       "0 NaT                      NaN                          \n",
       "1 2023-10-03 12:49:23.878  katherine.silliman@noaa.gov  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Drop samples with unwanted sample types  \n",
    "\n",
    "Often with eDNA projects, we have control samples that are sequenced along with our survey samples. These can include filtering distilled water, using pure water instead of DNA in a PCR or DNA extraction protocol, or a mock community of known microbial taxa. Controls can help identify and mitigate contaminant DNA in our samples, but are not useful for biodiversity platforms like OBIS. You can select which sample_type values to drop with the `skip_sample_types` parameter."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Drop samples with unwanted sample types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "samps_to_remove = data['sample_data']['sample_type'].isin(params['skip_sample_types'])\n",
    "#data['sample_data'][samps_to_remove]\n",
    "# list of samples to drop\n",
    "samples_to_drop = data['sample_data']['sample_name'][samps_to_remove]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can view the list of samples to be dropped below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26     GOMECC4_Blank_DIW_20210915_A\n",
       "27     GOMECC4_Blank_DIW_20210915_B\n",
       "28     GOMECC4_Blank_DIW_20210915_C\n",
       "200    GOMECC4_Blank_DIW_20210930_A\n",
       "201    GOMECC4_Blank_DIW_20210930_B\n",
       "202    GOMECC4_Blank_DIW_20210930_C\n",
       "334    GOMECC4_Blank_DIW_20211011_A\n",
       "335    GOMECC4_Blank_DIW_20211011_B\n",
       "336    GOMECC4_Blank_DIW_20211011_C\n",
       "409    GOMECC4_Blank_DIW_20211016_A\n",
       "410    GOMECC4_Blank_DIW_20211016_B\n",
       "411    GOMECC4_Blank_DIW_20211016_C\n",
       "484       GOMECC4_ExtractionBlank_1\n",
       "485      GOMECC4_ExtractionBlank_11\n",
       "486      GOMECC4_ExtractionBlank_12\n",
       "487       GOMECC4_ExtractionBlank_3\n",
       "488       GOMECC4_ExtractionBlank_5\n",
       "489       GOMECC4_ExtractionBlank_7\n",
       "490       GOMECC4_ExtractionBlank_9\n",
       "491            GOMECC4_MSUControl_1\n",
       "492            GOMECC4_MSUControl_2\n",
       "493            GOMECC4_MSUControl_3\n",
       "494            GOMECC4_MSUControl_4\n",
       "495            GOMECC4_MSUControl_5\n",
       "496            GOMECC4_MSUControl_6\n",
       "497            GOMECC4_MSUControl_7\n",
       "498       GOMECC4_NegativeControl_1\n",
       "499       GOMECC4_NegativeControl_2\n",
       "500       GOMECC4_PositiveControl_1\n",
       "501       GOMECC4_PositiveControl_2\n",
       "Name: sample_name, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_to_drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove samples from sample_data sheet\n",
    "data['sample_data'] = data['sample_data'][~samps_to_remove]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove samples from prep_data\n",
    "prep_samps_to_remove = data['prep_data']['sample_name'].isin(samples_to_drop)\n",
    "data['prep_data'] = data['prep_data'][~prep_samps_to_remove]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['seawater'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the sample_type values left in your sample_data. We only want seawater.\n",
    "data['sample_data']['sample_type'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Drop columns with all NAs  \n",
    "\n",
    "If your project data file has columns with only NAs, this code will check for those, provide their column headers for verification, then remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which have all NAs?\n",
    "dropped = pd.DataFrame()\n",
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    res = pd.Series(data[sheet].columns[data[sheet].isnull().all(0)],\n",
    "                name=sheet)\n",
    "    dropped=pd.concat([dropped,res],axis=1)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which columns in each sheet have only NA values?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_data</th>\n",
       "      <th>prep_data</th>\n",
       "      <th>analysis_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>samp_size</td>\n",
       "      <td>drive_location</td>\n",
       "      <td>sop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_data       prep_data analysis_data\n",
       "0   samp_size  drive_location           sop"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dropped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you are fine with leaving these columns out, proceed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    data[sheet].dropna(axis=1, how='all',inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's check which columns have missing values in some of the rows. These should be filled in on the Excel sheet with the appropriate term ('not applicable', 'missing', or 'not collected'). Alternatively, you can drop the column if it is not needed for submission to OBIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which columns have missing data (NAs) in some rows\n",
    "some = pd.DataFrame()\n",
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    res = pd.Series(data[sheet].columns[data[sheet].isnull().any()].tolist(),\n",
    "                name=sheet)\n",
    "    some=pd.concat([some,res],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_data</th>\n",
       "      <th>prep_data</th>\n",
       "      <th>analysis_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>notes_bottle_metadata</td>\n",
       "      <td>date_sheet_modified</td>\n",
       "      <td>date_sheet_modified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>date_sheet_modified</td>\n",
       "      <td>modified_by</td>\n",
       "      <td>modified_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>modified_by</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample_data            prep_data        analysis_data\n",
       "0  notes_bottle_metadata  date_sheet_modified  date_sheet_modified\n",
       "1    date_sheet_modified          modified_by          modified_by\n",
       "2            modified_by                  NaN                  NaN"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "some"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop columns with any missing data\n",
    "for sheet in ['sample_data','prep_data','analysis_data']:\n",
    "    data[sheet].dropna(axis=1, how='any',inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load data dictionary Excel file \n",
    "This Excel file is used as a data dictionary for converting between terms used in the project data Excel file and Darwin Core terms for submission to OBIS. Currently, we are only preparing an Occurrence core file and a DNA-derived extension file, with Event information in the Occurrence file. Future versions of this workflow will prepare an extendedMeasurementOrFact file as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in data dictionary excel file\n",
    "dwc_data = pd.read_excel(\n",
    "    params['md_excel'], \n",
    "    ['event','occurrence','dna'],\n",
    "    index_col=0, na_values=[\"\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AOML_term</th>\n",
       "      <th>AOML_file</th>\n",
       "      <th>DwC_definition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DwC_term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>eventID</th>\n",
       "      <td>sample_name</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:eventID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventDate</th>\n",
       "      <td>collection_date_local</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.tdwg.org/terms/#dwc:eventDate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplingProtocol</th>\n",
       "      <td>collection_method</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locationID</th>\n",
       "      <td>station</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLatitude</th>\n",
       "      <td>decimalLatitude</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:Location. Positive values are north of the Equator, negative values are south of it. Legal values lie between -90 and 90, inclusive. https://dwc.tdwg.org/terms/#dwc:decimalLatitude</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              AOML_term    AOML_file  \\\n",
       "DwC_term                                               \n",
       "eventID           sample_name            sample_data   \n",
       "eventDate         collection_date_local  sample_data   \n",
       "samplingProtocol  collection_method      sample_data   \n",
       "locationID        station                sample_data   \n",
       "decimalLatitude   decimalLatitude        sample_data   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                          DwC_definition  \n",
       "DwC_term                                                                                                                                                                                                                                                                                                                                                  \n",
       "eventID           An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:eventID                                                                                                                                                                               \n",
       "eventDate         this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.tdwg.org/terms/#dwc:eventDate                                                                                                                                                           \n",
       "samplingProtocol  The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.                                                                                                                                                                                                                                       \n",
       "locationID        An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.                                                                                                                                                                                                 \n",
       "decimalLatitude   The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:Location. Positive values are north of the Equator, negative values are south of it. Legal values lie between -90 and 90, inclusive. https://dwc.tdwg.org/terms/#dwc:decimalLatitude  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example of a sheet in the data dictionary\n",
    "dwc_data['event'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load ASV data  \n",
    "The ASV data files have one row for each unique amplicon sequence variants (ASVs). There is one ASV file for eacher marker sequences. They contain the ASV DNA sequence, a unique hash identifier the taxonomic assignment for each ASV, the confidence given that assignment by the naive-bayes classifier, and then the number of reads observed in each sample. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| column name    | definition                                                                                                                                                                                                                                                                                                                                                                                              |\n",
    "|----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| featureid      | A hash of the ASV sequence, used as a unique identifier for the ASV.                                                                                                                                                                                                                                                                                                                                    |\n",
    "| sequence       | The DNA sequence of the ASV                                                                                                                                                                                                                                                                                                                                                                             |\n",
    "| taxonomy       | The full taxonomy assigned to an ASV sequence. This string could be formatted in very different ways depending on the reference database used during classification, however it should always be in reverse rank order separated by ;. We provide examples for how to process results from a Silva classifier and the PR2 18S classifier. For other taxonomy formats, the code will need to be adapted. |\n",
    "| Confidence     | This is the confidence score assigned the taxonomic classification with a naive-bayes classifier.                                                                                                                                                                                                                                                                                                       |\n",
    "| sample columns | The next columns each represent a sample (or eventID), and the number of reads for that ASV observed in the sample.                                                                                                                                                                                                                                                                                     |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in ASV tables, looping through amplicons\n",
    "asv_tables = {}\n",
    "\n",
    "for gene in params['datafiles'].keys():\n",
    "    asv_tables[gene] = pd.read_table(params['datafiles'][gene])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['16S V4-V5', '18S V9'])"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTTCAGGTGTAGCGGTGAAATGCGTAGAGATCTGAAGGAATACCGGTGGCGAAGGCGGCCCCCTGGACAGATACTGACACTCAGATGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCTACTTGGAGGTTGTGGCCTTGAGCCGTGGCTTTCGGAGCTAACGCGTTAAGTAGACCGCCTGGGGAGTACGGTCGCAAGATTA</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTTCCGGTGTAGCGGTGAAATGCGTAGATATCGGAAGGAATACCAGTGGCGAAGGCGGCTTACTGGACCAATACTGACGCTGAGGTGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTCGTAGTCCATGCCGTAAACGATGTGTGTTAGACGTTGGAAATTTATTTTCAGTGTCGCAGCGAAAGCAGTAAACACACCGCCTGGGGAGTACGACCGCAAGGTTA</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTCCCAGTGTAGAGGTGAAATGCGTAGATATCGGGAAGAACACCAGTGGCGAAGGCGCTCTGCTGGGCCATCACTGACGCTCATGGACGAAAGCCAGGGGAGCGAAAGGGATTAGATACCCCTGTAGTCCTGGCCGTAAACGATGAACACTAGGTGTCGGGGGAATCGACCCCCTCGGTGTCGTAGCCAACGCGTTAAGTGTTCCGCCTGGGGAGTACGCACGCAAGTGTG</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTCCTGGTGTAGAGGTGAAATTCGTAGATATCAGGAGGAACACCGGTGGCGAAGGCGACCACCTGGACGGTGACTGACGCTGAGGTGCGAAAGCATGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCATGCCGTAAACGATGAGTACTAGGCGCTGCGGGTATTGACCCCTGCGGTGCCGAAGTTAACGCATTAAGTACTCCGCCTGGGAAGTACGGCCGCAAGGTTA</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTCACAGTGTAGCGGTGGAATGCGTAGATATTGTGAAGAATACCAATGGCGAAGGCAGCTTCCTGGTTCTGTACTGACACTGAGGTGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGGTCACTAGCTGTTTGGACTTCGGTCTGAGTGGCTAAGCGAAAGTGATAAGTGACCCACCTGGGGAGTACGTTCGCAAGAATG</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                 sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCATGCAGGTGGTTTGTTAAGTCAGATGTGAAAGCCCGGGGCTCAACCTCGGAATTGCATTTGAAACTGGCAGACTAGAGTACTGTAGAGGGGGGTAGAATTTCAGGTGTAGCGGTGAAATGCGTAGAGATCTGAAGGAATACCGGTGGCGAAGGCGGCCCCCTGGACAGATACTGACACTCAGATGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCTACTTGGAGGTTGTGGCCTTGAGCCGTGGCTTTCGGAGCTAACGCGTTAAGTAGACCGCCTGGGGAGTACGGTCGCAAGATTA    \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGGTATTTAAGTCGGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATCCGAAACTATTTAACTAGAGTATGGGAGAGGTAAGTAGAATTTCCGGTGTAGCGGTGAAATGCGTAGATATCGGAAGGAATACCAGTGGCGAAGGCGGCTTACTGGACCAATACTGACGCTGAGGTGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTCGTAGTCCATGCCGTAAACGATGTGTGTTAGACGTTGGAAATTTATTTTCAGTGTCGCAGCGAAAGCAGTAAACACACCGCCTGGGGAGTACGACCGCAAGGTTA       \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCGCGTAGGCGGTCTTCTAAGTTAGGCGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGCTTAATACTGGAAGACTAGAAAACGGAAGAGGGTAGTGGAATTCCCAGTGTAGAGGTGAAATGCGTAGATATCGGGAAGAACACCAGTGGCGAAGGCGCTCTGCTGGGCCATCACTGACGCTCATGGACGAAAGCCAGGGGAGCGAAAGGGATTAGATACCCCTGTAGTCCTGGCCGTAAACGATGAACACTAGGTGTCGGGGGAATCGACCCCCTCGGTGTCGTAGCCAACGCGTTAAGTGTTCCGCCTGGGGAGTACGCACGCAAGTGTG   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCGCGTAGGCGGCCAAATAAGTCTGATGTGAAGGCCCAGGGCTCAACCCTGGAAGTGCATCGGAAACTGTTTGGCTCGAGTCCCGGAGGGGGTGGTGGAATTCCTGGTGTAGAGGTGAAATTCGTAGATATCAGGAGGAACACCGGTGGCGAAGGCGACCACCTGGACGGTGACTGACGCTGAGGTGCGAAAGCATGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCATGCCGTAAACGATGAGTACTAGGCGCTGCGGGTATTGACCCCTGCGGTGCCGAAGTTAACGCATTAAGTACTCCGCCTGGGAAGTACGGCCGCAAGGTTA   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTTTTAAGTTGGATGTGAAAGCCCTGGGCTCAACCTAGGAACTGCATCCAAAACTAGATGACTAGAGTACGAAAGAGGGAAGTAGAATTCACAGTGTAGCGGTGGAATGCGTAGATATTGTGAAGAATACCAATGGCGAAGGCAGCTTCCTGGTTCTGTACTGACACTGAGGTGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGGTCACTAGCTGTTTGGACTTCGGTCTGAGTGGCTAAGCGAAAGTGATAAGTGACCCACCTGGGGAGTACGTTCGCAAGAATG       \n",
       "\n",
       "                                                                                                             taxonomy  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio                   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__HOC36; f__HOC36; g__HOC36; s__Candidatus_Thioglobus       \n",
       "2  d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Synechococcales; f__Cyanobiaceae; g__Cyanobium_PCC-6307         \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o__Myxococcales; f__Myxococcaceae; g__P3OB-42; s__uncultured_bacterium   \n",
       "4  d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__SAR86_clade; f__SAR86_clade; g__SAR86_clade               \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0  0.978926    0                       0                        \n",
       "1  0.881698    0                       0                        \n",
       "2  0.762793    0                       0                        \n",
       "3  0.997619    0                       0                        \n",
       "4  0.999961    0                       0                        \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C  GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  \\\n",
       "0  0                       0                        0                         \n",
       "1  0                       0                        0                         \n",
       "2  0                       0                        0                         \n",
       "3  0                       0                        0                         \n",
       "4  0                       0                        0                         \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_C  GOMECC4_27N_Sta1_Surface_A  \\\n",
       "0  0                        0                            \n",
       "1  0                        0                            \n",
       "2  0                        0                            \n",
       "3  0                        0                            \n",
       "4  0                        0                            \n",
       "\n",
       "   GOMECC4_27N_Sta1_Surface_B  GOMECC4_27N_Sta4_DCM_A  GOMECC4_27N_Sta4_DCM_B  \\\n",
       "0  0                           0                       0                        \n",
       "1  0                           0                       0                        \n",
       "2  0                           0                       0                        \n",
       "3  0                           0                       0                        \n",
       "4  0                           0                       0                        \n",
       "\n",
       "   GOMECC4_27N_Sta4_DCM_C  GOMECC4_27N_Sta4_Deep_A  GOMECC4_27N_Sta4_Deep_B  \\\n",
       "0  0                       0                        0                         \n",
       "1  0                       0                        0                         \n",
       "2  0                       0                        0                         \n",
       "3  0                       0                        0                         \n",
       "4  0                       0                        0                         \n",
       "\n",
       "   GOMECC4_27N_Sta4_Deep_C  GOMECC4_27N_Sta4_Surface_A  \\\n",
       "0  0                        0                            \n",
       "1  0                        0                            \n",
       "2  0                        0                            \n",
       "3  0                        0                            \n",
       "4  0                        0                            \n",
       "\n",
       "   GOMECC4_27N_Sta4_Surface_B  \n",
       "0  25                          \n",
       "1  0                           \n",
       "2  0                           \n",
       "3  0                           \n",
       "4  0                           "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S V4-V5'].iloc[:,0:20].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert to Occurrence file\n",
    "In order to link the DNA-derived extension metadata to our OBIS occurrence records, we have to use the Occurrence core. An \n",
    "For this data set, a `parentEvent` is a filtered water sample that was DNA extracted from a bigger niskin grab, a sequencing library from that DNA extraction is an `event`, and an `occurrence` is an ASV observed within a library. We will have an an occurence file, a DNA derived data file, and a measurements file.  \n",
    "**Define files**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Sampling event info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AOML_term</th>\n",
       "      <th>AOML_file</th>\n",
       "      <th>DwC_definition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DwC_term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>eventID</th>\n",
       "      <td>sample_name</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:eventID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventDate</th>\n",
       "      <td>collection_date_local</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.tdwg.org/terms/#dwc:eventDate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplingProtocol</th>\n",
       "      <td>collection_method</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locationID</th>\n",
       "      <td>station</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLatitude</th>\n",
       "      <td>decimalLatitude</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:Location. Positive values are north of the Equator, negative values are south of it. Legal values lie between -90 and 90, inclusive. https://dwc.tdwg.org/terms/#dwc:decimalLatitude</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLongitude</th>\n",
       "      <td>decimalLongitude</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The geographic longitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:Location. Positive values are east of the Greenwich Meridian, negative values are west of it. Legal values lie between -180 and 180, inclusive. https://dwc.tdwg.org/list/#dwc_decimalLongitude</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geodeticDatum</th>\n",
       "      <td>none</td>\n",
       "      <td>pipeline</td>\n",
       "      <td>The ellipsoid, geodetic datum, or spatial reference system (SRS) upon which the geographic coordinates given in dwc:decimalLatitude and dwc:decimalLongitude are based.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>countryCode</th>\n",
       "      <td>none</td>\n",
       "      <td>pipeline</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <td>depth</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <td>derived: depth</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>datasetID</th>\n",
       "      <td>project_id_external</td>\n",
       "      <td>study_data</td>\n",
       "      <td>An identifier for the set of data. May be a global unique identifier or an identifier specific to a collection or institution.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>waterBody</th>\n",
       "      <td>derived</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the Getty Thesaurus of Geographic Names.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locality</th>\n",
       "      <td>geo_loc_name</td>\n",
       "      <td>sample_data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventRemarks</th>\n",
       "      <td>derived: controls_used</td>\n",
       "      <td>analysis_data</td>\n",
       "      <td>Comments or notes about the dwc:Event.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   AOML_term      AOML_file  \\\n",
       "DwC_term                                                      \n",
       "eventID               sample_name             sample_data     \n",
       "eventDate             collection_date_local   sample_data     \n",
       "samplingProtocol      collection_method       sample_data     \n",
       "locationID            station                 sample_data     \n",
       "decimalLatitude       decimalLatitude         sample_data     \n",
       "decimalLongitude      decimalLongitude        sample_data     \n",
       "geodeticDatum         none                    pipeline        \n",
       "countryCode           none                    pipeline        \n",
       "minimumDepthInMeters  depth                   sample_data     \n",
       "maximumDepthInMeters  derived: depth          sample_data     \n",
       "datasetID             project_id_external     study_data      \n",
       "waterBody             derived                 sample_data     \n",
       "locality              geo_loc_name            sample_data     \n",
       "eventRemarks          derived: controls_used  analysis_data   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                          DwC_definition  \n",
       "DwC_term                                                                                                                                                                                                                                                                                                                                                                  \n",
       "eventID               An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:eventID                                                                                                                                                                                           \n",
       "eventDate             this is the date-time when the dwc:Event was recorded. Recommended best practice is to use a date that conforms to ISO 8601-1:2019. https://dwc.tdwg.org/terms/#dwc:eventDate                                                                                                                                                                       \n",
       "samplingProtocol      The names of, references to, or descriptions of the methods or protocols used during a dwc:Event.                                                                                                                                                                                                                                                   \n",
       "locationID            An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.                                                                                                                                                                                                             \n",
       "decimalLatitude       The geographic latitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:Location. Positive values are north of the Equator, negative values are south of it. Legal values lie between -90 and 90, inclusive. https://dwc.tdwg.org/terms/#dwc:decimalLatitude              \n",
       "decimalLongitude      The geographic longitude (in decimal degrees, using the spatial reference system given in dwc:geodeticDatum) of the geographic center of a dcterms:Location. Positive values are east of the Greenwich Meridian, negative values are west of it. Legal values lie between -180 and 180, inclusive. https://dwc.tdwg.org/list/#dwc_decimalLongitude  \n",
       "geodeticDatum         The ellipsoid, geodetic datum, or spatial reference system (SRS) upon which the geographic coordinates given in dwc:decimalLatitude and dwc:decimalLongitude are based.                                                                                                                                                                             \n",
       "countryCode           NaN                                                                                                                                                                                                                                                                                                                                                 \n",
       "minimumDepthInMeters  NaN                                                                                                                                                                                                                                                                                                                                                 \n",
       "maximumDepthInMeters  NaN                                                                                                                                                                                                                                                                                                                                                 \n",
       "datasetID             An identifier for the set of data. May be a global unique identifier or an identifier specific to a collection or institution.                                                                                                                                                                                                                      \n",
       "waterBody             The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the Getty Thesaurus of Geographic Names.                                                                                                                                                               \n",
       "locality              NaN                                                                                                                                                                                                                                                                                                                                                 \n",
       "eventRemarks          Comments or notes about the dwc:Event.                                                                                                                                                                                                                                                                                                              "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dwc_data['event']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "event_dict = dwc_data['event'].to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AOML_term': 'sample_name',\n",
       " 'AOML_file': 'sample_data',\n",
       " 'DwC_definition': 'An identifier for the set of information associated with a dwc:Event (something that occurs at a place and time). https://dwc.tdwg.org/terms/#dwc:eventID'}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_dict['eventID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "maximumDepthInMeters {'AOML_term': 'derived: depth', 'AOML_file': 'sample_data', 'DwC_definition': nan}\n",
      "waterBody {'AOML_term': 'derived', 'AOML_file': 'sample_data', 'DwC_definition': 'The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the Getty Thesaurus of Geographic Names.'}\n"
     ]
    }
   ],
   "source": [
    "# check which event terms are not in sample_data sheet\n",
    "for key in event_dict.keys():\n",
    "    if event_dict[key]['AOML_file'] == 'sample_data':\n",
    "        if event_dict[key]['AOML_term'] not in data['sample_data'].columns:\n",
    "            print(key,event_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_f/3hfwkwps2rq9q60vkv4fnd_n9rf1vk/T/ipykernel_16627/2121370784.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Atlantic Ocean' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  data['sample_data'].loc[data['sample_data']['geo_loc_name'].str.contains(\"Atlantic Ocean\"), 'waterBody']= \"Atlantic Ocean\"\n"
     ]
    }
   ],
   "source": [
    "# custom add waterBody\n",
    "\n",
    "data['sample_data'].loc[data['sample_data']['geo_loc_name'].str.contains(\"Atlantic Ocean\"), 'waterBody']= \"Atlantic Ocean\"\n",
    "data['sample_data'].loc[data['sample_data']['geo_loc_name'].str.contains(\"Gulf\"), 'waterBody']= \"Mexico, Gulf of\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change locationID to line_id+station\n",
    "data['sample_data']['station'] = data['sample_data']['line_id']+ \"_\"+data['sample_data']['station'] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename sample_data columns to fit DwC standard\n",
    "gen = (x for x in event_dict.keys() if event_dict[x]['AOML_file'] == 'sample_data')\n",
    "rename_dict = {}\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[event_dict[x]['AOML_term']] = x\n",
    "\n",
    "event_sample = data['sample_data'].rename(columns=rename_dict)\n",
    "event_sample = event_sample.drop(columns=[col for col in event_sample if col not in rename_dict.values()])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add minimumDepthInMeters\n",
    "#remove m in depth\n",
    "event_sample['minimumDepthInMeters'] = event_sample['minimumDepthInMeters'].str.strip(\" m\")\n",
    "event_sample['maximumDepthInMeters'] = event_sample['minimumDepthInMeters']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "3  GOMECC4_27N_Sta1_DCM_A   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "4  GOMECC4_27N_Sta1_DCM_B   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "1  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "2  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "3  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "4  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters  \n",
       "0  618                  \n",
       "1  618                  \n",
       "2  618                  \n",
       "3  49                   \n",
       "4  49                   "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "event_sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add amplicon_sequenced back \n",
    "event_sample['amplicon_sequenced'] = data['sample_data']['amplicon_sequenced']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now add an event for each sequencing library, with replicate water sample as the parentEvent.  \n",
    "\n",
    "**Future Update**: make this a for loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>parentEventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "4  GOMECC4_27N_Sta1_DCM_B_16S   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "1  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "2  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "3  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "4  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters  amplicon_sequenced            parentEventID  \n",
       "0  618                  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_A  \n",
       "1  618                  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_B  \n",
       "2  618                  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_C  \n",
       "3  49                   16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_DCM_A   \n",
       "4  49                   16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_DCM_B   "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "child_data_16S = event_sample[event_sample['amplicon_sequenced'].str.contains('16S V4-V5')].copy()\n",
    "child_data_16S['parentEventID'] = child_data_16S['eventID']\n",
    "child_data_16S['eventID'] = child_data_16S['eventID']+\"_16S\"\n",
    "child_data_16S.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>amplicon_sequenced</th>\n",
       "      <th>parentEventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_18S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>16S V4-V5 | 18S V9</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_18S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_18S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_18S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_18S   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "4  GOMECC4_27N_Sta1_DCM_B_18S   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "1  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "2  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "3  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "4  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters  amplicon_sequenced            parentEventID  \n",
       "0  618                  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_A  \n",
       "1  618                  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_B  \n",
       "2  618                  16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_Deep_C  \n",
       "3  49                   16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_DCM_A   \n",
       "4  49                   16S V4-V5 | 18S V9  GOMECC4_27N_Sta1_DCM_B   "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "child_data_18S = event_sample[event_sample['amplicon_sequenced'].str.contains('18S V9')].copy()\n",
    "child_data_18S['parentEventID'] = child_data_18S['eventID']\n",
    "child_data_18S['eventID'] = child_data_18S['eventID']+\"_18S\"\n",
    "child_data_18S.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is your full event file\n",
    "all_event_data = pd.concat([child_data_16S,child_data_18S],axis=0,ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_event_data = all_event_data.drop(columns=['amplicon_sequenced'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>GOMECC4_MSUControl_5_18S</td>\n",
       "      <td>not applicable_not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>GOMECC4_MSUControl_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>GOMECC4_MSUControl_6_18S</td>\n",
       "      <td>not applicable_not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>GOMECC4_MSUControl_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>GOMECC4_MSUControl_7_18S</td>\n",
       "      <td>not applicable_not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>GOMECC4_MSUControl_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>GOMECC4_NegativeControl_1_18S</td>\n",
       "      <td>not applicable_not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>GOMECC4_NegativeControl_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>GOMECC4_NegativeControl_2_18S</td>\n",
       "      <td>not applicable_not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>GOMECC4_NegativeControl_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            eventID                     locationID  \\\n",
       "997   GOMECC4_MSUControl_5_18S       not applicable_not applicable   \n",
       "998   GOMECC4_MSUControl_6_18S       not applicable_not applicable   \n",
       "999   GOMECC4_MSUControl_7_18S       not applicable_not applicable   \n",
       "1000  GOMECC4_NegativeControl_1_18S  not applicable_not applicable   \n",
       "1001  GOMECC4_NegativeControl_2_18S  not applicable_not applicable   \n",
       "\n",
       "           eventDate minimumDepthInMeters        locality decimalLatitude  \\\n",
       "997   not applicable  not applicable       not applicable  not applicable   \n",
       "998   not applicable  not applicable       not applicable  not applicable   \n",
       "999   not applicable  not applicable       not applicable  not applicable   \n",
       "1000  not applicable  not applicable       not applicable  not applicable   \n",
       "1001  not applicable  not applicable       not applicable  not applicable   \n",
       "\n",
       "     decimalLongitude samplingProtocol waterBody maximumDepthInMeters  \\\n",
       "997   not applicable   not applicable   NaN       not applicable        \n",
       "998   not applicable   not applicable   NaN       not applicable        \n",
       "999   not applicable   not applicable   NaN       not applicable        \n",
       "1000  not applicable   not applicable   NaN       not applicable        \n",
       "1001  not applicable   not applicable   NaN       not applicable        \n",
       "\n",
       "                  parentEventID  \n",
       "997   GOMECC4_MSUControl_5       \n",
       "998   GOMECC4_MSUControl_6       \n",
       "999   GOMECC4_MSUControl_7       \n",
       "1000  GOMECC4_NegativeControl_1  \n",
       "1001  GOMECC4_NegativeControl_2  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_event_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "geodeticDatum {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': 'The ellipsoid, geodetic datum, or spatial reference system (SRS) upon which the geographic coordinates given in dwc:decimalLatitude and dwc:decimalLongitude are based.'}\n",
      "countryCode {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': nan}\n",
      "datasetID {'AOML_term': 'project_id_external', 'AOML_file': 'study_data', 'DwC_definition': 'An identifier for the set of data. May be a global unique identifier or an identifier specific to a collection or institution.'}\n",
      "eventRemarks {'AOML_term': 'derived: controls_used', 'AOML_file': 'analysis_data', 'DwC_definition': 'Comments or notes about the dwc:Event.'}\n"
     ]
    }
   ],
   "source": [
    "for key in event_dict.keys():\n",
    "    if event_dict[key]['AOML_file'] != 'sample_data':\n",
    "        print(key,event_dict[key])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "countryCode, leave blank because it spans multiple countries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "#datasetID\n",
    "all_event_data['datasetID'] = data['study_data']['project_id_external'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "#geodeticDatum\n",
    "all_event_data['geodeticDatum'] = \"WGS84\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>618</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>618</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>2021-09-14T11:00-04:00</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID locationID               eventDate  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S  27N_Sta1   2021-09-14T11:00-04:00   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "4  GOMECC4_27N_Sta1_DCM_B_16S   27N_Sta1   2021-09-14T11:00-04:00   \n",
       "\n",
       "  minimumDepthInMeters                                     locality  \\\n",
       "0  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "1  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "2  618                  USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "3  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "4  49                   USA: Atlantic Ocean, east of Florida (27 N)   \n",
       "\n",
       "  decimalLatitude decimalLongitude samplingProtocol       waterBody  \\\n",
       "0  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "1  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "2  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "3  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "4  26.997         -79.618           CTD rosette      Atlantic Ocean   \n",
       "\n",
       "  maximumDepthInMeters            parentEventID          datasetID  \\\n",
       "0  618                  GOMECC4_27N_Sta1_Deep_A  noaa-aoml-gomecc4   \n",
       "1  618                  GOMECC4_27N_Sta1_Deep_B  noaa-aoml-gomecc4   \n",
       "2  618                  GOMECC4_27N_Sta1_Deep_C  noaa-aoml-gomecc4   \n",
       "3  49                   GOMECC4_27N_Sta1_DCM_A   noaa-aoml-gomecc4   \n",
       "4  49                   GOMECC4_27N_Sta1_DCM_B   noaa-aoml-gomecc4   \n",
       "\n",
       "  geodeticDatum  \n",
       "0  WGS84         \n",
       "1  WGS84         \n",
       "2  WGS84         \n",
       "3  WGS84         \n",
       "4  WGS84         "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_event_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Occurrence file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "occ = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 18S"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### drop unwanted samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_tables['18S V9'] = asv_tables['18S V9'].drop(columns=samples_to_drop,errors='ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta4_Surface_C</th>\n",
       "      <th>GOMECC4_27N_Sta6_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta6_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta6_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta6_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta6_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta6_Deep_C</th>\n",
       "      <th>GOMECC4_27N_Sta6_Surface_A</th>\n",
       "      <th>GOMECC4_27N_Sta6_Surface_B</th>\n",
       "      <th>GOMECC4_27N_Sta6_Surface_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_DCM_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_DCM_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_DCM_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_Deep_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_Deep_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_Deep_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_Surface_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_Surface_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta63_Surface_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_DCM_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_DCM_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_DCM_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_Surface_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_Surface_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_Surface_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_Deep_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_Deep_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta66_Deep_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_DCM_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_DCM_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_DCM_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_Deep_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_Deep_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_Deep_C</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_Surface_A</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_Surface_B</th>\n",
       "      <th>GOMECC4_BROWNSVILLE_Sta71_Surface_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_DCM_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_DCM_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_DCM_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_Deep_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_Deep_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_Deep_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_Surface_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_Surface_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta90_Surface_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_DCM_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_DCM_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_DCM_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_Deep_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_Deep_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_Deep_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_Surface_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_Surface_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta91_Surface_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta93_Deep_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta93_Deep_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta93_Deep_C</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta93_Surface_A</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta93_Surface_B</th>\n",
       "      <th>GOMECC4_CAMPECHE_Sta93_Surface_C</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_DCM_A</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_DCM_B</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_DCM_C</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_Deep_A</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_Deep_B</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_Deep_C</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_Surface_A</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_Surface_B</th>\n",
       "      <th>GOMECC4_CANCUN_Sta117_Surface_C</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_DCM_A</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_DCM_B</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_DCM_C</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_Deep_A</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_Deep_B</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_Deep_C</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_Surface_A</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_Surface_B</th>\n",
       "      <th>GOMECC4_CANCUN_Sta118_Surface_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_DCM_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_DCM_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_DCM_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_Deep_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_Deep_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_Deep_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_Surface_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_Surface_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta131_Surface_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_DCM_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_DCM_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_DCM_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_Deep_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_Deep_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_Deep_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_Surface_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_Surface_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta132_Surface_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_DCM_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_DCM_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_DCM_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_Deep_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_Deep_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_Deep_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_Surface_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_Surface_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta135_Surface_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_DCM_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_DCM_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_DCM_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_Deep_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_Deep_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_Deep_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_Surface_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_Surface_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta140_Surface_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_DCM_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_DCM_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_DCM_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_Deep_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_Deep_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_Deep_C</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_Surface_A</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_Surface_B</th>\n",
       "      <th>GOMECC4_CAPECORAL_Sta141_Surface_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_DCM_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_DCM_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_DCM_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_Deep_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_Deep_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_Deep_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_Surface_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_Surface_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta107_Surface_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_Deep_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_DCM_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_DCM_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_Deep_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_Deep_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_DCM_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_Surface_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_Surface_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta109_Surface_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_DCM_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_DCM_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_DCM_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_Deep_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_Deep_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_Deep_C</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_Surface_A</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_Surface_B</th>\n",
       "      <th>GOMECC4_CATOCHE_Sta115_Surface_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_DCM_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_DCM_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_DCM_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_Deep_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_Deep_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_Deep_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_Surface_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_Surface_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta121_Surface_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_DCM_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_DCM_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_DCM_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_Deep_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_Deep_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_Deep_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_Surface_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_Surface_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta122_Surface_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_DCM_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_DCM_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_DCM_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_Deep_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_Deep_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_Deep_C</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_Surface_A</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_Surface_B</th>\n",
       "      <th>GOMECC4_FLSTRAITS_Sta123_Surface_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_DCM_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_DCM_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_DCM_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_Deep_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_Deep_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_Deep_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_Surface_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_Surface_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta49_Surface_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_DCM_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_DCM_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_DCM_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_Deep_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_Deep_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_Deep_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_Surface_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_Surface_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta50_Surface_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_DCM_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_DCM_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_DCM_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_Deep_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_Deep_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_Deep_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_Surface_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_Surface_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta54_Surface_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_DCM_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_DCM_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_DCM_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_Deep_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_Deep_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_Deep_C</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_Surface_A</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_Surface_B</th>\n",
       "      <th>GOMECC4_GALVESTON_Sta59_Surface_C</th>\n",
       "      <th>GOMECC4_LA_Sta38_DCM_A</th>\n",
       "      <th>GOMECC4_LA_Sta38_DCM_B</th>\n",
       "      <th>GOMECC4_LA_Sta38_DCM_C</th>\n",
       "      <th>GOMECC4_LA_Sta38_Deep_A</th>\n",
       "      <th>GOMECC4_LA_Sta38_Deep_B</th>\n",
       "      <th>GOMECC4_LA_Sta38_Deep_C</th>\n",
       "      <th>GOMECC4_LA_Sta38_Surface_A</th>\n",
       "      <th>GOMECC4_LA_Sta38_Surface_B</th>\n",
       "      <th>GOMECC4_LA_Sta38_Surface_C</th>\n",
       "      <th>GOMECC4_LA_Sta39_DCM_A</th>\n",
       "      <th>GOMECC4_LA_Sta39_DCM_B</th>\n",
       "      <th>GOMECC4_LA_Sta39_DCM_C</th>\n",
       "      <th>GOMECC4_LA_Sta39_Deep_A</th>\n",
       "      <th>GOMECC4_LA_Sta39_Deep_B</th>\n",
       "      <th>GOMECC4_LA_Sta39_Deep_C</th>\n",
       "      <th>GOMECC4_LA_Sta39_Surface_A</th>\n",
       "      <th>GOMECC4_LA_Sta39_Surface_B</th>\n",
       "      <th>GOMECC4_LA_Sta39_Surface_C</th>\n",
       "      <th>GOMECC4_LA_Sta40_DCM_A</th>\n",
       "      <th>GOMECC4_LA_Sta40_DCM_B</th>\n",
       "      <th>GOMECC4_LA_Sta40_DCM_C</th>\n",
       "      <th>GOMECC4_LA_Sta40_Deep_A</th>\n",
       "      <th>GOMECC4_LA_Sta40_Deep_B</th>\n",
       "      <th>GOMECC4_LA_Sta40_Deep_C</th>\n",
       "      <th>GOMECC4_LA_Sta40_Surface_A</th>\n",
       "      <th>GOMECC4_LA_Sta40_Surface_B</th>\n",
       "      <th>GOMECC4_LA_Sta40_Surface_C</th>\n",
       "      <th>GOMECC4_LA_Sta45_DCM_A</th>\n",
       "      <th>GOMECC4_LA_Sta45_DCM_B</th>\n",
       "      <th>GOMECC4_LA_Sta45_DCM_C</th>\n",
       "      <th>GOMECC4_LA_Sta45_Deep_A</th>\n",
       "      <th>GOMECC4_LA_Sta45_Deep_B</th>\n",
       "      <th>GOMECC4_LA_Sta45_Deep_C</th>\n",
       "      <th>GOMECC4_LA_Sta45_Surface_A</th>\n",
       "      <th>GOMECC4_LA_Sta45_Surface_B</th>\n",
       "      <th>GOMECC4_LA_Sta45_Surface_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_DCM_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_DCM_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_DCM_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_Deep_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_Deep_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_Deep_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_Surface_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_Surface_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta94_Surface_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_DCM_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_DCM_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_DCM_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_Deep_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_Deep_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_Deep_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_Surface_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_Surface_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta97_Surface_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_DCM_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_DCM_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_DCM_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_Deep_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_Deep_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_Deep_C</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_Surface_A</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_Surface_B</th>\n",
       "      <th>GOMECC4_MERIDA_Sta98_Surface_C</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_DCM_A</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_DCM_B</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_DCM_C</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_Deep_A</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_Deep_B</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_Deep_C</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_Surface_A</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_Surface_B</th>\n",
       "      <th>GOMECC4_PAISNP_Sta61_Surface_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_DCM_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_DCM_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_DCM_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_Deep_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_Deep_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_Deep_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_Surface_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_Surface_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta19_Surface_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_DCM_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_DCM_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_DCM_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_Deep_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_Deep_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_Deep_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_Surface_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_Surface_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta21_Surface_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_DCM_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_DCM_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_DCM_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_Deep_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_Deep_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_Deep_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_Surface_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_Surface_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta23_Surface_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_DCM_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_DCM_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_DCM_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_Deep_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_Deep_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_Deep_C</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_Surface_A</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_Surface_B</th>\n",
       "      <th>GOMECC4_PANAMACITY_Sta28_Surface_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_DCM_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_DCM_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_DCM_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_Deep_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_Deep_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_Deep_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_Surface_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_Surface_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta31_Surface_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_DCM_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_DCM_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_DCM_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_Deep_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_Deep_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_Deep_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_Surface_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_Surface_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta33_Surface_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_DCM_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_DCM_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_DCM_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_Deep_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_Deep_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_Deep_C</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_Surface_A</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_Surface_B</th>\n",
       "      <th>GOMECC4_PENSACOLA_Sta35_Surface_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_DCM_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_DCM_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_DCM_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_Deep_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_Deep_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_Deep_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_Surface_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_Surface_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta10_Surface_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_DCM_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_DCM_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_DCM_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_Deep_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_Deep_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_Deep_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_Surface_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_Surface_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta13_Surface_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta18_Deep_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta18_Deep_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta18_Deep_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta18_Surface_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta18_Surface_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta18_Surface_C</th>\n",
       "      <th>GOMECC4_TAMPA_Sta9_Sediment_A</th>\n",
       "      <th>GOMECC4_TAMPA_Sta9_Sediment_B</th>\n",
       "      <th>GOMECC4_TAMPA_Sta9_Sediment_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_DCM_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_DCM_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_DCM_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_Deep_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_Deep_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_Deep_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_Surface_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_Surface_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta76_Surface_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_DCM_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_DCM_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_DCM_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_Deep_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_Deep_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_Deep_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_Surface_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_Surface_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta79_Surface_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_DCM_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_DCM_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_DCM_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_Deep_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_Deep_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_Deep_C</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_Surface_A</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_Surface_B</th>\n",
       "      <th>GOMECC4_TAMPICO_Sta82_Surface_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_DCM_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_DCM_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_DCM_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_Deep_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_Deep_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_Deep_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_Surface_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta85_Surface_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_DCM_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_DCM_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_DCM_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_Deep_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_Deep_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_Deep_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_Surface_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_Surface_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta87_Surface_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_DCM_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_DCM_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_DCM_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_Deep_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_Deep_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_Deep_C</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_Surface_A</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_Surface_B</th>\n",
       "      <th>GOMECC4_VERACRUZ_Sta89_Surface_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_DCM_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_DCM_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_DCM_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_Deep_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_Deep_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_Deep_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_Surface_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_Surface_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_Surface_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_DCM_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_DCM_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_DCM_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Deep_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Deep_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Deep_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_DCM_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_DCM_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_DCM_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_Deep_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_Deep_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_Deep_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_Surface_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_Surface_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_Surface_C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>1516</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4257</td>\n",
       "      <td>2005</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9527</td>\n",
       "      <td>1934</td>\n",
       "      <td>2037</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1696</td>\n",
       "      <td>645</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4479</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>149</td>\n",
       "      <td>0</td>\n",
       "      <td>11573</td>\n",
       "      <td>7644</td>\n",
       "      <td>3480</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>3617</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10276</td>\n",
       "      <td>7354</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4055</td>\n",
       "      <td>7992</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>23</td>\n",
       "      <td>38</td>\n",
       "      <td>14</td>\n",
       "      <td>7300</td>\n",
       "      <td>13</td>\n",
       "      <td>6136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>89</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1579</td>\n",
       "      <td>2340</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6956</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2531</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>150</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>308</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>999</td>\n",
       "      <td>0</td>\n",
       "      <td>989</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1524</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>8001</td>\n",
       "      <td>0</td>\n",
       "      <td>2820</td>\n",
       "      <td>4403</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1148</td>\n",
       "      <td>0</td>\n",
       "      <td>618</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>617</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4062</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>8804</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8245</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2776</td>\n",
       "      <td>0</td>\n",
       "      <td>1485</td>\n",
       "      <td>1471</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5763</td>\n",
       "      <td>8484</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>128</td>\n",
       "      <td>41</td>\n",
       "      <td>2342</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>4341</td>\n",
       "      <td>6942</td>\n",
       "      <td>3170</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>36</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>729</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1145</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>928</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1873</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1847</td>\n",
       "      <td>1481</td>\n",
       "      <td>2783</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10052</td>\n",
       "      <td>9663</td>\n",
       "      <td>11791</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1418</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>963</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3440</td>\n",
       "      <td>1978</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1470</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3576</td>\n",
       "      <td>0</td>\n",
       "      <td>804</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>353</td>\n",
       "      <td>0</td>\n",
       "      <td>584</td>\n",
       "      <td>2310</td>\n",
       "      <td>0</td>\n",
       "      <td>13111</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>333</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>571</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1155</td>\n",
       "      <td>4719</td>\n",
       "      <td>194</td>\n",
       "      <td>1451</td>\n",
       "      <td>0</td>\n",
       "      <td>1268</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>53</td>\n",
       "      <td>2294</td>\n",
       "      <td>0</td>\n",
       "      <td>2744</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>111</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>8480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>962</td>\n",
       "      <td>316</td>\n",
       "      <td>548</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>613</td>\n",
       "      <td>561</td>\n",
       "      <td>434</td>\n",
       "      <td>0</td>\n",
       "      <td>397</td>\n",
       "      <td>297</td>\n",
       "      <td>76</td>\n",
       "      <td>916</td>\n",
       "      <td>1444</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>174</td>\n",
       "      <td>186</td>\n",
       "      <td>249</td>\n",
       "      <td>138</td>\n",
       "      <td>117</td>\n",
       "      <td>842</td>\n",
       "      <td>322</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>85</td>\n",
       "      <td>370</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>77</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>326</td>\n",
       "      <td>2506</td>\n",
       "      <td>10</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>380</td>\n",
       "      <td>0</td>\n",
       "      <td>468</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4536</td>\n",
       "      <td>3888</td>\n",
       "      <td>0</td>\n",
       "      <td>655</td>\n",
       "      <td>171</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>539</td>\n",
       "      <td>915</td>\n",
       "      <td>590</td>\n",
       "      <td>1828</td>\n",
       "      <td>203</td>\n",
       "      <td>164</td>\n",
       "      <td>361</td>\n",
       "      <td>148</td>\n",
       "      <td>483</td>\n",
       "      <td>1851</td>\n",
       "      <td>1936</td>\n",
       "      <td>0</td>\n",
       "      <td>223</td>\n",
       "      <td>3289</td>\n",
       "      <td>1215</td>\n",
       "      <td>3992</td>\n",
       "      <td>1958</td>\n",
       "      <td>1898</td>\n",
       "      <td>56</td>\n",
       "      <td>400</td>\n",
       "      <td>786</td>\n",
       "      <td>116</td>\n",
       "      <td>6</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>741</td>\n",
       "      <td>1888</td>\n",
       "      <td>740</td>\n",
       "      <td>88</td>\n",
       "      <td>158</td>\n",
       "      <td>96</td>\n",
       "      <td>1020</td>\n",
       "      <td>305</td>\n",
       "      <td>288</td>\n",
       "      <td>371</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>702</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>380</td>\n",
       "      <td>1461</td>\n",
       "      <td>1283</td>\n",
       "      <td>893</td>\n",
       "      <td>2339</td>\n",
       "      <td>1854</td>\n",
       "      <td>1000</td>\n",
       "      <td>237</td>\n",
       "      <td>99</td>\n",
       "      <td>657</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>163</td>\n",
       "      <td>135</td>\n",
       "      <td>550</td>\n",
       "      <td>2086</td>\n",
       "      <td>3315</td>\n",
       "      <td>1302</td>\n",
       "      <td>2470</td>\n",
       "      <td>30</td>\n",
       "      <td>129</td>\n",
       "      <td>795</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>2966</td>\n",
       "      <td>1438</td>\n",
       "      <td>0</td>\n",
       "      <td>663</td>\n",
       "      <td>273</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2731</td>\n",
       "      <td>2071</td>\n",
       "      <td>2386</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>198</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>238</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>59</td>\n",
       "      <td>1312</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>132</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>266</td>\n",
       "      <td>341</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>396</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>96</td>\n",
       "      <td>293</td>\n",
       "      <td>50</td>\n",
       "      <td>214</td>\n",
       "      <td>399</td>\n",
       "      <td>0</td>\n",
       "      <td>1880</td>\n",
       "      <td>0</td>\n",
       "      <td>679</td>\n",
       "      <td>1002</td>\n",
       "      <td>1082</td>\n",
       "      <td>28</td>\n",
       "      <td>19</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>526</td>\n",
       "      <td>655</td>\n",
       "      <td>412</td>\n",
       "      <td>0</td>\n",
       "      <td>506</td>\n",
       "      <td>827</td>\n",
       "      <td>769</td>\n",
       "      <td>192</td>\n",
       "      <td>1977</td>\n",
       "      <td>2583</td>\n",
       "      <td>2521</td>\n",
       "      <td>736</td>\n",
       "      <td>2091</td>\n",
       "      <td>2944</td>\n",
       "      <td>111</td>\n",
       "      <td>1243</td>\n",
       "      <td>33</td>\n",
       "      <td>244</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>1511</td>\n",
       "      <td>4301</td>\n",
       "      <td>1672</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>31</td>\n",
       "      <td>1019</td>\n",
       "      <td>1736</td>\n",
       "      <td>398</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>37</td>\n",
       "      <td>250</td>\n",
       "      <td>1199</td>\n",
       "      <td>299</td>\n",
       "      <td>192</td>\n",
       "      <td>322</td>\n",
       "      <td>327</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>557</td>\n",
       "      <td>29</td>\n",
       "      <td>1085</td>\n",
       "      <td>1732</td>\n",
       "      <td>3539</td>\n",
       "      <td>1424</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>163</td>\n",
       "      <td>512</td>\n",
       "      <td>2093</td>\n",
       "      <td>37</td>\n",
       "      <td>18</td>\n",
       "      <td>47</td>\n",
       "      <td>5</td>\n",
       "      <td>126</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>888</td>\n",
       "      <td>20</td>\n",
       "      <td>540</td>\n",
       "      <td>885</td>\n",
       "      <td>192</td>\n",
       "      <td>527</td>\n",
       "      <td>496</td>\n",
       "      <td>136</td>\n",
       "      <td>327</td>\n",
       "      <td>1093</td>\n",
       "      <td>190</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>436</td>\n",
       "      <td>10</td>\n",
       "      <td>1367</td>\n",
       "      <td>1708</td>\n",
       "      <td>0</td>\n",
       "      <td>6272</td>\n",
       "      <td>4382</td>\n",
       "      <td>82</td>\n",
       "      <td>210</td>\n",
       "      <td>36</td>\n",
       "      <td>268</td>\n",
       "      <td>3249</td>\n",
       "      <td>3755</td>\n",
       "      <td>524</td>\n",
       "      <td>125</td>\n",
       "      <td>6</td>\n",
       "      <td>494</td>\n",
       "      <td>16</td>\n",
       "      <td>120</td>\n",
       "      <td>9</td>\n",
       "      <td>42</td>\n",
       "      <td>87</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>526</td>\n",
       "      <td>1593</td>\n",
       "      <td>246</td>\n",
       "      <td>2499</td>\n",
       "      <td>1207</td>\n",
       "      <td>1160</td>\n",
       "      <td>553</td>\n",
       "      <td>634</td>\n",
       "      <td>2675</td>\n",
       "      <td>2379</td>\n",
       "      <td>1728</td>\n",
       "      <td>463</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1790</td>\n",
       "      <td>832</td>\n",
       "      <td>1860</td>\n",
       "      <td>64</td>\n",
       "      <td>52</td>\n",
       "      <td>1133</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>808</td>\n",
       "      <td>108</td>\n",
       "      <td>612</td>\n",
       "      <td>595</td>\n",
       "      <td>826</td>\n",
       "      <td>1128</td>\n",
       "      <td>3659</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>992</td>\n",
       "      <td>1878</td>\n",
       "      <td>487</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1534</td>\n",
       "      <td>3186</td>\n",
       "      <td>1040</td>\n",
       "      <td>267</td>\n",
       "      <td>221</td>\n",
       "      <td>1512</td>\n",
       "      <td>0</td>\n",
       "      <td>534</td>\n",
       "      <td>9</td>\n",
       "      <td>134</td>\n",
       "      <td>268</td>\n",
       "      <td>0</td>\n",
       "      <td>274</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>181</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1096</td>\n",
       "      <td>617</td>\n",
       "      <td>196</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>464</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "      <td>62</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1448</td>\n",
       "      <td>335</td>\n",
       "      <td>1997</td>\n",
       "      <td>2398</td>\n",
       "      <td>734</td>\n",
       "      <td>2585</td>\n",
       "      <td>821</td>\n",
       "      <td>120</td>\n",
       "      <td>1047</td>\n",
       "      <td>1096</td>\n",
       "      <td>694</td>\n",
       "      <td>674</td>\n",
       "      <td>112</td>\n",
       "      <td>64</td>\n",
       "      <td>74</td>\n",
       "      <td>2850</td>\n",
       "      <td>2926</td>\n",
       "      <td>2455</td>\n",
       "      <td>0</td>\n",
       "      <td>103</td>\n",
       "      <td>89</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1841</td>\n",
       "      <td>3754</td>\n",
       "      <td>2194</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>580</td>\n",
       "      <td>3094</td>\n",
       "      <td>544</td>\n",
       "      <td>577</td>\n",
       "      <td>0</td>\n",
       "      <td>820</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2826</td>\n",
       "      <td>447</td>\n",
       "      <td>1695</td>\n",
       "      <td>632</td>\n",
       "      <td>1133</td>\n",
       "      <td>1999</td>\n",
       "      <td>1321</td>\n",
       "      <td>417</td>\n",
       "      <td>378</td>\n",
       "      <td>0</td>\n",
       "      <td>150</td>\n",
       "      <td>17</td>\n",
       "      <td>1385</td>\n",
       "      <td>1253</td>\n",
       "      <td>2696</td>\n",
       "      <td>2413</td>\n",
       "      <td>2758</td>\n",
       "      <td>385</td>\n",
       "      <td>410</td>\n",
       "      <td>2987</td>\n",
       "      <td>2746</td>\n",
       "      <td>1562</td>\n",
       "      <td>731</td>\n",
       "      <td>2951</td>\n",
       "      <td>1359</td>\n",
       "      <td>511</td>\n",
       "      <td>2022</td>\n",
       "      <td>0</td>\n",
       "      <td>1412</td>\n",
       "      <td>777</td>\n",
       "      <td>65</td>\n",
       "      <td>287</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1530</td>\n",
       "      <td>0</td>\n",
       "      <td>954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5d4df37251121c08397c6fbc27b06175</td>\n",
       "      <td>GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCCTGGCGGTTAACGCTGCCTAGTTGGCGAAAAGACGACCAAACTGTAGCACTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Sinocalanus;Sinocalanus_sinensis;</td>\n",
       "      <td>0.992300</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>300</td>\n",
       "      <td>0</td>\n",
       "      <td>864</td>\n",
       "      <td>408</td>\n",
       "      <td>286</td>\n",
       "      <td>395</td>\n",
       "      <td>324</td>\n",
       "      <td>1893</td>\n",
       "      <td>1700</td>\n",
       "      <td>5115</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1553</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>384</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>3671</td>\n",
       "      <td>972</td>\n",
       "      <td>215</td>\n",
       "      <td>634</td>\n",
       "      <td>3144</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>135</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>1214</td>\n",
       "      <td>0</td>\n",
       "      <td>1031</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>3492</td>\n",
       "      <td>856</td>\n",
       "      <td>681</td>\n",
       "      <td>455</td>\n",
       "      <td>602</td>\n",
       "      <td>3420</td>\n",
       "      <td>0</td>\n",
       "      <td>566</td>\n",
       "      <td>2563</td>\n",
       "      <td>4019</td>\n",
       "      <td>0</td>\n",
       "      <td>2730</td>\n",
       "      <td>2120</td>\n",
       "      <td>2584</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1347</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>187</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>413</td>\n",
       "      <td>0</td>\n",
       "      <td>217</td>\n",
       "      <td>566</td>\n",
       "      <td>1107</td>\n",
       "      <td>141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>266</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>69</td>\n",
       "      <td>470</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>353</td>\n",
       "      <td>9</td>\n",
       "      <td>1704</td>\n",
       "      <td>392</td>\n",
       "      <td>2487</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>85</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1403</td>\n",
       "      <td>1930</td>\n",
       "      <td>3023</td>\n",
       "      <td>4658</td>\n",
       "      <td>1322</td>\n",
       "      <td>3726</td>\n",
       "      <td>3049</td>\n",
       "      <td>1159</td>\n",
       "      <td>2491</td>\n",
       "      <td>524</td>\n",
       "      <td>557</td>\n",
       "      <td>293</td>\n",
       "      <td>0</td>\n",
       "      <td>356</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>687</td>\n",
       "      <td>315</td>\n",
       "      <td>617</td>\n",
       "      <td>1506</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>324</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>115</td>\n",
       "      <td>132</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>304</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>774</td>\n",
       "      <td>331</td>\n",
       "      <td>1803</td>\n",
       "      <td>21</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>2013</td>\n",
       "      <td>142</td>\n",
       "      <td>1162</td>\n",
       "      <td>2987</td>\n",
       "      <td>6178</td>\n",
       "      <td>2887</td>\n",
       "      <td>2617</td>\n",
       "      <td>3531</td>\n",
       "      <td>1536</td>\n",
       "      <td>1468</td>\n",
       "      <td>2136</td>\n",
       "      <td>1079</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3297</td>\n",
       "      <td>319</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>1523</td>\n",
       "      <td>754</td>\n",
       "      <td>888</td>\n",
       "      <td>48</td>\n",
       "      <td>1630</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1773</td>\n",
       "      <td>11</td>\n",
       "      <td>418</td>\n",
       "      <td>107</td>\n",
       "      <td>295</td>\n",
       "      <td>511</td>\n",
       "      <td>354</td>\n",
       "      <td>104</td>\n",
       "      <td>1393</td>\n",
       "      <td>3355</td>\n",
       "      <td>3011</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>419</td>\n",
       "      <td>5994</td>\n",
       "      <td>5230</td>\n",
       "      <td>4584</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>682</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2169</td>\n",
       "      <td>6181</td>\n",
       "      <td>1781</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1596</td>\n",
       "      <td>0</td>\n",
       "      <td>342</td>\n",
       "      <td>1591</td>\n",
       "      <td>278</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>969</td>\n",
       "      <td>3922</td>\n",
       "      <td>1257</td>\n",
       "      <td>524</td>\n",
       "      <td>428</td>\n",
       "      <td>399</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>355</td>\n",
       "      <td>674</td>\n",
       "      <td>1104</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>955</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>477</td>\n",
       "      <td>0</td>\n",
       "      <td>1335</td>\n",
       "      <td>0</td>\n",
       "      <td>1315</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1848</td>\n",
       "      <td>765</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3010</td>\n",
       "      <td>421</td>\n",
       "      <td>269</td>\n",
       "      <td>387</td>\n",
       "      <td>0</td>\n",
       "      <td>1240</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2364</td>\n",
       "      <td>2408</td>\n",
       "      <td>0</td>\n",
       "      <td>4214</td>\n",
       "      <td>1233</td>\n",
       "      <td>1296</td>\n",
       "      <td>1271</td>\n",
       "      <td>1025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2391</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>842</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "      <td>989</td>\n",
       "      <td>446</td>\n",
       "      <td>825</td>\n",
       "      <td>4441</td>\n",
       "      <td>3916</td>\n",
       "      <td>4151</td>\n",
       "      <td>1088</td>\n",
       "      <td>203</td>\n",
       "      <td>2121</td>\n",
       "      <td>2015</td>\n",
       "      <td>0</td>\n",
       "      <td>356</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f863f671a575c6ab587e8de0190d3335</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCCAGGCGGGTCGCCCTGCCTGGTCTACGGGAAGACGACCAAACTGTAGTGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Paracalanus;Paracalanus_parvus;</td>\n",
       "      <td>0.998393</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>2668</td>\n",
       "      <td>2423</td>\n",
       "      <td>2604</td>\n",
       "      <td>1920</td>\n",
       "      <td>2277</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>466</td>\n",
       "      <td>3362</td>\n",
       "      <td>3622</td>\n",
       "      <td>1765</td>\n",
       "      <td>1249</td>\n",
       "      <td>372</td>\n",
       "      <td>502</td>\n",
       "      <td>3168</td>\n",
       "      <td>1682</td>\n",
       "      <td>523</td>\n",
       "      <td>804</td>\n",
       "      <td>441</td>\n",
       "      <td>2007</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>1472</td>\n",
       "      <td>0</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1298</td>\n",
       "      <td>3681</td>\n",
       "      <td>3210</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>1342</td>\n",
       "      <td>1839</td>\n",
       "      <td>58</td>\n",
       "      <td>366</td>\n",
       "      <td>1409</td>\n",
       "      <td>482</td>\n",
       "      <td>786</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1655</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>472</td>\n",
       "      <td>0</td>\n",
       "      <td>2718</td>\n",
       "      <td>436</td>\n",
       "      <td>282</td>\n",
       "      <td>0</td>\n",
       "      <td>564</td>\n",
       "      <td>1386</td>\n",
       "      <td>751</td>\n",
       "      <td>831</td>\n",
       "      <td>1472</td>\n",
       "      <td>393</td>\n",
       "      <td>752</td>\n",
       "      <td>738</td>\n",
       "      <td>678</td>\n",
       "      <td>138</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>592</td>\n",
       "      <td>254</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>328</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>528</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>406</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1050</td>\n",
       "      <td>499</td>\n",
       "      <td>1427</td>\n",
       "      <td>1143</td>\n",
       "      <td>1051</td>\n",
       "      <td>1381</td>\n",
       "      <td>1635</td>\n",
       "      <td>2096</td>\n",
       "      <td>1389</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2189</td>\n",
       "      <td>166</td>\n",
       "      <td>2287</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1027</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>450</td>\n",
       "      <td>1240</td>\n",
       "      <td>935</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>243</td>\n",
       "      <td>3436</td>\n",
       "      <td>392</td>\n",
       "      <td>671</td>\n",
       "      <td>142</td>\n",
       "      <td>338</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1761</td>\n",
       "      <td>273</td>\n",
       "      <td>1444</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>933</td>\n",
       "      <td>355</td>\n",
       "      <td>1285</td>\n",
       "      <td>1379</td>\n",
       "      <td>1959</td>\n",
       "      <td>1337</td>\n",
       "      <td>410</td>\n",
       "      <td>1778</td>\n",
       "      <td>1360</td>\n",
       "      <td>931</td>\n",
       "      <td>1352</td>\n",
       "      <td>323</td>\n",
       "      <td>62</td>\n",
       "      <td>165</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>444</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>948</td>\n",
       "      <td>6</td>\n",
       "      <td>323</td>\n",
       "      <td>1567</td>\n",
       "      <td>574</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2475</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>836</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>498</td>\n",
       "      <td>214</td>\n",
       "      <td>53</td>\n",
       "      <td>273</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>117</td>\n",
       "      <td>93</td>\n",
       "      <td>106</td>\n",
       "      <td>5592</td>\n",
       "      <td>477</td>\n",
       "      <td>3707</td>\n",
       "      <td>2674</td>\n",
       "      <td>3967</td>\n",
       "      <td>2291</td>\n",
       "      <td>958</td>\n",
       "      <td>2013</td>\n",
       "      <td>1677</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>669</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>477</td>\n",
       "      <td>0</td>\n",
       "      <td>3185</td>\n",
       "      <td>49</td>\n",
       "      <td>1620</td>\n",
       "      <td>943</td>\n",
       "      <td>795</td>\n",
       "      <td>54</td>\n",
       "      <td>2328</td>\n",
       "      <td>821</td>\n",
       "      <td>2364</td>\n",
       "      <td>1353</td>\n",
       "      <td>1305</td>\n",
       "      <td>545</td>\n",
       "      <td>1416</td>\n",
       "      <td>1628</td>\n",
       "      <td>4270</td>\n",
       "      <td>162</td>\n",
       "      <td>114</td>\n",
       "      <td>187</td>\n",
       "      <td>392</td>\n",
       "      <td>42</td>\n",
       "      <td>623</td>\n",
       "      <td>1672</td>\n",
       "      <td>1014</td>\n",
       "      <td>1925</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>648</td>\n",
       "      <td>2448</td>\n",
       "      <td>1580</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>428</td>\n",
       "      <td>372</td>\n",
       "      <td>815</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>991</td>\n",
       "      <td>428</td>\n",
       "      <td>799</td>\n",
       "      <td>0</td>\n",
       "      <td>221</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1592</td>\n",
       "      <td>865</td>\n",
       "      <td>1066</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>30</td>\n",
       "      <td>365</td>\n",
       "      <td>189</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>491</td>\n",
       "      <td>1558</td>\n",
       "      <td>825</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3862</td>\n",
       "      <td>3369</td>\n",
       "      <td>4562</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>800</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1878</td>\n",
       "      <td>1095</td>\n",
       "      <td>1284</td>\n",
       "      <td>1718</td>\n",
       "      <td>1199</td>\n",
       "      <td>1236</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>603</td>\n",
       "      <td>905</td>\n",
       "      <td>0</td>\n",
       "      <td>978</td>\n",
       "      <td>453</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>165</td>\n",
       "      <td>759</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>776</td>\n",
       "      <td>1190</td>\n",
       "      <td>846</td>\n",
       "      <td>1627</td>\n",
       "      <td>281</td>\n",
       "      <td>949</td>\n",
       "      <td>1377</td>\n",
       "      <td>711</td>\n",
       "      <td>677</td>\n",
       "      <td>896</td>\n",
       "      <td>1505</td>\n",
       "      <td>135</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>1164</td>\n",
       "      <td>2272</td>\n",
       "      <td>2208</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1103</td>\n",
       "      <td>157</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>938</td>\n",
       "      <td>305</td>\n",
       "      <td>0</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>275</td>\n",
       "      <td>654</td>\n",
       "      <td>632</td>\n",
       "      <td>1277</td>\n",
       "      <td>1960</td>\n",
       "      <td>1344</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1132</td>\n",
       "      <td>376</td>\n",
       "      <td>1174</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>343</td>\n",
       "      <td>181</td>\n",
       "      <td>74</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>588</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>518</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>22</td>\n",
       "      <td>190</td>\n",
       "      <td>115</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>338</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>234</td>\n",
       "      <td>964</td>\n",
       "      <td>121</td>\n",
       "      <td>129</td>\n",
       "      <td>111</td>\n",
       "      <td>0</td>\n",
       "      <td>196</td>\n",
       "      <td>0</td>\n",
       "      <td>422</td>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>792</td>\n",
       "      <td>6</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "      <td>1217</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>1233</td>\n",
       "      <td>0</td>\n",
       "      <td>1081</td>\n",
       "      <td>540</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>2005</td>\n",
       "      <td>248</td>\n",
       "      <td>1599</td>\n",
       "      <td>704</td>\n",
       "      <td>341</td>\n",
       "      <td>144</td>\n",
       "      <td>133</td>\n",
       "      <td>203</td>\n",
       "      <td>13</td>\n",
       "      <td>42</td>\n",
       "      <td>500</td>\n",
       "      <td>1318</td>\n",
       "      <td>452</td>\n",
       "      <td>477</td>\n",
       "      <td>820</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>118</td>\n",
       "      <td>61</td>\n",
       "      <td>265</td>\n",
       "      <td>480</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>94</td>\n",
       "      <td>3473</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1615</td>\n",
       "      <td>119</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>850</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>382</td>\n",
       "      <td>32</td>\n",
       "      <td>11</td>\n",
       "      <td>204</td>\n",
       "      <td>529</td>\n",
       "      <td>0</td>\n",
       "      <td>334</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>796</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>174</td>\n",
       "      <td>216</td>\n",
       "      <td>328</td>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1931</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>826</td>\n",
       "      <td>855</td>\n",
       "      <td>348</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>157</td>\n",
       "      <td>1383</td>\n",
       "      <td>825</td>\n",
       "      <td>1942</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>287</td>\n",
       "      <td>1768</td>\n",
       "      <td>5452</td>\n",
       "      <td>45</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>544</td>\n",
       "      <td>687</td>\n",
       "      <td>586</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>77</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>40</td>\n",
       "      <td>4651</td>\n",
       "      <td>3589</td>\n",
       "      <td>1438</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>88</td>\n",
       "      <td>0</td>\n",
       "      <td>557</td>\n",
       "      <td>1758</td>\n",
       "      <td>2206</td>\n",
       "      <td>72</td>\n",
       "      <td>957</td>\n",
       "      <td>4</td>\n",
       "      <td>46</td>\n",
       "      <td>546</td>\n",
       "      <td>184</td>\n",
       "      <td>2611</td>\n",
       "      <td>833</td>\n",
       "      <td>77</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>437</td>\n",
       "      <td>334</td>\n",
       "      <td>1678</td>\n",
       "      <td>64</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1558</td>\n",
       "      <td>1700</td>\n",
       "      <td>1669</td>\n",
       "      <td>1849</td>\n",
       "      <td>1217</td>\n",
       "      <td>2387</td>\n",
       "      <td>53</td>\n",
       "      <td>256</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>1030</td>\n",
       "      <td>214</td>\n",
       "      <td>1605</td>\n",
       "      <td>202</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>1854</td>\n",
       "      <td>1139</td>\n",
       "      <td>2005</td>\n",
       "      <td>107</td>\n",
       "      <td>459</td>\n",
       "      <td>70</td>\n",
       "      <td>1632</td>\n",
       "      <td>700</td>\n",
       "      <td>347</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>77</td>\n",
       "      <td>0</td>\n",
       "      <td>374</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>591</td>\n",
       "      <td>852</td>\n",
       "      <td>457</td>\n",
       "      <td>321</td>\n",
       "      <td>647</td>\n",
       "      <td>189</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>413</td>\n",
       "      <td>161</td>\n",
       "      <td>642</td>\n",
       "      <td>966</td>\n",
       "      <td>850</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>280</td>\n",
       "      <td>345</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1055</td>\n",
       "      <td>881</td>\n",
       "      <td>103</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>3</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>520</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1922</td>\n",
       "      <td>1218</td>\n",
       "      <td>2227</td>\n",
       "      <td>759</td>\n",
       "      <td>886</td>\n",
       "      <td>819</td>\n",
       "      <td>440</td>\n",
       "      <td>300</td>\n",
       "      <td>486</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>791</td>\n",
       "      <td>2453</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1520</td>\n",
       "      <td>175</td>\n",
       "      <td>1392</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>253</td>\n",
       "      <td>1240</td>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>1423</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1439</td>\n",
       "      <td>744</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>130</td>\n",
       "      <td>100</td>\n",
       "      <td>529</td>\n",
       "      <td>206</td>\n",
       "      <td>0</td>\n",
       "      <td>1008</td>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>739</td>\n",
       "      <td>1743</td>\n",
       "      <td>526</td>\n",
       "      <td>0</td>\n",
       "      <td>1953</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2107</td>\n",
       "      <td>3962</td>\n",
       "      <td>1627</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "2  5d4df37251121c08397c6fbc27b06175   \n",
       "3  f863f671a575c6ab587e8de0190d3335   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "\n",
       "                                                                                                                           sequence  \\\n",
       "0  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCCTGGCGGATTACTCTGCCTGGCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC    \n",
       "1  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGTAGTCGGATCACTCTGACTGCCTGGCGGGAAGACGACCAAACTGTAGCGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC    \n",
       "2  GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCCTGGCGGTTAACGCTGCCTAGTTGGCGAAAAGACGACCAAACTGTAGCACTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC    \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCCAGGCGGGTCGCCCTGCCTGGTCTACGGGAAGACGACCAAACTGTAGTGTTTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC    \n",
       "4  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAAGATAGTCGCAAGACTACCTTTTCTCCGGAAAGACTTTCAAACTTGAGCGTCTAGAGGAAGTAAAAGTCGTAACAAGGTTTCC   \n",
       "\n",
       "                                                                                                       taxonomy  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Neocalanus;Neocalanus_cristatus;        \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Clausocalanus;Clausocalanus_furcatus;   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Sinocalanus;Sinocalanus_sinensis;       \n",
       "3  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Paracalanus;Paracalanus_parvus;         \n",
       "4  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropoda;Crustacea;Maxillopoda;Acrocalanus;Acrocalanus_sp.;            \n",
       "\n",
       "   Confidence  GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  \\\n",
       "0  0.922099    1516                    0                        \n",
       "1  0.999947    962                     316                      \n",
       "2  0.992300    0                       4                        \n",
       "3  0.998393    0                       0                        \n",
       "4  0.779948    1164                    2272                     \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_C  GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  \\\n",
       "0  0                       6                        0                         \n",
       "1  548                     19                       10                        \n",
       "2  0                       12                       5                         \n",
       "3  0                       0                        0                         \n",
       "4  2208                    2                        0                         \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_C  GOMECC4_27N_Sta1_Surface_A  \\\n",
       "0  0                        0                            \n",
       "1  0                        0                            \n",
       "2  0                        0                            \n",
       "3  0                        0                            \n",
       "4  0                        0                            \n",
       "\n",
       "   GOMECC4_27N_Sta1_Surface_B  GOMECC4_27N_Sta4_DCM_A  GOMECC4_27N_Sta4_DCM_B  \\\n",
       "0  4257                        2005                    0                        \n",
       "1  0                           613                     561                      \n",
       "2  0                           9                       0                        \n",
       "3  0                           0                       0                        \n",
       "4  0                           0                       0                        \n",
       "\n",
       "   GOMECC4_27N_Sta4_DCM_C  GOMECC4_27N_Sta4_Deep_A  GOMECC4_27N_Sta4_Deep_B  \\\n",
       "0  14                      0                        0                         \n",
       "1  434                     0                        397                       \n",
       "2  0                       0                        0                         \n",
       "3  5                       0                        0                         \n",
       "4  0                       0                        2                         \n",
       "\n",
       "   GOMECC4_27N_Sta4_Deep_C  GOMECC4_27N_Sta4_Surface_A  \\\n",
       "0  0                        9527                         \n",
       "1  297                      76                           \n",
       "2  0                        0                            \n",
       "3  0                        92                           \n",
       "4  0                        1103                         \n",
       "\n",
       "   GOMECC4_27N_Sta4_Surface_B  GOMECC4_27N_Sta4_Surface_C  \\\n",
       "0  1934                        2037                         \n",
       "1  916                         1444                         \n",
       "2  300                         0                            \n",
       "3  2668                        2423                         \n",
       "4  157                         0                            \n",
       "\n",
       "   GOMECC4_27N_Sta6_DCM_A  GOMECC4_27N_Sta6_DCM_B  GOMECC4_27N_Sta6_DCM_C  \\\n",
       "0  0                       0                       11                       \n",
       "1  140                     0                       0                        \n",
       "2  864                     408                     286                      \n",
       "3  2604                    1920                    2277                     \n",
       "4  104                     938                     305                      \n",
       "\n",
       "   GOMECC4_27N_Sta6_Deep_A  GOMECC4_27N_Sta6_Deep_B  GOMECC4_27N_Sta6_Deep_C  \\\n",
       "0  0                        0                        1696                      \n",
       "1  174                      186                      249                       \n",
       "2  395                      324                      1893                      \n",
       "3  0                        0                        466                       \n",
       "4  0                        53                       0                         \n",
       "\n",
       "   GOMECC4_27N_Sta6_Surface_A  GOMECC4_27N_Sta6_Surface_B  \\\n",
       "0  645                         0                            \n",
       "1  138                         117                          \n",
       "2  1700                        5115                         \n",
       "3  3362                        3622                         \n",
       "4  275                         654                          \n",
       "\n",
       "   GOMECC4_27N_Sta6_Surface_C  GOMECC4_BROWNSVILLE_Sta63_DCM_A  \\\n",
       "0  0                           0                                 \n",
       "1  842                         322                               \n",
       "2  0                           0                                 \n",
       "3  1765                        1249                              \n",
       "4  632                         1277                              \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta63_DCM_B  GOMECC4_BROWNSVILLE_Sta63_DCM_C  \\\n",
       "0  0                                0                                 \n",
       "1  20                               0                                 \n",
       "2  1553                             251                               \n",
       "3  372                              502                               \n",
       "4  1960                             1344                              \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta63_Deep_A  GOMECC4_BROWNSVILLE_Sta63_Deep_B  \\\n",
       "0  0                                 0                                  \n",
       "1  85                                370                                \n",
       "2  0                                 0                                  \n",
       "3  3168                              1682                               \n",
       "4  0                                 0                                  \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta63_Deep_C  GOMECC4_BROWNSVILLE_Sta63_Surface_A  \\\n",
       "0  4479                              0                                     \n",
       "1  58                                0                                     \n",
       "2  384                               67                                    \n",
       "3  523                               804                                   \n",
       "4  0                                 1132                                  \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta63_Surface_B  GOMECC4_BROWNSVILLE_Sta63_Surface_C  \\\n",
       "0  0                                    0                                     \n",
       "1  0                                    7                                     \n",
       "2  0                                    0                                     \n",
       "3  441                                  2007                                  \n",
       "4  376                                  1174                                  \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta66_DCM_A  GOMECC4_BROWNSVILLE_Sta66_DCM_B  \\\n",
       "0  0                                149                               \n",
       "1  0                                77                                \n",
       "2  0                                43                                \n",
       "3  0                                186                               \n",
       "4  0                                20                                \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta66_DCM_C  GOMECC4_BROWNSVILLE_Sta66_Surface_A  \\\n",
       "0  0                                11573                                 \n",
       "1  0                                95                                    \n",
       "2  3                                0                                     \n",
       "3  0                                1472                                  \n",
       "4  343                              181                                   \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta66_Surface_B  GOMECC4_BROWNSVILLE_Sta66_Surface_C  \\\n",
       "0  7644                                 3480                                  \n",
       "1  326                                  2506                                  \n",
       "2  0                                    0                                     \n",
       "3  0                                    269                                   \n",
       "4  74                                   44                                    \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta66_Deep_A  GOMECC4_BROWNSVILLE_Sta66_Deep_B  \\\n",
       "0  0                                 0                                  \n",
       "1  10                                918                                \n",
       "2  0                                 7                                  \n",
       "3  0                                 0                                  \n",
       "4  0                                 0                                  \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta66_Deep_C  GOMECC4_BROWNSVILLE_Sta71_DCM_A  \\\n",
       "0  0                                 0                                 \n",
       "1  0                                 380                               \n",
       "2  0                                 0                                 \n",
       "3  0                                 0                                 \n",
       "4  0                                 0                                 \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta71_DCM_B  GOMECC4_BROWNSVILLE_Sta71_DCM_C  \\\n",
       "0  0                                0                                 \n",
       "1  0                                468                               \n",
       "2  0                                3                                 \n",
       "3  0                                0                                 \n",
       "4  0                                0                                 \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta71_Deep_A  GOMECC4_BROWNSVILLE_Sta71_Deep_B  \\\n",
       "0  0                                 0                                  \n",
       "1  15                                0                                  \n",
       "2  15                                0                                  \n",
       "3  0                                 0                                  \n",
       "4  0                                 0                                  \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta71_Deep_C  GOMECC4_BROWNSVILLE_Sta71_Surface_A  \\\n",
       "0  0                                 40                                    \n",
       "1  0                                 4536                                  \n",
       "2  0                                 0                                     \n",
       "3  0                                 1298                                  \n",
       "4  0                                 588                                   \n",
       "\n",
       "   GOMECC4_BROWNSVILLE_Sta71_Surface_B  GOMECC4_BROWNSVILLE_Sta71_Surface_C  \\\n",
       "0  0                                    0                                     \n",
       "1  3888                                 0                                     \n",
       "2  0                                    12                                    \n",
       "3  3681                                 3210                                  \n",
       "4  7                                    0                                     \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta90_DCM_A  GOMECC4_CAMPECHE_Sta90_DCM_B  \\\n",
       "0  0                             0                              \n",
       "1  655                           171                            \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  518                           0                              \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta90_DCM_C  GOMECC4_CAMPECHE_Sta90_Deep_A  \\\n",
       "0  0                             0                               \n",
       "1  0                             0                               \n",
       "2  0                             0                               \n",
       "3  0                             0                               \n",
       "4  0                             1                               \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta90_Deep_B  GOMECC4_CAMPECHE_Sta90_Deep_C  \\\n",
       "0  0                              0                               \n",
       "1  0                              8                               \n",
       "2  0                              3                               \n",
       "3  0                              0                               \n",
       "4  0                              3                               \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta90_Surface_A  GOMECC4_CAMPECHE_Sta90_Surface_B  \\\n",
       "0  19                                3617                               \n",
       "1  539                               915                                \n",
       "2  0                                 0                                  \n",
       "3  0                                 44                                 \n",
       "4  22                                190                                \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta90_Surface_C  GOMECC4_CAMPECHE_Sta91_DCM_A  \\\n",
       "0  15                                0                              \n",
       "1  590                               1828                           \n",
       "2  0                                 0                              \n",
       "3  0                                 0                              \n",
       "4  115                               0                              \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta91_DCM_B  GOMECC4_CAMPECHE_Sta91_DCM_C  \\\n",
       "0  0                             0                              \n",
       "1  203                           164                            \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  0                             0                              \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta91_Deep_A  GOMECC4_CAMPECHE_Sta91_Deep_B  \\\n",
       "0  0                              0                               \n",
       "1  361                            148                             \n",
       "2  0                              0                               \n",
       "3  0                              4                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta91_Deep_C  GOMECC4_CAMPECHE_Sta91_Surface_A  \\\n",
       "0  0                              0                                  \n",
       "1  483                            1851                               \n",
       "2  0                              0                                  \n",
       "3  0                              0                                  \n",
       "4  0                              338                                \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta91_Surface_B  GOMECC4_CAMPECHE_Sta91_Surface_C  \\\n",
       "0  10276                             7354                               \n",
       "1  1936                              0                                  \n",
       "2  0                                 0                                  \n",
       "3  33                                1342                               \n",
       "4  0                                 0                                  \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta93_Deep_A  GOMECC4_CAMPECHE_Sta93_Deep_B  \\\n",
       "0  0                              12                              \n",
       "1  223                            3289                            \n",
       "2  9                              3671                            \n",
       "3  1839                           58                              \n",
       "4  234                            964                             \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta93_Deep_C  GOMECC4_CAMPECHE_Sta93_Surface_A  \\\n",
       "0  0                              0                                  \n",
       "1  1215                           3992                               \n",
       "2  972                            215                                \n",
       "3  366                            1409                               \n",
       "4  121                            129                                \n",
       "\n",
       "   GOMECC4_CAMPECHE_Sta93_Surface_B  GOMECC4_CAMPECHE_Sta93_Surface_C  \\\n",
       "0  0                                 0                                  \n",
       "1  1958                              1898                               \n",
       "2  634                               3144                               \n",
       "3  482                               786                                \n",
       "4  111                               0                                  \n",
       "\n",
       "   GOMECC4_CANCUN_Sta117_DCM_A  GOMECC4_CANCUN_Sta117_DCM_B  \\\n",
       "0  0                            0                             \n",
       "1  56                           400                           \n",
       "2  0                            0                             \n",
       "3  0                            0                             \n",
       "4  196                          0                             \n",
       "\n",
       "   GOMECC4_CANCUN_Sta117_DCM_C  GOMECC4_CANCUN_Sta117_Deep_A  \\\n",
       "0  0                            0                              \n",
       "1  786                          116                            \n",
       "2  0                            32                             \n",
       "3  0                            0                              \n",
       "4  422                          18                             \n",
       "\n",
       "   GOMECC4_CANCUN_Sta117_Deep_B  GOMECC4_CANCUN_Sta117_Deep_C  \\\n",
       "0  0                             0                              \n",
       "1  6                             42                             \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  5                             0                              \n",
       "\n",
       "   GOMECC4_CANCUN_Sta117_Surface_A  GOMECC4_CANCUN_Sta117_Surface_B  \\\n",
       "0  0                                4055                              \n",
       "1  0                                0                                 \n",
       "2  0                                0                                 \n",
       "3  0                                1655                              \n",
       "4  0                                792                               \n",
       "\n",
       "   GOMECC4_CANCUN_Sta117_Surface_C  GOMECC4_CANCUN_Sta118_DCM_A  \\\n",
       "0  7992                             20                            \n",
       "1  0                                741                           \n",
       "2  0                                0                             \n",
       "3  0                                0                             \n",
       "4  6                                69                            \n",
       "\n",
       "   GOMECC4_CANCUN_Sta118_DCM_B  GOMECC4_CANCUN_Sta118_DCM_C  \\\n",
       "0  0                            90                            \n",
       "1  1888                         740                           \n",
       "2  0                            59                            \n",
       "3  0                            0                             \n",
       "4  0                            1217                          \n",
       "\n",
       "   GOMECC4_CANCUN_Sta118_Deep_A  GOMECC4_CANCUN_Sta118_Deep_B  \\\n",
       "0  23                            38                             \n",
       "1  88                            158                            \n",
       "2  0                             135                            \n",
       "3  0                             0                              \n",
       "4  20                            0                              \n",
       "\n",
       "   GOMECC4_CANCUN_Sta118_Deep_C  GOMECC4_CANCUN_Sta118_Surface_A  \\\n",
       "0  14                            7300                              \n",
       "1  96                            1020                              \n",
       "2  40                            0                                 \n",
       "3  140                           0                                 \n",
       "4  64                            0                                 \n",
       "\n",
       "   GOMECC4_CANCUN_Sta118_Surface_B  GOMECC4_CANCUN_Sta118_Surface_C  \\\n",
       "0  13                               6136                              \n",
       "1  305                              288                               \n",
       "2  1214                             0                                 \n",
       "3  472                              0                                 \n",
       "4  1233                             0                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta131_DCM_A  GOMECC4_CAPECORAL_Sta131_DCM_B  \\\n",
       "0  0                               0                                \n",
       "1  371                             8                                \n",
       "2  1031                            0                                \n",
       "3  2718                            436                              \n",
       "4  1081                            540                              \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta131_DCM_C  GOMECC4_CAPECORAL_Sta131_Deep_A  \\\n",
       "0  21                              89                                \n",
       "1  0                               702                               \n",
       "2  19                              0                                 \n",
       "3  282                             0                                 \n",
       "4  93                              0                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta131_Deep_B  GOMECC4_CAPECORAL_Sta131_Deep_C  \\\n",
       "0  0                                0                                 \n",
       "1  0                                47                                \n",
       "2  3492                             856                               \n",
       "3  564                              1386                              \n",
       "4  2005                             248                               \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta131_Surface_A  GOMECC4_CAPECORAL_Sta131_Surface_B  \\\n",
       "0  0                                   0                                    \n",
       "1  42                                  0                                    \n",
       "2  681                                 455                                  \n",
       "3  751                                 831                                  \n",
       "4  1599                                704                                  \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta131_Surface_C  GOMECC4_CAPECORAL_Sta132_DCM_A  \\\n",
       "0  0                                   0                                \n",
       "1  380                                 1461                             \n",
       "2  602                                 3420                             \n",
       "3  1472                                393                              \n",
       "4  341                                 144                              \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta132_DCM_B  GOMECC4_CAPECORAL_Sta132_DCM_C  \\\n",
       "0  0                               0                                \n",
       "1  1283                            893                              \n",
       "2  0                               566                              \n",
       "3  752                             738                              \n",
       "4  133                             203                              \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta132_Deep_A  GOMECC4_CAPECORAL_Sta132_Deep_B  \\\n",
       "0  0                                0                                 \n",
       "1  2339                             1854                              \n",
       "2  2563                             4019                              \n",
       "3  678                              138                               \n",
       "4  13                               42                                \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta132_Deep_C  GOMECC4_CAPECORAL_Sta132_Surface_A  \\\n",
       "0  0                                0                                    \n",
       "1  1000                             237                                  \n",
       "2  0                                2730                                 \n",
       "3  0                                0                                    \n",
       "4  500                              1318                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta132_Surface_B  GOMECC4_CAPECORAL_Sta132_Surface_C  \\\n",
       "0  0                                   0                                    \n",
       "1  99                                  657                                  \n",
       "2  2120                                2584                                 \n",
       "3  592                                 254                                  \n",
       "4  452                                 477                                  \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta135_DCM_A  GOMECC4_CAPECORAL_Sta135_DCM_B  \\\n",
       "0  0                               0                                \n",
       "1  17                              0                                \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  820                             0                                \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta135_DCM_C  GOMECC4_CAPECORAL_Sta135_Deep_A  \\\n",
       "0  0                               0                                 \n",
       "1  163                             135                               \n",
       "2  0                               0                                 \n",
       "3  0                               0                                 \n",
       "4  0                               0                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta135_Deep_B  GOMECC4_CAPECORAL_Sta135_Deep_C  \\\n",
       "0  0                                0                                 \n",
       "1  550                              2086                              \n",
       "2  0                                0                                 \n",
       "3  0                                0                                 \n",
       "4  0                                0                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta135_Surface_A  GOMECC4_CAPECORAL_Sta135_Surface_B  \\\n",
       "0  0                                   1061                                 \n",
       "1  3315                                1302                                 \n",
       "2  0                                   0                                    \n",
       "3  0                                   0                                    \n",
       "4  118                                 61                                   \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta135_Surface_C  GOMECC4_CAPECORAL_Sta140_DCM_A  \\\n",
       "0  0                                   0                                \n",
       "1  2470                                30                               \n",
       "2  0                                   0                                \n",
       "3  0                                   0                                \n",
       "4  265                                 480                              \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta140_DCM_B  GOMECC4_CAPECORAL_Sta140_DCM_C  \\\n",
       "0  0                               0                                \n",
       "1  129                             795                              \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  0                               0                                \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta140_Deep_A  GOMECC4_CAPECORAL_Sta140_Deep_B  \\\n",
       "0  0                                0                                 \n",
       "1  0                                13                                \n",
       "2  11                               0                                 \n",
       "3  31                               60                                \n",
       "4  45                               0                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta140_Deep_C  GOMECC4_CAPECORAL_Sta140_Surface_A  \\\n",
       "0  0                                1579                                 \n",
       "1  2966                             1438                                 \n",
       "2  0                                0                                    \n",
       "3  0                                0                                    \n",
       "4  0                                0                                    \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta140_Surface_B  GOMECC4_CAPECORAL_Sta140_Surface_C  \\\n",
       "0  2340                                0                                    \n",
       "1  0                                   663                                  \n",
       "2  1347                                0                                    \n",
       "3  328                                 0                                    \n",
       "4  34                                  94                                   \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta141_DCM_A  GOMECC4_CAPECORAL_Sta141_DCM_B  \\\n",
       "0  0                               0                                \n",
       "1  273                             0                                \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  3473                            0                                \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta141_DCM_C  GOMECC4_CAPECORAL_Sta141_Deep_A  \\\n",
       "0  6956                            30                                \n",
       "1  19                              0                                 \n",
       "2  0                               0                                 \n",
       "3  0                               0                                 \n",
       "4  0                               7                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta141_Deep_B  GOMECC4_CAPECORAL_Sta141_Deep_C  \\\n",
       "0  0                                0                                 \n",
       "1  0                                0                                 \n",
       "2  0                                0                                 \n",
       "3  0                                0                                 \n",
       "4  0                                0                                 \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta141_Surface_A  GOMECC4_CAPECORAL_Sta141_Surface_B  \\\n",
       "0  0                                   0                                    \n",
       "1  2731                                2071                                 \n",
       "2  0                                   0                                    \n",
       "3  528                                 0                                    \n",
       "4  1615                                119                                  \n",
       "\n",
       "   GOMECC4_CAPECORAL_Sta141_Surface_C  GOMECC4_CATOCHE_Sta107_DCM_A  \\\n",
       "0  2531                                0                              \n",
       "1  2386                                9                              \n",
       "2  0                                   0                              \n",
       "3  0                                   0                              \n",
       "4  0                                   0                              \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta107_DCM_B  GOMECC4_CATOCHE_Sta107_DCM_C  \\\n",
       "0  0                             0                              \n",
       "1  18                            0                              \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  0                             0                              \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta107_Deep_A  GOMECC4_CATOCHE_Sta107_Deep_B  \\\n",
       "0  0                              21                              \n",
       "1  198                            40                              \n",
       "2  187                            0                               \n",
       "3  406                            0                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta107_Deep_C  GOMECC4_CATOCHE_Sta107_Surface_A  \\\n",
       "0  150                            0                                  \n",
       "1  0                              238                                \n",
       "2  0                              0                                  \n",
       "3  5                              0                                  \n",
       "4  0                              850                                \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta107_Surface_B  GOMECC4_CATOCHE_Sta107_Surface_C  \\\n",
       "0  0                                 0                                  \n",
       "1  0                                 13                                 \n",
       "2  0                                 0                                  \n",
       "3  0                                 0                                  \n",
       "4  0                                 94                                 \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta109_Deep_C  GOMECC4_CATOCHE_Sta109_DCM_B  \\\n",
       "0  308                            0                              \n",
       "1  59                             1312                           \n",
       "2  0                              0                              \n",
       "3  21                             0                              \n",
       "4  0                              2                              \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta109_DCM_C  GOMECC4_CATOCHE_Sta109_Deep_A  \\\n",
       "0  0                             25                              \n",
       "1  60                            0                               \n",
       "2  0                             25                              \n",
       "3  0                             0                               \n",
       "4  382                           32                              \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta109_Deep_B  GOMECC4_CATOCHE_Sta109_DCM_A  \\\n",
       "0  0                              12                             \n",
       "1  0                              0                              \n",
       "2  0                              0                              \n",
       "3  0                              7                              \n",
       "4  11                             204                            \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta109_Surface_A  GOMECC4_CATOCHE_Sta109_Surface_B  \\\n",
       "0  0                                 0                                  \n",
       "1  0                                 0                                  \n",
       "2  0                                 0                                  \n",
       "3  0                                 0                                  \n",
       "4  529                               0                                  \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta109_Surface_C  GOMECC4_CATOCHE_Sta115_DCM_A  \\\n",
       "0  7                                 0                              \n",
       "1  0                                 0                              \n",
       "2  0                                 413                            \n",
       "3  0                                 1050                           \n",
       "4  334                               0                              \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta115_DCM_B  GOMECC4_CATOCHE_Sta115_DCM_C  \\\n",
       "0  0                             999                            \n",
       "1  132                           0                              \n",
       "2  0                             217                            \n",
       "3  499                           1427                           \n",
       "4  0                             0                              \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta115_Deep_A  GOMECC4_CATOCHE_Sta115_Deep_B  \\\n",
       "0  0                              989                             \n",
       "1  65                             266                             \n",
       "2  566                            1107                            \n",
       "3  1143                           1051                            \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta115_Deep_C  GOMECC4_CATOCHE_Sta115_Surface_A  \\\n",
       "0  0                              0                                  \n",
       "1  341                            0                                  \n",
       "2  141                            0                                  \n",
       "3  1381                           1635                               \n",
       "4  0                              0                                  \n",
       "\n",
       "   GOMECC4_CATOCHE_Sta115_Surface_B  GOMECC4_CATOCHE_Sta115_Surface_C  \\\n",
       "0  0                                 1524                               \n",
       "1  82                                45                                 \n",
       "2  0                                 0                                  \n",
       "3  2096                              1389                               \n",
       "4  0                                 0                                  \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta121_DCM_A  GOMECC4_FLSTRAITS_Sta121_DCM_B  \\\n",
       "0  0                               10                               \n",
       "1  0                               0                                \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  0                               4                                \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta121_DCM_C  GOMECC4_FLSTRAITS_Sta121_Deep_A  \\\n",
       "0  0                               0                                 \n",
       "1  0                               0                                 \n",
       "2  0                               0                                 \n",
       "3  0                               49                                \n",
       "4  0                               0                                 \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta121_Deep_B  GOMECC4_FLSTRAITS_Sta121_Deep_C  \\\n",
       "0  0                                0                                 \n",
       "1  80                               0                                 \n",
       "2  0                                27                                \n",
       "3  0                                48                                \n",
       "4  0                                0                                 \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta121_Surface_A  GOMECC4_FLSTRAITS_Sta121_Surface_B  \\\n",
       "0  0                                   0                                    \n",
       "1  396                                 0                                    \n",
       "2  266                                 0                                    \n",
       "3  0                                   0                                    \n",
       "4  796                                 0                                    \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta121_Surface_C  GOMECC4_FLSTRAITS_Sta122_DCM_A  \\\n",
       "0  10                                  0                                \n",
       "1  10                                  96                               \n",
       "2  7                                   0                                \n",
       "3  71                                  0                                \n",
       "4  0                                   174                              \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta122_DCM_B  GOMECC4_FLSTRAITS_Sta122_DCM_C  \\\n",
       "0  0                               11                               \n",
       "1  293                             50                               \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  216                             328                              \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta122_Deep_A  GOMECC4_FLSTRAITS_Sta122_Deep_B  \\\n",
       "0  13                               0                                 \n",
       "1  214                              399                               \n",
       "2  56                               0                                 \n",
       "3  81                               0                                 \n",
       "4  70                               0                                 \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta122_Deep_C  GOMECC4_FLSTRAITS_Sta122_Surface_A  \\\n",
       "0  15                               0                                    \n",
       "1  0                                1880                                 \n",
       "2  9                                0                                    \n",
       "3  0                                0                                    \n",
       "4  7                                1931                                 \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta122_Surface_B  GOMECC4_FLSTRAITS_Sta122_Surface_C  \\\n",
       "0  8001                                0                                    \n",
       "1  0                                   679                                  \n",
       "2  0                                   0                                    \n",
       "3  0                                   0                                    \n",
       "4  0                                   0                                    \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta123_DCM_A  GOMECC4_FLSTRAITS_Sta123_DCM_B  \\\n",
       "0  2820                            4403                             \n",
       "1  1002                            1082                             \n",
       "2  69                              470                              \n",
       "3  2189                            166                              \n",
       "4  826                             855                              \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta123_DCM_C  GOMECC4_FLSTRAITS_Sta123_Deep_A  \\\n",
       "0  0                               62                                \n",
       "1  28                              19                                \n",
       "2  0                               0                                 \n",
       "3  2287                            14                                \n",
       "4  348                             12                                \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta123_Deep_B  GOMECC4_FLSTRAITS_Sta123_Deep_C  \\\n",
       "0  0                                0                                 \n",
       "1  69                               0                                 \n",
       "2  0                                0                                 \n",
       "3  0                                0                                 \n",
       "4  0                                0                                 \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta123_Surface_A  GOMECC4_FLSTRAITS_Sta123_Surface_B  \\\n",
       "0  1067                                0                                    \n",
       "1  87                                  0                                    \n",
       "2  353                                 9                                    \n",
       "3  1027                                0                                    \n",
       "4  0                                   7                                    \n",
       "\n",
       "   GOMECC4_FLSTRAITS_Sta123_Surface_C  GOMECC4_GALVESTON_Sta49_DCM_A  \\\n",
       "0  0                                   0                               \n",
       "1  526                                 655                             \n",
       "2  1704                                392                             \n",
       "3  0                                   450                             \n",
       "4  157                                 1383                            \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta49_DCM_B  GOMECC4_GALVESTON_Sta49_DCM_C  \\\n",
       "0  0                              0                               \n",
       "1  412                            0                               \n",
       "2  2487                           0                               \n",
       "3  1240                           935                             \n",
       "4  825                            1942                            \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta49_Deep_A  GOMECC4_GALVESTON_Sta49_Deep_B  \\\n",
       "0  0                               0                                \n",
       "1  506                             827                              \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  0                               0                                \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta49_Deep_C  GOMECC4_GALVESTON_Sta49_Surface_A  \\\n",
       "0  0                               0                                   \n",
       "1  769                             192                                 \n",
       "2  0                               0                                   \n",
       "3  243                             3436                                \n",
       "4  0                               287                                 \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta49_Surface_B  GOMECC4_GALVESTON_Sta49_Surface_C  \\\n",
       "0  1148                               0                                   \n",
       "1  1977                               2583                                \n",
       "2  5                                  0                                   \n",
       "3  392                                671                                 \n",
       "4  1768                               5452                                \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta50_DCM_A  GOMECC4_GALVESTON_Sta50_DCM_B  \\\n",
       "0  618                            0                               \n",
       "1  2521                           736                             \n",
       "2  85                             48                              \n",
       "3  142                            338                             \n",
       "4  45                             46                              \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta50_DCM_C  GOMECC4_GALVESTON_Sta50_Deep_A  \\\n",
       "0  0                              617                              \n",
       "1  2091                           2944                             \n",
       "2  0                              0                                \n",
       "3  0                              0                                \n",
       "4  0                              0                                \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta50_Deep_B  GOMECC4_GALVESTON_Sta50_Deep_C  \\\n",
       "0  0                               0                                \n",
       "1  111                             1243                             \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  0                               0                                \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta50_Surface_A  GOMECC4_GALVESTON_Sta50_Surface_B  \\\n",
       "0  0                                  4062                                \n",
       "1  33                                 244                                 \n",
       "2  0                                  0                                   \n",
       "3  1761                               273                                 \n",
       "4  544                                687                                 \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta50_Surface_C  GOMECC4_GALVESTON_Sta54_DCM_A  \\\n",
       "0  0                                  0                               \n",
       "1  20                                 0                               \n",
       "2  0                                  0                               \n",
       "3  1444                               0                               \n",
       "4  586                                0                               \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta54_DCM_B  GOMECC4_GALVESTON_Sta54_DCM_C  \\\n",
       "0  0                              0                               \n",
       "1  0                              0                               \n",
       "2  0                              0                               \n",
       "3  0                              0                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta54_Deep_A  GOMECC4_GALVESTON_Sta54_Deep_B  \\\n",
       "0  0                               7                                \n",
       "1  0                               4                                \n",
       "2  0                               8                                \n",
       "3  0                               0                                \n",
       "4  0                               12                               \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta54_Deep_C  GOMECC4_GALVESTON_Sta54_Surface_A  \\\n",
       "0  0                               8804                                \n",
       "1  7                               1511                                \n",
       "2  0                               1403                                \n",
       "3  0                               0                                   \n",
       "4  0                               73                                  \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta54_Surface_B  GOMECC4_GALVESTON_Sta54_Surface_C  \\\n",
       "0  0                                  0                                   \n",
       "1  4301                               1672                                \n",
       "2  1930                               3023                                \n",
       "3  933                                355                                 \n",
       "4  0                                  0                                   \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta59_DCM_A  GOMECC4_GALVESTON_Sta59_DCM_B  \\\n",
       "0  0                              0                               \n",
       "1  0                              12                              \n",
       "2  4658                           1322                            \n",
       "3  1285                           1379                            \n",
       "4  0                              77                              \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta59_DCM_C  GOMECC4_GALVESTON_Sta59_Deep_A  \\\n",
       "0  0                              0                                \n",
       "1  0                              0                                \n",
       "2  3726                           3049                             \n",
       "3  1959                           1337                             \n",
       "4  0                              0                                \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta59_Deep_B  GOMECC4_GALVESTON_Sta59_Deep_C  \\\n",
       "0  101                             0                                \n",
       "1  13                              0                                \n",
       "2  1159                            2491                             \n",
       "3  410                             1778                             \n",
       "4  42                              0                                \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta59_Surface_A  GOMECC4_GALVESTON_Sta59_Surface_B  \\\n",
       "0  0                                  0                                   \n",
       "1  0                                  18                                  \n",
       "2  524                                557                                 \n",
       "3  1360                               931                                 \n",
       "4  14                                 12                                  \n",
       "\n",
       "   GOMECC4_GALVESTON_Sta59_Surface_C  GOMECC4_LA_Sta38_DCM_A  \\\n",
       "0  68                                 0                        \n",
       "1  31                                 1019                     \n",
       "2  293                                0                        \n",
       "3  1352                               323                      \n",
       "4  40                                 4651                     \n",
       "\n",
       "   GOMECC4_LA_Sta38_DCM_B  GOMECC4_LA_Sta38_DCM_C  GOMECC4_LA_Sta38_Deep_A  \\\n",
       "0  0                       19                      0                         \n",
       "1  1736                    398                     0                         \n",
       "2  356                     18                      18                        \n",
       "3  62                      165                     16                        \n",
       "4  3589                    1438                    0                         \n",
       "\n",
       "   GOMECC4_LA_Sta38_Deep_B  GOMECC4_LA_Sta38_Deep_C  \\\n",
       "0  0                        0                         \n",
       "1  81                       37                        \n",
       "2  0                        687                       \n",
       "3  6                        444                       \n",
       "4  0                        79                        \n",
       "\n",
       "   GOMECC4_LA_Sta38_Surface_A  GOMECC4_LA_Sta38_Surface_B  \\\n",
       "0  8245                        0                            \n",
       "1  250                         1199                         \n",
       "2  315                         617                          \n",
       "3  71                          0                            \n",
       "4  88                          0                            \n",
       "\n",
       "   GOMECC4_LA_Sta38_Surface_C  GOMECC4_LA_Sta39_DCM_A  GOMECC4_LA_Sta39_DCM_B  \\\n",
       "0  14                          0                       26                       \n",
       "1  299                         192                     322                      \n",
       "2  1506                        0                       0                        \n",
       "3  1337                        0                       0                        \n",
       "4  557                         1758                    2206                     \n",
       "\n",
       "   GOMECC4_LA_Sta39_DCM_C  GOMECC4_LA_Sta39_Deep_A  GOMECC4_LA_Sta39_Deep_B  \\\n",
       "0  0                       14                       0                         \n",
       "1  327                     38                       0                         \n",
       "2  16                      324                      0                         \n",
       "3  0                       948                      6                         \n",
       "4  72                      957                      4                         \n",
       "\n",
       "   GOMECC4_LA_Sta39_Deep_C  GOMECC4_LA_Sta39_Surface_A  \\\n",
       "0  0                        0                            \n",
       "1  23                       557                          \n",
       "2  0                        115                          \n",
       "3  323                      1567                         \n",
       "4  46                       546                          \n",
       "\n",
       "   GOMECC4_LA_Sta39_Surface_B  GOMECC4_LA_Sta39_Surface_C  \\\n",
       "0  0                           2776                         \n",
       "1  29                          1085                         \n",
       "2  132                         9                            \n",
       "3  574                         0                            \n",
       "4  184                         2611                         \n",
       "\n",
       "   GOMECC4_LA_Sta40_DCM_A  GOMECC4_LA_Sta40_DCM_B  GOMECC4_LA_Sta40_DCM_C  \\\n",
       "0  0                       1485                    1471                     \n",
       "1  1732                    3539                    1424                     \n",
       "2  0                       0                       0                        \n",
       "3  0                       2475                    0                        \n",
       "4  833                     77                      0                        \n",
       "\n",
       "   GOMECC4_LA_Sta40_Deep_A  GOMECC4_LA_Sta40_Deep_B  GOMECC4_LA_Sta40_Deep_C  \\\n",
       "0  0                        0                        0                         \n",
       "1  0                        0                        18                        \n",
       "2  0                        304                      0                         \n",
       "3  0                        836                      0                         \n",
       "4  0                        9                        10                        \n",
       "\n",
       "   GOMECC4_LA_Sta40_Surface_A  GOMECC4_LA_Sta40_Surface_B  \\\n",
       "0  5763                        8484                         \n",
       "1  163                         512                          \n",
       "2  0                           0                            \n",
       "3  0                           0                            \n",
       "4  437                         334                          \n",
       "\n",
       "   GOMECC4_LA_Sta40_Surface_C  GOMECC4_LA_Sta45_DCM_A  GOMECC4_LA_Sta45_DCM_B  \\\n",
       "0  14                          0                       0                        \n",
       "1  2093                        37                      18                       \n",
       "2  0                           774                     331                      \n",
       "3  0                           498                     214                      \n",
       "4  1678                        64                      53                       \n",
       "\n",
       "   GOMECC4_LA_Sta45_DCM_C  GOMECC4_LA_Sta45_Deep_A  GOMECC4_LA_Sta45_Deep_B  \\\n",
       "0  0                       0                        128                       \n",
       "1  47                      5                        126                       \n",
       "2  1803                    21                       43                        \n",
       "3  53                      273                      0                         \n",
       "4  0                       24                       35                        \n",
       "\n",
       "   GOMECC4_LA_Sta45_Deep_C  GOMECC4_LA_Sta45_Surface_A  \\\n",
       "0  41                       2342                         \n",
       "1  0                        19                           \n",
       "2  0                        2013                         \n",
       "3  0                        117                          \n",
       "4  0                        27                           \n",
       "\n",
       "   GOMECC4_LA_Sta45_Surface_B  GOMECC4_LA_Sta45_Surface_C  \\\n",
       "0  0                           0                            \n",
       "1  0                           14                           \n",
       "2  142                         1162                         \n",
       "3  93                          106                          \n",
       "4  13                          0                            \n",
       "\n",
       "   GOMECC4_MERIDA_Sta94_DCM_A  GOMECC4_MERIDA_Sta94_DCM_B  \\\n",
       "0  0                           0                            \n",
       "1  888                         20                           \n",
       "2  2987                        6178                         \n",
       "3  5592                        477                          \n",
       "4  1558                        1700                         \n",
       "\n",
       "   GOMECC4_MERIDA_Sta94_DCM_C  GOMECC4_MERIDA_Sta94_Deep_A  \\\n",
       "0  0                           0                             \n",
       "1  540                         885                           \n",
       "2  2887                        2617                          \n",
       "3  3707                        2674                          \n",
       "4  1669                        1849                          \n",
       "\n",
       "   GOMECC4_MERIDA_Sta94_Deep_B  GOMECC4_MERIDA_Sta94_Deep_C  \\\n",
       "0  0                            0                             \n",
       "1  192                          527                           \n",
       "2  3531                         1536                          \n",
       "3  3967                         2291                          \n",
       "4  1217                         2387                          \n",
       "\n",
       "   GOMECC4_MERIDA_Sta94_Surface_A  GOMECC4_MERIDA_Sta94_Surface_B  \\\n",
       "0  0                               49                               \n",
       "1  496                             136                              \n",
       "2  1468                            2136                             \n",
       "3  958                             2013                             \n",
       "4  53                              256                              \n",
       "\n",
       "   GOMECC4_MERIDA_Sta94_Surface_C  GOMECC4_MERIDA_Sta97_DCM_A  \\\n",
       "0  0                               0                            \n",
       "1  327                             1093                         \n",
       "2  1079                            0                            \n",
       "3  1677                            0                            \n",
       "4  0                               0                            \n",
       "\n",
       "   GOMECC4_MERIDA_Sta97_DCM_B  GOMECC4_MERIDA_Sta97_DCM_C  \\\n",
       "0  0                           0                            \n",
       "1  190                         0                            \n",
       "2  0                           0                            \n",
       "3  13                          0                            \n",
       "4  0                           0                            \n",
       "\n",
       "   GOMECC4_MERIDA_Sta97_Deep_A  GOMECC4_MERIDA_Sta97_Deep_B  \\\n",
       "0  0                            15                            \n",
       "1  0                            436                           \n",
       "2  0                            0                             \n",
       "3  0                            0                             \n",
       "4  0                            94                            \n",
       "\n",
       "   GOMECC4_MERIDA_Sta97_Deep_C  GOMECC4_MERIDA_Sta97_Surface_A  \\\n",
       "0  4341                         6942                             \n",
       "1  10                           1367                             \n",
       "2  0                            3297                             \n",
       "3  669                          0                                \n",
       "4  0                            0                                \n",
       "\n",
       "   GOMECC4_MERIDA_Sta97_Surface_B  GOMECC4_MERIDA_Sta97_Surface_C  \\\n",
       "0  3170                            14                               \n",
       "1  1708                            0                                \n",
       "2  319                             0                                \n",
       "3  0                               0                                \n",
       "4  0                               0                                \n",
       "\n",
       "   GOMECC4_MERIDA_Sta98_DCM_A  GOMECC4_MERIDA_Sta98_DCM_B  \\\n",
       "0  0                           0                            \n",
       "1  6272                        4382                         \n",
       "2  9                           26                           \n",
       "3  0                           477                          \n",
       "4  51                          560                          \n",
       "\n",
       "   GOMECC4_MERIDA_Sta98_DCM_C  GOMECC4_MERIDA_Sta98_Deep_A  \\\n",
       "0  0                           0                             \n",
       "1  82                          210                           \n",
       "2  0                           1523                          \n",
       "3  0                           3185                          \n",
       "4  0                           1030                          \n",
       "\n",
       "   GOMECC4_MERIDA_Sta98_Deep_B  GOMECC4_MERIDA_Sta98_Deep_C  \\\n",
       "0  18                           36                            \n",
       "1  36                           268                           \n",
       "2  754                          888                           \n",
       "3  49                           1620                          \n",
       "4  214                          1605                          \n",
       "\n",
       "   GOMECC4_MERIDA_Sta98_Surface_A  GOMECC4_MERIDA_Sta98_Surface_B  \\\n",
       "0  51                              0                                \n",
       "1  3249                            3755                             \n",
       "2  48                              1630                             \n",
       "3  943                             795                              \n",
       "4  202                             40                               \n",
       "\n",
       "   GOMECC4_MERIDA_Sta98_Surface_C  GOMECC4_PAISNP_Sta61_DCM_A  \\\n",
       "0  40                              0                            \n",
       "1  524                             125                          \n",
       "2  0                               0                            \n",
       "3  54                              2328                         \n",
       "4  0                               1854                         \n",
       "\n",
       "   GOMECC4_PAISNP_Sta61_DCM_B  GOMECC4_PAISNP_Sta61_DCM_C  \\\n",
       "0  0                           78                           \n",
       "1  6                           494                          \n",
       "2  1773                        11                           \n",
       "3  821                         2364                         \n",
       "4  1139                        2005                         \n",
       "\n",
       "   GOMECC4_PAISNP_Sta61_Deep_A  GOMECC4_PAISNP_Sta61_Deep_B  \\\n",
       "0  0                            0                             \n",
       "1  16                           120                           \n",
       "2  418                          107                           \n",
       "3  1353                         1305                          \n",
       "4  107                          459                           \n",
       "\n",
       "   GOMECC4_PAISNP_Sta61_Deep_C  GOMECC4_PAISNP_Sta61_Surface_A  \\\n",
       "0  0                            0                                \n",
       "1  9                            42                               \n",
       "2  295                          511                              \n",
       "3  545                          1416                             \n",
       "4  70                           1632                             \n",
       "\n",
       "   GOMECC4_PAISNP_Sta61_Surface_B  GOMECC4_PAISNP_Sta61_Surface_C  \\\n",
       "0  0                               729                              \n",
       "1  87                              17                               \n",
       "2  354                             104                              \n",
       "3  1628                            4270                             \n",
       "4  700                             347                              \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta19_DCM_A  GOMECC4_PANAMACITY_Sta19_DCM_B  \\\n",
       "0  0                               0                                \n",
       "1  0                               0                                \n",
       "2  1393                            3355                             \n",
       "3  162                             114                              \n",
       "4  0                               0                                \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta19_DCM_C  GOMECC4_PANAMACITY_Sta19_Deep_A  \\\n",
       "0  0                               15                                \n",
       "1  0                               0                                 \n",
       "2  3011                            0                                 \n",
       "3  187                             392                               \n",
       "4  0                               0                                 \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta19_Deep_B  GOMECC4_PANAMACITY_Sta19_Deep_C  \\\n",
       "0  0                                0                                 \n",
       "1  0                                0                                 \n",
       "2  56                               419                               \n",
       "3  42                               623                               \n",
       "4  0                                0                                 \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta19_Surface_A  GOMECC4_PANAMACITY_Sta19_Surface_B  \\\n",
       "0  0                                   0                                    \n",
       "1  8                                   12                                   \n",
       "2  5994                                5230                                 \n",
       "3  1672                                1014                                 \n",
       "4  0                                   0                                    \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta19_Surface_C  GOMECC4_PANAMACITY_Sta21_DCM_A  \\\n",
       "0  0                                   0                                \n",
       "1  0                                   526                              \n",
       "2  4584                                0                                \n",
       "3  1925                                0                                \n",
       "4  0                                   38                               \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta21_DCM_B  GOMECC4_PANAMACITY_Sta21_DCM_C  \\\n",
       "0  0                               0                                \n",
       "1  1593                            246                              \n",
       "2  0                               682                              \n",
       "3  0                               0                                \n",
       "4  0                               0                                \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta21_Deep_A  GOMECC4_PANAMACITY_Sta21_Deep_B  \\\n",
       "0  0                                0                                 \n",
       "1  2499                             1207                              \n",
       "2  0                                0                                 \n",
       "3  0                                0                                 \n",
       "4  2                                0                                 \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta21_Deep_C  GOMECC4_PANAMACITY_Sta21_Surface_A  \\\n",
       "0  0                                1145                                 \n",
       "1  1160                             553                                  \n",
       "2  0                                2169                                 \n",
       "3  0                                648                                  \n",
       "4  19                               0                                    \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta21_Surface_B  GOMECC4_PANAMACITY_Sta21_Surface_C  \\\n",
       "0  0                                   0                                    \n",
       "1  634                                 2675                                 \n",
       "2  6181                                1781                                 \n",
       "3  2448                                1580                                 \n",
       "4  95                                  77                                   \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta23_DCM_A  GOMECC4_PANAMACITY_Sta23_DCM_B  \\\n",
       "0  0                               0                                \n",
       "1  2379                            1728                             \n",
       "2  0                               0                                \n",
       "3  0                               0                                \n",
       "4  0                               374                              \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta23_DCM_C  GOMECC4_PANAMACITY_Sta23_Deep_A  \\\n",
       "0  0                               0                                 \n",
       "1  463                             0                                 \n",
       "2  8                               0                                 \n",
       "3  0                               65                                \n",
       "4  0                               0                                 \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta23_Deep_B  GOMECC4_PANAMACITY_Sta23_Deep_C  \\\n",
       "0  0                                0                                 \n",
       "1  7                                0                                 \n",
       "2  1596                             0                                 \n",
       "3  0                                8                                 \n",
       "4  0                                0                                 \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta23_Surface_A  GOMECC4_PANAMACITY_Sta23_Surface_B  \\\n",
       "0  61                                  0                                    \n",
       "1  1790                                832                                  \n",
       "2  342                                 1591                                 \n",
       "3  428                                 372                                  \n",
       "4  591                                 852                                  \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta23_Surface_C  GOMECC4_PANAMACITY_Sta28_DCM_A  \\\n",
       "0  0                                   0                                \n",
       "1  1860                                64                               \n",
       "2  278                                 0                                \n",
       "3  815                                 0                                \n",
       "4  457                                 321                              \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta28_DCM_B  GOMECC4_PANAMACITY_Sta28_DCM_C  \\\n",
       "0  928                             0                                \n",
       "1  52                              1133                             \n",
       "2  0                               4                                \n",
       "3  0                               0                                \n",
       "4  647                             189                              \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta28_Deep_A  GOMECC4_PANAMACITY_Sta28_Deep_B  \\\n",
       "0  0                                0                                 \n",
       "1  9                                0                                 \n",
       "2  0                                0                                 \n",
       "3  0                                0                                 \n",
       "4  0                                9                                 \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta28_Deep_C  GOMECC4_PANAMACITY_Sta28_Surface_A  \\\n",
       "0  0                                0                                    \n",
       "1  0                                55                                   \n",
       "2  5                                969                                  \n",
       "3  0                                991                                  \n",
       "4  5                                0                                    \n",
       "\n",
       "   GOMECC4_PANAMACITY_Sta28_Surface_B  GOMECC4_PANAMACITY_Sta28_Surface_C  \\\n",
       "0  140                                 0                                    \n",
       "1  808                                 108                                  \n",
       "2  3922                                1257                                 \n",
       "3  428                                 799                                  \n",
       "4  413                                 161                                  \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta31_DCM_A  GOMECC4_PENSACOLA_Sta31_DCM_B  \\\n",
       "0  10                             0                               \n",
       "1  612                            595                             \n",
       "2  524                            428                             \n",
       "3  0                              221                             \n",
       "4  642                            966                             \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta31_DCM_C  GOMECC4_PENSACOLA_Sta31_Deep_A  \\\n",
       "0  0                              0                                \n",
       "1  826                            1128                             \n",
       "2  399                            0                                \n",
       "3  0                              0                                \n",
       "4  850                            0                                \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta31_Deep_B  GOMECC4_PENSACOLA_Sta31_Deep_C  \\\n",
       "0  0                               0                                \n",
       "1  3659                            0                                \n",
       "2  0                               0                                \n",
       "3  0                               6                                \n",
       "4  0                               0                                \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta31_Surface_A  GOMECC4_PENSACOLA_Sta31_Surface_B  \\\n",
       "0  0                                  0                                   \n",
       "1  0                                  0                                   \n",
       "2  355                                674                                 \n",
       "3  1592                               865                                 \n",
       "4  280                                345                                 \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta31_Surface_C  GOMECC4_PENSACOLA_Sta33_DCM_A  \\\n",
       "0  0                                  0                               \n",
       "1  30                                 992                             \n",
       "2  1104                               23                              \n",
       "3  1066                               0                               \n",
       "4  20                                 0                               \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta33_DCM_B  GOMECC4_PENSACOLA_Sta33_DCM_C  \\\n",
       "0  0                              0                               \n",
       "1  1878                           487                             \n",
       "2  0                              0                               \n",
       "3  0                              0                               \n",
       "4  0                              61                              \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta33_Deep_A  GOMECC4_PENSACOLA_Sta33_Deep_B  \\\n",
       "0  0                               0                                \n",
       "1  0                               0                                \n",
       "2  0                               955                              \n",
       "3  0                               17                               \n",
       "4  5                               0                                \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta33_Deep_C  GOMECC4_PENSACOLA_Sta33_Surface_A  \\\n",
       "0  0                               0                                   \n",
       "1  0                               1534                                \n",
       "2  0                               0                                   \n",
       "3  30                              365                                 \n",
       "4  0                               1055                                \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta33_Surface_B  GOMECC4_PENSACOLA_Sta33_Surface_C  \\\n",
       "0  0                                  1873                                \n",
       "1  3186                               1040                                \n",
       "2  0                                  0                                   \n",
       "3  189                                0                                   \n",
       "4  881                                103                                 \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta35_DCM_A  GOMECC4_PENSACOLA_Sta35_DCM_B  \\\n",
       "0  0                              0                               \n",
       "1  267                            221                             \n",
       "2  0                              0                               \n",
       "3  0                              0                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta35_DCM_C  GOMECC4_PENSACOLA_Sta35_Deep_A  \\\n",
       "0  0                              0                                \n",
       "1  1512                           0                                \n",
       "2  0                              0                                \n",
       "3  0                              0                                \n",
       "4  49                             3                                \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta35_Deep_B  GOMECC4_PENSACOLA_Sta35_Deep_C  \\\n",
       "0  6                               0                                \n",
       "1  534                             9                                \n",
       "2  477                             0                                \n",
       "3  0                               0                                \n",
       "4  193                             0                                \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta35_Surface_A  GOMECC4_PENSACOLA_Sta35_Surface_B  \\\n",
       "0  1847                               1481                                \n",
       "1  134                                268                                 \n",
       "2  1335                               0                                   \n",
       "3  491                                1558                                \n",
       "4  95                                 0                                   \n",
       "\n",
       "   GOMECC4_PENSACOLA_Sta35_Surface_C  GOMECC4_TAMPA_Sta10_DCM_A  \\\n",
       "0  2783                               0                           \n",
       "1  0                                  274                         \n",
       "2  1315                               0                           \n",
       "3  825                                0                           \n",
       "4  0                                  0                           \n",
       "\n",
       "   GOMECC4_TAMPA_Sta10_DCM_B  GOMECC4_TAMPA_Sta10_DCM_C  \\\n",
       "0  0                          0                           \n",
       "1  0                          0                           \n",
       "2  12                         0                           \n",
       "3  0                          0                           \n",
       "4  0                          0                           \n",
       "\n",
       "   GOMECC4_TAMPA_Sta10_Deep_A  GOMECC4_TAMPA_Sta10_Deep_B  \\\n",
       "0  0                           7                            \n",
       "1  4                           0                            \n",
       "2  0                           0                            \n",
       "3  0                           0                            \n",
       "4  0                           0                            \n",
       "\n",
       "   GOMECC4_TAMPA_Sta10_Deep_C  GOMECC4_TAMPA_Sta10_Surface_A  \\\n",
       "0  22                          0                               \n",
       "1  0                           181                             \n",
       "2  0                           1848                            \n",
       "3  0                           3862                            \n",
       "4  0                           0                               \n",
       "\n",
       "   GOMECC4_TAMPA_Sta10_Surface_B  GOMECC4_TAMPA_Sta10_Surface_C  \\\n",
       "0  0                              0                               \n",
       "1  5                              0                               \n",
       "2  765                            0                               \n",
       "3  3369                           4562                            \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_TAMPA_Sta13_DCM_A  GOMECC4_TAMPA_Sta13_DCM_B  \\\n",
       "0  0                          0                           \n",
       "1  0                          0                           \n",
       "2  0                          0                           \n",
       "3  0                          0                           \n",
       "4  0                          0                           \n",
       "\n",
       "   GOMECC4_TAMPA_Sta13_DCM_C  GOMECC4_TAMPA_Sta13_Deep_A  \\\n",
       "0  0                          0                            \n",
       "1  1096                       617                          \n",
       "2  0                          0                            \n",
       "3  0                          8                            \n",
       "4  0                          0                            \n",
       "\n",
       "   GOMECC4_TAMPA_Sta13_Deep_B  GOMECC4_TAMPA_Sta13_Deep_C  \\\n",
       "0  0                           0                            \n",
       "1  196                         180                          \n",
       "2  0                           0                            \n",
       "3  0                           0                            \n",
       "4  0                           0                            \n",
       "\n",
       "   GOMECC4_TAMPA_Sta13_Surface_A  GOMECC4_TAMPA_Sta13_Surface_B  \\\n",
       "0  10052                          9663                            \n",
       "1  0                              464                             \n",
       "2  0                              0                               \n",
       "3  800                            0                               \n",
       "4  520                            0                               \n",
       "\n",
       "   GOMECC4_TAMPA_Sta13_Surface_C  GOMECC4_TAMPA_Sta18_Deep_A  \\\n",
       "0  11791                          0                            \n",
       "1  0                              0                            \n",
       "2  0                              3010                         \n",
       "3  0                              1878                         \n",
       "4  0                              0                            \n",
       "\n",
       "   GOMECC4_TAMPA_Sta18_Deep_B  GOMECC4_TAMPA_Sta18_Deep_C  \\\n",
       "0  0                           0                            \n",
       "1  0                           0                            \n",
       "2  421                         269                          \n",
       "3  1095                        1284                         \n",
       "4  0                           0                            \n",
       "\n",
       "   GOMECC4_TAMPA_Sta18_Surface_A  GOMECC4_TAMPA_Sta18_Surface_B  \\\n",
       "0  0                              0                               \n",
       "1  5                              0                               \n",
       "2  387                            0                               \n",
       "3  1718                           1199                            \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_TAMPA_Sta18_Surface_C  GOMECC4_TAMPA_Sta9_Sediment_A  \\\n",
       "0  0                              1418                            \n",
       "1  52                             62                              \n",
       "2  1240                           0                               \n",
       "3  1236                           0                               \n",
       "4  0                              4                               \n",
       "\n",
       "   GOMECC4_TAMPA_Sta9_Sediment_B  GOMECC4_TAMPA_Sta9_Sediment_C  \\\n",
       "0  0                              0                               \n",
       "1  0                              0                               \n",
       "2  0                              0                               \n",
       "3  0                              0                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta76_DCM_A  GOMECC4_TAMPICO_Sta76_DCM_B  \\\n",
       "0  0                            963                           \n",
       "1  1448                         335                           \n",
       "2  0                            2364                          \n",
       "3  603                          905                           \n",
       "4  1922                         1218                          \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta76_DCM_C  GOMECC4_TAMPICO_Sta76_Deep_A  \\\n",
       "0  0                            0                              \n",
       "1  1997                         2398                           \n",
       "2  2408                         0                              \n",
       "3  0                            978                            \n",
       "4  2227                         759                            \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta76_Deep_B  GOMECC4_TAMPICO_Sta76_Deep_C  \\\n",
       "0  0                             0                              \n",
       "1  734                           2585                           \n",
       "2  4214                          1233                           \n",
       "3  453                           180                            \n",
       "4  886                           819                            \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta76_Surface_A  GOMECC4_TAMPICO_Sta76_Surface_B  \\\n",
       "0  3440                             1978                              \n",
       "1  821                              120                               \n",
       "2  1296                             1271                              \n",
       "3  0                                165                               \n",
       "4  440                              300                               \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta76_Surface_C  GOMECC4_TAMPICO_Sta79_DCM_A  \\\n",
       "0  0                                7                             \n",
       "1  1047                             1096                          \n",
       "2  1025                             0                             \n",
       "3  759                              0                             \n",
       "4  486                              220                           \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta79_DCM_B  GOMECC4_TAMPICO_Sta79_DCM_C  \\\n",
       "0  0                            1470                          \n",
       "1  694                          674                           \n",
       "2  0                            0                             \n",
       "3  0                            0                             \n",
       "4  0                            0                             \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta79_Deep_A  GOMECC4_TAMPICO_Sta79_Deep_B  \\\n",
       "0  0                             0                              \n",
       "1  112                           64                             \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  0                             0                              \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta79_Deep_C  GOMECC4_TAMPICO_Sta79_Surface_A  \\\n",
       "0  0                             0                                 \n",
       "1  74                            2850                              \n",
       "2  0                             0                                 \n",
       "3  0                             0                                 \n",
       "4  0                             791                               \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta79_Surface_B  GOMECC4_TAMPICO_Sta79_Surface_C  \\\n",
       "0  0                                0                                 \n",
       "1  2926                             2455                              \n",
       "2  0                                0                                 \n",
       "3  0                                0                                 \n",
       "4  2453                             24                                \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta82_DCM_A  GOMECC4_TAMPICO_Sta82_DCM_B  \\\n",
       "0  0                            0                             \n",
       "1  0                            103                           \n",
       "2  0                            0                             \n",
       "3  0                            0                             \n",
       "4  0                            0                             \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta82_DCM_C  GOMECC4_TAMPICO_Sta82_Deep_A  \\\n",
       "0  0                            0                              \n",
       "1  89                           75                             \n",
       "2  0                            0                              \n",
       "3  0                            0                              \n",
       "4  0                            0                              \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta82_Deep_B  GOMECC4_TAMPICO_Sta82_Deep_C  \\\n",
       "0  0                             0                              \n",
       "1  0                             0                              \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  6                             2                              \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta82_Surface_A  GOMECC4_TAMPICO_Sta82_Surface_B  \\\n",
       "0  0                                0                                 \n",
       "1  1841                             3754                              \n",
       "2  0                                2391                              \n",
       "3  83                               0                                 \n",
       "4  1520                             175                               \n",
       "\n",
       "   GOMECC4_TAMPICO_Sta82_Surface_C  GOMECC4_VERACRUZ_Sta85_DCM_A  \\\n",
       "0  0                                3576                           \n",
       "1  2194                             0                              \n",
       "2  0                                0                              \n",
       "3  0                                0                              \n",
       "4  1392                             0                              \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta85_DCM_B  GOMECC4_VERACRUZ_Sta85_DCM_C  \\\n",
       "0  0                             804                            \n",
       "1  131                           0                              \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  0                             0                              \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta85_Deep_A  GOMECC4_VERACRUZ_Sta85_Deep_B  \\\n",
       "0  0                              0                               \n",
       "1  0                              0                               \n",
       "2  0                              0                               \n",
       "3  0                              0                               \n",
       "4  3                              0                               \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta85_Deep_C  GOMECC4_VERACRUZ_Sta85_Surface_A  \\\n",
       "0  0                              353                                \n",
       "1  580                            3094                               \n",
       "2  15                             18                                 \n",
       "3  85                             0                                  \n",
       "4  121                            253                                \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta85_Surface_B  GOMECC4_VERACRUZ_Sta87_DCM_A  \\\n",
       "0  0                                 584                            \n",
       "1  544                               577                            \n",
       "2  0                                 0                              \n",
       "3  0                                 0                              \n",
       "4  1240                              0                              \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta87_DCM_B  GOMECC4_VERACRUZ_Sta87_DCM_C  \\\n",
       "0  2310                          0                              \n",
       "1  0                             820                            \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  244                           1423                           \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta87_Deep_A  GOMECC4_VERACRUZ_Sta87_Deep_B  \\\n",
       "0  13111                          0                               \n",
       "1  0                              0                               \n",
       "2  0                              45                              \n",
       "3  0                              0                               \n",
       "4  0                              2                               \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta87_Deep_C  GOMECC4_VERACRUZ_Sta87_Surface_A  \\\n",
       "0  0                              0                                  \n",
       "1  0                              0                                  \n",
       "2  0                              0                                  \n",
       "3  0                              0                                  \n",
       "4  0                              0                                  \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta87_Surface_B  GOMECC4_VERACRUZ_Sta87_Surface_C  \\\n",
       "0  0                                 12                                 \n",
       "1  2826                              447                                \n",
       "2  842                               0                                  \n",
       "3  11                                0                                  \n",
       "4  0                                 0                                  \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta89_DCM_A  GOMECC4_VERACRUZ_Sta89_DCM_B  \\\n",
       "0  0                             0                              \n",
       "1  1695                          632                            \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  0                             0                              \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta89_DCM_C  GOMECC4_VERACRUZ_Sta89_Deep_A  \\\n",
       "0  0                             333                             \n",
       "1  1133                          1999                            \n",
       "2  124                           989                             \n",
       "3  0                             0                               \n",
       "4  0                             0                               \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta89_Deep_B  GOMECC4_VERACRUZ_Sta89_Deep_C  \\\n",
       "0  0                              0                               \n",
       "1  1321                           417                             \n",
       "2  446                            825                             \n",
       "3  0                              0                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta89_Surface_A  GOMECC4_VERACRUZ_Sta89_Surface_B  \\\n",
       "0  0                                 0                                  \n",
       "1  378                               0                                  \n",
       "2  4441                              3916                               \n",
       "3  776                               1190                               \n",
       "4  0                                 0                                  \n",
       "\n",
       "   GOMECC4_VERACRUZ_Sta89_Surface_C  GOMECC4_YUCATAN_Sta100_DCM_A  \\\n",
       "0  0                                 32                             \n",
       "1  150                               17                             \n",
       "2  4151                              1088                           \n",
       "3  846                               1627                           \n",
       "4  0                                 0                              \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta100_DCM_B  GOMECC4_YUCATAN_Sta100_DCM_C  \\\n",
       "0  0                             0                              \n",
       "1  1385                          1253                           \n",
       "2  203                           2121                           \n",
       "3  281                           949                            \n",
       "4  1439                          744                            \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta100_Deep_A  GOMECC4_YUCATAN_Sta100_Deep_B  \\\n",
       "0  571                            0                               \n",
       "1  2696                           2413                            \n",
       "2  2015                           0                               \n",
       "3  1377                           711                             \n",
       "4  0                              5                               \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta100_Deep_C  GOMECC4_YUCATAN_Sta100_Surface_A  \\\n",
       "0  0                              1155                               \n",
       "1  2758                           385                                \n",
       "2  356                            0                                  \n",
       "3  677                            896                                \n",
       "4  0                              130                                \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta100_Surface_B  GOMECC4_YUCATAN_Sta100_Surface_C  \\\n",
       "0  4719                              194                                \n",
       "1  410                               2987                               \n",
       "2  0                                 67                                 \n",
       "3  1505                              135                                \n",
       "4  100                               529                                \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_DCM_A  GOMECC4_YUCATAN_Sta102_DCM_B  \\\n",
       "0  1451                          0                              \n",
       "1  2746                          1562                           \n",
       "2  0                             0                              \n",
       "3  0                             0                              \n",
       "4  206                           0                              \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_DCM_C  GOMECC4_YUCATAN_Sta102_Deep_A  \\\n",
       "0  1268                          0                               \n",
       "1  731                           2951                            \n",
       "2  0                             101                             \n",
       "3  0                             112                             \n",
       "4  1008                          70                              \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Deep_B  GOMECC4_YUCATAN_Sta102_Deep_C  \\\n",
       "0  61                             53                              \n",
       "1  1359                           511                             \n",
       "2  8                              20                              \n",
       "3  0                              0                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Surface_A  GOMECC4_YUCATAN_Sta102_Surface_B  \\\n",
       "0  2294                              0                                  \n",
       "1  2022                              0                                  \n",
       "2  0                                 0                                  \n",
       "3  0                                 0                                  \n",
       "4  739                               1743                               \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Surface_C  GOMECC4_YUCATAN_Sta106_DCM_A  \\\n",
       "0  2744                              0                              \n",
       "1  1412                              777                            \n",
       "2  0                                 0                              \n",
       "3  0                                 0                              \n",
       "4  526                               0                              \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta106_DCM_B  GOMECC4_YUCATAN_Sta106_DCM_C  \\\n",
       "0  0                             0                              \n",
       "1  65                            287                            \n",
       "2  17                            0                              \n",
       "3  0                             0                              \n",
       "4  1953                          0                              \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta106_Deep_A  GOMECC4_YUCATAN_Sta106_Deep_B  \\\n",
       "0  111                            0                               \n",
       "1  0                              0                               \n",
       "2  0                              0                               \n",
       "3  47                             0                               \n",
       "4  0                              0                               \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta106_Deep_C  GOMECC4_YUCATAN_Sta106_Surface_A  \\\n",
       "0  0                              30                                 \n",
       "1  0                              1530                               \n",
       "2  0                              0                                  \n",
       "3  75                             0                                  \n",
       "4  0                              2107                               \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta106_Surface_B  GOMECC4_YUCATAN_Sta106_Surface_C  \n",
       "0  0                                 8480                              \n",
       "1  0                                 954                               \n",
       "2  0                                 0                                 \n",
       "3  0                                 0                                 \n",
       "4  3962                              1627                              "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['18S V9'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### format taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to automate this? Everyone's taxonomy might be different?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#18S \n",
    "taxa_ranks_18S = ['domain','supergroup','division','subdivision','class','order','family','genus','species']\n",
    "\n",
    "asv_tables['18S V9'][['domain','supergroup','division','subdivision','class','order','family','genus','species']] = [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\"]\n",
    "for index, row in asv_tables['18S V9'].iterrows():\n",
    "    taxa = row['taxonomy'].split(\";\")\n",
    "    for i in range(0,len(taxa)):\n",
    "        if i < len(taxa_ranks_18S):\n",
    "            asv_tables['18S V9'].loc[index,taxa_ranks_18S[i]] = taxa[i]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>...</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta106_Surface_C</th>\n",
       "      <th>domain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>subdivision</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>1516</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8480</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Neocalanus_cristatus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>962</td>\n",
       "      <td>316</td>\n",
       "      <td>548</td>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>954</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Clausocalanus_furcatus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5d4df37251121c08397c6fbc27b06175</td>\n",
       "      <td>GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.992300</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Sinocalanus</td>\n",
       "      <td>Sinocalanus_sinensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f863f671a575c6ab587e8de0190d3335</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.998393</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Paracalanus_parvus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>1164</td>\n",
       "      <td>2272</td>\n",
       "      <td>2208</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1627</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Acrocalanus_sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>287</td>\n",
       "      <td>414</td>\n",
       "      <td>335</td>\n",
       "      <td>195</td>\n",
       "      <td>228</td>\n",
       "      <td>298</td>\n",
       "      <td>...</td>\n",
       "      <td>373</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>d70494a723d85d66aa88d2d8a975aeec</td>\n",
       "      <td>GCTACTACCGATTGAATGGTTCCGTGAATTCTTGAGATCGGCGCGG...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta</td>\n",
       "      <td>0.992451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>250</td>\n",
       "      <td>323</td>\n",
       "      <td>194</td>\n",
       "      <td>51</td>\n",
       "      <td>59</td>\n",
       "      <td>55</td>\n",
       "      <td>...</td>\n",
       "      <td>305</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bbaaf7bb4e71c80de970677779e3bf3a</td>\n",
       "      <td>GCTACTACCGATTGAATGGTTTAGTGAGATCTTCGGATTGGCACAA...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Cnidaria...</td>\n",
       "      <td>0.864777</td>\n",
       "      <td>212</td>\n",
       "      <td>50</td>\n",
       "      <td>237</td>\n",
       "      <td>552</td>\n",
       "      <td>1278</td>\n",
       "      <td>480</td>\n",
       "      <td>...</td>\n",
       "      <td>67</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>Cnidaria_X</td>\n",
       "      <td>Hydrozoa</td>\n",
       "      <td>Sulculeolaria</td>\n",
       "      <td>Sulculeolaria_quadrivalvis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7a8324bb4448b65f7adc73d70e5901da</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTATTTGGACTGGGCCTT...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.992088</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>405</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Delibus</td>\n",
       "      <td>Delibus_sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>fe57d269b68215339f9003c7f3ee7a02</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTCAG...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Corycaeus</td>\n",
       "      <td>Corycaeus_speciosus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>fcaab82b1754b0701a7d6dfc963e5fd6</td>\n",
       "      <td>GCTCCTACCGATGAATGTTCCGATGAAATTTTCGGAGAGTGCCTGT...</td>\n",
       "      <td>Eukaryota;TSAR;Rhizaria;Radiolaria;Polycystine...</td>\n",
       "      <td>0.996328</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Rhizaria</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>Polycystinea</td>\n",
       "      <td>Spumellaria</td>\n",
       "      <td>Astrosphaeridae</td>\n",
       "      <td>Heliosphaera</td>\n",
       "      <td>Heliosphaera_sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>9f8c0544a864bb0f0a14a4b747f18eaf</td>\n",
       "      <td>GCTCCTACCGATTGAATGGTCCGGTGAAGCGTTTGGACTATGGTTT...</td>\n",
       "      <td>Eukaryota;Archaeplastida;Chlorophyta;Chlorophy...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Archaeplastida</td>\n",
       "      <td>Chlorophyta</td>\n",
       "      <td>Chlorophyta_X</td>\n",
       "      <td>Mamiellophyceae</td>\n",
       "      <td>Mamiellales</td>\n",
       "      <td>Bathycoccaceae</td>\n",
       "      <td>Ostreococcus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1b13e125454669ac7f22760665bcbdc0</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.999858</td>\n",
       "      <td>268</td>\n",
       "      <td>306</td>\n",
       "      <td>290</td>\n",
       "      <td>135</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>179</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>8bbe50c736574078a3a79144290d0234</td>\n",
       "      <td>GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCCTTCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>d785682c0a83be9275e095d76cabbe36</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.818596</td>\n",
       "      <td>89</td>\n",
       "      <td>136</td>\n",
       "      <td>119</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>129</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>Gyrodinium</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ab5a60662ae7e1e268811d578973fe58</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCACGGACTGTTTGCA...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.927008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Metridia</td>\n",
       "      <td>Metridia_longa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>4eeed41258406c33ec57c50936d14a75</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.974467</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Paracalanus_parvus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>e069d06675910c19a1f20ea9c566d632</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.917869</td>\n",
       "      <td>0</td>\n",
       "      <td>2355</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>174</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Delibus</td>\n",
       "      <td>Delibus_nudus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>edbe2e93cd941256132c9d6b7521743d</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTCAG...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Corycaeus</td>\n",
       "      <td>Corycaeus_speciosus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 485 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           featureid  \\\n",
       "0   36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1   4e38e8ced9070952b314e1880bede1ca   \n",
       "2   5d4df37251121c08397c6fbc27b06175   \n",
       "3   f863f671a575c6ab587e8de0190d3335   \n",
       "4   2a31e5c01634165da99e7381279baa75   \n",
       "5   ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "6   d70494a723d85d66aa88d2d8a975aeec   \n",
       "7   fa1f1a97dd4ae7c826009186bad26384   \n",
       "8   bbaaf7bb4e71c80de970677779e3bf3a   \n",
       "9   7a8324bb4448b65f7adc73d70e5901da   \n",
       "10  fe57d269b68215339f9003c7f3ee7a02   \n",
       "11  fcaab82b1754b0701a7d6dfc963e5fd6   \n",
       "12  9f8c0544a864bb0f0a14a4b747f18eaf   \n",
       "13  1b13e125454669ac7f22760665bcbdc0   \n",
       "14  8bbe50c736574078a3a79144290d0234   \n",
       "15  d785682c0a83be9275e095d76cabbe36   \n",
       "16  ab5a60662ae7e1e268811d578973fe58   \n",
       "17  4eeed41258406c33ec57c50936d14a75   \n",
       "18  e069d06675910c19a1f20ea9c566d632   \n",
       "19  edbe2e93cd941256132c9d6b7521743d   \n",
       "\n",
       "                                             sequence  \\\n",
       "0   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...   \n",
       "1   GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...   \n",
       "2   GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCTTTCC...   \n",
       "3   GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "4   GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...   \n",
       "5   GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "6   GCTACTACCGATTGAATGGTTCCGTGAATTCTTGAGATCGGCGCGG...   \n",
       "7   GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...   \n",
       "8   GCTACTACCGATTGAATGGTTTAGTGAGATCTTCGGATTGGCACAA...   \n",
       "9   GCTACTACCGATTGAACGTTTTAGTGAGGTATTTGGACTGGGCCTT...   \n",
       "10  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTCAG...   \n",
       "11  GCTCCTACCGATGAATGTTCCGATGAAATTTTCGGAGAGTGCCTGT...   \n",
       "12  GCTCCTACCGATTGAATGGTCCGGTGAAGCGTTTGGACTATGGTTT...   \n",
       "13  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "14  GCTACTACCGATTGAGTGTTTTAGTGAGGTCCTCGGATTGCCTTCC...   \n",
       "15  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "16  GCTACTACCGATTGAACGTTTTAGTGAGGTCCACGGACTGTTTGCA...   \n",
       "17  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "18  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTGAGCC...   \n",
       "19  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTCAG...   \n",
       "\n",
       "                                             taxonomy  Confidence  \\\n",
       "0   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.922099   \n",
       "1   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.999947   \n",
       "2   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.992300   \n",
       "3   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.998393   \n",
       "4   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.779948   \n",
       "5   Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.999931   \n",
       "6                       Eukaryota;Obazoa;Opisthokonta    0.992451   \n",
       "7   Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.986908   \n",
       "8   Eukaryota;Obazoa;Opisthokonta;Metazoa;Cnidaria...    0.864777   \n",
       "9   Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.992088   \n",
       "10  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.999999   \n",
       "11  Eukaryota;TSAR;Rhizaria;Radiolaria;Polycystine...    0.996328   \n",
       "12  Eukaryota;Archaeplastida;Chlorophyta;Chlorophy...    1.000000   \n",
       "13  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.999858   \n",
       "14  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.999996   \n",
       "15  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.818596   \n",
       "16  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.927008   \n",
       "17  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.974467   \n",
       "18  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.917869   \n",
       "19  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    1.000000   \n",
       "\n",
       "    GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  GOMECC4_27N_Sta1_DCM_C  \\\n",
       "0                     1516                       0                       0   \n",
       "1                      962                     316                     548   \n",
       "2                        0                       4                       0   \n",
       "3                        0                       0                       0   \n",
       "4                     1164                    2272                    2208   \n",
       "5                      287                     414                     335   \n",
       "6                        0                       0                       0   \n",
       "7                      250                     323                     194   \n",
       "8                      212                      50                     237   \n",
       "9                        0                       0                       0   \n",
       "10                       0                       0                       0   \n",
       "11                       0                       0                       0   \n",
       "12                       0                       0                       4   \n",
       "13                     268                     306                     290   \n",
       "14                       0                       0                       0   \n",
       "15                      89                     136                     119   \n",
       "16                       0                       0                       0   \n",
       "17                       0                       0                       0   \n",
       "18                       0                    2355                       0   \n",
       "19                       0                       0                       0   \n",
       "\n",
       "    GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  GOMECC4_27N_Sta1_Deep_C  \\\n",
       "0                         6                        0                        0   \n",
       "1                        19                       10                        0   \n",
       "2                        12                        5                        0   \n",
       "3                         0                        0                        0   \n",
       "4                         2                        0                        0   \n",
       "5                       195                      228                      298   \n",
       "6                         4                        0                        0   \n",
       "7                        51                       59                       55   \n",
       "8                       552                     1278                      480   \n",
       "9                        15                        0                        0   \n",
       "10                        0                        0                        0   \n",
       "11                        0                        0                        0   \n",
       "12                        9                        0                        0   \n",
       "13                      135                      168                        0   \n",
       "14                        0                        0                        0   \n",
       "15                        0                        0                        0   \n",
       "16                       10                        0                        0   \n",
       "17                       31                        0                        0   \n",
       "18                        0                        0                        0   \n",
       "19                        0                        0                        0   \n",
       "\n",
       "    ...  GOMECC4_YUCATAN_Sta106_Surface_C     domain      supergroup  \\\n",
       "0   ...                              8480  Eukaryota          Obazoa   \n",
       "1   ...                               954  Eukaryota          Obazoa   \n",
       "2   ...                                 0  Eukaryota          Obazoa   \n",
       "3   ...                                 0  Eukaryota          Obazoa   \n",
       "4   ...                              1627  Eukaryota          Obazoa   \n",
       "5   ...                               373  Eukaryota            TSAR   \n",
       "6   ...                                 0  Eukaryota          Obazoa   \n",
       "7   ...                               305  Eukaryota            TSAR   \n",
       "8   ...                                67  Eukaryota          Obazoa   \n",
       "9   ...                               405  Eukaryota          Obazoa   \n",
       "10  ...                                 0  Eukaryota          Obazoa   \n",
       "11  ...                                 0  Eukaryota            TSAR   \n",
       "12  ...                                 0  Eukaryota  Archaeplastida   \n",
       "13  ...                               179  Eukaryota            TSAR   \n",
       "14  ...                                 0  Eukaryota          Obazoa   \n",
       "15  ...                               129  Eukaryota            TSAR   \n",
       "16  ...                                 0  Eukaryota          Obazoa   \n",
       "17  ...                                 0  Eukaryota          Obazoa   \n",
       "18  ...                               174  Eukaryota          Obazoa   \n",
       "19  ...                                 0  Eukaryota          Obazoa   \n",
       "\n",
       "        division     subdivision            class          order  \\\n",
       "0   Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "1   Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "2   Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "3   Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "4   Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "5      Alveolata  Dinoflagellata      Dinophyceae            NaN   \n",
       "6   Opisthokonta             NaN              NaN            NaN   \n",
       "7      Alveolata  Dinoflagellata      Dinophyceae  Gymnodiniales   \n",
       "8   Opisthokonta         Metazoa         Cnidaria     Cnidaria_X   \n",
       "9   Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "10  Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "11      Rhizaria      Radiolaria     Polycystinea    Spumellaria   \n",
       "12   Chlorophyta   Chlorophyta_X  Mamiellophyceae    Mamiellales   \n",
       "13     Alveolata  Dinoflagellata      Dinophyceae            NaN   \n",
       "14  Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "15     Alveolata  Dinoflagellata      Dinophyceae  Gymnodiniales   \n",
       "16  Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "17  Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "18  Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "19  Opisthokonta         Metazoa       Arthropoda      Crustacea   \n",
       "\n",
       "             family          genus                     species  \n",
       "0       Maxillopoda     Neocalanus        Neocalanus_cristatus  \n",
       "1       Maxillopoda  Clausocalanus      Clausocalanus_furcatus  \n",
       "2       Maxillopoda    Sinocalanus        Sinocalanus_sinensis  \n",
       "3       Maxillopoda    Paracalanus          Paracalanus_parvus  \n",
       "4       Maxillopoda    Acrocalanus             Acrocalanus_sp.  \n",
       "5               NaN            NaN                         NaN  \n",
       "6               NaN            NaN                         NaN  \n",
       "7    Gymnodiniaceae            NaN                         NaN  \n",
       "8          Hydrozoa  Sulculeolaria  Sulculeolaria_quadrivalvis  \n",
       "9       Maxillopoda        Delibus                 Delibus_sp.  \n",
       "10      Maxillopoda      Corycaeus         Corycaeus_speciosus  \n",
       "11  Astrosphaeridae   Heliosphaera            Heliosphaera_sp.  \n",
       "12   Bathycoccaceae   Ostreococcus                         NaN  \n",
       "13              NaN            NaN                         NaN  \n",
       "14      Maxillopoda            NaN                         NaN  \n",
       "15   Gymnodiniaceae     Gyrodinium                         NaN  \n",
       "16      Maxillopoda       Metridia              Metridia_longa  \n",
       "17      Maxillopoda    Paracalanus          Paracalanus_parvus  \n",
       "18      Maxillopoda        Delibus               Delibus_nudus  \n",
       "19      Maxillopoda      Corycaeus         Corycaeus_speciosus  \n",
       "\n",
       "[20 rows x 485 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# replace None with NA\n",
    "asv_tables['18S rRNA'] = asv_tables['18S rRNA'].fillna(value=np.nan)\n",
    "## Replace 'unknown', 'unassigned', etc. in species and taxonomy columns with NaN\n",
    "\n",
    "asv_tables['18S rRNA'][taxa_ranks_18S] = asv_tables['18S rRNA'][taxa_ranks_18S].replace({'unassigned':np.nan,\n",
    "                            'Unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan,\n",
    "                               '':np.nan})\n",
    "asv_tables['18S rRNA'].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace _,- with space, remove sp. \n",
    "\n",
    "asv_tables['18S rRNA'][taxa_ranks_18S] = asv_tables['18S rRNA'][taxa_ranks_18S].replace('_',' ',regex=True)\n",
    "asv_tables['18S rRNA'][taxa_ranks_18S] = asv_tables['18S rRNA'][taxa_ranks_18S].replace(' sp\\.','',regex=True)\n",
    "asv_tables['18S rRNA'][taxa_ranks_18S] = asv_tables['18S rRNA'][taxa_ranks_18S].replace(' spp\\.','',regex=True)\n",
    "asv_tables['18S rRNA'][taxa_ranks_18S] = asv_tables['18S rRNA'][taxa_ranks_18S].replace('-',' ',regex=True)\n",
    "asv_tables['18S rRNA'][taxa_ranks_18S] = asv_tables['18S rRNA'][taxa_ranks_18S].replace('\\/',' ',regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24067, 485)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['18S rRNA'].shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ['18S rRNA'] = pd.melt(asv_tables['18S rRNA'],id_vars=['featureid','sequence','taxonomy','Confidence','domain','supergroup','division','subdivision','class','order','family','genus','species'],\n",
    "               var_name='eventID',value_name='organismQuantity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11359624, 15)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S rRNA'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(146232, 15)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)\n",
    "\n",
    "occ['18S rRNA'] = occ['18S rRNA'][occ['18S rRNA']['organismQuantity'] > 0]\n",
    "print(occ['18S rRNA'].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>subdivision</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "5  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "7  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...   \n",
       "1  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...   \n",
       "4  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...   \n",
       "5  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "7  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...   \n",
       "\n",
       "                                            taxonomy  Confidence     domain  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.922099  Eukaryota   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.999947  Eukaryota   \n",
       "4  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.779948  Eukaryota   \n",
       "5  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.999931  Eukaryota   \n",
       "7  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.986908  Eukaryota   \n",
       "\n",
       "  supergroup      division     subdivision        class          order  \\\n",
       "0     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "1     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "4     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "5       TSAR     Alveolata  Dinoflagellata  Dinophyceae            NaN   \n",
       "7       TSAR     Alveolata  Dinoflagellata  Dinophyceae  Gymnodiniales   \n",
       "\n",
       "           family          genus                 species  \\\n",
       "0     Maxillopoda     Neocalanus    Neocalanus cristatus   \n",
       "1     Maxillopoda  Clausocalanus  Clausocalanus furcatus   \n",
       "4     Maxillopoda    Acrocalanus             Acrocalanus   \n",
       "5             NaN            NaN                     NaN   \n",
       "7  Gymnodiniaceae            NaN                     NaN   \n",
       "\n",
       "                  eventID  organismQuantity  \n",
       "0  GOMECC4_27N_Sta1_DCM_A              1516  \n",
       "1  GOMECC4_27N_Sta1_DCM_A               962  \n",
       "4  GOMECC4_27N_Sta1_DCM_A              1164  \n",
       "5  GOMECC4_27N_Sta1_DCM_A               287  \n",
       "7  GOMECC4_27N_Sta1_DCM_A               250  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S rRNA'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add occurenceID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create an occurrenceID that will uniquely identify each ASV observed within a water sample\n",
    "\n",
    "occ['18S rRNA']['occurrenceID'] = occ['18S rRNA']['featureid']\n",
    "occ['18S rRNA']['occurrenceID'] = occ['18S rRNA']['eventID'] + '_occ' + occ['18S rRNA']['occurrenceID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>subdivision</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "5  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "7  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...   \n",
       "1  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...   \n",
       "4  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...   \n",
       "5  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "7  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...   \n",
       "\n",
       "                                            taxonomy  Confidence     domain  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.922099  Eukaryota   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.999947  Eukaryota   \n",
       "4  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.779948  Eukaryota   \n",
       "5  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.999931  Eukaryota   \n",
       "7  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.986908  Eukaryota   \n",
       "\n",
       "  supergroup      division     subdivision        class          order  \\\n",
       "0     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "1     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "4     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "5       TSAR     Alveolata  Dinoflagellata  Dinophyceae            NaN   \n",
       "7       TSAR     Alveolata  Dinoflagellata  Dinophyceae  Gymnodiniales   \n",
       "\n",
       "           family          genus                 species  \\\n",
       "0     Maxillopoda     Neocalanus    Neocalanus cristatus   \n",
       "1     Maxillopoda  Clausocalanus  Clausocalanus furcatus   \n",
       "4     Maxillopoda    Acrocalanus             Acrocalanus   \n",
       "5             NaN            NaN                     NaN   \n",
       "7  Gymnodiniaceae            NaN                     NaN   \n",
       "\n",
       "                  eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A               962   \n",
       "4  GOMECC4_27N_Sta1_DCM_A              1164   \n",
       "5  GOMECC4_27N_Sta1_DCM_A               287   \n",
       "7  GOMECC4_27N_Sta1_DCM_A               250   \n",
       "\n",
       "                                        occurrenceID  \n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...  \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...  \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...  \n",
       "5  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...  \n",
       "7  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S rRNA'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 16S"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### 1st, format ASV file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### drop unwanted samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_tables['16S rRNA'] = asv_tables['16S rRNA'].drop(columns=samples_to_drop,errors='ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>...</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta100_Surface_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_DCM_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_DCM_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_DCM_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Deep_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Deep_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Deep_C</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o_...</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 476 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...   \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCG...   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "\n",
       "                                            taxonomy  Confidence  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.978926   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.881698   \n",
       "2  d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...    0.762793   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o_...    0.997619   \n",
       "4  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.999961   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  GOMECC4_27N_Sta1_DCM_C  \\\n",
       "0                       0                       0                       0   \n",
       "1                       0                       0                       0   \n",
       "2                       0                       0                       0   \n",
       "3                       0                       0                       0   \n",
       "4                       0                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  GOMECC4_27N_Sta1_Deep_C  \\\n",
       "0                        0                        0                        0   \n",
       "1                        0                        0                        0   \n",
       "2                        0                        0                        0   \n",
       "3                        0                        0                        0   \n",
       "4                        0                        0                        0   \n",
       "\n",
       "   ...  GOMECC4_YUCATAN_Sta100_Surface_C  GOMECC4_YUCATAN_Sta102_DCM_A  \\\n",
       "0  ...                                 5                             0   \n",
       "1  ...                                 0                             0   \n",
       "2  ...                                 0                             0   \n",
       "3  ...                                 0                             0   \n",
       "4  ...                                 0                             0   \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_DCM_B  GOMECC4_YUCATAN_Sta102_DCM_C  \\\n",
       "0                             0                             0   \n",
       "1                             0                             0   \n",
       "2                             0                             0   \n",
       "3                             0                             0   \n",
       "4                             0                             0   \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Deep_A  GOMECC4_YUCATAN_Sta102_Deep_B  \\\n",
       "0                              0                              0   \n",
       "1                              0                              0   \n",
       "2                              0                              0   \n",
       "3                              0                              0   \n",
       "4                              0                              0   \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Deep_C  GOMECC4_YUCATAN_Sta102_Surface_A  \\\n",
       "0                              0                                11   \n",
       "1                              0                                 0   \n",
       "2                              0                                 0   \n",
       "3                              0                                 0   \n",
       "4                              0                                 0   \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Surface_B  GOMECC4_YUCATAN_Sta102_Surface_C  \n",
       "0                                 0                                 7  \n",
       "1                                 0                                 0  \n",
       "2                                 0                                 0  \n",
       "3                                 0                                 0  \n",
       "4                                 0                                 0  \n",
       "\n",
       "[5 rows x 476 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S rRNA'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Vibrionales; f__Vibrionaceae; g__Vibrio'"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S rRNA']['taxonomy'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxa_ranks_16S = ['domain','phylum','class','order','family','genus','species']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>...</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_C</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Vibrionales</td>\n",
       "      <td>f__Vibrionaceae</td>\n",
       "      <td>g__Vibrio</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__HOC36</td>\n",
       "      <td>f__HOC36</td>\n",
       "      <td>g__HOC36</td>\n",
       "      <td>s__Candidatus_Thioglobus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Cyanobacteria</td>\n",
       "      <td>c__Cyanobacteriia</td>\n",
       "      <td>o__Synechococcales</td>\n",
       "      <td>f__Cyanobiaceae</td>\n",
       "      <td>g__Cyanobium_PCC-6307</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o_...</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__Myxococcia</td>\n",
       "      <td>o__Myxococcales</td>\n",
       "      <td>f__Myxococcaceae</td>\n",
       "      <td>g__P3OB-42</td>\n",
       "      <td>s__uncultured_bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__SAR86_clade</td>\n",
       "      <td>f__SAR86_clade</td>\n",
       "      <td>g__SAR86_clade</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 483 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...   \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCG...   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "\n",
       "                                            taxonomy  Confidence  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.978926   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.881698   \n",
       "2  d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...    0.762793   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o_...    0.997619   \n",
       "4  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.999961   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  GOMECC4_27N_Sta1_DCM_C  \\\n",
       "0                       0                       0                       0   \n",
       "1                       0                       0                       0   \n",
       "2                       0                       0                       0   \n",
       "3                       0                       0                       0   \n",
       "4                       0                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  GOMECC4_27N_Sta1_Deep_C  \\\n",
       "0                        0                        0                        0   \n",
       "1                        0                        0                        0   \n",
       "2                        0                        0                        0   \n",
       "3                        0                        0                        0   \n",
       "4                        0                        0                        0   \n",
       "\n",
       "   ...  GOMECC4_YUCATAN_Sta102_Surface_A  GOMECC4_YUCATAN_Sta102_Surface_B  \\\n",
       "0  ...                                11                                 0   \n",
       "1  ...                                 0                                 0   \n",
       "2  ...                                 0                                 0   \n",
       "3  ...                                 0                                 0   \n",
       "4  ...                                 0                                 0   \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Surface_C       domain             phylum  \\\n",
       "0                                 7  d__Bacteria  p__Proteobacteria   \n",
       "1                                 0  d__Bacteria  p__Proteobacteria   \n",
       "2                                 0  d__Bacteria   p__Cyanobacteria   \n",
       "3                                 0  d__Bacteria     p__Myxococcota   \n",
       "4                                 0  d__Bacteria  p__Proteobacteria   \n",
       "\n",
       "                    class               order            family  \\\n",
       "0  c__Gammaproteobacteria      o__Vibrionales   f__Vibrionaceae   \n",
       "1  c__Gammaproteobacteria            o__HOC36          f__HOC36   \n",
       "2       c__Cyanobacteriia  o__Synechococcales   f__Cyanobiaceae   \n",
       "3           c__Myxococcia     o__Myxococcales  f__Myxococcaceae   \n",
       "4  c__Gammaproteobacteria      o__SAR86_clade    f__SAR86_clade   \n",
       "\n",
       "                   genus                   species  \n",
       "0              g__Vibrio                      None  \n",
       "1               g__HOC36  s__Candidatus_Thioglobus  \n",
       "2  g__Cyanobium_PCC-6307                      None  \n",
       "3             g__P3OB-42   s__uncultured_bacterium  \n",
       "4         g__SAR86_clade                      None  \n",
       "\n",
       "[5 rows x 483 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S rRNA'][['domain','phylum','class','order','family','genus','species']] = asv_tables['16S rRNA']['taxonomy'].str.split(\"; \",expand=True)\n",
    "asv_tables['16S rRNA'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_tables['16S rRNA']['domain'] = asv_tables['16S rRNA']['domain'].str.replace(\"d__\", \"\")\n",
    "asv_tables['16S rRNA']['phylum'] = asv_tables['16S rRNA']['phylum'].str.replace(\"p__\", \"\")\n",
    "asv_tables['16S rRNA']['class'] = asv_tables['16S rRNA']['class'].str.replace(\"c__\", \"\")\n",
    "asv_tables['16S rRNA']['order'] = asv_tables['16S rRNA']['order'].str.replace(\"o__\", \"\")\n",
    "asv_tables['16S rRNA']['family'] = asv_tables['16S rRNA']['family'].str.replace(\"f__\", \"\")\n",
    "asv_tables['16S rRNA']['genus'] = asv_tables['16S rRNA']['genus'].str.replace(\"g__\", \"\")\n",
    "asv_tables['16S rRNA']['species'] = asv_tables['16S rRNA']['species'].str.replace(\"s__\", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_DCM_C</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_A</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_B</th>\n",
       "      <th>GOMECC4_27N_Sta1_Deep_C</th>\n",
       "      <th>...</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_A</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_B</th>\n",
       "      <th>GOMECC4_YUCATAN_Sta102_Surface_C</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00006f0784f7dbb2f162408abb6da629</td>\n",
       "      <td>TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.978926</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000094731d4984ed41435a1bf65b7ef2</td>\n",
       "      <td>TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.881698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>HOC36</td>\n",
       "      <td>HOC36</td>\n",
       "      <td>HOC36</td>\n",
       "      <td>Candidatus_Thioglobus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0001a3c11fcef1b1b8f4c72942efbbac</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>0.762793</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Cyanobium_PCC-6307</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0001ceef5162e6d689ef30418cfcc164</td>\n",
       "      <td>TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o_...</td>\n",
       "      <td>0.997619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>Myxococcia</td>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>P3OB-42</td>\n",
       "      <td>uncultured_bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000235534662df05bb30219a4b978dac</td>\n",
       "      <td>TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>SAR86_clade</td>\n",
       "      <td>SAR86_clade</td>\n",
       "      <td>SAR86_clade</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 483 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00006f0784f7dbb2f162408abb6da629   \n",
       "1  000094731d4984ed41435a1bf65b7ef2   \n",
       "2  0001a3c11fcef1b1b8f4c72942efbbac   \n",
       "3  0001ceef5162e6d689ef30418cfcc164   \n",
       "4  000235534662df05bb30219a4b978dac   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...   \n",
       "1  TACAGAGAGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "2  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "3  TACAGAGGGTGCAAGCGTTGTTCGGAATCATTGGGCGTAAAGCGCG...   \n",
       "4  TACGGAAGGTCCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "\n",
       "                                            taxonomy  Confidence  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.978926   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.881698   \n",
       "2  d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...    0.762793   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o_...    0.997619   \n",
       "4  d__Bacteria; p__Proteobacteria; c__Gammaproteo...    0.999961   \n",
       "\n",
       "   GOMECC4_27N_Sta1_DCM_A  GOMECC4_27N_Sta1_DCM_B  GOMECC4_27N_Sta1_DCM_C  \\\n",
       "0                       0                       0                       0   \n",
       "1                       0                       0                       0   \n",
       "2                       0                       0                       0   \n",
       "3                       0                       0                       0   \n",
       "4                       0                       0                       0   \n",
       "\n",
       "   GOMECC4_27N_Sta1_Deep_A  GOMECC4_27N_Sta1_Deep_B  GOMECC4_27N_Sta1_Deep_C  \\\n",
       "0                        0                        0                        0   \n",
       "1                        0                        0                        0   \n",
       "2                        0                        0                        0   \n",
       "3                        0                        0                        0   \n",
       "4                        0                        0                        0   \n",
       "\n",
       "   ...  GOMECC4_YUCATAN_Sta102_Surface_A  GOMECC4_YUCATAN_Sta102_Surface_B  \\\n",
       "0  ...                                11                                 0   \n",
       "1  ...                                 0                                 0   \n",
       "2  ...                                 0                                 0   \n",
       "3  ...                                 0                                 0   \n",
       "4  ...                                 0                                 0   \n",
       "\n",
       "   GOMECC4_YUCATAN_Sta102_Surface_C    domain          phylum  \\\n",
       "0                                 7  Bacteria  Proteobacteria   \n",
       "1                                 0  Bacteria  Proteobacteria   \n",
       "2                                 0  Bacteria   Cyanobacteria   \n",
       "3                                 0  Bacteria     Myxococcota   \n",
       "4                                 0  Bacteria  Proteobacteria   \n",
       "\n",
       "                 class            order         family               genus  \\\n",
       "0  Gammaproteobacteria      Vibrionales   Vibrionaceae              Vibrio   \n",
       "1  Gammaproteobacteria            HOC36          HOC36               HOC36   \n",
       "2       Cyanobacteriia  Synechococcales   Cyanobiaceae  Cyanobium_PCC-6307   \n",
       "3           Myxococcia     Myxococcales  Myxococcaceae             P3OB-42   \n",
       "4  Gammaproteobacteria      SAR86_clade    SAR86_clade         SAR86_clade   \n",
       "\n",
       "                 species  \n",
       "0                    NaN  \n",
       "1  Candidatus_Thioglobus  \n",
       "2                    NaN  \n",
       "3   uncultured_bacterium  \n",
       "4                    NaN  \n",
       "\n",
       "[5 rows x 483 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# replace None with NA\n",
    "asv_tables['16S rRNA'] = asv_tables['16S rRNA'].fillna(value=np.nan)\n",
    "## Replace 'unknown', 'unassigned', etc. in species and taxonomy columns with NaN\n",
    "\n",
    "asv_tables['16S rRNA'][taxa_ranks_16S] = asv_tables['16S rRNA'][taxa_ranks_16S].replace({'unassigned':np.nan,'Unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan,\n",
    "                               '':np.nan})\n",
    "asv_tables['16S rRNA'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace _,- with space, remove sp. \n",
    "\n",
    "asv_tables['16S rRNA'][taxa_ranks_16S] = asv_tables['16S rRNA'][taxa_ranks_16S].replace('_',' ',regex=True)\n",
    "asv_tables['16S rRNA'][taxa_ranks_16S] = asv_tables['16S rRNA'][taxa_ranks_16S].replace(' sp\\.','',regex=True)\n",
    "asv_tables['16S rRNA'][taxa_ranks_16S] = asv_tables['16S rRNA'][taxa_ranks_16S].replace('-',' ',regex=True)\n",
    "asv_tables['16S rRNA'][taxa_ranks_16S] = asv_tables['16S rRNA'][taxa_ranks_16S].replace(' spp\\.','',regex=True)\n",
    "asv_tables['16S rRNA'][taxa_ranks_16S] = asv_tables['16S rRNA'][taxa_ranks_16S].replace('\\/',' ',regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Melt asv_tables to long format\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65048, 483)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_tables['16S rRNA'].shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ['16S rRNA'] = pd.melt(asv_tables['16S rRNA'],id_vars=['featureid','sequence','taxonomy','Confidence','domain','phylum','class','order','family','genus','species'],\n",
    "               var_name='eventID',value_name='organismQuantity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30702656, 13)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S rRNA'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(165158, 13)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)\n",
    "\n",
    "occ['16S rRNA'] = occ['16S rRNA'][occ['16S rRNA']['organismQuantity'] > 0]\n",
    "print(occ['16S rRNA'].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create an occurrenceID that will uniquely identify each ASV observed within a water sample\n",
    "\n",
    "occ['16S rRNA']['occurrenceID'] = occ['16S rRNA']['featureid']\n",
    "occ['16S rRNA']['occurrenceID'] = occ['16S rRNA']['eventID'] + '_16S_occ' + occ['16S rRNA']['occurrenceID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Parvibaculales</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>uncultured marine</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Puniceispirillales</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Thermoplasmatota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            featureid  \\\n",
       "182  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "225  00e6c13fe86364a5084987093afa1916   \n",
       "347  015dad1fafca90944d905beb2a980bc3   \n",
       "412  019c88c6ade406f731954f38e3461564   \n",
       "719  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                              sequence  \\\n",
       "182  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "225  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "347  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...   \n",
       "412  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...   \n",
       "719  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...   \n",
       "\n",
       "                                              taxonomy  Confidence    domain  \\\n",
       "182  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.832190  Bacteria   \n",
       "225  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.867040  Bacteria   \n",
       "347  d__Archaea; p__Thermoplasmatota; c__Thermoplas...    1.000000   Archaea   \n",
       "412  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.952911  Bacteria   \n",
       "719  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...    0.818195  Bacteria   \n",
       "\n",
       "                            phylum                          class  \\\n",
       "182                 Proteobacteria            Alphaproteobacteria   \n",
       "225                 Proteobacteria            Alphaproteobacteria   \n",
       "347               Thermoplasmatota                 Thermoplasmata   \n",
       "412                 Proteobacteria            Alphaproteobacteria   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                             order                         family  \\\n",
       "182                 Parvibaculales                   OCS116 clade   \n",
       "225             Puniceispirillales                   SAR116 clade   \n",
       "347                Marine Group II                Marine Group II   \n",
       "412                  Rickettsiales                   Mitochondria   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                             genus               species  \\\n",
       "182                   OCS116 clade     uncultured marine   \n",
       "225                   SAR116 clade                   NaN   \n",
       "347                Marine Group II                   NaN   \n",
       "412                   Mitochondria  uncultured bacterium   \n",
       "719  Marinimicrobia (SAR406 clade)  uncultured bacterium   \n",
       "\n",
       "                    eventID  organismQuantity  \\\n",
       "182  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "225  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "347  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "412  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "719  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                          occurrenceID  \n",
       "182  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...  \n",
       "225  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...  \n",
       "347  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...  \n",
       "412  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...  \n",
       "719  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S rRNA'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### WORMS conversion. \n",
    "Note, can't use multiprocessing in a Jupyter notebook, need multiprocess. See [here](https://stackoverflow.com/questions/41385708/multiprocessing-example-giving-attributeerror)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data providers for this dataset used the [NCBI taxonomy database](https://www.ncbi.nlm.nih.gov/taxonomy) as their reference database when assigning taxonomies to ASVs. **It's important to note** that this taxonomy database is not a taxonomic authority, and the taxonomic ranks for any given scientific name on WoRMS may not directly compare. There are ongoing discussions about this problem (see [this](https://github.com/iobis/Project-team-Genetic-Data/issues/5) GitHub issue). At the moment, I don't see a way to definitively ensure that a given scientific name actually has the same taxonomic ranks on both platforms without going case-by-case.\n",
    "\n",
    "In addition, there are still names in the data that will not match on WoRMS at all, despite appearing to be Linnaean names. This is because the name may not have been fully and officially adopted by the scientific community. I therefore need a system for searching through the higher taxonomic ranks given, finding the lowest one that will match on WoRMS, and putting that name in the `scientificName` column. The following few code blocks do this - they're clunky, but they were sufficient for this data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "import multiprocess as mp\n",
    "import pyworms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_worms_from_scientific_name(tax_df, ordered_rank_columns, queue,full_tax_column=\"taxonomy\", like=False, marine_only=False,full_tax_vI = False):\n",
    "    matches = []\n",
    "    w_ranks = ['kingdom','phylum','class','order','family','genus']\n",
    "    for index, row in tax_df.iterrows():\n",
    "        full_tax = row[full_tax_column]\n",
    "        if full_tax_vI:\n",
    "            row_data = {'full_tax':full_tax,'verbatimIdentification': full_tax}\n",
    "        else:   \n",
    "            row_data = {'full_tax':full_tax,'verbatimIdentification': 'Null'}\n",
    "        for i in ordered_rank_columns:\n",
    "            rank = i\n",
    "            old_name = row[i]\n",
    "            if pd.isna(old_name):\n",
    "                continue \n",
    "            else:\n",
    "                row_data.update({'old_taxonRank': rank, 'old name': old_name})\n",
    "                if row_data['verbatimIdentification'] == 'Null':\n",
    "                    row_data['verbatimIdentification'] = old_name\n",
    "                s_match = pyworms.aphiaRecordsByName(old_name,like=like,marine_only=marine_only)\n",
    "                #time.sleep(1)\n",
    "                if s_match == None:\n",
    "                    row_data['scientificName'] = \"No match\"\n",
    "                    row_data['scientificNameID'] = \"None\"\n",
    "                    print(old_name+\": No match, \"+rank)\n",
    "                    continue\n",
    "                elif len(s_match) > 1:\n",
    "                    mult = []\n",
    "                    for m in s_match:\n",
    "                        if m['status'] == 'accepted':\n",
    "                            mult = mult + [m]\n",
    "                    if len(mult) > 1:\n",
    "                        row_data['scientificName'] = \"Multiple matches\"\n",
    "                        row_data['scientificNameID'] = \"None\"\n",
    "                        print(old_name+\": Multiple matches, \"+rank+\" \")\n",
    "                    elif len(mult) < 1:\n",
    "                        row_data['scientificName'] = \"Multiple unaccepted matches\"\n",
    "                        row_data['scientificNameID'] = \"None\"\n",
    "                        print(old_name+\": Multiple unaccepted matches, \"+rank+\" \")\n",
    "                    elif len(mult) == 1:\n",
    "                        row_data['scientificName'] = mult[0]['scientificname']\n",
    "                        row_data['scientificNameID'] = mult[0]['lsid']\n",
    "                        row_data.update(dict(zip(w_ranks, [mult[0].get(key) for key in w_ranks])))\n",
    "                        row_data.update({'taxonRank': mult[0]['rank']})\n",
    "                        break\n",
    "                elif len(s_match) == 1:\n",
    "                    if s_match[0]['status'] == 'accepted':\n",
    "                        row_data['scientificName'] = s_match[0]['scientificname']\n",
    "                        row_data['scientificNameID'] = s_match[0]['lsid']\n",
    "                        row_data.update(dict(zip(w_ranks, [s_match[0].get(key) for key in w_ranks])))\n",
    "                        row_data.update({'taxonRank': s_match[0]['rank']})\n",
    "                        break\n",
    "                    elif s_match[0]['status'] == 'unaccepted':\n",
    "                        valid_name = s_match[0]['valid_name']\n",
    "                        if valid_name != None:\n",
    "                            v_match = pyworms.aphiaRecordsByName(valid_name,like=like,marine_only=marine_only)\n",
    "                            row_data['scientificName'] = v_match[0]['scientificname']\n",
    "                            row_data['scientificNameID'] = v_match[0]['lsid']\n",
    "                            row_data.update(dict(zip(w_ranks, [v_match[0].get(key) for key in w_ranks])))\n",
    "                            row_data.update({'taxonRank': v_match[0]['rank']})\n",
    "                            print(old_name+\": Unaccepted, using \"+valid_name+\", \"+rank+\" \")\n",
    "                        else:\n",
    "                            print(old_name+\": Unaccepted, no valid name, \"+rank+\" \")\n",
    "        matches += [row_data]\n",
    "    matches = pd.DataFrame.from_dict(matches)\n",
    "    queue.put(matches)\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_worms_from_scientific_name_parallel(tax_df, ordered_rank_columns, full_tax_column=\"taxonomy\",like=False, marine_only=False,full_tax_vI = False,n_proc=0):\n",
    "    queue = mp.Queue()\n",
    "    if n_proc == 0:\n",
    "    # create as many processes as there are CPUs on your machine\n",
    "        num_processes = mp.cpu_count()\n",
    "    else:\n",
    "        num_processes = n_proc\n",
    "        \n",
    "    # calculate the chunk size as an integer\n",
    "    chunk_size = int(tax_df.shape[0]/num_processes)\n",
    "    procs = []\n",
    "    for job in range(num_processes):\n",
    "        start = job * chunk_size\n",
    "        end = start + chunk_size\n",
    "        df_chunk = tax_df.iloc[start:end]\n",
    "        proc = mp.Process(\n",
    "            target=get_worms_from_scientific_name,\n",
    "            args=(df_chunk,ordered_rank_columns, queue,full_tax_column,like,marine_only,full_tax_vI)\n",
    "        )\n",
    "        procs.append(proc)\n",
    "        proc.start()\n",
    "    \n",
    "    new_df = pd.DataFrame()\n",
    "    for _ in procs:\n",
    "        new_df = pd.concat([new_df,queue.get()])\n",
    "    \n",
    "    #new_df = queue.get()\n",
    "    \n",
    "    for proc in procs:\n",
    "        proc.join()\n",
    "    \n",
    "    return new_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Had some [issues with the parallelization](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr) on Mac M1. Adding 'OBJC_DISABLE_INITIALIZE_FORK_SAFETY = YES' to .bash_profile and then [This](https://github.com/python/cpython/issues/74570) fixed it.   \n",
    "Try to run without the bash_profile fix LATER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"no_proxy\"]=\"*\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 16S worms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Species level IDs might be trash, [see here](https://forum.qiime2.org/t/processing-filtering-and-evaluating-the-silva-database-and-other-reference-sequence-data-with-rescript/15494), so look at genus and up."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_16S = asv_tables['16S rRNA'][['taxonomy','domain','phylum','class','order','family','genus','species']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_16S = tax_16S.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2729, 8)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_16S.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Arenicellaceae: No match, familyBlfdi19: No match, genus\n",
      "\n",
      "Blfdi19: No match, family\n",
      "Blfdi19: No match, order\n",
      "Mitochondria: No match, genus\n",
      "Mitochondria: No match, family\n",
      "uncultured: No match, genusuncultured: No match, genus\n",
      "\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Arenicellales: No match, orderPolyangia: No match, class\n",
      "\n",
      "Candidatus Tenderia: No match, genus\n",
      "HOC36: No match, genus\n",
      "HOC36: No match, family\n",
      "HOC36: No match, order\n",
      "vadinHA49: No match, genus\n",
      "Tenderiaceae: No match, familyvadinHA49: No match, familyMyxococcota: No match, phylumOM60(NOR5) clade: No match, genus\n",
      "\n",
      "\n",
      "\n",
      "vadinHA49: No match, order\n",
      "vadinHA49: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Cyanobium PCC 6307: No match, genus\n",
      "Halieaceae: No match, familyTenderiales: No match, orderZixibacteria: No match, genus\n",
      "\n",
      "\n",
      "Zixibacteria: No match, family\n",
      "Zixibacteria: No match, order\n",
      "Zixibacteria: No match, class\n",
      "Zixibacteria: No match, phylum\n",
      "Cellvibrionales: No match, order\n",
      "Cyanobiaceae: No match, family\n",
      "BD1 7 clade: No match, genus\n",
      "MSB 1D1: No match, genus\n",
      "vadinHA49: No match, genus\n",
      "vadinHA49: No match, familyChloroplast: No match, genus\n",
      "\n",
      "vadinHA49: No match, orderChloroplast: No match, family\n",
      "\n",
      "vadinHA49: No match, class\n",
      "Chloroplast: No match, order\n",
      "Spongiibacteraceae: No match, family\n",
      "Nitrosococcaceae: No match, familyPlanctomycetota: No match, phylum\n",
      "\n",
      "Cyanobacteriia: No match, class\n",
      "Chitinophagales: No match, orderCellvibrionales: No match, order\n",
      "\n",
      "P3OB 42: No match, genus\n",
      "Phaeodactylibacter: No match, genusNitrosococcales: No match, order\n",
      "\n",
      "Aquibacter: No match, genus\n",
      "uncultured: No match, genus\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Omnitrophales: No match, genus\n",
      "Omnitrophales: No match, family\n",
      "Omnitrophales: No match, order\n",
      "uncultured: No match, genus\n",
      "Methylophagaceae: No match, familyOmnitrophia: No match, classArenicellaceae: No match, family\n",
      "\n",
      "\n",
      "NS7 marine group: No match, genusLitoribacillus: No match, genusCandidatus Nitrosopelagicus: No match, genusNitrosococcales: No match, order\n",
      "\n",
      "\n",
      "\n",
      "NS7 marine group: No match, family\n",
      "Arenicellales: No match, orderuncultured: No match, genus\n",
      "\n",
      "Aliikangiella: No match, genusuncultured: No match, family\n",
      "\n",
      "uncultured: No match, order\n",
      "uncultured: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genusDesulfobacterota: No match, phylumKangiellaceae: No match, family\n",
      "\n",
      "\n",
      "uncultured: No match, family\n",
      "Schekmanbacteria: No match, genusSUP05 cluster: No match, genus\n",
      "\n",
      "Schekmanbacteria: No match, family\n",
      "Schekmanbacteria: No match, order\n",
      "Schekmanbacteria: No match, class\n",
      "Schekmanbacteria: No match, phylum\n",
      "Ardenticatenales: No match, order\n",
      "P.palmC41: No match, genus\n",
      "P.palmC41: No match, family\n",
      "Thioglobaceae: No match, familyP.palmC41: No match, order\n",
      "\n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, family\n",
      "Thiomicrospirales: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "uncultured: No match, genus\n",
      "SBR1031: No match, genus\n",
      "SBR1031: No match, family\n",
      "SBR1031: No match, order\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Omnitrophaceae: No match, family\n",
      "Cyanobacteriia: No match, class\n",
      "Omnitrophales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Prochlorococcus MIT9313: No match, genus\n",
      "Omnitrophia: No match, class\n",
      "Chloroplast: No match, genusCyanobiaceae: No match, family\n",
      "\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Cyanobacteriia: No match, classVerrucomicrobiota: No match, phylum\n",
      "\n",
      "Pla3 lineage: No match, genus\n",
      "Pla3 lineage: No match, family\n",
      "Pla3 lineage: No match, order\n",
      "Pla3 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "cvE6: No match, genus\n",
      "cvE6: No match, family\n",
      "MSB 1D1: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Micavibrionales: No match, order\n",
      "RCP2 54: No match, genus\n",
      "Candidatus Actinomarina: No match, genusRCP2 54: No match, family\n",
      "\n",
      "RCP2 54: No match, order\n",
      "Nitrosococcales: No match, orderRCP2 54: No match, class\n",
      "\n",
      "EV818SWSAP88: No match, genusRCP2 54: No match, phylum\n",
      "\n",
      "EV818SWSAP88: No match, family\n",
      "EV818SWSAP88: No match, order\n",
      "Aliiroseovarius: No match, genus\n",
      "uncultured: No match, genus\n",
      "Luteivirga: No match, genus\n",
      "PAUC26f: No match, genusuncultured: No match, genus\n",
      "\n",
      "SAR116 clade: No match, genusuncultured: No match, family\n",
      "PAUC26f: No match, family\n",
      "\n",
      "SAR116 clade: No match, familyPAUC26f: No match, order\n",
      "\n",
      "Actinomarinaceae: No match, family\n",
      "Amylibacter: No match, genus\n",
      "Puniceispirillales: No match, orderAcidobacteriae: No match, classArdenticatenales: No match, order\n",
      "\n",
      "\n",
      "Actinomarinales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "uncultured: No match, genus\n",
      "MD3 55: No match, genus\n",
      "Acidobacteriota: No match, phylumAcidimicrobiia: No match, class\n",
      "\n",
      "Leptospiraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Fokiniaceae: No match, family\n",
      "NB1 j: No match, genusCandidatus Kaiserbacteria: No match, genus\n",
      "\n",
      "Oligoflexales: No match, orderCandidatus Kaiserbacteria: No match, familyNB1 j: No match, family\n",
      "\n",
      "\n",
      "Candidatus Kaiserbacteria: No match, orderActinobacteriota: No match, phylumNB1 j: No match, order\n",
      "\n",
      "\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylumLeptospirales: No match, order\n",
      "\n",
      "Oligoflexia: No match, classParcubacteria: No match, class\n",
      "\n",
      "Leptospirae: No match, classQipengyuania: No match, genusCellvibrionaceae: No match, genus\n",
      "\n",
      "\n",
      "Cellvibrionaceae: No match, family\n",
      "BD2 11 terrestrial group: No match, genus\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "Spirochaetota: No match, phylum\n",
      "Patescibacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Bdellovibrionota: No match, phylum\n",
      "Cellvibrionales: No match, order\n",
      "uncultured: No match, genus\n",
      "Clade I: No match, family\n",
      "Gemmatimonadota: No match, phylum\n",
      "SAR11 clade: No match, order\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, familyDissulfuribacteraceae: No match, family\n",
      "\n",
      "Marine Group II: No match, order\n",
      "Marine Group III: No match, genus\n",
      "Nitrospira: Multiple matches, genus \n",
      "Dissulfuribacterales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Magnetospira: No match, genus\n",
      "Dissulfuribacteria: No match, classMarinimicrobia (SAR406 clade): No match, genusAB1: No match, genus\n",
      "\n",
      "\n",
      "Marinimicrobia (SAR406 clade): No match, familyAB1: No match, family\n",
      "\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "SAR11 clade: No match, order\n",
      "Magnetospiraceae: No match, family\n",
      "Obscuribacteraceae: No match, genus\n",
      "Obscuribacteraceae: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Desulfobacterota: No match, phylumChloroplast: No match, family\n",
      "\n",
      "PAUC34f: No match, genusuncultured: No match, genus\n",
      "Chloroplast: No match, order\n",
      "uncultured: No match, familyPAUC34f: No match, family\n",
      "\n",
      "\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "Obscuribacterales: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "SAR11 clade: No match, order\n",
      "MB11C04 marine group: No match, genus\n",
      "Vampirivibrionia: No match, class\n",
      "KD4 96: No match, genus\n",
      "KD4 96: No match, family\n",
      "KD4 96: No match, order\n",
      "KD4 96: No match, class\n",
      "Pedosphaeraceae: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "[Caedibacter] taeniospiralis group: No match, genus\n",
      "Pedosphaerales: No match, order\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "FS140 16B 02 marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "Vermiphilaceae: No match, genus\n",
      "Vermiphilaceae: No match, family\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Babeliales: No match, order\n",
      "NS7 marine group: No match, genus\n",
      "NS7 marine group: No match, family\n",
      "Bythopirellula: No match, genus\n",
      "Puniceispirillales: No match, order\n",
      "Coxiella: Multiple matches, genus \n",
      "Candidatus Peribacteria: No match, genus\n",
      "Candidatus Peribacteria: No match, family\n",
      "Candidatus Peribacteria: No match, order\n",
      "Pirellulaceae: No match, family\n",
      "Babeliae: No match, class\n",
      "JGI 0000069 P22: No match, genus\n",
      "JGI 0000069 P22: No match, family\n",
      "JGI 0000069 P22: No match, order\n",
      "Prochlorococcus MIT9313: No match, genusGracilibacteria: No match, class\n",
      "\n",
      "Pirellulales: No match, orderCandidatus Cryptoprodotis: No match, genusDependentiae: No match, phylum\n",
      "\n",
      "\n",
      "Gracilibacteria: No match, class\n",
      "Cyanobiaceae: No match, family\n",
      "Rubripirellula: No match, genus\n",
      "Patescibacteria: No match, phylum\n",
      "MVP 88: No match, genus\n",
      "MVP 88: No match, family\n",
      "MVP 88: No match, order\n",
      "Patescibacteria: No match, phylum\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Pirellulaceae: No match, family\n",
      "TK17: No match, genus\n",
      "TK17: No match, family\n",
      "TK17: No match, order\n",
      "TK17: No match, class\n",
      "TK10: No match, genus\n",
      "TK10: No match, family\n",
      "TK10: No match, order\n",
      "TK10: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Elusimicrobia: Multiple matches, class \n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "Pirellulales: No match, order\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Ilumatobacteraceae: No match, familyClade II: No match, genus\n",
      "\n",
      "Clade II: No match, family\n",
      "Elusimicrobiota: No match, phylum\n",
      "Sva0996 marine group: No match, genus\n",
      "SAR11 clade: No match, order\n",
      "Microtrichales: No match, order\n",
      "NS10 marine group: No match, genus\n",
      "NS2b marine group: No match, genus\n",
      "Microtrichaceae: No match, familyJTB23: No match, genus\n",
      "\n",
      "JTB23: No match, family\n",
      "JTB23: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Microtrichales: No match, order\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "Candidatus Terrybacteria: No match, genus\n",
      "Candidatus Terrybacteria: No match, family\n",
      "Prochlorococcus MIT9313: No match, genusCandidatus Terrybacteria: No match, order\n",
      "\n",
      "Actinobacteriota: No match, phylumPirellulaceae: No match, genus\n",
      "\n",
      "Pirellulaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Acidimicrobiia: No match, class\n",
      "SAR11 clade: No match, order\n",
      "Parcubacteria: No match, class\n",
      "Cyanobiaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "uncultured: No match, genus\n",
      "Actinobacteriota: No match, phylum\n",
      "Patescibacteria: No match, phylumuncultured: No match, genus\n",
      "\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Ilumatobacteraceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "JL ETNP F27: No match, genus\n",
      "Marinifilaceae: No match, family\n",
      "Oligoflexales: No match, order\n",
      "Nitrospinota: No match, genus\n",
      "Nitrospinota: No match, family\n",
      "Nitrospinota: No match, order\n",
      "Nitrospinota: No match, class\n",
      "Nitrospinota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Oligoflexia: No match, class\n",
      "mle1 8: No match, genus\n",
      "mle1 8: No match, family\n",
      "mle1 8: No match, order\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Limibaculum: No match, genus\n",
      "uncultured: No match, genus\n",
      "Bdellovibrionota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Limimaricola: No match, genus\n",
      "Clade IV: No match, genus\n",
      "Clade IV: No match, family\n",
      "Dehalococcoidia: No match, class\n",
      "Rubinisphaeraceae: No match, family\n",
      "Chitinophagales: No match, order\n",
      "Rubinisphaeraceae: No match, family\n",
      "Halobacteriovorax: No match, genus\n",
      "SAR11 clade: No match, order\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "uncultured: No match, genus\n",
      "WS2: No match, genus\n",
      "WS2: No match, family\n",
      "WS2: No match, order\n",
      "WS2: No match, class\n",
      "WS2: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Rubinisphaeraceae: No match, family\n",
      "Magnetospiraceae: No match, family\n",
      "Candidatus Megaira: No match, genus\n",
      "MB11C04 marine group: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Candidatus Amoebophilus: No match, genus\n",
      "OM43 clade: No match, genus\n",
      "OM182 clade: No match, genus\n",
      "OM182 clade: No match, family\n",
      "OM182 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Amoebophilaceae: No match, family\n",
      "Prochlorococcus MIT9313: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Gemmataceae: No match, family\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "HgCo23: No match, genus\n",
      "HgCo23: No match, family\n",
      "HgCo23: No match, order\n",
      "NS2b marine group: No match, genus\n",
      "Sva0996 marine group: No match, genus\n",
      "Gemmatales: No match, order\n",
      "Mucilaginibacter: No match, genusCandidatus Nitrosopumilus: No match, genus\n",
      "\n",
      "Omnitrophales: No match, genusSalinirepens: No match, genus\n",
      "\n",
      "Omnitrophales: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Clade Ia: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Crocinitomicaceae: No match, family\n",
      "Omnitrophia: No match, class\n",
      "Microtrichales: No match, order\n",
      "CL500 3: No match, genus\n",
      "Boseongicola: No match, genus\n",
      "Verrucomicrobiota: No match, phylum\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "Acidimicrobiia: No match, class\n",
      "Marinicella: No match, genus\n",
      "AKYG587: No match, genus\n",
      "Hydrogenedensaceae: No match, genus\n",
      "Hydrogenedensaceae: No match, family\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Pla4 lineage: No match, genus\n",
      "Pla4 lineage: No match, family\n",
      "Pla4 lineage: No match, order\n",
      "Pla4 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "uncultured: No match, genus\n",
      "Hydrogenedentiales: No match, order\n",
      "GW2011 GWC1 47 15: No match, genus\n",
      "GW2011 GWC1 47 15: No match, family\n",
      "Steroidobacteraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Cyanobacteriia: No match, class\n",
      "Subgroup 23: No match, genus\n",
      "Woesearchaeales: No match, order\n",
      "Flavicella: No match, genus\n",
      "Hydrogenedentia: No match, class\n",
      "Chloroplast: No match, genusMethyloligellaceae: No match, familySteroidobacterales: No match, order\n",
      "\n",
      "\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Thermoanaerobaculaceae: No match, family\n",
      "Thioglobaceae: No match, family\n",
      "Thiomicrospirales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Hydrogenedentes: No match, phylum\n",
      "Desulfuromonadia: No match, class\n",
      "OM182 clade: No match, genus\n",
      "OM182 clade: No match, family\n",
      "OM182 clade: No match, order\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Thermoanaerobaculales: No match, order\n",
      "Defluviicoccales: No match, order\n",
      "Vicingus: No match, genus\n",
      "Desulfobacterota: No match, phylum\n",
      "Subgroup 13: No match, genus\n",
      "Subgroup 13: No match, family\n",
      "Subgroup 13: No match, order\n",
      "Thermoanaerobaculia: No match, class\n",
      "PB19: No match, genus\n",
      "PB19: No match, family\n",
      "PB19: No match, order\n",
      "Heimdallarchaeia: No match, genus\n",
      "Heimdallarchaeia: No match, family\n",
      "Heimdallarchaeia: No match, order\n",
      "Heimdallarchaeia: No match, class\n",
      "0319 6G20: No match, genus\n",
      "0319 6G20: No match, family\n",
      "uncultured: No match, genus0319 6G20: No match, order\n",
      "\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Acidobacteriae: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Desulfuromonadia: No match, class\n",
      "Asgardarchaeota: No match, phylum\n",
      "Nitrincolaceae: No match, familyAT s2 59: No match, genus\n",
      "\n",
      "AT s2 59: No match, family\n",
      "AT s2 59: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Acidobacteriota: No match, phylum\n",
      "Desulfobacterota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Roseobacter clade NAC11 7 lineage: No match, genus\n",
      "OM27 clade: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "IMCC26256: No match, genus\n",
      "IMCC26256: No match, family\n",
      "IMCC26256: No match, order\n",
      "Motiliproteus: No match, genus\n",
      "Candidatus Fritschea: No match, genus\n",
      "uncultured: No match, genus\n",
      "Acidimicrobiia: No match, class\n",
      "Candidatus Woesebacteria: No match, genus\n",
      "Candidatus Woesebacteria: No match, family\n",
      "Candidatus Woesebacteria: No match, order\n",
      "Nitrincolaceae: No match, family\n",
      "SCGC AAA011 D5: No match, genus\n",
      "SCGC AAA011 D5: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "Actinobacteriota: No match, phylum\n",
      "Microgenomatia: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "NS5 marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "Taibaiella: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Arenicellaceae: No match, family\n",
      "Cyanobacteriia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Arenicellales: No match, order\n",
      "uncultured: No match, genus\n",
      "LS NOB: No match, genus\n",
      "uncultured: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Vicinamibacteraceae: No match, genus\n",
      "Vicinamibacteraceae: No match, family\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Woeseia: No match, genus\n",
      "Sandaracinaceae: No match, family\n",
      "Napoli 4B 65: No match, genus\n",
      "Napoli 4B 65: No match, family\n",
      "Napoli 4B 65: No match, order\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genusVicinamibacterales: No match, order\n",
      "\n",
      "bacteriap25: No match, genus\n",
      "bacteriap25: No match, family\n",
      "bacteriap25: No match, order\n",
      "bacteriap25: No match, class\n",
      "Woeseiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Polyangiales: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Nitrincolaceae: No match, family\n",
      "Vicinamibacteria: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Steroidobacterales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Polyangia: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Chitinophagales: No match, order\n",
      "Myxococcota: No match, phylum\n",
      "DEV007: No match, genus\n",
      "DEV007: No match, family\n",
      "uncultured: No match, genus\n",
      "DEV007: No match, genus\n",
      "DEV007: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Sva0996 marine group: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Thermodesulfovibrionia: No match, class\n",
      "uncultured: No match, genus\n",
      "Stappiaceae: No match, genus\n",
      "Stappiaceae: No match, family\n",
      "Phreatobacter: No match, genusMarinimicrobia (SAR406 clade): No match, genus\n",
      "\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Nitrospirota: No match, phylum\n",
      "Absconditabacteriales (SR1): No match, genus\n",
      "Absconditabacteriales (SR1): No match, family\n",
      "Absconditabacteriales (SR1): No match, order\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "SUP05 cluster: No match, genus\n",
      "Thioglobaceae: No match, family\n",
      "Thiomicrospirales: No match, order\n",
      "NS11 12 marine group: No match, genus\n",
      "NS11 12 marine group: No match, family\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "S25 593: No match, genus\n",
      "S25 593: No match, family\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "uncultured: No match, genus\n",
      "Subgroup 15: No match, genus\n",
      "Subgroup 15: No match, family\n",
      "Subgroup 15: No match, order\n",
      "Clade Ib: No match, genus\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "OM27 clade: No match, genus\n",
      "Acidobacteriae: No match, class\n",
      "koll11: No match, genus\n",
      "koll11: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "AqS1: No match, genusClade Ib: No match, genus\n",
      "\n",
      "Nitrosococcaceae: No match, familyClade I: No match, family\n",
      "\n",
      "Nitrosococcales: No match, orderSAR11 clade: No match, order\n",
      "\n",
      "HIMB11: No match, genus\n",
      "Acidobacteriota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Aliiroseovarius: No match, genus\n",
      "Sedimentitalea: No match, genus\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "uncultured: No match, genus\n",
      "AT s3 28: No match, genus\n",
      "AT s3 28: No match, family\n",
      "AT s3 28: No match, order\n",
      "AT s3 28: No match, class\n",
      "JTB23: No match, genus\n",
      "JTB23: No match, family\n",
      "JTB23: No match, order\n",
      "DEV007: No match, genus\n",
      "DEV007: No match, family\n",
      "Arctic95B 14: No match, genus\n",
      "Arctic95B 14: No match, family\n",
      "SM1A02: No match, genus\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Victivallales: No match, order\n",
      "Babeliales: No match, order\n",
      "AT s3 44: No match, genus\n",
      "AT s3 44: No match, family\n",
      "AT s3 44: No match, order\n",
      "uncultured: No match, genus\n",
      "Cyanobium PCC 6307: No match, genus\n",
      "Thiohalorhabdaceae: No match, familyLentisphaeria: No match, classuncultured: No match, genus\n",
      "\n",
      "\n",
      "Verrucomicrobiota: No match, phylumuncultured: No match, family\n",
      "\n",
      "Babeliae: No match, class\n",
      "Thiomicrorhabdus: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "Paracaedibacteraceae: No match, family\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Thiohalorhabdales: No match, order\n",
      "Dependentiae: No match, phylum\n",
      "Thiomicrospiraceae: No match, family\n",
      "KI89A clade: No match, genusParacaedibacterales: No match, order\n",
      "\n",
      "KI89A clade: No match, family\n",
      "NS7 marine group: No match, genusKI89A clade: No match, order\n",
      "\n",
      "NS7 marine group: No match, family\n",
      "PAUC26f: No match, genus\n",
      "PAUC26f: No match, family\n",
      "PAUC26f: No match, order\n",
      "Dadabacteriales: No match, genus\n",
      "Dadabacteriales: No match, family\n",
      "Dadabacteriales: No match, order\n",
      "Thiomicrospirales: No match, order\n",
      "Pseudohongiella: No match, genus\n",
      "Acidobacteriae: No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Prochlorococcus MIT9313: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "Dadabacteriia: No match, class\n",
      "PS1 clade: No match, genus\n",
      "PS1 clade: No match, family\n",
      "Acidobacteriota: No match, phylum\n",
      "Oligoflexia: No match, classPseudohongiellaceae: No match, family\n",
      "\n",
      "Bdellovibrionota: No match, phylum\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "BD7 11: No match, order\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Hydrogenedensaceae: No match, genus\n",
      "Hydrogenedensaceae: No match, family\n",
      "Dadabacteria: No match, phylum\n",
      "Parvibaculales: No match, order\n",
      "SGST604: No match, genus\n",
      "SGST604: No match, familyGa0077536: No match, genus\n",
      "\n",
      "SGST604: No match, orderGa0077536: No match, family\n",
      "\n",
      "Ga0077536: No match, orderSGST604: No match, class\n",
      "\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Hydrogenedentiales: No match, order\n",
      "Sva0485: No match, genus\n",
      "Sva0485: No match, family\n",
      "Sva0485: No match, order\n",
      "Sva0485: No match, class\n",
      "Sva0485: No match, phylum\n",
      "Unknown Family: No match, familyMarine Group III: No match, genus\n",
      "\n",
      "MB11C04 marine group: No match, genus\n",
      "Pseudorhodoferax: No match, genus\n",
      "Hydrogenedentia: No match, class\n",
      "UBA10353 marine group: No match, genusQipengyuania: No match, genus\n",
      "\n",
      "UBA10353 marine group: No match, family\n",
      "uncultured: No match, genusUBA10353 marine group: No match, order\n",
      "\n",
      "uncultured: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "Hydrogenedentes: No match, phylum\n",
      "Omnitrophales: No match, genus\n",
      "Omnitrophales: No match, familyPedosphaerales: No match, order\n",
      "\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Latescibacteraceae: No match, genus\n",
      "Latescibacteraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Bryobacter: No match, genus\n",
      "Latescibacterales: No match, order\n",
      "UBA12409: No match, genus\n",
      "UBA12409: No match, family\n",
      "Babeliales: No match, order\n",
      "Babeliae: No match, class\n",
      "Dependentiae: No match, phylum\n",
      "Oligoflexales: No match, order\n",
      "CCM11a: No match, genusCL500 3: No match, genus\n",
      "\n",
      "Bryobacteraceae: No match, familyLatescibacteria: No match, classCCM11a: No match, family\n",
      "\n",
      "\n",
      "CCM11a: No match, order\n",
      "Candidatus Abawacabacteria: No match, genus\n",
      "Candidatus Abawacabacteria: No match, family\n",
      "Candidatus Abawacabacteria: No match, order\n",
      "SAR92 clade: No match, genus\n",
      "Oligoflexia: No match, class\n",
      "Kistimonas: No match, genusSchleiferia: No match, genus\n",
      "\n",
      "Bryobacterales: No match, order\n",
      "Acidobacteriae: No match, classLatescibacterota: No match, phylum\n",
      "\n",
      "Acidobacteriota: No match, phylum\n",
      "Gracilibacteria: No match, class\n",
      "Porticoccaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Bdellovibrionota: No match, phylum\n",
      "Schleiferiaceae: No match, familySAR86 clade: No match, genus\n",
      "Candidatus Nitrosopumilus: No match, genus\n",
      "Endozoicomonadaceae: No match, family\n",
      "SAR86 clade: No match, family\n",
      "\n",
      "SAR86 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Patescibacteria: No match, phylum\n",
      "Polyangiales: No match, order\n",
      "SCGC AAA164 E04: No match, genus\n",
      "Gaiellales: No match, order\n",
      "Sericytochromatia: No match, genus\n",
      "Sericytochromatia: No match, family\n",
      "Sericytochromatia: No match, order\n",
      "Sericytochromatia: No match, class\n",
      "Polyangia: No match, class\n",
      "Pedosphaeraceae: No match, familyMyxococcota: No match, phylum\n",
      "\n",
      "uncultured: No match, genus\n",
      "Thermoleophilia: No match, classUnknown Family: No match, familySM1A02: No match, genus\n",
      "\n",
      "\n",
      "Pedosphaerales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Actinobacteriota: No match, phylum\n",
      "SUP05 cluster: No match, genus\n",
      "Candidatus Hepatoplasma: No match, genus\n",
      "uncultured: No match, genus\n",
      "Chitinophagales: No match, order\n",
      "Bradymonadales: No match, genus\n",
      "Entomoplasmatales Incertae Sedis: No match, familyBradymonadales: No match, family\n",
      "\n",
      "Bradymonadales: No match, order\n",
      "S085: No match, genusThioglobaceae: No match, family\n",
      "\n",
      "S085: No match, family\n",
      "S085: No match, order\n",
      "JGI 0000069 P22: No match, genus\n",
      "SAR324 clade(Marine group B): No match, genusJGI 0000069 P22: No match, family\n",
      "\n",
      "SAR324 clade(Marine group B): No match, familyJGI 0000069 P22: No match, order\n",
      "\n",
      "SAR324 clade(Marine group B): No match, orderGracilibacteria: No match, class\n",
      "\n",
      "SAR324 clade(Marine group B): No match, classPatescibacteria: No match, phylum\n",
      "\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "Desulfuromonadia: No match, classThiomicrospirales: No match, order\n",
      "\n",
      "Dehalococcoidia: No match, classuncultured: No match, genus\n",
      "\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, familyS25 593: No match, genus\n",
      "\n",
      "S25 593: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "P0X4b2H11: No match, genus\n",
      "Magnetospiraceae: No match, familyP0X4b2H11: No match, family\n",
      "\n",
      "P0X4b2H11: No match, order\n",
      "uncultured: No match, genuskoll11: No match, genus\n",
      "\n",
      "koll11: No match, family\n",
      "Deferrisoma: No match, genus\n",
      "Desulfobacterota: No match, phylum\n",
      "Pseudohongiella: No match, genus\n",
      "Vicinamibacterales: No match, order\n",
      "Candidatus Actinomarina: No match, genusCellvibrionaceae: No match, family\n",
      "\n",
      "Omnitrophales: No match, order\n",
      "uncultured: No match, genus\n",
      "Defferrisomataceae: No match, family\n",
      "Pseudohongiellaceae: No match, family\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Cellvibrionales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Actinomarinaceae: No match, familyDefferrisomatales: No match, order\n",
      "\n",
      "[Caedibacter] taeniospiralis group: No match, genus\n",
      "KF JG30 B3: No match, genus\n",
      "Actinomarinales: No match, orderCalditrichaceae: No match, genus\n",
      "KF JG30 B3: No match, family\n",
      "Acidimicrobiia: No match, class\n",
      "Calditrichaceae: No match, family\n",
      "Actinobacteriota: No match, phylum\n",
      "\n",
      "uncultured: No match, genus\n",
      "Defferrisomatia: No match, class\n",
      "EPR3968 O8a Bc78: No match, genus\n",
      "EPR3968 O8a Bc78: No match, family\n",
      "EPR3968 O8a Bc78: No match, order\n",
      "Sva0996 marine group: No match, genus\n",
      "Calditrichales: No match, order\n",
      "Luminiphilus: No match, genus\n",
      "Sumerlaea: No match, genus\n",
      "Deferrisomatota: No match, phylum\n",
      "Microtrichaceae: No match, family\n",
      "BD2 11 terrestrial group: No match, genus\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "Halieaceae: No match, familyCalditrichia: No match, classHIMB11: No match, genus\n",
      "\n",
      "\n",
      "Sumerlaeaceae: No match, familyCellvibrionales: No match, order\n",
      "\n",
      "Lentimonas: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "MB11C04 marine group: No match, genus\n",
      "SCGC AAA164 E04: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "Pedosphaerales: No match, order\n",
      "Sumerlaeales: No match, orderCalditrichota: No match, phylum\n",
      "\n",
      "SCGC AAA011 D5: No match, genus\n",
      "SCGC AAA011 D5: No match, family\n",
      "uncultured: No match, genus\n",
      "Phaeodactylibacter: No match, genusSumerlaeia: No match, class\n",
      "\n",
      "Woesearchaeales: No match, order\n",
      "uncultured: No match, genus\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Candidatus Actinomarina: No match, genus\n",
      "Actinomarinaceae: No match, family\n",
      "Actinomarinales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Sumerlaeota: No match, phylum\n",
      "Crocinitomicaceae: No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "HOC36: No match, genus\n",
      "HOC36: No match, family\n",
      "HOC36: No match, order\n",
      "Profundimonas: No match, genus\n",
      "Nitrincolaceae: No match, family\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "uncultured: No match, genus\n",
      "PS1 clade: No match, genus\n",
      "PS1 clade: No match, family\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Marine Benthic Group A: No match, genus\n",
      "Marine Benthic Group A: No match, family\n",
      "Marine Benthic Group A: No match, order\n",
      "PHOS HE36: No match, genus\n",
      "PHOS HE36: No match, family\n",
      "Iainarchaeales: No match, genus\n",
      "Iainarchaeales: No match, family\n",
      "Iainarchaeales: No match, order\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "MB11C04 marine group: No match, genus\n",
      "PB19: No match, genus\n",
      "PB19: No match, family\n",
      "PB19: No match, order\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Parvibaculales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Nitrososphaeria: No match, class\n",
      "Omnitrophaceae: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Ignavibacteriales: No match, order\n",
      "Maricurvus: No match, genus\n",
      "Iainarchaeia: No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Cyanobiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "AKYH767: No match, genusIgnavibacteria: No match, class\n",
      "\n",
      "AKYH767: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Lentimonas: No match, genus\n",
      "uncultured: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "Cellvibrionaceae: No match, familyBD7 11: No match, order\n",
      "\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Iainarchaeota: No match, phylum\n",
      "WCHB1 41: No match, genus\n",
      "WCHB1 41: No match, family\n",
      "WCHB1 41: No match, order\n",
      "Gimesiaceae: No match, family\n",
      "Bacteroidota: No match, phylum\n",
      "Cellvibrionales: No match, orderNS5 marine group: No match, genus\n",
      "\n",
      "Kiritimatiellae: No match, class\n",
      "Candidatus Endobugula: No match, genus\n",
      "Cyanobium PCC 6307: No match, genusCellvibrionaceae: No match, family\n",
      "\n",
      "Cellvibrionales: No match, order\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Cm1 21: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "Yoonia Loktanella: No match, genus\n",
      "Sva0081 sediment group: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Limibacillus: No match, genus\n",
      "JTB23: No match, genus\n",
      "JTB23: No match, family\n",
      "JTB23: No match, order\n",
      "Cellvibrionaceae: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Desulfosarcinaceae: No match, family\n",
      "Singulisphaera: No match, genus\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "Nitrosococcales: No match, order\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "BD7 11: No match, order\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "AB1: No match, genus\n",
      "AB1: No match, family\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "uncultured: No match, genus\n",
      "Isosphaeraceae: No match, family\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Candidatus Bealeia: No match, genus\n",
      "Synechocystis PCC 6803: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Sumerlaea: No match, genus\n",
      "Sumerlaeaceae: No match, family\n",
      "Sumerlaeales: No match, order\n",
      "Sumerlaeia: No match, class\n",
      "Sumerlaeota: No match, phylum\n",
      "Isosphaerales: No match, order\n",
      "PS1 clade: No match, genus\n",
      "PS1 clade: No match, family\n",
      "028H05 P BN P5: No match, genus\n",
      "028H05 P BN P5: No match, family\n",
      "028H05 P BN P5: No match, order\n",
      "028H05 P BN P5: No match, class\n",
      "Defluviicoccales: No match, order\n",
      "uncultured: No match, genus\n",
      "SCGC AAA164 E04: No match, genus\n",
      "Parvibaculales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "SM2D12: No match, genus\n",
      "SM2D12: No match, family\n",
      "SAR92 clade: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Luminiphilus: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Veillonella: No match, genus\n",
      "uncultured: No match, genus\n",
      "S085: No match, genus\n",
      "S085: No match, family\n",
      "S085: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "uncultured: No match, genus\n",
      "Porticoccaceae: No match, family\n",
      "Pedosphaerales: No match, orderCellvibrionales: No match, order\n",
      "\n",
      "OM75 clade: No match, genus\n",
      "Pir4 lineage: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Synechococcus CC9902: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Nisaeaceae: No match, family\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "SAR116 clade: No match, family\n",
      "Thalassobaculales: No match, order\n",
      "SBR1031: No match, genus\n",
      "SBR1031: No match, family\n",
      "SBR1031: No match, order\n",
      "Pirellulales: No match, order\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "uncultured: No match, genus\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "Candidatus Nitrosopelagicus: No match, genus\n",
      "uncultured: No match, genus\n",
      "Puniceispirillales: No match, order\n",
      "Parahaliea: No match, genusuncultured: No match, genus\n",
      "\n",
      "Methyloligellaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "AT s3 44: No match, genusEdaphobaculum: No match, genusHalieaceae: No match, family\n",
      "\n",
      "\n",
      "AT s3 44: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "AT s3 44: No match, order\n",
      "Synechocystis CCALA 700: No match, genus\n",
      "uncultured: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Sediminibacterium: No match, genus\n",
      "uncultured: No match, genusRichelia HH01: No match, genus\n",
      "\n",
      "Puia: No match, genusuncultured: No match, family\n",
      "\n",
      "Ga0077536: No match, genus\n",
      "Ga0077536: No match, family\n",
      "Ga0077536: No match, order\n",
      "uncultured: No match, genus\n",
      "Clade III: No match, genus\n",
      "Clade III: No match, family\n",
      "SAR11 clade: No match, order\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Oligoflexales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "Halobacteriovorax: No match, genus\n",
      "Parvibaculaceae: No match, family\n",
      "Parvibaculales: No match, order\n",
      "Woeseia: No match, genus\n",
      "Oligoflexia: No match, class\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "R76 B128: No match, genus\n",
      "Nitrospira: Multiple matches, genus \n",
      "Synechococcus CC9902: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Woeseiaceae: No match, family\n",
      "Bdellovibrionota: No match, phylum\n",
      "Kiritimatiellaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Thalassotalea: No match, genus\n",
      "Steroidobacterales: No match, order\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Clade Ib: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "SUP05 cluster: No match, genus\n",
      "Thioglobaceae: No match, family\n",
      "Thiomicrospirales: No match, order\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "Candidatus Actinomarina: No match, genusuncultured: No match, genus\n",
      "\n",
      "Kiritimatiellales: No match, order\n",
      "Pseudohongiella: No match, genus\n",
      "Methanofastidiosales: No match, order\n",
      "Actinomarinaceae: No match, family\n",
      "Clade I: No match, family\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Pseudohongiellaceae: No match, family\n",
      "Candidatus Megaira: No match, genus\n",
      "Sinobacterium: No match, genus\n",
      "Actinomarinales: No match, order\n",
      "SAR11 clade: No match, order\n",
      "OM60(NOR5) clade: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Sumerlaea: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Urania 1B 19 marine sediment group: No match, genus\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "LD29: No match, genus\n",
      "Sumerlaeaceae: No match, family\n",
      "Candidatus Megaira: No match, genus\n",
      "Cellvibrionales: No match, orderSubgroup 13: No match, genus\n",
      "\n",
      "Subgroup 13: No match, familyClade III: No match, genusKD3 93: No match, genus\n",
      "\n",
      "\n",
      "Subgroup 13: No match, orderClade III: No match, familyKD3 93: No match, family\n",
      "\n",
      "\n",
      "SAR11 clade: No match, orderAcidobacteriae: No match, class\n",
      "\n",
      "Acidobacteriota: No match, phylum\n",
      "Chthoniobacteraceae: No match, family\n",
      "Sumerlaeales: No match, order\n",
      "Atopobium: No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, familyGallicola: No match, genus\n",
      "\n",
      "Chthoniobacterales: No match, order\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "Coxiella: Multiple matches, genus \n",
      "Sumerlaeia: No match, class\n",
      "Atopobiaceae: No match, family\n",
      "OM60(NOR5) clade: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Peptostreptococcales Tissierellales: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Portibacter: No match, genus\n",
      "OCS116 clade: No match, genus\n",
      "OCS116 clade: No match, family\n",
      "Sumerlaeota: No match, phylum\n",
      "NS4 marine group: No match, genus\n",
      "C1 B045: No match, genus\n",
      "Parvibaculales: No match, order\n",
      "OM60(NOR5) clade: No match, genus\n",
      "uncultured: No match, genus\n",
      "Omnitrophales: No match, genusPorticoccaceae: No match, family\n",
      "\n",
      "Omnitrophales: No match, familyCellvibrionales: No match, order\n",
      "\n",
      "Omnitrophales: No match, order\n",
      "Algitalea: No match, genus\n",
      "Oxalobacteraceae: No match, family\n",
      "Mitochondria: No match, genus\n",
      "Mitochondria: No match, family\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "OM27 clade: No match, genus\n",
      "uncultured: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Nitrincolaceae: No match, family\n",
      "P.palmC41: No match, genus\n",
      "P.palmC41: No match, family\n",
      "P.palmC41: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Arcobacteraceae: No match, family\n",
      "SM1A02: No match, genus\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "Desulfocapsaceae: No match, family\n",
      "Prochlorococcus MIT9313: No match, genus\n",
      "uncultured: No match, genus\n",
      "Cyanobiaceae: No match, familyPla3 lineage: No match, genus\n",
      "\n",
      "Pla3 lineage: No match, family\n",
      "Pla3 lineage: No match, order\n",
      "Pla3 lineage: No match, class\n",
      "Desulfobulbales: No match, orderPlanctomycetota: No match, phylum\n",
      "\n",
      "Crassaminicella: No match, genus\n",
      "Sva0081 sediment group: No match, genus\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "Arctic95B 14: No match, genus\n",
      "Arctic95B 14: No match, family\n",
      "Bacteroidetes VC2.1 Bac22: No match, genus\n",
      "Bacteroidetes VC2.1 Bac22: No match, family\n",
      "Bacteroidetes VC2.1 Bac22: No match, order\n",
      "Desulfobulbia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Desulfosarcinaceae: No match, family\n",
      "Thermotaleaceae: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Profundimonas: No match, genus\n",
      "Nitrincolaceae: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "[Synechococcus] spongiarum group: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "Victivallales: No match, order\n",
      "MBMPE27: No match, genus\n",
      "MBMPE27: No match, family\n",
      "MBMPE27: No match, order\n",
      "Omnitrophaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Candidatus Bacilloplasma: No match, genus\n",
      "Sandaracinus: No match, genus\n",
      "Lentisphaeria: No match, class\n",
      "Omnitrophales: No match, order\n",
      "Magnetospiraceae: No match, family\n",
      "Sva0996 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "Sandaracinaceae: No match, familyVerrucomicrobiota: No match, phylum\n",
      "\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Microtrichaceae: No match, family\n",
      "Halobacteriovorax: No match, genus\n",
      "Polyangiales: No match, order\n",
      "SS1 B 02 17: No match, genus\n",
      "SS1 B 02 17: No match, family\n",
      "SS1 B 02 17: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Candidatus Kerfeldbacteria: No match, genus\n",
      "Candidatus Kerfeldbacteria: No match, family\n",
      "Candidatus Kerfeldbacteria: No match, order\n",
      "Microtrichales: No match, order\n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, family\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Polyangia: No match, class\n",
      "Poribacteria: No match, genus\n",
      "Poribacteria: No match, family\n",
      "Poribacteria: No match, order\n",
      "Poribacteria: No match, class\n",
      "Poribacteria: No match, phylum\n",
      "Acidimicrobiia: No match, class\n",
      "ABY1: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Methanofastidiosales: No match, order\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "Myxococcota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Defluviicoccales: No match, order\n",
      "SAR11 clade: No match, order\n",
      "vadinHA49: No match, genus\n",
      "vadinHA49: No match, family\n",
      "vadinHA49: No match, order\n",
      "vadinHA49: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Litoribacillus: No match, genus\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Amoebophilaceae: No match, family\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Sandaracinaceae: No match, family\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylumBalneolaceae: No match, family\n",
      "\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Defluviicoccales: No match, order\n",
      "Coxiella: Multiple matches, genus \n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Polyangiales: No match, order\n",
      "OCS116 clade: No match, genusBalneolales: No match, order\n",
      "\n",
      "OCS116 clade: No match, family\n",
      "Parvibaculales: No match, order\n",
      "Morganellaceae: No match, genus\n",
      "Morganellaceae: No match, family\n",
      "Lachnotalea: No match, genus\n",
      "Micavibrionales: No match, order\n",
      "Polyangia: No match, class\n",
      "Rhodothermia: No match, classNS4 marine group: No match, genus\n",
      "\n",
      "NS5 marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "HgCo23: No match, genus\n",
      "HgCo23: No match, family\n",
      "HgCo23: No match, order\n",
      "Myxococcota: No match, phylum\n",
      "Clade Ia: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Mitochondria: No match, genus\n",
      "Mitochondria: No match, family\n",
      "Bacteroidota: No match, phylum\n",
      "Profundimonas: No match, genus\n",
      "Pla3 lineage: No match, genus\n",
      "Pla3 lineage: No match, family\n",
      "Pla3 lineage: No match, order\n",
      "Pla3 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Lineage IV: No match, genusLS NOB: No match, genus\n",
      "\n",
      "Lineage IV: No match, family\n",
      "Lineage IV: No match, order\n",
      "Elusimicrobia: Multiple matches, class \n",
      "Elusimicrobiota: No match, phylum\n",
      "Nitrincolaceae: No match, family\n",
      "bacteriap25: No match, genus\n",
      "bacteriap25: No match, family\n",
      "bacteriap25: No match, order\n",
      "bacteriap25: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Blfdi19: No match, genus\n",
      "Blfdi19: No match, family\n",
      "Blfdi19: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "mle1 7: No match, genus\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "uncultured: No match, genus\n",
      "Nitrincolaceae: No match, family\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Acidobacteriota: No match, phylum\n",
      "Defluviicoccales: No match, order\n",
      "Sumerlaea: No match, genus\n",
      "Pir4 lineage: No match, genus\n",
      "Candidatus Hemipteriphilus: No match, genus\n",
      "uncultured: No match, genus\n",
      "UBA10353 marine group: No match, genus\n",
      "UBA10353 marine group: No match, family\n",
      "UBA10353 marine group: No match, order\n",
      "Sumerlaeaceae: No match, family\n",
      "Pirellulaceae: No match, family\n",
      "Candidatus Hepatoplasma: No match, genus\n",
      "EPR3968 O8a Bc78: No match, genus\n",
      "EPR3968 O8a Bc78: No match, family\n",
      "EPR3968 O8a Bc78: No match, order\n",
      "Gracilibacteria: No match, genus\n",
      "Gracilibacteria: No match, family\n",
      "Gracilibacteria: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Sumerlaeales: No match, order\n",
      "Pirellulales: No match, order\n",
      "Entomoplasmatales Incertae Sedis: No match, family\n",
      "Clade Ib: No match, genus\n",
      "Urania 1B 19 marine sediment group: No match, genus\n",
      "NS3a marine group: No match, genus\n",
      "AB1: No match, genus\n",
      "AB1: No match, family\n",
      "Sumerlaeia: No match, class\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Subgroup 10: No match, genus\n",
      "C86: No match, genus\n",
      "C86: No match, family\n",
      "C86: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Sumerlaeota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Thermoanaerobaculaceae: No match, family\n",
      "P9X2b3D02: No match, genus\n",
      "P9X2b3D02: No match, family\n",
      "P9X2b3D02: No match, order\n",
      "P9X2b3D02: No match, class\n",
      "BIrii41: No match, genus\n",
      "BIrii41: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Sumerlaeia: No match, class\n",
      "Sumerlaeota: No match, phylum\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, familyPAUC34f: No match, genus\n",
      "\n",
      "Woesearchaeales: No match, orderPAUC34f: No match, family\n",
      "\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Thermoanaerobaculales: No match, order\n",
      "Nitrospinota: No match, phylum\n",
      "Unknown Family: No match, genus\n",
      "Unknown Family: No match, family\n",
      "uncultured: No match, genus\n",
      "eub62A3: No match, genus\n",
      "eub62A3: No match, family\n",
      "eub62A3: No match, order\n",
      "BIrii41: No match, genus\n",
      "BIrii41: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Thermoanaerobaculia: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, familyCellvibrionaceae: No match, family\n",
      "\n",
      "Arctic97B 4 marine group: No match, orderCellvibrionales: No match, order\n",
      "\n",
      "PLTA13: No match, genus\n",
      "PLTA13: No match, family\n",
      "PLTA13: No match, order\n",
      "Ga0077536: No match, genus\n",
      "Ga0077536: No match, family\n",
      "Ga0077536: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Caenarcaniphilales: No match, genus\n",
      "Caenarcaniphilales: No match, family\n",
      "Caenarcaniphilales: No match, order\n",
      "PS1 clade: No match, genusJL ETNP F27: No match, genus\n",
      "\n",
      "PS1 clade: No match, family\n",
      "Parvibaculales: No match, order\n",
      "Vicinamibacteraceae: No match, family\n",
      "Vampirivibrionia: No match, class\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Candidatus Amoebophilus: No match, genus\n",
      "Vicinamibacterales: No match, order\n",
      "AqS1: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Nitrosococcales: No match, order\n",
      "Dadabacteriales: No match, genusPusillimonas: No match, genus\n",
      "\n",
      "Dadabacteriales: No match, family\n",
      "Dadabacteriales: No match, order\n",
      "Amoebophilaceae: No match, family\n",
      "Marinicaulis: No match, genus\n",
      "Dadabacteriia: No match, class\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "DEV007: No match, genus\n",
      "DEV007: No match, family\n",
      "Dadabacteria: No match, phylum\n",
      "Arcobacteraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Subgroup 22: No match, genusuncultured: No match, genus\n",
      "\n",
      "Magnetospiraceae: No match, familySubgroup 22: No match, family\n",
      "\n",
      "uncultured: No match, genusSubgroup 22: No match, order\n",
      "\n",
      "uncultured: No match, familySubgroup 22: No match, class\n",
      "\n",
      "Acidobacteriota: No match, phylumVicinamibacterales: No match, order\n",
      "\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "uncultured: No match, genus\n",
      "NS2b marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "Candidatus Nitrosopelagicus: No match, genus\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, genus\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, family\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, order\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, class\n",
      "Cellvibrionaceae: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Rubripirellula: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "AT s2 59: No match, genusPirellulales: No match, order\n",
      "\n",
      "AT s2 59: No match, family\n",
      "AT s2 59: No match, order\n",
      "uncultured: No match, genus\n",
      "Aenigmarchaeota: No match, phylum\n",
      "Cellvibrionales: No match, order\n",
      "NS7 marine group: No match, genus\n",
      "NS7 marine group: No match, family\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Terasakiellaceae: No match, family\n",
      "AB1: No match, genus\n",
      "AB1: No match, family\n",
      "MWH UniP1 aquatic group: No match, genusPAUC34f: No match, genus\n",
      "\n",
      "MWH UniP1 aquatic group: No match, familyPAUC34f: No match, family\n",
      "\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "WPS 2: No match, genus\n",
      "WPS 2: No match, family\n",
      "WPS 2: No match, order\n",
      "WPS 2: No match, class\n",
      "WPS 2: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Wenyingzhuangia: No match, genus\n",
      "SM1A02: No match, genus\n",
      "Bdellovibrionota: No match, phylum\n",
      "Synechococcus PCC 7336: No match, genus\n",
      "Pseudobacteriovorax: No match, genus\n",
      "Oligoflexales: No match, family\n",
      "Oligoflexales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Eurycoccales Incertae Sedis: No match, family\n",
      "Chryseobacterium: No match, genus\n",
      "CCM11a: No match, genus\n",
      "CCM11a: No match, family\n",
      "CCM11a: No match, orderCandidatus Paraholospora: No match, genus\n",
      "\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "EC3: No match, genus\n",
      "EC3: No match, family\n",
      "EC3: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Eurycoccales: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Weeksellaceae: No match, family\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "uncultured: No match, genus\n",
      "BD2 3: No match, genus\n",
      "BD2 3: No match, family\n",
      "Eel 36e1D6: No match, genus\n",
      "Eel 36e1D6: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "PeM15: No match, genus\n",
      "PeM15: No match, family\n",
      "PeM15: No match, orderCandidatus Colwellbacteria: No match, genus\n",
      "\n",
      "Candidatus Colwellbacteria: No match, family\n",
      "Candidatus Colwellbacteria: No match, order\n",
      "Parcubacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Methyloligellaceae: No match, family\n",
      "Pseudofulvibacter: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Victivallales: No match, order\n",
      "Arenicella: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, genus\n",
      "MWH UniP1 aquatic group: No match, genus\n",
      "MWH UniP1 aquatic group: No match, family\n",
      "Actinobacteria: Multiple matches, class Hydrogenedensaceae: No match, genus\n",
      "\n",
      "Candidatus Hepatincola: No match, genusActinobacteriota: No match, phylumHydrogenedensaceae: No match, family\n",
      "\n",
      "\n",
      "Candidatus Hepatincola: No match, family\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, orderArenicellaceae: No match, family\n",
      "\n",
      "Cyanobacteriia: No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "NS4 marine group: No match, genus\n",
      "wb1 A12: No match, genusHydrogenedentiales: No match, order\n",
      "\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Arenicellales: No match, order\n",
      "Sva0996 marine group: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Margulisbacteria: No match, genus\n",
      "Margulisbacteria: No match, family\n",
      "Margulisbacteria: No match, order\n",
      "Margulisbacteria: No match, class\n",
      "Margulisbacteria: No match, phylum\n",
      "OM27 clade: No match, genus\n",
      "Hydrogenedentia: No match, class\n",
      "Synechocystis PCC 6803: No match, genusMethylomirabilaceae: No match, family\n",
      "\n",
      "DS001: No match, genusuncultured: No match, genus\n",
      "\n",
      "FS140 16B 02 marine group: No match, genus\n",
      "Sva0996 marine group: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Candidatus Actinomarina: No match, genus\n",
      "Actinomarinaceae: No match, family\n",
      "Actinomarinales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Hydrogenedentes: No match, phylum\n",
      "Methylomirabilales: No match, order\n",
      "Leptotrichiaceae: No match, family\n",
      "Clade IV: No match, genus\n",
      "Clade IV: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Clade Ib: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, genus\n",
      "Cm1 21: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Nitrosococcales: No match, order\n",
      "Methylomirabilia: No match, class\n",
      "UBA10353 marine group: No match, genus\n",
      "UBA10353 marine group: No match, family\n",
      "UBA10353 marine group: No match, order\n",
      "Lentimonas: No match, genus\n",
      "uncultured: No match, genus\n",
      "SCGC AAA164 E04: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "Pedosphaerales: No match, order\n",
      "Subgroup 10: No match, genus\n",
      "Thermoanaerobaculaceae: No match, family\n",
      "Thermoanaerobaculales: No match, order\n",
      "Thermoanaerobaculia: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Parvibaculales: No match, order\n",
      "Methylomirabilota: No match, phylum\n",
      "Parvibaculaceae: No match, familyNS9 marine group: No match, genus\n",
      "\n",
      "Parvibaculales: No match, orderNS9 marine group: No match, family\n",
      "\n",
      "Marine Group II: No match, genusuncultured: No match, genus\n",
      "\n",
      "Clade I: No match, familyMarine Group II: No match, family\n",
      "\n",
      "Marine Group II: No match, orderSAR11 clade: No match, order\n",
      "\n",
      "P3OB 42: No match, genusuncultured: No match, genus\n",
      "\n",
      "BD1 7 clade: No match, genus\n",
      "Subgroup 17: No match, genusSediminispirochaeta: No match, genus\n",
      "\n",
      "Subgroup 17: No match, familyPla3 lineage: No match, genus\n",
      "\n",
      "Subgroup 17: No match, orderPla3 lineage: No match, family\n",
      "\n",
      "Vicinamibacteria: No match, classPla3 lineage: No match, order\n",
      "\n",
      "Acidobacteriota: No match, phylumPla3 lineage: No match, class\n",
      "\n",
      "Planctomycetota: No match, phylumuncultured: No match, genus\n",
      "\n",
      "NS9 marine group: No match, genusuncultured: No match, family\n",
      "\n",
      "Defluviicoccales: No match, orderNS9 marine group: No match, family\n",
      "\n",
      "Clade II: No match, genusLatescibacteraceae: No match, genus\n",
      "\n",
      "Clade II: No match, family\n",
      "Latescibacteraceae: No match, familySAR11 clade: No match, order\n",
      "\n",
      "Trichodesmium IMS101: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "NS4 marine group: No match, genus\n",
      "MB11C04 marine group: No match, genus\n",
      "CL500 3: No match, genusHaloplasma: No match, genus\n",
      "\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Latescibacterales: No match, orderVicinamibacterales: No match, order\n",
      "\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Vermiphilaceae: No match, genus\n",
      "Vermiphilaceae: No match, family\n",
      "Pla4 lineage: No match, genusHaloplasmataceae: No match, family\n",
      "\n",
      "Latescibacteria: No match, classPla4 lineage: No match, family\n",
      "\n",
      "Pla4 lineage: No match, order\n",
      "Pla4 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Candidatus Wolfebacteria: No match, genus\n",
      "Candidatus Wolfebacteria: No match, family\n",
      "Candidatus Wolfebacteria: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Gemmataceae: No match, family\n",
      "Babeliales: No match, order\n",
      "Haloplasmatales: No match, orderParcubacteria: No match, class\n",
      "\n",
      "Patescibacteria: No match, phylum\n",
      "NS5 marine group: No match, genus\n",
      "NS4 marine group: No match, genus\n",
      "Coxiella: Multiple matches, genus \n",
      "Latescibacterota: No match, phylumuncultured: No match, genus\n",
      "\n",
      "Ardenticatenales: No match, order\n",
      "Gemmatales: No match, order\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "Babeliae: No match, class\n",
      "Bacilli: No match, class\n",
      "Tistlia: No match, genus\n",
      "SM23 30: No match, genus\n",
      "SM23 30: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Lentimicrobiaceae: No match, genus\n",
      "Lentimicrobiaceae: No match, family\n",
      "Dependentiae: No match, phylum\n",
      "MSBL9: No match, order\n",
      "Chloroplast: No match, genusuncultured: No match, genus\n",
      "SAR202 clade: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Microtrichaceae: No match, family\n",
      "\n",
      "SAR202 clade: No match, familyChloroplast: No match, order\n",
      "Microtrichales: No match, orderSAR202 clade: No match, order\n",
      "\n",
      "Acidimicrobiia: No match, class\n",
      "Cyanobacteriia: No match, class\n",
      "\n",
      "Dehalococcoidia: No match, classActinobacteriota: No match, phylum\n",
      "uncultured: No match, genusChloroflexi: Multiple matches, phylum \n",
      "\n",
      "\n",
      "Micavibrionaceae: No match, family\n",
      "Micavibrionales: No match, order\n",
      "Kapabacteriales: No match, genus\n",
      "Kapabacteriales: No match, family\n",
      "Kapabacteriales: No match, order\n",
      "PAUC43f marine benthic group: No match, genus\n",
      "PAUC43f marine benthic group: No match, family\n",
      "PAUC43f marine benthic group: No match, order\n",
      "PAUC43f marine benthic group: No match, class\n",
      "Sedimenticolaceae: No match, family\n",
      "vadinHA49: No match, genus\n",
      "vadinHA49: No match, family\n",
      "vadinHA49: No match, order\n",
      "vadinHA49: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Candidatus Kuenenbacteria: No match, genus\n",
      "Candidatus Kuenenbacteria: No match, family\n",
      "Candidatus Kuenenbacteria: No match, order\n",
      "OM60(NOR5) clade: No match, genus\n",
      "Kapabacteria: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "Ilumatobacteraceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Rubinisphaeraceae: No match, family\n",
      "ABY1: No match, class\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Bacteroidota: No match, phylumoc32: No match, genus\n",
      "\n",
      "Polyangia: No match, class\n",
      "Criblamydiaceae: No match, genus\n",
      "Criblamydiaceae: No match, family\n",
      "Patescibacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Candidatus Altiarchaeum: No match, genus\n",
      "Myxococcota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Romboutsia: No match, genus\n",
      "S25 593: No match, genus\n",
      "S25 593: No match, family\n",
      "Myxococcota: No match, phylum\n",
      "Methylomonadaceae: No match, family\n",
      "Altiarchaeaceae: No match, family\n",
      "Flavicella: No match, genus\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "BD7 11: No match, order\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "NS7 marine group: No match, genusHalomicrobiaceae: No match, family\n",
      "\n",
      "NS7 marine group: No match, family\n",
      "Marine Benthic Group A: No match, genus\n",
      "Marine Benthic Group A: No match, family\n",
      "Marine Benthic Group A: No match, order\n",
      "Nitrososphaeria: No match, class\n",
      "Pla4 lineage: No match, genus\n",
      "Pla4 lineage: No match, family\n",
      "Pla4 lineage: No match, order\n",
      "Pla4 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Salinirepens: No match, genus\n",
      "uncultured: No match, genus\n",
      "Altiarchaeales: No match, order\n",
      "Halobacterales: No match, order\n",
      "Crocinitomicaceae: No match, familyMVP 88: No match, genus\n",
      "\n",
      "MVP 88: No match, family\n",
      "MVP 88: No match, order\n",
      "Aestuariicella: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Altiarchaeia: No match, class\n",
      "OCS116 clade: No match, genus\n",
      "OCS116 clade: No match, family\n",
      "Parvibaculales: No match, order\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "Omnitrophales: No match, genus\n",
      "Omnitrophales: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylumSB 5: No match, genus\n",
      "\n",
      "SB 5: No match, familyMitochondria: No match, genusAEGEAN 169 marine group: No match, genus\n",
      "\n",
      "\n",
      "Mitochondria: No match, familyAEGEAN 169 marine group: No match, family\n",
      "\n",
      "Marinimicrobia (SAR406 clade): No match, genusNS7 marine group: No match, genus\n",
      "\n",
      "Marinimicrobia (SAR406 clade): No match, familyNS7 marine group: No match, family\n",
      "\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "uncultured: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Elusimicrobia: Multiple matches, class \n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Sumerlaeia: No match, class\n",
      "Sumerlaeota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Desulfosarcinaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Gaiellales: No match, order\n",
      "Thermoleophilia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Altiarchaeota: No match, phylum\n",
      "Aquibacter: No match, genus\n",
      "Elusimicrobiota: No match, phylum\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "Nitrospira: Multiple matches, genus \n",
      "Subgroup 10: No match, genus\n",
      "Draconibacterium: No match, genus\n",
      "053A03 B DI P58: No match, genus\n",
      "053A03 B DI P58: No match, family\n",
      "053A03 B DI P58: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Thermoanaerobaculaceae: No match, family\n",
      "67 14: No match, genusProlixibacteraceae: No match, family\n",
      "\n",
      "67 14: No match, family\n",
      "Bradymonadales: No match, genus\n",
      "Arctic97B 4 marine group: No match, genusBradymonadales: No match, family\n",
      "Arctic97B 4 marine group: No match, family\n",
      "\n",
      "Bradymonadales: No match, orderArctic97B 4 marine group: No match, order\n",
      "\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Lentilitoribacter: No match, genus\n",
      "Thermoanaerobaculales: No match, order\n",
      "Fuerstia: No match, genus\n",
      "oc32: No match, genus\n",
      "Caproiciproducens: No match, genus\n",
      "HOC36: No match, genus\n",
      "HOC36: No match, family\n",
      "HOC36: No match, order\n",
      "Thermoanaerobaculia: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Rubinisphaeraceae: No match, family\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Marine Group III: No match, genus\n",
      "Parcubacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Arctic95B 14: No match, genus\n",
      "Arctic95B 14: No match, family\n",
      "Victivallales: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Bacteroidetes VC2.1 Bac22: No match, genus\n",
      "Bacteroidetes VC2.1 Bac22: No match, family\n",
      "Bacteroidetes VC2.1 Bac22: No match, orderuncultured: No match, genus\n",
      "\n",
      "uncultured: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Weeksellaceae: No match, familyChloroplast: No match, family\n",
      "\n",
      "uncultured: No match, genusChloroplast: No match, order\n",
      "\n",
      "Magnetospiraceae: No match, familyCyanobacteriia: No match, class\n",
      "\n",
      "Kiritimatiellaceae: No match, family\n",
      "Candidatus Endoecteinascidia: No match, genus\n",
      "Actinomarinales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Latescibacterota: No match, genus\n",
      "Latescibacterota: No match, family\n",
      "Latescibacterota: No match, order\n",
      "Latescibacterota: No match, class\n",
      "Latescibacterota: No match, phylum\n",
      "NS10 marine group: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "Kiritimatiellales: No match, order\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "Thalassotalea: No match, genus\n",
      "Marine Methylotrophic Group 2: No match, genus\n",
      "Methylomonadaceae: No match, family\n",
      "Magnetospiraceae: No match, familyCandidatus Gortzia: No match, genus\n",
      "\n",
      "NS10 marine group: No match, genus\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "OM182 clade: No match, genus\n",
      "OM182 clade: No match, family\n",
      "OM182 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "Crocinitomicaceae: No match, family\n",
      "uncultured: No match, genusuncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "\n",
      "uncultured: No match, order\n",
      "PB19: No match, genus\n",
      "PB19: No match, family\n",
      "PB19: No match, order\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "OM190: No match, genus\n",
      "OM190: No match, familyRubripirellula: No match, genus\n",
      "\n",
      "OM190: No match, order\n",
      "Pirellulaceae: No match, familyOM190: No match, class\n",
      "Pirellulales: No match, order\n",
      "\n",
      "Planctomycetota: No match, phylum\n",
      "Subgroup 21: No match, genus\n",
      "Subgroup 21: No match, family\n",
      "Subgroup 21: No match, order\n",
      "Subgroup 21: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Burkholderia Caballeronia Paraburkholderia: No match, genus\n",
      "LS NOB: No match, genus\n",
      "PeM15: No match, genus\n",
      "PeM15: No match, family\n",
      "PeM15: No match, order\n",
      "Sedimentitalea: No match, genus\n",
      "EPR3968 O8a Bc78: No match, genus\n",
      "EPR3968 O8a Bc78: No match, family\n",
      "EPR3968 O8a Bc78: No match, order\n",
      "Candidatus Actinomarina: No match, genus\n",
      "Actinomarinaceae: No match, family\n",
      "Actinomarinales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Actinobacteria: Multiple matches, class Cyanobacteriia: No match, class\n",
      "\n",
      "Actinobacteriota: No match, phylumCellvibrionaceae: No match, family\n",
      "\n",
      "P3OB 42: No match, genusCellvibrionales: No match, order\n",
      "\n",
      "A714019: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "Entotheonellaceae: No match, genus\n",
      "Entotheonellaceae: No match, family\n",
      "I3A: No match, genus\n",
      "I3A: No match, family\n",
      "I3A: No match, order\n",
      "I3A: No match, genus\n",
      "I3A: No match, family\n",
      "I3A: No match, order\n",
      "Clade IV: No match, genus\n",
      "SAR202 clade: No match, genusClade IV: No match, family\n",
      "\n",
      "SAR202 clade: No match, familySAR11 clade: No match, order\n",
      "\n",
      "SAR202 clade: No match, order\n",
      "Entotheonellales: No match, order\n",
      "NS10 marine group: No match, genus\n",
      "JG30 KF CM66: No match, genus\n",
      "OCS116 clade: No match, genusJG30 KF CM66: No match, family\n",
      "\n",
      "OCS116 clade: No match, familyJG30 KF CM66: No match, order\n",
      "\n",
      "Parvibaculales: No match, orderJG30 KF CM66: No match, class\n",
      "\n",
      "NB1 j: No match, genusChloroflexi: Multiple matches, phylum \n",
      "\n",
      "NB1 j: No match, familyuncultured: No match, genus\n",
      "\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "WCHB1 41: No match, genus\n",
      "WCHB1 41: No match, family\n",
      "WCHB1 41: No match, order\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Coxiella: Multiple matches, genus \n",
      "SM2D12: No match, genus\n",
      "SM2D12: No match, family\n",
      "Dehalococcoidia: No match, classEntotheonellia: No match, class\n",
      "\n",
      "Lineage IIc: No match, genus\n",
      "Lineage IIc: No match, family\n",
      "Lineage IIc: No match, order\n",
      "Lineage IIc: No match, class\n",
      "Elusimicrobiota: No match, phylum\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, orderEntotheonellaeota: No match, phylum\n",
      "\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "Pla4 lineage: No match, genus\n",
      "Pla4 lineage: No match, family\n",
      "Pla4 lineage: No match, order\n",
      "Pla4 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "uncultured: No match, genus\n",
      "Chloroflexi: Multiple matches, phylum Pirellulaceae: No match, family\n",
      "\n",
      "uncultured: No match, genusPirellulales: No match, order\n",
      "\n",
      "Ilumatobacteraceae: No match, familyPAUC26f: No match, genus\n",
      "\n",
      "Microtrichales: No match, orderPAUC26f: No match, family\n",
      "\n",
      "Acidimicrobiia: No match, classPAUC26f: No match, order\n",
      "\n",
      "Actinobacteriota: No match, phylumAcidobacteriae: No match, class\n",
      "\n",
      "uncultured: No match, genusAcidobacteriota: No match, phylum\n",
      "\n",
      "uncultured: No match, family\n",
      "Subgroup 26: No match, genus\n",
      "Subgroup 26: No match, family\n",
      "Subgroup 26: No match, order\n",
      "Subgroup 26: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Subgroup 22: No match, genus\n",
      "Subgroup 22: No match, family\n",
      "Subgroup 22: No match, order\n",
      "Subgroup 22: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Magnetospira: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "uncultured: No match, genus\n",
      "PAUC43f marine benthic group: No match, genus\n",
      "PAUC43f marine benthic group: No match, family\n",
      "PAUC43f marine benthic group: No match, order\n",
      "PAUC43f marine benthic group: No match, class\n",
      "Actinomarinales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Gemmataceae: No match, family\n",
      "Gemmatales: No match, order\n",
      "uncultured: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Kistimonas: No match, genus\n",
      "Candidatus Ovatusbacter: No match, genusJL ETNP F27: No match, genusMethyloprofundus: No match, genus\n",
      "\n",
      "\n",
      "Unknown Family: No match, familyGemmatimonadota: No match, phylumPS1 clade: No match, genus\n",
      "\n",
      "\n",
      "PS1 clade: No match, family\n",
      "Parvibaculales: No match, orderCandidatus Jidaibacter: No match, genus\n",
      "\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "PB19: No match, genus\n",
      "PB19: No match, family\n",
      "PB19: No match, order\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Endozoicomonadaceae: No match, family\n",
      "Sumerlaea: No match, genus\n",
      "Sumerlaeaceae: No match, family\n",
      "Sumerlaeales: No match, order\n",
      "Sumerlaeia: No match, class\n",
      "Sumerlaeota: No match, phylum\n",
      "Methylomonadaceae: No match, family\n",
      "Planctomycetota: No match, phylumcvE6: No match, genus\n",
      "\n",
      "PeM15: No match, genuscvE6: No match, family\n",
      "\n",
      "PeM15: No match, family\n",
      "PeM15: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "PAUC34f: No match, genus\n",
      "PAUC34f: No match, family\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "Pir4 lineage: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "VHS B3 70: No match, genus\n",
      "VHS B3 70: No match, family\n",
      "Actinobacteria: Multiple matches, class \n",
      "VHS B3 70: No match, orderActinobacteriota: No match, phylum\n",
      "\n",
      "Polyangia: No match, class\n",
      "uncultured: No match, genusMyxococcota: No match, phylum\n",
      "\n",
      "BD2 11 terrestrial group: No match, genusSubgroup 11: No match, genus\n",
      "BD2 11 terrestrial group: No match, family\n",
      "Subgroup 11: No match, family\n",
      "\n",
      "BD2 11 terrestrial group: No match, order\n",
      "Subgroup 11: No match, order\n",
      "BD2 11 terrestrial group: No match, classSubgroup 11: No match, class\n",
      "\n",
      "Gemmatimonadota: No match, phylumAcidobacteriota: No match, phylum\n",
      "\n",
      "uncultured: No match, genusuncultured: No match, genus\n",
      "\n",
      "Gimesiaceae: No match, familyMicrotrichaceae: No match, family\n",
      "\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Aquabacterium: No match, genus\n",
      "TK10: No match, genus\n",
      "TK10: No match, familyUBA10353 marine group: No match, genusCaedibacteraceae: No match, family\n",
      "\n",
      "\n",
      "TK10: No match, orderUBA10353 marine group: No match, family\n",
      "\n",
      "TK10: No match, classUBA10353 marine group: No match, order\n",
      "\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Pla3 lineage: No match, genus\n",
      "Pla3 lineage: No match, family\n",
      "Pla3 lineage: No match, order\n",
      "Pla3 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "NS5 marine group: No match, genus\n",
      "NS7 marine group: No match, genus\n",
      "NS7 marine group: No match, family\n",
      "NS4 marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "Eukaryota: No match, domain\n",
      "OM27 clade: No match, genus\n",
      "uncultured: No match, genus\n",
      "Caedibacterales: No match, order\n",
      "Phaeocystidibacter: No match, genusHelcococcus: No match, genus\n",
      "\n",
      "MB11C04 marine group: No match, genus\n",
      "OM182 clade: No match, genus\n",
      "OM182 clade: No match, family\n",
      "OM182 clade: No match, order\n",
      "A714019: No match, genus\n",
      "Vicingus: No match, genus\n",
      "Peptostreptococcales Tissierellales: No match, familyBlfdi19: No match, genusuncultured: No match, genus\n",
      "\n",
      "\n",
      "Peptostreptococcales Tissierellales: No match, orderBlfdi19: No match, family\n",
      "\n",
      "Blfdi19: No match, order\n",
      "Polyangia: No match, class\n",
      "MD3 55: No match, genusMyxococcota: No match, phylum\n",
      "\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Magnetospira: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "[Ruminococcus] gnavus group: No match, genus\n",
      "uncultured: No match, genus\n",
      "Candidatus Endobugula: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "uncultured: No match, genus\n",
      "SG8 4: No match, genus\n",
      "SG8 4: No match, family\n",
      "Zixibacteria: No match, genus\n",
      "Zixibacteria: No match, family\n",
      "Zixibacteria: No match, order\n",
      "Zixibacteria: No match, class\n",
      "Zixibacteria: No match, phylum\n",
      "Fokiniaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "Flavicella: No match, genusSubgroup 10: No match, genus\n",
      "\n",
      "MSBL9: No match, order\n",
      "OM60(NOR5) clade: No match, genus\n",
      "Desulfocapsaceae: No match, family\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, orderThermoanaerobaculaceae: No match, family\n",
      "\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Candidatus Hepatoplasma: No match, genus\n",
      "Entomoplasmatales Incertae Sedis: No match, family\n",
      "Desulfobulbales: No match, orderUBA12411: No match, genus\n",
      "\n",
      "UBA12411: No match, family\n",
      "Babeliales: No match, order\n",
      "Babeliae: No match, class\n",
      "Dependentiae: No match, phylum\n",
      "Thermoanaerobaculales: No match, order\n",
      "SS1 B 02 17: No match, genus\n",
      "SS1 B 02 17: No match, family\n",
      "SS1 B 02 17: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Subgroup 5: No match, genus\n",
      "Subgroup 5: No match, family\n",
      "Subgroup 5: No match, order\n",
      "Subgroup 5: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Desulfobulbia: No match, classhgcI clade: No match, genus\n",
      "\n",
      "S085: No match, genusThermoanaerobaculia: No match, class\n",
      "\n",
      "S085: No match, familyAcidobacteriota: No match, phylum\n",
      "\n",
      "S085: No match, order\n",
      "Pla3 lineage: No match, genusDehalococcoidia: No match, class\n",
      "\n",
      "Pla3 lineage: No match, familyMitochondria: No match, genusChloroflexi: Multiple matches, phylum \n",
      "\n",
      "\n",
      "Pla3 lineage: No match, orderMitochondria: No match, familyChloroplast: No match, genus\n",
      "\n",
      "Pla3 lineage: No match, class\n",
      "Chloroplast: No match, family\n",
      "Planctomycetota: No match, phylum\n",
      "\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "NS4 marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Desulfobacterota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Methyloligellaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Clade II: No match, genusCandidatus Nitrosopumilus: No match, genus\n",
      "\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Candidatus Berkiella: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Subgroup 13: No match, genus\n",
      "Subgroup 13: No match, family\n",
      "Subgroup 13: No match, order\n",
      "Acidobacteriae: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Puniceispirillales: No match, order\n",
      "Sva0996 marine group: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Spirochaeta 2: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Candidatus Magasanikbacteria: No match, genus\n",
      "Candidatus Magasanikbacteria: No match, family\n",
      "Candidatus Magasanikbacteria: No match, order\n",
      "OCS116 clade: No match, genus\n",
      "OCS116 clade: No match, family\n",
      "Parvibaculales: No match, order\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Omnitrophaceae: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "Subgroup 17: No match, genus\n",
      "Subgroup 17: No match, family\n",
      "Subgroup 17: No match, order\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Vibrionimonas: No match, genus\n",
      "uncultured: No match, genus\n",
      "ABY1: No match, classSubgroup 9: No match, genus\n",
      "\n",
      "Subgroup 9: No match, family\n",
      "Subgroup 9: No match, order\n",
      "Candidatus Peregrinibacteria: No match, genus\n",
      "Candidatus Peregrinibacteria: No match, family\n",
      "Candidatus Peregrinibacteria: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylumPeM15: No match, genus\n",
      "\n",
      "PeM15: No match, family\n",
      "PeM15: No match, order\n",
      "Patescibacteria: No match, phylum\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "S25 593: No match, genus\n",
      "S25 593: No match, family\n",
      "JTB23: No match, genus\n",
      "JTB23: No match, family\n",
      "JTB23: No match, order\n",
      "Actinobacteria: Multiple matches, class \n",
      "Actinobacteriota: No match, phylum\n",
      "FS140 16B 02 marine group: No match, genus\n",
      "OM27 clade: No match, genus\n",
      "Arctic95B 14: No match, genusMSB 4B10: No match, genus\n",
      "\n",
      "Arctic95B 14: No match, familyMSB 4B10: No match, family\n",
      "\n",
      "MSB 4B10: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "SUP05 cluster: No match, genus\n",
      "Thioglobaceae: No match, family\n",
      "Thiomicrospirales: No match, order\n",
      "uncultured: No match, genus\n",
      "Zixibacteria: No match, genus\n",
      "Zixibacteria: No match, family\n",
      "Zixibacteria: No match, order\n",
      "Zixibacteria: No match, class\n",
      "Zixibacteria: No match, phylum\n",
      "Victivallales: No match, order\n",
      "Bradymonadales: No match, genus\n",
      "Bradymonadales: No match, family\n",
      "Bradymonadales: No match, order\n",
      "uncultured: No match, genus\n",
      "IheB3 7: No match, genus\n",
      "BD7 8: No match, genus\n",
      "BD7 8: No match, family\n",
      "BD7 8: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Latescibacterota: No match, genus\n",
      "Latescibacterota: No match, family\n",
      "Latescibacterota: No match, order\n",
      "Latescibacterota: No match, class\n",
      "Latescibacterota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Candidatus Nitrosopumilus: No match, genus\n",
      "Synechococcus CC9902: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "Eel 36e1D6: No match, genus\n",
      "Eel 36e1D6: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Desulfuromonadia: No match, class\n",
      "Melioribacteraceae: No match, family\n",
      "Candidatus Endoecteinascidia: No match, genus\n",
      "P3OB 42: No match, genus\n",
      "Desulfobacterota: No match, phylum\n",
      "Vicingus: No match, genus\n",
      "DEV007: No match, genus\n",
      "DEV007: No match, family\n",
      "Ignavibacteriales: No match, order\n",
      "Sediminispirochaeta: No match, genus\n",
      "Cm1 21: No match, genus\n",
      "Ignavibacteria: No match, class\n",
      "Bacteroidota: No match, phylum\n",
      "Bacteroidetes BD2 2: No match, genus\n",
      "Bacteroidetes BD2 2: No match, family\n",
      "OM27 clade: No match, genus\n",
      "Iainarchaeales: No match, genus\n",
      "Iainarchaeales: No match, family\n",
      "Iainarchaeales: No match, order\n",
      "Nitrosococcaceae: No match, family\n",
      "Sedimenticola: No match, genus\n",
      "Lentimicrobiaceae: No match, genus\n",
      "Lentimicrobiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Chloroplast: No match, genusCellvibrionaceae: No match, family\n",
      "\n",
      "Chloroplast: No match, familyCellvibrionales: No match, order\n",
      "\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Iainarchaeia: No match, class\n",
      "Nitrosococcales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "WCHB1 41: No match, genus\n",
      "WCHB1 41: No match, family\n",
      "WCHB1 41: No match, order\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Sedimenticolaceae: No match, family\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, genus\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, family\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, order\n",
      "Deep Sea Euryarchaeotic Group(DSEG): No match, class\n",
      "Clostridiisalibacter: No match, genus\n",
      "Iainarchaeota: No match, phylum\n",
      "Gimesiaceae: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Aenigmarchaeota: No match, phylum\n",
      "0319 6G20: No match, genus\n",
      "0319 6G20: No match, family\n",
      "0319 6G20: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Salinimicrobium: No match, genus\n",
      "Candidatus Actinomarina: No match, genus\n",
      "Actinomarinaceae: No match, family\n",
      "Actinomarinales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "KI89A clade: No match, genus\n",
      "KI89A clade: No match, family\n",
      "KI89A clade: No match, order\n",
      "Luminiphilus: No match, genusGa0077536: No match, genus\n",
      "Halieaceae: No match, family\n",
      "\n",
      "Ga0077536: No match, family\n",
      "Cellvibrionales: No match, orderGa0077536: No match, order\n",
      "\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, familyBD2 11 terrestrial group: No match, genus\n",
      "\n",
      "Chloroplast: No match, orderBD2 11 terrestrial group: No match, family\n",
      "\n",
      "Cyanobacteriia: No match, class\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Amoebophilaceae: No match, family\n",
      "Synechococcus PCC 7336: No match, genus\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "MD3 55: No match, genus\n",
      "Fokiniaceae: No match, family\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "Ascidiaceihabitans: No match, genus\n",
      "uncultured: No match, genus\n",
      "Sandaracinaceae: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "Alloprevotella: No match, genusSAR202 clade: No match, order\n",
      "\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Magnetospira: No match, genus\n",
      "endosymbionts: No match, genus\n",
      "Eurycoccales Incertae Sedis: No match, family\n",
      "Dadabacteriales: No match, genus\n",
      "Dadabacteriales: No match, family\n",
      "Dadabacteriales: No match, order\n",
      "KI89A clade: No match, genus\n",
      "Methylobacterium Methylorubrum: No match, genusKI89A clade: No match, family\n",
      "\n",
      "KI89A clade: No match, order\n",
      "Magnetospiraceae: No match, family\n",
      "Thiomicrospiraceae: No match, family\n",
      "Thiomicrospirales: No match, order\n",
      "Eurycoccales: No match, order\n",
      "Cyanobacteriia: No match, classDadabacteriia: No match, class\n",
      "\n",
      "Salinirepens: No match, genus\n",
      "Flexithrix: No match, genus\n",
      "uncultured: No match, genus\n",
      "SB 5: No match, genus\n",
      "SB 5: No match, family\n",
      "Dadabacteria: No match, phylum\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Crocinitomicaceae: No match, family\n",
      "0319 6G20: No match, genus\n",
      "0319 6G20: No match, family\n",
      "uncultured: No match, genus0319 6G20: No match, orderuncultured: No match, genus\n",
      "\n",
      "\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "BD7 11: No match, order\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Coxiella: Multiple matches, genus \n",
      "Gimesiaceae: No match, familyuncultured: No match, genus\n",
      "\n",
      "Unknown Family: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Sumerlaeia: No match, class\n",
      "Sumerlaeota: No match, phylum\n",
      "Porticoccaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "NS5 marine group: No match, genus\n",
      "Hydrothermarchaeales: No match, genusuncultured: No match, genus\n",
      "Hydrothermarchaeales: No match, family\n",
      "Clade I: No match, family\n",
      "\n",
      "Hydrothermarchaeales: No match, orderSAR11 clade: No match, order\n",
      "\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "Babeliales: No match, genus\n",
      "Babeliales: No match, family\n",
      "Luminiphilus: No match, genusBabeliales: No match, order\n",
      "\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Limnochordia: No match, class\n",
      "CCM11a: No match, genus\n",
      "Bdellovibrionota: No match, phylumCCM11a: No match, family\n",
      "\n",
      "CCM11a: No match, order\n",
      "Hydrothermarchaeia: No match, class\n",
      "Babeliae: No match, class\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, orderPlanctomicrobium: No match, genus\n",
      "\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Hydrothermarchaeota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Ga0077536: No match, genus\n",
      "Ga0077536: No match, family\n",
      "Ga0077536: No match, order\n",
      "uncultured: No match, genus\n",
      "Rubinisphaeraceae: No match, family\n",
      "Dependentiae: No match, phylum\n",
      "Reyranella: No match, genus\n",
      "Omnitrophaceae: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "FS140 16B 02 marine group: No match, genus\n",
      "Diplorickettsiaceae: No match, family\n",
      "PS1 clade: No match, genus\n",
      "PS1 clade: No match, family\n",
      "Parvibaculales: No match, order\n",
      "uncultured: No match, genus\n",
      "NS10 marine group: No match, genus\n",
      "Omnitrophales: No match, order\n",
      "Reyranellaceae: No match, family\n",
      "PAUC34f: No match, genus\n",
      "PAUC34f: No match, family\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "Diplorickettsiales: No match, order\n",
      "Mitochondria: No match, genus\n",
      "Mitochondria: No match, family\n",
      "LS NOB: No match, genusOmnitrophia: No match, class\n",
      "\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Reyranellales: No match, order\n",
      "Clade III: No match, genus\n",
      "Clade III: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Microtrichaceae: No match, family\n",
      "Lentimonas: No match, genus\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "67 14: No match, genusSAR11 clade: No match, order\n",
      "67 14: No match, family\n",
      "\n",
      "Halarcobacter: No match, genus\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Marine Group III: No match, genus\n",
      "uncultured: No match, genus\n",
      "Ilumatobacteraceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "NS5 marine group: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Nitrosococcales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Arcobacteraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Candidatus Lloydbacteria: No match, genus\n",
      "Candidatus Lloydbacteria: No match, family\n",
      "Candidatus Lloydbacteria: No match, order\n",
      "RCP2 54: No match, genus\n",
      "RCP2 54: No match, family\n",
      "RCP2 54: No match, order\n",
      "RCP2 54: No match, class\n",
      "RCP2 54: No match, phylum\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Subgroup 22: No match, genus\n",
      "Subgroup 22: No match, family\n",
      "Subgroup 22: No match, order\n",
      "Subgroup 22: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Cyanobium PCC 6307: No match, genus\n",
      "uncultured: No match, genus\n",
      "Psychrosphaera: No match, genus\n",
      "uncultured: No match, genus\n",
      "Parcubacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Blfdi19: No match, genus\n",
      "Blfdi19: No match, family\n",
      "Blfdi19: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "HOC36: No match, genus\n",
      "HOC36: No match, family\n",
      "HOC36: No match, order\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Margulisbacteria: No match, genus\n",
      "Margulisbacteria: No match, family\n",
      "Margulisbacteria: No match, order\n",
      "Margulisbacteria: No match, class\n",
      "Margulisbacteria: No match, phylum\n",
      "Candidatus Actinomarina: No match, genus\n",
      "Actinomarinaceae: No match, family\n",
      "Actinomarinales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Cyanobiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Profundimonas: No match, genus\n",
      "uncultured: No match, genusNitrincolaceae: No match, family\n",
      "\n",
      "uncultured: No match, family\n",
      "uncultured: No match, genus\n",
      "Sva0996 marine group: No match, genus\n",
      "Candidatus Nitrosopelagicus: No match, genusPRD18C08: No match, genus\n",
      "\n",
      "PRD18C08: No match, family\n",
      "Victivallales: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Thalassobaculales: No match, order\n",
      "uncultured: No match, genus\n",
      "Microtrichaceae: No match, familyThalassobaculales: No match, order\n",
      "\n",
      "MSBL8: No match, genus\n",
      "MSBL8: No match, family\n",
      "[Caedibacter] taeniospiralis group: No match, genus\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Coxiella: Multiple matches, genus \n",
      "KD4 96: No match, genus\n",
      "KD4 96: No match, family\n",
      "KD4 96: No match, order\n",
      "KD4 96: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "[Caedibacter] taeniospiralis group: No match, genus\n",
      "Magnetospira: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Cloacimonadales: No match, order\n",
      "NS2b marine group: No match, genus\n",
      "WPS 2: No match, genus\n",
      "WPS 2: No match, family\n",
      "WPS 2: No match, order\n",
      "WPS 2: No match, class\n",
      "WPS 2: No match, phylum\n",
      "Kapabacteriales: No match, genus\n",
      "Kapabacteriales: No match, family\n",
      "Kapabacteriales: No match, order\n",
      "Cloacimonadia: No match, class\n",
      "Perspicuibacter: No match, genus\n",
      "MD3 55: No match, genus\n",
      "Candidatus Berkiella: No match, genus\n",
      "HIMB11: No match, genus\n",
      "Kapabacteria: No match, class\n",
      "Arenicellaceae: No match, family\n",
      "Fokiniaceae: No match, family\n",
      "Unknown Family: No match, family\n",
      "Candidatus Yanofskybacteria: No match, genus\n",
      "Candidatus Yanofskybacteria: No match, family\n",
      "Candidatus Yanofskybacteria: No match, order\n",
      "Parcubacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Bacteroidota: No match, phylum\n",
      "Candidatus Nitrosopelagicus: No match, genus\n",
      "Sva0996 marine group: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Arenicellales: No match, order\n",
      "Cloacimonadota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Candidatus Amoebophilus: No match, genus\n",
      "uncultured: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Gimesiaceae: No match, family\n",
      "OM43 clade: No match, genus\n",
      "Candidatus Peribacteria: No match, genus\n",
      "Candidatus Peribacteria: No match, family\n",
      "Candidatus Peribacteria: No match, order\n",
      "Amoebophilaceae: No match, family\n",
      "bacteriap25: No match, genus\n",
      "bacteriap25: No match, family\n",
      "bacteriap25: No match, order\n",
      "bacteriap25: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Halobacteriovorax: No match, genus\n",
      "Anaerococcus: No match, genus\n",
      "Peptostreptococcales Tissierellales: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Clade II: No match, genus\n",
      "Clade II: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "BD1 7 clade: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "OM182 clade: No match, genus\n",
      "OM182 clade: No match, family\n",
      "OM182 clade: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Woeseia: No match, genus\n",
      "UBA10353 marine group: No match, genus\n",
      "UBA10353 marine group: No match, family\n",
      "UBA10353 marine group: No match, order\n",
      "Rikenellaceae RC9 gut group: No match, genus\n",
      "SCGC AAA164 E04: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "Pedosphaerales: No match, order\n",
      "P9X2b3D02: No match, genus\n",
      "P9X2b3D02: No match, family\n",
      "uncultured: No match, genusP9X2b3D02: No match, order\n",
      "\n",
      "uncultured: No match, familyP9X2b3D02: No match, class\n",
      "\n",
      "uncultured: No match, orderNitrospinota: No match, phylum\n",
      "\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "WS1: No match, genus\n",
      "WS1: No match, family\n",
      "WS1: No match, order\n",
      "WS1: No match, class\n",
      "WS1: No match, phylum\n",
      "Woeseiaceae: No match, family\n",
      "P.palmC41: No match, genus\n",
      "P.palmC41: No match, family\n",
      "P.palmC41: No match, order\n",
      "Rubripirellula: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Crocinitomicaceae: No match, family\n",
      "Trichodesmium IMS101: No match, genusEel 36e1D6: No match, genus\n",
      "\n",
      "Eel 36e1D6: No match, family\n",
      "SBR1031: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Steroidobacterales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Roseimarinus: No match, genus\n",
      "ADurb.Bin180: No match, genus\n",
      "ADurb.Bin180: No match, family\n",
      "ADurb.Bin180: No match, order\n",
      "uncultured: No match, genus\n",
      "Rubinisphaeraceae: No match, family\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "SAR202 clade: No match, genus\n",
      "SAR202 clade: No match, family\n",
      "SAR202 clade: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Coxiella: Multiple matches, genus \n",
      "Portibacter: No match, genus\n",
      "Subgroup 18: No match, genus\n",
      "Subgroup 18: No match, family\n",
      "Subgroup 18: No match, order\n",
      "Subgroup 18: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Prolixibacteraceae: No match, family\n",
      "Umboniibacter: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Cellvibrionales: No match, orderThermoplasmatota: No match, phylum\n",
      "\n",
      "SG8 4: No match, genus\n",
      "SG8 4: No match, family\n",
      "MSBL9: No match, order\n",
      "Pseudohongiella: No match, genus\n",
      "uncultured: No match, genus\n",
      "Pajaroellobacter: No match, genus\n",
      "Schizothrix LEGE 07164: No match, genus\n",
      "Sva0996 marine group: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Pseudohongiellaceae: No match, family\n",
      "Lineage IIb: No match, genus\n",
      "Lineage IIb: No match, family\n",
      "Lineage IIb: No match, order\n",
      "Lineage IIb: No match, class\n",
      "Babeliales: No match, genus\n",
      "Babeliales: No match, family\n",
      "Babeliales: No match, order\n",
      "Gracilibacteria: No match, genus\n",
      "Gracilibacteria: No match, family\n",
      "Gracilibacteria: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "A714019: No match, genus\n",
      "Spirulina DRTO 55.2: No match, genus\n",
      "Elusimicrobiota: No match, phylum\n",
      "Clade Ia: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "Babeliae: No match, class\n",
      "HgCo23: No match, genus\n",
      "HgCo23: No match, family\n",
      "HgCo23: No match, order\n",
      "BD1 7 clade: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "S25 593: No match, genus\n",
      "S25 593: No match, family\n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, family\n",
      "Prochlorococcus MIT9313: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "PAUC34f: No match, genus\n",
      "PAUC34f: No match, family\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "S25 593: No match, genus\n",
      "S25 593: No match, family\n",
      "Clade I: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Paraspirulinaceae: No match, family\n",
      "Christensenellaceae: No match, familyuncultured: No match, genus\n",
      "\n",
      "Dependentiae: No match, phylum\n",
      "IheB2 23: No match, genus\n",
      "Methylomonadaceae: No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "Synechococcus CC9902: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "FS140 16B 02 marine group: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "Cyanobacteriales: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Christensenellales: No match, order\n",
      "Desulfuromonadia: No match, order\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Malaciobacter: No match, genus\n",
      "Arcobacteraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "uncultured: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "JG30 KF CM66: No match, genus\n",
      "JG30 KF CM66: No match, family\n",
      "JG30 KF CM66: No match, order\n",
      "JG30 KF CM66: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "OM75 clade: No match, genus\n",
      "Vicinamibacteraceae: No match, genus\n",
      "Vicinamibacteraceae: No match, familyEV818SWSAP88: No match, genus\n",
      "\n",
      "EV818SWSAP88: No match, family\n",
      "EV818SWSAP88: No match, order\n",
      "Actinobacteria: Multiple matches, class \n",
      "Actinobacteriota: No match, phylum\n",
      "SCGC AAA286 E23: No match, genus\n",
      "SCGC AAA286 E23: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "Nisaeaceae: No match, family\n",
      "Vicinamibacterales: No match, order\n",
      "Group 1.1c: No match, genus\n",
      "Group 1.1c: No match, family\n",
      "Group 1.1c: No match, order\n",
      "Morganellaceae: No match, family\n",
      "Rubinisphaeraceae: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "Pir4 lineage: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "SAR116 clade: No match, genus\n",
      "Thalassobaculales: No match, orderSAR116 clade: No match, family\n",
      "\n",
      "Marine Group II: No match, genusPuniceispirillales: No match, order\n",
      "\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "KI89A clade: No match, genus\n",
      "KI89A clade: No match, family\n",
      "KI89A clade: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Arenicellaceae: No match, family\n",
      "Arenicellales: No match, order\n",
      "Woeseia: No match, genus\n",
      "Woeseiaceae: No match, family\n",
      "Steroidobacterales: No match, order\n",
      "MB11C04 marine group: No match, genus\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Nitrososphaeria: No match, class\n",
      "MND1: No match, genus\n",
      "MB11C04 marine group: No match, genus\n",
      "BD2 11 terrestrial group: No match, genus\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Subgroup 26: No match, genus\n",
      "Subgroup 26: No match, family\n",
      "Subgroup 26: No match, order\n",
      "Subgroup 26: No match, classObscuribacteraceae: No match, genus\n",
      "\n",
      "Acidobacteriota: No match, phylum\n",
      "Obscuribacteraceae: No match, family\n",
      "Margulisbacteria: No match, genus\n",
      "Margulisbacteria: No match, family\n",
      "Margulisbacteria: No match, order\n",
      "Margulisbacteria: No match, class\n",
      "Margulisbacteria: No match, phylum\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "A4b: No match, genusTK30: No match, genus\n",
      "\n",
      "TK30: No match, familyA4b: No match, family\n",
      "\n",
      "TK30: No match, order\n",
      "TK30: No match, class\n",
      "Obscuribacterales: No match, order\n",
      "Vampirivibrionia: No match, class\n",
      "Arenicellaceae: No match, genus\n",
      "Arenicellaceae: No match, family\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "SBR1031: No match, order\n",
      "uncultured: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Rubripirellula: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "0319 6G20: No match, genus\n",
      "0319 6G20: No match, family\n",
      "Candidatus Hepatoplasma: No match, genus0319 6G20: No match, order\n",
      "\n",
      "Oligoflexia: No match, classEntomoplasmatales Incertae Sedis: No match, family\n",
      "\n",
      "Bdellovibrionota: No match, phylum\n",
      "UBA10353 marine group: No match, genus\n",
      "UBA10353 marine group: No match, family\n",
      "UBA10353 marine group: No match, order\n",
      "Arenicellales: No match, order\n",
      "Pla4 lineage: No match, genus\n",
      "Pla4 lineage: No match, family\n",
      "Pla4 lineage: No match, order\n",
      "Pla4 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "PAUC34f: No match, genus\n",
      "PAUC34f: No match, family\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "JGI 0000069 P22: No match, genus\n",
      "JGI 0000069 P22: No match, family\n",
      "JGI 0000069 P22: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "KI89A clade: No match, genus\n",
      "KI89A clade: No match, family\n",
      "KI89A clade: No match, order\n",
      "PAUC26f: No match, genus\n",
      "SM23 31: No match, genusPAUC26f: No match, family\n",
      "Subgroup 26: No match, genus\n",
      "\n",
      "Subgroup 26: No match, familyPAUC26f: No match, order\n",
      "\n",
      "Subgroup 26: No match, order\n",
      "Subgroup 26: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "DSSD61: No match, genus\n",
      "Latescibacterota: No match, genus\n",
      "Latescibacterota: No match, family\n",
      "Latescibacterota: No match, order\n",
      "Latescibacterota: No match, class\n",
      "Latescibacterota: No match, phylum\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "NS2b marine group: No match, genus\n",
      "Calditrichaceae: No match, family\n",
      "Acidobacteriae: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "mle1 8: No match, genus\n",
      "mle1 8: No match, family\n",
      "mle1 8: No match, order\n",
      "uncultured: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Poseidonibacter: No match, genus\n",
      "Nitrospira: Multiple matches, genus \n",
      "Sandaracinus: No match, genus\n",
      "Calditrichales: No match, order\n",
      "Arcobacteraceae: No match, family\n",
      "Sandaracinaceae: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Subgroup 17: No match, genus\n",
      "Subgroup 17: No match, family\n",
      "Subgroup 17: No match, order\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Candidatus Peregrinibacteria: No match, genus\n",
      "Candidatus Peregrinibacteria: No match, familyCalditrichia: No match, class\n",
      "\n",
      "Candidatus Peregrinibacteria: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "JTB23: No match, genus\n",
      "JTB23: No match, family\n",
      "JTB23: No match, order\n",
      "Clade IV: No match, genus\n",
      "Clade IV: No match, family\n",
      "SAR11 clade: No match, order\n",
      "P3OB 42: No match, genus\n",
      "Bathyarchaeia: No match, genus\n",
      "Bathyarchaeia: No match, family\n",
      "Bathyarchaeia: No match, order\n",
      "Bathyarchaeia: No match, class\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Calditrichota: No match, phylum\n",
      "Micavibrionaceae: No match, family\n",
      "Roseobacter clade NAC11 7 lineage: No match, genus\n",
      "EPR3968 O8a Bc78: No match, genus\n",
      "EPR3968 O8a Bc78: No match, family\n",
      "EPR3968 O8a Bc78: No match, order\n",
      "Ascidiaceihabitans: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Bauldia: No match, genus\n",
      "Micavibrionales: No match, order\n",
      "uncultured: No match, genus\n",
      "OM182 clade: No match, genus\n",
      "OM182 clade: No match, family\n",
      "OM182 clade: No match, order\n",
      "Prochlorothrix PCC 9006: No match, genus\n",
      "MSB 4B10: No match, genus\n",
      "MSB 4B10: No match, family\n",
      "MSB 4B10: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Pseudohongiella: No match, genusPS1 clade: No match, genus\n",
      "\n",
      "PS1 clade: No match, familyPseudohongiellaceae: No match, family\n",
      "\n",
      "Parvibaculales: No match, order\n",
      "OCS116 clade: No match, genus\n",
      "OCS116 clade: No match, family\n",
      "Parvibaculales: No match, order\n",
      "uncultured: No match, genus\n",
      "Bacilli: No match, class\n",
      "uncultured: No match, genus\n",
      "Balneolaceae: No match, family\n",
      "Lenti 02: No match, genus\n",
      "Lenti 02: No match, family\n",
      "ML602M 17: No match, genusAminicenantales: No match, genus\n",
      "\n",
      "ML602M 17: No match, familyBalneolales: No match, order\n",
      "Aminicenantales: No match, family\n",
      "ML602M 17: No match, order\n",
      "\n",
      "Aminicenantales: No match, order\n",
      "uncultured: No match, genus\n",
      "Pleionea: No match, genus\n",
      "Oligosphaerales: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Bacteroidetes BD2 2: No match, genus\n",
      "Marine Methylotrophic Group 3: No match, genus\n",
      "Bacteroidetes BD2 2: No match, familyAminicenantia: No match, classRhodothermia: No match, class\n",
      "\n",
      "\n",
      "Acidobacteriota: No match, phylumBacteroidota: No match, phylum\n",
      "\n",
      "SAR11 clade: No match, genus\n",
      "SAR11 clade: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Desulfocapsaceae: No match, family\n",
      "Kangiellaceae: No match, family\n",
      "Methylophagaceae: No match, familyPla3 lineage: No match, genusNS9 marine group: No match, genus\n",
      "\n",
      "\n",
      "NS9 marine group: No match, familyNitrosococcales: No match, orderPla3 lineage: No match, family\n",
      "\n",
      "\n",
      "uncultured: No match, genusPla3 lineage: No match, order\n",
      "\n",
      "uncultured: No match, familyPla3 lineage: No match, class\n",
      "\n",
      "Puniceispirillales: No match, orderPlanctomycetota: No match, phylum\n",
      "\n",
      "SM23 31: No match, genus\n",
      "MWH UniP1 aquatic group: No match, genusDesulfobulbales: No match, order\n",
      "\n",
      "MWH UniP1 aquatic group: No match, family\n",
      "SAR324 clade(Marine group B): No match, genusBathyarchaeia: No match, genusCalditrichaceae: No match, family\n",
      "\n",
      "Candidatus Endoecteinascidia: No match, genus\n",
      "SAR324 clade(Marine group B): No match, familyBathyarchaeia: No match, family\n",
      "\n",
      "\n",
      "Bathyarchaeia: No match, orderSAR324 clade(Marine group B): No match, order\n",
      "\n",
      "Bathyarchaeia: No match, classSAR324 clade(Marine group B): No match, class\n",
      "\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "Desulfobulbia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "uncultured: No match, genusCyanobacteriia: No match, class\n",
      "\n",
      "Calditrichales: No match, order\n",
      "Fimbriimonadaceae: No match, genus\n",
      "Fimbriimonadaceae: No match, family\n",
      "Gimesiaceae: No match, family\n",
      "JG30 KF CM45: No match, genus\n",
      "JG30 KF CM45: No match, family\n",
      "Calditrichia: No match, class\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Margulisbacteria: No match, genusUCG 004: No match, genus\n",
      "\n",
      "Margulisbacteria: No match, family\n",
      "Margulisbacteria: No match, order\n",
      "Margulisbacteria: No match, class\n",
      "Margulisbacteria: No match, phylum\n",
      "Coxiella: Multiple matches, genus \n",
      "Fimbriimonadales: No match, order\n",
      "Fangia: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "PLTA13: No match, genusCalditrichota: No match, phylum\n",
      "\n",
      "PLTA13: No match, family\n",
      "PLTA13: No match, order\n",
      "Fimbriimonadia: No match, class\n",
      "Erysipelatoclostridiaceae: No match, family\n",
      "RS62 marine group: No match, genus\n",
      "Lineage IV: No match, genus\n",
      "Lineage IV: No match, family\n",
      "Lineage IV: No match, order\n",
      "HTCC5015: No match, genusGW2011 GWC1 47 15: No match, genus\n",
      "Armatimonadota: No match, phylum\n",
      "Arenicellaceae: No match, familyBD7 8: No match, genus\n",
      "uncultured: No match, genusGW2011 GWC1 47 15: No match, family\n",
      "\n",
      "\n",
      "\n",
      "OM60(NOR5) clade: No match, genusArenicellales: No match, orderBD7 8: No match, family\n",
      "uncultured: No match, family\n",
      "\n",
      "\n",
      "Halieaceae: No match, familyBD7 8: No match, orderUrania 1B 19 marine sediment group: No match, genus\n",
      "\n",
      "\n",
      "Cellvibrionales: No match, orderuncultured: No match, genus\n",
      "\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Omnitrophaceae: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Elusimicrobia: Multiple matches, class \n",
      "Woesearchaeales: No match, order\n",
      "BD2 11 terrestrial group: No match, genus\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, orderLimibaculum: No match, genus\n",
      "\n",
      "BD2 11 terrestrial group: No match, classuncultured: No match, genus\n",
      "\n",
      "Gemmatimonadota: No match, phylum\n",
      "Elusimicrobiota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "OM27 clade: No match, genus\n",
      "Elusimicrobiota: No match, phylum\n",
      "Vicinamibacterales: No match, order\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "Mitochondria: No match, genus\n",
      "Mitochondria: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "PB19: No match, genus\n",
      "PB19: No match, family\n",
      "PB19: No match, order\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Clade III: No match, genus\n",
      "Clade III: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Oligoflexales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Candidatus Nitrosopumilus: No match, genus\n",
      "Verruc 01: No match, genus\n",
      "FCPU426: No match, genus\n",
      "FCPU426: No match, family053A03 B DI P58: No match, genus\n",
      "\n",
      "053A03 B DI P58: No match, familyFCPU426: No match, order\n",
      "\n",
      "FCPU426: No match, class053A03 B DI P58: No match, order\n",
      "\n",
      "FCPU426: No match, phylumOligoflexia: No match, class\n",
      "\n",
      "Bdellovibrionota: No match, phylumCellvibrionales: No match, order\n",
      "\n",
      "uncultured: No match, genus\n",
      "AT s3 44: No match, genusMethanomethylophilaceae: No match, family\n",
      "\n",
      "AT s3 44: No match, family\n",
      "AT s3 44: No match, order\n",
      "uncultured: No match, genus\n",
      "Candidatus Aquiluna: No match, genus\n",
      "Caedibacteraceae: No match, family\n",
      "Pla3 lineage: No match, genus\n",
      "Pla3 lineage: No match, family\n",
      "Pla3 lineage: No match, order\n",
      "Pla3 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Candidatus Endoecteinascidia: No match, genus\n",
      "Methanomassiliicoccales: No match, order\n",
      "Caedibacterales: No match, order\n",
      "HglApr721: No match, genus\n",
      "HglApr721: No match, family\n",
      "HglApr721: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "uncultured: No match, class\n",
      "Annamia HOs24: No match, genus\n",
      "UBA10353 marine group: No match, genus\n",
      "UBA10353 marine group: No match, family\n",
      "UBA10353 marine group: No match, order\n",
      "Candidatus Endoecteinascidia: No match, genus\n",
      "Marine Group III: No match, genus\n",
      "Dermabacter: No match, genus\n",
      "Sva0485: No match, genus\n",
      "Sva0485: No match, family\n",
      "Sva0485: No match, order\n",
      "Sva0485: No match, class\n",
      "Sva0485: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Armatimonadota: No match, phylumParahaliea: No match, genus\n",
      "\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "IS 44: No match, genus\n",
      "Lentimicrobiaceae: No match, genus\n",
      "Lentimicrobiaceae: No match, family\n",
      "Ignavibacteriales: No match, order\n",
      "Ignavibacteria: No match, class\n",
      "Bacteroidota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Arenicella: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "DTB120: No match, genusC1 B045: No match, genus\n",
      "\n",
      "Porticoccaceae: No match, familyDTB120: No match, family\n",
      "\n",
      "Cellvibrionales: No match, orderDTB120: No match, order\n",
      "\n",
      "DTB120: No match, class\n",
      "DTB120: No match, phylum\n",
      "Arenicellaceae: No match, family\n",
      "Candidatus Nitrosopumilus: No match, genus\n",
      "uncultured: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Thermodesulfovibrionia: No match, class\n",
      "Pseudahrensia: No match, genus\n",
      "AB1: No match, genus\n",
      "AB1: No match, family\n",
      "AKAU4049: No match, genus\n",
      "AKAU4049: No match, family\n",
      "AKAU4049: No match, order\n",
      "AKAU4049: No match, class\n",
      "Arenicellales: No match, order\n",
      "bacteriap25: No match, genus\n",
      "bacteriap25: No match, family\n",
      "bacteriap25: No match, order\n",
      "bacteriap25: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "Varunaivibrio: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Omnitrophales: No match, genus\n",
      "Omnitrophales: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Magnetospira: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Nitrospirota: No match, phylum\n",
      "Silvanigrella: No match, genus\n",
      "uncultured: No match, genus\n",
      "Gemmatimonadota: No match, phylum\n",
      "Rubripirellula: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Aureitalea: No match, genus\n",
      "Silvanigrellaceae: No match, familyMitochondria: No match, genus\n",
      "\n",
      "Mitochondria: No match, family\n",
      "Marine Benthic Group A: No match, genus\n",
      "Marine Benthic Group A: No match, family\n",
      "Marine Benthic Group A: No match, order\n",
      "Nitrososphaeria: No match, class\n",
      "uncultured: No match, genus\n",
      "Sandaracinaceae: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Unknown Family: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Limnochordaceae: No match, genus\n",
      "Limnochordaceae: No match, family\n",
      "Candidatus Megaira: No match, genus\n",
      "R76 B128: No match, genus\n",
      "Silvanigrellales: No match, orderCandidatus Finniella: No match, genus\n",
      "\n",
      "Limnochordales: No match, order\n",
      "Hydrogenedensaceae: No match, genus\n",
      "Hydrogenedensaceae: No match, family\n",
      "Hydrogenedentiales: No match, order\n",
      "Hydrogenedentia: No match, class\n",
      "Hydrogenedentes: No match, phylum\n",
      "koll11: No match, genus\n",
      "koll11: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Kiritimatiellaceae: No match, family\n",
      "Paracaedibacteraceae: No match, family\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Limnochordia: No match, class\n",
      "Urania 1B 19 marine sediment group: No match, genus\n",
      "SM23 30: No match, genus\n",
      "SM23 30: No match, family\n",
      "Kiritimatiellales: No match, order\n",
      "Caproiciproducens: No match, genus\n",
      "Paracaedibacterales: No match, order\n",
      "uncultured: No match, genus\n",
      "FS140 16B 02 marine group: No match, genus\n",
      "S25 593: No match, genus\n",
      "S25 593: No match, family\n",
      "Halioglobus: No match, genus\n",
      "MSBL9: No match, order\n",
      "Kiritimatiellae: No match, classSva0485: No match, genus\n",
      "\n",
      "Sva0485: No match, familyVerrucomicrobiota: No match, phylum\n",
      "\n",
      "Sva0485: No match, orderuncultured: No match, genus\n",
      "\n",
      "Sva0485: No match, class\n",
      "Sva0485: No match, phylum\n",
      "uncultured: No match, genus\n",
      "KD3 93: No match, genus\n",
      "NS2b marine group: No match, genusKD3 93: No match, family\n",
      "\n",
      "uncultured: No match, genusuncultured: No match, genus\n",
      "\n",
      "Pir4 lineage: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Candidatus Jidaibacter: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Roseobacter clade CHAB I 5 lineage: No match, genus\n",
      "bacteriap25: No match, genusPAUC34f: No match, genus\n",
      "\n",
      "bacteriap25: No match, familyPAUC34f: No match, familyNB1 j: No match, genus\n",
      "\n",
      "\n",
      "bacteriap25: No match, orderPAUC34f: No match, orderNB1 j: No match, family\n",
      "\n",
      "\n",
      "bacteriap25: No match, classPAUC34f: No match, classNB1 j: No match, order\n",
      "\n",
      "\n",
      "Myxococcota: No match, phylumPAUC34f: No match, phylumNB1 j: No match, class\n",
      "\n",
      "\n",
      "NB1 j: No match, phylumuncultured: No match, genus\n",
      "\n",
      "Sedimenticolaceae: No match, family\n",
      "Aurantivirga: No match, genus\n",
      "uncultured: No match, genus\n",
      "Omnitrophaceae: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Parahaliea: No match, genus\n",
      "PAUC34f: No match, genusHalieaceae: No match, family\n",
      "\n",
      "PAUC34f: No match, familyCellvibrionales: No match, order\n",
      "PAUC34f: No match, order\n",
      "\n",
      "uncultured: No match, genusPAUC34f: No match, class\n",
      "\n",
      "uncultured: No match, familyPAUC34f: No match, phylum\n",
      "uncultured: No match, order\n",
      "JGI 0000069 P22: No match, genus\n",
      "\n",
      "uncultured: No match, genusJGI 0000069 P22: No match, family\n",
      "\n",
      "JGI 0000069 P22: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Vicingus: No match, genus\n",
      "P13 46: No match, genus\n",
      "P13 46: No match, family\n",
      "Kistimonas: No match, genus\n",
      "Endozoicomonadaceae: No match, family\n",
      "Pseudofulvibacter: No match, genus\n",
      "Verruc 01: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Dadabacteriales: No match, genus\n",
      "Dadabacteriales: No match, family\n",
      "Dadabacteriales: No match, order\n",
      "Magnetospira: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Rs M47: No match, genus\n",
      "Rs M47: No match, family\n",
      "Rs M47: No match, order\n",
      "Rs M47: No match, class\n",
      "Elusimicrobiota: No match, phylum\n",
      "SEEP SRB1: No match, genus\n",
      "Dadabacteriia: No match, classDEV007: No match, genus\n",
      "\n",
      "DEV007: No match, family\n",
      "LCP 89: No match, genus\n",
      "LCP 89: No match, family\n",
      "LCP 89: No match, order\n",
      "LCP 89: No match, class\n",
      "LCP 89: No match, phylum\n",
      "TG3: No match, genus\n",
      "TG3: No match, family\n",
      "Desulfosarcinaceae: No match, familyuncultured: No match, genusCandidatus Nitrosopelagicus: No match, genus\n",
      "\n",
      "\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Dadabacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Thalassobaculales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Richelia HH01: No match, genus\n",
      "Methylotenera: No match, genus\n",
      "uncultured: No match, genus\n",
      "Paludibacteraceae: No match, family\n",
      "Hungateiclostridiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "4572 13: No match, genus\n",
      "4572 13: No match, family\n",
      "Acidocella: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genusuncultured: No match, family\n",
      "\n",
      "Methanofastidiosales: No match, order\n",
      "Gaiellales: No match, order\n",
      "MSBL9: No match, order\n",
      "SS1 B 02 17: No match, genus\n",
      "SS1 B 02 17: No match, family\n",
      "SS1 B 02 17: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genusSAR324 clade(Marine group B): No match, genus\n",
      "\n",
      "Arcobacteraceae: No match, familySAR324 clade(Marine group B): No match, family\n",
      "\n",
      "Chloroplast: No match, genusSAR324 clade(Marine group B): No match, order\n",
      "\n",
      "Chloroplast: No match, familySAR324 clade(Marine group B): No match, class\n",
      "\n",
      "Chloroplast: No match, orderGracilibacteria: No match, classSAR324 clade(Marine group B): No match, phylum\n",
      "\n",
      "\n",
      "Cyanobacteriia: No match, classPatescibacteria: No match, phylumNB1 j: No match, genus\n",
      "\n",
      "\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "Desulfoprunum: No match, genusNB1 j: No match, class\n",
      "\n",
      "NB1 j: No match, phylum\n",
      "Thermoleophilia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Endozoicomonadaceae: No match, familyDesulfocapsaceae: No match, family\n",
      "\n",
      "Oligosphaeraceae: No match, genusNS9 marine group: No match, genusUBA12411: No match, genus\n",
      "\n",
      "\n",
      "Oligosphaeraceae: No match, familyNS9 marine group: No match, familyUBA12411: No match, family\n",
      "\n",
      "\n",
      "Babeliales: No match, orderSAR324 clade(Marine group B): No match, genus\n",
      "\n",
      "Babeliae: No match, classSAR324 clade(Marine group B): No match, family\n",
      "\n",
      "Dependentiae: No match, phylumSAR324 clade(Marine group B): No match, order\n",
      "\n",
      "NS10 marine group: No match, genusSAR324 clade(Marine group B): No match, class\n",
      "\n",
      "uncultured: No match, genusSAR324 clade(Marine group B): No match, phylum\n",
      "\n",
      "Desulfosarcinaceae: No match, familyBD2 11 terrestrial group: No match, genus\n",
      "\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "PAUC34f: No match, genus\n",
      "PAUC34f: No match, family\n",
      "PAUC34f: No match, order\n",
      "PAUC34f: No match, class\n",
      "PAUC34f: No match, phylum\n",
      "LD1 PA32: No match, genus\n",
      "LD1 PA32: No match, family\n",
      "LD1 PA32: No match, order\n",
      "Aquabacterium: No match, genus\n",
      "Oligosphaerales: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "KI89A clade: No match, genusDesulfobulbales: No match, order\n",
      "\n",
      "KI89A clade: No match, family\n",
      "KI89A clade: No match, order\n",
      "OM182 clade: No match, genus\n",
      "OM182 clade: No match, family\n",
      "OM182 clade: No match, order\n",
      "AEGEAN 169 marine group: No match, genusAcaryochloris MBIC11017: No match, genus\n",
      "\n",
      "AEGEAN 169 marine group: No match, family\n",
      "Lentimicrobiaceae: No match, genusDesulfobulbia: No match, class\n",
      "\n",
      "Lentimicrobiaceae: No match, familyDesulfobacterota: No match, phylum\n",
      "\n",
      "AncK6: No match, genus\n",
      "AncK6: No match, family\n",
      "AncK6: No match, order\n",
      "AncK6: No match, class\n",
      "AncK6: No match, phylumUmboniibacter: No match, genus\n",
      "\n",
      "NB1 j: No match, genusCellvibrionaceae: No match, family\n",
      "\n",
      "NB1 j: No match, familyCellvibrionales: No match, order\n",
      "\n",
      "NB1 j: No match, orderuncultured: No match, genus\n",
      "\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Nitrincolaceae: No match, family\n",
      "koll11: No match, genus\n",
      "koll11: No match, familyuncultured: No match, genus\n",
      "\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Bradymonadales: No match, genus\n",
      "Bradymonadales: No match, family\n",
      "Bradymonadales: No match, order\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Fimbriimonadales: No match, genus\n",
      "Fimbriimonadales: No match, family\n",
      "Fimbriimonadales: No match, order\n",
      "Draconibacterium: No match, genus\n",
      "Synechococcus CC9902: No match, genusThiohalophilus: No match, genus\n",
      "\n",
      "Woeseia: No match, genusCyanobiaceae: No match, family\n",
      "\n",
      "uncultured: No match, genus\n",
      "Fimbriimonadia: No match, class\n",
      "Armatimonadota: No match, phylum\n",
      "Prolixibacteraceae: No match, family\n",
      "Diplorickettsiaceae: No match, family\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Oligoflexales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "NS10 marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "UBA10353 marine group: No match, genus\n",
      "Woeseiaceae: No match, familyThioalkalispiraceae: No match, familyUBA10353 marine group: No match, family\n",
      "\n",
      "\n",
      "UBA10353 marine group: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "OM43 clade: No match, genus\n",
      "Diplorickettsiales: No match, order\n",
      "AEGEAN 169 marine group: No match, genus\n",
      "AEGEAN 169 marine group: No match, family\n",
      "0319 6G20: No match, genus\n",
      "0319 6G20: No match, familySteroidobacterales: No match, order\n",
      "Ectothiorhodospirales: No match, order\n",
      "0319 6G20: No match, order\n",
      "uncultured: No match, genus\n",
      "\n",
      "Oligoflexia: No match, class\n",
      "Micavibrionaceae: No match, familyBdellovibrionota: No match, phylum\n",
      "Micavibrionales: No match, order\n",
      "\n",
      "Pedosphaeraceae: No match, genusLS NOB: No match, genus\n",
      "\n",
      "Pedosphaeraceae: No match, family\n",
      "Pedosphaerales: No match, order\n",
      "Micavibrionales: No match, order\n",
      "Lineage IIc: No match, genus\n",
      "Lineage IIc: No match, family\n",
      "Lineage IIc: No match, order\n",
      "Lineage IIc: No match, classCaenarcaniphilales: No match, genus\n",
      "\n",
      "Caenarcaniphilales: No match, family\n",
      "Caenarcaniphilales: No match, order\n",
      "Marine Group II: No match, genus\n",
      "Marine Group II: No match, family\n",
      "Marine Group II: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Elusimicrobiota: No match, phylumVampirivibrionia: No match, class\n",
      "\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Chitinophagales: No match, order\n",
      "IS 44: No match, genusBIrii41: No match, genus\n",
      "\n",
      "BIrii41: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Pla3 lineage: No match, genus\n",
      "Pla3 lineage: No match, family\n",
      "Pla3 lineage: No match, order\n",
      "Pla3 lineage: No match, class\n",
      "MWH UniP1 aquatic group: No match, genus\n",
      "A4b: No match, genusMWH UniP1 aquatic group: No match, family\n",
      "\n",
      "A4b: No match, family\n",
      "Candidatus Megaira: No match, genusSBR1031: No match, order\n",
      "\n",
      "Spirulina DRTO 55.2: No match, genusWoeseia: No match, genus\n",
      "\n",
      "Woeseiaceae: No match, family\n",
      "Steroidobacterales: No match, order\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "EC3: No match, genusWoeseia: No match, genus\n",
      "\n",
      "Paraspirulinaceae: No match, familyLentimonas: No match, genusSubgroup 5: No match, genusEC3: No match, family\n",
      "\n",
      "\n",
      "\n",
      "Subgroup 5: No match, familyEC3: No match, order\n",
      "\n",
      "Subgroup 5: No match, order\n",
      "Latescibacterota: No match, genusSubgroup 5: No match, class\n",
      "\n",
      "Latescibacterota: No match, family\n",
      "Acidobacteriota: No match, phylumLatescibacterota: No match, order\n",
      "\n",
      "Latescibacterota: No match, class\n",
      "Latescibacterota: No match, phylum\n",
      "BD2 7: No match, genus\n",
      "BD2 7: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Arcobacteraceae: No match, familyCyanobacteriales: No match, order\n",
      "Woeseiaceae: No match, family\n",
      "\n",
      "Cyanobacteriia: No match, class\n",
      "Steroidobacterales: No match, order\n",
      "uncultured: No match, genus\n",
      "BD2 11 terrestrial group: No match, genus\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "Candidatus Nitrosopumilus: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "P3OB 42: No match, genus\n",
      "Gemmatimonadota: No match, phylum\n",
      "R76 B128: No match, genus\n",
      "Kiritimatiellaceae: No match, family\n",
      "Kiritimatiellales: No match, order\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylumMD2896 B216: No match, genus\n",
      "\n",
      "MD2896 B216: No match, familySAR202 clade: No match, genus\n",
      "\n",
      "MD2896 B216: No match, orderSAR202 clade: No match, family\n",
      "\n",
      "MD2896 B216: No match, classSAR202 clade: No match, order\n",
      "\n",
      "Patescibacteria: No match, phylumDehalococcoidia: No match, class\n",
      "\n",
      "Chloroflexi: Multiple matches, phylum Vicingus: No match, genus\n",
      "\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Ardenticatenales: No match, order\n",
      "Candidatus Aquiluna: No match, genus\n",
      "Desulfosarcinaceae: No match, familyuncultured: No match, genus\n",
      "\n",
      "Clade Ia: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "Aliikangiella: No match, genus\n",
      "Candidatus Magasanikbacteria: No match, genus\n",
      "Candidatus Magasanikbacteria: No match, family\n",
      "Candidatus Magasanikbacteria: No match, order\n",
      "Terasakiellaceae: No match, familyNS11 12 marine group: No match, genus\n",
      "\n",
      "DEV007: No match, genusNS11 12 marine group: No match, family\n",
      "\n",
      "DEV007: No match, family\n",
      "CL500 3: No match, genus\n",
      "uncultured: No match, genus\n",
      "Kangiellaceae: No match, family\n",
      "ABY1: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Kangiellaceae: No match, family\n",
      "Candidatus Berkiella: No match, genus\n",
      "Unknown Family: No match, family\n",
      "uncultured: No match, genus\n",
      "BD1 7 clade: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, orderChloroplast: No match, genus\n",
      "\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Bathyarchaeia: No match, genus\n",
      "Bathyarchaeia: No match, family\n",
      "Bathyarchaeia: No match, order\n",
      "Bathyarchaeia: No match, class\n",
      "Trichodesmium IMS101: No match, genus\n",
      "SPG12 343 353 B69: No match, genus\n",
      "SPG12 343 353 B69: No match, family\n",
      "SPG12 343 353 B69: No match, order\n",
      "SPG12 343 353 B69: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "NS4 marine group: No match, genus\n",
      "EPR3968 O8a Bc78: No match, genus\n",
      "EPR3968 O8a Bc78: No match, family\n",
      "EPR3968 O8a Bc78: No match, order\n",
      "Subgroup 22: No match, genus\n",
      "Subgroup 22: No match, family\n",
      "Subgroup 22: No match, order\n",
      "Subgroup 22: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Candidatus Nitrosopelagicus: No match, genus\n",
      "Desulfobacterota: No match, phylum\n",
      "AKAU3564 sediment group: No match, genus\n",
      "AKAU3564 sediment group: No match, family\n",
      "Methyloligellaceae: No match, genus\n",
      "Methyloligellaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "Halioglobus: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "JTB23: No match, genus\n",
      "JTB23: No match, family\n",
      "JTB23: No match, order\n",
      "Synechocystis PCC 6803: No match, genus\n",
      "A4b: No match, genusSUP05 cluster: No match, genus\n",
      "\n",
      "A4b: No match, family\n",
      "SBR1031: No match, order\n",
      "MBMPE27: No match, genus\n",
      "MBMPE27: No match, family\n",
      "MBMPE27: No match, order\n",
      "Eel 36e1D6: No match, genusMarine Benthic Group A: No match, genus\n",
      "\n",
      "Eel 36e1D6: No match, familyMarine Benthic Group A: No match, family\n",
      "\n",
      "Polyangiales: No match, orderMarine Benthic Group A: No match, order\n",
      "\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Actinobacteriota: No match, phylum\n",
      "Gracilibacteria: No match, genus\n",
      "Gracilibacteria: No match, family\n",
      "Gracilibacteria: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Kistimonas: No match, genus\n",
      "Thioglobaceae: No match, family\n",
      "Nitrososphaeria: No match, class\n",
      "BD2 11 terrestrial group: No match, genus\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "Arcticiflavibacter: No match, genus\n",
      "Thiomicrospirales: No match, order\n",
      "Endozoicomonadaceae: No match, family\n",
      "67 14: No match, genus\n",
      "67 14: No match, family\n",
      "AKAU4049: No match, genus\n",
      "AKAU4049: No match, family\n",
      "AKAU4049: No match, order\n",
      "AKAU4049: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "Pla3 lineage: No match, genus\n",
      "Pla3 lineage: No match, family\n",
      "Pla3 lineage: No match, order\n",
      "Pla3 lineage: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "SCGC AAA011 D5: No match, genus\n",
      "SCGC AAA011 D5: No match, family\n",
      "CCM11a: No match, genus\n",
      "CCM11a: No match, family\n",
      "CCM11a: No match, order\n",
      "hgcI clade: No match, genus\n",
      "Pir4 lineage: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Woesearchaeales: No match, order\n",
      "Sva0081 sediment group: No match, genus\n",
      "Trichodesmium IMS101: No match, genusvadinHA49: No match, genus\n",
      "\n",
      "vadinHA49: No match, family\n",
      "vadinHA49: No match, order\n",
      "vadinHA49: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Desulfosarcinaceae: No match, family\n",
      "Vicinamibacteraceae: No match, genus\n",
      "Vicinamibacteraceae: No match, family\n",
      "Vicinamibacterales: No match, order\n",
      "Vicinamibacteria: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "SCGC AAA164 E04: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "Pedosphaerales: No match, order\n",
      "uncultured: No match, genus\n",
      "Ardenticatenales: No match, order\n",
      "Leptospiraceae: No match, family\n",
      "Unknown Family: No match, genus\n",
      "Unknown Family: No match, family\n",
      "uncultured: No match, genus\n",
      "Criblamydiaceae: No match, genusFokiniaceae: No match, family\n",
      "\n",
      "Criblamydiaceae: No match, family\n",
      "Leptospirales: No match, order\n",
      "SCGC AAA164 E04: No match, genus\n",
      "uncultured: No match, genus\n",
      "SAR86 clade: No match, genus\n",
      "SAR86 clade: No match, family\n",
      "SAR86 clade: No match, order\n",
      "Ectothiorhodospirales: No match, order\n",
      "Leptospirae: No match, class\n",
      "Pedosphaeraceae: No match, family\n",
      "LD1 PA32: No match, genus\n",
      "LD1 PA32: No match, family\n",
      "LD1 PA32: No match, order\n",
      "JL ETNP F27: No match, genus\n",
      "uncultured: No match, genus\n",
      "NS10 marine group: No match, genusuncultured: No match, genus\n",
      "\n",
      "uncultured: No match, genus\n",
      "Unknown Family: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Spirochaetota: No match, phylum\n",
      "Pedosphaerales: No match, order\n",
      "Woesearchaeales: No match, order\n",
      "Marinicella: No match, genus\n",
      "C1 B045: No match, genus\n",
      "uncultured: No match, genus\n",
      "Pseudahrensia: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Candidatus Nitrosopumilus: No match, genus\n",
      "AT s2 59: No match, genus\n",
      "AT s2 59: No match, family\n",
      "AT s2 59: No match, order\n",
      "Porticoccaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "BIrii41: No match, genus\n",
      "BIrii41: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "uncultured: No match, genus\n",
      "Unknown Family: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Nitrososphaeria: No match, class\n",
      "NS11 12 marine group: No match, genus\n",
      "NS11 12 marine group: No match, family\n",
      "Thermodesulfovibrionia: No match, class\n",
      "B2M28: No match, genus\n",
      "B2M28: No match, family\n",
      "B2M28: No match, order\n",
      "P3OB 42: No match, genusCognatishimia: No match, genus\n",
      "\n",
      "Magnetovibrio: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Sandaracinaceae: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Nitrospirota: No match, phylum\n",
      "Silvanigrella: No match, genusDadabacteriales: No match, genus\n",
      "\n",
      "Gastranaerophilales: No match, genusDadabacteriales: No match, family\n",
      "\n",
      "Gastranaerophilales: No match, familyDadabacteriales: No match, order\n",
      "\n",
      "Gastranaerophilales: No match, orderDadabacteriia: No match, class\n",
      "\n",
      "Vampirivibrionia: No match, classDadabacteria: No match, phylum\n",
      "\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "MND1: No match, genus\n",
      "Candidatus Berkiella: No match, genus\n",
      "Silvanigrellaceae: No match, family\n",
      "Clostridium sensu stricto 1: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Silvanigrellales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Rubrivirga: No match, genus\n",
      "Coxiella: Multiple matches, genus \n",
      "IMCC26207: No match, genusNitrospira: Multiple matches, genus \n",
      "\n",
      "uncultured: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Rhodococcus: Multiple matches, genus \n",
      "NS5 marine group: No match, genusAliiglaciecola: No match, genus\n",
      "\n",
      "Hydrogenedensaceae: No match, genus\n",
      "Hydrogenedensaceae: No match, family\n",
      "Hydrogenedentiales: No match, order\n",
      "Hydrogenedentia: No match, class\n",
      "Hydrogenedentes: No match, phylum\n",
      "Moduliflexaceae: No match, genus\n",
      "Moduliflexaceae: No match, family\n",
      "Portibacter: No match, genus\n",
      "Temperatibacter: No match, genus\n",
      "uncultured: No match, genus\n",
      "Clade Ia: No match, genus\n",
      "Moduliflexales: No match, order\n",
      "Temperatibacteraceae: No match, family\n",
      "Anaerovoracaceae: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Moduliflexia: No match, class\n",
      "IS 44: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Modulibacteria: No match, phylum\n",
      "JG30 KF CM66: No match, genus\n",
      "JG30 KF CM66: No match, family\n",
      "JG30 KF CM66: No match, order\n",
      "JG30 KF CM66: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Cocleimonas: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Subgroup 22: No match, genus\n",
      "Subgroup 22: No match, family\n",
      "Subgroup 22: No match, order\n",
      "Subgroup 22: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Pedosphaeraceae: No match, genus\n",
      "Pedosphaeraceae: No match, family\n",
      "Pedosphaerales: No match, order\n",
      "Margulisbacteria: No match, genus\n",
      "Margulisbacteria: No match, family\n",
      "Margulisbacteria: No match, order\n",
      "Margulisbacteria: No match, class\n",
      "Margulisbacteria: No match, phylum\n",
      "Thiohalorhabdaceae: No match, family\n",
      "Balneolaceae: No match, family\n",
      "Lineage IV: No match, genus\n",
      "Lineage IV: No match, family\n",
      "Lineage IV: No match, order\n",
      "Trichodesmium IMS101: No match, genus\n",
      "Luteibaculum: No match, genus\n",
      "Thiohalorhabdales: No match, order\n",
      "Balneolales: No match, order\n",
      "Elusimicrobia: Multiple matches, class \n",
      "Elusimicrobiota: No match, phylum\n",
      "SAR116 clade: No match, genus\n",
      "SAR116 clade: No match, family\n",
      "Puniceispirillales: No match, order\n",
      "GIF3: No match, genus\n",
      "uncultured: No match, genus\n",
      "GIF3: No match, familyuncultured: No match, family\n",
      "GIF3: No match, order\n",
      "\n",
      "Actinomarinales: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Acidimicrobiia: No match, classChloroflexi: Multiple matches, phylum \n",
      "\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "CHAB XI 27: No match, genus\n",
      "CHAB XI 27: No match, family\n",
      "Rhodothermia: No match, classCHAB XI 27: No match, order\n",
      "\n",
      "B2706 C7: No match, genus\n",
      "Blastocatella: No match, genusWCHB1 41: No match, genus\n",
      "\n",
      "WCHB1 41: No match, family\n",
      "WCHB1 41: No match, order\n",
      "Balneolaceae: No match, familyBacteroidota: No match, phylum\n",
      "\n",
      "uncultured: No match, genus\n",
      "Kiritimatiellae: No match, classBlastocatellaceae: No match, family\n",
      "\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Terasakiellaceae: No match, family\n",
      "Balneolales: No match, order\n",
      "NS5 marine group: No match, genus\n",
      "Blastocatellales: No match, order\n",
      "LS NOB: No match, genus\n",
      "Yoonia Loktanella: No match, genus\n",
      "NS5 marine group: No match, genus\n",
      "Blastocatellia: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "type III: No match, genus\n",
      "uncultured: No match, familytype III: No match, family\n",
      "\n",
      "uncultured: No match, order\n",
      "Rhodothermia: No match, class\n",
      "Bacteroidota: No match, phylum\n",
      "[Desulfobacterium] catecholicum group: No match, genus\n",
      "Desulfocapsaceae: No match, family\n",
      "Desulfobulbales: No match, order\n",
      "Desulfobulbia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Acidobacteriae: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Cyanobium PCC 6307: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "SCGC AB 179 E04: No match, genus\n",
      "SCGC AB 179 E04: No match, family\n",
      "SCGC AB 179 E04: No match, order\n",
      "LD1 PA32: No match, genus\n",
      "LD1 PA32: No match, family\n",
      "LD1 PA32: No match, order\n",
      "Planctomicrobium: No match, genus\n",
      "Rubinisphaeraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Thermoflexus: No match, genus\n",
      "Nitrososphaeria: No match, class\n",
      "Thermoflexaceae: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "bacteriap25: No match, genus\n",
      "bacteriap25: No match, family\n",
      "bacteriap25: No match, order\n",
      "bacteriap25: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Lentimonas: No match, genus\n",
      "Pseudohongiella: No match, genus\n",
      "Pseudohongiellaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Chitinophagales: No match, orderThermoflexales: No match, order\n",
      "\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Izemoplasmatales: No match, genus\n",
      "Izemoplasmatales: No match, family\n",
      "Izemoplasmatales: No match, order\n",
      "uncultured: No match, genus\n",
      "Sandaracinaceae: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "NS5 marine group: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Micavibrionales: No match, order\n",
      "Latescibacteraceae: No match, genus\n",
      "Latescibacteraceae: No match, family\n",
      "Latescibacterales: No match, order\n",
      "Latescibacteria: No match, class\n",
      "Latescibacterota: No match, phylum\n",
      "Puniceispirillales: No match, order\n",
      "Bacilli: No match, class\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Oligoflexales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Sva0485: No match, genus\n",
      "Sva0485: No match, family\n",
      "Sva0485: No match, order\n",
      "Sva0485: No match, class\n",
      "Sva0485: No match, phylum\n",
      "P9X2b3D02: No match, genusNoviherbaspirillum: No match, genus\n",
      "\n",
      "Oxalobacteraceae: No match, familyP9X2b3D02: No match, family\n",
      "\n",
      "P9X2b3D02: No match, order\n",
      "P9X2b3D02: No match, class\n",
      "OM75 clade: No match, genus\n",
      "uncultured: No match, genus\n",
      "MSBL3: No match, genus\n",
      "Nitrospinota: No match, phylumKiritimatiellaceae: No match, family\n",
      "\n",
      "Kiritimatiellales: No match, order\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Nisaeaceae: No match, family\n",
      "Thiohalorhabdaceae: No match, family\n",
      "Vampirovibrionaceae: No match, genus\n",
      "Vampirovibrionaceae: No match, family\n",
      "Thalassotalea: No match, genus\n",
      "Gracilibacteria: No match, genus\n",
      "Gracilibacteria: No match, family\n",
      "Gracilibacteria: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Thalassobaculales: No match, order\n",
      "Thiohalorhabdales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Oligoflexales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "NS5 marine group: No match, genus\n",
      "Vampirovibrionales: No match, order\n",
      "Romboutsia: No match, genus\n",
      "N9D0: No match, genus\n",
      "N9D0: No match, family\n",
      "N9D0: No match, order\n",
      "N9D0: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Vampirivibrionia: No match, class\n",
      "Vicingus: No match, genus\n",
      "PAUC43f marine benthic group: No match, genus\n",
      "PAUC43f marine benthic group: No match, family\n",
      "PAUC43f marine benthic group: No match, order\n",
      "PAUC43f marine benthic group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "MBMPE27: No match, genus\n",
      "MBMPE27: No match, family\n",
      "MBMPE27: No match, order\n",
      "Silvanigrella: No match, genus\n",
      "Nitriliruptoraceae: No match, genus\n",
      "Nitriliruptoraceae: No match, family\n",
      "Limnochordaceae: No match, genus\n",
      "Limnochordaceae: No match, family\n",
      "Fusicatenibacter: No match, genus\n",
      "Silvanigrellaceae: No match, family\n",
      "Nitriliruptorales: No match, order\n",
      "Actinobacteria: Multiple matches, class \n",
      "Actinobacteriota: No match, phylum\n",
      "Limnochordales: No match, order\n",
      "Acidibacter: No match, genus\n",
      "Unknown Family: No match, family\n",
      "PAUC43f marine benthic group: No match, genus\n",
      "PAUC43f marine benthic group: No match, family\n",
      "PAUC43f marine benthic group: No match, order\n",
      "PAUC43f marine benthic group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "Silvanigrellales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Limnochordia: No match, class\n",
      "Blfdi19: No match, genus\n",
      "Blfdi19: No match, family\n",
      "Blfdi19: No match, order\n",
      "SAR324 clade(Marine group B): No match, genus\n",
      "SAR324 clade(Marine group B): No match, family\n",
      "SAR324 clade(Marine group B): No match, order\n",
      "SAR324 clade(Marine group B): No match, class\n",
      "SAR324 clade(Marine group B): No match, phylum\n",
      "uncultured: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Candidatus Hepatoplasma: No match, genus\n",
      "Entomoplasmatales Incertae Sedis: No match, family\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "Latescibacterota: No match, genus\n",
      "Latescibacterota: No match, family\n",
      "Latescibacterota: No match, order\n",
      "Latescibacterota: No match, class\n",
      "Latescibacterota: No match, phylum\n",
      "Polyangia: No match, class\n",
      "Unknown Family: No match, genus\n",
      "Unknown Family: No match, family\n",
      "AT s2 59: No match, genus\n",
      "AT s2 59: No match, family\n",
      "AT s2 59: No match, order\n",
      "SM23 30: No match, genus\n",
      "SM23 30: No match, family\n",
      "Aenigmarchaeales: No match, genus\n",
      "Aenigmarchaeales: No match, family\n",
      "Aenigmarchaeales: No match, order\n",
      "Myxococcota: No match, phylum\n",
      "MSBL9: No match, order\n",
      "Aenigmarchaeia: No match, class\n",
      "Aenigmarchaeota: No match, phylum\n",
      "Caenarcaniphilales: No match, genus\n",
      "Caenarcaniphilales: No match, family\n",
      "Caenarcaniphilales: No match, order\n",
      "Marine Methylotrophic Group 3: No match, genus\n",
      "Methylophagaceae: No match, family\n",
      "Nitrosococcales: No match, order\n",
      "uncultured: No match, genus\n",
      "SZB85: No match, genus\n",
      "028H05 P BN P5: No match, genus\n",
      "028H05 P BN P5: No match, family\n",
      "028H05 P BN P5: No match, order\n",
      "028H05 P BN P5: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Vampirivibrionia: No match, class\n",
      "NS11 12 marine group: No match, genus\n",
      "NS11 12 marine group: No match, family\n",
      "Nitrosococcaceae: No match, family\n",
      "P9X2b3D02: No match, genus\n",
      "P9X2b3D02: No match, family\n",
      "P9X2b3D02: No match, order\n",
      "P9X2b3D02: No match, class\n",
      "Desulfatiglans: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "uncultured: No match, genus\n",
      "Nitrospinota: No match, phylumNitrosococcales: No match, order\n",
      "\n",
      "uncultured: No match, genus\n",
      "Desulfatiglandaceae: No match, family\n",
      "[Eubacterium] eligens group: No match, genus\n",
      "bacteriap25: No match, genus\n",
      "bacteriap25: No match, family\n",
      "bacteriap25: No match, order\n",
      "bacteriap25: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "SCM38C39 marine group: No match, genus\n",
      "Desulfatiglandales: No match, order\n",
      "Dasania: No match, genus\n",
      "Desulfobacteria: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "BIrii41: No match, genus\n",
      "BIrii41: No match, family\n",
      "CL500 3: No match, genus\n",
      "Thermoplasmatota: No match, family\n",
      "Thermoplasmatota: No match, order\n",
      "Thermoplasmatota: No match, class\n",
      "Thermoplasmatota: No match, phylum\n",
      "Epulopiscium: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Chitinophagales: No match, orderPolyangiales: No match, order\n",
      "\n",
      "Polyangia: No match, classuncultured: No match, genus\n",
      "\n",
      "Myxococcota: No match, phylumuncultured: No match, family\n",
      "\n",
      "Halieaceae: No match, genusChitinophagales: No match, order\n",
      "\n",
      "Halieaceae: No match, familyuncultured: No match, genus\n",
      "\n",
      "Cellvibrionales: No match, order\n",
      "EV818SWSAP88: No match, genus\n",
      "EV818SWSAP88: No match, family\n",
      "EV818SWSAP88: No match, order\n",
      "uncultured: No match, genus\n",
      "Escherichia Shigella: No match, genus\n",
      "Micavibrionaceae: No match, family\n",
      "Aminicenantales: No match, genusMicavibrionales: No match, order\n",
      "\n",
      "Aminicenantales: No match, family\n",
      "Pseudohongiella: No match, genusAminicenantales: No match, order\n",
      "\n",
      "Pseudohongiellaceae: No match, family\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "BD7 11: No match, order\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "Aureisphaera: No match, genus\n",
      "Aminicenantia: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Micavibrionaceae: No match, family\n",
      "Micavibrionales: No match, order\n",
      "A4b: No match, genus\n",
      "A4b: No match, family\n",
      "Woeseia: No match, genus\n",
      "Roseobacter clade NAC11 7 lineage: No match, genus\n",
      "MD3 55: No match, genus\n",
      "Fokiniaceae: No match, family\n",
      "Aliiglaciecola: No match, genus\n",
      "SAR92 clade: No match, genus\n",
      "Porticoccaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Agaribacterium: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "uncultured: No match, genus\n",
      "Nitrincolaceae: No match, family\n",
      "uncultured: No match, genusWoeseiaceae: No match, family\n",
      "\n",
      "uncultured: No match, family\n",
      "SBR1031: No match, order\n",
      "Lentimonas: No match, genus\n",
      "Candidatus Ryanbacteria: No match, genus\n",
      "Candidatus Ryanbacteria: No match, family\n",
      "Candidatus Ryanbacteria: No match, order\n",
      "Methanofastidiosales: No match, orderSteroidobacterales: No match, order\n",
      "\n",
      "MD2904 B13: No match, genus\n",
      "MD2904 B13: No match, family\n",
      "MD2904 B13: No match, order\n",
      "Candidatus Omnitrophus: No match, genus\n",
      "Omnitrophaceae: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Parcubacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Aquibacter: No match, genusSBR1031: No match, genus\n",
      "\n",
      "SBR1031: No match, family\n",
      "SBR1031: No match, order\n",
      "Deferrisoma: No match, genus\n",
      "W5053: No match, genus\n",
      "Babeliaceae: No match, genus\n",
      "Waddlia: No match, genusBabeliaceae: No match, family\n",
      "\n",
      "Babeliales: No match, order\n",
      "Babeliae: No match, class\n",
      "Dependentiae: No match, phylum\n",
      "Defferrisomataceae: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Verruc 01: No match, genus\n",
      "uncultured: No match, genusWaddliaceae: No match, family\n",
      "\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Aestuariispira: No match, genus\n",
      "Defferrisomatales: No match, order\n",
      "Cyanobacteriales: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Limibaculum: No match, genus\n",
      "Leptospiraceae: No match, family\n",
      "Terasakiellaceae: No match, family\n",
      "Omnitrophales: No match, genus\n",
      "Omnitrophales: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "uncultured: No match, genus\n",
      "Subgroup 26: No match, genus\n",
      "Subgroup 26: No match, family\n",
      "Subgroup 26: No match, order\n",
      "Subgroup 26: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Defferrisomatia: No match, class\n",
      "Leptospirales: No match, order\n",
      "Annamia HOs24: No match, genus\n",
      "Zixibacteria: No match, genus\n",
      "Zixibacteria: No match, family\n",
      "Zixibacteria: No match, order\n",
      "Zixibacteria: No match, class\n",
      "Zixibacteria: No match, phylum\n",
      "Deferrisomatota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Leptospirae: No match, class\n",
      "Bauldia: No match, genus\n",
      "Portibacter: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Defluviicoccales: No match, order\n",
      "uncultured: No match, genus\n",
      "Balneolaceae: No match, family\n",
      "Balneolales: No match, order\n",
      "Rhodothermia: No match, class\n",
      "Bacteroidota: No match, phylum\n",
      "Blfdi19: No match, genus\n",
      "Blfdi19: No match, family\n",
      "Blfdi19: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Spirochaetota: No match, phylum\n",
      "Fokiniaceae: No match, family\n",
      "PAUC26f: No match, genus\n",
      "PAUC26f: No match, family\n",
      "PAUC26f: No match, order\n",
      "Acidobacteriae: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Arctic95B 14: No match, genus\n",
      "Arctic95B 14: No match, family\n",
      "Victivallales: No match, order\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Gemmataceae: No match, familyCarboxylicivirga: No match, genus\n",
      "\n",
      "Micrococcales: No match, order\n",
      "Marinilabiliaceae: No match, familyUrania 1B 19 marine sediment group: No match, genus\n",
      "\n",
      "Gemmatales: No match, order\n",
      "Actinobacteria: Multiple matches, class \n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Chitinophagales: No match, order\n",
      "HOC36: No match, genus\n",
      "HOC36: No match, family\n",
      "HOC36: No match, order\n",
      "AT s2 59: No match, genus\n",
      "AT s2 59: No match, family\n",
      "AT s2 59: No match, order\n",
      "PAUC43f marine benthic group: No match, genus\n",
      "PAUC43f marine benthic group: No match, family\n",
      "PAUC43f marine benthic group: No match, order\n",
      "PAUC43f marine benthic group: No match, class\n",
      "Gemmatimonadota: No match, phylum\n",
      "OM60(NOR5) clade: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Marine Methylotrophic Group 2: No match, genus\n",
      "Anaeromicrobium: No match, genus\n",
      "Peptostreptococcales Tissierellales: No match, familyPseudofulvibacter: No match, genus\n",
      "\n",
      "NS4 marine group: No match, genusPeptostreptococcales Tissierellales: No match, order\n",
      "\n",
      "Methylomonadaceae: No match, family\n",
      "Pseudobacteriovorax: No match, genus\n",
      "053A03 B DI P58: No match, genus\n",
      "ZOR0006: No match, genus053A03 B DI P58: No match, family\n",
      "\n",
      "053A03 B DI P58: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Nannocystis: No match, genus\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "Sh765B AG 111: No match, genus\n",
      "Oligoflexales: No match, familySh765B AG 111: No match, family\n",
      "\n",
      "Oligoflexales: No match, orderSh765B AG 111: No match, order\n",
      "\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Cocleimonas: No match, genus\n",
      "IS 44: No match, genus\n",
      "uncultured: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Acidibacter: No match, genus\n",
      "Unknown Family: No match, family\n",
      "LS NOB: No match, genus\n",
      "NS5 marine group: No match, genus\n",
      "Candidatus Endoecteinascidia: No match, genus\n",
      "Bdellovibrionota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Thermicanus: No match, genus\n",
      "BIrii41: No match, genus\n",
      "BIrii41: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Thermicanaceae: No match, family\n",
      "MB11C04 marine group: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, classTK10: No match, genus\n",
      "\n",
      "TK10: No match, family\n",
      "TK10: No match, order\n",
      "TK10: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Candidatus Captivus: No match, genus\n",
      "Paracaedibacteraceae: No match, family\n",
      "Paracaedibacterales: No match, order\n",
      "Thermicanales: No match, order\n",
      "PB19: No match, genus\n",
      "PB19: No match, family\n",
      "PB19: No match, order\n",
      "Candidatus Endobugula: No match, genus\n",
      "Cellvibrionaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "JG30 KF CM66: No match, genus\n",
      "JG30 KF CM66: No match, family\n",
      "JG30 KF CM66: No match, order\n",
      "JG30 KF CM66: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Bacilli: No match, class\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Bacteroidetes VC2.1 Bac22: No match, genus\n",
      "Bacteroidetes VC2.1 Bac22: No match, family\n",
      "Bacteroidetes VC2.1 Bac22: No match, order\n",
      "Candidatus Iainarchaeum: No match, genus\n",
      "UBA4486: No match, genus\n",
      "UBA4486: No match, family\n",
      "UBA4486: No match, order\n",
      "uncultured: No match, genus\n",
      "Thalassobaculales: No match, order\n",
      "T34: No match, genus\n",
      "T34: No match, family\n",
      "uncultured: No match, genus\n",
      "Bacteroidota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Diplorickettsiaceae: No match, family\n",
      "Diplorickettsiales: No match, order\n",
      "uncultured: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Iainarchaeales: No match, family\n",
      "Iainarchaeales: No match, order\n",
      "Lineage IIb: No match, genus\n",
      "Lineage IIb: No match, family\n",
      "Lineage IIb: No match, order\n",
      "Lineage IIb: No match, class\n",
      "Geminicoccaceae: No match, family\n",
      "AKYH767: No match, genus\n",
      "AKYH767: No match, family\n",
      "uncultured: No match, genus\n",
      "Rubinisphaeraceae: No match, family\n",
      "Candidatus Endoecteinascidia: No match, genus\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "BD7 11: No match, order\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Iainarchaeia: No match, class\n",
      "Pirellulales: No match, order\n",
      "JGI 0000069 P22: No match, genus\n",
      "JGI 0000069 P22: No match, family\n",
      "JGI 0000069 P22: No match, order\n",
      "Gracilibacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Elusimicrobiota: No match, phylum\n",
      "N9D0: No match, genus\n",
      "N9D0: No match, family\n",
      "N9D0: No match, order\n",
      "N9D0: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Tistrellales: No match, order\n",
      "Dasania: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Iainarchaeota: No match, phylum\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM190: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "Clade I: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "uncultured: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "uncultured: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Unknown Family: No match, genus\n",
      "Unknown Family: No match, family\n",
      "Clostridium sensu stricto 1: No match, genus\n",
      "Geminocystis PCC 6308: No match, genusBabeliaceae: No match, genus\n",
      "\n",
      "Babeliaceae: No match, family\n",
      "Babeliales: No match, order\n",
      "Babeliae: No match, class\n",
      "Dependentiae: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Arctic97B 4 marine group: No match, genus\n",
      "Arctic97B 4 marine group: No match, family\n",
      "Arctic97B 4 marine group: No match, order\n",
      "PeM15: No match, genus\n",
      "PeM15: No match, family\n",
      "PeM15: No match, order\n",
      "Actinobacteria: Multiple matches, class \n",
      "Actinobacteriota: No match, phylum\n",
      "BD7 11: No match, genus\n",
      "BD7 11: No match, family\n",
      "BD7 11: No match, order\n",
      "BD7 11: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Bathyarchaeia: No match, genus\n",
      "Bathyarchaeia: No match, family\n",
      "Bathyarchaeia: No match, order\n",
      "Bathyarchaeia: No match, class\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "uncultured: No match, class\n",
      "pItb vmat 80: No match, genusDesulfobacterota: No match, phylum\n",
      "\n",
      "pItb vmat 80: No match, family\n",
      "pItb vmat 80: No match, order\n",
      "Candidatus Komeilibacteria: No match, genus\n",
      "Candidatus Komeilibacteria: No match, family\n",
      "Candidatus Komeilibacteria: No match, order\n",
      "ABY1: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "Candidatus Nitrosopelagicus: No match, genus\n",
      "wb1 A12: No match, genus\n",
      "Methylomirabilaceae: No match, family\n",
      "Methylomirabilales: No match, order\n",
      "Methylomirabilia: No match, class\n",
      "Methylomirabilota: No match, phylum\n",
      "Magnetospira: No match, genus\n",
      "Magnetospiraceae: No match, family\n",
      "Sumerlaea: No match, genus\n",
      "Sumerlaeaceae: No match, family\n",
      "Sumerlaeales: No match, order\n",
      "Sumerlaeia: No match, class\n",
      "Sumerlaeota: No match, phylum\n",
      "Dadabacteriales: No match, genus\n",
      "Dadabacteriales: No match, family\n",
      "Dadabacteriales: No match, order\n",
      "Candidatus Berkiella: No match, genus\n",
      "Unknown Family: No match, family\n",
      "CHAB XI 27: No match, genus\n",
      "CHAB XI 27: No match, family\n",
      "CHAB XI 27: No match, order\n",
      "0319 6G20: No match, genus\n",
      "0319 6G20: No match, family\n",
      "0319 6G20: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Oligoflexales: No match, genus\n",
      "Oligoflexales: No match, family\n",
      "Oligoflexales: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "NB1 j: No match, genus\n",
      "NB1 j: No match, family\n",
      "NB1 j: No match, order\n",
      "NB1 j: No match, class\n",
      "NB1 j: No match, phylum\n",
      "Woesearchaeales: No match, genus\n",
      "Woesearchaeales: No match, family\n",
      "Woesearchaeales: No match, order\n",
      "SEEP SRB1: No match, genus\n",
      "Desulfosarcinaceae: No match, family\n",
      "Dadabacteriia: No match, class\n",
      "Pseudolabrys: No match, genus\n",
      "Dadabacteria: No match, phylum\n",
      "AqS1: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Nitrosococcales: No match, order\n",
      "Milano WF1B 03: No match, genus\n",
      "Anaerovorax: No match, genus\n",
      "Anaerovoracaceae: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Candidatus Chisholmbacteria: No match, genus\n",
      "Candidatus Chisholmbacteria: No match, family\n",
      "Candidatus Chisholmbacteria: No match, order\n",
      "Microgenomatia: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "OM27 clade: No match, genus\n",
      "Methylomonadaceae: No match, family\n",
      "Spirochaeta 2: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Prochlorococcus MIT9313: No match, genus\n",
      "Cyanobiaceae: No match, family\n",
      "SJA 15: No match, genus\n",
      "SJA 15: No match, family\n",
      "SJA 15: No match, order\n",
      "Candidatus Hepatoplasma: No match, genus\n",
      "uncultured: No match, genus\n",
      "Aliikangiella: No match, genus\n",
      "Kangiellaceae: No match, family\n",
      "BD2 7: No match, genusHungateiclostridiaceae: No match, family\n",
      "\n",
      "BD2 7: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Entomoplasmatales Incertae Sedis: No match, family\n",
      "Marinifilaceae: No match, family\n",
      "Marine Methylotrophic Group 2: No match, genus\n",
      "Marine Benthic Group A: No match, genus\n",
      "Marine Benthic Group A: No match, family\n",
      "Marine Benthic Group A: No match, order\n",
      "Roseobacter clade NAC11 7 lineage: No match, genus\n",
      "Latescibacteraceae: No match, genus\n",
      "Latescibacteraceae: No match, family\n",
      "Nitrososphaeria: No match, classMethylomonadaceae: No match, family\n",
      "\n",
      "Pir4 lineage: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Prolixibacteraceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Latescibacterales: No match, order\n",
      "Oceanobacterium: No match, genus\n",
      "uncultured: No match, genus\n",
      "Latescibacteria: No match, class\n",
      "Candidatus Obscuribacter: No match, genusuncultured: No match, genus\n",
      "\n",
      "Calditrichaceae: No match, genus\n",
      "Calditrichaceae: No match, family\n",
      "Calditrichales: No match, order\n",
      "Calditrichia: No match, class\n",
      "Calditrichota: No match, phylum\n",
      "Latescibacterota: No match, phylum\n",
      "Obscuribacteraceae: No match, family\n",
      "SG8 4: No match, genus\n",
      "SG8 4: No match, family\n",
      "Obscuribacterales: No match, order\n",
      "Vampirivibrionia: No match, class\n",
      "Schekmanbacteria: No match, genus\n",
      "Schekmanbacteria: No match, family\n",
      "Schekmanbacteria: No match, order\n",
      "Schekmanbacteria: No match, class\n",
      "Schekmanbacteria: No match, phylum\n",
      "OM75 clade: No match, genus\n",
      "Nisaeaceae: No match, family\n",
      "Thalassobaculales: No match, order\n",
      "MSBL9: No match, order\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "OM190: No match, genus\n",
      "OM190: No match, family\n",
      "OM190: No match, order\n",
      "OM27 clade: No match, genusOM190: No match, class\n",
      "\n",
      "Planctomycetota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Vicingus: No match, genus\n",
      "Patescibacteria: No match, phylum\n",
      "Marinimicrobia (SAR406 clade): No match, genus\n",
      "Marinimicrobia (SAR406 clade): No match, family\n",
      "Marinimicrobia (SAR406 clade): No match, order\n",
      "Marinimicrobia (SAR406 clade): No match, class\n",
      "Marinimicrobia (SAR406 clade): No match, phylum\n",
      "MBNT15: No match, genus\n",
      "MBNT15: No match, family\n",
      "MBNT15: No match, order\n",
      "MBNT15: No match, class\n",
      "MBNT15: No match, phylum\n",
      "Lentimonas: No match, genus\n",
      "HOC36: No match, genus\n",
      "HOC36: No match, family\n",
      "HOC36: No match, order\n",
      "JG30 KF CM66: No match, genus\n",
      "JG30 KF CM66: No match, family\n",
      "JG30 KF CM66: No match, order\n",
      "JG30 KF CM66: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Calorithrix: No match, genus\n",
      "S085: No match, genus\n",
      "S085: No match, family\n",
      "S085: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "4572 13: No match, genus\n",
      "4572 13: No match, family\n",
      "MSBL9: No match, order\n",
      "A714019: No match, genus\n",
      "NS11 12 marine group: No match, genus\n",
      "NS11 12 marine group: No match, familyCalditrichaceae: No match, family\n",
      "Psychrosphaera: No match, genus\n",
      "\n",
      "Calditrichales: No match, order\n",
      "HTCC5015: No match, genus\n",
      "Arenicellaceae: No match, family\n",
      "Arenicellales: No match, order\n",
      "Candidatus Yanofskybacteria: No match, genus\n",
      "Candidatus Yanofskybacteria: No match, family\n",
      "Candidatus Yanofskybacteria: No match, order\n",
      "Parcubacteria: No match, class\n",
      "Patescibacteria: No match, phylum\n",
      "uncultured: No match, genus\n",
      "Calditrichia: No match, classArenicellaceae: No match, family\n",
      "\n",
      "Arenicellales: No match, order\n",
      "Calditrichota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "NKB5: No match, genus\n",
      "NKB5: No match, family\n",
      "NKB5: No match, order\n",
      "Salinirepens: No match, genus\n",
      "Crocinitomicaceae: No match, family\n",
      "Cutibacterium: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Geothermarchaeaceae: No match, genus\n",
      "Geothermarchaeaceae: No match, family\n",
      "Woeseia: No match, genusKapabacteriales: No match, genus\n",
      "\n",
      "Woeseiaceae: No match, familyKapabacteriales: No match, family\n",
      "\n",
      "Steroidobacterales: No match, orderKapabacteriales: No match, order\n",
      "\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Caldiarchaeales: No match, order\n",
      "Nitrososphaeria: No match, class\n",
      "Kapabacteria: No match, class\n",
      "Bacteroidota: No match, phylum\n",
      "Granulosicoccales: No match, order\n",
      "Macellibacteroides: No match, genus\n",
      "MSBL3: No match, genus\n",
      "Richelia HH01: No match, genus\n",
      "Tannerellaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, genus\n",
      "Kiritimatiellaceae: No match, family\n",
      "MSBL3: No match, genus\n",
      "Kiritimatiellaceae: No match, family\n",
      "Kiritimatiellales: No match, order\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Salinicola: No match, genus\n",
      "Kiritimatiellales: No match, order\n",
      "C86: No match, genus\n",
      "C86: No match, family\n",
      "C86: No match, order\n",
      "Nitrospira: Multiple matches, genus \n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "Nordella: No match, genus\n",
      "FS142 36B 02: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Nitrosococcales: No match, order\n",
      "hgcI clade: No match, genus\n",
      "uncultured: No match, genus\n",
      "Sandaracinaceae: No match, family\n",
      "Polyangiales: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Unknown Family: No match, genus\n",
      "Unknown Family: No match, family\n",
      "uncultured: No match, genus\n",
      "BD7 8: No match, genus\n",
      "BD7 8: No match, family\n",
      "BD7 8: No match, order\n",
      "Vallitalea: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "Acuticoccus: No match, genus\n",
      "Unknown Family: No match, family\n",
      "IMCC26256: No match, genus\n",
      "IMCC26256: No match, family\n",
      "IMCC26256: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Desulfobulbales: No match, order\n",
      "Desulfobulbia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "UASB TL25: No match, genus\n",
      "UASB TL25: No match, family\n",
      "UASB TL25: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Erysipelatoclostridium: No match, genus\n",
      "Edaphobaculum: No match, genus\n",
      "Erysipelatoclostridiaceae: No match, family\n",
      "Chalicogloea CCALA 975: No match, genus\n",
      "IMCC26207: No match, genus\n",
      "Microtrichaceae: No match, family\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "C86: No match, genus\n",
      "C86: No match, family\n",
      "C86: No match, order\n",
      "P.palmC41: No match, genus\n",
      "P.palmC41: No match, family\n",
      "P.palmC41: No match, order\n",
      "0319 6G20: No match, genus\n",
      "0319 6G20: No match, family\n",
      "0319 6G20: No match, order\n",
      "Oligoflexia: No match, class\n",
      "Bdellovibrionota: No match, phylum\n",
      "Lentisphaeria: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "SAR11 clade: No match, order\n",
      "NS11 12 marine group: No match, genus\n",
      "NS11 12 marine group: No match, family\n",
      "WCHB1 41: No match, genus\n",
      "WCHB1 41: No match, family\n",
      "WCHB1 41: No match, order\n",
      "AncK6: No match, genus\n",
      "AncK6: No match, family\n",
      "AncK6: No match, order\n",
      "AncK6: No match, class\n",
      "AncK6: No match, phylum\n",
      "Kiritimatiellae: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "BD1 7 clade: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Rubidimonas: No match, genus\n",
      "Candidatus Raymondbacteria: No match, genus\n",
      "Candidatus Raymondbacteria: No match, family\n",
      "Silicimonas: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Halocynthiibacter: No match, genus\n",
      "Geothermomicrobium: No match, genus\n",
      "Dadabacteriales: No match, genus\n",
      "Dadabacteriales: No match, family\n",
      "Dadabacteriales: No match, order\n",
      "Dadabacteriia: No match, class\n",
      "Dadabacteria: No match, phylum\n",
      "AKAU3564 sediment group: No match, genus\n",
      "AKAU3564 sediment group: No match, family\n",
      "Sinobacterium: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "NS5 marine group: No match, genus\n",
      "vadinHA49: No match, genus\n",
      "vadinHA49: No match, family\n",
      "vadinHA49: No match, order\n",
      "vadinHA49: No match, class\n",
      "Planctomycetota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, genus\n",
      "Candidatus Latescibacter: No match, genus\n",
      "Latescibacteraceae: No match, family\n",
      "Latescibacterales: No match, order\n",
      "Latescibacteria: No match, class\n",
      "Latescibacterota: No match, phylum\n",
      "Nitrososphaeria: No match, class\n",
      "uncultured: No match, genus\n",
      "Peptostreptococcales Tissierellales: No match, family\n",
      "Peptostreptococcales Tissierellales: No match, order\n",
      "Verruc 01: No match, genus\n",
      "Allorhizobium Neorhizobium Pararhizobium Rhizobium: No match, genus\n",
      "Blastocatella: No match, genus\n",
      "Blastocatellaceae: No match, family\n",
      "Blastocatellales: No match, order\n",
      "Blastocatellia: No match, class\n",
      "Acidobacteriota: No match, phylum\n",
      "Thaumasiovibrio: No match, genus\n",
      "MD2894 B20: No match, genus\n",
      "MD2894 B20: No match, family\n",
      "MD2894 B20: No match, order\n",
      "NS9 marine group: No match, genus\n",
      "NS9 marine group: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Elusimicrobia: Multiple matches, class \n",
      "Elusimicrobiota: No match, phylum\n",
      "Bradymonadales: No match, genus\n",
      "Bradymonadales: No match, family\n",
      "Bradymonadales: No match, order\n",
      "BD2 11 terrestrial group: No match, genusOM43 clade: No match, genus\n",
      "\n",
      "BD2 11 terrestrial group: No match, family\n",
      "BD2 11 terrestrial group: No match, order\n",
      "BD2 11 terrestrial group: No match, class\n",
      "Desulfuromonadia: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "Nisaeaceae: No match, family\n",
      "Thalassobaculales: No match, order\n",
      "Gemmatimonadota: No match, phylum\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Verruc 01: No match, genus\n",
      "Calditrichaceae: No match, genus\n",
      "Calditrichaceae: No match, family\n",
      "Calditrichales: No match, order\n",
      "Calditrichia: No match, class\n",
      "Calditrichota: No match, phylum\n",
      "S1 3 65: No match, genus\n",
      "Woeseia: No match, genusS1 3 65: No match, family\n",
      "\n",
      "S1 3 65: No match, orderWoeseiaceae: No match, family\n",
      "\n",
      "Steroidobacterales: No match, order\n",
      "Clostridia vadinBB60 group: No match, genus\n",
      "Clostridia vadinBB60 group: No match, family\n",
      "Clostridia vadinBB60 group: No match, order\n",
      "NS4 marine group: No match, genusRoseimarinus: No match, genus\n",
      "\n",
      "Prolixibacteraceae: No match, family\n",
      "Salinimicrobium: No match, genus\n",
      "IheB3 7: No match, genus\n",
      "Melioribacteraceae: No match, family\n",
      "Ignavibacteriales: No match, order\n",
      "Ignavibacteria: No match, class\n",
      "Bacteroidota: No match, phylum\n",
      "uncultured: No match, genus\n",
      "WD2101 soil group: No match, genus\n",
      "WD2101 soil group: No match, family\n",
      "Marinilabiliaceae: No match, family\n",
      "Tepidisphaerales: No match, order\n",
      "Deferrisoma: No match, genus\n",
      "Defferrisomataceae: No match, familyHalobacteriovorax: No match, genus\n",
      "\n",
      "Defferrisomatales: No match, orderWenyingzhuangia: No match, genus\n",
      "\n",
      "uncultured: No match, genus\n",
      "Defferrisomatia: No match, classIlumatobacteraceae: No match, family\n",
      "\n",
      "Microtrichales: No match, order\n",
      "Acidimicrobiia: No match, class\n",
      "Actinobacteriota: No match, phylum\n",
      "Deferrisomatota: No match, phylum\n",
      "SM1A02: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "NS11 12 marine group: No match, genus\n",
      "NS11 12 marine group: No match, family\n",
      "Nitrospira: Multiple matches, genus \n",
      "LS NOB: No match, genus\n",
      "uncultured: No match, genus\n",
      "Symphothece PCC 7002: No match, genus\n",
      "Phaeodactylibacter: No match, genus\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "Chitinophagales: No match, order\n",
      "uncultured: No match, genus\n",
      "uncultured: No match, family\n",
      "uncultured: No match, order\n",
      "uncultured: No match, class\n",
      "Desulfobacterota: No match, phylum\n",
      "FS142 36B 02: No match, genus\n",
      "Nitrosococcaceae: No match, family\n",
      "Nitrosococcales: No match, order\n",
      "37 13: No match, genus\n",
      "37 13: No match, family\n",
      "Chitinophagales: No match, order\n",
      "uncultured: No match, genus\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "uncultured: No match, genus\n",
      "Blastocatellaceae: No match, family\n",
      "Chloroplast: No match, genus\n",
      "Chloroplast: No match, family\n",
      "Chloroplast: No match, order\n",
      "Cyanobacteriia: No match, class\n",
      "Omnitrophales: No match, genus\n",
      "Omnitrophales: No match, family\n",
      "Omnitrophales: No match, order\n",
      "Omnitrophia: No match, class\n",
      "Verrucomicrobiota: No match, phylum\n",
      "DEV007: No match, genus\n",
      "DEV007: No match, family\n",
      "Fusibacter: No match, genusBlastocatellales: No match, order\n",
      "\n",
      "Blastocatellia: No match, classFusibacteraceae: No match, family\n",
      "\n",
      "Acidobacteriota: No match, phylumPeptostreptococcales Tissierellales: No match, order\n",
      "\n",
      "Pir4 lineage: No match, genus\n",
      "Pirellulaceae: No match, family\n",
      "Pirellulales: No match, order\n",
      "S25 593: No match, genus\n",
      "S25 593: No match, family\n",
      "uncultured: No match, genus\n",
      "Microscillaceae: No match, familyCraurococcus Caldovatus: No match, genus\n",
      "\n",
      "Polyangia: No match, genus\n",
      "Polyangia: No match, family\n",
      "Polyangia: No match, order\n",
      "Polyangia: No match, class\n",
      "Myxococcota: No match, phylum\n",
      "Emcibacter: No match, genus\n",
      "Emcibacteraceae: No match, family\n",
      "Emcibacterales: No match, order\n",
      "uncultured: No match, genus\n",
      "Leptotrichiaceae: No match, family\n",
      "uncultured: No match, genus\n",
      "Candidatus Bacilloplasma: No match, genus\n",
      "Spongiibacteraceae: No match, genus\n",
      "Spongiibacteraceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "vadinBA26: No match, genus\n",
      "vadinBA26: No match, family\n",
      "vadinBA26: No match, order\n",
      "Dehalococcoidia: No match, class\n",
      "Chloroflexi: Multiple matches, phylum \n",
      "Marimicrobium: No match, genus\n",
      "Halieaceae: No match, family\n",
      "Cellvibrionales: No match, order\n",
      "Hyphococcus: No match, genus\n",
      "uncultured: No match, genus\n",
      "Labrys: Multiple matches, genus \n",
      "Labraceae: No match, family\n",
      "Thiohalophilus: No match, genus\n",
      "Thioalkalispiraceae: No match, family\n",
      "Ectothiorhodospirales: No match, order\n",
      "Fervidicella: No match, genus\n",
      "Caloramatoraceae: No match, family\n",
      "JL ETNP F27: No match, genus\n",
      "Clade Ia: No match, genus\n",
      "Clade I: No match, family\n",
      "SAR11 clade: No match, order\n",
      "Flavivirga: No match, genus\n",
      "Roseimarinus: No match, genus\n",
      "Prolixibacteraceae: No match, family\n",
      "Litoribrevibacter: No match, genus\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    worms_16s = get_worms_from_scientific_name_parallel(\n",
    "    tax_df = tax_16S,ordered_rank_columns=['genus','family','order','class','phylum','domain'],\n",
    "    full_tax_column=\"taxonomy\",full_tax_vI=True,n_proc=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>genus</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:480248</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>class</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>order</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:345514</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanophyceae</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o_...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o_...</td>\n",
       "      <td>family</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570956</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>class</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>d__Archaea; p__Crenarchaeota; c__Nitrososphaer...</td>\n",
       "      <td>d__Archaea; p__Crenarchaeota; c__Nitrososphaer...</td>\n",
       "      <td>family</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:559432</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Thaumarchaeota</td>\n",
       "      <td>Thaumarchaeota incertae sedis</td>\n",
       "      <td>Nitrosopumilales</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>class</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>phylum</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:146537</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Phylum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>d__Bacteria; p__Planctomycetota; c__Pla3_linea...</td>\n",
       "      <td>d__Bacteria; p__Planctomycetota; c__Pla3_linea...</td>\n",
       "      <td>domain</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>d__Bacteria; p__Actinobacteriota; c__Acidimicr...</td>\n",
       "      <td>d__Bacteria; p__Actinobacteriota; c__Acidimicr...</td>\n",
       "      <td>domain</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>genus</td>\n",
       "      <td>Comamonas</td>\n",
       "      <td>Comamonas</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:569917</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Comamonadaceae</td>\n",
       "      <td>Comamonas</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>class</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>phylum</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:146537</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Phylum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>domain</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>class</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             full_tax  \\\n",
       "0   d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "1   d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "2   d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...   \n",
       "3   d__Bacteria; p__Myxococcota; c__Myxococcia; o_...   \n",
       "4   d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "5   d__Archaea; p__Crenarchaeota; c__Nitrososphaer...   \n",
       "6   d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "7   d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...   \n",
       "8   d__Bacteria; p__Planctomycetota; c__Pla3_linea...   \n",
       "9   d__Bacteria; p__Actinobacteriota; c__Acidimicr...   \n",
       "10  d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "11  d__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
       "12  d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...   \n",
       "13  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...   \n",
       "14  d__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
       "\n",
       "                               verbatimIdentification old_taxonRank  \\\n",
       "0   d__Bacteria; p__Proteobacteria; c__Gammaproteo...         genus   \n",
       "1   d__Bacteria; p__Proteobacteria; c__Gammaproteo...         class   \n",
       "2   d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...         order   \n",
       "3   d__Bacteria; p__Myxococcota; c__Myxococcia; o_...        family   \n",
       "4   d__Bacteria; p__Proteobacteria; c__Gammaproteo...         class   \n",
       "5   d__Archaea; p__Crenarchaeota; c__Nitrososphaer...        family   \n",
       "6   d__Bacteria; p__Proteobacteria; c__Gammaproteo...         class   \n",
       "7   d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...        phylum   \n",
       "8   d__Bacteria; p__Planctomycetota; c__Pla3_linea...        domain   \n",
       "9   d__Bacteria; p__Actinobacteriota; c__Acidimicr...        domain   \n",
       "10  d__Bacteria; p__Proteobacteria; c__Gammaproteo...         genus   \n",
       "11  d__Bacteria; p__Proteobacteria; c__Alphaproteo...         class   \n",
       "12  d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...        phylum   \n",
       "13  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...        domain   \n",
       "14  d__Bacteria; p__Proteobacteria; c__Alphaproteo...         class   \n",
       "\n",
       "               old name       scientificName  \\\n",
       "0                Vibrio               Vibrio   \n",
       "1   Gammaproteobacteria  Gammaproteobacteria   \n",
       "2       Synechococcales      Synechococcales   \n",
       "3         Myxococcaceae        Myxococcaceae   \n",
       "4   Gammaproteobacteria  Gammaproteobacteria   \n",
       "5     Nitrosopumilaceae    Nitrosopumilaceae   \n",
       "6   Gammaproteobacteria  Gammaproteobacteria   \n",
       "7         Cyanobacteria        Cyanobacteria   \n",
       "8              Bacteria             Bacteria   \n",
       "9              Bacteria             Bacteria   \n",
       "10            Comamonas            Comamonas   \n",
       "11  Alphaproteobacteria  Alphaproteobacteria   \n",
       "12        Cyanobacteria        Cyanobacteria   \n",
       "13             Bacteria             Bacteria   \n",
       "14  Alphaproteobacteria  Alphaproteobacteria   \n",
       "\n",
       "                             scientificNameID   kingdom          phylum  \\\n",
       "0   urn:lsid:marinespecies.org:taxname:480248  Bacteria  Proteobacteria   \n",
       "1   urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "2   urn:lsid:marinespecies.org:taxname:345514  Bacteria   Cyanobacteria   \n",
       "3   urn:lsid:marinespecies.org:taxname:570956  Bacteria  Proteobacteria   \n",
       "4   urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "5   urn:lsid:marinespecies.org:taxname:559432   Archaea  Thaumarchaeota   \n",
       "6   urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "7   urn:lsid:marinespecies.org:taxname:146537  Bacteria   Cyanobacteria   \n",
       "8        urn:lsid:marinespecies.org:taxname:6  Bacteria            None   \n",
       "9        urn:lsid:marinespecies.org:taxname:6  Bacteria            None   \n",
       "10  urn:lsid:marinespecies.org:taxname:569917  Bacteria  Proteobacteria   \n",
       "11  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "12  urn:lsid:marinespecies.org:taxname:146537  Bacteria   Cyanobacteria   \n",
       "13       urn:lsid:marinespecies.org:taxname:6  Bacteria            None   \n",
       "14  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "\n",
       "                            class             order             family  \\\n",
       "0             Gammaproteobacteria       Vibrionales       Vibrionaceae   \n",
       "1             Gammaproteobacteria              None               None   \n",
       "2                    Cyanophyceae   Synechococcales               None   \n",
       "3             Deltaproteobacteria      Myxococcales      Myxococcaceae   \n",
       "4             Gammaproteobacteria              None               None   \n",
       "5   Thaumarchaeota incertae sedis  Nitrosopumilales  Nitrosopumilaceae   \n",
       "6             Gammaproteobacteria              None               None   \n",
       "7                            None              None               None   \n",
       "8                            None              None               None   \n",
       "9                            None              None               None   \n",
       "10             Betaproteobacteria   Burkholderiales     Comamonadaceae   \n",
       "11            Alphaproteobacteria              None               None   \n",
       "12                           None              None               None   \n",
       "13                           None              None               None   \n",
       "14            Alphaproteobacteria              None               None   \n",
       "\n",
       "        genus taxonRank  \n",
       "0      Vibrio     Genus  \n",
       "1        None     Class  \n",
       "2        None     Order  \n",
       "3        None    Family  \n",
       "4        None     Class  \n",
       "5        None    Family  \n",
       "6        None     Class  \n",
       "7        None    Phylum  \n",
       "8        None   Kingdom  \n",
       "9        None   Kingdom  \n",
       "10  Comamonas     Genus  \n",
       "11       None     Class  \n",
       "12       None    Phylum  \n",
       "13       None   Kingdom  \n",
       "14       None     Class  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>d__Eukaryota</td>\n",
       "      <td>d__Eukaryota</td>\n",
       "      <td>domain</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         full_tax verbatimIdentification old_taxonRank   old name  \\\n",
       "242  d__Eukaryota           d__Eukaryota        domain  Eukaryota   \n",
       "\n",
       "    scientificName scientificNameID kingdom phylum class order family genus  \\\n",
       "242       No match             None     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "\n",
       "    taxonRank  \n",
       "242       NaN  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s[worms_16s[\"scientificName\"]==\"No match\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_16s.loc[worms_16s[\"scientificName\"]==\"No match\",'scientificName'] = \"Biota\"\n",
    "worms_16s.loc[worms_16s[\"scientificName\"]==\"Biota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      full_tax verbatimIdentification old_taxonRank old name scientificName  \\\n",
       "97  Unassigned             Unassigned           NaN      NaN            NaN   \n",
       "\n",
       "   scientificNameID kingdom phylum class order family genus taxonRank  \n",
       "97              NaN     NaN    NaN   NaN   NaN    NaN   NaN       NaN  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s[worms_16s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 13)\n",
      "(0, 13)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(worms_16s[worms_16s['scientificName'].isna() == True].shape)\n",
    "worms_16s.loc[worms_16s['scientificName'].isna() == True,'scientificName'] = 'incertae sedis'\n",
    "worms_16s.loc[worms_16s['scientificName'] == 'incertae sedis','scientificNameID'] =  'urn:lsid:marinespecies.org:taxname:12'\n",
    "print(worms_16s[worms_16s['scientificName'].isna() == True].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_16s.to_csv(\"../gomecc_v2_processed/worms_16S_matching.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:480248</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:345514</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanophyceae</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o_...</td>\n",
       "      <td>d__Bacteria; p__Myxococcota; c__Myxococcia; o_...</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570956</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Myxococcales</td>\n",
       "      <td>Myxococcaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:393018</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            full_tax  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "2  d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o_...   \n",
       "4  d__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
       "\n",
       "                              verbatimIdentification       scientificName  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Gammaproteo...               Vibrio   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Gammaproteo...  Gammaproteobacteria   \n",
       "2  d__Bacteria; p__Cyanobacteria; c__Cyanobacteri...      Synechococcales   \n",
       "3  d__Bacteria; p__Myxococcota; c__Myxococcia; o_...        Myxococcaceae   \n",
       "4  d__Bacteria; p__Proteobacteria; c__Gammaproteo...  Gammaproteobacteria   \n",
       "\n",
       "                            scientificNameID   kingdom          phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:480248  Bacteria  Proteobacteria   \n",
       "1  urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "2  urn:lsid:marinespecies.org:taxname:345514  Bacteria   Cyanobacteria   \n",
       "3  urn:lsid:marinespecies.org:taxname:570956  Bacteria  Proteobacteria   \n",
       "4  urn:lsid:marinespecies.org:taxname:393018  Bacteria  Proteobacteria   \n",
       "\n",
       "                 class            order         family   genus taxonRank  \n",
       "0  Gammaproteobacteria      Vibrionales   Vibrionaceae  Vibrio     Genus  \n",
       "1  Gammaproteobacteria             None           None    None     Class  \n",
       "2         Cyanophyceae  Synechococcales           None    None     Order  \n",
       "3  Deltaproteobacteria     Myxococcales  Myxococcaceae    None    Family  \n",
       "4  Gammaproteobacteria             None           None    None     Class  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_16s.drop(columns=['old name','old_taxonRank'],inplace=True)\n",
    "worms_16s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Parvibaculales</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>OCS116 clade</td>\n",
       "      <td>uncultured marine</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Puniceispirillales</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>SAR116 clade</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Thermoplasmatota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>Marine Group II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>Marinimicrobia (SAR406 clade)</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            featureid  \\\n",
       "182  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "225  00e6c13fe86364a5084987093afa1916   \n",
       "347  015dad1fafca90944d905beb2a980bc3   \n",
       "412  019c88c6ade406f731954f38e3461564   \n",
       "719  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                              sequence  \\\n",
       "182  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "225  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "347  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...   \n",
       "412  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...   \n",
       "719  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...   \n",
       "\n",
       "                                              taxonomy  Confidence    domain  \\\n",
       "182  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.832190  Bacteria   \n",
       "225  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.867040  Bacteria   \n",
       "347  d__Archaea; p__Thermoplasmatota; c__Thermoplas...    1.000000   Archaea   \n",
       "412  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.952911  Bacteria   \n",
       "719  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...    0.818195  Bacteria   \n",
       "\n",
       "                            phylum                          class  \\\n",
       "182                 Proteobacteria            Alphaproteobacteria   \n",
       "225                 Proteobacteria            Alphaproteobacteria   \n",
       "347               Thermoplasmatota                 Thermoplasmata   \n",
       "412                 Proteobacteria            Alphaproteobacteria   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                             order                         family  \\\n",
       "182                 Parvibaculales                   OCS116 clade   \n",
       "225             Puniceispirillales                   SAR116 clade   \n",
       "347                Marine Group II                Marine Group II   \n",
       "412                  Rickettsiales                   Mitochondria   \n",
       "719  Marinimicrobia (SAR406 clade)  Marinimicrobia (SAR406 clade)   \n",
       "\n",
       "                             genus               species  \\\n",
       "182                   OCS116 clade     uncultured marine   \n",
       "225                   SAR116 clade                   NaN   \n",
       "347                Marine Group II                   NaN   \n",
       "412                   Mitochondria  uncultured bacterium   \n",
       "719  Marinimicrobia (SAR406 clade)  uncultured bacterium   \n",
       "\n",
       "                    eventID  organismQuantity  \\\n",
       "182  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "225  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "347  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "412  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "719  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                          occurrenceID  \n",
       "182  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...  \n",
       "225  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...  \n",
       "347  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...  \n",
       "412  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...  \n",
       "719  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S rRNA'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Occurrence and worms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(165158, 14)"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['16S rRNA'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Order</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...   \n",
       "\n",
       "                                            taxonomy  Confidence  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.832190   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.867040   \n",
       "2  d__Archaea; p__Thermoplasmatota; c__Thermoplas...    1.000000   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.952911   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...    0.818195   \n",
       "\n",
       "                  eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "1  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "2  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "3  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "4  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...   \n",
       "\n",
       "                              verbatimIdentification       scientificName  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "2  d__Archaea; p__Thermoplasmatota; c__Thermoplas...       Thermoplasmata   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteo...        Rickettsiales   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...             Bacteria   \n",
       "\n",
       "                            scientificNameID   kingdom          phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "1  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "2  urn:lsid:marinespecies.org:taxname:416268   Archaea   Euryarchaeota   \n",
       "3  urn:lsid:marinespecies.org:taxname:570969  Bacteria  Proteobacteria   \n",
       "4       urn:lsid:marinespecies.org:taxname:6  Bacteria            None   \n",
       "\n",
       "                 class          order family genus taxonRank  \n",
       "0  Alphaproteobacteria           None   None  None     Class  \n",
       "1  Alphaproteobacteria           None   None  None     Class  \n",
       "2       Thermoplasmata           None   None  None     Class  \n",
       "3  Alphaproteobacteria  Rickettsiales   None  None     Order  \n",
       "4                 None           None   None  None   Kingdom  "
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "occ16_test = occ['16S rRNA'].copy()\n",
    "occ16_test.drop(columns=['domain','phylum','class','order','family','genus','species'],inplace=True)\n",
    "#occ16_test.drop(columns=['old name'],inplace=True)\n",
    "\n",
    "occ16_test = occ16_test.merge(worms_16s, how='left', left_on ='taxonomy', right_on='full_tax')\n",
    "occ16_test.drop(columns='full_tax', inplace=True)\n",
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_gene</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>controls_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>411</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392695 | https://github.com/ao...</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>260</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392706 | https://pr2-database....</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target_gene  ampliconSize trim_method                    cluster_method  \\\n",
       "0    16S rRNA           411    cutadapt  Tourmaline; qiime2-2021.2; dada2   \n",
       "1    18S rRNA           260    cutadapt  Tourmaline; qiime2-2021.2; dada2   \n",
       "\n",
       "  pid_clustering                                  taxa_class_method  \\\n",
       "0            ASV  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1            ASV  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                         taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "1  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...   \n",
       "\n",
       "                             code_repo  \\\n",
       "0  https://github.com/aomlomics/gomecc   \n",
       "1  https://github.com/aomlomics/gomecc   \n",
       "\n",
       "                            identificationReferences  \\\n",
       "0  10.5281/zenodo.8392695 | https://github.com/ao...   \n",
       "1  10.5281/zenodo.8392706 | https://pr2-database....   \n",
       "\n",
       "                                       controls_used  \n",
       "0  12 distilled water blanks | 2 PCR no-template ...  \n",
       "1  12 distilled water blanks | 2 PCR no-template ...  "
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '16S rRNA','taxa_class_method'].item()\n",
    "occ16_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '16S rRNA','taxa_ref_db'].item()\n",
    "\n",
    "occ16_test['identificationRemarks'] = occ16_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+occ16_test['Confidence'].astype(str) +\", against reference database: \"+occ16_test['taxa_ref_db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.832189583, against reference database: Silva SSU Ref NR 99 v138.1; 515f-926r region; 10.5281/zenodo.8392695'"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test['identificationRemarks'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### taxonID, basisOfRecord, eventID, nameAccordingTo, organismQuantityType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['taxonID'] = 'ASV:'+occ16_test['featureid']\n",
    "occ16_test['basisOfRecord'] = 'MaterialSample'\n",
    "occ16_test['nameAccordingTo'] = \"WoRMS\"\n",
    "occ16_test['organismQuantityType'] = \"DNA sequence reads\"\n",
    "occ16_test['recordedBy'] = data['study_data']['recordedBy'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### associatedSequences, materialSampleID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_name', 'library_id', 'title', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'lib_layout', 'platform',\n",
       "       'instrument_model', 'design_description', 'filetype', 'filename',\n",
       "       'filename2', 'biosample_accession', 'sra_accession', 'seq_method',\n",
       "       'nucl_acid_ext', 'target_gene', 'target_subfragment',\n",
       "       'pcr_primer_forward', 'pcr_primer_reverse', 'pcr_primer_name_forward',\n",
       "       'pcr_primer_name_reverse', 'pcr_primer_reference', 'pcr_cond',\n",
       "       'nucl_acid_amp', 'adapters', 'mid_barcode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test = occ16_test.merge(data['prep_data'].loc[data['prep_data']['target_gene'] == '16S rRNA',['sample_name','sra_accession','biosample_accession']], how='left', left_on ='eventID', right_on='sample_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>...</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>biosample_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.832190</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>...</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.867040</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>...</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>...</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.952911</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>...</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>0.818195</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>...</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>SRR26148187</td>\n",
       "      <td>SAMN37516094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  00c4c1c65d8669ed9f07abe149f9a01d   \n",
       "1  00e6c13fe86364a5084987093afa1916   \n",
       "2  015dad1fafca90944d905beb2a980bc3   \n",
       "3  019c88c6ade406f731954f38e3461564   \n",
       "4  02dfb0869af4bf549d290d48e66e2196   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...   \n",
       "\n",
       "                                            taxonomy  Confidence  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.832190   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.867040   \n",
       "2  d__Archaea; p__Thermoplasmatota; c__Thermoplas...    1.000000   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.952911   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...    0.818195   \n",
       "\n",
       "                  eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A                18   \n",
       "1  GOMECC4_27N_Sta1_DCM_A                36   \n",
       "2  GOMECC4_27N_Sta1_DCM_A                49   \n",
       "3  GOMECC4_27N_Sta1_DCM_A                 2   \n",
       "4  GOMECC4_27N_Sta1_DCM_A                 3   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...   \n",
       "\n",
       "                              verbatimIdentification       scientificName  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "2  d__Archaea; p__Thermoplasmatota; c__Thermoplas...       Thermoplasmata   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteo...        Rickettsiales   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...             Bacteria   \n",
       "\n",
       "                            scientificNameID  ...  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:392750  ...   \n",
       "1  urn:lsid:marinespecies.org:taxname:392750  ...   \n",
       "2  urn:lsid:marinespecies.org:taxname:416268  ...   \n",
       "3  urn:lsid:marinespecies.org:taxname:570969  ...   \n",
       "4       urn:lsid:marinespecies.org:taxname:6  ...   \n",
       "\n",
       "                                         taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "\n",
       "                               identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman   \n",
       "\n",
       "              sample_name sra_accession biosample_accession  \n",
       "0  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "1  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "2  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "3  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "4  GOMECC4_27N_Sta1_DCM_A   SRR26148187        SAMN37516094  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### eventID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['eventID'] = occ16_test['eventID']+\"_16S\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sampleSize by total number of reads per sample\n",
    "x = asv_tables['16S rRNA'].sum(numeric_only=True).astype('int')\n",
    "x.index = x.index+\"_16S\"\n",
    "occ16_test['occurrenceRemarks'] = \"Total sampleSize in DNA sequence reads: \"+occ16_test['eventID'].map(x).astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ16_test.drop(columns=['sample_name','featureid','taxonomy','Confidence','taxa_class_method','taxa_ref_db'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test['associatedSequences'] = occ16_test['sra_accession']+' | '+ occ16_test['biosample_accession']+' | '+data['study_data']['bioproject_accession'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_test.rename(columns={'biosample_accession': 'materialSampleID',\n",
    "                  'sequence': 'DNA_sequence'},inplace=True)\n",
    "                   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ16_test.drop(columns=['sra_accession'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DNA_sequence', 'eventID', 'organismQuantity', 'occurrenceID',\n",
       "       'verbatimIdentification', 'scientificName', 'scientificNameID',\n",
       "       'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'taxonRank',\n",
       "       'identificationRemarks', 'taxonID', 'basisOfRecord', 'nameAccordingTo',\n",
       "       'organismQuantityType', 'recordedBy', 'materialSampleID',\n",
       "       'occurrenceRemarks', 'associatedSequences'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>occurrenceRemarks</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:00c4c1c65d8669ed9f07abe149f9a01d</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 16187</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:00e6c13fe86364a5084987093afa1916</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 16187</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>...</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:015dad1fafca90944d905beb2a980bc3</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 16187</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>Order</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:019c88c6ade406f731954f38e3461564</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 16187</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Kingdom</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:02dfb0869af4bf549d290d48e66e2196</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 16187</td>\n",
       "      <td>SRR26148187 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        DNA_sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S                18   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S                36   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S                49   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S                 2   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S                 3   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...   \n",
       "\n",
       "                              verbatimIdentification       scientificName  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "2  d__Archaea; p__Thermoplasmatota; c__Thermoplas...       Thermoplasmata   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteo...        Rickettsiales   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...             Bacteria   \n",
       "\n",
       "                            scientificNameID   kingdom          phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "1  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "2  urn:lsid:marinespecies.org:taxname:416268   Archaea   Euryarchaeota   \n",
       "3  urn:lsid:marinespecies.org:taxname:570969  Bacteria  Proteobacteria   \n",
       "4       urn:lsid:marinespecies.org:taxname:6  Bacteria            None   \n",
       "\n",
       "                 class  ... taxonRank  \\\n",
       "0  Alphaproteobacteria  ...     Class   \n",
       "1  Alphaproteobacteria  ...     Class   \n",
       "2       Thermoplasmata  ...     Class   \n",
       "3  Alphaproteobacteria  ...     Order   \n",
       "4                 None  ...   Kingdom   \n",
       "\n",
       "                               identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:00c4c1c65d8669ed9f07abe149f9a01d  MaterialSample           WoRMS   \n",
       "1  ASV:00e6c13fe86364a5084987093afa1916  MaterialSample           WoRMS   \n",
       "2  ASV:015dad1fafca90944d905beb2a980bc3  MaterialSample           WoRMS   \n",
       "3  ASV:019c88c6ade406f731954f38e3461564  MaterialSample           WoRMS   \n",
       "4  ASV:02dfb0869af4bf549d290d48e66e2196  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "                               occurrenceRemarks  \\\n",
       "0  Total sampleSize in DNA sequence reads: 16187   \n",
       "1  Total sampleSize in DNA sequence reads: 16187   \n",
       "2  Total sampleSize in DNA sequence reads: 16187   \n",
       "3  Total sampleSize in DNA sequence reads: 16187   \n",
       "4  Total sampleSize in DNA sequence reads: 16187   \n",
       "\n",
       "                        associatedSequences  \n",
       "0  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "1  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "2  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "3  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "4  SRR26148187 | SAMN37516094 | PRJNA887898  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge event and occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>locationID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>59</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>59</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>59</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C_18S</td>\n",
       "      <td>CAPECORAL_Sta141</td>\n",
       "      <td>2021-10-20T12:47-04:00</td>\n",
       "      <td>4</td>\n",
       "      <td>USA: Gulf of Mexico</td>\n",
       "      <td>25.574</td>\n",
       "      <td>-84.843</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Mexico, Gulf of</td>\n",
       "      <td>4</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    eventID        locationID  \\\n",
       "939      GOMECC4_CAPECORAL_Sta141_DCM_B_18S  CAPECORAL_Sta141   \n",
       "940      GOMECC4_CAPECORAL_Sta141_DCM_C_18S  CAPECORAL_Sta141   \n",
       "941  GOMECC4_CAPECORAL_Sta141_Surface_A_18S  CAPECORAL_Sta141   \n",
       "942  GOMECC4_CAPECORAL_Sta141_Surface_B_18S  CAPECORAL_Sta141   \n",
       "943  GOMECC4_CAPECORAL_Sta141_Surface_C_18S  CAPECORAL_Sta141   \n",
       "\n",
       "                  eventDate minimumDepthInMeters             locality  \\\n",
       "939  2021-10-20T12:47-04:00                   59  USA: Gulf of Mexico   \n",
       "940  2021-10-20T12:47-04:00                   59  USA: Gulf of Mexico   \n",
       "941  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "942  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "943  2021-10-20T12:47-04:00                    4  USA: Gulf of Mexico   \n",
       "\n",
       "    decimalLatitude decimalLongitude samplingProtocol        waterBody  \\\n",
       "939          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "940          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "941          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "942          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "943          25.574          -84.843      CTD rosette  Mexico, Gulf of   \n",
       "\n",
       "    maximumDepthInMeters                       parentEventID  \\\n",
       "939                   59      GOMECC4_CAPECORAL_Sta141_DCM_B   \n",
       "940                   59      GOMECC4_CAPECORAL_Sta141_DCM_C   \n",
       "941                    4  GOMECC4_CAPECORAL_Sta141_Surface_A   \n",
       "942                    4  GOMECC4_CAPECORAL_Sta141_Surface_B   \n",
       "943                    4  GOMECC4_CAPECORAL_Sta141_Surface_C   \n",
       "\n",
       "             datasetID geodeticDatum  \n",
       "939  noaa-aoml-gomecc4         WGS84  \n",
       "940  noaa-aoml-gomecc4         WGS84  \n",
       "941  noaa-aoml-gomecc4         WGS84  \n",
       "942  noaa-aoml-gomecc4         WGS84  \n",
       "943  noaa-aoml-gomecc4         WGS84  "
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_event_data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_merged = occ16_test.merge(all_event_data,how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>...</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>18</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>36</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:392750</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...</td>\n",
       "      <td>d__Archaea; p__Thermoplasmatota; c__Thermoplas...</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:416268</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>2</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...</td>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:570969</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>3</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...</td>\n",
       "      <td>d__Bacteria; p__Marinimicrobia_(SAR406_clade);...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:6</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        DNA_sequence  \\\n",
       "0  TACGGAGGGGGCTAACGTTGTTCGGAATTACTGGGCGTAAAGCGCG...   \n",
       "1  TACGAAGGGGGCGAGCGTTGTTCGGAATTACTGGGCGTAAAGGGCG...   \n",
       "2  TACCGGCGCCTCAAGTGGTAGTCGCTTTTATTGGGCCTAAAACGTC...   \n",
       "3  TACAGGAGGGACGAGTGTTACTCGGAATGATTAGGCGTAAAGGGTC...   \n",
       "4  TACGAGGGGTGCTAGCGTTGTCCGGAATAACTGGGCGTAAAGGGTC...   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S                18   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S                36   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S                49   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S                 2   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S                 3   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_16S_occ00c4c1c65d8669ed...   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_16S_occ00e6c13fe86364a5...   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_16S_occ015dad1fafca9094...   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_16S_occ019c88c6ade406f7...   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_16S_occ02dfb0869af4bf54...   \n",
       "\n",
       "                              verbatimIdentification       scientificName  \\\n",
       "0  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "1  d__Bacteria; p__Proteobacteria; c__Alphaproteo...  Alphaproteobacteria   \n",
       "2  d__Archaea; p__Thermoplasmatota; c__Thermoplas...       Thermoplasmata   \n",
       "3  d__Bacteria; p__Proteobacteria; c__Alphaproteo...        Rickettsiales   \n",
       "4  d__Bacteria; p__Marinimicrobia_(SAR406_clade);...             Bacteria   \n",
       "\n",
       "                            scientificNameID   kingdom          phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "1  urn:lsid:marinespecies.org:taxname:392750  Bacteria  Proteobacteria   \n",
       "2  urn:lsid:marinespecies.org:taxname:416268   Archaea   Euryarchaeota   \n",
       "3  urn:lsid:marinespecies.org:taxname:570969  Bacteria  Proteobacteria   \n",
       "4       urn:lsid:marinespecies.org:taxname:6  Bacteria            None   \n",
       "\n",
       "                 class  ... minimumDepthInMeters  \\\n",
       "0  Alphaproteobacteria  ...                   49   \n",
       "1  Alphaproteobacteria  ...                   49   \n",
       "2       Thermoplasmata  ...                   49   \n",
       "3  Alphaproteobacteria  ...                   49   \n",
       "4                 None  ...                   49   \n",
       "\n",
       "                                      locality decimalLatitude  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "\n",
       "  decimalLongitude samplingProtocol       waterBody maximumDepthInMeters  \\\n",
       "0          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "1          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "2          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "3          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "4          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "\n",
       "            parentEventID          datasetID geodeticDatum  \n",
       "0  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "1  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "2  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "3  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "4  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ16_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ16_merged.drop(columns=['DNA_sequence']).to_csv(\"../gomecc_v2_processed/occurrence_16S.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 18S worms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "18S PR2 database provides WORMS IDs for species that are in worms. We will read in that file, assign known worms ids, the do a search for unannotated taxa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "pr2_18S = pd.read_excel(\"../../../databases/18S_PR2/pr2_v5.0.0_SSU/pr2_version_5.0.0_taxonomy.xlsx\",\n",
    "    index_col=None, na_values=[\"\"])\n",
    "pr2_18S = pr2_18S.dropna(subset=['worms_id'])\n",
    "pr2_18S['worms_id'] = pr2_18S['worms_id'].astype('int').astype('str')\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('_',' ',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace(' sp\\.','',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace(' spp\\.','',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('-',' ',regex=True)\n",
    "pr2_18S['species'] = pr2_18S['species'].replace('\\/',' ',regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "pr2_18S_dict = dict(zip(pr2_18S.species,pr2_18S.worms_id))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'614894'"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(pr2_18S_dict['Aphanocapsa feldmannii'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### code to get record from aphia id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "import multiprocess as mp\n",
    "import pyworms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_worms_from_aphiaid_or_name(tax_df, worms_dict,ordered_rank_columns, queue,full_tax_column=\"taxonomy\", like=False, marine_only=False,full_tax_vI = False):\n",
    "    matches = []\n",
    "    w_ranks = ['kingdom','phylum','class','order','family','genus']\n",
    "    for index, row in tax_df.iterrows():\n",
    "        full_tax = row[full_tax_column]\n",
    "        if row['species'] in worms_dict.keys():\n",
    "            row_data = {'full_tax':full_tax,'verbatimIdentification': row['species']}\n",
    "            aid = worms_dict[row['species']]\n",
    "            record = pyworms.aphiaRecordByAphiaID(aid)\n",
    "            row_data.update({'taxonRank': 'species', 'old name': 'aphiaID'})\n",
    "            if record['status'] == 'accepted':\n",
    "                row_data['scientificName'] = record['scientificname']\n",
    "                row_data['scientificNameID'] = record['lsid']\n",
    "                row_data.update(dict(zip(w_ranks, [record.get(key) for key in w_ranks])))\n",
    "                row_data.update({'taxonRank': record['rank']})\n",
    "            elif record['status'] == 'unaccepted':\n",
    "                valid_name = record['valid_name']\n",
    "                if valid_name != None:\n",
    "                    v_match = pyworms.aphiaRecordsByName(valid_name,like=like,marine_only=marine_only)\n",
    "                    row_data['scientificName'] = v_match[0]['scientificname']\n",
    "                    row_data['scientificNameID'] = v_match[0]['lsid']\n",
    "                    row_data.update(dict(zip(w_ranks, [v_match[0].get(key) for key in w_ranks])))\n",
    "                    row_data.update({'taxonRank': v_match[0]['rank']})\n",
    "                    print(aid+\": Unaccepted, using \"+valid_name)\n",
    "                else:\n",
    "                    print(aid+\": Unaccepted, no valid name \")\n",
    "        else:\n",
    "            if full_tax_vI:\n",
    "                row_data = {'full_tax':full_tax,'verbatimIdentification': full_tax}\n",
    "            else:   \n",
    "                row_data = {'full_tax':full_tax,'verbatimIdentification': 'Null'}\n",
    "            for i in ordered_rank_columns:\n",
    "                rank = i\n",
    "                old_name = row[i]\n",
    "                if pd.isna(old_name):\n",
    "                    continue \n",
    "                else:\n",
    "                    row_data.update({'old_taxonRank': rank, 'old name': old_name})\n",
    "                    if row_data['verbatimIdentification'] == 'Null':\n",
    "                        row_data['verbatimIdentification'] = old_name\n",
    "                    s_match = pyworms.aphiaRecordsByName(old_name,like=like,marine_only=marine_only)\n",
    "                    #time.sleep(1)\n",
    "                    if s_match == None:\n",
    "                        row_data['scientificName'] = \"No match\"\n",
    "                        row_data['scientificNameID'] = \"None\"\n",
    "                        print(old_name+\": No match, \"+rank)\n",
    "                        continue\n",
    "                    elif len(s_match) > 1:\n",
    "                        mult = []\n",
    "                        for m in s_match:\n",
    "                            if m['status'] == 'accepted':\n",
    "                                mult = mult + [m]\n",
    "                        if len(mult) > 1:\n",
    "                            row_data['scientificName'] = \"Multiple matches\"\n",
    "                            row_data['scientificNameID'] = \"None\"\n",
    "                            print(old_name+\": Multiple matches, \"+rank+\" \")\n",
    "                        elif len(mult) < 1:\n",
    "                            row_data['scientificName'] = \"Multiple unaccepted matches\"\n",
    "                            row_data['scientificNameID'] = \"None\"\n",
    "                            print(old_name+\": Multiple unaccepted matches, \"+rank+\" \")\n",
    "                        elif len(mult) == 1:\n",
    "                            row_data['scientificName'] = mult[0]['scientificname']\n",
    "                            row_data['scientificNameID'] = mult[0]['lsid']\n",
    "                            row_data.update(dict(zip(w_ranks, [mult[0].get(key) for key in w_ranks])))\n",
    "                            row_data.update({'taxonRank': mult[0]['rank']})\n",
    "                            break\n",
    "                    elif len(s_match) == 1:\n",
    "                        if s_match[0]['status'] == 'accepted':\n",
    "                            row_data['scientificName'] = s_match[0]['scientificname']\n",
    "                            row_data['scientificNameID'] = s_match[0]['lsid']\n",
    "                            row_data.update(dict(zip(w_ranks, [s_match[0].get(key) for key in w_ranks])))\n",
    "                            row_data.update({'taxonRank': s_match[0]['rank']})\n",
    "                            break\n",
    "                        elif s_match[0]['status'] == 'unaccepted':\n",
    "                            valid_name = s_match[0]['valid_name']\n",
    "                            if valid_name != None:\n",
    "                                v_match = pyworms.aphiaRecordsByName(valid_name,like=like,marine_only=marine_only)\n",
    "                                row_data['scientificName'] = v_match[0]['scientificname']\n",
    "                                row_data['scientificNameID'] = v_match[0]['lsid']\n",
    "                                row_data.update(dict(zip(w_ranks, [v_match[0].get(key) for key in w_ranks])))\n",
    "                                row_data.update({'taxonRank': v_match[0]['rank']})\n",
    "                                print(old_name+\": Unaccepted, using \"+valid_name+\", \"+rank+\" \")\n",
    "                            else:\n",
    "                                print(old_name+\": Unaccepted, no valid name, \"+rank+\" \")\n",
    "        matches += [row_data]\n",
    "    matches = pd.DataFrame.from_dict(matches)\n",
    "    queue.put(matches)\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_worms_from_aphiaid_or_name_parallel(tax_df, worms_dict,ordered_rank_columns, full_tax_column=\"taxonomy\",like=False, marine_only=False,full_tax_vI = False,n_proc=0):\n",
    "    queue = mp.Queue()\n",
    "    if n_proc == 0:\n",
    "    # create as many processes as there are CPUs on your machine\n",
    "        num_processes = mp.cpu_count()\n",
    "    else:\n",
    "        num_processes = n_proc\n",
    "        \n",
    "    # calculate the chunk size as an integer\n",
    "    chunk_size = int(tax_df.shape[0]/num_processes)\n",
    "    procs = []\n",
    "    for job in range(num_processes):\n",
    "        start = job * chunk_size\n",
    "        end = start + chunk_size\n",
    "        df_chunk = tax_df.iloc[start:end]\n",
    "        proc = mp.Process(\n",
    "            target=get_worms_from_aphiaid_or_name,\n",
    "            args=(df_chunk,worms_dict,ordered_rank_columns, queue,full_tax_column,like,marine_only,full_tax_vI)\n",
    "        )\n",
    "        procs.append(proc)\n",
    "        proc.start()\n",
    "    \n",
    "    new_df = pd.DataFrame()\n",
    "    for _ in procs:\n",
    "        new_df = pd.concat([new_df,queue.get()])\n",
    "    \n",
    "    #new_df = queue.get()\n",
    "    \n",
    "    for proc in procs:\n",
    "        proc.join()\n",
    "    \n",
    "    return new_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Had some [issues with the parallelization](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr) on Mac M1. Adding 'OBJC_DISABLE_INITIALIZE_FORK_SAFETY = YES' to .bash_profile and then [This](https://github.com/python/cpython/issues/74570) fixed it.   \n",
    "Try to run without the bash_profile fix LATER."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"no_proxy\"]=\"*\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_18S = asv_tables['18S rRNA'][['taxonomy','domain','supergroup','division','subdivision','class','order','family','genus','species']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1374, 10)"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_18S = tax_18S.drop_duplicates()\n",
    "tax_18S.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aspergillus penicillioides: No match, speciesProtoscenium cf intricatum: No match, species\n",
      "\n",
      "Euglypha acanthophora: No match, species\n",
      "RAD B X Group IVe X: No match, species\n",
      "RAD B X Group IVe X: No match, genus\n",
      "RAD B X Group IVe: No match, family\n",
      "Eimeriida: No match, order\n",
      "Nibbleromonas: No match, genus\n",
      "RAD B X: No match, order\n",
      "Coccidiomorphea: No match, class\n",
      "MAST 12A: No match, species\n",
      "MAST 12A: No match, genus\n",
      "Solemya reidi: Unaccepted, using Petrasma pervernicosa, species \n",
      "Nibbleridae: No match, family\n",
      "RAD B: No match, class\n",
      "MAST 12: No match, family\n",
      "Nibbleridida: No match, order\n",
      "Opalozoa X: No match, order\n",
      "Malus x: No match, species\n",
      "Nibbleridea: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Euduboscquella cachoni: No match, species\n",
      "Obazoa: No match, supergroup\n",
      "Malus: No match, genus\n",
      "Nibbleridia X: No match, subdivision\n",
      "Pectinoida: No match, familyEmbryophyceae XX: No match, family\n",
      "\n",
      "Nibbleridia: No match, division\n",
      "Skeletonema menzellii: No match, species\n",
      "Embryophyceae X: No match, order\n",
      "Provora: No match, supergroup\n",
      "Dictyochales X: No match, species\n",
      "Embryophyceae: No match, classDictyochales X: No match, genus\n",
      "\n",
      "Archaeplastida: No match, supergroup\n",
      "Streptophyta X: No match, subdivision\n",
      "Oligotrichida XX: No match, species\n",
      "Oligotrichida XX: No match, genus\n",
      "Haplosporidium littoralis: No match, species\n",
      "MAST 4D: No match, species\n",
      "MAST 4D: No match, genus\n",
      "Oligotrichida X: No match, family\n",
      "MAST 4: No match, family\n",
      "Uroleptus gallina: No match, species\n",
      "Chlamydomonadales X: No match, family\n",
      "Maxillopoda: Unaccepted, no valid name, family \n",
      "Sagenista X: No match, order\n",
      "Dinophyceae XX: No match, family\n",
      "Choreotrichida Tintinnina: No match, order\n",
      "Siphonosphaera cyathina: No match, species\n",
      "Dinophyceae X: No match, order\n",
      "Siphonosphaera: Multiple matches, genus \n",
      "MAST 3I: No match, species\n",
      "MAST 3I: No match, genus\n",
      "MAST 3: No match, family\n",
      "Maxillopoda X: No match, species\n",
      "Maxillopoda X: No match, genus\n",
      "Maxillopoda: Unaccepted, no valid name, family \n",
      "Dino Group II Clade 30 X: No match, species\n",
      "Dino Group II Clade 30 X: No match, genus\n",
      "Nanomonadea: No match, order\n",
      "Amphibelone: No match, genus\n",
      "Dino Group II Clade 30: No match, family\n",
      "Acantharea F2: No match, family\n",
      "Dino Group II: No match, order\n",
      "Acantharea F: No match, order\n",
      "Cordagalma cordiforme: Unaccepted, using Cordagalma ordinatum, species \n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Chlorochromonas: Unaccepted, using Ochromonas, species \n",
      "Chlorochromonas: Unaccepted, using Ochromonas, genus \n",
      "Caecitellaceae X: No match, species\n",
      "Caecitellaceae X: No match, genus\n",
      "Caecitellaceae: No match, family\n",
      "Dino Group II Clade 4 X: No match, species\n",
      "Dino Group II Clade 4 X: No match, genus\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Oikopleura dioica: No match, species\n",
      "Anoecales: No match, order\n",
      "Dino Group II Clade 4: No match, family\n",
      "Diplonemidae: No match, family\n",
      "Dino Group II: No match, order\n",
      "Chlorellales X: No match, family\n",
      "Haliphthorales X: No match, species\n",
      "Haliphthorales X: No match, genus\n",
      "LOS7N I lineage: No match, family\n",
      "Thalassophysa pelagica: No match, speciesLychnaspis E2 giltschii: No match, species\n",
      "\n",
      "Haliphthorales: No match, family\n",
      "Echinodermata XXX: No match, species\n",
      "Echinodermata XXX: No match, genusDino Group II Clade 16 X: No match, species\n",
      "\n",
      "Dino Group II Clade 16 X: No match, genus\n",
      "Lychnaspis E2: No match, genus\n",
      "Peronosporomycetes X: No match, order\n",
      "Dino Group II Clade 16: No match, family\n",
      "Echinodermata XX: No match, family\n",
      "Acantharea E2: No match, family\n",
      "Peronosporomycetes: No match, class\n",
      "Rhodococcus rhodochrous: No match, species\n",
      "Echinodermata X: No match, order\n",
      "Dino Group II: No match, order\n",
      "Acantharea E: No match, order\n",
      "Gyrista: No match, subdivision\n",
      "Eustigmatophyceae XX: No match, family\n",
      "Rhodococcus: Multiple matches, genus \n",
      "Stramenopiles: No match, division\n",
      "Eustigmatophyceae X: No match, order\n",
      "Parmales env 2 X: No match, speciesAcantharea: Unaccepted, using Acantharia, class \n",
      "\n",
      "Parmales env 2 X: No match, genusRadiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "\n",
      "TSAR: No match, supergroup\n",
      "Gigartinales X: No match, family\n",
      "Parmales env 2: No match, family\n",
      "Dino Group II Clade 1 X: No match, species\n",
      "Peronosporomycetes XXX: No match, speciesDino Group II Clade 1 X: No match, genus\n",
      "\n",
      "Peronosporomycetes XXX: No match, genus\n",
      "Gregarinomorphea: No match, class\n",
      "Dino Group II Clade 1: No match, family\n",
      "Dino Group II: No match, orderPeronosporomycetes XX: No match, family\n",
      "\n",
      "Peronosporomycetes X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Microrhizoidea pickettheapsiorum: No match, species\n",
      "Picozoa X: No match, subdivision\n",
      "Microrhizoidea: No match, genus\n",
      "Dino Group II Clade 7 X: No match, species\n",
      "Dino Group II Clade 7 X: No match, genus\n",
      "Dino Group II Clade 7: No match, family\n",
      "Dino Group II Clade 33 X: No match, speciesDino Group II: No match, order\n",
      "\n",
      "Dino Group II Clade 33 X: No match, genus\n",
      "Rotifera XXX: No match, species\n",
      "Rotifera XXX: No match, genus\n",
      "Cryptista: Unaccepted, using Cryptophyta, supergroup \n",
      "Rotifera XX: No match, familyDino Group II Clade 33: No match, familyRAD B X Group IVd X: No match, species\n",
      "\n",
      "\n",
      "Eupelagonemidae X: No match, speciesDino Group II: No match, orderRAD B X Group IVd X: No match, genus\n",
      "\n",
      "\n",
      "Eupelagonemidae X: No match, genus\n",
      "MAST 8: No match, family\n",
      "Eupelagonemidae: No match, family\n",
      "Rotifera X: No match, orderRAD B X Group IVd: No match, family\n",
      "\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Ochromonadales clade VII X: No match, species\n",
      "Ochromonadales clade VII X: No match, genus\n",
      "Sagenista X: No match, order\n",
      "MAST 1B: No match, species\n",
      "MAST 1B: No match, genus\n",
      "Ochromonadales clade VII: No match, family\n",
      "MAST 1: No match, family\n",
      "Plagiopylea X: No match, order\n",
      "Gyrista XX: No match, order\n",
      "Chloroparvula B2: No match, species\n",
      "Gyrista X: No match, class\n",
      "Eupetalomonads XX: No match, species\n",
      "Eupetalomonads XX: No match, genus\n",
      "Metazoa: Unaccepted, using Animalia, subdivision \n",
      "Obazoa: No match, supergroup\n",
      "Gyrista: No match, subdivision\n",
      "Eupetalomonads X: No match, family\n",
      "Stramenopiles: No match, division\n",
      "Nemertea XX: No match, family\n",
      "Peronosporomycetes X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Dino Group I Clade 1 X: No match, speciesStramenopiles: No match, division\n",
      "\n",
      "Dino Group I Clade 1 X: No match, genusTSAR: No match, supergroup\n",
      "\n",
      "TSAR: No match, supergroup\n",
      "MAST 3: No match, family\n",
      "Nemertea X: No match, order\n",
      "Dino Group I Clade 1: No match, family\n",
      "RAD B X Group Ic X: No match, speciesTeratosphaeria: No match, genus\n",
      "\n",
      "RAD B X Group Ic X: No match, genus\n",
      "Nanomonadea: No match, order\n",
      "Bicoecea XXX: No match, species\n",
      "Bicoecea XXX: No match, genusDino Group I: No match, order\n",
      "\n",
      "RAD B X Group Ic: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Bicoecea XX: No match, family\n",
      "Dino Group II Clade 26 X: No match, species\n",
      "Dino Group II Clade 26 X: No match, genus\n",
      "Dino Group I Clade 4 X: No match, species\n",
      "Dino Group I Clade 4 X: No match, genusBicoecea X: No match, order\n",
      "\n",
      "Dino Group II Clade 26: No match, family\n",
      "Dino Group II: No match, order\n",
      "Ventrifissuridae: No match, familyDino Group I Clade 4: No match, family\n",
      "\n",
      "Dino Group I: No match, order\n",
      "Peromelissa tricuspidata: No match, species\n",
      "Picozoa XXXXX: No match, species\n",
      "Picozoa XXXXX: No match, genus\n",
      "Dino Group II Clade 5 X: No match, species\n",
      "Dino Group II Clade 5 X: No match, genus\n",
      "Picozoa XXXX: No match, family\n",
      "Dino Group II Clade 5: No match, family\n",
      "Dino Group II: No match, orderBlastocladiomycetes: No match, family\n",
      "\n",
      "Haptophyta XXXXX: No match, species\n",
      "Picozoa XXX: No match, order\n",
      "Haptophyta XXXXX: No match, genus\n",
      "Blastocladiomycotina: No match, order\n",
      "RAD B X Group IVb X: No match, species\n",
      "Ochromonadales clade VIII X: No match, species\n",
      "RAD B X Group IVb X: No match, genus\n",
      "Ochromonadales clade VIII X: No match, genus\n",
      "Picozoa XX: No match, class\n",
      "Haptophyta XXXX: No match, family\n",
      "Ochromonadales clade VIII: No match, familyPicozoa X: No match, subdivisionBlastocladiomycota: No match, class\n",
      "\n",
      "\n",
      "Haptophyta XXX: No match, order\n",
      "RAD B X Group IVb: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Haptophyta XX: No match, classSpumellaria 3 X: No match, species\n",
      "\n",
      "Spumellaria 3 X: No match, genus\n",
      "Gamasiphis pulchellus: No match, speciesAcantharea 4: No match, orderSpumellaria 3: No match, family\n",
      "\n",
      "\n",
      "Haptophyta X: No match, subdivision\n",
      "Gamasiphis: No match, genus\n",
      "Acantharea: Unaccepted, using Acantharia, class Acantharea 1 XX: No match, species\n",
      "\n",
      "Acantharea 1 XX: No match, genus\n",
      "Cryptosporidium1: No match, genusAcantharea 1 X: No match, family\n",
      "Acantharea C: No match, order\n",
      "\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Acantharea 1: No match, order\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "LC104 lineage X: No match, species\n",
      "Blastodinium mangini: No match, species\n",
      "LC104 lineage X: No match, genus\n",
      "LC104 lineage: No match, family\n",
      "Pyloniidae 3: No match, species\n",
      "Pyloniidae 3: No match, genus\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Chlorarachniophyceae X: No match, order\n",
      "MAST 10 X: No match, species\n",
      "MAST 10 X: No match, genus\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision MAST 10: No match, family\n",
      "\n",
      "Sagenista X: No match, order\n",
      "Chlorarachniophyceae: Unaccepted, using Chlorarachnea, class \n",
      "MAST 8A: No match, species\n",
      "MAST 8A: No match, genus\n",
      "Scoloplos johnstonei: Unaccepted, using Leodamas johnstonei, species \n",
      "Heterocapsa nei rotundata: No match, species\n",
      "MAST 8: No match, family\n",
      "Dolichomastigaceae A: No match, species\n",
      "Dolichomastigaceae A: No match, genus\n",
      "Dino Group II Clade 32 X: No match, species\n",
      "Dino Group II Clade 32 X: No match, genus\n",
      "Sagenista X: No match, order\n",
      "Dino Group II Clade 32: No match, family\n",
      "Dino Group II: No match, order\n",
      "Williopsis pratensis: No match, species\n",
      "Williopsis: No match, genus\n",
      "Telonemia Group 2 X: No match, species\n",
      "Telonemia Group 2 X: No match, genus\n",
      "Kefersteinia cirrata: Unaccepted, using Psamathe fusca, species \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Cnidaria X: No match, order\n",
      "Telonemia Group 2: No match, family\n",
      "Kefersteinia: Multiple unaccepted matches, genus \n",
      "Antetintinnidium mucicola: No match, species\n",
      "Telonemia XXX: No match, order\n",
      "Annelida XX: No match, family\n",
      "Typosyllis: Unaccepted, using Syllis, species \n",
      "Typosyllis: Unaccepted, using Syllis, genus \n",
      "Antetintinnidium: No match, genusTelonemia XX: No match, class\n",
      "\n",
      "Annelida X: No match, order\n",
      "Annelida XX: No match, family\n",
      "Phaeodarea: No match, classTelonemia X: No match, subdivision\n",
      "\n",
      "Mataza lineage X: No match, species\n",
      "Mataza lineage X: No match, genus\n",
      "Annelida X: No match, order\n",
      "Telonemia: No match, division\n",
      "Mataza lineage: No match, family\n",
      "TSAR: No match, supergroup\n",
      "Filosa Thecofilosea X: No match, order\n",
      "Collophidium serpentinum: No match, species\n",
      "Filosa Thecofilosea: No match, class\n",
      "Craniata X: No match, order\n",
      "Colpodea X: No match, order\n",
      "Collophidium: No match, genus\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Centroplasthelida XXXXX: No match, species\n",
      "Centroplasthelida XXXXX: No match, genus\n",
      "Collophidiidae: No match, family\n",
      "Suessiaceae X: No match, species\n",
      "Suessiaceae X: No match, genus\n",
      "Hemiselmis aquamarina: No match, species\n",
      "Dino Group II Clade 3 X: No match, species\n",
      "Dino Group II Clade 3 X: No match, genus\n",
      "Centroplasthelida XXXX: No match, family\n",
      "Dino Group II Clade 3: No match, family\n",
      "Dino Group II: No match, order\n",
      "Centroplasthelida XXX: No match, order\n",
      "Leptophryidae: No match, family\n",
      "Dino Group II Clade 27 X: No match, species\n",
      "Dino Group II Clade 27 X: No match, genus\n",
      "Centroplasthelida XX: No match, class\n",
      "Vampyrellida: No match, order\n",
      "RAD B X Group Ie X: No match, species\n",
      "RAD B X Group Ie X: No match, genus\n",
      "Micromonas bravo B2: No match, species\n",
      "Dino Group II Clade 27: No match, family\n",
      "Goussia6 janae: No match, speciesDino Group II: No match, order\n",
      "\n",
      "Centroplasthelida X: No match, subdivision\n",
      "RAD B X Group Ie: No match, family\n",
      "Cryptomonadales:nucl: No match, order\n",
      "Goussia6: No match, genus\n",
      "Centroplasthelida: No match, division\n",
      "RAD B X: No match, order\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Haptista: Unaccepted, no valid name, supergroup \n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Mariager Fjord lineage X: No match, speciesCryptophyta X:nucl: No match, subdivision\n",
      "\n",
      "Mariager Fjord lineage X: No match, genus\n",
      "Gigartacon C1 muelleri: No match, species\n",
      "Mariager Fjord lineage: No match, family\n",
      "Cryptophyta:nucl: No match, division\n",
      "Gigartacon C1: No match, genus\n",
      "Cryptista:nucl: No match, supergroup\n",
      "Variosea X: No match, order\n",
      "Thraustochytriaceae HK10: No match, species\n",
      "Thraustochytriaceae HK10: No match, genus\n",
      "Acantharea C1: No match, family\n",
      "Stephanoecidae Group D X: No match, species\n",
      "Stephanoecidae Group D X: No match, genus\n",
      "Prymnesiophyceae: Unaccepted, using Coccolithophyceae, class \n",
      "Acantharea C: No match, order\n",
      "Dipodomys ordii: No match, species\n",
      "Haptophyta X: No match, subdivision\n",
      "Stephanoecidae Group D: No match, family\n",
      "Dipodomys: No match, genus\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Liosphaeroidea XX: No match, species\n",
      "Liosphaeroidea XX: No match, genus\n",
      "MAST 7C: No match, species\n",
      "MAST 7C: No match, genus\n",
      "Liosphaeroidea X: No match, family\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Prymnesiophyceae Clade B3 X: No match, species\n",
      "Prymnesiophyceae Clade B3 X: No match, genus\n",
      "Prymnesiophyceae Clade B3: No match, family\n",
      "Nusuttodinium desymbiontum: No match, species\n",
      "Pelagomonadaceae clade D1: No match, species\n",
      "Pseudoplatyophrya nana: No match, speciesPelagomonadaceae clade D1: No match, genus\n",
      "\n",
      "Gyrista: No match, subdivision\n",
      "Pseudoplatyophrya: No match, genus\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group I Clade 5 X: No match, species\n",
      "Dino Group I Clade 5 X: No match, genus\n",
      "Coccidiomorphea: No match, class\n",
      "Philaenus spumarius: No match, species\n",
      "Lophophaenidae: No match, family\n",
      "Dino Group I Clade 5: No match, family\n",
      "Discocephalus pararotatorius: No match, species\n",
      "Dino Group I: No match, order\n",
      "Stephanoecidae Group E X: No match, species\n",
      "Stephanoecidae Group E X: No match, genus\n",
      "Philaenus: No match, genus\n",
      "Nebulomonas marisrubri: No match, species\n",
      "Prasino Clade 9 XXX: No match, species\n",
      "Prasino Clade 9 XXX: No match, genus\n",
      "Stephanoecidae Group E: No match, family\n",
      "Nebulomonas: No match, genus\n",
      "Haptophyta X: No match, subdivision\n",
      "Prasino Clade 9 XX: No match, familyChlorellales X: No match, family\n",
      "\n",
      "Nebulidae: No match, family\n",
      "Dino Group II Clade 57 X: No match, species\n",
      "Dino Group II Clade 57 X: No match, genus\n",
      "Thraustochytriaceae LAB17: No match, species\n",
      "Thraustochytriaceae LAB17: No match, genus\n",
      "Prasino Clade 9 X: No match, order\n",
      "Rogersella griseliniae: No match, species\n",
      "Nebulidida: No match, order\n",
      "Dino Group II Clade 57: No match, family\n",
      "Prasino Clade 9: No match, class\n",
      "Apionsoma misakianum: No match, species\n",
      "Rogersella: No match, genus\n",
      "Nebulidea: No match, class\n",
      "Dino Group II: No match, order\n",
      "Chlorophyta X: No match, subdivision\n",
      "Nebulidia X: No match, subdivision\n",
      "Diplonemida DSPDII X: No match, species\n",
      "Diplonemida DSPDII X: No match, genus\n",
      "Fabomonas tropica: No match, species\n",
      "Nebulidia: No match, division\n",
      "Provora: No match, supergroup\n",
      "Chytridiales XX: No match, species\n",
      "Chytridiales XX: No match, genus\n",
      "Dino Group V XX: No match, species\n",
      "Dino Group V XX: No match, genus\n",
      "Diplonemida DSPDII: No match, family\n",
      "Ancyromonadida Group 1 X: No match, genus\n",
      "Chytridiales X: No match, family\n",
      "Dino Group V X: No match, family\n",
      "Ancyromonadida Group 1: No match, family\n",
      "Nemastomatales X: No match, family\n",
      "Dino Group V: No match, order\n",
      "RAD A X: No match, order\n",
      "Ancyromonadida XXX: No match, order\n",
      "Dino Group III XX: No match, speciesRAD A: No match, class\n",
      "\n",
      "Dino Group III XX: No match, genusRadiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "\n",
      "Crustomastigaceae AB: No match, species\n",
      "Crustomastigaceae AB: No match, genus\n",
      "Ancyromonadida XX: No match, class\n",
      "Micromonas clade B4: No match, species\n",
      "Crustomastigaceae: No match, family\n",
      "Dino Group III X: No match, familyCorallinales X: No match, family\n",
      "\n",
      "Bryozoa XX: No match, family\n",
      "Ancyromonadida X: No match, subdivision\n",
      "Dino Group III: No match, order\n",
      "Bryozoa X: No match, order\n",
      "MAST 1D: No match, species\n",
      "MAST 1D: No match, genus\n",
      "Paraphysomonadales Clade EC1Ga X: No match, species\n",
      "Paraphysomonadales Clade EC1Ga X: No match, genus\n",
      "MAST 1: No match, family\n",
      "Paraphysomonadales Clade EC1Ga: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Bicosoeca vacillans: No match, species\n",
      "Olpidiopsis porphyrae: No match, species\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivisionTSAR: No match, supergroup\n",
      "\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Alveolata XXXXX: No match, species\n",
      "Alveolata XXXXX: No match, genus\n",
      "Drosophila ananassae: No match, species\n",
      "Conchoecia annelida: No match, species\n",
      "Conchoecia annelida: No match, genus\n",
      "Spumellaria 6 X: No match, species\n",
      "Spumellaria 6 X: No match, genus\n",
      "Alveolata XXXX: No match, family\n",
      "Annelida XX: No match, family\n",
      "Spumellaria 6: No match, familyAlveolata XXX: No match, order\n",
      "\n",
      "Monosigidae Group B X: No match, speciesAnnelida X: No match, order\n",
      "\n",
      "Monosigidae Group B X: No match, genus\n",
      "Alveolata XX: No match, class\n",
      "Alveolata X: No match, subdivision\n",
      "Breviata lineage X: No match, species\n",
      "Breviata lineage X: No match, genus\n",
      "Perkinsus qugwadi: Unaccepted, using Perkinsus qugwadii, species \n",
      "MAST 4C: No match, species\n",
      "MAST 4C: No match, genus\n",
      "Breviata lineage: No match, family\n",
      "Ctenophora XX: No match, family\n",
      "Monosigidae Group B: No match, family\n",
      "MAST 4: No match, family\n",
      "Breviatea XXX: No match, order\n",
      "Ctenophora X: No match, order\n",
      "Strobilidiidae X: No match, species\n",
      "Strobilidiidae X: No match, genus\n",
      "Sagenista X: No match, order\n",
      "Breviatea XX: No match, class\n",
      "Ctenophora: Multiple matches, class \n",
      "Breviatea X: No match, subdivision\n",
      "Diplonema: Multiple matches, species \n",
      "Diplonema: Multiple matches, genus \n",
      "Diplonemidae: No match, family\n",
      "MOCH 5 XXX: No match, species\n",
      "MOCH 5 XXX: No match, genus\n",
      "Breviatea: Multiple matches, division \n",
      "Metazoa: Unaccepted, using Animalia, subdivision \n",
      "Obazoa: No match, supergroupMOCH 5 XX: No match, family\n",
      "\n",
      "Chaetoceros debilis 2: No match, species\n",
      "Labyrinthulaceae ANT10 3: No match, speciesPrymnesiophyceae Clade E XX: No match, species\n",
      "\n",
      "Labyrinthulaceae ANT10 3: No match, genusPrymnesiophyceae Clade E XX: No match, genus\n",
      "\n",
      "MOCH 5 X: No match, order\n",
      "Obazoa: No match, supergroup\n",
      "Vermamoebidae: No match, family\n",
      "Prymnesiophyceae Clade E X: No match, family\n",
      "MOCH 5: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Hyphodontia rimosissima: No match, species\n",
      "Prymnesiophyceae Clade E: No match, order\n",
      "Sagenista X: No match, order\n",
      "Spongopylidae: No match, family\n",
      "Prymnesiophyceae: Unaccepted, using Coccolithophyceae, class \n",
      "Haptophyta X: No match, subdivision\n",
      "Florenciellales X: No match, species\n",
      "Florenciellales X: No match, genus\n",
      "Hexaconus serratus: No match, species\n",
      "Aureobasidium pullulans: Multiple matches, species \n",
      "Hexaconus: No match, genus\n",
      "Acantharea E4: No match, family\n",
      "Acantharea E: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Demodex: No match, genus\n",
      "MAST 7E: No match, species\n",
      "MAST 7E: No match, genus\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "Acantharea 4a X: No match, species\n",
      "Acantharea 4a X: No match, genus\n",
      "Melkoniana: No match, species\n",
      "Melkoniana: No match, genus\n",
      "Acantharea 4a: No match, family\n",
      "Cephalopoda X: No match, family\n",
      "Ventricleftida XX: No match, species\n",
      "Ventricleftida XX: No match, genus\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "Acantharea 4: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Ventricleftida X: No match, family\n",
      "Dino Group I Clade 7 X: No match, species\n",
      "Dino Group I Clade 7 X: No match, genus\n",
      "Flectonema neradi: No match, species\n",
      "Dino Group I Clade 7: No match, family\n",
      "Dino Group I: No match, order\n",
      "Acantharea F3: No match, family\n",
      "MAST 8C: No match, species\n",
      "MAST 8C: No match, genus\n",
      "MAST 8: No match, family\n",
      "Sagenista X: No match, order\n",
      "Flectonema: No match, genus\n",
      "MAST 3B: No match, species\n",
      "MAST 3B: No match, genus\n",
      "Acantharea F: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Diplonemidae: No match, family\n",
      "Thalassomyces fagei: Unaccepted, using Amallocystis fagei, species \n",
      "MAST 3: No match, family\n",
      "Nanomonadea: No match, order\n",
      "Chrysocampanula spinifera: Unaccepted, using Chrysochromulina spinifera, species \n",
      "Dino Group II Clade 35 X: No match, species\n",
      "Dino Group II Clade 35 X: No match, genus\n",
      "RAD B X Group IV XX: No match, species\n",
      "RAD B X Group IV XX: No match, genus\n",
      "Dino Group II Clade 35: No match, family\n",
      "Dino Group II: No match, order\n",
      "Adelina grylli: No match, species\n",
      "RAD B X Group IV X: No match, family\n",
      "Chrysocampanula: Unaccepted, using Chrysochromulina, genus RAD B X: No match, order\n",
      "\n",
      "RAD B: No match, class\n",
      "Micromonas clade B5: No match, speciesRadiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "\n",
      "Prymnesiales X: No match, familyMAST 7D: No match, species\n",
      "\n",
      "MAST 7D: No match, genus\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "Chlorellales X: No match, family\n",
      "Cannobotryidae X: No match, species\n",
      "Cannobotryidae X: No match, genus\n",
      "Dino Group II Clade 39 X: No match, species\n",
      "Dino Group II Clade 39 X: No match, genus\n",
      "Phyllostaurus B2 echinoides: No match, species\n",
      "Knufia cryptophialidica: No match, species\n",
      "RAD A Xb X: No match, species\n",
      "RAD A Xb X: No match, genus\n",
      "Papiliotrema laurentii: No match, species\n",
      "Dino Group II Clade 39: No match, family\n",
      "Dino Group II: No match, order\n",
      "Microhelida XX: No match, species\n",
      "Microhelida XX: No match, genus\n",
      "Phyllostaurus B2: No match, genus\n",
      "RAD A Xb: No match, family\n",
      "Microhelida X: No match, familyVentrifissuridae X: No match, species\n",
      "\n",
      "Ventrifissuridae X: No match, genus\n",
      "RAD A X: No match, order\n",
      "Acantharea B2: No match, family\n",
      "Microhelida: No match, order\n",
      "RAD A: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Acantharea B: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Pterocystida: No match, class\n",
      "MAST 1A: No match, species\n",
      "MAST 1A: No match, genus\n",
      "MAST 1: No match, family\n",
      "MAST 4B: No match, speciesGyrista XX: No match, order\n",
      "\n",
      "MAST 4B: No match, genusGyrista X: No match, class\n",
      "\n",
      "MAST 4: No match, familyGyrista: No match, subdivision\n",
      "\n",
      "Sagenista X: No match, orderStramenopiles: No match, division\n",
      "\n",
      "Stramenopiles: No match, divisionTSAR: No match, supergroup\n",
      "\n",
      "TSAR: No match, supergroup\n",
      "Centroplasthelida X: No match, subdivision\n",
      "Minorisa minuta: No match, species\n",
      "Amphibelone heteracanthum: No match, species\n",
      "Ventrifissuridae: No match, family\n",
      "Centroplasthelida: No match, division\n",
      "Minorisa: No match, genus\n",
      "MAST 22D: No match, species\n",
      "MAST 22D: No match, genus\n",
      "Pseudo nitzschia: No match, species\n",
      "Pseudo nitzschia: No match, genus\n",
      "Amphibelone: No match, genus\n",
      "Haptista: Unaccepted, no valid name, supergroup \n",
      "Minorisa lineage: No match, family\n",
      "MAST 22: No match, family\n",
      "Acantharea F2: No match, family\n",
      "Chlorarachnida: No match, order\n",
      "Bicosoecida: No match, orderAcantharea F: No match, orderExobasidiomycetes X: No match, species\n",
      "\n",
      "\n",
      "Monosigidae Group A X: No match, speciesExobasidiomycetes X: No match, genusAcantharea: Unaccepted, using Acantharia, class \n",
      "\n",
      "\n",
      "Monosigidae Group A X: No match, genusRadiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "\n",
      "Gorilla gorilla: No match, species\n",
      "Dino Group II Clade 6 X: No match, speciesMonosigidae Group A: No match, family\n",
      "\n",
      "Exobasidiomycetes: No match, familyDino Group II Clade 6 X: No match, genus\n",
      "\n",
      "Gorilla: No match, genus\n",
      "Chlorarachniophyceae: Unaccepted, using Chlorarachnea, class \n",
      "Dino Group II Clade 6: No match, family\n",
      "Dino Group II: No match, order\n",
      "Gastrotricha XX: No match, family\n",
      "Craniata XX: No match, family\n",
      "Clausidium vancouverense: No match, species\n",
      "Gastrotricha X: No match, order\n",
      "Craniata X: No match, order\n",
      "Pseudo nitzschia: No match, genus\n",
      "MAST 7B: No match, species\n",
      "MAST 7B: No match, genus\n",
      "Labyrinthulomycetes LAB15 X: No match, speciesMAST 12: No match, family\n",
      "\n",
      "Labyrinthulomycetes LAB15 X: No match, genus\n",
      "Acantharea F: No match, order\n",
      "MAST 7: No match, family\n",
      "Sagenista X: No match, order\n",
      "Cryptomonas:nucl: No match, genus\n",
      "Opalozoa X: No match, order\n",
      "Labyrinthulomycetes LAB15: No match, family\n",
      "Kathablepharidida XX: No match, species\n",
      "Kathablepharidida XX: No match, genus\n",
      "Cryptomonadales X:nucl: No match, family\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Kathablepharidida X: No match, familyStramenopiles X Group 7 XXX: No match, species\n",
      "\n",
      "Stramenopiles X Group 7 XXX: No match, genus\n",
      "Cryptomonadales:nucl: No match, order\n",
      "Stramenopiles X Group 7 XX: No match, family\n",
      "Kathablepharidida: No match, orderLabyrinthulomycetes: Unaccepted, using Labyrinthulea, order \n",
      "\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Stramenopiles X Group 7 X: No match, order\n",
      "RAD A Xe X: No match, species\n",
      "RAD A Xe X: No match, genusKathablepharidea: No match, class\n",
      "\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "RAD A Xe: No match, family\n",
      "RAD A X: No match, orderStramenopiles X Group 7: No match, class\n",
      "\n",
      "RAD A: No match, classKathablepharida: No match, subdivision\n",
      "\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Chaetoceros brevis 2: No match, species\n",
      "Cryptophyta:nucl: No match, division\n",
      "Stramenopiles X: No match, subdivision\n",
      "Kathablepharidacea: No match, divisionStramenopiles: No match, division\n",
      "\n",
      "TSAR: No match, supergroup\n",
      "Cryptista:nucl: No match, supergroup\n",
      "MAST 12B: No match, species\n",
      "MAST 12B: No match, genus\n",
      "MAST 12: No match, family\n",
      "Opalozoa X: No match, order\n",
      "Cryptista: Unaccepted, using Cryptophyta, supergroup Sphaeropleales X: No match, family\n",
      "\n",
      "Ostreococcus clade B: No match, species\n",
      "Acantharea 4d X: No match, species\n",
      "Acantharea 4d X: No match, genus\n",
      "Dactylodinium arachnoides: No match, species\n",
      "Salpa: Multiple matches, genus Acantharea 4d: No match, family\n",
      "\n",
      "Acantharea 4: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Dino Group I Clade 2 X: No match, species\n",
      "Dino Group I Clade 2 X: No match, genus\n",
      "Ceraceomyces americanus: No match, species\n",
      "Dino Group I Clade 2: No match, family\n",
      "Dino Group I: No match, order\n",
      "Chaetoceros pumilum: No match, species\n",
      "Ceraceomyces: No match, genus\n",
      "Hemichordata XX: No match, family\n",
      "Hemichordata X: No match, order\n",
      "Dino Group II Clade 22 X: No match, species\n",
      "Dino Group II Clade 22 X: No match, genus\n",
      "Galeomma takii: Unaccepted, using Lepirodes layardi, species \n",
      "Malassezia sympodialis: No match, species\n",
      "Dino Group II Clade 22: No match, familyHaliphthorales: No match, family\n",
      "\n",
      "Dino Group II: No match, order\n",
      "Galeomma: Multiple matches, genus \n",
      "Malassezia: No match, genus\n",
      "Rhinomonas nottbecki: No match, species\n",
      "Peronosporomycetes X: No match, order\n",
      "Exobasidiomycetes: No match, family\n",
      "Peronosporomycetes: No match, class\n",
      "Micromonas commoda A1: No match, species\n",
      "Pedinellales X: No match, species\n",
      "Pedinellales X: No match, genus\n",
      "Rhodomonas:nucl: No match, species\n",
      "Rhodomonas:nucl: No match, genus\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, divisionPyrenomonadaceae:nucl: No match, family\n",
      "\n",
      "TSAR: No match, supergroup\n",
      "Pyrenomonadales:nucl: No match, order\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "Cryptophyta:nucl: No match, division\n",
      "Cryptista:nucl: No match, supergroup\n",
      "Voromonas pontica: No match, species\n",
      "Acanthoecida XX: No match, species\n",
      "Acanthoecida XX: No match, genus\n",
      "Prymnesiophyceae Clade D XX: No match, species\n",
      "Prymnesiophyceae Clade D XX: No match, genus\n",
      "Voromonas: No match, genus\n",
      "Dictyochophyceae X: No match, order\n",
      "Acanthoecida X: No match, familyParatetrahymena parawassi: No match, species\n",
      "\n",
      "Prymnesiophyceae Clade D X: No match, family\n",
      "Colpodellaceae: No match, family\n",
      "Prymnesiophyceae Clade D: No match, order\n",
      "Prymnesiophyceae: Unaccepted, using Coccolithophyceae, class \n",
      "Haptophyta X: No match, subdivision\n",
      "Brassica rapa: No match, species\n",
      "Ctenophora XXX: No match, species\n",
      "Ctenophora XXX: No match, genus\n",
      "Ctenophora XX: No match, family\n",
      "Ctenophora X: No match, order\n",
      "Paradinidae: No match, family\n",
      "Ctenophora: Multiple matches, class \n",
      "Metazoa: Unaccepted, using Animalia, subdivision \n",
      "Obazoa: No match, supergroup\n",
      "Sargassum wightii: Unaccepted, using Sargassum swartzii, species \n",
      "Perkinsida XXX: No match, species\n",
      "Perkinsida XXX: No match, genus\n",
      "Dino Group II Clade 2 X: No match, species\n",
      "Dino Group II Clade 2 X: No match, genus\n",
      "Perkinsida XX: No match, family\n",
      "Chaetoceros muellerii: No match, species\n",
      "Dino Group II Clade 2: No match, family\n",
      "Dino Group II: No match, order\n",
      "Bacillariophyceae XXX: No match, species\n",
      "Bacillariophyceae XXX: No match, genus\n",
      "Perkinsida X: No match, order\n",
      "Rhizopus oryzae: No match, species\n",
      "Haptophyta Clade HAP4 XXX: No match, species\n",
      "Haptophyta Clade HAP4 XXX: No match, genus\n",
      "Bacillariophyceae XX: No match, family\n",
      "Haptophyta Clade HAP4 XX: No match, family\n",
      "Bacillariophyceae X: No match, order\n",
      "Conchellium capsula: No match, species\n",
      "Mediocremonas mediterraneus: No match, species\n",
      "Filosa Imbricatea XXX: No match, species\n",
      "Filosa Imbricatea XXX: No match, genus\n",
      "Haptophyta Clade HAP4 X: No match, order\n",
      "Conchellium: No match, genusMediocremonas: No match, genus\n",
      "\n",
      "Telonemia Group 2: No match, family\n",
      "Filosa Imbricatea XX: No match, family\n",
      "Haptophyta Clade HAP4: No match, class\n",
      "Telonemia XXX: No match, order\n",
      "Developea XX: No match, family\n",
      "Conchariidae: No match, family\n",
      "Filosa Imbricatea X: No match, order\n",
      "Haptophyta X: No match, subdivision\n",
      "Telonemia XX: No match, class\n",
      "Developea X: No match, order\n",
      "Filosa Imbricatea: No match, class\n",
      "Telonemia X: No match, subdivision\n",
      "Developea: No match, class\n",
      "Phascolion strombus: Multiple unaccepted matches, species \n",
      "Malassezia globosa: No match, species\n",
      "Telonemia: No match, divisionGyrista: No match, subdivision\n",
      "\n",
      "TSAR: No match, supergroup\n",
      "Malassezia: No match, genus\n",
      "Stramenopiles: No match, division\n",
      "Thoracosphaeraceae X: No match, species\n",
      "Thoracosphaeraceae X: No match, genus\n",
      "Exobasidiomycetes: No match, family\n",
      "Dino Group II Clade 17 X: No match, species\n",
      "Dino Group II Clade 17 X: No match, genus\n",
      "TSAR: No match, supergroup\n",
      "CONThreeP XXX: No match, species\n",
      "CONThreeP XXX: No match, genus\n",
      "Dino Group II Clade 17: No match, family\n",
      "Dino Group II: No match, order\n",
      "Pterocystida: No match, class\n",
      "Symbiontida XXX: No match, species\n",
      "Symbiontida XXX: No match, genus\n",
      "CONThreeP XX: No match, family\n",
      "Appendicularia X: No match, species\n",
      "Appendicularia X: No match, genus\n",
      "Centroplasthelida X: No match, subdivision\n",
      "RAD C Xa X: No match, species\n",
      "RAD C Xa X: No match, genus\n",
      "Mediophyceae XXX: No match, speciesSymbiontida XX: No match, family\n",
      "\n",
      "Mediophyceae XXX: No match, genus\n",
      "CONThreeP X: No match, order\n",
      "Appendicularia: Multiple matches, family \n",
      "Centroplasthelida: No match, division\n",
      "RAD C Xa: No match, family\n",
      "Mediophyceae XX: No match, family\n",
      "Symbiontida X: No match, order\n",
      "CONThreeP: No match, class\n",
      "Haptista: Unaccepted, no valid name, supergroup Urochordata X: No match, order\n",
      "\n",
      "RAD C X: No match, order\n",
      "Mediophyceae X: No match, order\n",
      "Symbiontida: No match, class\n",
      "RAD C: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Euglenozoa: Multiple matches, subdivision \n",
      "Mediophyceae: Unaccepted, no valid name, class \n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Ventrifissura: No match, species\n",
      "Ventrifissura: No match, genus\n",
      "Dino Group II Clade 13 X: No match, species\n",
      "Dino Group II Clade 13 X: No match, genus\n",
      "Urochordata: Unaccepted, using Tunicata, class \n",
      "Metazoa: Unaccepted, using Animalia, subdivision \n",
      "Obazoa: No match, supergroup\n",
      "Echinamoeba silvestris: No match, species\n",
      "Ventrifissuridae: No match, family\n",
      "Dino Group II Clade 13: No match, family\n",
      "Dino Group II: No match, order\n",
      "Metasychis disparidentata: No match, species\n",
      "Nassellaria 3 X: No match, species\n",
      "Nassellaria 3 X: No match, genus\n",
      "RAD A Xc X: No match, species\n",
      "RAD A Xc X: No match, genus\n",
      "Fibrophrys: No match, species\n",
      "Fibrophrys: No match, genus\n",
      "Nassellaria 3: No match, family\n",
      "RAD A Xc: No match, family\n",
      "RAD A X: No match, order\n",
      "RAD A: No match, classChoreotrichida Tintinnina X: No match, family\n",
      "\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision Choreotrichida Tintinnina: No match, order\n",
      "\n",
      "Kodamaea ohmeri: No match, species\n",
      "Euduboscquella crenulata: No match, species\n",
      "Amphifilaceae: No match, family\n",
      "Turbellaria: Unaccepted, no valid name, order \n",
      "Kodamaea: No match, genus\n",
      "Chytridiomycota XXX: No match, species\n",
      "Chytridiomycota XXX: No match, genus\n",
      "Actinocyclus: Multiple matches, species \n",
      "Actinocyclus: Multiple matches, genus \n",
      "Labyrinthulomycetes: Unaccepted, using Labyrinthulea, order \n",
      "Chytridiomycota XX: No match, family\n",
      "Opisthokonta XXXXX: No match, species\n",
      "Chytridiomycota X: No match, orderOpisthokonta XXXXX: No match, genus\n",
      "\n",
      "MAST 11 X: No match, species\n",
      "MAST 11 X: No match, genus\n",
      "Opisthokonta XXXX: No match, family\n",
      "MAST 11: No match, family\n",
      "Sagenista X: No match, order\n",
      "Diplonema: Multiple matches, genus \n",
      "Diplonemidae: No match, family\n",
      "Microciona prolifera: Unaccepted, using Clathria (Clathria) prolifera, species \n",
      "Opisthokonta XXX: No match, order\n",
      "Rhodotorula lamellibrachiae: No match, species\n",
      "Dino Group II Clade 10 and 11 X: No match, species\n",
      "Dino Group II Clade 10 and 11 X: No match, genus\n",
      "MAST 2: No match, family\n",
      "Ebriacea: No match, order\n",
      "Opisthokonta XX: No match, class\n",
      "Dino Group II Clade 10 and 11: No match, family\n",
      "Dino Group II: No match, order\n",
      "Gyrista XX: No match, order\n",
      "Filosa Thecofilosea: No match, class\n",
      "Microciona: Unaccepted, using Clathria (Microciona), genus Opisthokonta X: No match, subdivision\n",
      "\n",
      "Obazoa: No match, supergroup\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Haptista: Unaccepted, no valid name, supergroup \n",
      "Dipodascus ingens: No match, species\n",
      "Echinodermata XX: No match, family\n",
      "Chondria: Multiple matches, species \n",
      "Chondria: Multiple matches, genus \n",
      "Echinodermata X: No match, order\n",
      "Dino Group II: No match, order\n",
      "CCW10 lineage X: No match, species\n",
      "Thalassema thalassemum: Unaccepted, using Thalassema thalassema, species \n",
      "CCW10 lineage X: No match, genus\n",
      "RAD A Xa X: No match, species\n",
      "RAD A Xa X: No match, genus\n",
      "RAD B X Group IVa X: No match, species\n",
      "RAD B X Group IVa X: No match, genus\n",
      "RAD A Xa: No match, familyCCW10 lineage: No match, family\n",
      "\n",
      "RAD A X: No match, order\n",
      "RAD A: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "RAD B X Group IVa: No match, family\n",
      "Rhizosphaera: No match, species\n",
      "Rhizosphaera: No match, genus\n",
      "RAD B X: No match, order\n",
      "Rhizosphaeridae: No match, family\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Dino Group II Clade 20 X: No match, species\n",
      "Dino Group II Clade 20 X: No match, genus\n",
      "Trentepohliales X: No match, family\n",
      "MAST 2C: No match, speciesDino Group II Clade 38 X: No match, species\n",
      "\n",
      "MAST 2C: No match, genusDino Group II Clade 38 X: No match, genus\n",
      "\n",
      "MAST 2: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group II Clade 20: No match, family\n",
      "Dino Group II: No match, order\n",
      "Annelida XX: No match, family\n",
      "Annelida X: No match, order\n",
      "Margolisiellidae X: No match, species\n",
      "Margolisiellidae X: No match, genus\n",
      "Dino Group II Clade 38: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group II XX: No match, species\n",
      "Dino Group II XX: No match, genus\n",
      "Margolisiellidae: No match, family\n",
      "Dino Group II X: No match, family\n",
      "Dino Group II: No match, order\n",
      "Saccamminidae X: No match, species\n",
      "Saccamminidae X: No match, genus\n",
      "Paralecudina polymorpha: No match, species\n",
      "Marosporida: No match, order\n",
      "Coccidiomorphea: No match, class\n",
      "Syndiniales XXX: No match, species\n",
      "Syndiniales XXX: No match, genus\n",
      "Paralecudina: No match, genus\n",
      "Streptophyta X: No match, subdivision\n",
      "Syndiniales XX: No match, family\n",
      "Paralecudinids: No match, family\n",
      "Cryomonadida XX: No match, species\n",
      "Cryomonadida XX: No match, genusSyndiniales X: No match, order\n",
      "\n",
      "Thraustochytriaceae QPX: No match, species\n",
      "Thraustochytriaceae QPX: No match, genus\n",
      "Spizellomycetales XX: No match, speciesCryomonadida X: No match, family\n",
      "\n",
      "Spizellomycetales XX: No match, genus\n",
      "Dino Group II Clade 4: No match, family\n",
      "Dino Group II: No match, order\n",
      "Spizellomycetales X: No match, family\n",
      "Vannella contorta: No match, species\n",
      "Monosigidae Group A: No match, family\n",
      "Chaetoceros curvisetus 3: No match, species\n",
      "Dino Group II Clade 15 X: No match, species\n",
      "Dino Group II Clade 15 X: No match, genus\n",
      "Vampyrellida: No match, species\n",
      "Dino Group II Clade 15: No match, family\n",
      "Dino Group II: No match, order\n",
      "MAST 4E: No match, species\n",
      "MAST 4E: No match, genus\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "Vampyrellida clade B: No match, genus\n",
      "MAST 3E: No match, species\n",
      "MAST 3E: No match, genus\n",
      "MAST 3: No match, family\n",
      "Nanomonadea: No match, order\n",
      "Vampyrellida X: No match, family\n",
      "Vampyrellida: No match, order\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "MAST 3K: No match, species\n",
      "MAST 3K: No match, genus\n",
      "MAST 3: No match, family\n",
      "Nanomonadea: No match, order\n",
      "Dino Group II Clade 9 X: No match, species\n",
      "RAD B X Group II X: No match, speciesDino Group II Clade 9 X: No match, genus\n",
      "\n",
      "RAD B X Group II X: No match, genus\n",
      "Tintinnidae X: No match, species\n",
      "Tintinnidae X: No match, genus\n",
      "Stagonosporopsis cucurbitacearum: No match, species\n",
      "RAD B X Group II: No match, family\n",
      "Dino Group II Clade 9: No match, family\n",
      "RAD B X: No match, order\n",
      "Dino Group II: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Malassezia: No match, genus\n",
      "Dino Group I Clade 4: No match, family\n",
      "Exobasidiomycetes: No match, family\n",
      "Rhynchopus: Multiple matches, species \n",
      "Rhynchopus: Multiple matches, genus \n",
      "Dino Group I: No match, order\n",
      "Bullacta exarata: Unaccepted, using Bullacta caurina, species \n",
      "Diplonemidae: No match, family\n",
      "RAD B X Group IVc X: No match, species\n",
      "RAD B X Group IVc X: No match, genus\n",
      "Phalachroma: No match, genus\n",
      "Chlorarachnida: No match, order\n",
      "RAD B X Group IVc: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Cryptomonadales XX: No match, species\n",
      "Cryptomonadales XX: No match, genus\n",
      "Oxyphysiaceae: No match, family\n",
      "Paramoeba branchiphila: No match, species\n",
      "Cryptomonadales X: No match, family\n",
      "Chlorarachniophyceae: Unaccepted, using Chlorarachnea, class \n",
      "Chlorodendrales X: No match, species\n",
      "Chlorodendrales XX: No match, genus\n",
      "Cercozoa XXXX: No match, species\n",
      "Cercozoa XXXX: No match, genus\n",
      "Excentroconchidae X: No match, species\n",
      "Excentroconchidae X: No match, genus\n",
      "Cercozoa XXX: No match, family\n",
      "Scytomonadidae: No match, family\n",
      "Chaetocerotales: Unaccepted, using Chaetocerotanae, order \n",
      "Excentroconchidae: No match, family\n",
      "Cercozoa XX: No match, order\n",
      "Heliocephala: No match, genus\n",
      "Mediophyceae: Unaccepted, no valid name, class \n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group I Clade 3 X: No match, speciesCercozoa X: No match, class\n",
      "\n",
      "Dino Group I Clade 3 X: No match, genus\n",
      "Pezizomycotina X: No match, familyChrysophyceae XXX: No match, species\n",
      "Vexillifera tasmaniana: No match, species\n",
      "\n",
      "Chrysophyceae XXX: No match, genus\n",
      "Dino Group I Clade 3: No match, family\n",
      "Dino Group I: No match, order\n",
      "Chrysophyceae XX: No match, family\n",
      "MAST 1C: No match, species\n",
      "MAST 1C: No match, genus\n",
      "MAST 1: No match, familyStephanoecidae Group I: No match, family\n",
      "\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Chrysophyceae X: No match, order\n",
      "Dino Group II Clade 29 X: No match, species\n",
      "Dino Group II Clade 29 X: No match, genus\n",
      "MAST 4F: No match, species\n",
      "MAST 4F: No match, genus\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "Neoparamoeba branchiphila: No match, species\n",
      "Rhodophyta X: No match, subdivision\n",
      "Dino Group II Clade 29: No match, family\n",
      "Dino Group II: No match, order\n",
      "Euglenozoa: Multiple matches, subdivision \n",
      "Chaetoceros sp Clade Na26B1: No match, species\n",
      "Euglyphida XX: No match, species\n",
      "Euglyphida XX: No match, genus\n",
      "Euglyphida X: No match, familyMAST 3J: No match, species\n",
      "\n",
      "Parvilucifera prorocentri: No match, species\n",
      "MAST 3J: No match, genus\n",
      "MAST 3: No match, family\n",
      "Nanomonadea: No match, order\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "MAST 2D: No match, species\n",
      "MAST 2D: No match, genus\n",
      "Dino Group II Clade 12 X: No match, species\n",
      "Dino Group II Clade 12 X: No match, genus\n",
      "MAST 2: No match, family\n",
      "Neosiphonia: Multiple matches, species \n",
      "Neosiphonia: Multiple matches, genus \n",
      "Gyrista XX: No match, order\n",
      "Thaumatomastigidae: Unaccepted, using Thaumatomonadidae, species \n",
      "Thaumatomastigidae: Unaccepted, using Thaumatomonadidae, genus Dino Group II Clade 12: No match, family\n",
      "\n",
      "Dino Group II: No match, order\n",
      "Paracoccidioides: No match, genus\n",
      "Chrysophyceae Clade EC2H XX: No match, species\n",
      "Chrysophyceae Clade EC2H XX: No match, genus\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Chlorarachnida XX: No match, species\n",
      "Chlorarachnida XX: No match, genus\n",
      "Chrysophyceae Clade EC2H X: No match, family\n",
      "Paramoeba invadens: No match, species\n",
      "Chlorarachnida X: No match, family\n",
      "Frankliniella occidentalis: No match, species\n",
      "Chrysophyceae Clade EC2H: No match, order\n",
      "Chlorarachnida: No match, order\n",
      "Acantharea E XX: No match, species\n",
      "Acantharea E XX: No match, genus\n",
      "Acantharea E X: No match, family\n",
      "Tricolocamptra aff urnula: No match, species\n",
      "Dinophyceae XXX: No match, species\n",
      "Dinophyceae XXX: No match, genus\n",
      "Chlorarachniophyceae: Unaccepted, using Chlorarachnea, class \n",
      "Acantharea E: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Tricolocamptra: No match, genus\n",
      "Dinophyceae XX: No match, family\n",
      "Verrucomonas: No match, genus\n",
      "Theopiliidae: No match, family\n",
      "Dinophyceae X: No match, order\n",
      "Verrucomonadidae: No match, family\n",
      "Basal Cryptophyceae 2 X: No match, species\n",
      "Basal Cryptophyceae 2 X: No match, genus\n",
      "MAST 7A: No match, species\n",
      "MAST 7A: No match, genus\n",
      "Dictyocoryne truncata: No match, species\n",
      "MAST 7: No match, familyBasal Cryptophyceae 2: No match, family\n",
      "\n",
      "Sagenista X: No match, order\n",
      "Cryptophyceae X: No match, order\n",
      "Termitomyces heimii: No match, species\n",
      "Chaetoceros protuberans: Multiple unaccepted matches, species \n",
      "Gregarinomorphea X GRE1 XX: No match, species\n",
      "Gregarinomorphea X GRE1 XX: No match, genus\n",
      "Termitomyces: No match, genus\n",
      "Gregarinomorphea X GRE1 X: No match, family\n",
      "RAD B X Group If X: No match, species\n",
      "RAD B X Group If X: No match, genus\n",
      "Gregarinomorphea X GRE1: No match, order\n",
      "Stephanoecidae Group I X: No match, species\n",
      "Stephanoecidae Group I X: No match, genus\n",
      "RAD B X Group If: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Gregarinomorphea: No match, class\n",
      "Stephanoecidae Group I: No match, family\n",
      "RAD B X Group III X: No match, species\n",
      "RAD B X Group III X: No match, genus\n",
      "Perichlamydium venustum: No match, species\n",
      "RAD B X Group III: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Perichlamydium: No match, genus\n",
      "Stylodictyidae: No match, family\n",
      "Pterocorys clausa: No match, species\n",
      "Pelagodinium beii: No match, species\n",
      "Parmales env 3 X: No match, species\n",
      "Parmales env 3 X: No match, genus\n",
      "Parmales env 3: No match, family\n",
      "Chlorellales X: No match, family\n",
      "Pinus taeda: No match, species\n",
      "Haptophyta Clade HAP2 XXX: No match, species\n",
      "Haptophyta Clade HAP2 XXX: No match, genus\n",
      "Thaumatomonadida XX: No match, species\n",
      "Thaumatomonadida XX: No match, genus\n",
      "Haptophyta Clade HAP2 XX: No match, family\n",
      "Thaumatomonadida X: No match, family\n",
      "Paradinidae X: No match, species\n",
      "Paradinidae X: No match, genus\n",
      "Haptophyta Clade HAP2 X: No match, order\n",
      "Paradinidae: No match, family\n",
      "Haptophyta Clade HAP2: No match, class\n",
      "Haptophyta X: No match, subdivision\n",
      "Protaspa lineage X: No match, species\n",
      "Protaspa lineage X: No match, genus\n",
      "Chromadorea X: No match, family\n",
      "Mytiloida: Unaccepted, using Mytiloides, family \n",
      "Haliommatidium: No match, species\n",
      "Haliommatidium: No match, genus\n",
      "Paramoeba atlantica: No match, species\n",
      "Protaspa lineage: No match, family\n",
      "Acantharea F3: No match, family\n",
      "Acantharea F: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Leidyana2: No match, species\n",
      "Leidyana2: No match, genus\n",
      "NPK2 lineage X: No match, species\n",
      "NPK2 lineage X: No match, genus\n",
      "Litholophus spB: No match, species\n",
      "NPK2 lineage: No match, family\n",
      "Chlorarachniophyceae X: No match, order\n",
      "Chlorarachniophyceae: Unaccepted, using Chlorarachnea, class \n",
      "Litholophus: No match, genus\n",
      "Stigmostylus ferrusi: No match, species\n",
      "Rhodotorula toruloides: No match, species\n",
      "Apusomonadidae Group 1 XX: No match, species\n",
      "Apusomonadidae Group 1 XX: No match, genus\n",
      "Acantharea C3: No match, family\n",
      "Stigmostylus: No match, genus\n",
      "Stylodictya subtilis: No match, species\n",
      "Apusomonadidae Group 1 X: No match, family\n",
      "Hexalonchidae: No match, family\n",
      "Vampyrellida: No match, orderAcantharea C: No match, order\n",
      "\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Apusomonadidae Group 1: No match, order\n",
      "Ochromonadales clade XII X: No match, species\n",
      "Ochromonadales clade XII X: No match, genus\n",
      "Scaphopoda X: No match, family\n",
      "Ochromonadales clade XII: No match, family\n",
      "Sagitta elegans: Unaccepted, using Parasagitta elegans, species \n",
      "Mediophyceae: Unaccepted, no valid name, class \n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Bjerkandera adusta: No match, species\n",
      "Novel clade 2 X: No match, species\n",
      "Novel clade 2 X: No match, genus\n",
      "Bjerkandera: No match, genus\n",
      "Novel clade 2: No match, family\n",
      "MAST 3F: No match, species\n",
      "MAST 3F: No match, genus\n",
      "MAST 3: No match, family\n",
      "Nanomonadea: No match, order\n",
      "Vannella ebro: No match, species\n",
      "Filosa Imbricatea X: No match, order\n",
      "Filosa Imbricatea: No match, class\n",
      "Tunicaraptor unikontum: No match, species\n",
      "MOCH 4 XXX: No match, species\n",
      "MOCH 4 XXX: No match, genus\n",
      "Tunicaraptor: No match, genus\n",
      "MOCH 4 XX: No match, familyPrasinoderma singulare: No match, species\n",
      "\n",
      "Spumellaria 4b X: No match, species\n",
      "Spumellaria 4b X: No match, genus\n",
      "Colpodella angusta: No match, species\n",
      "Opisthokonta XXXX: No match, family\n",
      "Phoronis architecta: Unaccepted, using Phoronis psammophila, species \n",
      "MOCH 4 X: No match, order\n",
      "Exobasidiomycetes: No match, family\n",
      "Spumellaria 4b: No match, family\n",
      "Opisthokonta XXX: No match, order\n",
      "Hydrozoa X: No match, species\n",
      "Hydrozoa X: No match, genus\n",
      "MOCH 4: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Opisthokonta XX: No match, class\n",
      "Penicillium capsulatum: No match, species\n",
      "Opisthokonta X: No match, subdivision\n",
      "Schizodiscus: Multiple unaccepted matches, species \n",
      "Schizodiscus: Multiple unaccepted matches, genus \n",
      "Spongopylidae: No match, family\n",
      "Hematodinium perezi: No match, species\n",
      "Protoperidinium tricingulatum: Unaccepted, using Islandinium tricingulatum, species \n",
      "Radiolaria XXXX: No match, species\n",
      "Radiolaria XXXX: No match, genus\n",
      "Cryptomonadales X: No match, family\n",
      "Obazoa: No match, supergroup\n",
      "Ventrifissura artocarpoidea: No match, species\n",
      "Radiolaria XXX: No match, family\n",
      "Mycale fibrexilis: Unaccepted, using Mycale (Carmia) fibrexilis, species Ventrifissura: No match, genus\n",
      "\n",
      "Ventrifissuridae: No match, family\n",
      "Radiolaria XX: No match, order\n",
      "Derxomyces: No match, genus\n",
      "Dino Group II Clade 52 XX: No match, species\n",
      "Dino Group II Clade 52 XX: No match, genus\n",
      "Radiolaria X: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Dino Group II X: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group II Clade 55 X: No match, species\n",
      "Dino Group II Clade 55 X: No match, genus\n",
      "Euchitoniidae: No match, family\n",
      "Dino Group II Clade 55: No match, family\n",
      "Dino Group II: No match, order\n",
      "Eutintinnus fraknoi: No match, speciesCentroplasthelida X: No match, subdivision\n",
      "\n",
      "Chattonaria mesnili: No match, species\n",
      "Halocrusticida baliensis: No match, species\n",
      "Centroplasthelida: No match, division\n",
      "Chattonaria: No match, genus\n",
      "Corallicolidae: No match, family\n",
      "Halocrusticida: No match, genus\n",
      "Haliphthorales: No match, family\n",
      "Peronosporomycetes X: No match, order\n",
      "Peronosporomycetes: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Haptista: Unaccepted, no valid name, supergroup \n",
      "Protaspa lineage: No match, family\n",
      "MAST 8B: No match, species\n",
      "MAST 8B: No match, genus\n",
      "Gregarinomorphea XX: No match, family\n",
      "Eimeriida: No match, order\n",
      "MAST 2 X: No match, species\n",
      "MAST 2 X: No match, genus\n",
      "MAST 2: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "MAST 8: No match, family\n",
      "Sagenista X: No match, order\n",
      "Gregarinomorphea X: No match, order\n",
      "Coccidiomorphea: No match, class\n",
      "Pterista: No match, order\n",
      "Gregarinomorphea: No match, class\n",
      "Paralecudina anankea: No match, species\n",
      "Pterocystida: No match, class\n",
      "Centroplasthelida X: No match, subdivision\n",
      "Centroplasthelida: No match, division\n",
      "Haptista: Unaccepted, no valid name, supergroup \n",
      "Labyrinthulomycetes LAB6: No match, species\n",
      "MAST 2B: No match, speciesLabyrinthulomycetes LAB6: No match, genus\n",
      "\n",
      "MAST 2B: No match, genus\n",
      "MAST 2: No match, family\n",
      "Gyrista XX: No match, order\n",
      "Gyrista X: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Paralecudina: No match, genus\n",
      "Dino Group II Clade 40 X: No match, species\n",
      "Dino Group II Clade 40 X: No match, genus\n",
      "Labyrinthulomycetes LAB1 6 8: No match, family\n",
      "Paralecudinids: No match, family\n",
      "Dino Group II Clade 40: No match, familyPrymnesiophyceae Clade B5 X: No match, species\n",
      "\n",
      "Prymnesiophyceae Clade B5 X: No match, genusDino Group II: No match, order\n",
      "\n",
      "Prymnesiophyceae Clade B5: No match, family\n",
      "Labyrinthulomycetes: Unaccepted, using Labyrinthulea, order \n",
      "Stephanoecidae Group H X: No match, species\n",
      "Stephanoecidae Group H X: No match, genus\n",
      "Chroomonas:nucl: No match, species\n",
      "Chroomonas:nucl: No match, genus\n",
      "Stephanoecidae Group H: No match, family\n",
      "Hemiselmidaceae:nucl: No match, family\n",
      "Strombidiidae X: No match, species\n",
      "Telonemia Group 1 X: No match, speciesStrombidiidae X: No match, genusMAST 3D: No match, species\n",
      "\n",
      "\n",
      "Telonemia Group 1 X: No match, genus\n",
      "MAST 3D: No match, genus\n",
      "MAST 3: No match, familyMAST 6 X: No match, species\n",
      "\n",
      "Nanomonadea: No match, orderMAST 6 X: No match, genus\n",
      "\n",
      "Cryptomonadales:nucl: No match, order\n",
      "Ichthyobodonidae X: No match, species\n",
      "Pterocanium campanella: No match, speciesIchthyobodonidae X: No match, genusMAST 6: No match, familyTelonemia Group 1: No match, family\n",
      "\n",
      "\n",
      "\n",
      "Telonemia XXX: No match, order\n",
      "Telonemia XX: No match, class\n",
      "Telonemia X: No match, subdivision\n",
      "Telonemia: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Botuliforma benthica: No match, species\n",
      "Ichthyobodonidae: No match, family\n",
      "Pseudophyllomitidae: No match, order\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "Botuliforma: No match, genusCryptophyta:nucl: No match, division\n",
      "\n",
      "Dinophysiaceae: Unaccepted, using Dinophysaceae, family \n",
      "Botuliformidae: No match, family\n",
      "Cryptista:nucl: No match, supergroup\n",
      "Asterodon dilatatus: Unaccepted, using Diplodontias dilatatus, species \n",
      "Ebriacea: No match, order\n",
      "RAD B X Group Ib X: No match, species\n",
      "RAD B X Group Ib X: No match, genus\n",
      "Favella: Multiple matches, genus \n",
      "Filosa Thecofilosea: No match, class\n",
      "MAST 3 X: No match, species\n",
      "MAST 3 X: No match, genus\n",
      "RAD B X Group Ib: No match, familyMAST 3: No match, family\n",
      "\n",
      "RAD B X: No match, orderNanomonadea: No match, order\n",
      "\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Favellidae: No match, family\n",
      "Choreotrichida Tintinnina: No match, order\n",
      "Asterodon: Unaccepted, using Odontaster, genus \n",
      "Dino Group II Clade 21 X: No match, species\n",
      "Dino Group II Clade 21 X: No match, genus\n",
      "Echinodermata XX: No match, family\n",
      "Dino Group I: No match, order\n",
      "Dino Group II Clade 21: No match, familyDifficilina paramertis: No match, species\n",
      "\n",
      "Dino Group II: No match, order\n",
      "Echinodermata X: No match, order\n",
      "Tridentulus: Unaccepted, using Tridentula, species \n",
      "Tridentulus: Unaccepted, using Tridentula, genus Appendicularia: Multiple matches, family \n",
      "\n",
      "Chromadorea X: No match, familyUrochordata X: No match, order\n",
      "\n",
      "Urochordata: Unaccepted, using Tunicata, class \n",
      "Metazoa: Unaccepted, using Animalia, subdivision \n",
      "Obazoa: No match, supergroup\n",
      "Hexalonchidae: No match, familyColpodida X: No match, species\n",
      "\n",
      "Colpodida X: No match, genus\n",
      "OLIGO5 XX: No match, species\n",
      "OLIGO5 XX: No match, genus\n",
      "Polycladida X: No match, species\n",
      "Polycladida X: No match, genus\n",
      "OLIGO5 X: No match, family\n",
      "Selaginopsis cornigera: No match, speciesPhaeophyceae XX: No match, family\n",
      "\n",
      "OLIGO5: No match, order\n",
      "Phaeophyceae X: No match, orderDino Group I XX: No match, species\n",
      "\n",
      "Dino Group I XX: No match, genus\n",
      "Telonemia Group 1: No match, family\n",
      "Dino Group I X: No match, family\n",
      "Dino Group I: No match, order\n",
      "Selaginopsis: Unaccepted, using Thuiaria, genus \n",
      "Crassostrea gigas: Unaccepted, using Magallana gigas, species \n",
      "Telonemia XXX: No match, order\n",
      "Telonemia XX: No match, class\n",
      "Dino Group I Clade 8 X: No match, species\n",
      "Dino Group I Clade 8 X: No match, genus\n",
      "Telonemia X: No match, subdivision\n",
      "Sey055 lineage X: No match, speciesDino Group I Clade 8: No match, family\n",
      "\n",
      "Dino Group II Clade 44 X: No match, speciesSey055 lineage X: No match, genusDino Group I: No match, order\n",
      "\n",
      "\n",
      "Dino Group II Clade 44 X: No match, genus\n",
      "Bathypecten vulcani: Unaccepted, using Catillopecten vulcani, species \n",
      "Telonemia: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Dino Group II Clade 44: No match, family\n",
      "Dino Group II: No match, order\n",
      "Sey055 lineage: No match, family\n",
      "Vampyrellida: No match, order\n",
      "Ustilaginomycetes X: No match, species\n",
      "Dino Group I Clade 6 X: No match, speciesUstilaginomycetes X: No match, genus\n",
      "\n",
      "Dino Group I Clade 6 X: No match, genus\n",
      "Wickerhamomyces anomalus: No match, species\n",
      "Bathypecten: Unaccepted, using Catillopecten, genus \n",
      "Dino Group I Clade 6: No match, family\n",
      "Dino Group I: No match, order\n",
      "Coscinodiscus radiatus: Multiple unaccepted matches, species \n",
      "Pectinoida: No match, familyWickerhamomyces: No match, genus\n",
      "\n",
      "Leidyana2: No match, genus\n",
      "Pichia membranifaciens: No match, species\n",
      "Ochroconis musae: No match, species\n",
      "Dinophyta XXXX: No match, species\n",
      "Astrosphaeridae X: No match, speciesDinophyta XXXX: No match, genus\n",
      "\n",
      "Astrosphaeridae X: No match, genusAnurofeca: No match, species\n",
      "\n",
      "Anurofeca: No match, genus\n",
      "Ochroconis: No match, genus\n",
      "Dinophyta XXX: No match, family\n",
      "Pseudoperkinsidae X: No match, family\n",
      "Monothalamids: No match, class\n",
      "Dinophyta XX: No match, order\n",
      "Pseudoperkinsidae: No match, order\n",
      "Dinophyta X: No match, class\n",
      "Ichthyophonida: No match, class\n",
      "TAGIRI1 lineage X: No match, species\n",
      "TAGIRI1 lineage X: No match, genus\n",
      "TAGIRI1 lineage: No match, family\n",
      "Ebriacea: No match, order\n",
      "Dino Group II Clade 23 X: No match, speciesFilosa Thecofilosea: No match, class\n",
      "\n",
      "Dino Group II Clade 23 X: No match, genus\n",
      "Dino Group II Clade 23: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group IV: No match, orderRAD B X Group Ia X: No match, species\n",
      "\n",
      "RAD B X Group Ia X: No match, genus\n",
      "Dictyocha speculum: Unaccepted, using Octactis speculum, species \n",
      "RAD B X Group Ia: No match, family\n",
      "RAD B X: No match, orderIchthyobodonidae: No match, family\n",
      "\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Stramenopiles XXXXX: No match, species\n",
      "Stramenopiles XXXXX: No match, genus\n",
      "Stramenopiles XXXX: No match, family\n",
      "Allovahlkampfia spelaea: No match, species\n",
      "Stramenopiles XXX: No match, order\n",
      "Stramenopiles XX: No match, class\n",
      "Allovahlkampfia: No match, genus\n",
      "Lophophaenidae X: No match, species\n",
      "Lophophaenidae X: No match, genus\n",
      "Lophophaenidae: No match, familyLabyrinthulomycetes: Unaccepted, using Labyrinthulea, order \n",
      "\n",
      "Ventrifissura foliiformis: No match, species\n",
      "Ventrifissura: No match, genus\n",
      "Ventrifissuridae: No match, family\n",
      "Acrasidae: No match, family\n",
      "Stramenopiles X: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Cryptomonadales X:nucl: No match, family\n",
      "Corallinales XX: No match, species\n",
      "Corallinales XX: No match, genus\n",
      "Corallinales X: No match, family\n",
      "Watanabea Clade X: No match, family\n",
      "Cryptomonadales:nucl: No match, order\n",
      "Foraminifera XXXX: No match, species\n",
      "Foraminifera XXXX: No match, genus\n",
      "Watanabea Clade: No match, order\n",
      "MAST 9A: No match, species\n",
      "MAST 9A: No match, genus\n",
      "Cryptophyceae:nucl: No match, class\n",
      "Dino Group II Clade 25 X: No match, species\n",
      "Dino Group II Clade 25 X: No match, genus\n",
      "Foraminifera XXX: No match, family\n",
      "MAST 9: No match, family\n",
      "Sagenista X: No match, order\n",
      "Cryptophyta X:nucl: No match, subdivision\n",
      "Dino Group II Clade 25: No match, family\n",
      "Dino Group II: No match, order\n",
      "Foraminifera XX: No match, order\n",
      "Actinocyclus curvatulus: Multiple unaccepted matches, species \n",
      "Cryptophyta:nucl: No match, division\n",
      "Foraminifera X: No match, class\n",
      "Actinocyclus: Multiple matches, genus \n",
      "Cryptista:nucl: No match, supergroup\n",
      "Rhizosphaeridae: No match, family\n",
      "Dino Group II Clade 8 X: No match, species\n",
      "Dino Group II Clade 8 X: No match, genus\n",
      "Dino Group II Clade 8: No match, family\n",
      "Dino Group II: No match, order\n",
      "MAST 8D: No match, species\n",
      "MAST 8D: No match, genus\n",
      "MAST 8: No match, family\n",
      "Sagenista X: No match, order\n",
      "Porosporidae X: No match, species\n",
      "Porosporidae X: No match, genus\n",
      "RAD B X Group Id X: No match, species\n",
      "RAD B X Group Id X: No match, genus\n",
      "Ochromonadales clade V X: No match, species\n",
      "Ochromonadales clade V X: No match, genus\n",
      "RAD B X Group Id: No match, family\n",
      "RAD B X: No match, order\n",
      "RAD B: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Ochromonadales clade V: No match, family\n",
      "Obazoa: No match, supergroup\n",
      "Pterocystida XXX: No match, species\n",
      "Pterocystida XXX: No match, genus\n",
      "Filosa Thecofilosea: No match, class\n",
      "Pterocystida XX: No match, family\n",
      "Sarcinochrysidaceae X: No match, species\n",
      "Sarcinochrysidaceae X: No match, genus\n",
      "Pterocystida X: No match, order\n",
      "Pterocystida: No match, class\n",
      "Echinamoeba exundans: No match, species\n",
      "Ircinia felix f. felix: Unaccepted, using Ircinia felix, species \n",
      "Centroplasthelida X: No match, subdivision\n",
      "Centroplasthelida: No match, division\n",
      "Leptocylindrus danicus: Multiple unaccepted matches, species \n",
      "Haptista: Unaccepted, no valid name, supergroup \n",
      "MAST 4A: No match, species\n",
      "MAST 4A: No match, genus\n",
      "MAST 4: No match, family\n",
      "Sagenista X: No match, order\n",
      "Liburna glaciale: No match, species\n",
      "Dino Group I Clade 3: No match, family\n",
      "Dino Group I: No match, order\n",
      "Dino Group II Clade 18 X: No match, species\n",
      "Dino Group II Clade 18 X: No match, genus\n",
      "Dino Group II Clade 28 X: No match, species\n",
      "Dino Group II Clade 28 X: No match, genus\n",
      "Dino Group II Clade 18: No match, family\n",
      "Dino Group II: No match, order\n",
      "Dino Group II Clade 28: No match, family\n",
      "Dino Group II: No match, order\n",
      "Galactomyces: No match, species\n",
      "Galactomyces: No match, genus\n",
      "Chrysomerophyceae XX: No match, family\n",
      "Dino Group II Clade 14 X: No match, species\n",
      "Dino Group II Clade 14 X: No match, genus\n",
      "Chrysomerophyceae X: No match, order\n",
      "Dino Group II Clade 14: No match, family\n",
      "Dino Group II: No match, order\n",
      "MAST 3C: No match, species\n",
      "MAST 3C: No match, genus\n",
      "MAST 3: No match, family\n",
      "Nanomonadea: No match, order\n",
      "Malassezia slooffiae: No match, species\n",
      "Collozoum amoeboides: No match, species\n",
      "Malassezia: No match, genus\n",
      "RAD C Xb X: No match, species\n",
      "RAD C Xb X: No match, genus\n",
      "Exobasidiomycetes: No match, family\n",
      "RAD C Xb: No match, family\n",
      "RAD C X: No match, order\n",
      "RAD C: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Ichthyophonus irregularis: No match, species\n",
      "MOCH 2 XXX: No match, species\n",
      "MOCH 2 XXX: No match, genus\n",
      "MOCH 2 XX: No match, family\n",
      "MOCH 2 X: No match, order\n",
      "MOCH 2: No match, class\n",
      "Gyrista: No match, subdivision\n",
      "Stramenopiles: No match, division\n",
      "TSAR: No match, supergroup\n",
      "Chrysophyceae Clade EC2I XX: No match, species\n",
      "Chrysophyceae Clade EC2I XX: No match, genus\n",
      "Eugregarinorida EUG1 X: No match, species\n",
      "Eugregarinorida EUG1 X: No match, genus\n",
      "Chrysophyceae Clade EC2I X: No match, family\n",
      "Eugregarinorida EUG1: No match, family\n",
      "Chrysophyceae Clade EC2I: No match, order\n",
      "Embryophyceae XX: No match, family\n",
      "Diplonema papillatum: No match, species\n",
      "Embryophyceae X: No match, order\n",
      "Diplonema: Multiple matches, genus \n",
      "Embryophyceae: No match, class\n",
      "Diplonemidae: No match, family\n",
      "Streptophyta X: No match, subdivision\n",
      "Amphifilaceae: No match, family\n",
      "Labyrinthulomycetes: Unaccepted, using Labyrinthulea, order \n",
      "Chaunacanthid: No match, species\n",
      "Chaunacanthid: No match, genus\n",
      "Acantharea C4: No match, family\n",
      "Acantharea C: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "RAD A Xg X: No match, species\n",
      "RAD A Xg X: No match, genus\n",
      "RAD A Xg: No match, family\n",
      "RAD A X: No match, order\n",
      "RAD A: No match, class\n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Acantharea C3: No match, family\n",
      "Acantharea C: No match, order\n",
      "Acantharea: Unaccepted, using Acantharia, class \n",
      "Radiolaria: Unaccepted, using Radiozoa, subdivision \n",
      "Spumellaria 1 X: No match, species\n",
      "Spumellaria 1 X: No match, genus\n",
      "Spumellaria 1: No match, family\n",
      "Chlorophyta X: No match, subdivision\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    worms_18s = get_worms_from_aphiaid_or_name_parallel(\n",
    "    tax_df = tax_18S,worms_dict=pr2_18S_dict,ordered_rank_columns=['species','genus','family','order','class','subdivision','division','supergroup'],\n",
    "    full_tax_column=\"taxonomy\",full_tax_vI=True,n_proc=6)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...</td>\n",
       "      <td>genus</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:100211</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Trichocomaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...</td>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...</td>\n",
       "      <td>species</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:106286</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Cryptophyta</td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td>Cryptomonadales</td>\n",
       "      <td>Cryptomonadaceae</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...</td>\n",
       "      <td>species</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:101195</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligotrichea</td>\n",
       "      <td>Oligotrichida</td>\n",
       "      <td>Strombidiidae</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida...</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aphiaID</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:131155</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>Polychaeta</td>\n",
       "      <td>Spionida</td>\n",
       "      <td>Spionidae</td>\n",
       "      <td>Prionospio</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...</td>\n",
       "      <td>class</td>\n",
       "      <td>Opalozoa</td>\n",
       "      <td>Opalozoa</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:582466</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Bigyra</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Subphylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            full_tax  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...   \n",
       "2  Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...   \n",
       "3  Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida...   \n",
       "4  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...   \n",
       "\n",
       "                              verbatimIdentification old_taxonRank  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...         genus   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...       species   \n",
       "2  Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...       species   \n",
       "3                                   Prionospio dubia           NaN   \n",
       "4  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...         class   \n",
       "\n",
       "      old name    scientificName                           scientificNameID  \\\n",
       "0  Aspergillus       Aspergillus  urn:lsid:marinespecies.org:taxname:100211   \n",
       "1   Goniomonas        Goniomonas  urn:lsid:marinespecies.org:taxname:106286   \n",
       "2  Strombidium       Strombidium  urn:lsid:marinespecies.org:taxname:101195   \n",
       "3      aphiaID  Prionospio dubia  urn:lsid:marinespecies.org:taxname:131155   \n",
       "4     Opalozoa          Opalozoa  urn:lsid:marinespecies.org:taxname:582466   \n",
       "\n",
       "     kingdom       phylum           class            order            family  \\\n",
       "0      Fungi   Ascomycota  Eurotiomycetes       Eurotiales    Trichocomaceae   \n",
       "1  Chromista  Cryptophyta   Cryptophyceae  Cryptomonadales  Cryptomonadaceae   \n",
       "2  Chromista   Ciliophora    Oligotrichea    Oligotrichida     Strombidiidae   \n",
       "3   Animalia     Annelida      Polychaeta         Spionida         Spionidae   \n",
       "4  Chromista       Bigyra            None             None              None   \n",
       "\n",
       "         genus  taxonRank  \n",
       "0  Aspergillus      Genus  \n",
       "1   Goniomonas      Genus  \n",
       "2  Strombidium      Genus  \n",
       "3   Prionospio    Species  \n",
       "4         None  Subphylum  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Haptista', 'Archaeplastida', 'TSAR', 'Cryptista:nucl', 'Obazoa',\n",
       "       'Provora'], dtype=object)"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s[\"scientificName\"]==\"No match\"]['old name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centropla...</td>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centropla...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;Archaeplastida</td>\n",
       "      <td>Eukaryota;Archaeplastida</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Archaeplastida</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...</td>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Cryptista:nucl</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Eukaryota;TSAR</td>\n",
       "      <td>Eukaryota;TSAR</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenopho...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenopho...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;O...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;O...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...</td>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Ni...</td>\n",
       "      <td>Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Ni...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Provora</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centropla...</td>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centropla...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebuli...</td>\n",
       "      <td>Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebuli...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Provora</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviat...</td>\n",
       "      <td>Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviat...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...</td>\n",
       "      <td>Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centropla...</td>\n",
       "      <td>Eukaryota;Haptista;Centroplasthelida;Centropla...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...</td>\n",
       "      <td>Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Cryptista:nucl</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>Eukaryota;Obazoa</td>\n",
       "      <td>Eukaryota;Obazoa</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>No match</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              full_tax  \\\n",
       "121  Eukaryota;Haptista;Centroplasthelida;Centropla...   \n",
       "4                             Eukaryota;Archaeplastida   \n",
       "19   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...   \n",
       "46   Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...   \n",
       "60                                      Eukaryota;TSAR   \n",
       "65   Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenopho...   \n",
       "81   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...   \n",
       "119  Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;O...   \n",
       "177  Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae   \n",
       "203  Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...   \n",
       "4    Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Ni...   \n",
       "48   Eukaryota;Haptista;Centroplasthelida;Centropla...   \n",
       "62   Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebuli...   \n",
       "72   Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviat...   \n",
       "108  Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...   \n",
       "120  Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...   \n",
       "188  Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...   \n",
       "191  Eukaryota;Haptista;Centroplasthelida;Centropla...   \n",
       "195  Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...   \n",
       "227                                   Eukaryota;Obazoa   \n",
       "\n",
       "                                verbatimIdentification old_taxonRank  \\\n",
       "121  Eukaryota;Haptista;Centroplasthelida;Centropla...    supergroup   \n",
       "4                             Eukaryota;Archaeplastida    supergroup   \n",
       "19   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...    supergroup   \n",
       "46   Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...    supergroup   \n",
       "60                                      Eukaryota;TSAR    supergroup   \n",
       "65   Eukaryota;Obazoa;Opisthokonta;Metazoa;Ctenopho...    supergroup   \n",
       "81   Eukaryota;TSAR;Stramenopiles;Gyrista;Gyrista_X...    supergroup   \n",
       "119  Eukaryota;Obazoa;Opisthokonta;Opisthokonta_X;O...    supergroup   \n",
       "177  Eukaryota;TSAR;Stramenopiles;Gyrista;Mediophyceae    supergroup   \n",
       "203  Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...    supergroup   \n",
       "4    Eukaryota;Provora;Nibbleridia;Nibbleridia_X;Ni...    supergroup   \n",
       "48   Eukaryota;Haptista;Centroplasthelida;Centropla...    supergroup   \n",
       "62   Eukaryota;Provora;Nebulidia;Nebulidia_X;Nebuli...    supergroup   \n",
       "72   Eukaryota;Obazoa;Breviatea;Breviatea_X;Breviat...    supergroup   \n",
       "108  Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...    supergroup   \n",
       "120  Eukaryota;TSAR;Telonemia;Telonemia_X;Telonemia...    supergroup   \n",
       "188  Eukaryota;TSAR;Stramenopiles;Gyrista;Peronospo...    supergroup   \n",
       "191  Eukaryota;Haptista;Centroplasthelida;Centropla...    supergroup   \n",
       "195  Eukaryota:nucl;Cryptista:nucl;Cryptophyta:nucl...    supergroup   \n",
       "227                                   Eukaryota;Obazoa    supergroup   \n",
       "\n",
       "           old name scientificName scientificNameID   kingdom phylum class  \\\n",
       "121        Haptista       No match             None       NaN    NaN   NaN   \n",
       "4    Archaeplastida       No match             None       NaN    NaN   NaN   \n",
       "19             TSAR       No match             None       NaN    NaN   NaN   \n",
       "46   Cryptista:nucl       No match             None       NaN    NaN   NaN   \n",
       "60             TSAR       No match             None       NaN    NaN   NaN   \n",
       "65           Obazoa       No match             None  Animalia   None  None   \n",
       "81             TSAR       No match             None       NaN    NaN   NaN   \n",
       "119          Obazoa       No match             None       NaN    NaN   NaN   \n",
       "177            TSAR       No match             None       NaN    NaN   NaN   \n",
       "203            TSAR       No match             None       NaN    NaN   NaN   \n",
       "4           Provora       No match             None       NaN    NaN   NaN   \n",
       "48         Haptista       No match             None       NaN    NaN   NaN   \n",
       "62          Provora       No match             None       NaN    NaN   NaN   \n",
       "72           Obazoa       No match             None       NaN    NaN   NaN   \n",
       "108            TSAR       No match             None       NaN    NaN   NaN   \n",
       "120            TSAR       No match             None       NaN    NaN   NaN   \n",
       "188            TSAR       No match             None       NaN    NaN   NaN   \n",
       "191        Haptista       No match             None       NaN    NaN   NaN   \n",
       "195  Cryptista:nucl       No match             None       NaN    NaN   NaN   \n",
       "227          Obazoa       No match             None       NaN    NaN   NaN   \n",
       "\n",
       "    order family genus taxonRank  \n",
       "121   NaN    NaN   NaN       NaN  \n",
       "4     NaN    NaN   NaN       NaN  \n",
       "19    NaN    NaN   NaN       NaN  \n",
       "46    NaN    NaN   NaN       NaN  \n",
       "60    NaN    NaN   NaN       NaN  \n",
       "65   None   None  None   Kingdom  \n",
       "81    NaN    NaN   NaN       NaN  \n",
       "119   NaN    NaN   NaN       NaN  \n",
       "177   NaN    NaN   NaN       NaN  \n",
       "203   NaN    NaN   NaN       NaN  \n",
       "4     NaN    NaN   NaN       NaN  \n",
       "48    NaN    NaN   NaN       NaN  \n",
       "62    NaN    NaN   NaN       NaN  \n",
       "72    NaN    NaN   NaN       NaN  \n",
       "108   NaN    NaN   NaN       NaN  \n",
       "120   NaN    NaN   NaN       NaN  \n",
       "188   NaN    NaN   NaN       NaN  \n",
       "191   NaN    NaN   NaN       NaN  \n",
       "195   NaN    NaN   NaN       NaN  \n",
       "227   NaN    NaN   NaN       NaN  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s[\"scientificName\"]==\"No match\"].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.loc[worms_18s[\"scientificName\"]==\"No match\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"scientificName\"]==\"Biota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Eukaryota;Haptista</td>\n",
       "      <td>Eukaryota;Haptista</td>\n",
       "      <td>supergroup</td>\n",
       "      <td>Haptista</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               full_tax verbatimIdentification old_taxonRank  old name  \\\n",
       "109          Unassigned             Unassigned           NaN       NaN   \n",
       "120  Eukaryota;Haptista     Eukaryota;Haptista    supergroup  Haptista   \n",
       "77            Eukaryota              Eukaryota           NaN       NaN   \n",
       "\n",
       "    scientificName scientificNameID kingdom phylum class order family genus  \\\n",
       "109            NaN              NaN     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "120            NaN              NaN     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "77             NaN              NaN     NaN    NaN   NaN   NaN    NaN   NaN   \n",
       "\n",
       "    taxonRank  \n",
       "109       NaN  \n",
       "120       NaN  \n",
       "77        NaN  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota;Haptista\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota;Haptista\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota\",'scientificName'] = \"Biota\"\n",
    "worms_18s.loc[worms_18s[\"full_tax\"]==\"Eukaryota\",'scientificNameID'] = \"urn:lsid:marinespecies.org:taxname:1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>old_taxonRank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       full_tax verbatimIdentification old_taxonRank old name scientificName  \\\n",
       "109  Unassigned             Unassigned           NaN      NaN            NaN   \n",
       "\n",
       "    scientificNameID kingdom phylum class order family genus taxonRank  \n",
       "109              NaN     NaN    NaN   NaN   NaN    NaN   NaN       NaN  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s['scientificName'].isna() == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 13)\n",
      "(0, 13)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(worms_18s[worms_18s['scientificName'].isna() == True].shape)\n",
    "worms_18s.loc[worms_18s['scientificName'].isna() == True,'scientificName'] = 'incertae sedis'\n",
    "worms_18s.loc[worms_18s['scientificName'] == 'incertae sedis','scientificNameID'] =  'urn:lsid:marinespecies.org:taxname:12'\n",
    "print(worms_18s[worms_18s['scientificName'].isna() == True].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(332, 13)"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s[worms_18s[\"old name\"]==\"aphiaID\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "worms_18s.to_csv(\"../gomecc_processed/worms_18S_matching.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>full_tax</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:100211</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Trichocomaceae</td>\n",
       "      <td>Aspergillus</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...</td>\n",
       "      <td>Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:106286</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Cryptophyta</td>\n",
       "      <td>Cryptophyceae</td>\n",
       "      <td>Cryptomonadales</td>\n",
       "      <td>Cryptomonadaceae</td>\n",
       "      <td>Goniomonas</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:101195</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Ciliophora</td>\n",
       "      <td>Oligotrichea</td>\n",
       "      <td>Oligotrichida</td>\n",
       "      <td>Strombidiidae</td>\n",
       "      <td>Strombidium</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida...</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>Prionospio dubia</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:131155</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>Polychaeta</td>\n",
       "      <td>Spionida</td>\n",
       "      <td>Spionidae</td>\n",
       "      <td>Prionospio</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...</td>\n",
       "      <td>Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...</td>\n",
       "      <td>Opalozoa</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:582466</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Bigyra</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Subphylum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            full_tax  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...   \n",
       "2  Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...   \n",
       "3  Eukaryota;Obazoa;Opisthokonta;Metazoa;Annelida...   \n",
       "4  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...   \n",
       "\n",
       "                              verbatimIdentification    scientificName  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Fungi;Ascomycota...       Aspergillus   \n",
       "1  Eukaryota;Cryptista;Cryptophyta;Cryptophyta_X;...        Goniomonas   \n",
       "2  Eukaryota;TSAR;Alveolata;Ciliophora;Spirotrich...       Strombidium   \n",
       "3                                   Prionospio dubia  Prionospio dubia   \n",
       "4  Eukaryota;TSAR;Stramenopiles;Bigyra;Opalozoa;O...          Opalozoa   \n",
       "\n",
       "                            scientificNameID    kingdom       phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:100211      Fungi   Ascomycota   \n",
       "1  urn:lsid:marinespecies.org:taxname:106286  Chromista  Cryptophyta   \n",
       "2  urn:lsid:marinespecies.org:taxname:101195  Chromista   Ciliophora   \n",
       "3  urn:lsid:marinespecies.org:taxname:131155   Animalia     Annelida   \n",
       "4  urn:lsid:marinespecies.org:taxname:582466  Chromista       Bigyra   \n",
       "\n",
       "            class            order            family        genus  taxonRank  \n",
       "0  Eurotiomycetes       Eurotiales    Trichocomaceae  Aspergillus      Genus  \n",
       "1   Cryptophyceae  Cryptomonadales  Cryptomonadaceae   Goniomonas      Genus  \n",
       "2    Oligotrichea    Oligotrichida     Strombidiidae  Strombidium      Genus  \n",
       "3      Polychaeta         Spionida         Spionidae   Prionospio    Species  \n",
       "4            None             None              None         None  Subphylum  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_18s.drop(columns=['old name','old_taxonRank'],inplace=True)\n",
    "worms_18s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>domain</th>\n",
       "      <th>supergroup</th>\n",
       "      <th>division</th>\n",
       "      <th>subdivision</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Obazoa</td>\n",
       "      <td>Opisthokonta</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Maxillopoda</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>TSAR</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "4  2a31e5c01634165da99e7381279baa75   \n",
       "5  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "7  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...   \n",
       "1  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...   \n",
       "4  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...   \n",
       "5  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "7  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...   \n",
       "\n",
       "                                            taxonomy  Confidence     domain  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.922099  Eukaryota   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.999947  Eukaryota   \n",
       "4  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.779948  Eukaryota   \n",
       "5  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.999931  Eukaryota   \n",
       "7  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.986908  Eukaryota   \n",
       "\n",
       "  supergroup      division     subdivision        class          order  \\\n",
       "0     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "1     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "4     Obazoa  Opisthokonta         Metazoa   Arthropoda      Crustacea   \n",
       "5       TSAR     Alveolata  Dinoflagellata  Dinophyceae            NaN   \n",
       "7       TSAR     Alveolata  Dinoflagellata  Dinophyceae  Gymnodiniales   \n",
       "\n",
       "           family          genus                 species  \\\n",
       "0     Maxillopoda     Neocalanus    Neocalanus cristatus   \n",
       "1     Maxillopoda  Clausocalanus  Clausocalanus furcatus   \n",
       "4     Maxillopoda    Acrocalanus             Acrocalanus   \n",
       "5             NaN            NaN                     NaN   \n",
       "7  Gymnodiniaceae            NaN                     NaN   \n",
       "\n",
       "                  eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A               962   \n",
       "4  GOMECC4_27N_Sta1_DCM_A              1164   \n",
       "5  GOMECC4_27N_Sta1_DCM_A               287   \n",
       "7  GOMECC4_27N_Sta1_DCM_A               250   \n",
       "\n",
       "                                        occurrenceID  \n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...  \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...  \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...  \n",
       "5  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...  \n",
       "7  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S rRNA'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Occurrence and worms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146232, 16)"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ['18S rRNA'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonRank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.922099</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Calanidae</td>\n",
       "      <td>Neocalanus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.999947</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Clausocalanidae</td>\n",
       "      <td>Clausocalanus</td>\n",
       "      <td>Species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>0.779948</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>Genus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Class</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>0.986908</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniales</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>None</td>\n",
       "      <td>Family</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  36aa75f9b28f5f831c2d631ba65c2bcb   \n",
       "1  4e38e8ced9070952b314e1880bede1ca   \n",
       "2  2a31e5c01634165da99e7381279baa75   \n",
       "3  ecee60339b2fb88ea6d1c8d18054bed4   \n",
       "4  fa1f1a97dd4ae7c826009186bad26384   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...   \n",
       "1  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...   \n",
       "3  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "4  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...   \n",
       "\n",
       "                                            taxonomy  Confidence  \\\n",
       "0  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.922099   \n",
       "1  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.999947   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...    0.779948   \n",
       "3  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.999931   \n",
       "4  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...    0.986908   \n",
       "\n",
       "                  eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A               962   \n",
       "2  GOMECC4_27N_Sta1_DCM_A              1164   \n",
       "3  GOMECC4_27N_Sta1_DCM_A               287   \n",
       "4  GOMECC4_27N_Sta1_DCM_A               250   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...   \n",
       "\n",
       "                              verbatimIdentification          scientificName  \\\n",
       "0                               Neocalanus cristatus    Neocalanus cristatus   \n",
       "1                             Clausocalanus furcatus  Clausocalanus furcatus   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...             Acrocalanus   \n",
       "3  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...             Dinophyceae   \n",
       "4  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...          Gymnodiniaceae   \n",
       "\n",
       "                            scientificNameID    kingdom      phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104470   Animalia  Arthropoda   \n",
       "1  urn:lsid:marinespecies.org:taxname:104503   Animalia  Arthropoda   \n",
       "2  urn:lsid:marinespecies.org:taxname:104192   Animalia  Arthropoda   \n",
       "3   urn:lsid:marinespecies.org:taxname:19542  Chromista     Myzozoa   \n",
       "4  urn:lsid:marinespecies.org:taxname:109410  Chromista     Myzozoa   \n",
       "\n",
       "         class          order           family          genus taxonRank  \n",
       "0     Copepoda      Calanoida        Calanidae     Neocalanus   Species  \n",
       "1     Copepoda      Calanoida  Clausocalanidae  Clausocalanus   Species  \n",
       "2     Copepoda      Calanoida    Paracalanidae    Acrocalanus     Genus  \n",
       "3  Dinophyceae           None             None           None     Class  \n",
       "4  Dinophyceae  Gymnodiniales   Gymnodiniaceae           None    Family  "
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get identificationRemarks\n",
    "occ18_test = occ['18S rRNA'].copy()\n",
    "occ18_test.drop(columns=['domain','supergroup','division','subdivision','class','order','family','genus','species'],inplace=True)\n",
    "#occ18_test.drop(columns=['old name'],inplace=True)\n",
    "\n",
    "occ18_test = occ18_test.merge(worms_18s, how='left', left_on ='taxonomy', right_on='full_tax')\n",
    "occ18_test.drop(columns='full_tax', inplace=True)\n",
    "occ18_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_gene</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>controls_used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>411</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392695 | https://github.com/ao...</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>260</td>\n",
       "      <td>cutadapt</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2</td>\n",
       "      <td>ASV</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...</td>\n",
       "      <td>https://github.com/aomlomics/gomecc</td>\n",
       "      <td>10.5281/zenodo.8392706 | https://pr2-database....</td>\n",
       "      <td>12 distilled water blanks | 2 PCR no-template ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target_gene  ampliconSize trim_method                    cluster_method  \\\n",
       "0    16S rRNA           411    cutadapt  Tourmaline; qiime2-2021.2; dada2   \n",
       "1    18S rRNA           260    cutadapt  Tourmaline; qiime2-2021.2; dada2   \n",
       "\n",
       "  pid_clustering                                  taxa_class_method  \\\n",
       "0            ASV  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1            ASV  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                         taxa_ref_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "1  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...   \n",
       "\n",
       "                             code_repo  \\\n",
       "0  https://github.com/aomlomics/gomecc   \n",
       "1  https://github.com/aomlomics/gomecc   \n",
       "\n",
       "                            identificationReferences  \\\n",
       "0  10.5281/zenodo.8392695 | https://github.com/ao...   \n",
       "1  10.5281/zenodo.8392706 | https://pr2-database....   \n",
       "\n",
       "                                       controls_used  \n",
       "0  12 distilled water blanks | 2 PCR no-template ...  \n",
       "1  12 distilled water blanks | 2 PCR no-template ...  "
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '18S rRNA','taxa_class_method'].item()\n",
    "occ18_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '18S rRNA','taxa_ref_db'].item()\n",
    "\n",
    "occ18_test['identificationRemarks'] = occ18_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+occ18_test['Confidence'].astype(str) +\", against reference database: \"+occ18_test['taxa_ref_db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; naive-bayes classifier, confidence (at lowest specified taxon): 0.92209885, against reference database: PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zenodo.8392706'"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test['identificationRemarks'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### taxonID, basisOfRecord, eventID, nameAccordingTo, organismQuantityType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['taxonID'] = 'ASV:'+occ18_test['featureid']\n",
    "occ18_test['basisOfRecord'] = 'MaterialSample'\n",
    "occ18_test['nameAccordingTo'] = \"WoRMS\"\n",
    "occ18_test['organismQuantityType'] = \"DNA sequence reads\"\n",
    "occ18_test['recordedBy'] = data['study_data']['recordedBy'].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### associatedSequences, materialSampleID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_name', 'library_id', 'title', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'lib_layout', 'platform',\n",
       "       'instrument_model', 'design_description', 'filetype', 'filename',\n",
       "       'filename2', 'biosample_accession', 'sra_accession', 'seq_method',\n",
       "       'nucl_acid_ext', 'target_gene', 'target_subfragment',\n",
       "       'pcr_primer_forward', 'pcr_primer_reverse', 'pcr_primer_name_forward',\n",
       "       'pcr_primer_name_reverse', 'pcr_primer_reference', 'pcr_cond',\n",
       "       'nucl_acid_amp', 'adapters', 'mid_barcode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['prep_data'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test = occ18_test.merge(data['prep_data'].loc[data['prep_data']['target_gene'] == '18S rRNA',['sample_name','sra_accession','biosample_accession']], how='left', left_on ='eventID', right_on='sample_name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### eventID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['eventID'] = occ18_test['eventID']+\"_18S\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sampleSize by total number of reads per sample\n",
    "x = asv_tables['18S rRNA'].sum(numeric_only=True).astype('int')\n",
    "x.index = x.index+\"_18S\"\n",
    "occ18_test['occurrenceRemarks'] = \"Total sampleSize in DNA sequence reads: \"+occ18_test['eventID'].map(x).astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ18_test.drop(columns=['sample_name','featureid','taxonomy','Confidence','taxa_class_method','taxa_ref_db'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test['associatedSequences'] = occ18_test['sra_accession']+' | '+ occ18_test['biosample_accession']+' | '+data['study_data']['bioproject_accession'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_test.rename(columns={'biosample_accession': 'materialSampleID',\n",
    "                  'sequence': 'DNA_sequence'},inplace=True)\n",
    "                   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop unnneeded columns\n",
    "occ18_test.drop(columns=['sra_accession'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DNA_sequence', 'eventID', 'organismQuantity', 'occurrenceID',\n",
       "       'verbatimIdentification', 'scientificName', 'scientificNameID',\n",
       "       'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'taxonRank',\n",
       "       'identificationRemarks', 'taxonID', 'basisOfRecord', 'nameAccordingTo',\n",
       "       'organismQuantityType', 'recordedBy', 'materialSampleID',\n",
       "       'occurrenceRemarks', 'associatedSequences'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonRank</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>recordedBy</th>\n",
       "      <th>materialSampleID</th>\n",
       "      <th>occurrenceRemarks</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>...</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:36aa75f9b28f5f831c2d631ba65c2bcb</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 9838</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>...</td>\n",
       "      <td>Species</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:4e38e8ced9070952b314e1880bede1ca</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 9838</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>...</td>\n",
       "      <td>Genus</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:2a31e5c01634165da99e7381279baa75</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 9838</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>...</td>\n",
       "      <td>Class</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:ecee60339b2fb88ea6d1c8d18054bed4</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 9838</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>...</td>\n",
       "      <td>Family</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classif...</td>\n",
       "      <td>ASV:fa1f1a97dd4ae7c826009186bad26384</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>Luke Thompson | Katherine Silliman</td>\n",
       "      <td>SAMN37516094</td>\n",
       "      <td>Total sampleSize in DNA sequence reads: 9838</td>\n",
       "      <td>SRR26161153 | SAMN37516094 | PRJNA887898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        DNA_sequence  \\\n",
       "0  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...   \n",
       "1  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...   \n",
       "3  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "4  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_18S              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_18S               962   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_18S              1164   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_18S               287   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_18S               250   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...   \n",
       "\n",
       "                              verbatimIdentification          scientificName  \\\n",
       "0                               Neocalanus cristatus    Neocalanus cristatus   \n",
       "1                             Clausocalanus furcatus  Clausocalanus furcatus   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...             Acrocalanus   \n",
       "3  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...             Dinophyceae   \n",
       "4  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...          Gymnodiniaceae   \n",
       "\n",
       "                            scientificNameID    kingdom      phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104470   Animalia  Arthropoda   \n",
       "1  urn:lsid:marinespecies.org:taxname:104503   Animalia  Arthropoda   \n",
       "2  urn:lsid:marinespecies.org:taxname:104192   Animalia  Arthropoda   \n",
       "3   urn:lsid:marinespecies.org:taxname:19542  Chromista     Myzozoa   \n",
       "4  urn:lsid:marinespecies.org:taxname:109410  Chromista     Myzozoa   \n",
       "\n",
       "         class  ... taxonRank  \\\n",
       "0     Copepoda  ...   Species   \n",
       "1     Copepoda  ...   Species   \n",
       "2     Copepoda  ...     Genus   \n",
       "3  Dinophyceae  ...     Class   \n",
       "4  Dinophyceae  ...    Family   \n",
       "\n",
       "                               identificationRemarks  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classif...   \n",
       "\n",
       "                                taxonID   basisOfRecord nameAccordingTo  \\\n",
       "0  ASV:36aa75f9b28f5f831c2d631ba65c2bcb  MaterialSample           WoRMS   \n",
       "1  ASV:4e38e8ced9070952b314e1880bede1ca  MaterialSample           WoRMS   \n",
       "2  ASV:2a31e5c01634165da99e7381279baa75  MaterialSample           WoRMS   \n",
       "3  ASV:ecee60339b2fb88ea6d1c8d18054bed4  MaterialSample           WoRMS   \n",
       "4  ASV:fa1f1a97dd4ae7c826009186bad26384  MaterialSample           WoRMS   \n",
       "\n",
       "  organismQuantityType                          recordedBy materialSampleID  \\\n",
       "0   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "1   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "2   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "3   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "4   DNA sequence reads  Luke Thompson | Katherine Silliman     SAMN37516094   \n",
       "\n",
       "                              occurrenceRemarks  \\\n",
       "0  Total sampleSize in DNA sequence reads: 9838   \n",
       "1  Total sampleSize in DNA sequence reads: 9838   \n",
       "2  Total sampleSize in DNA sequence reads: 9838   \n",
       "3  Total sampleSize in DNA sequence reads: 9838   \n",
       "4  Total sampleSize in DNA sequence reads: 9838   \n",
       "\n",
       "                        associatedSequences  \n",
       "0  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "1  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "2  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "3  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "4  SRR26161153 | SAMN37516094 | PRJNA887898  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge event and occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_merged = occ18_test.merge(all_event_data,how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>...</th>\n",
       "      <th>minimumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>waterBody</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>datasetID</th>\n",
       "      <th>geodeticDatum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1516</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>Neocalanus cristatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104470</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>962</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>Clausocalanus furcatus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104503</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>1164</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...</td>\n",
       "      <td>Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...</td>\n",
       "      <td>Acrocalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104192</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Copepoda</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>287</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:19542</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>250</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...</td>\n",
       "      <td>Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...</td>\n",
       "      <td>Gymnodiniaceae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:109410</td>\n",
       "      <td>Chromista</td>\n",
       "      <td>Myzozoa</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>USA: Atlantic Ocean, east of Florida (27 N)</td>\n",
       "      <td>26.997</td>\n",
       "      <td>-79.618</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>49</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A</td>\n",
       "      <td>noaa-aoml-gomecc4</td>\n",
       "      <td>WGS84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        DNA_sequence  \\\n",
       "0  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGCC...   \n",
       "1  GCTACTACCGATTGAACGTTTTAGTGAGGTCCTCGGACTGTTTGGT...   \n",
       "2  GCTACTACCGATTGGACGTTTTAGTGAGACATTTGGACTGGGTTAA...   \n",
       "3  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAG...   \n",
       "4  GCTCCTACCGATTGAGTGATCCGGTGAATAATTCGGACTGCAGCAA...   \n",
       "\n",
       "                      eventID  organismQuantity  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_18S              1516   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_18S               962   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_18S              1164   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_18S               287   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_18S               250   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_DCM_A_occ36aa75f9b28f5f831c2d...   \n",
       "1  GOMECC4_27N_Sta1_DCM_A_occ4e38e8ced9070952b314...   \n",
       "2  GOMECC4_27N_Sta1_DCM_A_occ2a31e5c01634165da99e...   \n",
       "3  GOMECC4_27N_Sta1_DCM_A_occecee60339b2fb88ea6d1...   \n",
       "4  GOMECC4_27N_Sta1_DCM_A_occfa1f1a97dd4ae7c82600...   \n",
       "\n",
       "                              verbatimIdentification          scientificName  \\\n",
       "0                               Neocalanus cristatus    Neocalanus cristatus   \n",
       "1                             Clausocalanus furcatus  Clausocalanus furcatus   \n",
       "2  Eukaryota;Obazoa;Opisthokonta;Metazoa;Arthropo...             Acrocalanus   \n",
       "3  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...             Dinophyceae   \n",
       "4  Eukaryota;TSAR;Alveolata;Dinoflagellata;Dinoph...          Gymnodiniaceae   \n",
       "\n",
       "                            scientificNameID    kingdom      phylum  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104470   Animalia  Arthropoda   \n",
       "1  urn:lsid:marinespecies.org:taxname:104503   Animalia  Arthropoda   \n",
       "2  urn:lsid:marinespecies.org:taxname:104192   Animalia  Arthropoda   \n",
       "3   urn:lsid:marinespecies.org:taxname:19542  Chromista     Myzozoa   \n",
       "4  urn:lsid:marinespecies.org:taxname:109410  Chromista     Myzozoa   \n",
       "\n",
       "         class  ... minimumDepthInMeters  \\\n",
       "0     Copepoda  ...                   49   \n",
       "1     Copepoda  ...                   49   \n",
       "2     Copepoda  ...                   49   \n",
       "3  Dinophyceae  ...                   49   \n",
       "4  Dinophyceae  ...                   49   \n",
       "\n",
       "                                      locality decimalLatitude  \\\n",
       "0  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "1  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "2  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "3  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "4  USA: Atlantic Ocean, east of Florida (27 N)          26.997   \n",
       "\n",
       "  decimalLongitude samplingProtocol       waterBody maximumDepthInMeters  \\\n",
       "0          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "1          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "2          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "3          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "4          -79.618      CTD rosette  Atlantic Ocean                   49   \n",
       "\n",
       "            parentEventID          datasetID geodeticDatum  \n",
       "0  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "1  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "2  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "3  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "4  GOMECC4_27N_Sta1_DCM_A  noaa-aoml-gomecc4         WGS84  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_merged.drop(columns=['DNA_sequence']).to_csv(\"../gomecc_v2_processed/occurrence_18S.tsv\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DNA_sequence', 'eventID', 'organismQuantity', 'occurrenceID',\n",
       "       'verbatimIdentification', 'scientificName', 'scientificNameID',\n",
       "       'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'taxonRank',\n",
       "       'identificationRemarks', 'taxonID', 'basisOfRecord', 'nameAccordingTo',\n",
       "       'organismQuantityType', 'recordedBy', 'materialSampleID',\n",
       "       'occurrenceRemarks', 'associatedSequences', 'locationID', 'eventDate',\n",
       "       'minimumDepthInMeters', 'locality', 'decimalLatitude',\n",
       "       'decimalLongitude', 'samplingProtocol', 'waterBody',\n",
       "       'maximumDepthInMeters', 'parentEventID', 'datasetID', 'geodeticDatum'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_merged.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ18_merged['sampleSize'] = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### combine 16s and 18s occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146232, 35)"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ18_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all = pd.concat([occ16_merged,occ18_merged],axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(311390, 35)"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all.drop(columns=['DNA_sequence']).to_csv(\"../gomecc_v2_processed/occurrence.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " temp = pd.DataFrame({'ticker' : ['spx 5/25/2001 p500', 'spx 5/25/2001 p600', 'spx 5/25/2001 p700']})\n",
    "temp['ticker'].str.split(' ').str[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all['sampleSize'] = [item[1] for item in occ_all['occurrenceRemarks'].str.split(\": \")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all['sampleSizeUnit'] = 'DNA sequence reads'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_all.drop(columns=['DNA_sequence']).to_csv(\"../gomecc_v2_processed/occurrence_sampsizeAdded.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNA-derived"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_dict = dwc_data['dna'].to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['eventID', 'occurrenceID', 'DNA_sequence', 'sop', 'nucl_acid_ext', 'samp_vol_we_dna_ext', 'samp_mat_process', 'nucl_acid_amp', 'target_gene', 'target_subfragment', 'ampliconSize', 'lib_layout', 'pcr_primer_forward', 'pcr_primer_reverse', 'pcr_primer_name_forward', 'pcr_primer_name_reverse', 'pcr_primer_reference', 'pcr_cond', 'seq_meth', 'otu_class_appr', 'otu_seq_comp_appr', 'otu_db', 'env_broad_scale', 'env_local_scale', 'env_medium', 'size_frac', 'concentration', 'concentrationUnit', 'samp_collect_device', 'source_mat_id'])"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_dict.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### sample_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samp_vol_we_dna_ext {'AOML_term': 'samp_vol_we_dna_ext', 'AOML_file': 'sample_data', 'DwC_definition': 'Volume (ml) or mass (g) of total collected sample processed for DNA extraction.MIXS:0000111', 'Example': nan}\n",
      "samp_mat_process {'AOML_term': 'samp_mat_process', 'AOML_file': 'sample_data', 'DwC_definition': 'Any processing applied to the sample during or after retrieving the sample from environment. This field accepts OBI, for a browser of OBI (v 2018-02-12) terms please see http://purl.bioontology.org/ontology/OBI', 'Example': nan}\n",
      "env_broad_scale {'AOML_term': 'env_broad_scale', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "env_local_scale {'AOML_term': 'env_local_scale', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "env_medium {'AOML_term': 'env_medium', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "size_frac {'AOML_term': 'size_frac', 'AOML_file': 'sample_data', 'DwC_definition': 'Filtering pore size used in sample preparation. Examples: 0-0.22 micrometer', 'Example': nan}\n",
      "concentration {'AOML_term': 'dna_conc', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "concentrationUnit {'AOML_term': 'derived: dna_conc', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "samp_collect_device {'AOML_term': 'samp_collect_device', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan}\n",
      "source_mat_id {'AOML_term': 'source_mat_id', 'AOML_file': 'sample_data', 'DwC_definition': 'used here for the nisken bottle sample', 'Example': nan}\n"
     ]
    }
   ],
   "source": [
    "# check if all event file terms are in sample_data\n",
    "for key in dna_dict.keys():\n",
    "    if dna_dict[key]['AOML_file'] == 'sample_data':\n",
    "        print(key,dna_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename sample_data columns to fit DwC standard\n",
    "rename_dict = {}\n",
    "gen = (x for x in dna_dict.keys() if dna_dict[x]['AOML_file'] == 'sample_data')\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[dna_dict[x]['AOML_term']] = x\n",
    "\n",
    "gen = (x for x in dna_dict.keys() if dna_dict[x]['AOML_file'] == 'prep_data')\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[dna_dict[x]['AOML_term']] = x\n",
    "\n",
    "gen = (x for x in dna_dict.keys() if dna_dict[x]['AOML_file'] == 'analysis_data')\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[dna_dict[x]['AOML_term']] = x\n",
    "\n",
    "dna_sample = data['sample_data'].rename(columns=rename_dict).copy()\n",
    "dna_prep = data['prep_data'].rename(columns=rename_dict).copy()\n",
    "dna_analysis = data['analysis_data'].rename(columns=rename_dict).copy()\n",
    "\n",
    "#dna_sample = dna_sample.drop(columns=[col for col in dna_sample if col not in rename_dict.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>...</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>waterBody</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID serial_number        cruise_id line_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S   GOMECC4_001  GOMECC-4 (2021)     27N   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S   GOMECC4_002  GOMECC-4 (2021)     27N   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S   GOMECC4_003  GOMECC-4 (2021)     27N   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_16S   GOMECC4_004  GOMECC-4 (2021)     27N   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_16S   GOMECC4_005  GOMECC-4 (2021)     27N   \n",
       "\n",
       "    station ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0  27N_Sta1             3                A  GOMECC4_27N_Sta1_Deep   \n",
       "1  27N_Sta1             3                B  GOMECC4_27N_Sta1_Deep   \n",
       "2  27N_Sta1             3                C  GOMECC4_27N_Sta1_Deep   \n",
       "3  27N_Sta1            14                A   GOMECC4_27N_Sta1_DCM   \n",
       "4  27N_Sta1            14                B   GOMECC4_27N_Sta1_DCM   \n",
       "\n",
       "                              biological_replicates extract_number  ...  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C      Plate4_52  ...   \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C      Plate4_60  ...   \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B      Plate4_62  ...   \n",
       "3    GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C      Plate4_53  ...   \n",
       "4    GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C      Plate4_46  ...   \n",
       "\n",
       "       samp_store_loc samp_store_temp         silicate  size_frac_low  \\\n",
       "0  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "1  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "2  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "3  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg  no pre-filter   \n",
       "4  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg  no pre-filter   \n",
       "\n",
       "  size_frac_up       temp  tot_alkalinity tot_depth_water_col transmittance  \\\n",
       "0      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "1      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "2      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "3      0.22 µm  28.592 °C    2371 µmol/kg               623 m         4.665   \n",
       "4      0.22 µm  28.592 °C    2371 µmol/kg               623 m         4.665   \n",
       "\n",
       "        waterBody  \n",
       "0  Atlantic Ocean  \n",
       "1  Atlantic Ocean  \n",
       "2  Atlantic Ocean  \n",
       "3  Atlantic Ocean  \n",
       "4  Atlantic Ocean  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_16 = dna_sample[dna_sample['amplicon_sequenced'].str.contains('16S rRNA')]\n",
    "dna_16['eventID'] = dna_16['eventID']+\"_16S\"\n",
    "dna_16.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>...</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>waterBody</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_18S</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_18S</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_18S</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_18S</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID serial_number        cruise_id line_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_18S   GOMECC4_001  GOMECC-4 (2021)     27N   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_18S   GOMECC4_002  GOMECC-4 (2021)     27N   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_18S   GOMECC4_003  GOMECC-4 (2021)     27N   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_18S   GOMECC4_004  GOMECC-4 (2021)     27N   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_18S   GOMECC4_005  GOMECC-4 (2021)     27N   \n",
       "\n",
       "    station ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0  27N_Sta1             3                A  GOMECC4_27N_Sta1_Deep   \n",
       "1  27N_Sta1             3                B  GOMECC4_27N_Sta1_Deep   \n",
       "2  27N_Sta1             3                C  GOMECC4_27N_Sta1_Deep   \n",
       "3  27N_Sta1            14                A   GOMECC4_27N_Sta1_DCM   \n",
       "4  27N_Sta1            14                B   GOMECC4_27N_Sta1_DCM   \n",
       "\n",
       "                              biological_replicates extract_number  ...  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C      Plate4_52  ...   \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C      Plate4_60  ...   \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B      Plate4_62  ...   \n",
       "3    GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C      Plate4_53  ...   \n",
       "4    GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C      Plate4_46  ...   \n",
       "\n",
       "       samp_store_loc samp_store_temp         silicate  size_frac_low  \\\n",
       "0  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "1  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "2  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "3  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg  no pre-filter   \n",
       "4  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg  no pre-filter   \n",
       "\n",
       "  size_frac_up       temp  tot_alkalinity tot_depth_water_col transmittance  \\\n",
       "0      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "1      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "2      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "3      0.22 µm  28.592 °C    2371 µmol/kg               623 m         4.665   \n",
       "4      0.22 µm  28.592 °C    2371 µmol/kg               623 m         4.665   \n",
       "\n",
       "        waterBody  \n",
       "0  Atlantic Ocean  \n",
       "1  Atlantic Ocean  \n",
       "2  Atlantic Ocean  \n",
       "3  Atlantic Ocean  \n",
       "4  Atlantic Ocean  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_18 = dna_sample[dna_sample['amplicon_sequenced'].str.contains('18S rRNA')]\n",
    "dna_18['eventID'] = dna_18['eventID']+\"_18S\"\n",
    "dna_18.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>serial_number</th>\n",
       "      <th>cruise_id</th>\n",
       "      <th>line_id</th>\n",
       "      <th>station</th>\n",
       "      <th>ctd_bottle_no</th>\n",
       "      <th>sample_replicate</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>biological_replicates</th>\n",
       "      <th>extract_number</th>\n",
       "      <th>...</th>\n",
       "      <th>samp_store_loc</th>\n",
       "      <th>samp_store_temp</th>\n",
       "      <th>silicate</th>\n",
       "      <th>size_frac_low</th>\n",
       "      <th>size_frac_up</th>\n",
       "      <th>temp</th>\n",
       "      <th>tot_alkalinity</th>\n",
       "      <th>tot_depth_water_col</th>\n",
       "      <th>transmittance</th>\n",
       "      <th>waterBody</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_001</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_52</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_16S</td>\n",
       "      <td>GOMECC4_002</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C</td>\n",
       "      <td>Plate4_60</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_C_16S</td>\n",
       "      <td>GOMECC4_003</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B</td>\n",
       "      <td>Plate4_62</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>20.3569 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>7.479 °C</td>\n",
       "      <td>2318.9 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.7221</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_16S</td>\n",
       "      <td>GOMECC4_004</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_53</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_16S</td>\n",
       "      <td>GOMECC4_005</td>\n",
       "      <td>GOMECC-4 (2021)</td>\n",
       "      <td>27N</td>\n",
       "      <td>27N_Sta1</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM</td>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C</td>\n",
       "      <td>Plate4_46</td>\n",
       "      <td>...</td>\n",
       "      <td>NOAA/AOML Room 248</td>\n",
       "      <td>-20 °C</td>\n",
       "      <td>1.05635 µmol/kg</td>\n",
       "      <td>no pre-filter</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>28.592 °C</td>\n",
       "      <td>2371 µmol/kg</td>\n",
       "      <td>623 m</td>\n",
       "      <td>4.665</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID serial_number        cruise_id line_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S   GOMECC4_001  GOMECC-4 (2021)     27N   \n",
       "1  GOMECC4_27N_Sta1_Deep_B_16S   GOMECC4_002  GOMECC-4 (2021)     27N   \n",
       "2  GOMECC4_27N_Sta1_Deep_C_16S   GOMECC4_003  GOMECC-4 (2021)     27N   \n",
       "3   GOMECC4_27N_Sta1_DCM_A_16S   GOMECC4_004  GOMECC-4 (2021)     27N   \n",
       "4   GOMECC4_27N_Sta1_DCM_B_16S   GOMECC4_005  GOMECC-4 (2021)     27N   \n",
       "\n",
       "    station ctd_bottle_no sample_replicate          source_mat_id  \\\n",
       "0  27N_Sta1             3                A  GOMECC4_27N_Sta1_Deep   \n",
       "1  27N_Sta1             3                B  GOMECC4_27N_Sta1_Deep   \n",
       "2  27N_Sta1             3                C  GOMECC4_27N_Sta1_Deep   \n",
       "3  27N_Sta1            14                A   GOMECC4_27N_Sta1_DCM   \n",
       "4  27N_Sta1            14                B   GOMECC4_27N_Sta1_DCM   \n",
       "\n",
       "                              biological_replicates extract_number  ...  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_B, GOMECC4_27N_Sta1_Deep_C      Plate4_52  ...   \n",
       "1  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_C      Plate4_60  ...   \n",
       "2  GOMECC4_27N_Sta1_Deep_A, GOMECC4_27N_Sta1_Deep_B      Plate4_62  ...   \n",
       "3    GOMECC4_27N_Sta1_DCM_B, GOMECC4_27N_Sta1_DCM_C      Plate4_53  ...   \n",
       "4    GOMECC4_27N_Sta1_DCM_A, GOMECC4_27N_Sta1_DCM_C      Plate4_46  ...   \n",
       "\n",
       "       samp_store_loc samp_store_temp         silicate  size_frac_low  \\\n",
       "0  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "1  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "2  NOAA/AOML Room 248          -20 °C  20.3569 µmol/kg  no pre-filter   \n",
       "3  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg  no pre-filter   \n",
       "4  NOAA/AOML Room 248          -20 °C  1.05635 µmol/kg  no pre-filter   \n",
       "\n",
       "  size_frac_up       temp  tot_alkalinity tot_depth_water_col transmittance  \\\n",
       "0      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "1      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "2      0.22 µm   7.479 °C  2318.9 µmol/kg               623 m        4.7221   \n",
       "3      0.22 µm  28.592 °C    2371 µmol/kg               623 m         4.665   \n",
       "4      0.22 µm  28.592 °C    2371 µmol/kg               623 m         4.665   \n",
       "\n",
       "        waterBody  \n",
       "0  Atlantic Ocean  \n",
       "1  Atlantic Ocean  \n",
       "2  Atlantic Ocean  \n",
       "3  Atlantic Ocean  \n",
       "4  Atlantic Ocean  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_sample = pd.concat([dna_16,dna_18],axis=0,ignore_index=True)\n",
    "dna_sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>...</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_1</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_10</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_11</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta49_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_12</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S</td>\n",
       "      <td>GOMECC16S_Plate1_13</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                eventID           library_id  \\\n",
       "4   GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S   GOMECC16S_Plate1_1   \n",
       "6     GOMECC4_GALVESTON_Sta54_DCM_B_16S  GOMECC16S_Plate1_10   \n",
       "8    GOMECC4_GALVESTON_Sta54_Deep_A_16S  GOMECC16S_Plate1_11   \n",
       "10   GOMECC4_GALVESTON_Sta49_Deep_A_16S  GOMECC16S_Plate1_12   \n",
       "12  GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S  GOMECC16S_Plate1_13   \n",
       "\n",
       "                                                title library_strategy  \\\n",
       "4   16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "6   16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "8   16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "10  16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "12  16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "\n",
       "   library_source library_selection lib_layout  platform instrument_model  \\\n",
       "4     METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "6     METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "8     METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "10    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "12    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "\n",
       "                                   design_description  ... target_subfragment  \\\n",
       "4   Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "6   Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "8   Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "10  Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "12  Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "\n",
       "     pcr_primer_forward    pcr_primer_reverse pcr_primer_name_forward  \\\n",
       "4   GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "6   GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "8   GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "10  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "12  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "\n",
       "   pcr_primer_name_reverse     pcr_primer_reference  \\\n",
       "4                     926R  10.1111/1462-2920.13023   \n",
       "6                     926R  10.1111/1462-2920.13023   \n",
       "8                     926R  10.1111/1462-2920.13023   \n",
       "10                    926R  10.1111/1462-2920.13023   \n",
       "12                    926R  10.1111/1462-2920.13023   \n",
       "\n",
       "                                             pcr_cond  \\\n",
       "4   initial denaturation:95_2;denaturation:95_0.75...   \n",
       "6   initial denaturation:95_2;denaturation:95_0.75...   \n",
       "8   initial denaturation:95_2;denaturation:95_0.75...   \n",
       "10  initial denaturation:95_2;denaturation:95_0.75...   \n",
       "12  initial denaturation:95_2;denaturation:95_0.75...   \n",
       "\n",
       "              nucl_acid_amp                                       adapters  \\\n",
       "4   10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "6   10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "8   10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "10  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "12  10.1111/1462-2920.13023  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT   \n",
       "\n",
       "              mid_barcode  \n",
       "4   missing: not provided  \n",
       "6   missing: not provided  \n",
       "8   missing: not provided  \n",
       "10  missing: not provided  \n",
       "12  missing: not provided  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prep_16 = dna_prep[dna_prep['target_gene'].str.contains('16S rRNA')].copy()\n",
    "prep_16['eventID'] = prep_16['eventID']+\"_16S\"\n",
    "prep_16.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>...</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_A_18S</td>\n",
       "      <td>GOMECC18S_Plate4_53</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_B_18S</td>\n",
       "      <td>GOMECC18S_Plate4_46</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GOMECC4_27N_Sta1_DCM_C_18S</td>\n",
       "      <td>GOMECC18S_Plate4_54</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_18S</td>\n",
       "      <td>GOMECC18S_Plate4_52</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_B_18S</td>\n",
       "      <td>GOMECC18S_Plate4_60</td>\n",
       "      <td>18S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V9</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID           library_id  \\\n",
       "1   GOMECC4_27N_Sta1_DCM_A_18S  GOMECC18S_Plate4_53   \n",
       "3   GOMECC4_27N_Sta1_DCM_B_18S  GOMECC18S_Plate4_46   \n",
       "5   GOMECC4_27N_Sta1_DCM_C_18S  GOMECC18S_Plate4_54   \n",
       "7  GOMECC4_27N_Sta1_Deep_A_18S  GOMECC18S_Plate4_52   \n",
       "9  GOMECC4_27N_Sta1_Deep_B_18S  GOMECC18S_Plate4_60   \n",
       "\n",
       "                                               title library_strategy  \\\n",
       "1  18S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "3  18S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "5  18S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "7  18S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "9  18S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "\n",
       "  library_source library_selection lib_layout  platform instrument_model  \\\n",
       "1    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "3    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "5    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "7    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "9    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "\n",
       "                                  design_description  ... target_subfragment  \\\n",
       "1  Samples were collected and filtered onto Steri...  ...                 V9   \n",
       "3  Samples were collected and filtered onto Steri...  ...                 V9   \n",
       "5  Samples were collected and filtered onto Steri...  ...                 V9   \n",
       "7  Samples were collected and filtered onto Steri...  ...                 V9   \n",
       "9  Samples were collected and filtered onto Steri...  ...                 V9   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse pcr_primer_name_forward  \\\n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "5   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "7   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "9   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse          pcr_primer_reference  \\\n",
       "1                   EukBr  10.1371/journal.pone.0006372   \n",
       "3                   EukBr  10.1371/journal.pone.0006372   \n",
       "5                   EukBr  10.1371/journal.pone.0006372   \n",
       "7                   EukBr  10.1371/journal.pone.0006372   \n",
       "9                   EukBr  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                            pcr_cond  \\\n",
       "1  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "3  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "5  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "7  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "9  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "\n",
       "                  nucl_acid_amp  \\\n",
       "1  10.1371/journal.pone.0006372   \n",
       "3  10.1371/journal.pone.0006372   \n",
       "5  10.1371/journal.pone.0006372   \n",
       "7  10.1371/journal.pone.0006372   \n",
       "9  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                        adapters            mid_barcode  \n",
       "1  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "3  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "5  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "7  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "9  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prep_18 = dna_prep[dna_prep['target_gene'].str.contains('18S rRNA')].copy()\n",
    "prep_18['eventID'] = prep_18['eventID']+\"_18S\"\n",
    "prep_18.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>...</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>adapters</th>\n",
       "      <th>mid_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_1</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_DCM_B_16S</td>\n",
       "      <td>GOMECC16S_Plate1_10</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta54_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_11</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_GALVESTON_Sta49_Deep_A_16S</td>\n",
       "      <td>GOMECC16S_Plate1_12</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S</td>\n",
       "      <td>GOMECC16S_Plate1_13</td>\n",
       "      <td>16S amplicon metabarcoding of marine metagenom...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>paired</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Samples were collected and filtered onto Steri...</td>\n",
       "      <td>...</td>\n",
       "      <td>V4-V5</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>CCGYCAATTYMTTTRAGTTT</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT</td>\n",
       "      <td>missing: not provided</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               eventID           library_id  \\\n",
       "0  GOMECC4_BROWNSVILLE_Sta66_DCM_B_16S   GOMECC16S_Plate1_1   \n",
       "1    GOMECC4_GALVESTON_Sta54_DCM_B_16S  GOMECC16S_Plate1_10   \n",
       "2   GOMECC4_GALVESTON_Sta54_Deep_A_16S  GOMECC16S_Plate1_11   \n",
       "3   GOMECC4_GALVESTON_Sta49_Deep_A_16S  GOMECC16S_Plate1_12   \n",
       "4  GOMECC4_BROWNSVILLE_Sta66_DCM_C_16S  GOMECC16S_Plate1_13   \n",
       "\n",
       "                                               title library_strategy  \\\n",
       "0  16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "1  16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "2  16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "3  16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "4  16S amplicon metabarcoding of marine metagenom...         AMPLICON   \n",
       "\n",
       "  library_source library_selection lib_layout  platform instrument_model  \\\n",
       "0    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "1    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "2    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "3    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "4    METAGENOMIC               PCR     paired  ILLUMINA   Illumina MiSeq   \n",
       "\n",
       "                                  design_description  ... target_subfragment  \\\n",
       "0  Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "1  Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "2  Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "3  Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "4  Samples were collected and filtered onto Steri...  ...              V4-V5   \n",
       "\n",
       "    pcr_primer_forward    pcr_primer_reverse pcr_primer_name_forward  \\\n",
       "0  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "1  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "2  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "3  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "4  GTGYCAGCMGCCGCGGTAA  CCGYCAATTYMTTTRAGTTT                  515F-Y   \n",
       "\n",
       "  pcr_primer_name_reverse     pcr_primer_reference  \\\n",
       "0                    926R  10.1111/1462-2920.13023   \n",
       "1                    926R  10.1111/1462-2920.13023   \n",
       "2                    926R  10.1111/1462-2920.13023   \n",
       "3                    926R  10.1111/1462-2920.13023   \n",
       "4                    926R  10.1111/1462-2920.13023   \n",
       "\n",
       "                                            pcr_cond            nucl_acid_amp  \\\n",
       "0  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "1  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "2  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "3  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "4  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "\n",
       "                                        adapters            mid_barcode  \n",
       "0  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "1  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "2  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "3  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "4  ACACTGACGACATGGTTCTACA;TACGGTAGCAGAGACTTGGTCT  missing: not provided  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_prep = pd.concat([prep_16,prep_18],axis=0,ignore_index=True)\n",
    "dna_prep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge prep and sample\n",
    "dna = dna_sample.merge(dna_prep, how='outer', on='eventID')\n",
    "dna = dna.merge(dna_analysis,how='outer',on='target_gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_values(['samp_vol_we_dna_ext', 'samp_mat_process', 'env_broad_scale', 'env_local_scale', 'env_medium', 'size_frac', 'concentration', 'concentrationUnit', 'samp_collect_device', 'source_mat_id', 'eventID', 'nucl_acid_ext', 'nucl_acid_amp', 'target_gene', 'target_subfragment', 'lib_layout', 'pcr_primer_forward', 'pcr_primer_reverse', 'pcr_primer_name_forward', 'pcr_primer_name_reverse', 'pcr_primer_reference', 'pcr_cond', 'seq_meth', 'sop', 'ampliconSize', 'otu_class_appr', 'otu_seq_comp_appr', 'otu_db'])"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rename_dict.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['serial_number',\n",
       " 'cruise_id',\n",
       " 'line_id',\n",
       " 'station',\n",
       " 'ctd_bottle_no',\n",
       " 'sample_replicate',\n",
       " 'biological_replicates',\n",
       " 'extract_number',\n",
       " 'sample_title',\n",
       " 'bioproject_accession',\n",
       " 'biosample_accession_x',\n",
       " 'study_id',\n",
       " 'study_title',\n",
       " 'amplicon_sequenced',\n",
       " 'metagenome_sequenced',\n",
       " 'organism',\n",
       " 'collection_date_local',\n",
       " 'collection_date',\n",
       " 'depth',\n",
       " 'geo_loc_name',\n",
       " 'lat_lon',\n",
       " 'decimalLatitude',\n",
       " 'decimalLongitude',\n",
       " 'sample_type',\n",
       " 'collection_method',\n",
       " 'basisOfRecord',\n",
       " 'cluster_16s',\n",
       " 'cluster_18s',\n",
       " 'notes_sampling',\n",
       " 'notes_bottle_metadata',\n",
       " 'line_position',\n",
       " 'offshore_inshore_200m_isobath',\n",
       " 'depth_category',\n",
       " 'ocean_acidification_status',\n",
       " 'seascape_class',\n",
       " 'seascape_probability',\n",
       " 'seascape_window',\n",
       " 'dna_sample_number',\n",
       " 'dna_yield',\n",
       " 'extraction_plate_name',\n",
       " 'extraction_well_number',\n",
       " 'extraction_well_position',\n",
       " 'ship_crs_expocode',\n",
       " 'woce_sect',\n",
       " 'ammonium',\n",
       " 'carbonate',\n",
       " 'diss_inorg_carb',\n",
       " 'diss_oxygen',\n",
       " 'fluor',\n",
       " 'hydrogen_ion',\n",
       " 'nitrate',\n",
       " 'nitrite',\n",
       " 'nitrate_plus_nitrite',\n",
       " 'omega_arag',\n",
       " 'pco2',\n",
       " 'ph',\n",
       " 'phosphate',\n",
       " 'pressure',\n",
       " 'salinity',\n",
       " 'samp_store_loc',\n",
       " 'samp_store_temp',\n",
       " 'silicate',\n",
       " 'size_frac_low',\n",
       " 'size_frac_up',\n",
       " 'temp',\n",
       " 'tot_alkalinity',\n",
       " 'tot_depth_water_col',\n",
       " 'transmittance',\n",
       " 'waterBody',\n",
       " 'library_id',\n",
       " 'title',\n",
       " 'library_strategy',\n",
       " 'library_source',\n",
       " 'library_selection',\n",
       " 'platform',\n",
       " 'instrument_model',\n",
       " 'design_description',\n",
       " 'filetype',\n",
       " 'filename',\n",
       " 'filename2',\n",
       " 'biosample_accession_y',\n",
       " 'sra_accession',\n",
       " 'adapters',\n",
       " 'mid_barcode',\n",
       " 'trim_method',\n",
       " 'cluster_method',\n",
       " 'pid_clustering',\n",
       " 'code_repo',\n",
       " 'identificationReferences',\n",
       " 'controls_used']"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[col for col in dna if col not in rename_dict.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = dna.drop(columns=[col for col in dna if col not in rename_dict.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>concentration</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_B_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2040 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.634 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM_C_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_DCM</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2080 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>2.307 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_A_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2100 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.286 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_B_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2000 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.831 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface_C_18S</td>\n",
       "      <td>GOMECC4_CAPECORAL_Sta141_Surface</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine photic zone [ENVO:00000209]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>2000 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>1.849 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>initial denaturation:94_3;denaturation:94_0.75...</td>\n",
       "      <td>10.1371/journal.pone.0006372</td>\n",
       "      <td>260</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    eventID                     source_mat_id  \\\n",
       "939      GOMECC4_CAPECORAL_Sta141_DCM_B_18S      GOMECC4_CAPECORAL_Sta141_DCM   \n",
       "940      GOMECC4_CAPECORAL_Sta141_DCM_C_18S      GOMECC4_CAPECORAL_Sta141_DCM   \n",
       "941  GOMECC4_CAPECORAL_Sta141_Surface_A_18S  GOMECC4_CAPECORAL_Sta141_Surface   \n",
       "942  GOMECC4_CAPECORAL_Sta141_Surface_B_18S  GOMECC4_CAPECORAL_Sta141_Surface   \n",
       "943  GOMECC4_CAPECORAL_Sta141_Surface_C_18S  GOMECC4_CAPECORAL_Sta141_Surface   \n",
       "\n",
       "                  env_broad_scale                     env_local_scale  \\\n",
       "939  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "940  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "941  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "942  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "943  marine biome [ENVO:00000447]  marine photic zone [ENVO:00000209]   \n",
       "\n",
       "                    env_medium samp_vol_we_dna_ext samp_collect_device  \\\n",
       "939  sea water [ENVO:00002149]             2040 ml       Niskin bottle   \n",
       "940  sea water [ENVO:00002149]             2080 ml       Niskin bottle   \n",
       "941  sea water [ENVO:00002149]             2100 ml       Niskin bottle   \n",
       "942  sea water [ENVO:00002149]             2000 ml       Niskin bottle   \n",
       "943  sea water [ENVO:00002149]             2000 ml       Niskin bottle   \n",
       "\n",
       "                                      samp_mat_process size_frac  \\\n",
       "939  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm   \n",
       "940  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm   \n",
       "941  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm   \n",
       "942  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm   \n",
       "943  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm   \n",
       "\n",
       "    concentration  ... pcr_primer_forward        pcr_primer_reverse  \\\n",
       "939   1.634 ng/µl  ...   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "940   2.307 ng/µl  ...   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "941   1.286 ng/µl  ...   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "942   1.831 ng/µl  ...   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "943   1.849 ng/µl  ...   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC   \n",
       "\n",
       "    pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "939                   1391f                   EukBr   \n",
       "940                   1391f                   EukBr   \n",
       "941                   1391f                   EukBr   \n",
       "942                   1391f                   EukBr   \n",
       "943                   1391f                   EukBr   \n",
       "\n",
       "             pcr_primer_reference  \\\n",
       "939  10.1371/journal.pone.0006372   \n",
       "940  10.1371/journal.pone.0006372   \n",
       "941  10.1371/journal.pone.0006372   \n",
       "942  10.1371/journal.pone.0006372   \n",
       "943  10.1371/journal.pone.0006372   \n",
       "\n",
       "                                              pcr_cond  \\\n",
       "939  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "940  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "941  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "942  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "943  initial denaturation:94_3;denaturation:94_0.75...   \n",
       "\n",
       "                    nucl_acid_amp ampliconSize  \\\n",
       "939  10.1371/journal.pone.0006372          260   \n",
       "940  10.1371/journal.pone.0006372          260   \n",
       "941  10.1371/journal.pone.0006372          260   \n",
       "942  10.1371/journal.pone.0006372          260   \n",
       "943  10.1371/journal.pone.0006372          260   \n",
       "\n",
       "                                     otu_seq_comp_appr  \\\n",
       "939  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "940  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "941  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "942  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "943  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                                otu_db  \n",
       "939  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...  \n",
       "940  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...  \n",
       "941  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...  \n",
       "942  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...  \n",
       "943  PR2 v5.0.1; V9 1391f-1510r region; 10.5281/zen...  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### merge with occurrenceID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(944, 25)"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ = dna.merge(occ_all[['eventID','occurrenceID','DNA_sequence']],how='left',on='eventID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(311390, 27)"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_occ.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>concentration</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9...</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b...</td>\n",
       "      <td>TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041...</td>\n",
       "      <td>TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705b...</td>\n",
       "      <td>TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038 ng/µl</td>\n",
       "      <td>...</td>\n",
       "      <td>515F-Y</td>\n",
       "      <td>926R</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db781...</td>\n",
       "      <td>TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID          source_mat_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "4  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "\n",
       "                  env_medium samp_vol_we_dna_ext samp_collect_device  \\\n",
       "0  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "1  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "2  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "3  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "4  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "\n",
       "                                    samp_mat_process size_frac  concentration  \\\n",
       "0  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm  0.08038 ng/µl   \n",
       "1  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm  0.08038 ng/µl   \n",
       "2  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm  0.08038 ng/µl   \n",
       "3  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm  0.08038 ng/µl   \n",
       "4  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm  0.08038 ng/µl   \n",
       "\n",
       "   ... pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "0  ...                  515F-Y                    926R   \n",
       "1  ...                  515F-Y                    926R   \n",
       "2  ...                  515F-Y                    926R   \n",
       "3  ...                  515F-Y                    926R   \n",
       "4  ...                  515F-Y                    926R   \n",
       "\n",
       "      pcr_primer_reference                                           pcr_cond  \\\n",
       "0  10.1111/1462-2920.13023  initial denaturation:95_2;denaturation:95_0.75...   \n",
       "1  10.1111/1462-2920.13023  initial denaturation:95_2;denaturation:95_0.75...   \n",
       "2  10.1111/1462-2920.13023  initial denaturation:95_2;denaturation:95_0.75...   \n",
       "3  10.1111/1462-2920.13023  initial denaturation:95_2;denaturation:95_0.75...   \n",
       "4  10.1111/1462-2920.13023  initial denaturation:95_2;denaturation:95_0.75...   \n",
       "\n",
       "             nucl_acid_amp ampliconSize  \\\n",
       "0  10.1111/1462-2920.13023          411   \n",
       "1  10.1111/1462-2920.13023          411   \n",
       "2  10.1111/1462-2920.13023          411   \n",
       "3  10.1111/1462-2920.13023          411   \n",
       "4  10.1111/1462-2920.13023          411   \n",
       "\n",
       "                                   otu_seq_comp_appr  \\\n",
       "0  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                              otu_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9...   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b...   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041...   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705b...   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db781...   \n",
       "\n",
       "                                        DNA_sequence  \n",
       "0  TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTT...  \n",
       "1  TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGA...  \n",
       "2  TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTC...  \n",
       "3  TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTG...  \n",
       "4  TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAG...  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ['concentration'] = dna_occ['concentration'].str.strip(\" ng/µl\")\n",
    "dna_occ['concentrationUnit'] = \"ng/µl\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sop {'AOML_term': 'sop', 'AOML_file': 'analysis_data', 'DwC_definition': 'Standard operating procedures used in assembly and/or annotation of genomes, metagenomes or environmental sequences. Or A reference to a well documented protocol, e.g. using protocols.io', 'Example': nan}\n",
      "otu_class_appr {'AOML_term': 'derived: cluster_method, pid_clustering', 'AOML_file': 'analysis_data', 'DwC_definition': 'Approach/algorithm when defining OTUs or ASVs, include version and parameters separated by semicolons', 'Example': '\"dada2; 1.14.0; ASV\"'}\n"
     ]
    }
   ],
   "source": [
    "# check if all DwC terms are in dna file\n",
    "for key in dna_dict.keys():\n",
    "    if key not in dna_occ.columns:\n",
    "        print(key,dna_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tourmaline; qiime2-2021.2; dada2'"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data']['cluster_method'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ['seq_meth'] = 'Illumina MiSeq 2x250'\n",
    "dna_occ['otu_class_appr']= data['analysis_data']['cluster_method'][0]+\"; \"+data['analysis_data']['pid_clustering'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sop {'AOML_term': 'sop', 'AOML_file': 'analysis_data', 'DwC_definition': 'Standard operating procedures used in assembly and/or annotation of genomes, metagenomes or environmental sequences. Or A reference to a well documented protocol, e.g. using protocols.io', 'Example': nan}\n"
     ]
    }
   ],
   "source": [
    "# check if all DwC terms are in dna file\n",
    "for key in dna_dict.keys():\n",
    "    if key not in dna_occ.columns:\n",
    "        print(key,dna_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>size_frac</th>\n",
       "      <th>concentration</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>pcr_cond</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>concentrationUnit</th>\n",
       "      <th>otu_class_appr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9...</td>\n",
       "      <td>TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTT...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b...</td>\n",
       "      <td>TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGA...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041...</td>\n",
       "      <td>TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTC...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705b...</td>\n",
       "      <td>TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTG...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>marine mesopelagic zone [ENVO:00000213]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>1920 ml</td>\n",
       "      <td>Niskin bottle</td>\n",
       "      <td>Pumped through Sterivex filter (0.22-µm) using...</td>\n",
       "      <td>0.22 µm</td>\n",
       "      <td>0.08038</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>initial denaturation:95_2;denaturation:95_0.75...</td>\n",
       "      <td>10.1111/1462-2920.13023</td>\n",
       "      <td>411</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; naive-bayes classifier</td>\n",
       "      <td>Silva SSU Ref NR 99 v138.1; 515f-926r region; ...</td>\n",
       "      <td>GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db781...</td>\n",
       "      <td>TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAG...</td>\n",
       "      <td>ng/µl</td>\n",
       "      <td>Tourmaline; qiime2-2021.2; dada2; ASV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       eventID          source_mat_id  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S  GOMECC4_27N_Sta1_Deep   \n",
       "\n",
       "                env_broad_scale                          env_local_scale  \\\n",
       "0  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "1  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "2  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "3  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "4  marine biome [ENVO:00000447]  marine mesopelagic zone [ENVO:00000213]   \n",
       "\n",
       "                  env_medium samp_vol_we_dna_ext samp_collect_device  \\\n",
       "0  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "1  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "2  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "3  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "4  sea water [ENVO:00002149]             1920 ml       Niskin bottle   \n",
       "\n",
       "                                    samp_mat_process size_frac concentration  \\\n",
       "0  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm       0.08038   \n",
       "1  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm       0.08038   \n",
       "2  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm       0.08038   \n",
       "3  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm       0.08038   \n",
       "4  Pumped through Sterivex filter (0.22-µm) using...   0.22 µm       0.08038   \n",
       "\n",
       "   ...     pcr_primer_reference  \\\n",
       "0  ...  10.1111/1462-2920.13023   \n",
       "1  ...  10.1111/1462-2920.13023   \n",
       "2  ...  10.1111/1462-2920.13023   \n",
       "3  ...  10.1111/1462-2920.13023   \n",
       "4  ...  10.1111/1462-2920.13023   \n",
       "\n",
       "                                            pcr_cond            nucl_acid_amp  \\\n",
       "0  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "1  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "2  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "3  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "4  initial denaturation:95_2;denaturation:95_0.75...  10.1111/1462-2920.13023   \n",
       "\n",
       "  ampliconSize                                  otu_seq_comp_appr  \\\n",
       "0          411  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "1          411  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "2          411  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "3          411  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "4          411  Tourmaline; qiime2-2021.2; naive-bayes classifier   \n",
       "\n",
       "                                              otu_db  \\\n",
       "0  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "1  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "2  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "3  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "4  Silva SSU Ref NR 99 v138.1; 515f-926r region; ...   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  GOMECC4_27N_Sta1_Deep_A_16S_occ009257b156ab4a9...   \n",
       "1  GOMECC4_27N_Sta1_Deep_A_16S_occ01398067b1d323b...   \n",
       "2  GOMECC4_27N_Sta1_Deep_A_16S_occ01770ea2fb7f041...   \n",
       "3  GOMECC4_27N_Sta1_Deep_A_16S_occ017dbdc8b62705b...   \n",
       "4  GOMECC4_27N_Sta1_Deep_A_16S_occ069f375524db781...   \n",
       "\n",
       "                                        DNA_sequence concentrationUnit  \\\n",
       "0  TACGAGGGGTGCTAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTT...             ng/µl   \n",
       "1  TACGGAGGGTGCAAGCGTTGTTCGGAATTATTGGGCGTAAAGCGGA...             ng/µl   \n",
       "2  TACGGAGGATCCAAGCGTTATCCGGATTTATTGGGTTTAAAGGGTC...             ng/µl   \n",
       "3  TACTAGGGGTGCAAGCGTTGTCCGGAATTACTGGGCGTAAAGGGTG...             ng/µl   \n",
       "4  TACGTAGGAGGCTAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAG...             ng/µl   \n",
       "\n",
       "                          otu_class_appr  \n",
       "0  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "1  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "2  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "3  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "4  Tourmaline; qiime2-2021.2; dada2; ASV  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_occ.to_csv(\"../gomecc_v2_processed/dna-derived.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I also wanted to persist the original name from the NCBI taxonomy database into the Darwin Core-converted data set. To do this, I queried the database based on the name in the original data to obtain its taxonomic ID number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get set up to query NCBI taxonomy \n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "# ----- Insert your email here -----\n",
    "Entrez.email = 'ksil91@gmail.com'\n",
    "# ----------------------------------\n",
    "\n",
    "# Get list of all databases available through this tool\n",
    "record = Entrez.read(Entrez.einfo())\n",
    "all_dbs = record['DbList']\n",
    "all_dbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ncbi_txid_from_name_parallel(all_names,n_proc=0):\n",
    "    all_names = all_names.unique()\n",
    "    queue = mp.Queue()\n",
    "    if n_proc == 0:\n",
    "    # create as many processes as there are CPUs on your machine\n",
    "        num_processes = mp.cpu_count()\n",
    "    else:\n",
    "        num_processes = n_proc\n",
    "        \n",
    "    # calculate the chunk size as an integer\n",
    "    chunk_size = int(len(all_names)/num_processes)\n",
    "    procs = []\n",
    "    for job in range(num_processes):\n",
    "        start = job * chunk_size\n",
    "        end = start + chunk_size\n",
    "        name_chunk = all_names[start:end]\n",
    "        proc = mp.Process(\n",
    "            target=get_ncbi_txid_from_name,\n",
    "            args=(name_chunk,queue)\n",
    "        )\n",
    "        procs.append(proc)\n",
    "        proc.start()\n",
    "    \n",
    "    names_dict = {}\n",
    "    for _ in procs:\n",
    "        names_dict.update(queue.get())\n",
    "    \n",
    "    #new_df = queue.get()\n",
    "    \n",
    "    for proc in procs:\n",
    "        proc.join()\n",
    "    \n",
    "    return names_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get NCBI taxIDs for each name in dataset ---- TAKES ~ 2 MINUTES FOR 300 RECORDS\n",
    "def get_ncbi_txid_from_name(names,queue):\n",
    "    \n",
    "    name_ncbiid_dict = {}\n",
    "    for name in names:\n",
    "        if name not in name_ncbiid_dict.keys():\n",
    "            taxon2 = '\"' + name + '\"'\n",
    "            handle = Entrez.esearch(db='taxonomy', retmax=10, term=taxon2)\n",
    "            record = Entrez.read(handle)\n",
    "            if not record['IdList'] or not record['IdList'][0]:\n",
    "                print(name+\": not found\")\n",
    "                name_ncbiid_dict[name] = \"None\"\n",
    "            else:\n",
    "                name_ncbiid_dict[name] = record['IdList'][0]\n",
    "            handle.close()\n",
    "    queue.put(name_ncbiid_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "n =worms_12s['verbatimIdentification'][0:10]\n",
    "new = get_ncbi_txid_from_name_parallel(n,7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                 Clupeidae\n",
       "1         Eucinostomus gula\n",
       "2    Oncorhynchus gorbuscha\n",
       "3          Lutjanus griseus\n",
       "4         Eucinostomus gula\n",
       "5          Cheilopogon agoo\n",
       "6    Oncorhynchus gorbuscha\n",
       "7    Oncorhynchus gorbuscha\n",
       "8          Scomber scombrus\n",
       "9        Haemulon plumierii\n",
       "Name: verbatimIdentification, dtype: object"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Oncorhynchus gorbuscha': '8017',\n",
       " 'Scomber scombrus': '13677',\n",
       " 'Eucinostomus gula': '435273',\n",
       " 'Haemulon plumierii': '334415',\n",
       " 'Clupeidae': '55118',\n",
       " 'Cheilopogon agoo': '123223',\n",
       " 'Lutjanus griseus': '40503'}"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                 Clupeidae\n",
       "1         Eucinostomus gula\n",
       "2    Oncorhynchus gorbuscha\n",
       "3          Lutjanus griseus\n",
       "4         Eucinostomus gula\n",
       "5          Cheilopogon agoo\n",
       "6    Oncorhynchus gorbuscha\n",
       "7    Oncorhynchus gorbuscha\n",
       "8          Scomber scombrus\n",
       "9        Haemulon plumierii\n",
       "Name: verbatimIdentification, dtype: object"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Oncorhynchus gorbuscha': '8017',\n",
       " 'Lutjanus griseus': '40503',\n",
       " 'Scomber scombrus': '13677',\n",
       " 'Haemulon plumierii': '334415',\n",
       " 'Eucinostomus gula': '435273',\n",
       " 'Cheilopogon agoo': '123223',\n",
       " 'Clupeidae': '55118'}"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get NCBI taxIDs for each name in dataset ---- TAKES ~ 2 MINUTES FOR 300 RECORDS\n",
    "\n",
    "name_ncbiid_dict = {}\n",
    "\n",
    "for name in worms_12s['verbatimIdentification']:\n",
    "    if name not in name_ncbiid_dict.keys():\n",
    "        taxon2 = '\"' + name + '\"'\n",
    "        handle = Entrez.esearch(db='taxonomy', retmax=10, term=taxon2)\n",
    "        record = Entrez.read(handle)\n",
    "        if not record['IdList'] or not record['IdList'][0]:\n",
    "            print(name+\": not found\")\n",
    "            name_ncbiid_dict[name] = \"None\"\n",
    "        else:\n",
    "            name_ncbiid_dict[name] = record['IdList'][0]\n",
    "        handle.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look up the not found names by hand to get taxid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'None'"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_ncbiid_dict['Lobianchia sp. CBM:ZF:14789'] = '2057084'\n",
    "name_ncbiid_dict['Brotula sp. CBM:ZF:20276']= '2768194'\n",
    "name_ncbiid_dict['Scopelarchus sp. CBM:DNA No. 2000-014']= '2608515'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Clupeidae',\n",
       " 'Eucinostomus gula',\n",
       " 'Oncorhynchus gorbuscha',\n",
       " 'Lutjanus griseus',\n",
       " 'Cheilopogon agoo',\n",
       " 'Scomber scombrus',\n",
       " 'Haemulon plumierii',\n",
       " 'Archosargus probatocephalus',\n",
       " 'Actinopteri',\n",
       " 'Carangidae']"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(name_ncbiid_dict.keys())[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'55118'"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_ncbiid_dict['Clupeidae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>asv</th>\n",
       "      <th>rank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonConceptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>951a3746cb4c488d657fe3c64bdd3d75</td>\n",
       "      <td>family</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:125464</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Clupeiformes</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBI:txid55118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>f278a917dedcc2e0434be6ecb605a642</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>1d0fcad8f69709ed9ea8dd6e6a3c627a</td>\n",
       "      <td>species</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:127182</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Salmoniformes</td>\n",
       "      <td>Salmonidae</td>\n",
       "      <td>Oncorhynchus</td>\n",
       "      <td>NCBI:txid8017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>cc6451b251afdd97ef4f11e47abc53a7</td>\n",
       "      <td>species</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159797</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Lutjanidae</td>\n",
       "      <td>Lutjanus</td>\n",
       "      <td>NCBI:txid40503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>d16bc4ff5741c0ba4a5d0bb2e21e5589</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   verbatimIdentification                               asv     rank  \\\n",
       "0               Clupeidae  951a3746cb4c488d657fe3c64bdd3d75   family   \n",
       "1       Eucinostomus gula  f278a917dedcc2e0434be6ecb605a642  species   \n",
       "2  Oncorhynchus gorbuscha  1d0fcad8f69709ed9ea8dd6e6a3c627a  species   \n",
       "3        Lutjanus griseus  cc6451b251afdd97ef4f11e47abc53a7  species   \n",
       "4       Eucinostomus gula  d16bc4ff5741c0ba4a5d0bb2e21e5589  species   \n",
       "\n",
       "                 old name          scientificName  \\\n",
       "0               Clupeidae               Clupeidae   \n",
       "1       Eucinostomus gula       Eucinostomus gula   \n",
       "2  Oncorhynchus gorbuscha  Oncorhynchus gorbuscha   \n",
       "3        Lutjanus griseus        Lutjanus griseus   \n",
       "4       Eucinostomus gula       Eucinostomus gula   \n",
       "\n",
       "                            scientificNameID   kingdom    phylum      class  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:125464  Animalia  Chordata  Teleostei   \n",
       "1  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "2  urn:lsid:marinespecies.org:taxname:127182  Animalia  Chordata  Teleostei   \n",
       "3  urn:lsid:marinespecies.org:taxname:159797  Animalia  Chordata  Teleostei   \n",
       "4  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "\n",
       "                       order      family         genus   taxonConceptID  \n",
       "0               Clupeiformes   Clupeidae          None   NCBI:txid55118  \n",
       "1  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  \n",
       "2              Salmoniformes  Salmonidae  Oncorhynchus    NCBI:txid8017  \n",
       "3  Eupercaria incertae sedis  Lutjanidae      Lutjanus   NCBI:txid40503  \n",
       "4  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  "
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add NCBI taxonomy IDs under taxonConceptID\n",
    "\n",
    "# Map indicators that say no taxonomy was assigned to empty strings\n",
    "#name_ncbiid_dict['unassigned'], name_ncbiid_dict['s_'], name_ncbiid_dict['no_hit'], name_ncbiid_dict['unknown'], name_ncbiid_dict['g_'] = '', '', '', '', ''\n",
    "\n",
    "# Create column\n",
    "worms_12s['taxonConceptID']  = worms_12s['verbatimIdentification'].copy()\n",
    "worms_12s['taxonConceptID'].replace(name_ncbiid_dict, inplace=True)\n",
    "\n",
    "# Add remainder of text and clean\n",
    "worms_12s['taxonConceptID'] = 'NCBI:txid' + worms_12s['taxonConceptID']\n",
    "worms_12s['taxonConceptID'].replace('NCBI:txid', '', inplace=True)\n",
    "worms_12s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>asv</th>\n",
       "      <th>rank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonConceptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>951a3746cb4c488d657fe3c64bdd3d75</td>\n",
       "      <td>family</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:125464</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Clupeiformes</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBI:txid55118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>f278a917dedcc2e0434be6ecb605a642</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>1d0fcad8f69709ed9ea8dd6e6a3c627a</td>\n",
       "      <td>species</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:127182</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Salmoniformes</td>\n",
       "      <td>Salmonidae</td>\n",
       "      <td>Oncorhynchus</td>\n",
       "      <td>NCBI:txid8017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>cc6451b251afdd97ef4f11e47abc53a7</td>\n",
       "      <td>species</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159797</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Lutjanidae</td>\n",
       "      <td>Lutjanus</td>\n",
       "      <td>NCBI:txid40503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>d16bc4ff5741c0ba4a5d0bb2e21e5589</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   verbatimIdentification                               asv     rank  \\\n",
       "0               Clupeidae  951a3746cb4c488d657fe3c64bdd3d75   family   \n",
       "1       Eucinostomus gula  f278a917dedcc2e0434be6ecb605a642  species   \n",
       "2  Oncorhynchus gorbuscha  1d0fcad8f69709ed9ea8dd6e6a3c627a  species   \n",
       "3        Lutjanus griseus  cc6451b251afdd97ef4f11e47abc53a7  species   \n",
       "4       Eucinostomus gula  d16bc4ff5741c0ba4a5d0bb2e21e5589  species   \n",
       "\n",
       "                 old name          scientificName  \\\n",
       "0               Clupeidae               Clupeidae   \n",
       "1       Eucinostomus gula       Eucinostomus gula   \n",
       "2  Oncorhynchus gorbuscha  Oncorhynchus gorbuscha   \n",
       "3        Lutjanus griseus        Lutjanus griseus   \n",
       "4       Eucinostomus gula       Eucinostomus gula   \n",
       "\n",
       "                            scientificNameID   kingdom    phylum      class  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:125464  Animalia  Chordata  Teleostei   \n",
       "1  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "2  urn:lsid:marinespecies.org:taxname:127182  Animalia  Chordata  Teleostei   \n",
       "3  urn:lsid:marinespecies.org:taxname:159797  Animalia  Chordata  Teleostei   \n",
       "4  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "\n",
       "                       order      family         genus   taxonConceptID  \n",
       "0               Clupeiformes   Clupeidae          None   NCBI:txid55118  \n",
       "1  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  \n",
       "2              Salmoniformes  Salmonidae  Oncorhynchus    NCBI:txid8017  \n",
       "3  Eupercaria incertae sedis  Lutjanidae      Lutjanus   NCBI:txid40503  \n",
       "4  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  "
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_12s.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge Occurrence and worms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>rank</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonConceptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4008ca75e225240b385ec167d0e0c9b4</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Mugiliformes;Mu...</td>\n",
       "      <td>0.920924</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>29</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occ4008ca7...</td>\n",
       "      <td>Mugil curema</td>\n",
       "      <td>species</td>\n",
       "      <td>Mugil curema</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159416</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Mugiliformes</td>\n",
       "      <td>Mugilidae</td>\n",
       "      <td>Mugil</td>\n",
       "      <td>NCBI:txid48194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b8b3d9e1b83477e4927fd8d4c6f1cd2e</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCTA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Salmoniformes;S...</td>\n",
       "      <td>0.912878</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>4041</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occb8b3d9e...</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>species</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:127182</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Salmoniformes</td>\n",
       "      <td>Salmonidae</td>\n",
       "      <td>Oncorhynchus</td>\n",
       "      <td>NCBI:txid8017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bfe5abf81199e96bbc51918d76b77e8c</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGACCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Lutjaniformes;L...</td>\n",
       "      <td>0.991830</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>37</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occbfe5abf...</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>species</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159797</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Lutjanidae</td>\n",
       "      <td>Lutjanus</td>\n",
       "      <td>NCBI:txid40503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>208ffed2e333d449c29995ca62b808cc</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;NA;Opistognathi...</td>\n",
       "      <td>0.999966</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>47</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occ208ffed...</td>\n",
       "      <td>Lonchopisthus micrognathus</td>\n",
       "      <td>species</td>\n",
       "      <td>Lonchopisthus micrognathus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:281386</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Ovalentaria incertae sedis</td>\n",
       "      <td>Opistognathidae</td>\n",
       "      <td>Lonchopisthus</td>\n",
       "      <td>NCBI:txid1311555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2f6cd78f1f099544eb8f1b252fba924c</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGGGGCCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Spariformes;Spa...</td>\n",
       "      <td>0.993729</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>4</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occ2f6cd78...</td>\n",
       "      <td>Archosargus probatocephalus</td>\n",
       "      <td>species</td>\n",
       "      <td>Archosargus probatocephalus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159238</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Sparidae</td>\n",
       "      <td>Archosargus</td>\n",
       "      <td>NCBI:txid119682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  4008ca75e225240b385ec167d0e0c9b4   \n",
       "1  b8b3d9e1b83477e4927fd8d4c6f1cd2e   \n",
       "2  bfe5abf81199e96bbc51918d76b77e8c   \n",
       "3  208ffed2e333d449c29995ca62b808cc   \n",
       "4  2f6cd78f1f099544eb8f1b252fba924c   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...   \n",
       "1  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCTA...   \n",
       "2  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGACCCAA...   \n",
       "3  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...   \n",
       "4  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGGGGCCCAA...   \n",
       "\n",
       "                                            taxonomy  Confidence  \\\n",
       "0  Eukaryota;Chordata;Actinopteri;Mugiliformes;Mu...    0.920924   \n",
       "1  Eukaryota;Chordata;Actinopteri;Salmoniformes;S...    0.912878   \n",
       "2  Eukaryota;Chordata;Actinopteri;Lutjaniformes;L...    0.991830   \n",
       "3  Eukaryota;Chordata;Actinopteri;NA;Opistognathi...    0.999966   \n",
       "4  Eukaryota;Chordata;Actinopteri;Spariformes;Spa...    0.993729   \n",
       "\n",
       "                               eventID  organismQuantity  \\\n",
       "0  SEAMAP2019_Brownsville_Sta103_39m_A                29   \n",
       "1  SEAMAP2019_Brownsville_Sta103_39m_A              4041   \n",
       "2  SEAMAP2019_Brownsville_Sta103_39m_A                37   \n",
       "3  SEAMAP2019_Brownsville_Sta103_39m_A                47   \n",
       "4  SEAMAP2019_Brownsville_Sta103_39m_A                 4   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  SEAMAP2019_Brownsville_Sta103_39m_A_occ4008ca7...   \n",
       "1  SEAMAP2019_Brownsville_Sta103_39m_A_occb8b3d9e...   \n",
       "2  SEAMAP2019_Brownsville_Sta103_39m_A_occbfe5abf...   \n",
       "3  SEAMAP2019_Brownsville_Sta103_39m_A_occ208ffed...   \n",
       "4  SEAMAP2019_Brownsville_Sta103_39m_A_occ2f6cd78...   \n",
       "\n",
       "        verbatimIdentification     rank               scientificName  \\\n",
       "0                 Mugil curema  species                 Mugil curema   \n",
       "1       Oncorhynchus gorbuscha  species       Oncorhynchus gorbuscha   \n",
       "2             Lutjanus griseus  species             Lutjanus griseus   \n",
       "3   Lonchopisthus micrognathus  species   Lonchopisthus micrognathus   \n",
       "4  Archosargus probatocephalus  species  Archosargus probatocephalus   \n",
       "\n",
       "                            scientificNameID   kingdom    phylum      class  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:159416  Animalia  Chordata  Teleostei   \n",
       "1  urn:lsid:marinespecies.org:taxname:127182  Animalia  Chordata  Teleostei   \n",
       "2  urn:lsid:marinespecies.org:taxname:159797  Animalia  Chordata  Teleostei   \n",
       "3  urn:lsid:marinespecies.org:taxname:281386  Animalia  Chordata  Teleostei   \n",
       "4  urn:lsid:marinespecies.org:taxname:159238  Animalia  Chordata  Teleostei   \n",
       "\n",
       "                        order           family          genus  \\\n",
       "0                Mugiliformes        Mugilidae          Mugil   \n",
       "1               Salmoniformes       Salmonidae   Oncorhynchus   \n",
       "2   Eupercaria incertae sedis       Lutjanidae       Lutjanus   \n",
       "3  Ovalentaria incertae sedis  Opistognathidae  Lonchopisthus   \n",
       "4   Eupercaria incertae sedis         Sparidae    Archosargus   \n",
       "\n",
       "     taxonConceptID  \n",
       "0    NCBI:txid48194  \n",
       "1     NCBI:txid8017  \n",
       "2    NCBI:txid40503  \n",
       "3  NCBI:txid1311555  \n",
       "4   NCBI:txid119682  "
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get identificationRemarks\n",
    "occ12_test = occ['12S rRNA'].copy()\n",
    "occ12_test.drop(columns=['domain','phylum','class','order','family','genus','species'],inplace=True)\n",
    "#worms_12s.drop(columns=['old name'],inplace=True)\n",
    "\n",
    "occ12_test = occ12_test.merge(worms_12s, how='left', left_on ='featureid', right_on='asv')\n",
    "occ12_test.drop(columns='asv', inplace=True)\n",
    "occ12_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ12_test.to_csv(\"occ12_test.tsv\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_gene</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>sop</th>\n",
       "      <th>identificationReferences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>X</td>\n",
       "      <td>q2-cutadapt v2023.5 (--p-front-f FWD primer, -...</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>https://github.com/ksilnoaa/seamap-edna</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>X</td>\n",
       "      <td>q2-cutadapt v2023.5 (--p-front-f FWD primer, -...</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>https://github.com/ksilnoaa/seamap-edna</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12S rRNA</td>\n",
       "      <td>X</td>\n",
       "      <td>q2-cutadapt v2023.5 (--p-front-f FWD primer, -...</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>https://github.com/ksilnoaa/seamap-edna</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target_gene ampliconSize                                        trim_method  \\\n",
       "0    16S rRNA            X  q2-cutadapt v2023.5 (--p-front-f FWD primer, -...   \n",
       "1    18S rRNA            X  q2-cutadapt v2023.5 (--p-front-f FWD primer, -...   \n",
       "2    12S rRNA            X  q2-cutadapt v2023.5 (--p-front-f FWD primer, -...   \n",
       "\n",
       "  cluster_method pid_clustering taxa_class_method taxa_ref_db  \\\n",
       "0           test           test              test        test   \n",
       "1           test           test              test        test   \n",
       "2           test           test              test        test   \n",
       "\n",
       "                                 code_repo bioproject_accession   sop  \\\n",
       "0  https://github.com/ksilnoaa/seamap-edna                 test  test   \n",
       "1  https://github.com/ksilnoaa/seamap-edna                 test  test   \n",
       "2  https://github.com/ksilnoaa/seamap-edna                 test  test   \n",
       "\n",
       "  identificationReferences  \n",
       "0                     test  \n",
       "1                     test  \n",
       "2                     test  "
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ12_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '12S rRNA','taxa_class_method'].item()\n",
    "occ12_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '12S rRNA','taxa_ref_db'].item()\n",
    "\n",
    "occ12_test['identificationRemarks'] = occ12_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+str(occ12_test['Confidence']) +\", against reference database: \"+occ12_test['taxa_ref_db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    test, confidence (at lowest specified taxon): ...\n",
       "1    test, confidence (at lowest specified taxon): ...\n",
       "2    test, confidence (at lowest specified taxon): ...\n",
       "3    test, confidence (at lowest specified taxon): ...\n",
       "4    test, confidence (at lowest specified taxon): ...\n",
       "Name: identificationRemarks, dtype: object"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ12_test['identificationRemarks'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### now convert other occurrence info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### identificationRemarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_gene</th>\n",
       "      <th>ampliconSize</th>\n",
       "      <th>trim_method</th>\n",
       "      <th>cluster_method</th>\n",
       "      <th>pid_clustering</th>\n",
       "      <th>taxa_class_method</th>\n",
       "      <th>taxa_ref_db</th>\n",
       "      <th>code_repo</th>\n",
       "      <th>bioproject_accession</th>\n",
       "      <th>sop</th>\n",
       "      <th>identificationReferences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>X</td>\n",
       "      <td>q2-cutadapt v2023.5 (--p-front-f FWD primer, -...</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>https://github.com/ksilnoaa/seamap-edna</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18S rRNA</td>\n",
       "      <td>X</td>\n",
       "      <td>q2-cutadapt v2023.5 (--p-front-f FWD primer, -...</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>https://github.com/ksilnoaa/seamap-edna</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12S rRNA</td>\n",
       "      <td>X</td>\n",
       "      <td>q2-cutadapt v2023.5 (--p-front-f FWD primer, -...</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>https://github.com/ksilnoaa/seamap-edna</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target_gene ampliconSize                                        trim_method  \\\n",
       "0    16S rRNA            X  q2-cutadapt v2023.5 (--p-front-f FWD primer, -...   \n",
       "1    18S rRNA            X  q2-cutadapt v2023.5 (--p-front-f FWD primer, -...   \n",
       "2    12S rRNA            X  q2-cutadapt v2023.5 (--p-front-f FWD primer, -...   \n",
       "\n",
       "  cluster_method pid_clustering taxa_class_method taxa_ref_db  \\\n",
       "0           test           test              test        test   \n",
       "1           test           test              test        test   \n",
       "2           test           test              test        test   \n",
       "\n",
       "                                 code_repo bioproject_accession   sop  \\\n",
       "0  https://github.com/ksilnoaa/seamap-edna                 test  test   \n",
       "1  https://github.com/ksilnoaa/seamap-edna                 test  test   \n",
       "2  https://github.com/ksilnoaa/seamap-edna                 test  test   \n",
       "\n",
       "  identificationReferences  \n",
       "0                     test  \n",
       "1                     test  \n",
       "2                     test  "
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['analysis_data'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ12_test['taxa_class_method'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '12S rRNA','taxa_class_method'].item()\n",
    "occ12_test['taxa_ref_db'] = data['analysis_data'].loc[data['analysis_data']['target_gene'] == '12S rRNA','taxa_ref_db'].item()\n",
    "\n",
    "occ12_test['identificationRemarks'] = occ12_test['taxa_class_method'] +\", confidence (at lowest specified taxon): \"+str(occ12_test['Confidence']) +\", against reference database: \"+occ12_test['taxa_ref_db']\n",
    "#occ12_test.drop(columns=['Confidence','taxa_class_method','taxa_ref_db'],inplace=True)\n",
    "occ12_test.drop(columns=['Confidence','taxa_class_method','taxa_ref_db'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    test, confidence (at lowest specified taxon): ...\n",
       "1    test, confidence (at lowest specified taxon): ...\n",
       "2    test, confidence (at lowest specified taxon): ...\n",
       "3    test, confidence (at lowest specified taxon): ...\n",
       "4    test, confidence (at lowest specified taxon): ...\n",
       "Name: identificationRemarks, dtype: object"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ12_test['identificationRemarks'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### taxonID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ12_test['taxonID'] = 'ASV:'+occ12_test['featureid']\n",
    "occ12_test['basisOfRecord'] = 'MaterialSample'\n",
    "occ12_test['eventID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>eventID</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>rank</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>taxonID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4008ca75e225240b385ec167d0e0c9b4</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Mugiliformes;Mu...</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>29</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occ4008ca7...</td>\n",
       "      <td>Mugil curema</td>\n",
       "      <td>species</td>\n",
       "      <td>Mugil curema</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159416</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Mugiliformes</td>\n",
       "      <td>Mugilidae</td>\n",
       "      <td>Mugil</td>\n",
       "      <td>NCBI:txid48194</td>\n",
       "      <td>test, confidence (at lowest specified taxon): ...</td>\n",
       "      <td>ASV:4008ca75e225240b385ec167d0e0c9b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b8b3d9e1b83477e4927fd8d4c6f1cd2e</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCTA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Salmoniformes;S...</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>4041</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occb8b3d9e...</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>species</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:127182</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Salmoniformes</td>\n",
       "      <td>Salmonidae</td>\n",
       "      <td>Oncorhynchus</td>\n",
       "      <td>NCBI:txid8017</td>\n",
       "      <td>test, confidence (at lowest specified taxon): ...</td>\n",
       "      <td>ASV:b8b3d9e1b83477e4927fd8d4c6f1cd2e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bfe5abf81199e96bbc51918d76b77e8c</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGACCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Lutjaniformes;L...</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>37</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occbfe5abf...</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>species</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159797</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Lutjanidae</td>\n",
       "      <td>Lutjanus</td>\n",
       "      <td>NCBI:txid40503</td>\n",
       "      <td>test, confidence (at lowest specified taxon): ...</td>\n",
       "      <td>ASV:bfe5abf81199e96bbc51918d76b77e8c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>208ffed2e333d449c29995ca62b808cc</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;NA;Opistognathi...</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>47</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occ208ffed...</td>\n",
       "      <td>Lonchopisthus micrognathus</td>\n",
       "      <td>species</td>\n",
       "      <td>Lonchopisthus micrognathus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:281386</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Ovalentaria incertae sedis</td>\n",
       "      <td>Opistognathidae</td>\n",
       "      <td>Lonchopisthus</td>\n",
       "      <td>NCBI:txid1311555</td>\n",
       "      <td>test, confidence (at lowest specified taxon): ...</td>\n",
       "      <td>ASV:208ffed2e333d449c29995ca62b808cc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2f6cd78f1f099544eb8f1b252fba924c</td>\n",
       "      <td>GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGGGGCCCAA...</td>\n",
       "      <td>Eukaryota;Chordata;Actinopteri;Spariformes;Spa...</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>4</td>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A_occ2f6cd78...</td>\n",
       "      <td>Archosargus probatocephalus</td>\n",
       "      <td>species</td>\n",
       "      <td>Archosargus probatocephalus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159238</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Sparidae</td>\n",
       "      <td>Archosargus</td>\n",
       "      <td>NCBI:txid119682</td>\n",
       "      <td>test, confidence (at lowest specified taxon): ...</td>\n",
       "      <td>ASV:2f6cd78f1f099544eb8f1b252fba924c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          featureid  \\\n",
       "0  4008ca75e225240b385ec167d0e0c9b4   \n",
       "1  b8b3d9e1b83477e4927fd8d4c6f1cd2e   \n",
       "2  bfe5abf81199e96bbc51918d76b77e8c   \n",
       "3  208ffed2e333d449c29995ca62b808cc   \n",
       "4  2f6cd78f1f099544eb8f1b252fba924c   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...   \n",
       "1  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCTA...   \n",
       "2  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGACCCAA...   \n",
       "3  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAA...   \n",
       "4  GCCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGGGGCCCAA...   \n",
       "\n",
       "                                            taxonomy  \\\n",
       "0  Eukaryota;Chordata;Actinopteri;Mugiliformes;Mu...   \n",
       "1  Eukaryota;Chordata;Actinopteri;Salmoniformes;S...   \n",
       "2  Eukaryota;Chordata;Actinopteri;Lutjaniformes;L...   \n",
       "3  Eukaryota;Chordata;Actinopteri;NA;Opistognathi...   \n",
       "4  Eukaryota;Chordata;Actinopteri;Spariformes;Spa...   \n",
       "\n",
       "                               eventID  organismQuantity  \\\n",
       "0  SEAMAP2019_Brownsville_Sta103_39m_A                29   \n",
       "1  SEAMAP2019_Brownsville_Sta103_39m_A              4041   \n",
       "2  SEAMAP2019_Brownsville_Sta103_39m_A                37   \n",
       "3  SEAMAP2019_Brownsville_Sta103_39m_A                47   \n",
       "4  SEAMAP2019_Brownsville_Sta103_39m_A                 4   \n",
       "\n",
       "                                        occurrenceID  \\\n",
       "0  SEAMAP2019_Brownsville_Sta103_39m_A_occ4008ca7...   \n",
       "1  SEAMAP2019_Brownsville_Sta103_39m_A_occb8b3d9e...   \n",
       "2  SEAMAP2019_Brownsville_Sta103_39m_A_occbfe5abf...   \n",
       "3  SEAMAP2019_Brownsville_Sta103_39m_A_occ208ffed...   \n",
       "4  SEAMAP2019_Brownsville_Sta103_39m_A_occ2f6cd78...   \n",
       "\n",
       "        verbatimIdentification     rank               scientificName  \\\n",
       "0                 Mugil curema  species                 Mugil curema   \n",
       "1       Oncorhynchus gorbuscha  species       Oncorhynchus gorbuscha   \n",
       "2             Lutjanus griseus  species             Lutjanus griseus   \n",
       "3   Lonchopisthus micrognathus  species   Lonchopisthus micrognathus   \n",
       "4  Archosargus probatocephalus  species  Archosargus probatocephalus   \n",
       "\n",
       "                            scientificNameID   kingdom    phylum      class  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:159416  Animalia  Chordata  Teleostei   \n",
       "1  urn:lsid:marinespecies.org:taxname:127182  Animalia  Chordata  Teleostei   \n",
       "2  urn:lsid:marinespecies.org:taxname:159797  Animalia  Chordata  Teleostei   \n",
       "3  urn:lsid:marinespecies.org:taxname:281386  Animalia  Chordata  Teleostei   \n",
       "4  urn:lsid:marinespecies.org:taxname:159238  Animalia  Chordata  Teleostei   \n",
       "\n",
       "                        order           family          genus  \\\n",
       "0                Mugiliformes        Mugilidae          Mugil   \n",
       "1               Salmoniformes       Salmonidae   Oncorhynchus   \n",
       "2   Eupercaria incertae sedis       Lutjanidae       Lutjanus   \n",
       "3  Ovalentaria incertae sedis  Opistognathidae  Lonchopisthus   \n",
       "4   Eupercaria incertae sedis         Sparidae    Archosargus   \n",
       "\n",
       "     taxonConceptID                              identificationRemarks  \\\n",
       "0    NCBI:txid48194  test, confidence (at lowest specified taxon): ...   \n",
       "1     NCBI:txid8017  test, confidence (at lowest specified taxon): ...   \n",
       "2    NCBI:txid40503  test, confidence (at lowest specified taxon): ...   \n",
       "3  NCBI:txid1311555  test, confidence (at lowest specified taxon): ...   \n",
       "4   NCBI:txid119682  test, confidence (at lowest specified taxon): ...   \n",
       "\n",
       "                                taxonID  \n",
       "0  ASV:4008ca75e225240b385ec167d0e0c9b4  \n",
       "1  ASV:b8b3d9e1b83477e4927fd8d4c6f1cd2e  \n",
       "2  ASV:bfe5abf81199e96bbc51918d76b77e8c  \n",
       "3  ASV:208ffed2e333d449c29995ca62b808cc  \n",
       "4  ASV:2f6cd78f1f099544eb8f1b252fba924c  "
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ12_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge 12S and 18S occurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "occ_dict = dwc_data['occurrence'].to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['eventID', 'occurrenceID', 'basisOfRecord', 'eventDate', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'scientificName', 'scientificNameID', 'taxonID', 'nameAccordingTo', 'taxonConceptID', 'verbatimIdentification', 'identificationRemarks', 'identificationReferences', 'taxonRank', 'organismQuantity', 'organismQuantityType', 'associatedSequences', 'materialSampleID'])"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "basisOfRecord {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': nan, 'Example': nan, 'notes': nan}\n",
      "eventDate {'AOML_term': 'collection_date', 'AOML_file': 'sample_data', 'DwC_definition': nan, 'Example': nan, 'notes': nan}\n",
      "nameAccordingTo {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': nan, 'Example': nan, 'notes': nan}\n",
      "identificationReferences {'AOML_term': 'identificationReferences', 'AOML_file': 'analysis_data', 'DwC_definition': 'A list (concatenated and separated) of references (publication, global unique identifier, URI) used in the Identification. Recommended best practice is to separate the values in a list with space vertical bar space ( | ).', 'Example': nan, 'notes': nan}\n",
      "taxonRank {'AOML_term': 'derived', 'AOML_file': 'tourmaline', 'DwC_definition': 'The taxonomic rank of the most specific name in the dwc:scientificName. Recommended best practice is to use a controlled vocabulary. The taxon ranks of algae, fungi and plants are defined in the International Code of Nomenclature for algae, fungi, and plants (Schenzhen Code Articles H3.2, H4.4 and H.3.1).', 'Example': nan, 'notes': nan}\n",
      "organismQuantityType {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': 'Should be DNA sequence reads', 'Example': nan, 'notes': nan}\n",
      "associatedSequences {'AOML_term': 'associatedSequences', 'AOML_file': 'prep_data', 'DwC_definition': 'A list (concatenated and separated) of identifiers (publication, global unique identifier, URI) of genetic sequence information associated with the Occurrence. Could be used for linking to archived raw barcode reads and/or associated genome sequences, e.g. in a public repository.', 'Example': nan, 'notes': nan}\n",
      "materialSampleID {'AOML_term': 'bioproject_accession', 'AOML_file': 'prep_data', 'DwC_definition': 'An identifier for the MaterialSample (as opposed to a particular digital record of the material sample). Use the biosample ID if one was obtained from a nucleotide archive. In the absence of a persistent global unique identifier, construct one from a combination of identifiers in the record that will most closely make the materialSampleID globally unique.', 'Example': nan, 'notes': nan}\n"
     ]
    }
   ],
   "source": [
    "# check if all occurrence file terms are in sample_data\n",
    "for key in occ_dict.keys():\n",
    "    if key not in occ12_test.columns:\n",
    "        print(key,occ_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AOML_term': 'derived',\n",
       " 'AOML_file': 'tourmaline',\n",
       " 'DwC_definition': 'The scientific name ID of \"Balaenoptera musculus\" as per the WoRMS database.',\n",
       " 'Example': nan,\n",
       " 'notes': nan}"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ_dict['scientificNameID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'eventID': {'AOML_term': 'sample_name',\n",
       "  'AOML_file': 'sample_data',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'occurrenceID': {'AOML_term': 'derived: sample_name, featureid',\n",
       "  'AOML_file': 'tourmaline, sample_data',\n",
       "  'DwC_definition': 'A unique identifier for the occurrence, allowing the same occurrence to be recognized across dataset versions as well as through data downloads and use. May be a global unique identifier or an identifier specific to the data set.',\n",
       "  'Example': nan,\n",
       "  'notes': 'unique: with eventID, sample ID and ASV, generated in script'},\n",
       " 'basisOfRecord': {'AOML_term': 'none',\n",
       "  'AOML_file': 'pipeline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'eventDate': {'AOML_term': 'collection_date',\n",
       "  'AOML_file': 'sample_data',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'kingdom': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'phylum': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'class': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'order': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'family': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'genus': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'scientificName': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': 'Scientific name, preferably as listed in the WoRMS database. This differs from GBIF, where it is recommended to use the taxon name derived from the classification approach used.',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'scientificNameID': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': 'The scientific name ID of \"Balaenoptera musculus\" as per the WoRMS database.',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'taxonID': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': 'For eDNA data, it is recommended to use an MD5 hash of the sequence and prepend it with “ASV:”. See also ',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'nameAccordingTo': {'AOML_term': 'none',\n",
       "  'AOML_file': 'pipeline',\n",
       "  'DwC_definition': nan,\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'taxonConceptID': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': 'The NCBI ID linked to Balaenoptera musculus in the NCBI taxonomic database. Can also be a BIN-ID if BOLD was used for identification, or another ID from a different database.',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'verbatimIdentification': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': 'The name corresponding to the NCBI ID (Balaenoptera musculus) (or other ID). This does not necessarily correspond to the value in scientific name.',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'identificationRemarks': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'analysis_data: taxa_class_method, taxa_ref_db; tourmaline: confidence',\n",
       "  'DwC_definition': 'Specification of taxonomic identification process, ideally including data on applied algorithm and reference database, as well as on level of confidence in the resulting identification.',\n",
       "  'Example': 'RDP annotation confidence (at lowest specified taxon): 0.96, against reference database: GTDB',\n",
       "  'notes': nan},\n",
       " 'identificationReferences': {'AOML_term': 'identificationReferences',\n",
       "  'AOML_file': 'analysis_data',\n",
       "  'DwC_definition': 'A list (concatenated and separated) of references (publication, global unique identifier, URI) used in the Identification. Recommended best practice is to separate the values in a list with space vertical bar space ( | ).',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'taxonRank': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': 'The taxonomic rank of the most specific name in the dwc:scientificName. Recommended best practice is to use a controlled vocabulary. The taxon ranks of algae, fungi and plants are defined in the International Code of Nomenclature for algae, fungi, and plants (Schenzhen Code Articles H3.2, H4.4 and H.3.1).',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'organismQuantity': {'AOML_term': 'derived',\n",
       "  'AOML_file': 'tourmaline',\n",
       "  'DwC_definition': 'Number of reads of this OTU or ASV in the sample.',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'organismQuantityType': {'AOML_term': 'none',\n",
       "  'AOML_file': 'pipeline',\n",
       "  'DwC_definition': 'Should be DNA sequence reads',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'associatedSequences': {'AOML_term': 'associatedSequences',\n",
       "  'AOML_file': 'prep_data',\n",
       "  'DwC_definition': 'A list (concatenated and separated) of identifiers (publication, global unique identifier, URI) of genetic sequence information associated with the Occurrence. Could be used for linking to archived raw barcode reads and/or associated genome sequences, e.g. in a public repository.',\n",
       "  'Example': nan,\n",
       "  'notes': nan},\n",
       " 'materialSampleID': {'AOML_term': 'bioproject_accession',\n",
       "  'AOML_file': 'prep_data',\n",
       "  'DwC_definition': 'An identifier for the MaterialSample (as opposed to a particular digital record of the material sample). Use the biosample ID if one was obtained from a nucleotide archive. In the absence of a persistent global unique identifier, construct one from a combination of identifiers in the record that will most closely make the materialSampleID globally unique.',\n",
       "  'Example': nan,\n",
       "  'notes': nan}}"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occ_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "waterBody {'AOML_term': 'waterBody', 'AOML_file': 'sample_data', 'DwC_definition': 'The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the Getty Thesaurus of Geographic Names.'}\n"
     ]
    }
   ],
   "source": [
    "# check if all occurrence file terms are in sample_data\n",
    "for key in event_dict.keys():\n",
    "    if event_dict[key]['AOML_file'] == 'sample_data':\n",
    "        if event_dict[key]['AOML_term'] not in data['sample_data'].columns:\n",
    "            print(key,event_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AOML_term': 'waterBody',\n",
       " 'AOML_file': 'sample_data',\n",
       " 'DwC_definition': 'The name of the water body in which the dcterms:Location occurs.         Recommended best practice is to use a controlled vocabulary such as the Getty Thesaurus of Geographic Names.'}"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# decide to not add waterBody \n",
    "event_dict.pop('waterBody', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename sample_data columns to fit DwC standard\n",
    "gen = (x for x in event_dict.keys() if event_dict[x]['AOML_file'] == 'sample_data')\n",
    "rename_dict = {}\n",
    "for x in gen:\n",
    "    #print(x)\n",
    "    rename_dict[event_dict[x]['AOML_term']] = x\n",
    "\n",
    "event_sample = data['sample_data'].rename(columns=rename_dict)\n",
    "event_sample = event_sample.drop(columns=[col for col in event_sample if col not in rename_dict.values()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>parentEventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>maximumDepthInMeters</th>\n",
       "      <th>locality</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>samplingProtocol</th>\n",
       "      <th>eventRemarks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_A</td>\n",
       "      <td>Bot</td>\n",
       "      <td>2019-04-23T22:05:07-05:00</td>\n",
       "      <td>38.8 m</td>\n",
       "      <td>Gulf of Mexico: Brownsville</td>\n",
       "      <td>26.0262</td>\n",
       "      <td>-96.8373</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>not applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_B</td>\n",
       "      <td>Bot</td>\n",
       "      <td>2019-04-23T22:05:07-05:00</td>\n",
       "      <td>38.8 m</td>\n",
       "      <td>Gulf of Mexico: Brownsville</td>\n",
       "      <td>26.0262</td>\n",
       "      <td>-96.8373</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>not applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_39m_C</td>\n",
       "      <td>Bot</td>\n",
       "      <td>2019-04-23T22:05:07-05:00</td>\n",
       "      <td>38.8 m</td>\n",
       "      <td>Gulf of Mexico: Brownsville</td>\n",
       "      <td>26.0262</td>\n",
       "      <td>-96.8373</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>PM cast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_4m_A</td>\n",
       "      <td>Sur</td>\n",
       "      <td>2019-04-23T22:05:07-05:00</td>\n",
       "      <td>3.7 m</td>\n",
       "      <td>Gulf of Mexico: Brownsville</td>\n",
       "      <td>26.0262</td>\n",
       "      <td>-96.8373</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>not applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SEAMAP2019_Brownsville_Sta103_4m_B</td>\n",
       "      <td>Sur</td>\n",
       "      <td>2019-04-23T22:05:07-05:00</td>\n",
       "      <td>3.7 m</td>\n",
       "      <td>Gulf of Mexico: Brownsville</td>\n",
       "      <td>26.0262</td>\n",
       "      <td>-96.8373</td>\n",
       "      <td>CTD rosette</td>\n",
       "      <td>not applicable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               eventID parentEventID  \\\n",
       "0  SEAMAP2019_Brownsville_Sta103_39m_A           Bot   \n",
       "1  SEAMAP2019_Brownsville_Sta103_39m_B           Bot   \n",
       "2  SEAMAP2019_Brownsville_Sta103_39m_C           Bot   \n",
       "3   SEAMAP2019_Brownsville_Sta103_4m_A           Sur   \n",
       "4   SEAMAP2019_Brownsville_Sta103_4m_B           Sur   \n",
       "\n",
       "                   eventDate maximumDepthInMeters  \\\n",
       "0  2019-04-23T22:05:07-05:00               38.8 m   \n",
       "1  2019-04-23T22:05:07-05:00               38.8 m   \n",
       "2  2019-04-23T22:05:07-05:00               38.8 m   \n",
       "3  2019-04-23T22:05:07-05:00                3.7 m   \n",
       "4  2019-04-23T22:05:07-05:00                3.7 m   \n",
       "\n",
       "                      locality decimalLongitude decimalLongitude  \\\n",
       "0  Gulf of Mexico: Brownsville          26.0262         -96.8373   \n",
       "1  Gulf of Mexico: Brownsville          26.0262         -96.8373   \n",
       "2  Gulf of Mexico: Brownsville          26.0262         -96.8373   \n",
       "3  Gulf of Mexico: Brownsville          26.0262         -96.8373   \n",
       "4  Gulf of Mexico: Brownsville          26.0262         -96.8373   \n",
       "\n",
       "  samplingProtocol    eventRemarks  \n",
       "0      CTD rosette  not applicable  \n",
       "1      CTD rosette  not applicable  \n",
       "2      CTD rosette         PM cast  \n",
       "3      CTD rosette  not applicable  \n",
       "4      CTD rosette  not applicable  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "event_sample.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sampleSizeValue {'AOML_term': 'derived', 'AOML_file': 'tourmaline', 'DwC_definition': 'In the context of eDNA data, `sampleSizeValue` should be the total number of reads for a given sample.'}\n",
      "sampleSizeUnit {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': 'Should be DNA sequence reads'}\n",
      "locationID {'AOML_term': 'optional', 'AOML_file': '?', 'DwC_definition': 'An identifier for the set of dcterms:Location information. May be a global unique identifier or an identifier specific to the data set.'}\n",
      "geodeticDatum {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': 'The ellipsoid, geodetic datum, or spatial reference system (SRS) upon which the geographic coordinates given in dwc:decimalLatitude and dwc:decimalLongitude are based.'}\n",
      "countryCode {'AOML_term': 'none', 'AOML_file': 'pipeline', 'DwC_definition': nan}\n",
      "datasetID {'AOML_term': 'project_id', 'AOML_file': 'project_data', 'DwC_definition': 'An identifier for the set of data. May be a global unique identifier or an identifier specific to a collection or institution.'}\n"
     ]
    }
   ],
   "source": [
    "for key in event_dict.keys():\n",
    "    if event_dict[key]['AOML_file'] != 'sample_data':\n",
    "        print(key,event_dict[key])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the context of eDNA data, `sampleSizeValue` should be the total number of reads for a given sample."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is where you would add in that extra info. For now, just leave it out."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# boneyard\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I also wanted to persist the original name from the NCBI taxonomy database into the Darwin Core-converted data set. To do this, I queried the database based on the name in the original data to obtain its taxonomic ID number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get set up to query NCBI taxonomy \n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "# ----- Insert your email here -----\n",
    "Entrez.email = 'ksil91@gmail.com'\n",
    "# ----------------------------------\n",
    "\n",
    "# Get list of all databases available through this tool\n",
    "record = Entrez.read(Entrez.einfo())\n",
    "all_dbs = record['DbList']\n",
    "all_dbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ncbi_txid_from_name_parallel(all_names,n_proc=0):\n",
    "    all_names = all_names.unique()\n",
    "    queue = mp.Queue()\n",
    "    if n_proc == 0:\n",
    "    # create as many processes as there are CPUs on your machine\n",
    "        num_processes = mp.cpu_count()\n",
    "    else:\n",
    "        num_processes = n_proc\n",
    "        \n",
    "    # calculate the chunk size as an integer\n",
    "    chunk_size = int(len(all_names)/num_processes)\n",
    "    procs = []\n",
    "    for job in range(num_processes):\n",
    "        start = job * chunk_size\n",
    "        end = start + chunk_size\n",
    "        name_chunk = all_names[start:end]\n",
    "        proc = mp.Process(\n",
    "            target=get_ncbi_txid_from_name,\n",
    "            args=(name_chunk,queue)\n",
    "        )\n",
    "        procs.append(proc)\n",
    "        proc.start()\n",
    "    \n",
    "    names_dict = {}\n",
    "    for _ in procs:\n",
    "        names_dict.update(queue.get())\n",
    "    \n",
    "    #new_df = queue.get()\n",
    "    \n",
    "    for proc in procs:\n",
    "        proc.join()\n",
    "    \n",
    "    return names_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get NCBI taxIDs for each name in dataset ---- TAKES ~ 2 MINUTES FOR 300 RECORDS\n",
    "def get_ncbi_txid_from_name(names,queue):\n",
    "    \n",
    "    name_ncbiid_dict = {}\n",
    "    for name in names:\n",
    "        if name not in name_ncbiid_dict.keys():\n",
    "            taxon2 = '\"' + name + '\"'\n",
    "            handle = Entrez.esearch(db='taxonomy', retmax=10, term=taxon2)\n",
    "            record = Entrez.read(handle)\n",
    "            if not record['IdList'] or not record['IdList'][0]:\n",
    "                print(name+\": not found\")\n",
    "                name_ncbiid_dict[name] = \"None\"\n",
    "            else:\n",
    "                name_ncbiid_dict[name] = record['IdList'][0]\n",
    "            handle.close()\n",
    "    queue.put(name_ncbiid_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "n =worms_12s['verbatimIdentification'][0:10]\n",
    "new = get_ncbi_txid_from_name_parallel(n,7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                 Clupeidae\n",
       "1         Eucinostomus gula\n",
       "2    Oncorhynchus gorbuscha\n",
       "3          Lutjanus griseus\n",
       "4         Eucinostomus gula\n",
       "5          Cheilopogon agoo\n",
       "6    Oncorhynchus gorbuscha\n",
       "7    Oncorhynchus gorbuscha\n",
       "8          Scomber scombrus\n",
       "9        Haemulon plumierii\n",
       "Name: verbatimIdentification, dtype: object"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Oncorhynchus gorbuscha': '8017',\n",
       " 'Scomber scombrus': '13677',\n",
       " 'Eucinostomus gula': '435273',\n",
       " 'Haemulon plumierii': '334415',\n",
       " 'Clupeidae': '55118',\n",
       " 'Cheilopogon agoo': '123223',\n",
       " 'Lutjanus griseus': '40503'}"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                 Clupeidae\n",
       "1         Eucinostomus gula\n",
       "2    Oncorhynchus gorbuscha\n",
       "3          Lutjanus griseus\n",
       "4         Eucinostomus gula\n",
       "5          Cheilopogon agoo\n",
       "6    Oncorhynchus gorbuscha\n",
       "7    Oncorhynchus gorbuscha\n",
       "8          Scomber scombrus\n",
       "9        Haemulon plumierii\n",
       "Name: verbatimIdentification, dtype: object"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Oncorhynchus gorbuscha': '8017',\n",
       " 'Lutjanus griseus': '40503',\n",
       " 'Scomber scombrus': '13677',\n",
       " 'Haemulon plumierii': '334415',\n",
       " 'Eucinostomus gula': '435273',\n",
       " 'Cheilopogon agoo': '123223',\n",
       " 'Clupeidae': '55118'}"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get NCBI taxIDs for each name in dataset ---- TAKES ~ 2 MINUTES FOR 300 RECORDS\n",
    "\n",
    "name_ncbiid_dict = {}\n",
    "\n",
    "for name in worms_12s['verbatimIdentification']:\n",
    "    if name not in name_ncbiid_dict.keys():\n",
    "        taxon2 = '\"' + name + '\"'\n",
    "        handle = Entrez.esearch(db='taxonomy', retmax=10, term=taxon2)\n",
    "        record = Entrez.read(handle)\n",
    "        if not record['IdList'] or not record['IdList'][0]:\n",
    "            print(name+\": not found\")\n",
    "            name_ncbiid_dict[name] = \"None\"\n",
    "        else:\n",
    "            name_ncbiid_dict[name] = record['IdList'][0]\n",
    "        handle.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look up the not found names by hand to get taxid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'None'"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_ncbiid_dict['Lobianchia sp. CBM:ZF:14789'] = '2057084'\n",
    "name_ncbiid_dict['Brotula sp. CBM:ZF:20276']= '2768194'\n",
    "name_ncbiid_dict['Scopelarchus sp. CBM:DNA No. 2000-014']= '2608515'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Clupeidae',\n",
       " 'Eucinostomus gula',\n",
       " 'Oncorhynchus gorbuscha',\n",
       " 'Lutjanus griseus',\n",
       " 'Cheilopogon agoo',\n",
       " 'Scomber scombrus',\n",
       " 'Haemulon plumierii',\n",
       " 'Archosargus probatocephalus',\n",
       " 'Actinopteri',\n",
       " 'Carangidae']"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(name_ncbiid_dict.keys())[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'55118'"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_ncbiid_dict['Clupeidae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>asv</th>\n",
       "      <th>rank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonConceptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>951a3746cb4c488d657fe3c64bdd3d75</td>\n",
       "      <td>family</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:125464</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Clupeiformes</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBI:txid55118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>f278a917dedcc2e0434be6ecb605a642</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>1d0fcad8f69709ed9ea8dd6e6a3c627a</td>\n",
       "      <td>species</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:127182</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Salmoniformes</td>\n",
       "      <td>Salmonidae</td>\n",
       "      <td>Oncorhynchus</td>\n",
       "      <td>NCBI:txid8017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>cc6451b251afdd97ef4f11e47abc53a7</td>\n",
       "      <td>species</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159797</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Lutjanidae</td>\n",
       "      <td>Lutjanus</td>\n",
       "      <td>NCBI:txid40503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>d16bc4ff5741c0ba4a5d0bb2e21e5589</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   verbatimIdentification                               asv     rank  \\\n",
       "0               Clupeidae  951a3746cb4c488d657fe3c64bdd3d75   family   \n",
       "1       Eucinostomus gula  f278a917dedcc2e0434be6ecb605a642  species   \n",
       "2  Oncorhynchus gorbuscha  1d0fcad8f69709ed9ea8dd6e6a3c627a  species   \n",
       "3        Lutjanus griseus  cc6451b251afdd97ef4f11e47abc53a7  species   \n",
       "4       Eucinostomus gula  d16bc4ff5741c0ba4a5d0bb2e21e5589  species   \n",
       "\n",
       "                 old name          scientificName  \\\n",
       "0               Clupeidae               Clupeidae   \n",
       "1       Eucinostomus gula       Eucinostomus gula   \n",
       "2  Oncorhynchus gorbuscha  Oncorhynchus gorbuscha   \n",
       "3        Lutjanus griseus        Lutjanus griseus   \n",
       "4       Eucinostomus gula       Eucinostomus gula   \n",
       "\n",
       "                            scientificNameID   kingdom    phylum      class  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:125464  Animalia  Chordata  Teleostei   \n",
       "1  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "2  urn:lsid:marinespecies.org:taxname:127182  Animalia  Chordata  Teleostei   \n",
       "3  urn:lsid:marinespecies.org:taxname:159797  Animalia  Chordata  Teleostei   \n",
       "4  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "\n",
       "                       order      family         genus   taxonConceptID  \n",
       "0               Clupeiformes   Clupeidae          None   NCBI:txid55118  \n",
       "1  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  \n",
       "2              Salmoniformes  Salmonidae  Oncorhynchus    NCBI:txid8017  \n",
       "3  Eupercaria incertae sedis  Lutjanidae      Lutjanus   NCBI:txid40503  \n",
       "4  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  "
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add NCBI taxonomy IDs under taxonConceptID\n",
    "\n",
    "# Map indicators that say no taxonomy was assigned to empty strings\n",
    "#name_ncbiid_dict['unassigned'], name_ncbiid_dict['s_'], name_ncbiid_dict['no_hit'], name_ncbiid_dict['unknown'], name_ncbiid_dict['g_'] = '', '', '', '', ''\n",
    "\n",
    "# Create column\n",
    "worms_12s['taxonConceptID']  = worms_12s['verbatimIdentification'].copy()\n",
    "worms_12s['taxonConceptID'].replace(name_ncbiid_dict, inplace=True)\n",
    "\n",
    "# Add remainder of text and clean\n",
    "worms_12s['taxonConceptID'] = 'NCBI:txid' + worms_12s['taxonConceptID']\n",
    "worms_12s['taxonConceptID'].replace('NCBI:txid', '', inplace=True)\n",
    "worms_12s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>verbatimIdentification</th>\n",
       "      <th>asv</th>\n",
       "      <th>rank</th>\n",
       "      <th>old name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>taxonConceptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>951a3746cb4c488d657fe3c64bdd3d75</td>\n",
       "      <td>family</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:125464</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Clupeiformes</td>\n",
       "      <td>Clupeidae</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBI:txid55118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>f278a917dedcc2e0434be6ecb605a642</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>1d0fcad8f69709ed9ea8dd6e6a3c627a</td>\n",
       "      <td>species</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>Oncorhynchus gorbuscha</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:127182</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Salmoniformes</td>\n",
       "      <td>Salmonidae</td>\n",
       "      <td>Oncorhynchus</td>\n",
       "      <td>NCBI:txid8017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>cc6451b251afdd97ef4f11e47abc53a7</td>\n",
       "      <td>species</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>Lutjanus griseus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159797</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Lutjanidae</td>\n",
       "      <td>Lutjanus</td>\n",
       "      <td>NCBI:txid40503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>d16bc4ff5741c0ba4a5d0bb2e21e5589</td>\n",
       "      <td>species</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>Eucinostomus gula</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:159733</td>\n",
       "      <td>Animalia</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Teleostei</td>\n",
       "      <td>Eupercaria incertae sedis</td>\n",
       "      <td>Gerreidae</td>\n",
       "      <td>Eucinostomus</td>\n",
       "      <td>NCBI:txid435273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   verbatimIdentification                               asv     rank  \\\n",
       "0               Clupeidae  951a3746cb4c488d657fe3c64bdd3d75   family   \n",
       "1       Eucinostomus gula  f278a917dedcc2e0434be6ecb605a642  species   \n",
       "2  Oncorhynchus gorbuscha  1d0fcad8f69709ed9ea8dd6e6a3c627a  species   \n",
       "3        Lutjanus griseus  cc6451b251afdd97ef4f11e47abc53a7  species   \n",
       "4       Eucinostomus gula  d16bc4ff5741c0ba4a5d0bb2e21e5589  species   \n",
       "\n",
       "                 old name          scientificName  \\\n",
       "0               Clupeidae               Clupeidae   \n",
       "1       Eucinostomus gula       Eucinostomus gula   \n",
       "2  Oncorhynchus gorbuscha  Oncorhynchus gorbuscha   \n",
       "3        Lutjanus griseus        Lutjanus griseus   \n",
       "4       Eucinostomus gula       Eucinostomus gula   \n",
       "\n",
       "                            scientificNameID   kingdom    phylum      class  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:125464  Animalia  Chordata  Teleostei   \n",
       "1  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "2  urn:lsid:marinespecies.org:taxname:127182  Animalia  Chordata  Teleostei   \n",
       "3  urn:lsid:marinespecies.org:taxname:159797  Animalia  Chordata  Teleostei   \n",
       "4  urn:lsid:marinespecies.org:taxname:159733  Animalia  Chordata  Teleostei   \n",
       "\n",
       "                       order      family         genus   taxonConceptID  \n",
       "0               Clupeiformes   Clupeidae          None   NCBI:txid55118  \n",
       "1  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  \n",
       "2              Salmoniformes  Salmonidae  Oncorhynchus    NCBI:txid8017  \n",
       "3  Eupercaria incertae sedis  Lutjanidae      Lutjanus   NCBI:txid40503  \n",
       "4  Eupercaria incertae sedis   Gerreidae  Eucinostomus  NCBI:txid435273  "
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "worms_12s.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Worms "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use SOAP access to worms API. See examples for python [here](https://marinespecies.org/aphia.php?p=webservice&type=python)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "array_of_results_array = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "531014 Buccinum fusiforme Buccinum\n",
      "510389 Buccinum fusiforme Buccinum\n",
      "141433 Abra alba Abra\n",
      "1492457 Randia Randia\n",
      "1442158 Randomia Randomia\n"
     ]
    }
   ],
   "source": [
    "from suds import null, WebFault\n",
    "from suds.client import Client\n",
    "cl = Client('https://marinespecies.org/aphia.php?p=soap&wsdl=1')\n",
    "\n",
    "scinames = cl.factory.create('scientificnames')\n",
    "scinames[\"_arrayType\"] = \"string[]\"\n",
    "scinames[\"scientificname\"] = [\"Buccinum fusiforme\", \"Abra alba\",\"random\"]\n",
    "\n",
    "# like = wildcard after name\n",
    "array_of_results_array = cl.service.matchAphiaRecordsByNames(scinames, like=False, fuzzy=False, marine_only=False)\n",
    "for results_array in array_of_results_array:\n",
    "    for aphia_object in results_array:\n",
    "        print('%s %s %s' % (aphia_object.AphiaID, aphia_object.scientificname, aphia_object.genus))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "urn:lsid:marinespecies.org:taxname:141433 Abra alba Abra\n"
     ]
    }
   ],
   "source": [
    "for results_array in array_of_results_array:\n",
    "    for aphia_object in results_array:\n",
    "        if aphia_object.status == \"accepted\":\n",
    "            print('%s %s %s' % (aphia_object.lsid, aphia_object.scientificname, aphia_object.genus))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use [pyworms](https://pyworms.readthedocs.io/en/latest/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 1]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = [2]\n",
    "y = [1]\n",
    "x + y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "s_match = pyworms.aphiaRecordsByName(\"Abra alba\",like=False,marine_only=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AphiaID': 141433,\n",
       " 'url': 'https://www.marinespecies.org/aphia.php?p=taxdetails&id=141433',\n",
       " 'scientificname': 'Abra alba',\n",
       " 'authority': '(W. Wood, 1802)',\n",
       " 'status': 'accepted',\n",
       " 'unacceptreason': None,\n",
       " 'taxonRankID': 220,\n",
       " 'rank': 'Species',\n",
       " 'valid_AphiaID': 141433,\n",
       " 'valid_name': 'Abra alba',\n",
       " 'valid_authority': '(W. Wood, 1802)',\n",
       " 'parentNameUsageID': 138474,\n",
       " 'kingdom': 'Animalia',\n",
       " 'phylum': 'Mollusca',\n",
       " 'class': 'Bivalvia',\n",
       " 'order': 'Cardiida',\n",
       " 'family': 'Semelidae',\n",
       " 'genus': 'Abra',\n",
       " 'citation': 'MolluscaBase eds. (2023). MolluscaBase. Abra alba (W. Wood, 1802). Accessed through: World Register of Marine Species at: https://www.marinespecies.org/aphia.php?p=taxdetails&id=141433 on 2023-09-13',\n",
       " 'lsid': 'urn:lsid:marinespecies.org:taxname:141433',\n",
       " 'isMarine': 1,\n",
       " 'isBrackish': None,\n",
       " 'isFreshwater': None,\n",
       " 'isTerrestrial': None,\n",
       " 'isExtinct': None,\n",
       " 'match_type': 'exact',\n",
       " 'modified': '2010-09-23T10:34:21.967Z'}"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_match[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 710,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data = [{'ASV': '1', 'species': \"Abra alba\", 'genus': \"Abra\"},\n",
    "        {'ASV': '2',  'species': \"Caldanaerobacter fake\", 'genus': \"Caldanaerobacter\"},\n",
    "        {'ASV': '3',  'species': \"fake\",  'genus': \"fake\" },\n",
    "       {'ASV': '4',  'species': \"NaN\",  'genus': \"Caldanaerobacter\"}]\n",
    "df = pd.DataFrame.from_dict(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 711,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Abra alba</td>\n",
       "      <td>Abra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Caldanaerobacter fake</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>fake</td>\n",
       "      <td>fake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASV                species             genus\n",
       "0   1              Abra alba              Abra\n",
       "1   2  Caldanaerobacter fake  Caldanaerobacter\n",
       "2   3                   fake              fake\n",
       "3   4                    NaN  Caldanaerobacter"
      ]
     },
     "execution_count": 711,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2]"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(tax_df, ordered_rank_columns, ASV_column_name = \"NaN\", like=False, marine_only=False,verbose=False, n_proc=0):\n",
    "# don't forget to import\n",
    "    import pandas as pd\n",
    "    import multiprocessing\n",
    "    from functools import partial\n",
    "\n",
    "    if n_proc == 0:\n",
    "    # create as many processes as there are CPUs on your machine\n",
    "        num_processes = multiprocessing.cpu_count()\n",
    "    else:\n",
    "        num_processes = n_proc\n",
    "\n",
    "    # calculate the chunk size as an integer\n",
    "    chunk_size = int(df.shape[0]/num_processes)\n",
    "\n",
    "    # this solution was reworked from the above link.\n",
    "    # will work even if the length of the dataframe is not evenly divisible by num_processes\n",
    "    chunks = [tax_df.iloc[tax_df.index[i:i + chunk_size]] for i in range(0, tax_df.shape[0], chunk_size)]\n",
    "\n",
    "    # create our pool with `num_processes` processes\n",
    "    pool = multiprocessing.Pool(processes=num_processes)\n",
    "\n",
    "    # apply our function to each chunk in the list\n",
    "    func = partial(get_worms_from_scientific_name, ordered_rank_columns=ordered_rank_columns,ASV_column_name=ASV_column_name,like=like,marine_only=marine_only)\n",
    "    result = pool.map(func,chunks)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "    return result\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(df,['genus','species'],n_proc=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Abra alba</td>\n",
       "      <td>Abra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Caldanaerobacter fake</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>fake</td>\n",
       "      <td>fake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ASV                species             genus\n",
       "0   1              Abra alba              Abra\n",
       "1   2  Caldanaerobacter fake  Caldanaerobacter\n",
       "2   3                   fake              fake\n",
       "3   4                    NaN  Caldanaerobacter"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asv</th>\n",
       "      <th>rank</th>\n",
       "      <th>old name</th>\n",
       "      <th>new name</th>\n",
       "      <th>lsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>genus</td>\n",
       "      <td>Abra</td>\n",
       "      <td>Abra</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:138474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>genus</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:571044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>species</td>\n",
       "      <td>fake</td>\n",
       "      <td>No match</td>\n",
       "      <td>No match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>genus</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "      <td>Caldanaerobacter</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:571044</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  asv     rank          old name          new name  \\\n",
       "0   1    genus              Abra              Abra   \n",
       "1   2    genus  Caldanaerobacter  Caldanaerobacter   \n",
       "2   3  species              fake          No match   \n",
       "3   4    genus  Caldanaerobacter  Caldanaerobacter   \n",
       "\n",
       "                                        lsid  \n",
       "0  urn:lsid:marinespecies.org:taxname:138474  \n",
       "1  urn:lsid:marinespecies.org:taxname:571044  \n",
       "2                                   No match  \n",
       "3  urn:lsid:marinespecies.org:taxname:571044  "
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_worms_from_scientific_name(df,['genus','species'],ASV_column_name=\"ASV\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_worms_from_scientific_name(tax_df, ordered_rank_columns, ASV_column_name = \"NaN\", like=False, marine_only=False,verbose=False):\n",
    "    import time\n",
    "    matches = []\n",
    "    for index, row in df.iterrows():\n",
    "        asv = row[ASV_column_name]\n",
    "        for i in ordered_rank_columns:\n",
    "            rank = i\n",
    "            old_name = row[i]\n",
    "            row_data = {'asv':asv,'rank': rank, 'old name': old_name}\n",
    "            if pd.isna(old_name):\n",
    "                continue \n",
    "            else:\n",
    "                s_match = pyworms.aphiaRecordsByName(old_name,like=like,marine_only=marine_only)\n",
    "                #time.sleep(1)\n",
    "                if s_match == None:\n",
    "                    row_data['new name'] = \"No match\"\n",
    "                    row_data['lsid'] = \"No match\"\n",
    "                    continue\n",
    "                elif len(s_match) > 1:\n",
    "                    mult = []\n",
    "                    for m in s_match:\n",
    "                        if m['status'] == 'accepted':\n",
    "                            mult = mult + [m]\n",
    "                    if mult > 1:\n",
    "                        row_data['new name'] = \"Multiple matches\"\n",
    "                        row_data['lsid'] = \"Multiple matches\"\n",
    "                    elif len(mult) < 1:\n",
    "                        row_data['new name'] = \"Unaccepted\"\n",
    "                        row_data['lsid'] = \"Unaccepted\"\n",
    "                elif len(s_match) == 1:\n",
    "                    if s_match[0]['status'] == 'accepted':\n",
    "                        row_data['new name'] = s_match[0]['scientificname']\n",
    "                        row_data['lsid'] = s_match[0]['lsid']\n",
    "                        break\n",
    "                    elif s_match[0]['status'] == 'unaccepted':\n",
    "                        row_data['new name'] = \"Unaccepted\"\n",
    "                        row_data['lsid'] = \"Unaccepted\"\n",
    "        matches += [row_data]\n",
    "    matches = pd.DataFrame.from_dict(matches)\n",
    "    return matches\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AphiaID': 105,\n",
       " 'url': 'https://www.marinespecies.org/aphia.php?p=taxdetails&id=105',\n",
       " 'scientificname': 'Bivalvia',\n",
       " 'authority': 'Linnaeus, 1758',\n",
       " 'status': 'accepted',\n",
       " 'unacceptreason': None,\n",
       " 'taxonRankID': 60,\n",
       " 'rank': 'Class',\n",
       " 'valid_AphiaID': 105,\n",
       " 'valid_name': 'Bivalvia',\n",
       " 'valid_authority': 'Linnaeus, 1758',\n",
       " 'parentNameUsageID': 51,\n",
       " 'kingdom': 'Animalia',\n",
       " 'phylum': 'Mollusca',\n",
       " 'class': 'Bivalvia',\n",
       " 'order': None,\n",
       " 'family': None,\n",
       " 'genus': None,\n",
       " 'citation': 'WoRMS (2023). Bivalvia. Accessed at: https://www.marinespecies.org/aphia.php?p=taxdetails&id=105 on 2023-09-13',\n",
       " 'lsid': 'urn:lsid:marinespecies.org:taxname:105',\n",
       " 'isMarine': 1,\n",
       " 'isBrackish': 1,\n",
       " 'isFreshwater': 1,\n",
       " 'isTerrestrial': 0,\n",
       " 'isExtinct': 0,\n",
       " 'match_type': 'exact',\n",
       " 'modified': '2019-08-08T17:02:58.460Z'}"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyworms.aphiaRecordByExternalID(6544,\"ncbi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## sampleSizeValue\n",
    "\n",
    "count_by_seq = plate.groupby('Sequence_ID', as_index=False)['Reads'].sum()\n",
    "occ = occ.merge(count_by_seq, how='left', left_on='eventID', right_on='Sequence_ID')\n",
    "occ.drop(columns='Sequence_ID', inplace=True)\n",
    "occ.rename(columns={'Reads':'sampleSizeValue'}, inplace=True)\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "## eventID - the sample_name column in the plate dataframe uniquely identifies a water sample\n",
    "\n",
    "aterm = event_dict['eventID']['AOML_term']\n",
    "afile = event_dict['eventID']['AOML_file']\n",
    "\n",
    "event_df = pd.DataFrame({'eventID':data[afile][aterm]})\n",
    "\n",
    "for key in event_dict.keys():\n",
    "    \n",
    "    \n",
    "    res = pd.Series(data[params[sheet]].columns[data[params[sheet]].isnull().any()].tolist(),\n",
    "                name=sheet)\n",
    "    some=pd.concat([some,res],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2/20/14 15:33</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>4/21/16 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>4/21/16 14:39</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID      eventDate  decimalLatitude  decimalLongitude  \\\n",
       "0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848   \n",
       "1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848   \n",
       "2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848   \n",
       "3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846   \n",
       "4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846   \n",
       "\n",
       "                 env_broad_scale                env_local_scale  \\\n",
       "0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse                               pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "1                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "2                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "3                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "4                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "\n",
       "                                                 sop            seq_meth  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "\n",
       "  samp_vol_we_dna_ext                             nucl_acid_ext  \\\n",
       "0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "\n",
       "                              nucl_acid_amp  \n",
       "0  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "1  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "2  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "3  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "4  dx.doi.org/10.17504/protocols.io.n2vdge6  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merge with plate_meta to obtain columns that can be added directly from metadata\n",
    "\n",
    "metadata_cols = [\n",
    "    'seqID',\n",
    "    'eventDate', \n",
    "    'decimalLatitude', \n",
    "    'decimalLongitude',\n",
    "    'env_broad_scale',\n",
    "    'env_local_scale',\n",
    "    'env_medium',\n",
    "    'target_gene',\n",
    "    'primer_sequence_forward',\n",
    "    'primer_sequence_reverse',\n",
    "    'pcr_primer_name_forward',\n",
    "    'pcr_primer_name_reverse',\n",
    "    'pcr_primer_reference',\n",
    "    'sop',\n",
    "    'seq_meth',\n",
    "    'samp_vol_we_dna_ext',\n",
    "    'nucl_acid_ext', \n",
    "    'nucl_acid_amp',\n",
    "]\n",
    "\n",
    "dwc_cols = metadata_cols.copy()\n",
    "dwc_cols[0] = 'eventID'\n",
    "\n",
    "occ = occ.merge(meta[metadata_cols], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "occ.columns = dwc_cols\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \\\n",
       "0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse                               pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "1                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "2                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "3                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "4                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "\n",
       "                                                 sop            seq_meth  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   \n",
       "\n",
       "  samp_vol_we_dna_ext                             nucl_acid_ext  \\\n",
       "0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "\n",
       "                              nucl_acid_amp  \n",
       "0  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "1  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "2  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "3  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "4  dx.doi.org/10.17504/protocols.io.n2vdge6  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Format eventDate\n",
    "\n",
    "pst = pytz.timezone('America/Los_Angeles')\n",
    "eventDate = [pst.localize(datetime.strptime(dt, '%m/%d/%y %H:%M')).isoformat() for dt in occ['eventDate']]\n",
    "occ['eventDate'] = eventDate\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dx.doi.org/10.17504/protocols.io.xjufknw | dx.doi.org/10.17504/protocols.io.n2vdge6 | https://github.com/MBARI-BOG/BOG-Banzai-Dada2-Pipeline'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Clean sop\n",
    "\n",
    "occ['sop'] = occ['sop'].str.replace('|', ' | ', regex=False)\n",
    "occ['sop'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Update seq_meth\n",
    "\n",
    "occ['seq_meth'] = 'Illumina MiSeq 2x250'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>sop</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse pcr_primer_name_forward  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   \n",
       "\n",
       "  pcr_primer_name_reverse                               pcr_primer_reference  \\\n",
       "0                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "1                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "2                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "3                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "4                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   \n",
       "\n",
       "                                                 sop              seq_meth  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "\n",
       "  samp_vol_we_dna_ext                             nucl_acid_ext  \\\n",
       "0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "\n",
       "                              nucl_acid_amp  \n",
       "0  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "1  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "2  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "3  dx.doi.org/10.17504/protocols.io.n2vdge6  \n",
       "4  dx.doi.org/10.17504/protocols.io.n2vdge6  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Change column names as needed\n",
    "\n",
    "occ = occ.rename(columns = {'primer_sequence_forward':'pcr_primer_forward',\n",
    "                            'primer_sequence_reverse':'pcr_primer_reverse'})\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>sop</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>otu_class_appr</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw | dx....</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "                                                 sop              seq_meth  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  Illumina MiSeq 2x250   \n",
       "\n",
       "  samp_vol_we_dna_ext                             nucl_acid_ext  \\\n",
       "0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "\n",
       "                              nucl_acid_amp target_subfragment lib_layout  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "1  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "2  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "3  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "4  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "\n",
       "     otu_class_appr                                  otu_seq_comp_appr  \\\n",
       "0  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "1  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "2  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "3  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "4  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "\n",
       "           otu_db  \n",
       "0  Genbank nr;221  \n",
       "1  Genbank nr;221  \n",
       "2  Genbank nr;221  \n",
       "3  Genbank nr;221  \n",
       "4  Genbank nr;221  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add extension terms that weren't in metadata file (obtained by asking data provider)\n",
    "\n",
    "occ['target_subfragment'] = 'V9'\n",
    "occ['lib_layout'] = 'paired'\n",
    "occ['otu_class_appr'] = 'dada2;1.14.0;ASV'\n",
    "occ['otu_seq_comp_appr'] = 'blast;2.9.0+;80% identity;e-value cutoff: 0.00001 | MEGAN6;6.18.5;bitscore:100:2%'\n",
    "occ['otu_db'] = 'Genbank nr;221'\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>otu_class_appr</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_1_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_2_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_3_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_1_S_occ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina MiSeq 2x250</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_2_S_occ1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...              seq_meth  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Illumina MiSeq 2x250   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Illumina MiSeq 2x250   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Illumina MiSeq 2x250   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Illumina MiSeq 2x250   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Illumina MiSeq 2x250   \n",
       "\n",
       "  samp_vol_we_dna_ext                             nucl_acid_ext  \\\n",
       "0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "\n",
       "                              nucl_acid_amp target_subfragment lib_layout  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "1  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "2  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "3  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "4  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "\n",
       "     otu_class_appr                                  otu_seq_comp_appr  \\\n",
       "0  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "1  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "2  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "3  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "4  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "\n",
       "           otu_db               occurrenceID  \n",
       "0  Genbank nr;221  05114c01_12_edna_1_S_occ1  \n",
       "1  Genbank nr;221  05114c01_12_edna_2_S_occ1  \n",
       "2  Genbank nr;221  05114c01_12_edna_3_S_occ1  \n",
       "3  Genbank nr;221  11216c01_12_edna_1_S_occ1  \n",
       "4  Genbank nr;221  11216c01_12_edna_2_S_occ1  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Create an occurrenceID that will uniquely identify each ASV observed within a water sample\n",
    "\n",
    "occ['occurrenceID'] = plate.groupby('Sequence_ID')['ASV'].cumcount()+1\n",
    "occ['occurrenceID'] = occ['eventID'] + '_occ' + occ['occurrenceID'].astype(str)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>samp_vol_we_dna_ext</th>\n",
       "      <th>nucl_acid_ext</th>\n",
       "      <th>nucl_acid_amp</th>\n",
       "      <th>target_subfragment</th>\n",
       "      <th>lib_layout</th>\n",
       "      <th>otu_class_appr</th>\n",
       "      <th>otu_seq_comp_appr</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_1_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_2_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>100ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_3_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_1_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>1000ml</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.xjufknw</td>\n",
       "      <td>dx.doi.org/10.17504/protocols.io.n2vdge6</td>\n",
       "      <td>V9</td>\n",
       "      <td>paired</td>\n",
       "      <td>dada2;1.14.0;ASV</td>\n",
       "      <td>blast;2.9.0+;80% identity;e-value cutoff: 0.00...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_2_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ... samp_vol_we_dna_ext  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...               100ml   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...               100ml   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...               100ml   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...              1000ml   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...              1000ml   \n",
       "\n",
       "                              nucl_acid_ext  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "1  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "2  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "3  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "4  dx.doi.org/10.17504/protocols.io.xjufknw   \n",
       "\n",
       "                              nucl_acid_amp target_subfragment lib_layout  \\\n",
       "0  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "1  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "2  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "3  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "4  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   \n",
       "\n",
       "     otu_class_appr                                  otu_seq_comp_appr  \\\n",
       "0  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "1  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "2  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "3  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "4  dada2;1.14.0;ASV  blast;2.9.0+;80% identity;e-value cutoff: 0.00...   \n",
       "\n",
       "           otu_db               occurrenceID  \\\n",
       "0  Genbank nr;221  05114c01_12_edna_1_S_occ1   \n",
       "1  Genbank nr;221  05114c01_12_edna_2_S_occ1   \n",
       "2  Genbank nr;221  05114c01_12_edna_3_S_occ1   \n",
       "3  Genbank nr;221  11216c01_12_edna_1_S_occ1   \n",
       "4  Genbank nr;221  11216c01_12_edna_2_S_occ1   \n",
       "\n",
       "                                        DNA_sequence  \n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add DNA_sequence\n",
    "\n",
    "occ['DNA_sequence'] = plate['ASV']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_1_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_2_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_3_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_1_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_2_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...          otu_db  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "\n",
       "                occurrenceID  \\\n",
       "0  05114c01_12_edna_1_S_occ1   \n",
       "1  05114c01_12_edna_2_S_occ1   \n",
       "2  05114c01_12_edna_3_S_occ1   \n",
       "3  11216c01_12_edna_1_S_occ1   \n",
       "4  11216c01_12_edna_2_S_occ1   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add scientificName, taxonomic info\n",
    "\n",
    "occ['scientificName'] = plate['Species']\n",
    "occ['kingdom'] = plate['Kingdom']\n",
    "occ['phylum'] = plate['Phylum']\n",
    "occ['class'] = plate['Class']\n",
    "occ['order'] = plate['Order']\n",
    "occ['family'] = plate['Family']\n",
    "occ['genus'] = plate['Genus']\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the purpose of submitting data to OBIS, all the variations on missing data (e.g. \"unknown,\" \"no_hit,\" etc.) do not add information. We can replace these with NaN, which is easy to work with in pandas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>otu_db</th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>DNA_sequence</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_1_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_2_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>05114c01_12_edna_3_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_1_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Genbank nr;221</td>\n",
       "      <td>11216c01_12_edna_2_S_occ1</td>\n",
       "      <td>GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...          otu_db  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Genbank nr;221   \n",
       "\n",
       "                occurrenceID  \\\n",
       "0  05114c01_12_edna_1_S_occ1   \n",
       "1  05114c01_12_edna_2_S_occ1   \n",
       "2  05114c01_12_edna_3_S_occ1   \n",
       "3  11216c01_12_edna_1_S_occ1   \n",
       "4  11216c01_12_edna_2_S_occ1   \n",
       "\n",
       "                                        DNA_sequence scientificName  \\\n",
       "0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   \n",
       "\n",
       "     kingdom      phylum        class      order         family        genus  \n",
       "0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Replace 'unknown', 'unassigned', etc. in scientificName and taxonomy columns with NaN\n",
    "\n",
    "cols = ['scientificName', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "occ[cols] = occ[cols].replace({'unassigned':np.nan,\n",
    "                              's_':np.nan,\n",
    "                              'g_':np.nan,\n",
    "                              'unknown':np.nan,\n",
    "                              'no_hit':np.nan})\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "323\n"
     ]
    }
   ],
   "source": [
    "## Get unique species names\n",
    "\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)]  # remove NaN\n",
    "print(len(names))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OBIS uses the World Register of Marine Species (or [WoRMS](http://www.marinespecies.org/)) as it's taxonomic backbone, so scientific names have to be WoRMS-approved in order to show up as valid occurrences. But there are a number of entries in the `scientificName` column, like \"uncultured marine eukaryote,\" \"eukaryote clone OLI11007,\" and \"Acantharian sp. 6201,\" that are **not proper Linnaean species names**. Since these essentially indicate that a more precise name is unknown, it seemed reasonable to replace these with NaN as well. \n",
    "\n",
    "**NOTE: I used a simple rule to filter out non-Linnaean names, but it's important to check and see if any true species names are being removed.**\n",
    "\n",
    "To visually inspect names that are being filtered out, use:\n",
    "```python\n",
    "names = occ['scientificName'].unique()\n",
    "names = names[~pd.isnull(names)]  # remove NaN\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        print(name)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace non-Linnaean species names with NaN\n",
    "\n",
    "# Get non-Linnaean names\n",
    "non_latin_names = []\n",
    "for name in names:\n",
    "    words_in_name = name.split(' ')\n",
    "    if len(words_in_name) > 2:\n",
    "        non_latin_names.append(name)\n",
    "non_latin_names_dict = {i:np.nan for i in non_latin_names}\n",
    "\n",
    "# Add any names that didn't get caught in the simple filter\n",
    "non_latin_names_dict['phototrophic eukaryote'] = np.nan\n",
    "non_latin_names_dict['Candida <clade Candida/Lodderomyces clade>'] = np.nan\n",
    "\n",
    "# Replace\n",
    "occ['scientificName'].replace(non_latin_names_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition, many records **only give \"Eukaryota\" as the scientific name** (i.e. Eukaryota is in the kingdom field, and there is no more taxonomic information). These should be replaced with [Biota](http://marinespecies.org/aphia.php?p=taxdetails&id=1), which is WoRMS's most general taxonomic designation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Replace entries where kingdom = 'Eukaryota' with the WoRMS-approved 'Biota'\n",
    "\n",
    "occ.loc[occ['kingdom'] == 'Eukaryota', 'kingdom'] = 'Biota'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data providers for this dataset used the [NCBI taxonomy database](https://www.ncbi.nlm.nih.gov/taxonomy) as their reference database when assigning taxonomies to ASVs. **It's important to note** that this taxonomy database is not a taxonomic authority, and the taxonomic ranks for any given scientific name on WoRMS may not directly compare. There are ongoing discussions about this problem (see [this](https://github.com/iobis/Project-team-Genetic-Data/issues/5) GitHub issue). At the moment, I don't see a way to definitively ensure that a given scientific name actually has the same taxonomic ranks on both platforms without going case-by-case.\n",
    "\n",
    "In addition, there are still names in the data that will not match on WoRMS at all, despite appearing to be Linnaean names. This is because the name may not have been fully and officially adopted by the scientific community. I therefore need a system for searching through the higher taxonomic ranks given, finding the lowest one that will match on WoRMS, and putting that name in the `scientificName` column. The following few code blocks do this - they're clunky, but they were sufficient for this data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define functions for finding the lowest available taxonomic rank that will match on WoRMS\n",
    "\n",
    "def fill_lowest_taxon(df, cols):\n",
    "    \"\"\" Takes the occurrence pandas data frame and fills missing values in scientificName \n",
    "    with values from the first non-missing taxonomic rank column. The names of the taxonomic\n",
    "    rank columns are listed in cols. \"\"\"\n",
    "    \n",
    "    cols.reverse()\n",
    "    \n",
    "    for col in cols[:-1]:\n",
    "        df['scientificName'] = df['scientificName'].combine_first(df[col])\n",
    "    \n",
    "    cols.reverse()\n",
    "    \n",
    "    return(df)\n",
    "\n",
    "def find_not_matched(df, name_dict):\n",
    "    \"\"\" Takes the occurrence pandas data frame and name_dict matching scientificName values \n",
    "    with names on WoRMS and returns a list of names that did not match on WoRMS. \"\"\"\n",
    "    \n",
    "    not_matched = []\n",
    "    \n",
    "    for name in df['scientificName'].unique():\n",
    "        if name not in name_dict.keys():\n",
    "            not_matched.append(name)\n",
    "    \n",
    "    try:\n",
    "        not_matched.remove(np.nan)\n",
    "    except ValueError:\n",
    "        pass\n",
    "            \n",
    "    return(not_matched)\n",
    "\n",
    "def replace_not_matched(df, not_matched, cols):\n",
    "    \"\"\" Takes the occurrence pandas data frame and a list of scientificName values that \n",
    "    did not match on WoRMS and replaces those values with NaN in the columns specified by cols. \"\"\"\n",
    "    \n",
    "    df[cols] = df[cols].replace(not_matched, np.nan)\n",
    "    \n",
    "    return(df)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matching 756 names on WoRMS.\n",
      "Length of name_name_dict: 696\n",
      "Number of names not matched: 60\n",
      "Matching 35 names on WoRMS.\n",
      "Length of name_name_dict: 716\n",
      "Number of names not matched: 15\n",
      "Matching 11 names on WoRMS.\n",
      "Length of name_name_dict: 721\n",
      "Number of names not matched: 6\n",
      "Matching 3 names on WoRMS.\n",
      "Length of name_name_dict: 722\n",
      "Number of names not matched: 2\n",
      "Matching 1 names on WoRMS.\n",
      "Length of name_name_dict: 722\n",
      "Number of names not matched: 1\n",
      "Matching 0 names on WoRMS.\n",
      "Length of name_name_dict: 722\n",
      "Number of names not matched: 0\n"
     ]
    }
   ],
   "source": [
    "## Iterate to match lowest possible taxonomic rank on WoRMS (takes ~8 minutes when starting with ~750 names)\n",
    "\n",
    "# Note that cols (list of taxonomic column names) was defined in a previous code block \n",
    "\n",
    "# Initialize dictionaries\n",
    "name_name_dict = {}\n",
    "name_id_dict = {}\n",
    "name_taxid_dict = {}\n",
    "name_class_dict = {}\n",
    "\n",
    "# Initialize not_matched\n",
    "not_matched = [1]\n",
    "\n",
    "# Iterate\n",
    "while len(not_matched) > 0:\n",
    "    \n",
    "    # Step 1 - fill\n",
    "    occ = fill_lowest_taxon(occ, cols)\n",
    "\n",
    "    # Step 2 - get names to match\n",
    "    to_match = find_not_matched(occ, name_name_dict)\n",
    "\n",
    "    # Step 3 - match\n",
    "    print('Matching {num} names on WoRMS.'.format(num = len(to_match)))\n",
    "    name_id, name_name, name_taxid, name_class = WoRMS.run_get_worms_from_scientific_name(to_match, verbose_flag=False)\n",
    "    name_id_dict = {**name_id_dict, **name_id}\n",
    "    name_name_dict = {**name_name_dict, **name_name}\n",
    "    name_taxid_dict = {**name_taxid_dict, **name_taxid}\n",
    "    name_class_dict = {**name_class_dict, **name_class}\n",
    "    print('Length of name_name_dict: {length}'.format(length = len(name_name_dict)))\n",
    "\n",
    "    # Step 4 - get names that didn't match\n",
    "    not_matched = find_not_matched(occ, name_name_dict)\n",
    "    print('Number of names not matched: {num}'.format(num = len(not_matched)))\n",
    "\n",
    "    # Step 5 - replace these values with NaN\n",
    "    occ = replace_not_matched(occ, not_matched, cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are, I'm sure, a vast number of ways to improve on this. A couple that have crossed my mind are:\n",
    "- Use of .reverse() in fill_lowest_taxon()\n",
    "- Add in a progress bar\n",
    "- Consider better and/or additional stopping criteria. Importantly, what if not all names can be matched?\n",
    "- Could consider using pyworms instead of my custom WoRMS functions\n",
    "\n",
    "There are quite a few records where no taxonomic information was obtained at all (i.e., after this whole process, `scientificName` is still NaN). I set `scientificName` to 'Biota' for these records."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33360, 32)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(0, 32)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Change scientificName to Biota in cases where all taxonomic information is missing\n",
    "\n",
    "print(occ[occ['scientificName'].isna() == True].shape)\n",
    "occ.loc[occ['scientificName'].isna() == True, 'scientificName'] = 'Biota'\n",
    "occ[occ['scientificName'].isna() == True].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, during the above process, **I altered the taxonomy columns in order to obtain the best possible `scientificName` column**. I chose to re-populate these columns with the taxonomy from the original data set, rather than altering some of the names to match taxonomy retrieved from WoRMS. In this case, it seemed best to adhere as closely as possible to the original data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fix taxonomy columns\n",
    "\n",
    "# Replace with original data\n",
    "occ[cols[1:]] = plate[['Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus']].copy()\n",
    "\n",
    "# Replace missing data indicators in original data with empty strings ('')\n",
    "occ[cols[1:]] = occ[cols[1:]].replace({\n",
    "    'unassigned':'',\n",
    "    's_':'',\n",
    "    'g_':'',\n",
    "    'unknown':'',\n",
    "    'no_hit':''})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ... scientificName    kingdom  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   \n",
       "\n",
       "       phylum        class      order         family        genus  \\\n",
       "0  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add scientific name-related columns\n",
    "\n",
    "occ['scientificNameID'] = occ['scientificName']\n",
    "occ['scientificNameID'].replace(name_id_dict, inplace=True)\n",
    "\n",
    "occ['taxonID'] = occ['scientificName']\n",
    "occ['taxonID'].replace(name_taxid_dict, inplace=True)\n",
    "\n",
    "occ['scientificName'].replace(name_name_dict, inplace=True)\n",
    "\n",
    "occ['nameAccordingTo'] = 'WoRMS'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I also wanted to persist the original name from the NCBI taxonomy database into the Darwin Core-converted data set. To do this, I queried the database based on the name in the original data to obtain its taxonomic ID number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'biosystems', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get set up to query NCBI taxonomy \n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "# ----- Insert your email here -----\n",
    "Entrez.email = 'dianalg@mbari.org'\n",
    "# ----------------------------------\n",
    "\n",
    "# Get list of all databases available through this tool\n",
    "record = Entrez.read(Entrez.einfo())\n",
    "all_dbs = record['DbList']\n",
    "all_dbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get NCBI taxIDs for each name in dataset ---- TAKES ~ 2 MINUTES FOR 300 RECORDS\n",
    "\n",
    "name_ncbiid_dict = {}\n",
    "\n",
    "for name in names:\n",
    "    handle = Entrez.esearch(db='taxonomy', retmax=10, term=name)\n",
    "    record = Entrez.read(handle)\n",
    "    name_ncbiid_dict[name] = record['IdList'][0]\n",
    "    handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that this code will throw an IndexError (IndexError: list index out of range) if a term is not found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...    kingdom      phylum  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   \n",
       "\n",
       "         class      order         family        genus  \\\n",
       "0  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID  \n",
       "0                 \n",
       "1                 \n",
       "2                 \n",
       "3                 \n",
       "4                 \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add NCBI taxonomy IDs under taxonConceptID\n",
    "\n",
    "# Map indicators that say no taxonomy was assigned to empty strings\n",
    "name_ncbiid_dict['unassigned'], name_ncbiid_dict['s_'], name_ncbiid_dict['no_hit'], name_ncbiid_dict['unknown'], name_ncbiid_dict['g_'] = '', '', '', '', ''\n",
    "\n",
    "# Create column\n",
    "occ['taxonConceptID']  = plate['Species'].copy()\n",
    "occ['taxonConceptID'].replace(name_ncbiid_dict, inplace=True)\n",
    "\n",
    "# Add remainder of text and clean\n",
    "occ['taxonConceptID'] = 'NCBI:txid' + occ['taxonConceptID']\n",
    "occ['taxonConceptID'].replace('NCBI:txid', '', inplace=True)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...      phylum        class  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   \n",
       "\n",
       "       order         family        genus  \\\n",
       "0  Calanoida  Paracalanidae  Paracalanus   \n",
       "1  Calanoida  Paracalanidae  Paracalanus   \n",
       "2  Calanoida  Paracalanidae  Paracalanus   \n",
       "3  Calanoida  Paracalanidae  Paracalanus   \n",
       "4  Calanoida  Paracalanidae  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \n",
       "0                 unassigned, Genbank nr Release 221 September 2...  \n",
       "1                 unassigned, Genbank nr Release 221 September 2...  \n",
       "2                 unassigned, Genbank nr Release 221 September 2...  \n",
       "3                 unassigned, Genbank nr Release 221 September 2...  \n",
       "4                 unassigned, Genbank nr Release 221 September 2...  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## identificationRemarks\n",
    "\n",
    "# Get identificationRemarks\n",
    "occ = occ.merge(meta[['seqID', 'identificationRemarks']], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "\n",
    "# Add name that matched in GenBank - i.e. the species name from the original data\n",
    "occ['identificationRemarks'] = plate['Species'].copy() + ', ' + occ['identificationRemarks']\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Hexanauplia</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...        class      order  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   \n",
       "\n",
       "          family        genus                           scientificNameID  \\\n",
       "0  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "1  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "2  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "3  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "4  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   \n",
       "\n",
       "  taxonID nameAccordingTo taxonConceptID  \\\n",
       "0  104196           WoRMS                  \n",
       "1  104196           WoRMS                  \n",
       "2  104196           WoRMS                  \n",
       "3  104196           WoRMS                  \n",
       "4  104196           WoRMS                  \n",
       "\n",
       "                               identificationRemarks   basisOfRecord  \n",
       "0  unassigned, Genbank nr Release 221 September 2...  MaterialSample  \n",
       "1  unassigned, Genbank nr Release 221 September 2...  MaterialSample  \n",
       "2  unassigned, Genbank nr Release 221 September 2...  MaterialSample  \n",
       "3  unassigned, Genbank nr Release 221 September 2...  MaterialSample  \n",
       "4  unassigned, Genbank nr Release 221 September 2...  MaterialSample  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## basisOfRecord\n",
    "\n",
    "occ['basisOfRecord'] = 'MaterialSample'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Calanoida</td>\n",
       "      <td>Paracalanidae</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...      order         family  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   \n",
       "\n",
       "         genus                           scientificNameID taxonID  \\\n",
       "0  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   \n",
       "1  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   \n",
       "2  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   \n",
       "3  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   \n",
       "4  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   \n",
       "\n",
       "  nameAccordingTo taxonConceptID  \\\n",
       "0           WoRMS                  \n",
       "1           WoRMS                  \n",
       "2           WoRMS                  \n",
       "3           WoRMS                  \n",
       "4           WoRMS                  \n",
       "\n",
       "                               identificationRemarks   basisOfRecord  \\\n",
       "0  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "1  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "2  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "3  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "4  unassigned, Genbank nr Release 221 September 2...  MaterialSample   \n",
       "\n",
       "                            identificationReferences  \n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add identificationReferences \n",
    "\n",
    "occ = occ.merge(meta[['seqID', 'identificationReferences']], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "occ['identificationReferences'] = occ['identificationReferences'].str.replace('| ', ' | ', regex=False)\n",
    "\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>genus</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>Paracalanus</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...        genus  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord                           identificationReferences  \\\n",
       "0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "\n",
       "  organismQuantity organismQuantityType  \n",
       "0            14825   DNA sequence reads  \n",
       "1            16094   DNA sequence reads  \n",
       "2            22459   DNA sequence reads  \n",
       "3            19312   DNA sequence reads  \n",
       "4            16491   DNA sequence reads  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## organismQuantity (number of reads)\n",
    "\n",
    "occ['organismQuantity'] = plate['Reads']\n",
    "occ['organismQuantityType'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the context of eDNA data, `sampleSizeValue` should be the total number of reads for a given sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(280440, 42)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>scientificNameID</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>urn:lsid:marinespecies.org:taxname:104196</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ...  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   \n",
       "\n",
       "                            scientificNameID taxonID nameAccordingTo  \\\n",
       "0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord                           identificationReferences  \\\n",
       "0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "\n",
       "  organismQuantity organismQuantityType sampleSizeValue  \n",
       "0            14825   DNA sequence reads           85600  \n",
       "1            16094   DNA sequence reads           90702  \n",
       "2            22459   DNA sequence reads          130275  \n",
       "3            19312   DNA sequence reads          147220  \n",
       "4            16491   DNA sequence reads          121419  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeValue\n",
    "\n",
    "count_by_seq = plate.groupby('Sequence_ID', as_index=False)['Reads'].sum()\n",
    "occ = occ.merge(count_by_seq, how='left', left_on='eventID', right_on='Sequence_ID')\n",
    "occ.drop(columns='Sequence_ID', inplace=True)\n",
    "occ.rename(columns={'Reads':'sampleSizeValue'}, inplace=True)\n",
    "print(occ.shape)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>104196</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ... taxonID nameAccordingTo  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord                           identificationReferences  \\\n",
       "0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "\n",
       "  organismQuantity organismQuantityType sampleSizeValue      sampleSizeUnit  \n",
       "0            14825   DNA sequence reads           85600  DNA sequence reads  \n",
       "1            16094   DNA sequence reads           90702  DNA sequence reads  \n",
       "2            22459   DNA sequence reads          130275  DNA sequence reads  \n",
       "3            19312   DNA sequence reads          147220  DNA sequence reads  \n",
       "4            16491   DNA sequence reads          121419  DNA sequence reads  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sampleSizeUnit\n",
    "\n",
    "occ['sampleSizeUnit'] = 'DNA sequence reads'\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eventID</th>\n",
       "      <th>eventDate</th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>pcr_primer_forward</th>\n",
       "      <th>pcr_primer_reverse</th>\n",
       "      <th>...</th>\n",
       "      <th>nameAccordingTo</th>\n",
       "      <th>taxonConceptID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>basisOfRecord</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>organismQuantity</th>\n",
       "      <th>organismQuantityType</th>\n",
       "      <th>sampleSizeValue</th>\n",
       "      <th>sampleSizeUnit</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05114c01_12_edna_1_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14825</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>85600</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>05114c01_12_edna_2_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16094</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>90702</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05114c01_12_edna_3_S</td>\n",
       "      <td>2014-02-20T15:33:00-08:00</td>\n",
       "      <td>36.7958</td>\n",
       "      <td>-121.848</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22459</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>130275</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11216c01_12_edna_1_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>19312</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>147220</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11216c01_12_edna_2_S</td>\n",
       "      <td>2016-04-21T14:39:00-07:00</td>\n",
       "      <td>36.7962</td>\n",
       "      <td>-121.846</td>\n",
       "      <td>marine biome  (ENVO:00000447)</td>\n",
       "      <td>coastal water (ENVO:00001250)</td>\n",
       "      <td>waterborne particulate matter (ENVO:01000436)</td>\n",
       "      <td>18S</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td></td>\n",
       "      <td>unassigned, Genbank nr Release 221 September 2...</td>\n",
       "      <td>MaterialSample</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>16491</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>121419</td>\n",
       "      <td>DNA sequence reads</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                eventID                  eventDate  decimalLatitude  \\\n",
       "0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   \n",
       "3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   \n",
       "\n",
       "   decimalLongitude                env_broad_scale  \\\n",
       "0          -121.848  marine biome  (ENVO:00000447)   \n",
       "1          -121.848  marine biome  (ENVO:00000447)   \n",
       "2          -121.848  marine biome  (ENVO:00000447)   \n",
       "3          -121.846  marine biome  (ENVO:00000447)   \n",
       "4          -121.846  marine biome  (ENVO:00000447)   \n",
       "\n",
       "                 env_local_scale  \\\n",
       "0  coastal water (ENVO:00001250)   \n",
       "1  coastal water (ENVO:00001250)   \n",
       "2  coastal water (ENVO:00001250)   \n",
       "3  coastal water (ENVO:00001250)   \n",
       "4  coastal water (ENVO:00001250)   \n",
       "\n",
       "                                      env_medium target_gene  \\\n",
       "0  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "1  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "2  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "3  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "4  waterborne particulate matter (ENVO:01000436)         18S   \n",
       "\n",
       "  pcr_primer_forward        pcr_primer_reverse  ... nameAccordingTo  \\\n",
       "0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   \n",
       "\n",
       "  taxonConceptID                              identificationRemarks  \\\n",
       "0                 unassigned, Genbank nr Release 221 September 2...   \n",
       "1                 unassigned, Genbank nr Release 221 September 2...   \n",
       "2                 unassigned, Genbank nr Release 221 September 2...   \n",
       "3                 unassigned, Genbank nr Release 221 September 2...   \n",
       "4                 unassigned, Genbank nr Release 221 September 2...   \n",
       "\n",
       "    basisOfRecord                           identificationReferences  \\\n",
       "0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   \n",
       "\n",
       "  organismQuantity organismQuantityType sampleSizeValue      sampleSizeUnit  \\\n",
       "0            14825   DNA sequence reads           85600  DNA sequence reads   \n",
       "1            16094   DNA sequence reads           90702  DNA sequence reads   \n",
       "2            22459   DNA sequence reads          130275  DNA sequence reads   \n",
       "3            19312   DNA sequence reads          147220  DNA sequence reads   \n",
       "4            16491   DNA sequence reads          121419  DNA sequence reads   \n",
       "\n",
       "                            associatedSequences  \n",
       "0  NCBI BioProject accession number PRJNA433203  \n",
       "1  NCBI BioProject accession number PRJNA433203  \n",
       "2  NCBI BioProject accession number PRJNA433203  \n",
       "3  NCBI BioProject accession number PRJNA433203  \n",
       "4  NCBI BioProject accession number PRJNA433203  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## associatedSequences\n",
    "\n",
    "occ = occ.merge(meta[['seqID', 'associatedSequences']], how='left', left_on='eventID', right_on='seqID')\n",
    "occ.drop(columns='seqID', inplace=True)\n",
    "occ.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(64903, 44)\n"
     ]
    }
   ],
   "source": [
    "## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)\n",
    "\n",
    "occ = occ[occ['organismQuantity'] > 0]\n",
    "print(occ.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "eventID                     0\n",
       "eventDate                   0\n",
       "decimalLatitude             0\n",
       "decimalLongitude            0\n",
       "env_broad_scale             0\n",
       "env_local_scale             0\n",
       "env_medium                  0\n",
       "target_gene                 0\n",
       "pcr_primer_forward          0\n",
       "pcr_primer_reverse          0\n",
       "pcr_primer_name_forward     0\n",
       "pcr_primer_name_reverse     0\n",
       "pcr_primer_reference        0\n",
       "sop                         0\n",
       "seq_meth                    0\n",
       "samp_vol_we_dna_ext         0\n",
       "nucl_acid_ext               0\n",
       "nucl_acid_amp               0\n",
       "target_subfragment          0\n",
       "lib_layout                  0\n",
       "otu_class_appr              0\n",
       "otu_seq_comp_appr           0\n",
       "otu_db                      0\n",
       "occurrenceID                0\n",
       "DNA_sequence                0\n",
       "scientificName              0\n",
       "kingdom                     0\n",
       "phylum                      0\n",
       "class                       0\n",
       "order                       0\n",
       "family                      0\n",
       "genus                       0\n",
       "scientificNameID            0\n",
       "taxonID                     0\n",
       "nameAccordingTo             0\n",
       "taxonConceptID              0\n",
       "identificationRemarks       0\n",
       "basisOfRecord               0\n",
       "identificationReferences    0\n",
       "organismQuantity            0\n",
       "organismQuantityType        0\n",
       "sampleSizeValue             0\n",
       "sampleSizeUnit              0\n",
       "associatedSequences         0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Check for NaN values in string fields - if there are any, replace them with empty strings ('')\n",
    "\n",
    "occ.isna(). sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Divide into occurrence and DNADerivedDataExt\n",
    "\n",
    "ddd_cols = [\n",
    "    'eventID',\n",
    "    'occurrenceID',\n",
    "    'DNA_sequence',\n",
    "    'sop',\n",
    "    'nucl_acid_ext',\n",
    "    'samp_vol_we_dna_ext',\n",
    "    'nucl_acid_amp',\n",
    "    'target_gene',\n",
    "    'target_subfragment',\n",
    "    'lib_layout',\n",
    "    'pcr_primer_forward',\n",
    "    'pcr_primer_reverse',\n",
    "    'pcr_primer_name_forward',\n",
    "    'pcr_primer_name_reverse',\n",
    "    'pcr_primer_reference',\n",
    "    'seq_meth',\n",
    "    'otu_class_appr',\n",
    "    'otu_seq_comp_appr',\n",
    "    'otu_db',\n",
    "    'env_broad_scale',\n",
    "    'env_local_scale',\n",
    "    'env_medium',\n",
    "]\n",
    "\n",
    "DNADerivedData = occ[ddd_cols].copy()\n",
    "\n",
    "occ.drop(ddd_cols[2:], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save\n",
    "\n",
    "# Get path\n",
    "folder = os.getcwd().replace('src', 'processed')\n",
    "occ_filename = os.path.join(folder, 'occurrence.csv')\n",
    "ddd_filename = os.path.join(folder, 'dna_extension.csv')\n",
    "\n",
    "# Create folder\n",
    "if not os.path.exists(folder):\n",
    "    os.makedirs(folder)\n",
    "\n",
    "# Save\n",
    "# occ.to_csv(occ_filename, index=False, na_rep='NaN')\n",
    "# DNADerivedData.to_csv(ddd_filename, index=False, na_rep='NaN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Boneyard"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**update**\n",
    "Plate data contains the ASV sequence, the number of reads (number of times that ASV was observed in the sample), and the taxonomy associated with that ASV.\n",
    "\n",
    "| Column name| Column definition                                                                                                                                                                           |\n",
    "|------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| ASV        | The sequence of the Amplicon Sequence Variant observed                                                                                                                                       |\n",
    "| FilterID   | A unique identifier for the filter the sample was obtained from, composed of: <br>- cruise number <br>- CTD cast number<br>- CTD bottle number <br>- filter indicator <br>- replicate number |\n",
    "| Sequence_ID| The FilterID plus a letter indicating which plate the sample was on when sequenced                                                                                                           |\n",
    "| Reads      | The number of reads for the ASV                                                                                                                                                             |\n",
    "| Kingdom    | The Kingdom of the taxonomic identity assigned to the ASV, if known                                                                                                                          |\n",
    "| Phylum     | The Phylum of the taxonomic identity assigned to the ASV, if known                                                                                                                           |\n",
    "| Class      | The Class of the taxonomic identity assigned to the ASV, if known                                                                                                                            |\n",
    "| Order      | The Order of the taxonomic identity assigned to the ASV, if known                                                                                                                            |\n",
    "| Family     | The Family of the taxonomic identity assigned to the ASV, if known                                                                                                                           |\n",
    "| Genus      | The Genus of the taxonomic identity assigned to the ASV, if known                                                                                                                            |\n",
    "| Species    | The Species of the taxonomic identity assigned to the ASV, if known                                                                                                                          |\n",
    "\n",
    "Additionally, taxonomic columns may include the following designations:\n",
    "- **unknown** = GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic rank. I.e., either the name doesn't exist, or there isn't enough scientific consensus to give a name.\n",
    "- **no_hit** = BLAST did not find any hits for the ASV.\n",
    "- **unassigned** = The ASV got BLAST hits, but the post-processing program Megan6 didn't assign the ASV to any taxonomic group.\n",
    "- **g_** or **s_** = Megan6 assigned the ASV to a genus or species, but not with high enough confidence to include it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60, 67)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library</th>\n",
       "      <th>tag_sequence</th>\n",
       "      <th>primer_sequence_forward</th>\n",
       "      <th>primer_sequence_reverse</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>PlateID</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>pcr_primer_name_forward</th>\n",
       "      <th>pcr_primer_name_reverse</th>\n",
       "      <th>pcr_primer_reference</th>\n",
       "      <th>seq_meth</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>seqID</th>\n",
       "      <th>identificationRemarks</th>\n",
       "      <th>identificationReferences</th>\n",
       "      <th>FilterID</th>\n",
       "      <th>associatedSequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14213c01_12_eDNA_1</td>\n",
       "      <td>S1</td>\n",
       "      <td>ACGAGACTGATT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_1_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14213c01_12_eDNA_2</td>\n",
       "      <td>S2</td>\n",
       "      <td>GAATACCAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_2_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_2_S2_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_2_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14213c01_12_eDNA_3</td>\n",
       "      <td>S3</td>\n",
       "      <td>CGAGGGAAAGTC</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>14213c01_12_edna_3_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>14213c01_12_edna_3_S3_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>14213c01_12_edna_3_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>14213c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22013c01_12_eDNA_1</td>\n",
       "      <td>S4</td>\n",
       "      <td>GAACACTTTGGA</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>22013c01_12_edna_1_S4_L001_R1_001.fastq.gz</td>\n",
       "      <td>22013c01_12_edna_1_S4_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>22013c01_12_edna_1_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22013c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22013c01_12_eDNA_2</td>\n",
       "      <td>S5</td>\n",
       "      <td>ACTCACAGGAAT</td>\n",
       "      <td>GTACACACCGCCCGTC</td>\n",
       "      <td>TGATCCTTCTGCAGGTTCACCTAC</td>\n",
       "      <td>22013c01_12_edna_2_S5_L001_R1_001.fastq.gz</td>\n",
       "      <td>22013c01_12_edna_2_S5_L001_R2_001.fastq.gz</td>\n",
       "      <td>S</td>\n",
       "      <td>environmental</td>\n",
       "      <td>18S</td>\n",
       "      <td>...</td>\n",
       "      <td>1391f</td>\n",
       "      <td>EukBr</td>\n",
       "      <td>Amaral-Zettler LA, McCliment EA, Ducklow HW, H...</td>\n",
       "      <td>NGS Illumina Miseq</td>\n",
       "      <td>Stanford</td>\n",
       "      <td>22013c01_12_edna_2_S</td>\n",
       "      <td>Genbank nr Release 221 September 20 2017</td>\n",
       "      <td>https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...</td>\n",
       "      <td>22013c01_12_edna</td>\n",
       "      <td>NCBI BioProject accession number PRJNA433203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample_name library  tag_sequence primer_sequence_forward  \\\n",
       "0  14213c01_12_eDNA_1      S1  ACGAGACTGATT        GTACACACCGCCCGTC   \n",
       "1  14213c01_12_eDNA_2      S2  GAATACCAAGTC        GTACACACCGCCCGTC   \n",
       "2  14213c01_12_eDNA_3      S3  CGAGGGAAAGTC        GTACACACCGCCCGTC   \n",
       "3  22013c01_12_eDNA_1      S4  GAACACTTTGGA        GTACACACCGCCCGTC   \n",
       "4  22013c01_12_eDNA_2      S5  ACTCACAGGAAT        GTACACACCGCCCGTC   \n",
       "\n",
       "    primer_sequence_reverse                                          R1  \\\n",
       "0  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_1_S1_L001_R1_001.fastq.gz   \n",
       "1  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_2_S2_L001_R1_001.fastq.gz   \n",
       "2  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_3_S3_L001_R1_001.fastq.gz   \n",
       "3  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_1_S4_L001_R1_001.fastq.gz   \n",
       "4  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_2_S5_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                           R2 PlateID    sample_type  \\\n",
       "0  14213c01_12_edna_1_S1_L001_R2_001.fastq.gz       S  environmental   \n",
       "1  14213c01_12_edna_2_S2_L001_R2_001.fastq.gz       S  environmental   \n",
       "2  14213c01_12_edna_3_S3_L001_R2_001.fastq.gz       S  environmental   \n",
       "3  22013c01_12_edna_1_S4_L001_R2_001.fastq.gz       S  environmental   \n",
       "4  22013c01_12_edna_2_S5_L001_R2_001.fastq.gz       S  environmental   \n",
       "\n",
       "  target_gene  ...  pcr_primer_name_forward pcr_primer_name_reverse  \\\n",
       "0         18S  ...                    1391f                   EukBr   \n",
       "1         18S  ...                    1391f                   EukBr   \n",
       "2         18S  ...                    1391f                   EukBr   \n",
       "3         18S  ...                    1391f                   EukBr   \n",
       "4         18S  ...                    1391f                   EukBr   \n",
       "\n",
       "                                pcr_primer_reference            seq_meth  \\\n",
       "0  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "1  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "2  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "3  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "4  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   \n",
       "\n",
       "   sequencing_facility                 seqID  \\\n",
       "0             Stanford  14213c01_12_edna_1_S   \n",
       "1             Stanford  14213c01_12_edna_2_S   \n",
       "2             Stanford  14213c01_12_edna_3_S   \n",
       "3             Stanford  22013c01_12_edna_1_S   \n",
       "4             Stanford  22013c01_12_edna_2_S   \n",
       "\n",
       "                      identificationRemarks  \\\n",
       "0  Genbank nr Release 221 September 20 2017   \n",
       "1  Genbank nr Release 221 September 20 2017   \n",
       "2  Genbank nr Release 221 September 20 2017   \n",
       "3  Genbank nr Release 221 September 20 2017   \n",
       "4  Genbank nr Release 221 September 20 2017   \n",
       "\n",
       "                            identificationReferences          FilterID  \\\n",
       "0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   \n",
       "1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   \n",
       "2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   \n",
       "3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna   \n",
       "4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna   \n",
       "\n",
       "                            associatedSequences  \n",
       "0  NCBI BioProject accession number PRJNA433203  \n",
       "1  NCBI BioProject accession number PRJNA433203  \n",
       "2  NCBI BioProject accession number PRJNA433203  \n",
       "3  NCBI BioProject accession number PRJNA433203  \n",
       "4  NCBI BioProject accession number PRJNA433203  \n",
       "\n",
       "[5 rows x 67 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Plate metadata\n",
    "\n",
    "filename = os.getcwd().replace('src', os.path.join('raw', 'metadata_table.csv'))  \n",
    "meta = pd.read_csv(filename)\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metadata contains information on sample collection, DNA extraction, DNA amplification, and DNA sequencing. \n",
    "\n",
    "Definitions of relevant columns:\n",
    "\n",
    "| Column name              | Column definition                                                                       |\n",
    "|--------------------------|-----------------------------------------------------------------------------------------|\n",
    "| primer_sequence_forward  | The sequence of the forward primer used during PCR                                      |\n",
    "| primer_sequence_reverse  | The sequence of the reverse primer used during PCR                                      |\n",
    "| target_gene              | The gene being targeted for amplification during PCR                                    |\n",
    "| eventDate                | The date (and time, if available) the water sample was collected                        |\n",
    "| decimalLatitude          | The latitude in decimal degrees where the water sample was collected (WGS84)            |\n",
    "| decimalLongitude         | The longitude in decimal degrees where the water sample was collected (WGS84)           |\n",
    "| env_broad_scale          | The most broad descriptor of the environment from which the water sample was collected  |\n",
    "| env_local_scale          | A more specific descriptor of the environment from which the water sample was collected |\n",
    "| env_medium               | A descriptor of the medium from which the DNA was collected                             |\n",
    "| minimumDepthInMeters     | The minimum depth at which the water sample was collected                               |\n",
    "| maximumDepthInMeters     | The maximum depth at which the water sample was collected                               |\n",
    "| samp_vol_we_dna_ext      | The volume of the water sample that was processed during DNA extraction                 |\n",
    "| nucl_acid_ext            | Reference to the DNA extraction protocol                                                |\n",
    "| nucl_acid_amp            | Reference to the DNA amplification protocol                                             |\n",
    "| sop                      | Links or references to standard operating protocols used to obtain the data             |\n",
    "| pcr_primer_name_forward  | Name of the forward primer used during PCR                                              |\n",
    "| pcr_primer_name_reverse  | Name of the reverse primer used during PCR                                              |\n",
    "| pcr_primer_reference     | Reference for PCR primers                                                               |\n",
    "| seq_meth                 | The sequencing method used                                                              |\n",
    "| identificationRemarks    | Information on the taxonomic identification process                                     |\n",
    "| identificationReferences | References to procedures and/or code used during the taxonomic identification process   |\n",
    "| associatedSequences      | The identifier of the published raw DNA sequences from the water sample, if available   |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['sample_data']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:edna2obis]",
   "language": "python",
   "name": "conda-env-edna2obis-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
